{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "import seaborn as sns\n",
    "import xarray as xr\n",
    "\n",
    "# Seaborn customisation\n",
    "sns.set_theme()\n",
    "sns.set_style(\"whitegrid\")\n",
    "biomodal_palette = [\"#003B49\", \"#9CDBD9\", \"#F87C56\", \"#C0DF16\", \"#05868E\"]\n",
    "sns.set_palette(biomodal_palette)\n",
    "\n",
    "from modality.datasets import load_biomodal_dataset\n",
    "from modality.annotation import (\n",
    "    get_genes,\n",
    "    get_transcription_end_region,\n",
    "    get_tss_region,\n",
    "    get_exons,\n",
    "    get_introns,\n",
    "    get_five_prime_utrs,\n",
    "    get_three_prime_utrs,\n",
    "    get_transcripts,\n",
    "    get_cpg_islands,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load ES-E14 evoC dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from modality.contig_dataset import ContigDataset\n",
    "def unify_chr_names(ds: ContigDataset) -> ContigDataset:\n",
    "    # Extract the contig column\n",
    "    selected_column = ds[\"contig\"].values\n",
    "    \n",
    "    # Check if the contigs start with \"chr\" and remove it if they do\n",
    "    unified_contigs = np.array([contig[3:] if contig.startswith(\"chr\") else contig \n",
    "                                for contig in selected_column],dtype='U20')\n",
    "    \n",
    "    # Assign the unified contigs back to the dataset\n",
    "    ds = ds.assign_coords(contig=(\"pos\", unified_contigs))\n",
    "\n",
    "    ds.attrs[\"contigs\"] = [contig.replace(\"chr\", \"\") for contig in ds.attrs[\"contigs\"]]\n",
    "\n",
    "    \n",
    "    # Print unique values of the contig column\n",
    "    print(np.unique(ds[\"contig\"].values))\n",
    "    return ds\n",
    "\n",
    "\n",
    "def unify_slice_names(ds):\n",
    "    # Iterate through the dataset attributes and update slice names\n",
    "    for key in list(ds.attrs.keys()):\n",
    "        if key.startswith('slice_chr'):\n",
    "            new_key = key.replace('slice_chr', 'slice_')\n",
    "            ds.attrs[new_key] = ds.attrs.pop(key)\n",
    "    \n",
    "    return ds\n",
    "\n",
    "def load_data(dataset=\"\", contexts=[\"num_modc\"]):\n",
    "    #downloads modality.contig_dataset.ContigDataset object\n",
    "    if dataset != \"\":  \n",
    "        ds = ContigDataset.from_zarrz(dataset)\n",
    "        ds = unify_chr_names(ds)\n",
    "        ds = unify_slice_names(ds)\n",
    "    else:\n",
    "        ds = load_biomodal_dataset()\n",
    "        # ds = ContigDataset.from_zarrz(\"../ES-E14.zarrz\")\n",
    "        ds = ds.drop_vars([\"Input DNA Quantity (ng/sample)\", \"tech_replicate_number\"])  \n",
    "    ds = ds.sum(dim=\"sample_id\", keep_attrs=True)\n",
    "    ds = ds.expand_dims(dim=\"sample_id\", axis=1)\n",
    "    ds = ds.assign_coords(sample_id=[\"sample_0\"])\n",
    "    ds.assign_fractions(\n",
    "        numerators=contexts,\n",
    "        denominator=\"num_total_c\",\n",
    "        min_coverage=10,\n",
    "        inplace=True,\n",
    "    )\n",
    "    return ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['1' '10' '11' '12' '13' '14' '15' '16' '17' '18' '19' '2' '3' '4' '5' '6'\n",
      " '7' '8' '9' 'MT' 'X' 'Y']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-30 15:33:00 | WARNING | [modality/contig_dataset.py:340] \"sample_id\" coordinate is missing but expected to be set by many ContigDataset methods.\n",
      "It can be added using the `assign_coords` method, e.g. `ds.assign_coords(sample_id = (\"sample_id\", [\"sample1\", \"sample2\"]))`\n",
      "2024-09-30 15:33:01 | WARNING | [modality/contig_dataset.py:340] \"sample_id\" coordinate is missing but expected to be set by many ContigDataset methods.\n",
      "It can be added using the `assign_coords` method, e.g. `ds.assign_coords(sample_id = (\"sample_id\", [\"sample1\", \"sample2\"]))`\n"
     ]
    }
   ],
   "source": [
    "# ds = load_data(\"../data/CEGX_Run1485_CG.zarrz\", [\"num_modc\", \"num_mc\", \"num_hmc\"])\n",
    "# ds = load_data(\"../data/CEGX_Run1485_CG.zarrz\", [\"num_modc\"])\n",
    "## ds.to_zarrz(\"../data/CEGX_Run1485_CG_renamed.zarrz\")\n",
    "ds = load_data(\"../data/reik_e14_GRCm38_CG.zarrz\",[\"num_modc\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;\n",
       "Dimensions:       (pos: 26074280, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig        (pos) &lt;U20 &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;MT&#x27; &#x27;MT&#x27; &#x27;MT&#x27; &#x27;MT&#x27; &#x27;MT&#x27;\n",
       "    ref_position  (pos) int64 dask.array&lt;chunksize=(100000,), meta=np.ndarray&gt;\n",
       "    strand        (pos) &lt;U2 dask.array&lt;chunksize=(100000,), meta=np.ndarray&gt;\n",
       "  * sample_id     (sample_id) &lt;U8 &#x27;sample_0&#x27;\n",
       "    group         (sample_id) &lt;U8 &#x27;sample_0&#x27;\n",
       "Dimensions without coordinates: pos\n",
       "Data variables:\n",
       "    end           (pos, sample_id) uint64 dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;\n",
       "    num_c         (pos, sample_id) uint64 dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;\n",
       "    num_modc      (pos, sample_id) uint64 dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;\n",
       "    num_other     (pos, sample_id) uint64 dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;\n",
       "    num_total     (pos, sample_id) uint64 dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;\n",
       "    num_total_c   (pos, sample_id) uint64 dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;\n",
       "    percent_modc  (pos, sample_id) uint64 dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;\n",
       "    frac_modc     (pos, sample_id) float64 dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;\n",
       "Attributes:\n",
       "    context:            CG\n",
       "    context_sampling:   1.0\n",
       "    contigs:            [&#x27;1&#x27;, &#x27;2&#x27;, &#x27;3&#x27;, &#x27;4&#x27;, &#x27;5&#x27;, &#x27;6&#x27;, &#x27;7&#x27;, &#x27;8&#x27;, &#x27;9&#x27;, &#x27;10&#x27;, &#x27;...\n",
       "    coordinate_basis:   0\n",
       "    fasta_path:         GRCm38.fa.block.gz\n",
       "    input_path:         [&#x27;SRR5008113.genome.GRCm38_primary_assembly.dedup_CpG...\n",
       "    quant_type:         bedgraph\n",
       "    ref_name:           GRCm38\n",
       "    sample_ids:         [&#x27;SRR5008113&#x27;, &#x27;SRR5008115&#x27;]\n",
       "    slice_1:            slice(0, 1737408, 1)\n",
       "    slice_10:           slice(13946402, 15311424, 1)\n",
       "    slice_11:           slice(15311424, 16849294, 1)\n",
       "    slice_12:           slice(16849294, 18005730, 1)\n",
       "    slice_13:           slice(18005730, 19196682, 1)\n",
       "    slice_14:           slice(19196682, 20320810, 1)\n",
       "    slice_15:           slice(20320810, 21410730, 1)\n",
       "    slice_16:           slice(21410730, 22319618, 1)\n",
       "    slice_17:           slice(22319618, 23386164, 1)\n",
       "    slice_18:           slice(23386164, 24278174, 1)\n",
       "    slice_19:           slice(24278174, 25010342, 1)\n",
       "    slice_2:            slice(1737408, 3662080, 1)\n",
       "    slice_3:            slice(3662080, 5043804, 1)\n",
       "    slice_4:            slice(5043804, 6647958, 1)\n",
       "    slice_5:            slice(6647958, 8321048, 1)\n",
       "    slice_6:            slice(8321048, 9714434, 1)\n",
       "    slice_7:            slice(9714434, 11153720, 1)\n",
       "    slice_8:            slice(11153720, 12605040, 1)\n",
       "    slice_9:            slice(12605040, 13946402, 1)\n",
       "    slice_MT:           slice(26073706, 26074280, 1)\n",
       "    slice_X:            slice(25010342, 25760674, 1)\n",
       "    slice_Y:            slice(25760674, 26073706, 1)\n",
       "    frac_denominator:   num_total_c\n",
       "    frac_min_coverage:  10</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-8c5ecc2d-6647-4205-a772-c12acfede9ea' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-8c5ecc2d-6647-4205-a772-c12acfede9ea' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>pos</span>: 26074280</li><li><span class='xr-has-index'>sample_id</span>: 1</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-4d5e3716-5240-4278-bb8a-8e7f972b7cc5' class='xr-section-summary-in' type='checkbox'  checked><label for='section-4d5e3716-5240-4278-bb8a-8e7f972b7cc5' class='xr-section-summary' >Coordinates: <span>(5)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>contig</span></div><div class='xr-var-dims'>(pos)</div><div class='xr-var-dtype'>&lt;U20</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;MT&#x27; &#x27;MT&#x27; &#x27;MT&#x27;</div><input id='attrs-51be9695-0c3f-4309-9e28-43cc2221bd0d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-51be9695-0c3f-4309-9e28-43cc2221bd0d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1b0ac682-1e14-4b15-8d9d-98d139df0ddb' class='xr-var-data-in' type='checkbox'><label for='data-1b0ac682-1e14-4b15-8d9d-98d139df0ddb' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;, ..., &#x27;MT&#x27;, &#x27;MT&#x27;, &#x27;MT&#x27;], dtype=&#x27;&lt;U20&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>ref_position</span></div><div class='xr-var-dims'>(pos)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(100000,), meta=np.ndarray&gt;</div><input id='attrs-d28b4848-8481-468c-814a-0a4d3ab1b2d4' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d28b4848-8481-468c-814a-0a4d3ab1b2d4' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-56e3fea5-798f-44fd-bf6e-c6b09592993b' class='xr-var-data-in' type='checkbox'><label for='data-56e3fea5-798f-44fd-bf6e-c6b09592993b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 198.93 MiB </td>\n",
       "                        <td> 781.25 kiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (26074280,) </td>\n",
       "                        <td> (100000,) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 261 chunks in 2 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> int64 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"170\" height=\"75\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"120\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"25\" x2=\"120\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"5\" y1=\"0\" x2=\"5\" y2=\"25\" />\n",
       "  <line x1=\"12\" y1=\"0\" x2=\"12\" y2=\"25\" />\n",
       "  <line x1=\"18\" y1=\"0\" x2=\"18\" y2=\"25\" />\n",
       "  <line x1=\"24\" y1=\"0\" x2=\"24\" y2=\"25\" />\n",
       "  <line x1=\"31\" y1=\"0\" x2=\"31\" y2=\"25\" />\n",
       "  <line x1=\"37\" y1=\"0\" x2=\"37\" y2=\"25\" />\n",
       "  <line x1=\"44\" y1=\"0\" x2=\"44\" y2=\"25\" />\n",
       "  <line x1=\"50\" y1=\"0\" x2=\"50\" y2=\"25\" />\n",
       "  <line x1=\"56\" y1=\"0\" x2=\"56\" y2=\"25\" />\n",
       "  <line x1=\"63\" y1=\"0\" x2=\"63\" y2=\"25\" />\n",
       "  <line x1=\"69\" y1=\"0\" x2=\"69\" y2=\"25\" />\n",
       "  <line x1=\"75\" y1=\"0\" x2=\"75\" y2=\"25\" />\n",
       "  <line x1=\"81\" y1=\"0\" x2=\"81\" y2=\"25\" />\n",
       "  <line x1=\"88\" y1=\"0\" x2=\"88\" y2=\"25\" />\n",
       "  <line x1=\"94\" y1=\"0\" x2=\"94\" y2=\"25\" />\n",
       "  <line x1=\"100\" y1=\"0\" x2=\"100\" y2=\"25\" />\n",
       "  <line x1=\"107\" y1=\"0\" x2=\"107\" y2=\"25\" />\n",
       "  <line x1=\"113\" y1=\"0\" x2=\"113\" y2=\"25\" />\n",
       "  <line x1=\"120\" y1=\"0\" x2=\"120\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"60.000000\" y=\"45.412617\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >26074280</text>\n",
       "  <text x=\"140.000000\" y=\"12.706308\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(0,140.000000,12.706308)\">1</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>strand</span></div><div class='xr-var-dims'>(pos)</div><div class='xr-var-dtype'>&lt;U2</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(100000,), meta=np.ndarray&gt;</div><input id='attrs-90b3eac9-d7a0-4c55-ab4f-5b17d305f205' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-90b3eac9-d7a0-4c55-ab4f-5b17d305f205' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-660cb29d-caef-4217-a0ca-f2d3f8d5e5bc' class='xr-var-data-in' type='checkbox'><label for='data-660cb29d-caef-4217-a0ca-f2d3f8d5e5bc' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 198.93 MiB </td>\n",
       "                        <td> 781.25 kiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (26074280,) </td>\n",
       "                        <td> (100000,) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 261 chunks in 2 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> <U2 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"170\" height=\"75\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"120\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"25\" x2=\"120\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"5\" y1=\"0\" x2=\"5\" y2=\"25\" />\n",
       "  <line x1=\"12\" y1=\"0\" x2=\"12\" y2=\"25\" />\n",
       "  <line x1=\"18\" y1=\"0\" x2=\"18\" y2=\"25\" />\n",
       "  <line x1=\"24\" y1=\"0\" x2=\"24\" y2=\"25\" />\n",
       "  <line x1=\"31\" y1=\"0\" x2=\"31\" y2=\"25\" />\n",
       "  <line x1=\"37\" y1=\"0\" x2=\"37\" y2=\"25\" />\n",
       "  <line x1=\"44\" y1=\"0\" x2=\"44\" y2=\"25\" />\n",
       "  <line x1=\"50\" y1=\"0\" x2=\"50\" y2=\"25\" />\n",
       "  <line x1=\"56\" y1=\"0\" x2=\"56\" y2=\"25\" />\n",
       "  <line x1=\"63\" y1=\"0\" x2=\"63\" y2=\"25\" />\n",
       "  <line x1=\"69\" y1=\"0\" x2=\"69\" y2=\"25\" />\n",
       "  <line x1=\"75\" y1=\"0\" x2=\"75\" y2=\"25\" />\n",
       "  <line x1=\"81\" y1=\"0\" x2=\"81\" y2=\"25\" />\n",
       "  <line x1=\"88\" y1=\"0\" x2=\"88\" y2=\"25\" />\n",
       "  <line x1=\"94\" y1=\"0\" x2=\"94\" y2=\"25\" />\n",
       "  <line x1=\"100\" y1=\"0\" x2=\"100\" y2=\"25\" />\n",
       "  <line x1=\"107\" y1=\"0\" x2=\"107\" y2=\"25\" />\n",
       "  <line x1=\"113\" y1=\"0\" x2=\"113\" y2=\"25\" />\n",
       "  <line x1=\"120\" y1=\"0\" x2=\"120\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"60.000000\" y=\"45.412617\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >26074280</text>\n",
       "  <text x=\"140.000000\" y=\"12.706308\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(0,140.000000,12.706308)\">1</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample_id</span></div><div class='xr-var-dims'>(sample_id)</div><div class='xr-var-dtype'>&lt;U8</div><div class='xr-var-preview xr-preview'>&#x27;sample_0&#x27;</div><input id='attrs-ffd63156-6b48-4f02-b2ab-544eeacac77e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ffd63156-6b48-4f02-b2ab-544eeacac77e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5201f4f5-9216-4519-a4c7-3e635a1c8344' class='xr-var-data-in' type='checkbox'><label for='data-5201f4f5-9216-4519-a4c7-3e635a1c8344' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;sample_0&#x27;], dtype=&#x27;&lt;U8&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>group</span></div><div class='xr-var-dims'>(sample_id)</div><div class='xr-var-dtype'>&lt;U8</div><div class='xr-var-preview xr-preview'>&#x27;sample_0&#x27;</div><input id='attrs-518cecc7-8e3b-4ac6-8a90-836a2298f8b5' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-518cecc7-8e3b-4ac6-8a90-836a2298f8b5' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-608c4e29-40ff-46fc-896b-98c3722a0469' class='xr-var-data-in' type='checkbox'><label for='data-608c4e29-40ff-46fc-896b-98c3722a0469' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;sample_0&#x27;], dtype=&#x27;&lt;U8&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-1cf27865-2e96-4b3f-9b43-a1f828c974f9' class='xr-section-summary-in' type='checkbox'  checked><label for='section-1cf27865-2e96-4b3f-9b43-a1f828c974f9' class='xr-section-summary' >Data variables: <span>(8)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>end</span></div><div class='xr-var-dims'>(pos, sample_id)</div><div class='xr-var-dtype'>uint64</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;</div><input id='attrs-2bfa0fe3-583c-4ce6-9e87-06d8c0ac3447' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2bfa0fe3-583c-4ce6-9e87-06d8c0ac3447' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-f0a4e4aa-876a-4317-8f96-c3e9e054614b' class='xr-var-data-in' type='checkbox'><label for='data-f0a4e4aa-876a-4317-8f96-c3e9e054614b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 198.93 MiB </td>\n",
       "                        <td> 781.25 kiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (26074280, 1) </td>\n",
       "                        <td> (100000, 1) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 261 chunks in 6 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> uint64 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"75\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"25\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"5\" x2=\"25\" y2=\"5\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"25\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"18\" x2=\"25\" y2=\"18\" />\n",
       "  <line x1=\"0\" y1=\"24\" x2=\"25\" y2=\"24\" />\n",
       "  <line x1=\"0\" y1=\"31\" x2=\"25\" y2=\"31\" />\n",
       "  <line x1=\"0\" y1=\"37\" x2=\"25\" y2=\"37\" />\n",
       "  <line x1=\"0\" y1=\"44\" x2=\"25\" y2=\"44\" />\n",
       "  <line x1=\"0\" y1=\"50\" x2=\"25\" y2=\"50\" />\n",
       "  <line x1=\"0\" y1=\"56\" x2=\"25\" y2=\"56\" />\n",
       "  <line x1=\"0\" y1=\"63\" x2=\"25\" y2=\"63\" />\n",
       "  <line x1=\"0\" y1=\"69\" x2=\"25\" y2=\"69\" />\n",
       "  <line x1=\"0\" y1=\"75\" x2=\"25\" y2=\"75\" />\n",
       "  <line x1=\"0\" y1=\"81\" x2=\"25\" y2=\"81\" />\n",
       "  <line x1=\"0\" y1=\"88\" x2=\"25\" y2=\"88\" />\n",
       "  <line x1=\"0\" y1=\"94\" x2=\"25\" y2=\"94\" />\n",
       "  <line x1=\"0\" y1=\"100\" x2=\"25\" y2=\"100\" />\n",
       "  <line x1=\"0\" y1=\"107\" x2=\"25\" y2=\"107\" />\n",
       "  <line x1=\"0\" y1=\"113\" x2=\"25\" y2=\"113\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"25\" y1=\"0\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 25.412616514582485,0.0 25.412616514582485,120.0 0.0,120.0\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"12.706308\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >1</text>\n",
       "  <text x=\"45.412617\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,45.412617,60.000000)\">26074280</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_c</span></div><div class='xr-var-dims'>(pos, sample_id)</div><div class='xr-var-dtype'>uint64</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;</div><input id='attrs-f9ba6856-89f8-4db1-a840-bf8070674cc9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f9ba6856-89f8-4db1-a840-bf8070674cc9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c9e4f681-d7a1-4cbf-827f-732991e26aec' class='xr-var-data-in' type='checkbox'><label for='data-c9e4f681-d7a1-4cbf-827f-732991e26aec' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 198.93 MiB </td>\n",
       "                        <td> 781.25 kiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (26074280, 1) </td>\n",
       "                        <td> (100000, 1) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 261 chunks in 6 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> uint64 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"75\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"25\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"5\" x2=\"25\" y2=\"5\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"25\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"18\" x2=\"25\" y2=\"18\" />\n",
       "  <line x1=\"0\" y1=\"24\" x2=\"25\" y2=\"24\" />\n",
       "  <line x1=\"0\" y1=\"31\" x2=\"25\" y2=\"31\" />\n",
       "  <line x1=\"0\" y1=\"37\" x2=\"25\" y2=\"37\" />\n",
       "  <line x1=\"0\" y1=\"44\" x2=\"25\" y2=\"44\" />\n",
       "  <line x1=\"0\" y1=\"50\" x2=\"25\" y2=\"50\" />\n",
       "  <line x1=\"0\" y1=\"56\" x2=\"25\" y2=\"56\" />\n",
       "  <line x1=\"0\" y1=\"63\" x2=\"25\" y2=\"63\" />\n",
       "  <line x1=\"0\" y1=\"69\" x2=\"25\" y2=\"69\" />\n",
       "  <line x1=\"0\" y1=\"75\" x2=\"25\" y2=\"75\" />\n",
       "  <line x1=\"0\" y1=\"81\" x2=\"25\" y2=\"81\" />\n",
       "  <line x1=\"0\" y1=\"88\" x2=\"25\" y2=\"88\" />\n",
       "  <line x1=\"0\" y1=\"94\" x2=\"25\" y2=\"94\" />\n",
       "  <line x1=\"0\" y1=\"100\" x2=\"25\" y2=\"100\" />\n",
       "  <line x1=\"0\" y1=\"107\" x2=\"25\" y2=\"107\" />\n",
       "  <line x1=\"0\" y1=\"113\" x2=\"25\" y2=\"113\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"25\" y1=\"0\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 25.412616514582485,0.0 25.412616514582485,120.0 0.0,120.0\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"12.706308\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >1</text>\n",
       "  <text x=\"45.412617\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,45.412617,60.000000)\">26074280</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc</span></div><div class='xr-var-dims'>(pos, sample_id)</div><div class='xr-var-dtype'>uint64</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;</div><input id='attrs-daf66f6d-f35c-4ab6-b60e-2df6d0bf8634' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-daf66f6d-f35c-4ab6-b60e-2df6d0bf8634' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7657b2cf-40df-47da-adc6-6b5155c3d08a' class='xr-var-data-in' type='checkbox'><label for='data-7657b2cf-40df-47da-adc6-6b5155c3d08a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 198.93 MiB </td>\n",
       "                        <td> 781.25 kiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (26074280, 1) </td>\n",
       "                        <td> (100000, 1) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 261 chunks in 6 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> uint64 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"75\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"25\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"5\" x2=\"25\" y2=\"5\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"25\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"18\" x2=\"25\" y2=\"18\" />\n",
       "  <line x1=\"0\" y1=\"24\" x2=\"25\" y2=\"24\" />\n",
       "  <line x1=\"0\" y1=\"31\" x2=\"25\" y2=\"31\" />\n",
       "  <line x1=\"0\" y1=\"37\" x2=\"25\" y2=\"37\" />\n",
       "  <line x1=\"0\" y1=\"44\" x2=\"25\" y2=\"44\" />\n",
       "  <line x1=\"0\" y1=\"50\" x2=\"25\" y2=\"50\" />\n",
       "  <line x1=\"0\" y1=\"56\" x2=\"25\" y2=\"56\" />\n",
       "  <line x1=\"0\" y1=\"63\" x2=\"25\" y2=\"63\" />\n",
       "  <line x1=\"0\" y1=\"69\" x2=\"25\" y2=\"69\" />\n",
       "  <line x1=\"0\" y1=\"75\" x2=\"25\" y2=\"75\" />\n",
       "  <line x1=\"0\" y1=\"81\" x2=\"25\" y2=\"81\" />\n",
       "  <line x1=\"0\" y1=\"88\" x2=\"25\" y2=\"88\" />\n",
       "  <line x1=\"0\" y1=\"94\" x2=\"25\" y2=\"94\" />\n",
       "  <line x1=\"0\" y1=\"100\" x2=\"25\" y2=\"100\" />\n",
       "  <line x1=\"0\" y1=\"107\" x2=\"25\" y2=\"107\" />\n",
       "  <line x1=\"0\" y1=\"113\" x2=\"25\" y2=\"113\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"25\" y1=\"0\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 25.412616514582485,0.0 25.412616514582485,120.0 0.0,120.0\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"12.706308\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >1</text>\n",
       "  <text x=\"45.412617\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,45.412617,60.000000)\">26074280</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_other</span></div><div class='xr-var-dims'>(pos, sample_id)</div><div class='xr-var-dtype'>uint64</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;</div><input id='attrs-fe66b5c9-cfa5-46d1-bca5-3b97da334091' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-fe66b5c9-cfa5-46d1-bca5-3b97da334091' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-18686449-a83e-49ca-8f99-a7cee3447e86' class='xr-var-data-in' type='checkbox'><label for='data-18686449-a83e-49ca-8f99-a7cee3447e86' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 198.93 MiB </td>\n",
       "                        <td> 781.25 kiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (26074280, 1) </td>\n",
       "                        <td> (100000, 1) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 261 chunks in 6 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> uint64 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"75\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"25\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"5\" x2=\"25\" y2=\"5\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"25\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"18\" x2=\"25\" y2=\"18\" />\n",
       "  <line x1=\"0\" y1=\"24\" x2=\"25\" y2=\"24\" />\n",
       "  <line x1=\"0\" y1=\"31\" x2=\"25\" y2=\"31\" />\n",
       "  <line x1=\"0\" y1=\"37\" x2=\"25\" y2=\"37\" />\n",
       "  <line x1=\"0\" y1=\"44\" x2=\"25\" y2=\"44\" />\n",
       "  <line x1=\"0\" y1=\"50\" x2=\"25\" y2=\"50\" />\n",
       "  <line x1=\"0\" y1=\"56\" x2=\"25\" y2=\"56\" />\n",
       "  <line x1=\"0\" y1=\"63\" x2=\"25\" y2=\"63\" />\n",
       "  <line x1=\"0\" y1=\"69\" x2=\"25\" y2=\"69\" />\n",
       "  <line x1=\"0\" y1=\"75\" x2=\"25\" y2=\"75\" />\n",
       "  <line x1=\"0\" y1=\"81\" x2=\"25\" y2=\"81\" />\n",
       "  <line x1=\"0\" y1=\"88\" x2=\"25\" y2=\"88\" />\n",
       "  <line x1=\"0\" y1=\"94\" x2=\"25\" y2=\"94\" />\n",
       "  <line x1=\"0\" y1=\"100\" x2=\"25\" y2=\"100\" />\n",
       "  <line x1=\"0\" y1=\"107\" x2=\"25\" y2=\"107\" />\n",
       "  <line x1=\"0\" y1=\"113\" x2=\"25\" y2=\"113\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"25\" y1=\"0\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 25.412616514582485,0.0 25.412616514582485,120.0 0.0,120.0\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"12.706308\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >1</text>\n",
       "  <text x=\"45.412617\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,45.412617,60.000000)\">26074280</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total</span></div><div class='xr-var-dims'>(pos, sample_id)</div><div class='xr-var-dtype'>uint64</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;</div><input id='attrs-20c866c9-7f02-410d-93e1-aea6502a8463' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-20c866c9-7f02-410d-93e1-aea6502a8463' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0356b7c0-5ada-429b-a50e-954ea1ec875a' class='xr-var-data-in' type='checkbox'><label for='data-0356b7c0-5ada-429b-a50e-954ea1ec875a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 198.93 MiB </td>\n",
       "                        <td> 781.25 kiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (26074280, 1) </td>\n",
       "                        <td> (100000, 1) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 261 chunks in 6 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> uint64 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"75\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"25\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"5\" x2=\"25\" y2=\"5\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"25\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"18\" x2=\"25\" y2=\"18\" />\n",
       "  <line x1=\"0\" y1=\"24\" x2=\"25\" y2=\"24\" />\n",
       "  <line x1=\"0\" y1=\"31\" x2=\"25\" y2=\"31\" />\n",
       "  <line x1=\"0\" y1=\"37\" x2=\"25\" y2=\"37\" />\n",
       "  <line x1=\"0\" y1=\"44\" x2=\"25\" y2=\"44\" />\n",
       "  <line x1=\"0\" y1=\"50\" x2=\"25\" y2=\"50\" />\n",
       "  <line x1=\"0\" y1=\"56\" x2=\"25\" y2=\"56\" />\n",
       "  <line x1=\"0\" y1=\"63\" x2=\"25\" y2=\"63\" />\n",
       "  <line x1=\"0\" y1=\"69\" x2=\"25\" y2=\"69\" />\n",
       "  <line x1=\"0\" y1=\"75\" x2=\"25\" y2=\"75\" />\n",
       "  <line x1=\"0\" y1=\"81\" x2=\"25\" y2=\"81\" />\n",
       "  <line x1=\"0\" y1=\"88\" x2=\"25\" y2=\"88\" />\n",
       "  <line x1=\"0\" y1=\"94\" x2=\"25\" y2=\"94\" />\n",
       "  <line x1=\"0\" y1=\"100\" x2=\"25\" y2=\"100\" />\n",
       "  <line x1=\"0\" y1=\"107\" x2=\"25\" y2=\"107\" />\n",
       "  <line x1=\"0\" y1=\"113\" x2=\"25\" y2=\"113\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"25\" y1=\"0\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 25.412616514582485,0.0 25.412616514582485,120.0 0.0,120.0\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"12.706308\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >1</text>\n",
       "  <text x=\"45.412617\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,45.412617,60.000000)\">26074280</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c</span></div><div class='xr-var-dims'>(pos, sample_id)</div><div class='xr-var-dtype'>uint64</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;</div><input id='attrs-3425ae6f-b6be-45af-ba44-2c8774ee748d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3425ae6f-b6be-45af-ba44-2c8774ee748d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-337bc5c1-5672-4f4a-8b4e-6800f0fa2531' class='xr-var-data-in' type='checkbox'><label for='data-337bc5c1-5672-4f4a-8b4e-6800f0fa2531' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 198.93 MiB </td>\n",
       "                        <td> 781.25 kiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (26074280, 1) </td>\n",
       "                        <td> (100000, 1) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 261 chunks in 6 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> uint64 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"75\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"25\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"5\" x2=\"25\" y2=\"5\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"25\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"18\" x2=\"25\" y2=\"18\" />\n",
       "  <line x1=\"0\" y1=\"24\" x2=\"25\" y2=\"24\" />\n",
       "  <line x1=\"0\" y1=\"31\" x2=\"25\" y2=\"31\" />\n",
       "  <line x1=\"0\" y1=\"37\" x2=\"25\" y2=\"37\" />\n",
       "  <line x1=\"0\" y1=\"44\" x2=\"25\" y2=\"44\" />\n",
       "  <line x1=\"0\" y1=\"50\" x2=\"25\" y2=\"50\" />\n",
       "  <line x1=\"0\" y1=\"56\" x2=\"25\" y2=\"56\" />\n",
       "  <line x1=\"0\" y1=\"63\" x2=\"25\" y2=\"63\" />\n",
       "  <line x1=\"0\" y1=\"69\" x2=\"25\" y2=\"69\" />\n",
       "  <line x1=\"0\" y1=\"75\" x2=\"25\" y2=\"75\" />\n",
       "  <line x1=\"0\" y1=\"81\" x2=\"25\" y2=\"81\" />\n",
       "  <line x1=\"0\" y1=\"88\" x2=\"25\" y2=\"88\" />\n",
       "  <line x1=\"0\" y1=\"94\" x2=\"25\" y2=\"94\" />\n",
       "  <line x1=\"0\" y1=\"100\" x2=\"25\" y2=\"100\" />\n",
       "  <line x1=\"0\" y1=\"107\" x2=\"25\" y2=\"107\" />\n",
       "  <line x1=\"0\" y1=\"113\" x2=\"25\" y2=\"113\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"25\" y1=\"0\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 25.412616514582485,0.0 25.412616514582485,120.0 0.0,120.0\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"12.706308\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >1</text>\n",
       "  <text x=\"45.412617\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,45.412617,60.000000)\">26074280</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>percent_modc</span></div><div class='xr-var-dims'>(pos, sample_id)</div><div class='xr-var-dtype'>uint64</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;</div><input id='attrs-51893c2d-4be7-4c52-ae79-b64a3531104e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-51893c2d-4be7-4c52-ae79-b64a3531104e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b83b5213-9bf9-48ec-8a78-e27733a7cd64' class='xr-var-data-in' type='checkbox'><label for='data-b83b5213-9bf9-48ec-8a78-e27733a7cd64' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 198.93 MiB </td>\n",
       "                        <td> 781.25 kiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (26074280, 1) </td>\n",
       "                        <td> (100000, 1) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 261 chunks in 6 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> uint64 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"75\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"25\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"5\" x2=\"25\" y2=\"5\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"25\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"18\" x2=\"25\" y2=\"18\" />\n",
       "  <line x1=\"0\" y1=\"24\" x2=\"25\" y2=\"24\" />\n",
       "  <line x1=\"0\" y1=\"31\" x2=\"25\" y2=\"31\" />\n",
       "  <line x1=\"0\" y1=\"37\" x2=\"25\" y2=\"37\" />\n",
       "  <line x1=\"0\" y1=\"44\" x2=\"25\" y2=\"44\" />\n",
       "  <line x1=\"0\" y1=\"50\" x2=\"25\" y2=\"50\" />\n",
       "  <line x1=\"0\" y1=\"56\" x2=\"25\" y2=\"56\" />\n",
       "  <line x1=\"0\" y1=\"63\" x2=\"25\" y2=\"63\" />\n",
       "  <line x1=\"0\" y1=\"69\" x2=\"25\" y2=\"69\" />\n",
       "  <line x1=\"0\" y1=\"75\" x2=\"25\" y2=\"75\" />\n",
       "  <line x1=\"0\" y1=\"81\" x2=\"25\" y2=\"81\" />\n",
       "  <line x1=\"0\" y1=\"88\" x2=\"25\" y2=\"88\" />\n",
       "  <line x1=\"0\" y1=\"94\" x2=\"25\" y2=\"94\" />\n",
       "  <line x1=\"0\" y1=\"100\" x2=\"25\" y2=\"100\" />\n",
       "  <line x1=\"0\" y1=\"107\" x2=\"25\" y2=\"107\" />\n",
       "  <line x1=\"0\" y1=\"113\" x2=\"25\" y2=\"113\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"25\" y1=\"0\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 25.412616514582485,0.0 25.412616514582485,120.0 0.0,120.0\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"12.706308\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >1</text>\n",
       "  <text x=\"45.412617\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,45.412617,60.000000)\">26074280</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>frac_modc</span></div><div class='xr-var-dims'>(pos, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(100000, 1), meta=np.ndarray&gt;</div><input id='attrs-3cf74239-b4bc-4954-9ecd-b5751058cc43' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3cf74239-b4bc-4954-9ecd-b5751058cc43' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d92d01d9-cf0c-47b5-9c6e-a72d5768c806' class='xr-var-data-in' type='checkbox'><label for='data-d92d01d9-cf0c-47b5-9c6e-a72d5768c806' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 198.93 MiB </td>\n",
       "                        <td> 781.25 kiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (26074280, 1) </td>\n",
       "                        <td> (100000, 1) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 261 chunks in 16 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> float64 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"75\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"25\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"5\" x2=\"25\" y2=\"5\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"25\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"18\" x2=\"25\" y2=\"18\" />\n",
       "  <line x1=\"0\" y1=\"24\" x2=\"25\" y2=\"24\" />\n",
       "  <line x1=\"0\" y1=\"31\" x2=\"25\" y2=\"31\" />\n",
       "  <line x1=\"0\" y1=\"37\" x2=\"25\" y2=\"37\" />\n",
       "  <line x1=\"0\" y1=\"44\" x2=\"25\" y2=\"44\" />\n",
       "  <line x1=\"0\" y1=\"50\" x2=\"25\" y2=\"50\" />\n",
       "  <line x1=\"0\" y1=\"56\" x2=\"25\" y2=\"56\" />\n",
       "  <line x1=\"0\" y1=\"63\" x2=\"25\" y2=\"63\" />\n",
       "  <line x1=\"0\" y1=\"69\" x2=\"25\" y2=\"69\" />\n",
       "  <line x1=\"0\" y1=\"75\" x2=\"25\" y2=\"75\" />\n",
       "  <line x1=\"0\" y1=\"81\" x2=\"25\" y2=\"81\" />\n",
       "  <line x1=\"0\" y1=\"88\" x2=\"25\" y2=\"88\" />\n",
       "  <line x1=\"0\" y1=\"94\" x2=\"25\" y2=\"94\" />\n",
       "  <line x1=\"0\" y1=\"100\" x2=\"25\" y2=\"100\" />\n",
       "  <line x1=\"0\" y1=\"107\" x2=\"25\" y2=\"107\" />\n",
       "  <line x1=\"0\" y1=\"113\" x2=\"25\" y2=\"113\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"25\" y1=\"0\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 25.412616514582485,0.0 25.412616514582485,120.0 0.0,120.0\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"12.706308\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >1</text>\n",
       "  <text x=\"45.412617\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,45.412617,60.000000)\">26074280</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li></ul></div></li><li class='xr-section-item'><input id='section-a5d06f82-b382-46cc-b4e5-3ce863e99fe5' class='xr-section-summary-in' type='checkbox'  ><label for='section-a5d06f82-b382-46cc-b4e5-3ce863e99fe5' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>sample_id</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-da928866-8253-4d8a-9f34-1c8a1551c029' class='xr-index-data-in' type='checkbox'/><label for='index-da928866-8253-4d8a-9f34-1c8a1551c029' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;sample_0&#x27;], dtype=&#x27;object&#x27;, name=&#x27;sample_id&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-376ee634-0075-4f63-a08d-fbf1e6f39d2f' class='xr-section-summary-in' type='checkbox'  ><label for='section-376ee634-0075-4f63-a08d-fbf1e6f39d2f' class='xr-section-summary' >Attributes: <span>(33)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'><dt><span>context :</span></dt><dd>CG</dd><dt><span>context_sampling :</span></dt><dd>1.0</dd><dt><span>contigs :</span></dt><dd>[&#x27;1&#x27;, &#x27;2&#x27;, &#x27;3&#x27;, &#x27;4&#x27;, &#x27;5&#x27;, &#x27;6&#x27;, &#x27;7&#x27;, &#x27;8&#x27;, &#x27;9&#x27;, &#x27;10&#x27;, &#x27;11&#x27;, &#x27;12&#x27;, &#x27;13&#x27;, &#x27;14&#x27;, &#x27;15&#x27;, &#x27;16&#x27;, &#x27;17&#x27;, &#x27;18&#x27;, &#x27;19&#x27;, &#x27;X&#x27;, &#x27;Y&#x27;, &#x27;MT&#x27;]</dd><dt><span>coordinate_basis :</span></dt><dd>0</dd><dt><span>fasta_path :</span></dt><dd>GRCm38.fa.block.gz</dd><dt><span>input_path :</span></dt><dd>[&#x27;SRR5008113.genome.GRCm38_primary_assembly.dedup_CpG.bedGraph.gz&#x27;, &#x27;SRR5008115.genome.GRCm38_primary_assembly.dedup_CpG.bedGraph.gz&#x27;]</dd><dt><span>quant_type :</span></dt><dd>bedgraph</dd><dt><span>ref_name :</span></dt><dd>GRCm38</dd><dt><span>sample_ids :</span></dt><dd>[&#x27;SRR5008113&#x27;, &#x27;SRR5008115&#x27;]</dd><dt><span>slice_1 :</span></dt><dd>slice(0, 1737408, 1)</dd><dt><span>slice_10 :</span></dt><dd>slice(13946402, 15311424, 1)</dd><dt><span>slice_11 :</span></dt><dd>slice(15311424, 16849294, 1)</dd><dt><span>slice_12 :</span></dt><dd>slice(16849294, 18005730, 1)</dd><dt><span>slice_13 :</span></dt><dd>slice(18005730, 19196682, 1)</dd><dt><span>slice_14 :</span></dt><dd>slice(19196682, 20320810, 1)</dd><dt><span>slice_15 :</span></dt><dd>slice(20320810, 21410730, 1)</dd><dt><span>slice_16 :</span></dt><dd>slice(21410730, 22319618, 1)</dd><dt><span>slice_17 :</span></dt><dd>slice(22319618, 23386164, 1)</dd><dt><span>slice_18 :</span></dt><dd>slice(23386164, 24278174, 1)</dd><dt><span>slice_19 :</span></dt><dd>slice(24278174, 25010342, 1)</dd><dt><span>slice_2 :</span></dt><dd>slice(1737408, 3662080, 1)</dd><dt><span>slice_3 :</span></dt><dd>slice(3662080, 5043804, 1)</dd><dt><span>slice_4 :</span></dt><dd>slice(5043804, 6647958, 1)</dd><dt><span>slice_5 :</span></dt><dd>slice(6647958, 8321048, 1)</dd><dt><span>slice_6 :</span></dt><dd>slice(8321048, 9714434, 1)</dd><dt><span>slice_7 :</span></dt><dd>slice(9714434, 11153720, 1)</dd><dt><span>slice_8 :</span></dt><dd>slice(11153720, 12605040, 1)</dd><dt><span>slice_9 :</span></dt><dd>slice(12605040, 13946402, 1)</dd><dt><span>slice_MT :</span></dt><dd>slice(26073706, 26074280, 1)</dd><dt><span>slice_X :</span></dt><dd>slice(25010342, 25760674, 1)</dd><dt><span>slice_Y :</span></dt><dd>slice(25760674, 26073706, 1)</dd><dt><span>frac_denominator :</span></dt><dd>num_total_c</dd><dt><span>frac_min_coverage :</span></dt><dd>10</dd></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "ContigDataset(data_vars=Data variables:\n",
       "    end           (pos, sample_id) uint64 dask.array<chunksize=(100000, 1), meta=np.ndarray>\n",
       "    num_c         (pos, sample_id) uint64 dask.array<chunksize=(100000, 1), meta=np.ndarray>\n",
       "    num_modc      (pos, sample_id) uint64 dask.array<chunksize=(100000, 1), meta=np.ndarray>\n",
       "    num_other     (pos, sample_id) uint64 dask.array<chunksize=(100000, 1), meta=np.ndarray>\n",
       "    num_total     (pos, sample_id) uint64 dask.array<chunksize=(100000, 1), meta=np.ndarray>\n",
       "    num_total_c   (pos, sample_id) uint64 dask.array<chunksize=(100000, 1), meta=np.ndarray>\n",
       "    percent_modc  (pos, sample_id) uint64 dask.array<chunksize=(100000, 1), meta=np.ndarray>\n",
       "    frac_modc     (pos, sample_id) float64 dask.array<chunksize=(100000, 1), meta=np.ndarray>, coords=Coordinates:\n",
       "    contig        (pos) <U20 '1' '1' '1' '1' '1' ... 'MT' 'MT' 'MT' 'MT' 'MT'\n",
       "    ref_position  (pos) int64 dask.array<chunksize=(100000,), meta=np.ndarray>\n",
       "    strand        (pos) <U2 dask.array<chunksize=(100000,), meta=np.ndarray>\n",
       "  * sample_id     (sample_id) <U8 'sample_0'\n",
       "    group         (sample_id) <U8 'sample_0', attrs={'context': 'CG', 'context_sampling': 1.0, 'contigs': ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', 'X', 'Y', 'MT'], 'coordinate_basis': 0, 'fasta_path': 'GRCm38.fa.block.gz', 'input_path': ['SRR5008113.genome.GRCm38_primary_assembly.dedup_CpG.bedGraph.gz', 'SRR5008115.genome.GRCm38_primary_assembly.dedup_CpG.bedGraph.gz'], 'quant_type': 'bedgraph', 'ref_name': 'GRCm38', 'sample_ids': ['SRR5008113', 'SRR5008115'], 'slice_1': slice(0, 1737408, 1), 'slice_10': slice(13946402, 15311424, 1), 'slice_11': slice(15311424, 16849294, 1), 'slice_12': slice(16849294, 18005730, 1), 'slice_13': slice(18005730, 19196682, 1), 'slice_14': slice(19196682, 20320810, 1), 'slice_15': slice(20320810, 21410730, 1), 'slice_16': slice(21410730, 22319618, 1), 'slice_17': slice(22319618, 23386164, 1), 'slice_18': slice(23386164, 24278174, 1), 'slice_19': slice(24278174, 25010342, 1), 'slice_2': slice(1737408, 3662080, 1), 'slice_3': slice(3662080, 5043804, 1), 'slice_4': slice(5043804, 6647958, 1), 'slice_5': slice(6647958, 8321048, 1), 'slice_6': slice(8321048, 9714434, 1), 'slice_7': slice(9714434, 11153720, 1), 'slice_8': slice(11153720, 12605040, 1), 'slice_9': slice(12605040, 13946402, 1), 'slice_MT': slice(26073706, 26074280, 1), 'slice_X': slice(25010342, 25760674, 1), 'slice_Y': slice(25760674, 26073706, 1), 'frac_denominator': 'num_total_c', 'frac_min_coverage': 10})"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### subset `ContigDataset` by strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# modified set_contig_slices()\n",
    "import dask.array as da\n",
    "def set_contig_slices(xds: xr.Dataset) -> xr.Dataset:\n",
    "    \"\"\"\n",
    "    Given a dataset with a contig data array, compute the slices for each contig and\n",
    "    store them in the dataset attributes.\n",
    "\n",
    "    Args:\n",
    "        xds: xarray dataset to be processed\n",
    "\n",
    "    Returns:\n",
    "        input dataset with the contig slices set in the attributes\n",
    "    \"\"\"\n",
    "    print(\"Computing available chromosomes.\")\n",
    "\n",
    "    # Ensure contig is a Dask array\n",
    "    contig_data = da.from_array(xds.contig.data) if not isinstance(xds.contig.data, da.Array) else xds.contig.data\n",
    "\n",
    "    # Compute unique labels and the first index\n",
    "    contig_labels, contig_first_index = da.compute(*da.unique(contig_data, return_index=True))\n",
    "\n",
    "\n",
    "    # contig_labels, contig_first_index = da.compute(\n",
    "    #     *da.unique(xds.contig.data, return_index=True)\n",
    "    # )\n",
    "\n",
    "    # TODO sort by reference order.\n",
    "    contig_labels = contig_labels[np.argsort(contig_first_index)]\n",
    "    contig_first_index = contig_first_index[np.argsort(contig_first_index)]\n",
    "\n",
    "    # TODO mouse\n",
    "    \"\"\"\n",
    "    assert all(\n",
    "        contig_labels == valid_contigs\n",
    "    ), f\"{contig_labels} do not match expectation ({valid_contigs})\"\n",
    "    \"\"\"\n",
    "\n",
    "    last_index = len(xds.contig.data)\n",
    "\n",
    "    print(\n",
    "        f'Found available contigs: {\",\".join([x.replace(\"chr\",\"\") for x in contig_labels])}.'\n",
    "    )\n",
    "\n",
    "    xds.attrs[\"contigs\"] = [str(xx) for xx in contig_labels]\n",
    "\n",
    "    # for var in [\"contig\", \"icontig\"]:\n",
    "    #    if var in xds.variables:\n",
    "    #        xds = xds.drop([var])\n",
    "\n",
    "    if len(contig_labels) == 1:\n",
    "        xds.attrs[f\"slice_{contig_labels[0]}\"] = (0, last_index, 1)\n",
    "        return xds\n",
    "\n",
    "    # work out slices if multiple contigs\n",
    "    contig_first_index = dict(zip(contig_labels, contig_first_index))\n",
    "\n",
    "    for icontig, (contig, contig_next) in enumerate(\n",
    "        zip(contig_labels[:-1], contig_labels[1:])\n",
    "    ):\n",
    "        xds.attrs[f\"slice_{contig}\"] = (\n",
    "            contig_first_index[contig],\n",
    "            contig_first_index[contig_next],\n",
    "            1,\n",
    "        )\n",
    "    else:\n",
    "        xds.attrs[f\"slice_{contig_next}\"] = (\n",
    "            contig_first_index[contig_next],\n",
    "            last_index,\n",
    "            1,\n",
    "        )\n",
    "\n",
    "    return xds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from modality.contig_dataset import cast_result\n",
    "\n",
    "def refresh_slices(subset_data):\n",
    "    if subset_data.pos.size == 0:\n",
    "            raise ValueError(\n",
    "                \"Subset is empty. Please check the coverage values and method used.\"\n",
    "            )\n",
    "    else:\n",
    "        slices = [x for x in subset_data.attrs if x.startswith(\"slice\")]\n",
    "        for sl in slices:\n",
    "            subset_data.attrs.pop(sl)\n",
    "        return set_contig_slices(subset_data)\n",
    "\n",
    "def split_strand(data, strand):\n",
    "    \"\"\"\n",
    "    split xarray dataset into positive and negative strands\n",
    "    Parameters\n",
    "        data : xarray.Dataset\n",
    "        strand : str ('+', '-')\n",
    "    Returns\n",
    "        strand_data :xarray.Dataset\n",
    "    \"\"\"\n",
    "    strand_mask = (data['strand'].data == strand).compute()\n",
    "    masked_da = xr.DataArray(strand_mask, dims=data['strand'].dims, coords=data['strand'].coords)\n",
    "    strand_data = data.where(masked_da, drop=True)\n",
    "\n",
    "    # return strand_data\n",
    "    return cast_result(refresh_slices(strand_data.ds))\n",
    "\n",
    "def remove_low_coverage(data: xr, threshold: int) -> xr.Dataset:\n",
    "    \"\"\"\n",
    "    Removes data points with num_total_c less than the given threshold.\n",
    "\n",
    "    Parameters:\n",
    "    - data (xr.Dataset): The input dataset containing the variable `num_total_c`.\n",
    "    - threshold (int): The threshold value. Rows with `num_total_c` less than this will be removed.\n",
    "\n",
    "    Returns:\n",
    "    - xr.Dataset: The dataset with rows filtered by the threshold.\n",
    "    \"\"\"\n",
    "    # Ensure that num_total_c exists and has valid dimensions\n",
    "    if 'num_total_c' not in data.data_vars:\n",
    "        raise ValueError(\"The dataset does not contain the 'num_total_c' variable.\")\n",
    "    \n",
    "    coverage_mask = (data['num_total_c'].data >= threshold).compute()\n",
    "    masked_da = xr.DataArray(coverage_mask, dims=data['num_total_c'].dims, coords=data['num_total_c'].coords)\n",
    "    # Apply the mask to the dataset and drop positions that don't meet the threshold\n",
    "    filtered_data = data.where(masked_da, drop=True)\n",
    "    \n",
    "    return cast_result(refresh_slices(filtered_data.ds))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ds.attrs[\"contigs\"] = [contig.replace(\"chr\", \"\") for contig in strand_data.attrs[\"contigs\"]]\n",
    "# strand_mask = (ds['strand'].data == '+').compute()\n",
    "# masked_da = xr.DataArray(strand_mask, dims=ds['strand'].dims, coords=ds['strand'].coords)\n",
    "# strand_data = ds.where(masked_da, drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing available chromosomes.\n",
      "Found available contigs: 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,X,Y,MT.\n",
      "Computing available chromosomes.\n",
      "Found available contigs: 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,X,Y,MT.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-30 17:52:00 | INFO | [modality/contig_dataset.py:860] Writing to zipped zarr store (ZipStore) ../data/plus_strand_wgbs.zarrz\n",
      "2024-09-30 17:52:12 | INFO | [modality/contig_dataset.py:860] Writing to zipped zarr store (ZipStore) ../data/minus_strand_wgbs.zarrz\n"
     ]
    }
   ],
   "source": [
    "plus_strand = split_strand(ds, \"+\")\n",
    "minus_strand = split_strand( ds, \"-\")\n",
    "\n",
    "plus_strand.to_zarrz(\"../data/plus_strand_wgbs.zarrz\")\n",
    "minus_strand.to_zarrz(\"../data/minus_strand_wgbs.zarrz\")\n",
    "\n",
    "# get original dataset for comparison\n",
    "# ds_orig = load_data()\n",
    "# plus_strand_orig = split_strand(ds_orig, \"+\")\n",
    "# minus_strand_orig = split_strand(ds_orig, \"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plus_strand = ContigDataset.from_zarrz(\"../data/plus_strand.zarrz\")\n",
    "# minus_strand= ContigDataset.from_zarrz(\"../data/minus_strand.zarrz\")\n",
    "plus_strand = ContigDataset.from_zarrz(\"../data/plus_strand.zarrz\")\n",
    "minus_strand= ContigDataset.from_zarrz(\"../data/minus_strand.zarrz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### removing low coverage positions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Compute the dask array into a numpy array\n",
    "# num_total_c_computed = ds['num_total_c'].compute()\n",
    "\n",
    "# # Plot the distribution using seaborn or matplotlib\n",
    "# sns.histplot(num_total_c_computed, bins=50, kde=True)\n",
    "# plt.title(\"Distribution of num_total_c, full dataset\")\n",
    "# plt.xlabel(\"num_total_c values\")\n",
    "# plt.ylabel(\"Frequency\")\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "working with strand aware subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing available chromosomes.\n",
      "Found available contigs: 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,X,Y,M.\n",
      "Proportion of instances where num_total_c == 0 in plus_strand: 0.11373456826114078\n"
     ]
    }
   ],
   "source": [
    "plus_all = plus_strand[\"num_total_c\"].size\n",
    "plus_strand_cov = remove_low_coverage(plus_strand, 10)\n",
    "plus_filtered = plus_strand_cov[\"num_total_c\"].size\n",
    "\n",
    "print(f\"Proportion of instances where num_total_c == 0 in plus_strand: {(plus_all - plus_filtered)/plus_all}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing available chromosomes.\n",
      "Found available contigs: 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,X,Y,M.\n",
      "Proportion of instances where num_total_c == 0 in minus_strand: 0.11376635009671968\n"
     ]
    }
   ],
   "source": [
    "minus_all = minus_strand[\"num_total_c\"].size\n",
    "minus_strand_cov = remove_low_coverage(minus_strand, 10)\n",
    "minus_filtered = minus_strand_cov[\"num_total_c\"].size\n",
    "\n",
    "print(f\"Proportion of instances where num_total_c == 0 in minus_strand: {(minus_all - minus_filtered)/minus_all}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Compute the dask array into a numpy array\n",
    "# num_total_c_computed = plus_strand_cov['num_total_c'].compute()\n",
    "\n",
    "# # Plot the distribution using seaborn or matplotlib\n",
    "# sns.histplot(num_total_c_computed, bins=50, kde=True)\n",
    "# plt.title(\"Distribution of num_total_c, cleaned Plus\")\n",
    "# plt.xlabel(\"num_total_c values\")\n",
    "# plt.ylabel(\"Frequency\")\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "double-checking the counts of remaining positions per contig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "slice_1\n",
      "(1341826,)\n",
      "slice_2\n",
      "(1414344,)\n",
      "slice_3\n",
      "(1086495,)\n",
      "slice_4\n",
      "(1174582,)\n",
      "slice_5\n",
      "(1238054,)\n",
      "slice_6\n",
      "(1069125,)\n",
      "slice_7\n",
      "(1055303,)\n",
      "slice_8\n",
      "(1029525,)\n",
      "slice_9\n",
      "(999321,)\n",
      "slice_10\n",
      "(1029648,)\n",
      "slice_11\n",
      "(1098901,)\n",
      "slice_12\n",
      "(850345,)\n",
      "slice_13\n",
      "(898712,)\n",
      "slice_14\n",
      "(812156,)\n",
      "slice_15\n",
      "(791927,)\n",
      "slice_16\n",
      "(694729,)\n",
      "slice_17\n",
      "(768263,)\n",
      "slice_18\n",
      "(658766,)\n",
      "slice_19\n",
      "(514211,)\n",
      "slice_X\n",
      "(784165,)\n",
      "slice_Y\n",
      "(70093,)\n",
      "slice_M\n",
      "(217,)\n"
     ]
    }
   ],
   "source": [
    "for contig in [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,'X','Y','M']:\n",
    "    slice1 = plus_strand_cov.slices[f\"slice_{contig}\"]\n",
    "    print(f\"slice_{contig}\")\n",
    "    print(plus_strand_cov.ref_position[slice1].data.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Loading mm10 (GRCm38) reference genome from [GENCODE](https://www.gencodegenes.org/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-01 14:43:38 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------+------------+------------+-----------+-----------+-------+\n",
      "|   Chromosome | Source     | Type       |     Start |       End | +12   |\n",
      "|   (category) | (object)   | (object)   |   (int64) |   (int64) | ...   |\n",
      "|--------------+------------+------------+-----------+-----------+-------|\n",
      "|            1 | HAVANA     | gene       |   4807787 |   4848409 | ...   |\n",
      "|            1 | HAVANA     | gene       |   4807891 |   4886769 | ...   |\n",
      "|            1 | HAVANA     | gene       |   4857813 |   4897908 | ...   |\n",
      "|            1 | HAVANA     | gene       |   5070017 |   5162528 | ...   |\n",
      "|            1 | HAVANA     | gene       |   5588465 |   5606130 | ...   |\n",
      "+--------------+------------+------------+-----------+-----------+-------+\n",
      "Stranded PyRanges object has 5 rows and 17 columns from 1 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
      "12 hidden columns: Score, Strand, Phase, Id, Gene_id, Gene_type, Gene_name, Level, ... (+ 4 more.)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-01 14:43:51 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------------------------------------------------\n",
      " all genes: 21673\n",
      " unique gene id: 21673\n"
     ]
    }
   ],
   "source": [
    "# Get the gene annotations\n",
    "gene_filter = {\n",
    "    \"gene_type\": \"protein_coding\",\n",
    "    \"source\": \"HAVANA\",\n",
    "}\n",
    "\n",
    "genes = get_genes(\n",
    "    reference=\"mm10\",\n",
    "    as_pyranges=True,\n",
    "    filterby=gene_filter,\n",
    ")\n",
    "print(genes.head(5))\n",
    "\n",
    "# Get transcripts for mm10\n",
    "transcripts = get_transcripts(\n",
    "        reference=\"mm10\",\n",
    "        contig=None,\n",
    "        start=None,\n",
    "        end=None,\n",
    "        as_pyranges=False,\n",
    "    )\n",
    "\n",
    "print('-'*100)\n",
    "\n",
    "print(f\" all genes: {len(genes)}\")\n",
    "print(f\" unique gene id: {len(genes.df['Id'])}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Select primiary transcript for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003.15</td>\n",
       "      <td>ENSMUST00000000003.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028.15</td>\n",
       "      <td>ENSMUST00000000028.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000037.17</td>\n",
       "      <td>ENSMUST00000101113.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000049.11</td>\n",
       "      <td>ENSMUST00000000049.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene_id          transcript_id\n",
       "0   ENSMUSG00000000001.4   ENSMUST00000000001.4\n",
       "1  ENSMUSG00000000003.15  ENSMUST00000000003.13\n",
       "2  ENSMUSG00000000028.15  ENSMUST00000000028.13\n",
       "3  ENSMUSG00000000037.17   ENSMUST00000101113.8\n",
       "4  ENSMUSG00000000049.11   ENSMUST00000000049.5"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def select_transcript_based_on_tag(df):\n",
    "    # for each transcript in df, select the one with the highest priority tag\n",
    "    # priorities are:\n",
    "        # 1. 'basic,appris_principal_1,CCDS'\n",
    "        # 2. 'basic,appris_principal_1'\n",
    "        # 3. 'basic,CCDS'\n",
    "        # 4. 'basic'\n",
    "    # but with 'exp_conf' (experimentally confirmed) tag, the priority is higher.\n",
    "    \n",
    "    priorties = {\n",
    "        'basic,appris_principal_1,exp_conf,CCDS': 1,\n",
    "        'basic,appris_principal_1,CCDS': 1,\n",
    "        'basic,appris_principal_1,exp_conf': 3,\n",
    "        'basic,appris_principal_1': 4,\n",
    "        'basic,exp_conf,CCDS': 5,\n",
    "        'basic,CCDS': 6,\n",
    "        'basic,exp_conf': 7,\n",
    "        'basic': 8\n",
    "    }\n",
    "\n",
    "    # sort the dataframe by the priority of the tags\n",
    "    df['tag_priority'] = df.tag.map(priorties)\n",
    "\n",
    "    df = df.sort_values(by='tag_priority')\n",
    "\n",
    "    # drop duplicates, keeping the first one\n",
    "    df = df.drop_duplicates(subset='gene_id', keep='first')\n",
    "\n",
    "    return df[[\"gene_id\", \"transcript_id\"]]\n",
    "\n",
    "selected_transcripts = transcripts.groupby('gene_id').apply(\n",
    "    select_transcript_based_on_tag\n",
    "    ).reset_index(drop=True)\n",
    "selected_transcripts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 21541 genes with unique transcript ids\n"
     ]
    }
   ],
   "source": [
    "# Create dictionary for gene_id to transcript_id mapping\n",
    "gene_to_transcription = selected_transcripts.set_index('gene_id')['transcript_id'].to_dict()\n",
    "print(f\"There are {len(gene_to_transcription)} genes with unique transcript ids\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "review / drop genes with no transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all genes: 21673\n",
      "all genes with annotated transcripts: 21541\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807787</td>\n",
       "      <td>4848409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807891</td>\n",
       "      <td>4886769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5611216</td>\n",
       "      <td>OTTMUSG00000050100.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUST00000155020.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4857813</td>\n",
       "      <td>4897908</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5070017</td>\n",
       "      <td>5162528</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Atp6v1h</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1914864</td>\n",
       "      <td>OTTMUSG00000050145.9</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUST00000192847.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5588465</td>\n",
       "      <td>5606130</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21536</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>78835720</td>\n",
       "      <td>78838055</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20806</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434162</td>\n",
       "      <td>OTTMUSG00000046577.2</td>\n",
       "      <td></td>\n",
       "      <td>21668</td>\n",
       "      <td>ENSMUST00000190349.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21537</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>79148788</td>\n",
       "      <td>79151121</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20917</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434273</td>\n",
       "      <td>OTTMUSG00000046619.2</td>\n",
       "      <td></td>\n",
       "      <td>21669</td>\n",
       "      <td>ENSMUST00000188706.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21538</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>84562571</td>\n",
       "      <td>84564906</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21394</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434749</td>\n",
       "      <td>OTTMUSG00000045415.1</td>\n",
       "      <td></td>\n",
       "      <td>21670</td>\n",
       "      <td>ENSMUST00000189463.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21539</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>85528516</td>\n",
       "      <td>85530907</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434210</td>\n",
       "      <td>OTTMUSG00000042966.1</td>\n",
       "      <td></td>\n",
       "      <td>21671</td>\n",
       "      <td>ENSMUST00000181549.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21540</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89713423</td>\n",
       "      <td>89745531</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5440224</td>\n",
       "      <td>OTTMUSG00000047316.1</td>\n",
       "      <td></td>\n",
       "      <td>21672</td>\n",
       "      <td>ENSMUST00000188269.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21541 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +13   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 4807787   | 4848409   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4807891   | 4886769   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4857813   | 4897908   | ...   |\n",
       "| 1            | HAVANA     | gene       | 5070017   | 5162528   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 79148788  | 79151121  | ...   |\n",
       "| Y            | HAVANA     | gene       | 84562571  | 84564906  | ...   |\n",
       "| Y            | HAVANA     | gene       | 85528516  | 85530907  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89713423  | 89745531  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 21,541 rows and 18 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "13 hidden columns: Score, Strand, Phase, Id, Gene_id, Gene_type, Gene_name, Level, ... (+ 5 more.)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Function to map transcription_id\n",
    "def map_transcription_id(gene_id, gene_to_transcription):\n",
    "    return gene_to_transcription.get(gene_id, None)\n",
    "\n",
    "# Map transcription_id \n",
    "genes_df = genes.df\n",
    "genes_df['Transcript_id'] = genes_df['Gene_id'].map(lambda gene_id: map_transcription_id(gene_id, gene_to_transcription))\n",
    "genes_clean = pr.PyRanges(genes_df)\n",
    "\n",
    "# Find and Handle missing gene_ids\n",
    "missing_gene_ids = genes_clean.df[genes_clean.df['Transcript_id'].isna()]['Gene_id'].tolist()\n",
    "missing_df = genes_clean.df[genes_clean.df['Gene_id'].isin(missing_gene_ids)]\n",
    "genes_clean = genes_clean[~genes_clean.df['Transcript_id'].isna()]\n",
    "print(f\"all genes: {len(genes)}\")\n",
    "print(f\"all genes with annotated transcripts: {len(genes_clean)}\")\n",
    "genes_clean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Strand\n",
       "-    10801\n",
       "+    10740\n",
       ".        0\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_clean.Strand.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807787</td>\n",
       "      <td>4848409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807891</td>\n",
       "      <td>4886769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5611216</td>\n",
       "      <td>OTTMUSG00000050100.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUST00000155020.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4857813</td>\n",
       "      <td>4897908</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5070017</td>\n",
       "      <td>5162528</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Atp6v1h</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1914864</td>\n",
       "      <td>OTTMUSG00000050145.9</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUST00000192847.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5588465</td>\n",
       "      <td>5606130</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10735</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>87116820</td>\n",
       "      <td>87142916</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21477</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434832</td>\n",
       "      <td>OTTMUSG00000047083.1</td>\n",
       "      <td></td>\n",
       "      <td>21601</td>\n",
       "      <td>ENSMUST00000189543.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10736</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>87550965</td>\n",
       "      <td>87575690</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20906</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td></td>\n",
       "      <td>21602</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10737</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>88053314</td>\n",
       "      <td>88079493</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm28102</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td></td>\n",
       "      <td>21603</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10738</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89052805</td>\n",
       "      <td>89079014</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102045.1</td>\n",
       "      <td>ENSMUSG00000102045.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21294</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434649</td>\n",
       "      <td>OTTMUSG00000047309.1</td>\n",
       "      <td></td>\n",
       "      <td>21604</td>\n",
       "      <td>ENSMUST00000186443.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10739</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90401247</td>\n",
       "      <td>90433262</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000096178.7</td>\n",
       "      <td>ENSMUSG00000096178.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20837</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434193</td>\n",
       "      <td>OTTMUSG00000047352.1</td>\n",
       "      <td></td>\n",
       "      <td>21605</td>\n",
       "      <td>ENSMUST00000190558.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10740 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +13   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 4807787   | 4848409   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4807891   | 4886769   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4857813   | 4897908   | ...   |\n",
       "| 1            | HAVANA     | gene       | 5070017   | 5162528   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 87550965  | 87575690  | ...   |\n",
       "| Y            | HAVANA     | gene       | 88053314  | 88079493  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89052805  | 89079014  | ...   |\n",
       "| Y            | HAVANA     | gene       | 90401247  | 90433262  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 10,740 rows and 18 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "13 hidden columns: Score, Strand, Phase, Id, Gene_id, Gene_type, Gene_name, Level, ... (+ 5 more.)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_plus = genes_clean[genes_clean.Strand == '+']\n",
    "genes_minus = genes_clean[genes_clean.Strand == '-']\n",
    "genes_plus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3205900</td>\n",
       "      <td>3671497</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td></td>\n",
       "      <td>611</td>\n",
       "      <td>ENSMUST00000070533.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3999556</td>\n",
       "      <td>4409240</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025900.13</td>\n",
       "      <td>ENSMUSG00000025900.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>OTTMUSG00000049985.3</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>612</td>\n",
       "      <td>ENSMUST00000027032.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4490930</td>\n",
       "      <td>4497353</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025902.13</td>\n",
       "      <td>ENSMUSG00000025902.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:107543</td>\n",
       "      <td>OTTMUSG00000050014.7</td>\n",
       "      <td></td>\n",
       "      <td>613</td>\n",
       "      <td>ENSMUST00000027035.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4773205</td>\n",
       "      <td>4785738</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033845.13</td>\n",
       "      <td>ENSMUSG00000033845.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1351639</td>\n",
       "      <td>OTTMUSG00000029329.3</td>\n",
       "      <td></td>\n",
       "      <td>614</td>\n",
       "      <td>ENSMUST00000156816.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4909575</td>\n",
       "      <td>5070284</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000002459.17</td>\n",
       "      <td>ENSMUSG00000002459.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Rgs20</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1929866</td>\n",
       "      <td>OTTMUSG00000029338.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>615</td>\n",
       "      <td>ENSMUST00000002533.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10796</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>78835720</td>\n",
       "      <td>78838055</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20806</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434162</td>\n",
       "      <td>OTTMUSG00000046577.2</td>\n",
       "      <td></td>\n",
       "      <td>21668</td>\n",
       "      <td>ENSMUST00000190349.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10797</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>79148788</td>\n",
       "      <td>79151121</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20917</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434273</td>\n",
       "      <td>OTTMUSG00000046619.2</td>\n",
       "      <td></td>\n",
       "      <td>21669</td>\n",
       "      <td>ENSMUST00000188706.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10798</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>84562571</td>\n",
       "      <td>84564906</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21394</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434749</td>\n",
       "      <td>OTTMUSG00000045415.1</td>\n",
       "      <td></td>\n",
       "      <td>21670</td>\n",
       "      <td>ENSMUST00000189463.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10799</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>85528516</td>\n",
       "      <td>85530907</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434210</td>\n",
       "      <td>OTTMUSG00000042966.1</td>\n",
       "      <td></td>\n",
       "      <td>21671</td>\n",
       "      <td>ENSMUST00000181549.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10800</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89713423</td>\n",
       "      <td>89745531</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5440224</td>\n",
       "      <td>OTTMUSG00000047316.1</td>\n",
       "      <td></td>\n",
       "      <td>21672</td>\n",
       "      <td>ENSMUST00000188269.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10801 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +13   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 3205900   | 3671497   | ...   |\n",
       "| 1            | HAVANA     | gene       | 3999556   | 4409240   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4490930   | 4497353   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4773205   | 4785738   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 79148788  | 79151121  | ...   |\n",
       "| Y            | HAVANA     | gene       | 84562571  | 84564906  | ...   |\n",
       "| Y            | HAVANA     | gene       | 85528516  | 85530907  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89713423  | 89745531  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 10,801 rows and 18 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "13 hidden columns: Score, Strand, Phase, Id, Gene_id, Gene_type, Gene_name, Level, ... (+ 5 more.)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_minus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tags for 132 missing genes:\n",
      "Tag\n",
      "                                                 87\n",
      "overlapping_locus                                13\n",
      "reference_genome_error                           10\n",
      "fragmented_locus                                  9\n",
      "ncRNA_host,fragmented_locus                       2\n",
      "ncRNA_host                                        2\n",
      "overlapping_locus,reference_genome_error          2\n",
      "ncRNA_host,reference_genome_error                 2\n",
      "ncRNA_host,overlapping_locus                      2\n",
      "fragmented_locus,reference_genome_error           1\n",
      "fragmented_locus,overlapping_locus                1\n",
      "ncRNA_host,fragmented_locus,overlapping_locus     1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(f\"Tags for {len(missing_df)} missing genes:\")\n",
    "print(missing_df['Tag'].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract Regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### requires stitching"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 200bp before Transcription Start Site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default arguments for get_tss_region and get_transcription_end_region\n",
    "default_args = {\n",
    "    \"contig\": None,\n",
    "    \"start\": None,\n",
    "    \"end\": None,\n",
    "    \"reference\": \"mm10\",\n",
    "    \"as_pyranges\": True,\n",
    "    \"protein_coding\": True,\n",
    "    \"filterby\": None,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-01 14:52:19 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "before_tss regions with transcripts: 21541\n",
      "====================================================================================================\n"
     ]
    }
   ],
   "source": [
    "before_tss = get_tss_region(\n",
    "    start_offset=-200,\n",
    "    span=200,\n",
    "    **default_args,\n",
    ")\n",
    "\n",
    "# Map transcription_id \n",
    "before_tss_df = before_tss.df\n",
    "before_tss_df['Transcript_id'] = before_tss_df['Gene_id'].map(lambda gene_id: map_transcription_id(gene_id, gene_to_transcription))\n",
    "before_tss = pr.PyRanges(before_tss_df)\n",
    "before_tss = before_tss[~before_tss.df['Transcript_id'].isna()]\n",
    "\n",
    "print(f\"before_tss regions with transcripts: {len(before_tss)}\")\n",
    "print(\"=\"*100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807587</td>\n",
       "      <td>4807787</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807691</td>\n",
       "      <td>4807891</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5611216</td>\n",
       "      <td>OTTMUSG00000050100.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUST00000155020.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4857613</td>\n",
       "      <td>4857813</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5069817</td>\n",
       "      <td>5070017</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Atp6v1h</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1914864</td>\n",
       "      <td>OTTMUSG00000050145.9</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUST00000192847.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5588265</td>\n",
       "      <td>5588465</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10735</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>87116620</td>\n",
       "      <td>87116820</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21477</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434832</td>\n",
       "      <td>OTTMUSG00000047083.1</td>\n",
       "      <td></td>\n",
       "      <td>21601</td>\n",
       "      <td>ENSMUST00000189543.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10736</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>87550765</td>\n",
       "      <td>87550965</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20906</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td></td>\n",
       "      <td>21602</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10737</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>88053114</td>\n",
       "      <td>88053314</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm28102</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td></td>\n",
       "      <td>21603</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10738</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89052605</td>\n",
       "      <td>89052805</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102045.1</td>\n",
       "      <td>ENSMUSG00000102045.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21294</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434649</td>\n",
       "      <td>OTTMUSG00000047309.1</td>\n",
       "      <td></td>\n",
       "      <td>21604</td>\n",
       "      <td>ENSMUST00000186443.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10739</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90401047</td>\n",
       "      <td>90401247</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000096178.7</td>\n",
       "      <td>ENSMUSG00000096178.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20837</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434193</td>\n",
       "      <td>OTTMUSG00000047352.1</td>\n",
       "      <td></td>\n",
       "      <td>21605</td>\n",
       "      <td>ENSMUST00000190558.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10740 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +12   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 4807587   | 4807787   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4807691   | 4807891   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4857613   | 4857813   | ...   |\n",
       "| 1            | HAVANA     | gene       | 5069817   | 5070017   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 87550765  | 87550965  | ...   |\n",
       "| Y            | HAVANA     | gene       | 88053114  | 88053314  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89052605  | 89052805  | ...   |\n",
       "| Y            | HAVANA     | gene       | 90401047  | 90401247  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Unstranded PyRanges object has 10,740 rows and 17 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "12 hidden columns: Score, Phase, Id, Gene_id, Gene_type, Gene_name, Level, Mgi_id, ... (+ 4 more.)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#split strands\n",
    "before_tss_plus = pr.PyRanges(before_tss.df[before_tss.df[\"Strand\"]==\"+\"])\n",
    "before_tss_minus = pr.PyRanges(before_tss.df[before_tss.df[\"Strand\"]==\"-\"])\n",
    "\n",
    "before_tss_plus = before_tss_plus.unstrand()\n",
    "before_tss_minus = before_tss_minus.unstrand()\n",
    "before_tss_plus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3671497</td>\n",
       "      <td>3671697</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td></td>\n",
       "      <td>611</td>\n",
       "      <td>ENSMUST00000070533.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4409240</td>\n",
       "      <td>4409440</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025900.13</td>\n",
       "      <td>ENSMUSG00000025900.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>OTTMUSG00000049985.3</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>612</td>\n",
       "      <td>ENSMUST00000027032.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4497353</td>\n",
       "      <td>4497553</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025902.13</td>\n",
       "      <td>ENSMUSG00000025902.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:107543</td>\n",
       "      <td>OTTMUSG00000050014.7</td>\n",
       "      <td></td>\n",
       "      <td>613</td>\n",
       "      <td>ENSMUST00000027035.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4785738</td>\n",
       "      <td>4785938</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033845.13</td>\n",
       "      <td>ENSMUSG00000033845.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1351639</td>\n",
       "      <td>OTTMUSG00000029329.3</td>\n",
       "      <td></td>\n",
       "      <td>614</td>\n",
       "      <td>ENSMUST00000156816.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5070284</td>\n",
       "      <td>5070484</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000002459.17</td>\n",
       "      <td>ENSMUSG00000002459.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Rgs20</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1929866</td>\n",
       "      <td>OTTMUSG00000029338.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>615</td>\n",
       "      <td>ENSMUST00000002533.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10796</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>78838055</td>\n",
       "      <td>78838255</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20806</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434162</td>\n",
       "      <td>OTTMUSG00000046577.2</td>\n",
       "      <td></td>\n",
       "      <td>21668</td>\n",
       "      <td>ENSMUST00000190349.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10797</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>79151121</td>\n",
       "      <td>79151321</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20917</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434273</td>\n",
       "      <td>OTTMUSG00000046619.2</td>\n",
       "      <td></td>\n",
       "      <td>21669</td>\n",
       "      <td>ENSMUST00000188706.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10798</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>84564906</td>\n",
       "      <td>84565106</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21394</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434749</td>\n",
       "      <td>OTTMUSG00000045415.1</td>\n",
       "      <td></td>\n",
       "      <td>21670</td>\n",
       "      <td>ENSMUST00000189463.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10799</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>85530907</td>\n",
       "      <td>85531107</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434210</td>\n",
       "      <td>OTTMUSG00000042966.1</td>\n",
       "      <td></td>\n",
       "      <td>21671</td>\n",
       "      <td>ENSMUST00000181549.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10800</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89745531</td>\n",
       "      <td>89745731</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5440224</td>\n",
       "      <td>OTTMUSG00000047316.1</td>\n",
       "      <td></td>\n",
       "      <td>21672</td>\n",
       "      <td>ENSMUST00000188269.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10801 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +12   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 3671497   | 3671697   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4409240   | 4409440   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4497353   | 4497553   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4785738   | 4785938   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 79151121  | 79151321  | ...   |\n",
       "| Y            | HAVANA     | gene       | 84564906  | 84565106  | ...   |\n",
       "| Y            | HAVANA     | gene       | 85530907  | 85531107  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89745531  | 89745731  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Unstranded PyRanges object has 10,801 rows and 17 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "12 hidden columns: Score, Phase, Id, Gene_id, Gene_type, Gene_name, Level, Mgi_id, ... (+ 4 more.)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "before_tss_minus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1000bp after Transcription End Site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-01 14:52:46 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after_tes regions with transcripts: 21541\n",
      "====================================================================================================\n"
     ]
    }
   ],
   "source": [
    "after_tes = get_transcription_end_region(\n",
    "    start_offset=0,\n",
    "    span=1000,\n",
    "    **default_args,\n",
    ")\n",
    "\n",
    "# Map transcription_id \n",
    "after_tes_df = after_tes.df\n",
    "after_tes_df['Transcript_id'] = after_tes_df['Gene_id'].map(lambda gene_id: map_transcription_id(gene_id, gene_to_transcription))\n",
    "after_tes = pr.PyRanges(after_tes_df)\n",
    "after_tes = after_tes[~after_tes.df['Transcript_id'].isna()]\n",
    "\n",
    "print(f\"after_tes regions with transcripts: {len(after_tes)}\")\n",
    "print(\"=\"*100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4848409</td>\n",
       "      <td>4849409</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4886769</td>\n",
       "      <td>4887769</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5611216</td>\n",
       "      <td>OTTMUSG00000050100.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUST00000155020.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4897908</td>\n",
       "      <td>4898908</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5162528</td>\n",
       "      <td>5163528</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Atp6v1h</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1914864</td>\n",
       "      <td>OTTMUSG00000050145.9</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUST00000192847.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5606130</td>\n",
       "      <td>5607130</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10735</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>87142916</td>\n",
       "      <td>87143916</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21477</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434832</td>\n",
       "      <td>OTTMUSG00000047083.1</td>\n",
       "      <td></td>\n",
       "      <td>21601</td>\n",
       "      <td>ENSMUST00000189543.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10736</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>87575690</td>\n",
       "      <td>87576690</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20906</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td></td>\n",
       "      <td>21602</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10737</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>88079493</td>\n",
       "      <td>88080493</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm28102</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td></td>\n",
       "      <td>21603</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10738</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89079014</td>\n",
       "      <td>89080014</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102045.1</td>\n",
       "      <td>ENSMUSG00000102045.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21294</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434649</td>\n",
       "      <td>OTTMUSG00000047309.1</td>\n",
       "      <td></td>\n",
       "      <td>21604</td>\n",
       "      <td>ENSMUST00000186443.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10739</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90433262</td>\n",
       "      <td>90434262</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000096178.7</td>\n",
       "      <td>ENSMUSG00000096178.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20837</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434193</td>\n",
       "      <td>OTTMUSG00000047352.1</td>\n",
       "      <td></td>\n",
       "      <td>21605</td>\n",
       "      <td>ENSMUST00000190558.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10740 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +12   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 4848409   | 4849409   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4886769   | 4887769   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4897908   | 4898908   | ...   |\n",
       "| 1            | HAVANA     | gene       | 5162528   | 5163528   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 87575690  | 87576690  | ...   |\n",
       "| Y            | HAVANA     | gene       | 88079493  | 88080493  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89079014  | 89080014  | ...   |\n",
       "| Y            | HAVANA     | gene       | 90433262  | 90434262  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Unstranded PyRanges object has 10,740 rows and 17 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "12 hidden columns: Score, Phase, Id, Gene_id, Gene_type, Gene_name, Level, Mgi_id, ... (+ 4 more.)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#split strands\n",
    "after_tes_plus = pr.PyRanges(after_tes.df[after_tes.df[\"Strand\"]==\"+\"])\n",
    "after_tes_minus = pr.PyRanges(after_tes.df[after_tes.df[\"Strand\"]==\"-\"])\n",
    "\n",
    "after_tes_plus = after_tes_plus.unstrand()\n",
    "after_tes_minus = after_tes_minus.unstrand()\n",
    "after_tes_plus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3204900</td>\n",
       "      <td>3205900</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td></td>\n",
       "      <td>611</td>\n",
       "      <td>ENSMUST00000070533.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3998556</td>\n",
       "      <td>3999556</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025900.13</td>\n",
       "      <td>ENSMUSG00000025900.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>OTTMUSG00000049985.3</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>612</td>\n",
       "      <td>ENSMUST00000027032.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4489930</td>\n",
       "      <td>4490930</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025902.13</td>\n",
       "      <td>ENSMUSG00000025902.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:107543</td>\n",
       "      <td>OTTMUSG00000050014.7</td>\n",
       "      <td></td>\n",
       "      <td>613</td>\n",
       "      <td>ENSMUST00000027035.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4772205</td>\n",
       "      <td>4773205</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033845.13</td>\n",
       "      <td>ENSMUSG00000033845.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1351639</td>\n",
       "      <td>OTTMUSG00000029329.3</td>\n",
       "      <td></td>\n",
       "      <td>614</td>\n",
       "      <td>ENSMUST00000156816.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4908575</td>\n",
       "      <td>4909575</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000002459.17</td>\n",
       "      <td>ENSMUSG00000002459.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Rgs20</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1929866</td>\n",
       "      <td>OTTMUSG00000029338.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>615</td>\n",
       "      <td>ENSMUST00000002533.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10796</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>78834720</td>\n",
       "      <td>78835720</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20806</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434162</td>\n",
       "      <td>OTTMUSG00000046577.2</td>\n",
       "      <td></td>\n",
       "      <td>21668</td>\n",
       "      <td>ENSMUST00000190349.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10797</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>79147788</td>\n",
       "      <td>79148788</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20917</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434273</td>\n",
       "      <td>OTTMUSG00000046619.2</td>\n",
       "      <td></td>\n",
       "      <td>21669</td>\n",
       "      <td>ENSMUST00000188706.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10798</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>84561571</td>\n",
       "      <td>84562571</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21394</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434749</td>\n",
       "      <td>OTTMUSG00000045415.1</td>\n",
       "      <td></td>\n",
       "      <td>21670</td>\n",
       "      <td>ENSMUST00000189463.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10799</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>85527516</td>\n",
       "      <td>85528516</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434210</td>\n",
       "      <td>OTTMUSG00000042966.1</td>\n",
       "      <td></td>\n",
       "      <td>21671</td>\n",
       "      <td>ENSMUST00000181549.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10800</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89712423</td>\n",
       "      <td>89713423</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5440224</td>\n",
       "      <td>OTTMUSG00000047316.1</td>\n",
       "      <td></td>\n",
       "      <td>21672</td>\n",
       "      <td>ENSMUST00000188269.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10801 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +12   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 3204900   | 3205900   | ...   |\n",
       "| 1            | HAVANA     | gene       | 3998556   | 3999556   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4489930   | 4490930   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4772205   | 4773205   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 79147788  | 79148788  | ...   |\n",
       "| Y            | HAVANA     | gene       | 84561571  | 84562571  | ...   |\n",
       "| Y            | HAVANA     | gene       | 85527516  | 85528516  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89712423  | 89713423  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Unstranded PyRanges object has 10,801 rows and 17 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "12 hidden columns: Score, Phase, Id, Gene_id, Gene_type, Gene_name, Level, Mgi_id, ... (+ 4 more.)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "after_tes_minus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract 3`UTRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-01 14:53:03 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "strand shape: + (8075, 27), - (8134, 27)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>...</th>\n",
       "      <th>Level</th>\n",
       "      <th>Protein_id</th>\n",
       "      <th>Transcript_support_level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ccdsid</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Havana_transcript</th>\n",
       "      <th>Ont</th>\n",
       "      <th>Ranges_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>4845016</td>\n",
       "      <td>4846738</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000027036.10</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000027036.4</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14806.1</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>OTTMUST00000051162.1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>4896364</td>\n",
       "      <td>4897904</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000081551.13</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000080266.7</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>non_canonical_U12,basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS35505.1</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>OTTMUST00000111602.1</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>5602784</td>\n",
       "      <td>5606130</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000125105.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14809.1</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td>OTTMUST00000088255.1</td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>6274275</td>\n",
       "      <td>6276647</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000027040.12</td>\n",
       "      <td>ENSMUST00000027040.12</td>\n",
       "      <td>ENSMUSG00000025907.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000027040.6</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1341850</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS35507.1</td>\n",
       "      <td>OTTMUSG00000033467.12</td>\n",
       "      <td>OTTMUST00000084091.5</td>\n",
       "      <td></td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>6391104</td>\n",
       "      <td>6391115</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000133144.3</td>\n",
       "      <td>ENSMUST00000133144.3</td>\n",
       "      <td>ENSMUSG00000087247.3</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000137420.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:3645495</td>\n",
       "      <td>not_organism_supported,basic,appris_principal_...</td>\n",
       "      <td>CCDS56620.1</td>\n",
       "      <td>OTTMUSG00000050239.2</td>\n",
       "      <td>OTTMUST00000127625.2</td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8070</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>84772706</td>\n",
       "      <td>84772901</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000186110.1</td>\n",
       "      <td>ENSMUST00000186110.1</td>\n",
       "      <td>ENSMUSG00000099840.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140850.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434764</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000046908.1</td>\n",
       "      <td>OTTMUST00000121611.1</td>\n",
       "      <td></td>\n",
       "      <td>27079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8071</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>86087758</td>\n",
       "      <td>86087951</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000188754.1</td>\n",
       "      <td>ENSMUST00000188754.1</td>\n",
       "      <td>ENSMUSG00000100240.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000139858.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434176</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047031.1</td>\n",
       "      <td>OTTMUST00000121805.1</td>\n",
       "      <td></td>\n",
       "      <td>27080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8072</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>87142720</td>\n",
       "      <td>87142916</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000189543.6</td>\n",
       "      <td>ENSMUST00000189543.6</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140238.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434832</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047083.1</td>\n",
       "      <td>OTTMUST00000121876.1</td>\n",
       "      <td></td>\n",
       "      <td>27081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8073</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>87575495</td>\n",
       "      <td>87575690</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140305.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td>OTTMUST00000121968.1</td>\n",
       "      <td></td>\n",
       "      <td>27082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8074</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>88079297</td>\n",
       "      <td>88079493</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140242.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td>OTTMUST00000121980.1</td>\n",
       "      <td></td>\n",
       "      <td>27083</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8075 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------------+-----------+-------+\n",
       "| Chromosome   | Source     | Type            | Start     | +23   |\n",
       "| (category)   | (object)   | (object)        | (int64)   | ...   |\n",
       "|--------------+------------+-----------------+-----------+-------|\n",
       "| 1            | HAVANA     | three_prime_UTR | 4845016   | ...   |\n",
       "| 1            | HAVANA     | three_prime_UTR | 4896364   | ...   |\n",
       "| 1            | HAVANA     | three_prime_UTR | 5602784   | ...   |\n",
       "| 1            | HAVANA     | three_prime_UTR | 6274275   | ...   |\n",
       "| ...          | ...        | ...             | ...       | ...   |\n",
       "| Y            | HAVANA     | three_prime_UTR | 86087758  | ...   |\n",
       "| Y            | HAVANA     | three_prime_UTR | 87142720  | ...   |\n",
       "| Y            | HAVANA     | three_prime_UTR | 87575495  | ...   |\n",
       "| Y            | HAVANA     | three_prime_UTR | 88079297  | ...   |\n",
       "+--------------+------------+-----------------+-----------+-------+\n",
       "Unstranded PyRanges object has 8,075 rows and 27 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "23 hidden columns: End, Score, Phase, Id, Parent, Gene_id, Transcript_id, ... (+ 16 more.)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract with modality.get_three_prime_utrs\n",
    "three_prime_utrs = get_three_prime_utrs(reference=\"mm10\")\n",
    "three_prime_utrs = three_prime_utrs[\n",
    "    three_prime_utrs.Transcript_id.isin(selected_transcripts.transcript_id)\n",
    "    ]\n",
    "\n",
    "#split strands\n",
    "three_prime_utrs_plus = pr.PyRanges(three_prime_utrs.df[three_prime_utrs.df[\"Strand\"]==\"+\"])\n",
    "three_prime_utrs_plus = three_prime_utrs_plus.unstrand()\n",
    "\n",
    "three_prime_utrs_minus = pr.PyRanges(three_prime_utrs.df[three_prime_utrs.df[\"Strand\"]==\"-\"])\n",
    "three_prime_utrs_minus = three_prime_utrs_minus.unstrand()\n",
    "\n",
    "print(f\"strand shape: + {three_prime_utrs_plus.df.shape}, - {three_prime_utrs_minus.df.shape}\")\n",
    "three_prime_utrs_plus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract 5`UTRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-01 14:53:20 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "strand shape: + (11449, 27), - (11369, 27)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>...</th>\n",
       "      <th>Level</th>\n",
       "      <th>Protein_id</th>\n",
       "      <th>Transcript_support_level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ccdsid</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Havana_transcript</th>\n",
       "      <th>Ont</th>\n",
       "      <th>Ranges_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>4807822</td>\n",
       "      <td>4807912</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000027036.10</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000027036.4</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14806.1</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>OTTMUST00000051162.1</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>4857813</td>\n",
       "      <td>4857912</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000081551.13</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000080266.7</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>non_canonical_U12,basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS35505.1</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>OTTMUST00000111602.1</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>5588492</td>\n",
       "      <td>5588662</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000125105.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14809.1</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td>OTTMUST00000088255.1</td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>5589034</td>\n",
       "      <td>5589047</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000125105.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14809.1</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td>OTTMUST00000088255.1</td>\n",
       "      <td></td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>6214644</td>\n",
       "      <td>6214956</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000027040.12</td>\n",
       "      <td>ENSMUST00000027040.12</td>\n",
       "      <td>ENSMUSG00000025907.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000027040.6</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1341850</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS35507.1</td>\n",
       "      <td>OTTMUSG00000033467.12</td>\n",
       "      <td>OTTMUST00000084091.5</td>\n",
       "      <td></td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11444</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>87118231</td>\n",
       "      <td>87118244</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000189543.6</td>\n",
       "      <td>ENSMUST00000189543.6</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140238.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434832</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047083.1</td>\n",
       "      <td>OTTMUST00000121876.1</td>\n",
       "      <td></td>\n",
       "      <td>39793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11445</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>87550965</td>\n",
       "      <td>87551008</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140305.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td>OTTMUST00000121968.1</td>\n",
       "      <td></td>\n",
       "      <td>39794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11446</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>87552378</td>\n",
       "      <td>87552391</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140305.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td>OTTMUST00000121968.1</td>\n",
       "      <td></td>\n",
       "      <td>39795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11447</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>88053314</td>\n",
       "      <td>88053357</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140242.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td>OTTMUST00000121980.1</td>\n",
       "      <td></td>\n",
       "      <td>39796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11448</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>88054723</td>\n",
       "      <td>88054736</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140242.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td>OTTMUST00000121980.1</td>\n",
       "      <td></td>\n",
       "      <td>39797</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11449 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+----------------+-----------+-------+\n",
       "| Chromosome   | Source     | Type           | Start     | +23   |\n",
       "| (category)   | (object)   | (object)       | (int64)   | ...   |\n",
       "|--------------+------------+----------------+-----------+-------|\n",
       "| 1            | HAVANA     | five_prime_UTR | 4807822   | ...   |\n",
       "| 1            | HAVANA     | five_prime_UTR | 4857813   | ...   |\n",
       "| 1            | HAVANA     | five_prime_UTR | 5588492   | ...   |\n",
       "| 1            | HAVANA     | five_prime_UTR | 5589034   | ...   |\n",
       "| ...          | ...        | ...            | ...       | ...   |\n",
       "| Y            | HAVANA     | five_prime_UTR | 87550965  | ...   |\n",
       "| Y            | HAVANA     | five_prime_UTR | 87552378  | ...   |\n",
       "| Y            | HAVANA     | five_prime_UTR | 88053314  | ...   |\n",
       "| Y            | HAVANA     | five_prime_UTR | 88054723  | ...   |\n",
       "+--------------+------------+----------------+-----------+-------+\n",
       "Unstranded PyRanges object has 11,449 rows and 27 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "23 hidden columns: End, Score, Phase, Id, Parent, Gene_id, Transcript_id, ... (+ 16 more.)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract with modality.get_five_prime_utrs\n",
    "five_prime_utrs = get_five_prime_utrs(reference=\"mm10\")\n",
    "five_prime_utrs = five_prime_utrs[\n",
    "    five_prime_utrs.Transcript_id.isin(selected_transcripts.transcript_id)\n",
    "    ]\n",
    "\n",
    "#split strands\n",
    "five_prime_utrs_plus = pr.PyRanges(five_prime_utrs.df[five_prime_utrs.df[\"Strand\"]==\"+\"])\n",
    "five_prime_utrs_plus = five_prime_utrs_plus.unstrand()\n",
    "\n",
    "five_prime_utrs_minus = pr.PyRanges(five_prime_utrs.df[five_prime_utrs.df[\"Strand\"]==\"-\"])\n",
    "five_prime_utrs_minus = five_prime_utrs_minus.unstrand()\n",
    "print(f\"strand shape: + {five_prime_utrs_plus.df.shape}, - {five_prime_utrs_minus.df.shape}\")\n",
    "five_prime_utrs_plus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract exons and introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-01 14:53:47 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    }
   ],
   "source": [
    "# Exons:\n",
    "exons = get_exons(reference=\"mm10\")\n",
    "exons = exons[exons.Transcript_id.isin(selected_transcripts.transcript_id)]\n",
    "\n",
    "# exons split strands\n",
    "exons_plus = pr.PyRanges(exons.df[exons.df[\"Strand\"]==\"+\"])\n",
    "first_exons_plus = exons_plus[exons_plus.Exon_number == \"1\"].unstrand()\n",
    "exons_plus = exons_plus[exons_plus.Exon_number.astype(\"int\") > 1].unstrand()\n",
    "\n",
    "\n",
    "exons_minus = pr.PyRanges(exons.df[exons.df[\"Strand\"]==\"-\"])\n",
    "first_exons_minus = exons_minus[exons_minus.Exon_number == \"1\"].unstrand()\n",
    "exons_minus = exons_minus[exons_minus.Exon_number.astype(\"int\") > 1].unstrand()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-01 14:55:03 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n",
      "2024-10-01 14:55:26 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n",
      "2024-10-01 14:56:02 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n",
      "2024-10-01 14:56:17 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 8 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "#Introns:\n",
    "introns = get_introns(\n",
    "    reference=\"mm10\",\n",
    "    transcripts=selected_transcripts.transcript_id.values,\n",
    "    nb_workers=8,\n",
    ")\n",
    "\n",
    "# introns split strands\n",
    "introns_plus = pr.PyRanges(introns.df[introns.df[\"Strand\"]==\"+\"])\n",
    "first_introns_plus = introns_plus[introns_plus.Intron_number == \"1\"].unstrand()\n",
    "introns_plus = introns_plus[introns_plus.Intron_number.astype(\"int\") > 1].unstrand()\n",
    "\n",
    "\n",
    "introns_minus = pr.PyRanges(introns.df[introns.df[\"Strand\"]==\"-\"])\n",
    "first_introns_minus = introns_minus[introns_minus.Intron_number == \"1\"].unstrand()\n",
    "introns_minus = introns_minus[introns_minus.Intron_number.astype(\"int\") > 1].unstrand()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exons strand shape: + (90807, 27), - (90167, 27)\n",
      "Introns strand shape: + (85982, 20), - (85480, 20)\n",
      "First Exons strand shape: + (10740, 27), - (10801, 27)\n",
      "First Introns strand shape: + (10186, 20), - (10212, 20)\n"
     ]
    }
   ],
   "source": [
    "print(f\"Exons strand shape: + {exons_plus.df.shape}, - {exons_minus.df.shape}\")\n",
    "print(f\"Introns strand shape: + {introns_plus.df.shape}, - {introns_minus.df.shape}\")\n",
    "\n",
    "print(f\"First Exons strand shape: + {first_exons_plus.df.shape}, - {first_exons_minus.df.shape}\")\n",
    "print(f\"First Introns strand shape: + {first_introns_plus.df.shape}, - {first_introns_minus.df.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Source', 'Type', 'Start', 'End', 'Score', 'Strand',\n",
      "       'Phase', 'Id', 'Parent', 'Gene_id', 'Transcript_id', 'Gene_type',\n",
      "       'Gene_name', 'Transcript_type', 'Transcript_name', 'Exon_number',\n",
      "       'Exon_id', 'Level', 'Transcript_support_level', 'Mgi_id', 'Tag',\n",
      "       'Havana_gene', 'Havana_transcript', 'Protein_id', 'Ccdsid', 'Ont',\n",
      "       'Ranges_ID'],\n",
      "      dtype='object')\n",
      "Index(['Transcript_id', 'Level_1', 'Chromosome', 'Source', 'Type', 'Start',\n",
      "       'End', 'Score', 'Strand', 'Phase', 'Id', 'Gene_id', 'Gene_type',\n",
      "       'Gene_name', 'Level', 'Mgi_id', 'Havana_gene', 'Tag', 'Ranges_id',\n",
      "       'Intron_number', 'Ranges_ID'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(exons.columns) # get_exons don't return \"Transcript_support_level\", 'Havana_transcript', 'Protein_id', 'Ccdsid', 'Ont', 'Parent', 'Exon_id', 'Exon_number'\n",
    "print(introns.columns) # get_introns specific fields return 'Level_1'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Summarise 3\\` 5\\` UTRs exons and introns regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mean_across_region(df, var):\n",
    "    sum_var = df[f\"{var}_sum\"].sum()\n",
    "    sum_total_c = df[\"num_total_c_sum\"].sum()\n",
    "    return sum_var / sum_total_c\n",
    "def summarise_across_region(rdr, region):\n",
    "    \"\"\"\n",
    "    Summarise the methylation data across a region.\n",
    "\n",
    "    Parameters\n",
    "        rdr: pyranges object\n",
    "        region: str\n",
    "\n",
    "    Returns\n",
    "        grouped_df: pd.DataFrame\n",
    "    \"\"\"\n",
    "\n",
    "    df = rdr.to_dataframe().reset_index(level=\"sample_id\", drop=True)\n",
    "    grouped = df.groupby(\"Gene_id\")\n",
    "\n",
    "    mean_mc = grouped.apply(get_mean_across_region, var=\"num_mc\")\n",
    "    mean_hmc = grouped.apply(get_mean_across_region, var=\"num_hmc\")\n",
    "    mean_modc = grouped.apply(get_mean_across_region, var=\"num_modc\")\n",
    "    cpg_count = grouped[\"num_total_c_cpg_count\"].sum()\n",
    "    sum_c = grouped[\"num_total_c_sum\"].sum()\n",
    "    range_length = grouped[\"range_length\"].sum()\n",
    "    gene_name = grouped[\"Gene_name\"].first()\n",
    "    contig = grouped[\"contig\"].first()\n",
    "    strand = grouped[\"strand\"].first()\n",
    "    \n",
    "    \n",
    "    # Prepare list for storing primary_transcripts and handling missing gene_ids\n",
    "    primary_transcripts = []\n",
    "    missing_gene_ids = []\n",
    "\n",
    "    for gene_id in grouped.groups.keys():\n",
    "        if gene_id in gene_to_transcription:\n",
    "            primary_transcripts.append(gene_to_transcription[gene_id])\n",
    "        else:\n",
    "            primary_transcripts.append(None)\n",
    "            missing_gene_ids.append(gene_id)\n",
    "\n",
    "    # Print missing gene_ids\n",
    "    for gene_id in missing_gene_ids:\n",
    "        print(f\"Gene_ID {gene_id} not found in selected_transcripts\")\n",
    "\n",
    "    \n",
    "    grouped_df = pd.DataFrame(\n",
    "        {\n",
    "            \"mean_mc\": mean_mc,\n",
    "            \"mean_hmc\": mean_hmc,\n",
    "            \"mean_modc\": mean_modc,\n",
    "            \"cpg_count\": cpg_count,\n",
    "            \"total_c\" : sum_c,\n",
    "            \"range_length\": range_length,\n",
    "            \"Gene_name\": gene_name,\n",
    "            \"contig\": contig,\n",
    "            \"strand\": strand,\n",
    "            \"Transcript_id\": primary_transcripts,\n",
    "        }\n",
    "    )\n",
    "\n",
    "    grouped_df = grouped_df.reset_index()\n",
    "    grouped_df[\"Region\"] = region\n",
    "    return grouped_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "regions for plus strand:\n",
      "196313\n"
     ]
    }
   ],
   "source": [
    "five_prime_utrs_count = five_prime_utrs_plus.df.shape[0]\n",
    "exon_count = exons_plus.df.shape[0]\n",
    "intron_count = introns_plus.df.shape[0]\n",
    "three_prime_utrs_count = three_prime_utrs_plus.df.shape[0]\n",
    "\n",
    "print(\"regions for plus strand:\")\n",
    "print(five_prime_utrs_count+exon_count+intron_count+three_prime_utrs_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "regions for minus strand:\n",
      "195150\n"
     ]
    }
   ],
   "source": [
    "five_prime_utrs_count = five_prime_utrs_minus.df.shape[0]\n",
    "exon_count = exons_minus.df.shape[0]\n",
    "intron_count = introns_minus.df.shape[0]\n",
    "three_prime_utrs_count = three_prime_utrs_minus.df.shape[0]\n",
    "\n",
    "print(\"regions for minus strand:\")\n",
    "print(five_prime_utrs_count+exon_count+intron_count+three_prime_utrs_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'exons': +--------------+------------+------------+-----------+-----------+-------+\n",
       " | Chromosome   | Source     | Type       | Start     | End       | +22   |\n",
       " | (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       " |--------------+------------+------------+-----------+-----------+-------|\n",
       " | 1            | HAVANA     | exon       | 4808454   | 4808485   | ...   |\n",
       " | 1            | HAVANA     | exon       | 4828583   | 4828648   | ...   |\n",
       " | 1            | HAVANA     | exon       | 4828583   | 4828648   | ...   |\n",
       " | 1            | HAVANA     | exon       | 4830267   | 4830314   | ...   |\n",
       " | ...          | ...        | ...        | ...       | ...       | ...   |\n",
       " | Y            | HAVANA     | exon       | 90429592  | 90429709  | ...   |\n",
       " | Y            | HAVANA     | exon       | 90430460  | 90430559  | ...   |\n",
       " | Y            | HAVANA     | exon       | 90432572  | 90432670  | ...   |\n",
       " | Y            | HAVANA     | exon       | 90433029  | 90433262  | ...   |\n",
       " +--------------+------------+------------+-----------+-----------+-------+\n",
       " Unstranded PyRanges object has 90,807 rows and 27 columns from 21 chromosomes.\n",
       " For printing, the PyRanges was sorted on Chromosome.\n",
       " 22 hidden columns: Score, Phase, Id, Parent, Gene_id, Transcript_id, Gene_type, ... (+ 15 more.),\n",
       " 'introns': +-----------------------+-----------+--------------+------------+-------+\n",
       " | Transcript_id         | Level_1   | Chromosome   | Source     | +16   |\n",
       " | (object)              | (int64)   | (category)   | (object)   | ...   |\n",
       " |-----------------------+-----------+--------------+------------+-------|\n",
       " | ENSMUST00000000514.10 | 1         | 1            | HAVANA     | ...   |\n",
       " | ENSMUST00000000514.10 | 2         | 1            | HAVANA     | ...   |\n",
       " | ENSMUST00000000514.10 | 3         | 1            | HAVANA     | ...   |\n",
       " | ENSMUST00000000514.10 | 4         | 1            | HAVANA     | ...   |\n",
       " | ...                   | ...       | ...          | ...        | ...   |\n",
       " | ENSMUST00000194687.1  | 1         | Y            | HAVANA     | ...   |\n",
       " | ENSMUST00000195574.1  | 1         | Y            | HAVANA     | ...   |\n",
       " | ENSMUST00000195574.1  | 2         | Y            | HAVANA     | ...   |\n",
       " | ENSMUST00000195626.1  | 1         | Y            | HAVANA     | ...   |\n",
       " +-----------------------+-----------+--------------+------------+-------+\n",
       " Unstranded PyRanges object has 85,982 rows and 20 columns from 21 chromosomes.\n",
       " For printing, the PyRanges was sorted on Chromosome.\n",
       " 16 hidden columns: Type, Start, End, Score, Phase, Id, Gene_id, Gene_type, Gene_name, ... (+ 7 more.),\n",
       " 'five_prime_utrs': +--------------+------------+----------------+-----------+-------+\n",
       " | Chromosome   | Source     | Type           | Start     | +23   |\n",
       " | (category)   | (object)   | (object)       | (int64)   | ...   |\n",
       " |--------------+------------+----------------+-----------+-------|\n",
       " | 1            | HAVANA     | five_prime_UTR | 4807822   | ...   |\n",
       " | 1            | HAVANA     | five_prime_UTR | 4857813   | ...   |\n",
       " | 1            | HAVANA     | five_prime_UTR | 5588492   | ...   |\n",
       " | 1            | HAVANA     | five_prime_UTR | 5589034   | ...   |\n",
       " | ...          | ...        | ...            | ...       | ...   |\n",
       " | Y            | HAVANA     | five_prime_UTR | 87550965  | ...   |\n",
       " | Y            | HAVANA     | five_prime_UTR | 87552378  | ...   |\n",
       " | Y            | HAVANA     | five_prime_UTR | 88053314  | ...   |\n",
       " | Y            | HAVANA     | five_prime_UTR | 88054723  | ...   |\n",
       " +--------------+------------+----------------+-----------+-------+\n",
       " Unstranded PyRanges object has 11,449 rows and 27 columns from 21 chromosomes.\n",
       " For printing, the PyRanges was sorted on Chromosome.\n",
       " 23 hidden columns: End, Score, Phase, Id, Parent, Gene_id, Transcript_id, ... (+ 16 more.),\n",
       " 'three_prime_utrs': +--------------+------------+-----------------+-----------+-------+\n",
       " | Chromosome   | Source     | Type            | Start     | +23   |\n",
       " | (category)   | (object)   | (object)        | (int64)   | ...   |\n",
       " |--------------+------------+-----------------+-----------+-------|\n",
       " | 1            | HAVANA     | three_prime_UTR | 4845016   | ...   |\n",
       " | 1            | HAVANA     | three_prime_UTR | 4896364   | ...   |\n",
       " | 1            | HAVANA     | three_prime_UTR | 5602784   | ...   |\n",
       " | 1            | HAVANA     | three_prime_UTR | 6274275   | ...   |\n",
       " | ...          | ...        | ...             | ...       | ...   |\n",
       " | Y            | HAVANA     | three_prime_UTR | 86087758  | ...   |\n",
       " | Y            | HAVANA     | three_prime_UTR | 87142720  | ...   |\n",
       " | Y            | HAVANA     | three_prime_UTR | 87575495  | ...   |\n",
       " | Y            | HAVANA     | three_prime_UTR | 88079297  | ...   |\n",
       " +--------------+------------+-----------------+-----------+-------+\n",
       " Unstranded PyRanges object has 8,075 rows and 27 columns from 21 chromosomes.\n",
       " For printing, the PyRanges was sorted on Chromosome.\n",
       " 23 hidden columns: End, Score, Phase, Id, Parent, Gene_id, Transcript_id, ... (+ 16 more.)}"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict_regions_to_patch_plus = {\n",
    "    \"exons\": exons_plus,\n",
    "    \"introns\": introns_plus,\n",
    "    \"five_prime_utrs\": five_prime_utrs_plus,\n",
    "    \"three_prime_utrs\": three_prime_utrs_plus,\n",
    "}\n",
    "\n",
    "dict_regions_to_patch_minus = {\n",
    "    \"exons\": exons_minus,\n",
    "    \"introns\": introns_minus,\n",
    "    \"five_prime_utrs\": five_prime_utrs_minus,\n",
    "    \"three_prime_utrs\": three_prime_utrs_minus,\n",
    "}\n",
    "\n",
    "dict_regions_to_patch_plus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Build sense and antisense methylation summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarise_methylation_across_regions(plus_strand_cpgs, minus_strand_cpgs, regions_dict, region_strand, sense_methylation={}, antisense_methylation={}):\n",
    "    \"\"\"\n",
    "    Summarise the methylation data across multiple regions.\n",
    "\n",
    "    Parameters\n",
    "        rdr: pyranges object\n",
    "        regions_dict: dict\n",
    "\n",
    "    Returns\n",
    "        grouped_df: pd.DataFrame\n",
    "    \"\"\"\n",
    "    # For each region in + strand expressed genes, summarise the methylation data\n",
    "    for region, gr in regions_dict.items():\n",
    "        rds_plus = plus_strand_cpgs.reduce_byranges(\n",
    "                    ranges=gr.unstrand(), \n",
    "                    var=[\"num_mc\", \"num_hmc\", \"num_modc\", \"num_total_c\"]\n",
    "                )\n",
    "        \n",
    "        rds_minus = minus_strand_cpgs.reduce_byranges(\n",
    "                ranges=gr.unstrand(), \n",
    "                var=[\"num_mc\", \"num_hmc\", \"num_modc\", \"num_total_c\"]\n",
    "            )\n",
    "        \n",
    "        if region_strand == \"+\":\n",
    "            # sense to the + region\n",
    "            sense_plus_ds = rds_plus\n",
    "            plus_sense = ['+'] * sense_plus_ds.dims['ranges']\n",
    "            sense_plus_ds = sense_plus_ds.assign_coords(strand=('ranges', plus_sense))\n",
    "\n",
    "            # antisense to the + region\n",
    "            antisense_plus_ds = rds_minus\n",
    "            plus_antisense = ['+'] * antisense_plus_ds.dims['ranges']\n",
    "            antisense_plus_ds = antisense_plus_ds.assign_coords(strand=('ranges', plus_antisense))\n",
    "            \n",
    "            # Add the modified dataset to the dictionary\n",
    "            if region not in sense_methylation:\n",
    "                sense_methylation[region] = sense_plus_ds\n",
    "                antisense_methylation[region] = antisense_plus_ds\n",
    "            else:\n",
    "                sense_methylation[region] = xr.concat([sense_methylation[region], sense_plus_ds], dim=\"ranges\")\n",
    "                antisense_methylation[region] = xr.concat([antisense_methylation[region], antisense_plus_ds], dim=\"ranges\")\n",
    "        else:\n",
    "            # sense to the - region\n",
    "            sense_minus_ds = rds_minus\n",
    "            minus_sense = ['-'] * sense_minus_ds.dims['ranges']\n",
    "            sense_minus_ds = sense_minus_ds.assign_coords(strand=('ranges', minus_sense))\n",
    "            \n",
    "            # antisense to the - region\n",
    "            antisense_minus_ds = rds_plus\n",
    "            minus_antisense = ['-'] * antisense_minus_ds.dims['ranges']\n",
    "            antisense_minus_ds = antisense_minus_ds.assign_coords(strand=('ranges', minus_antisense))\n",
    "            \n",
    "\n",
    "            # Add the modified dataset to the dictionary\n",
    "            if region not in antisense_methylation:\n",
    "                sense_methylation[region] = sense_minus_ds\n",
    "                antisense_methylation[region] = antisense_minus_ds\n",
    "            else:\n",
    "                sense_methylation[region] = xr.concat([sense_methylation[region], sense_minus_ds], dim=\"ranges\")\n",
    "                antisense_methylation[region] = xr.concat([antisense_methylation[region], antisense_minus_ds], dim=\"ranges\")\n",
    "\n",
    "\n",
    "    return sense_methylation, antisense_methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "sense_methylation, antisense_methylation = summarise_methylation_across_regions(plus_strand_cov, minus_strand_cov, dict_regions_to_patch_plus, \"+\")\n",
    "sense_methylation, antisense_methylation = summarise_methylation_across_regions(plus_strand_cov, minus_strand_cov, dict_regions_to_patch_minus, \"-\", sense_methylation, antisense_methylation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;\n",
       "Dimensions:                   (ranges: 180974, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) &lt;U2 &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;\n",
       "    start                     (ranges) int64 4808454 4828583 ... 89743993\n",
       "    end                       (ranges) int64 4808485 4828648 ... 89744124\n",
       "    range_id                  (ranges) int64 5 9 13 14 ... 408327 408328 408329\n",
       "    num_contexts              (ranges) int64 1 0 0 0 0 1 1 0 ... 0 0 0 0 0 0 0 0\n",
       "    range_length              (ranges) int64 31 65 65 47 47 ... 34 66 34 66 131\n",
       "  * sample_id                 (sample_id) &lt;U8 &#x27;sample_0&#x27;\n",
       "    strand                    (ranges) &lt;U1 &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 126.0 0.0 ... 0.0 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 126.0 nan ... nan nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    Source                    (ranges) object &#x27;HAVANA&#x27; &#x27;HAVANA&#x27; ... &#x27;HAVANA&#x27;\n",
       "    Type                      (ranges) object &#x27;exon&#x27; &#x27;exon&#x27; ... &#x27;exon&#x27; &#x27;exon&#x27;\n",
       "    Score                     (ranges) object &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;\n",
       "    Phase                     (ranges) object &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;\n",
       "    Id                        (ranges) object &#x27;exon:ENSMUST00000027036.10:2&#x27; ...\n",
       "    Parent                    (ranges) object &#x27;ENSMUST00000027036.10&#x27; ... &#x27;EN...\n",
       "    Gene_id                   (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Transcript_id             (ranges) object &#x27;ENSMUST00000027036.10&#x27; ... &#x27;EN...\n",
       "    Gene_type                 (ranges) object &#x27;protein_coding&#x27; ... &#x27;protein_c...\n",
       "    Gene_name                 (ranges) object &#x27;Lypla1&#x27; &#x27;Lypla1&#x27; ... &#x27;Gm21996&#x27;\n",
       "    Transcript_type           (ranges) object &#x27;protein_coding&#x27; ... &#x27;protein_c...\n",
       "    Transcript_name           (ranges) object &#x27;Lypla1-201&#x27; ... &#x27;Gm21996-201&#x27;\n",
       "    Exon_number               (ranges) object &#x27;2&#x27; &#x27;3&#x27; &#x27;2&#x27; &#x27;4&#x27; ... &#x27;4&#x27; &#x27;3&#x27; &#x27;2&#x27;\n",
       "    Exon_id                   (ranges) object &#x27;ENSMUSE00001255376.1&#x27; ... &#x27;ENS...\n",
       "    Level                     (ranges) object &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;\n",
       "    Transcript_support_level  (ranges) object &#x27;1&#x27; &#x27;1&#x27; &#x27;5&#x27; &#x27;1&#x27; ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;\n",
       "    Mgi_id                    (ranges) object &#x27;MGI:1344588&#x27; ... &#x27;MGI:5440224&#x27;\n",
       "    Tag                       (ranges) object &#x27;basic,appris_principal_1,CCDS&#x27;...\n",
       "    Havana_gene               (ranges) object &#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTT...\n",
       "    Havana_transcript         (ranges) object &#x27;OTTMUST00000051162.1&#x27; ... &#x27;OTT...\n",
       "    Protein_id                (ranges) object &#x27;ENSMUSP00000027036.4&#x27; ... &#x27;ENS...\n",
       "    Ccdsid                    (ranges) object &#x27;CCDS14806.1&#x27; &#x27;CCDS14806.1&#x27; ... &#x27;&#x27;\n",
       "    Ont                       (ranges) object &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; ... &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27;</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-4a91a82b-693c-4e52-a11f-9bd7939112bc' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-4a91a82b-693c-4e52-a11f-9bd7939112bc' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>ranges</span>: 180974</li><li><span class='xr-has-index'>sample_id</span>: 1</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-17714fbc-8841-46d7-b65d-c165274b17d6' class='xr-section-summary-in' type='checkbox'  checked><label for='section-17714fbc-8841-46d7-b65d-c165274b17d6' class='xr-section-summary' >Coordinates: <span>(8)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>contig</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U2</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;</div><input id='attrs-81c22890-3ec5-4685-a93b-6bc9ce3bd281' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-81c22890-3ec5-4685-a93b-6bc9ce3bd281' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-2ade6b9d-032f-4a94-b61a-c25d09e14e34' class='xr-var-data-in' type='checkbox'><label for='data-2ade6b9d-032f-4a94-b61a-c25d09e14e34' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;, ..., &#x27;Y&#x27;, &#x27;Y&#x27;, &#x27;Y&#x27;], dtype=&#x27;&lt;U2&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>start</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4808454 4828583 ... 89743993</div><input id='attrs-2726a135-9a1c-4e0d-b7f7-94ea32c2a466' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2726a135-9a1c-4e0d-b7f7-94ea32c2a466' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1e36b060-2ae4-4054-8109-0f3b409dd07f' class='xr-var-data-in' type='checkbox'><label for='data-1e36b060-2ae4-4054-8109-0f3b409dd07f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4808454,  4828583,  4828583, ..., 89731936, 89732605, 89743993])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>end</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4808485 4828648 ... 89744124</div><input id='attrs-cc1c5c01-a0c5-49d5-84ab-6483ddd2796f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-cc1c5c01-a0c5-49d5-84ab-6483ddd2796f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c4f67249-164b-4eb9-9f22-aa7457db792d' class='xr-var-data-in' type='checkbox'><label for='data-c4f67249-164b-4eb9-9f22-aa7457db792d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4808485,  4828648,  4828648, ..., 89731970, 89732671, 89744124])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>5 9 13 14 ... 408327 408328 408329</div><input id='attrs-1407ef16-fe5b-48a2-be83-ba13ce843afa' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1407ef16-fe5b-48a2-be83-ba13ce843afa' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e73c1021-1553-4771-afd0-02927d4e5172' class='xr-var-data-in' type='checkbox'><label for='data-e73c1021-1553-4771-afd0-02927d4e5172' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([     5,      9,     13, ..., 408327, 408328, 408329])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_contexts</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>1 0 0 0 0 1 1 0 ... 0 0 0 0 0 0 0 0</div><input id='attrs-3385bbbd-2a33-4176-9a59-27c4fee2ac24' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3385bbbd-2a33-4176-9a59-27c4fee2ac24' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-bc174389-3688-4e74-b716-874dfdd5396b' class='xr-var-data-in' type='checkbox'><label for='data-bc174389-3688-4e74-b716-874dfdd5396b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([1, 0, 0, ..., 0, 0, 0])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_length</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>31 65 65 47 47 ... 34 66 34 66 131</div><input id='attrs-bb4832a0-caeb-4f92-a5ae-0d0e64b41d4e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-bb4832a0-caeb-4f92-a5ae-0d0e64b41d4e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c73607f5-e2b1-427f-86ca-ceb8a267f357' class='xr-var-data-in' type='checkbox'><label for='data-c73607f5-e2b1-427f-86ca-ceb8a267f357' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 31,  65,  65, ...,  34,  66, 131])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample_id</span></div><div class='xr-var-dims'>(sample_id)</div><div class='xr-var-dtype'>&lt;U8</div><div class='xr-var-preview xr-preview'>&#x27;sample_0&#x27;</div><input id='attrs-a95fed14-4817-4a16-9c38-d9382f06429a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-a95fed14-4817-4a16-9c38-d9382f06429a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d3fd2d15-671d-42fe-a5e9-3e32e9b88294' class='xr-var-data-in' type='checkbox'><label for='data-d3fd2d15-671d-42fe-a5e9-3e32e9b88294' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;sample_0&#x27;], dtype=&#x27;&lt;U8&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U1</div><div class='xr-var-preview xr-preview'>&#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;</div><input id='attrs-c6e35789-0dff-46b7-bc1d-b5b7cd494c4c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c6e35789-0dff-46b7-bc1d-b5b7cd494c4c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d8e02bb6-0d38-4ce5-9ad4-cca9ed78400b' class='xr-var-data-in' type='checkbox'><label for='data-d8e02bb6-0d38-4ce5-9ad4-cca9ed78400b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;+&#x27;, &#x27;+&#x27;, &#x27;+&#x27;, ..., &#x27;-&#x27;, &#x27;-&#x27;, &#x27;-&#x27;], dtype=&#x27;&lt;U1&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-365d6f6a-08a7-4c33-a3e2-955cb8d505e5' class='xr-section-summary-in' type='checkbox'  ><label for='section-365d6f6a-08a7-4c33-a3e2-955cb8d505e5' class='xr-section-summary' >Data variables: <span>(35)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0</div><input id='attrs-deae9234-12df-40dd-afba-fb7d87c40ed4' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-deae9234-12df-40dd-afba-fb7d87c40ed4' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-03c894a3-556b-4af9-bb68-a2a3fb87a584' class='xr-var-data-in' type='checkbox'><label for='data-03c894a3-556b-4af9-bb68-a2a3fb87a584' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 nan nan nan ... nan nan nan nan</div><input id='attrs-4aaadcc3-5307-4a76-8bcf-d0046ee1c94f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4aaadcc3-5307-4a76-8bcf-d0046ee1c94f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5328e847-e229-4da5-ac11-0addfa5e32ec' class='xr-var-data-in' type='checkbox'><label for='data-5328e847-e229-4da5-ac11-0addfa5e32ec' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [nan],\n",
       "       [nan],\n",
       "       ...,\n",
       "       [nan],\n",
       "       [nan],\n",
       "       [nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0</div><input id='attrs-5ce2aacf-b0d2-4738-97d3-dc9e0d709018' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5ce2aacf-b0d2-4738-97d3-dc9e0d709018' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-900434d4-6c2d-442a-a99e-6b3dd5fb95c0' class='xr-var-data-in' type='checkbox'><label for='data-900434d4-6c2d-442a-a99e-6b3dd5fb95c0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0</div><input id='attrs-839fd05a-ec8e-4779-9897-1eb8168e3627' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-839fd05a-ec8e-4779-9897-1eb8168e3627' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-096a6563-7b85-4eaa-8723-41d190647418' class='xr-var-data-in' type='checkbox'><label for='data-096a6563-7b85-4eaa-8723-41d190647418' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 nan nan nan ... nan nan nan nan</div><input id='attrs-771bbe78-5904-409f-8d20-e84903fdcfdb' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-771bbe78-5904-409f-8d20-e84903fdcfdb' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-72381ae5-da89-46d0-b583-07761968dcf4' class='xr-var-data-in' type='checkbox'><label for='data-72381ae5-da89-46d0-b583-07761968dcf4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [nan],\n",
       "       [nan],\n",
       "       ...,\n",
       "       [nan],\n",
       "       [nan],\n",
       "       [nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0</div><input id='attrs-768f2e21-849a-4cd2-8fcf-1d1e076d16a8' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-768f2e21-849a-4cd2-8fcf-1d1e076d16a8' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1f7e4c7d-d6c1-4784-8d6d-0503f012eaf4' class='xr-var-data-in' type='checkbox'><label for='data-1f7e4c7d-d6c1-4784-8d6d-0503f012eaf4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0</div><input id='attrs-7d9a04d9-358b-49b4-af6b-fd1a2c1352a3' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7d9a04d9-358b-49b4-af6b-fd1a2c1352a3' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7bd2624d-d894-467a-953d-ea3f8f3e1c93' class='xr-var-data-in' type='checkbox'><label for='data-7bd2624d-d894-467a-953d-ea3f8f3e1c93' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 nan nan nan ... nan nan nan nan</div><input id='attrs-48b4bace-0090-47ba-b857-4cd1786c1f26' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-48b4bace-0090-47ba-b857-4cd1786c1f26' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-694d6581-b721-4354-b323-70b7f2ad322e' class='xr-var-data-in' type='checkbox'><label for='data-694d6581-b721-4354-b323-70b7f2ad322e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [nan],\n",
       "       [nan],\n",
       "       ...,\n",
       "       [nan],\n",
       "       [nan],\n",
       "       [nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0</div><input id='attrs-9b76ca6c-d35b-4ff2-bb3a-b89dea9229a2' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-9b76ca6c-d35b-4ff2-bb3a-b89dea9229a2' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9aba1bc4-8c89-4d76-94ad-0e48925efcc1' class='xr-var-data-in' type='checkbox'><label for='data-9aba1bc4-8c89-4d76-94ad-0e48925efcc1' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>126.0 0.0 0.0 0.0 ... 0.0 0.0 0.0</div><input id='attrs-95bba6a3-90b9-4038-a370-87cc01f12097' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-95bba6a3-90b9-4038-a370-87cc01f12097' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8c984c4c-7a61-48c2-a489-e238e14760de' class='xr-var-data-in' type='checkbox'><label for='data-8c984c4c-7a61-48c2-a489-e238e14760de' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[126.],\n",
       "       [  0.],\n",
       "       [  0.],\n",
       "       ...,\n",
       "       [  0.],\n",
       "       [  0.],\n",
       "       [  0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>126.0 nan nan nan ... nan nan nan</div><input id='attrs-e7e77f22-fb75-454b-af53-2208bdf3cf9e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e7e77f22-fb75-454b-af53-2208bdf3cf9e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-61a04df9-6ae6-4197-8338-a01301993ae7' class='xr-var-data-in' type='checkbox'><label for='data-61a04df9-6ae6-4197-8338-a01301993ae7' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[126.],\n",
       "       [ nan],\n",
       "       [ nan],\n",
       "       ...,\n",
       "       [ nan],\n",
       "       [ nan],\n",
       "       [ nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0</div><input id='attrs-f32eb122-a2aa-4c47-aca5-a040573fff05' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f32eb122-a2aa-4c47-aca5-a040573fff05' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-45eca842-899b-4d7d-92be-ac405cbb0e4c' class='xr-var-data-in' type='checkbox'><label for='data-45eca842-899b-4d7d-92be-ac405cbb0e4c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Source</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;HAVANA&#x27; &#x27;HAVANA&#x27; ... &#x27;HAVANA&#x27;</div><input id='attrs-7b65eef9-101e-4ab1-bb80-4b571381e0e7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7b65eef9-101e-4ab1-bb80-4b571381e0e7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b34bb63e-5442-41d4-b20d-099410816205' class='xr-var-data-in' type='checkbox'><label for='data-b34bb63e-5442-41d4-b20d-099410816205' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;, ..., &#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;exon&#x27; &#x27;exon&#x27; ... &#x27;exon&#x27; &#x27;exon&#x27;</div><input id='attrs-1a44c896-e9f2-40db-b041-24fdadf11f99' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1a44c896-e9f2-40db-b041-24fdadf11f99' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-de166e9a-eb4d-4462-9c59-05547c3bc4b6' class='xr-var-data-in' type='checkbox'><label for='data-de166e9a-eb4d-4462-9c59-05547c3bc4b6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;exon&#x27;, &#x27;exon&#x27;, &#x27;exon&#x27;, ..., &#x27;exon&#x27;, &#x27;exon&#x27;, &#x27;exon&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Score</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;</div><input id='attrs-edca008d-497c-408e-9aa1-acb2079b63dc' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-edca008d-497c-408e-9aa1-acb2079b63dc' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-39e8b047-6e9f-472b-b570-f56bc206db2f' class='xr-var-data-in' type='checkbox'><label for='data-39e8b047-6e9f-472b-b570-f56bc206db2f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;, ..., &#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Phase</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;</div><input id='attrs-9824b320-e5c2-4c5b-ae3b-1624446d1ccb' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-9824b320-e5c2-4c5b-ae3b-1624446d1ccb' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-939ee04b-1789-4c25-83ac-19696ac756ca' class='xr-var-data-in' type='checkbox'><label for='data-939ee04b-1789-4c25-83ac-19696ac756ca' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;, ..., &#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;exon:ENSMUST00000027036.10:2&#x27; ....</div><input id='attrs-50f9fca2-bea4-42c6-a0fb-f024df501331' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-50f9fca2-bea4-42c6-a0fb-f024df501331' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b24a59a4-061a-4e83-bad6-ed2702340c75' class='xr-var-data-in' type='checkbox'><label for='data-b24a59a4-061a-4e83-bad6-ed2702340c75' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;exon:ENSMUST00000027036.10:2&#x27;, &#x27;exon:ENSMUST00000027036.10:3&#x27;,\n",
       "       &#x27;exon:ENSMUST00000155020.1:2&#x27;, ..., &#x27;exon:ENSMUST00000188269.1:4&#x27;,\n",
       "       &#x27;exon:ENSMUST00000188269.1:3&#x27;, &#x27;exon:ENSMUST00000188269.1:2&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Parent</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUST00000027036.10&#x27; ... &#x27;ENS...</div><input id='attrs-d17fd2ce-cce0-4ea1-bb64-43d0bbffb011' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d17fd2ce-cce0-4ea1-bb64-43d0bbffb011' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-cf46d46c-b06a-4b78-a75b-81c5789ad8e4' class='xr-var-data-in' type='checkbox'><label for='data-cf46d46c-b06a-4b78-a75b-81c5789ad8e4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUST00000027036.10&#x27;, &#x27;ENSMUST00000027036.10&#x27;,\n",
       "       &#x27;ENSMUST00000155020.1&#x27;, ..., &#x27;ENSMUST00000188269.1&#x27;,\n",
       "       &#x27;ENSMUST00000188269.1&#x27;, &#x27;ENSMUST00000188269.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-ee083739-dcbb-405a-8dbe-afe00f3b3c0b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ee083739-dcbb-405a-8dbe-afe00f3b3c0b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0602e20e-4b49-4e71-8942-3c9e113eb5f3' class='xr-var-data-in' type='checkbox'><label for='data-0602e20e-4b49-4e71-8942-3c9e113eb5f3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000025903.14&#x27;,\n",
       "       &#x27;ENSMUSG00000104217.1&#x27;, ..., &#x27;ENSMUSG00000100608.1&#x27;,\n",
       "       &#x27;ENSMUSG00000100608.1&#x27;, &#x27;ENSMUSG00000100608.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUST00000027036.10&#x27; ... &#x27;ENS...</div><input id='attrs-ad9c5f8e-d9da-49c5-9dc1-4177b069203e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ad9c5f8e-d9da-49c5-9dc1-4177b069203e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-eaafd4e1-7cf6-44f0-936b-96fe02999f0e' class='xr-var-data-in' type='checkbox'><label for='data-eaafd4e1-7cf6-44f0-936b-96fe02999f0e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUST00000027036.10&#x27;, &#x27;ENSMUST00000027036.10&#x27;,\n",
       "       &#x27;ENSMUST00000155020.1&#x27;, ..., &#x27;ENSMUST00000188269.1&#x27;,\n",
       "       &#x27;ENSMUST00000188269.1&#x27;, &#x27;ENSMUST00000188269.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;protein_coding&#x27; ... &#x27;protein_co...</div><input id='attrs-21243294-036c-426d-b270-158533ead681' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-21243294-036c-426d-b270-158533ead681' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7fcb1be0-7af8-4edc-80b4-e9942132fb63' class='xr-var-data-in' type='checkbox'><label for='data-7fcb1be0-7af8-4edc-80b4-e9942132fb63' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, ...,\n",
       "       &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Lypla1&#x27; &#x27;Lypla1&#x27; ... &#x27;Gm21996&#x27;</div><input id='attrs-01d68ac2-7a2a-46d8-9f62-95d1bbe49026' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-01d68ac2-7a2a-46d8-9f62-95d1bbe49026' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-04aa35ee-25de-4715-9e30-d5bd8bc4e9e1' class='xr-var-data-in' type='checkbox'><label for='data-04aa35ee-25de-4715-9e30-d5bd8bc4e9e1' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Lypla1&#x27;, &#x27;Lypla1&#x27;, &#x27;Gm37988&#x27;, ..., &#x27;Gm21996&#x27;, &#x27;Gm21996&#x27;,\n",
       "       &#x27;Gm21996&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;protein_coding&#x27; ... &#x27;protein_co...</div><input id='attrs-0b1d06ee-c82b-436d-94b8-2adac885eca8' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0b1d06ee-c82b-436d-94b8-2adac885eca8' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-67c8bcb8-bcf9-45e7-9ea8-61b8cdb824bb' class='xr-var-data-in' type='checkbox'><label for='data-67c8bcb8-bcf9-45e7-9ea8-61b8cdb824bb' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;nonsense_mediated_decay&#x27;, ...,\n",
       "       &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Lypla1-201&#x27; ... &#x27;Gm21996-201&#x27;</div><input id='attrs-05fc5744-059c-44ce-aa21-ca1a54e5f41e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-05fc5744-059c-44ce-aa21-ca1a54e5f41e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-fd8c4c9b-ab21-4f88-8025-bc7ce6e956db' class='xr-var-data-in' type='checkbox'><label for='data-fd8c4c9b-ab21-4f88-8025-bc7ce6e956db' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Lypla1-201&#x27;, &#x27;Lypla1-201&#x27;, &#x27;Gm37988-201&#x27;, ..., &#x27;Gm21996-201&#x27;,\n",
       "       &#x27;Gm21996-201&#x27;, &#x27;Gm21996-201&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;2&#x27; &#x27;3&#x27; &#x27;2&#x27; &#x27;4&#x27; ... &#x27;5&#x27; &#x27;4&#x27; &#x27;3&#x27; &#x27;2&#x27;</div><input id='attrs-1e15356f-51da-4a0d-8e1c-04262e384b3a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1e15356f-51da-4a0d-8e1c-04262e384b3a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1c3a6ed1-92eb-4fca-ab37-6a3c8365fd17' class='xr-var-data-in' type='checkbox'><label for='data-1c3a6ed1-92eb-4fca-ab37-6a3c8365fd17' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;2&#x27;, &#x27;3&#x27;, &#x27;2&#x27;, ..., &#x27;4&#x27;, &#x27;3&#x27;, &#x27;2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSE00001255376.1&#x27; ... &#x27;ENSM...</div><input id='attrs-4bea25d1-ff5a-4f15-9515-a15af6dc46be' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4bea25d1-ff5a-4f15-9515-a15af6dc46be' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-46dcd133-da80-48c0-bf64-1ba9af4a3aee' class='xr-var-data-in' type='checkbox'><label for='data-46dcd133-da80-48c0-bf64-1ba9af4a3aee' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSE00001255376.1&#x27;, &#x27;ENSMUSE00001222304.1&#x27;,\n",
       "       &#x27;ENSMUSE00001241075.1&#x27;, ..., &#x27;ENSMUSE00001333847.1&#x27;,\n",
       "       &#x27;ENSMUSE00001331494.1&#x27;, &#x27;ENSMUSE00001332880.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;</div><input id='attrs-971d62d8-aac9-4767-ad2d-b4777dd65812' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-971d62d8-aac9-4767-ad2d-b4777dd65812' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-6b9aff23-da3f-481f-a5a2-5517c94055a4' class='xr-var-data-in' type='checkbox'><label for='data-6b9aff23-da3f-481f-a5a2-5517c94055a4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;, ..., &#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_support_level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;5&#x27; &#x27;1&#x27; ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;</div><input id='attrs-75ec20b6-1f67-4020-93bb-dde0dbcb86b7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-75ec20b6-1f67-4020-93bb-dde0dbcb86b7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-95d5c213-9629-4a07-8f36-1d0499d0b255' class='xr-var-data-in' type='checkbox'><label for='data-95d5c213-9629-4a07-8f36-1d0499d0b255' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;5&#x27;, ..., &#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Mgi_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;MGI:1344588&#x27; ... &#x27;MGI:5440224&#x27;</div><input id='attrs-60d032ff-61aa-411c-975b-0342b3189af5' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-60d032ff-61aa-411c-975b-0342b3189af5' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d59a346c-82b3-4c6a-98c1-ddca2ba06eec' class='xr-var-data-in' type='checkbox'><label for='data-d59a346c-82b3-4c6a-98c1-ddca2ba06eec' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;MGI:1344588&#x27;, &#x27;MGI:1344588&#x27;, &#x27;MGI:5611216&#x27;, ..., &#x27;MGI:5440224&#x27;,\n",
       "       &#x27;MGI:5440224&#x27;, &#x27;MGI:5440224&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Tag</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;basic,appris_principal_1,CCDS&#x27; ...</div><input id='attrs-5126c88d-c103-4fc9-819f-dcb19d81aa85' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5126c88d-c103-4fc9-819f-dcb19d81aa85' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e1b86ac3-5e4c-4bf8-8432-1697a08a1bde' class='xr-var-data-in' type='checkbox'><label for='data-e1b86ac3-5e4c-4bf8-8432-1697a08a1bde' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;basic,appris_principal_1,CCDS&#x27;, &#x27;basic,appris_principal_1,CCDS&#x27;,\n",
       "       &#x27;basic,appris_principal_1,readthrough_transcript&#x27;, ...,\n",
       "       &#x27;basic,appris_principal_1&#x27;, &#x27;basic,appris_principal_1&#x27;,\n",
       "       &#x27;basic,appris_principal_1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_gene</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTTM...</div><input id='attrs-d2ec7c61-a8ff-4aed-b5f7-18f05702a1f0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d2ec7c61-a8ff-4aed-b5f7-18f05702a1f0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b6f555b4-525c-45ab-a2c6-95291f1315f9' class='xr-var-data-in' type='checkbox'><label for='data-b6f555b4-525c-45ab-a2c6-95291f1315f9' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;OTTMUSG00000021562.4&#x27;, &#x27;OTTMUSG00000021562.4&#x27;,\n",
       "       &#x27;OTTMUSG00000050100.1&#x27;, ..., &#x27;OTTMUSG00000047316.1&#x27;,\n",
       "       &#x27;OTTMUSG00000047316.1&#x27;, &#x27;OTTMUSG00000047316.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_transcript</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;OTTMUST00000051162.1&#x27; ... &#x27;OTTM...</div><input id='attrs-7b64ab93-2a99-4fd3-96b2-40e9d5494b5c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7b64ab93-2a99-4fd3-96b2-40e9d5494b5c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c3842657-001a-428e-b612-7d0b88198b6f' class='xr-var-data-in' type='checkbox'><label for='data-c3842657-001a-428e-b612-7d0b88198b6f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;OTTMUST00000051162.1&#x27;, &#x27;OTTMUST00000051162.1&#x27;,\n",
       "       &#x27;OTTMUST00000127419.1&#x27;, ..., &#x27;OTTMUST00000122318.1&#x27;,\n",
       "       &#x27;OTTMUST00000122318.1&#x27;, &#x27;OTTMUST00000122318.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Protein_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSP00000027036.4&#x27; ... &#x27;ENSM...</div><input id='attrs-54ad5e87-29ca-4a32-95e8-acbc31724be3' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-54ad5e87-29ca-4a32-95e8-acbc31724be3' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1714b99c-1619-4439-b8da-ca8fba51422d' class='xr-var-data-in' type='checkbox'><label for='data-1714b99c-1619-4439-b8da-ca8fba51422d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSP00000027036.4&#x27;, &#x27;ENSMUSP00000027036.4&#x27;,\n",
       "       &#x27;ENSMUSP00000136108.1&#x27;, ..., &#x27;ENSMUSP00000141173.1&#x27;,\n",
       "       &#x27;ENSMUSP00000141173.1&#x27;, &#x27;ENSMUSP00000141173.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ccdsid</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;CCDS14806.1&#x27; &#x27;CCDS14806.1&#x27; ... &#x27;&#x27;</div><input id='attrs-0ab2dca8-e5e4-4755-b6c5-a4570173fa95' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0ab2dca8-e5e4-4755-b6c5-a4570173fa95' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-54238803-d32a-4e9c-974c-51370c01e779' class='xr-var-data-in' type='checkbox'><label for='data-54238803-d32a-4e9c-974c-51370c01e779' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;CCDS14806.1&#x27;, &#x27;CCDS14806.1&#x27;, &#x27;&#x27;, ..., &#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ont</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; ... &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27;</div><input id='attrs-c85bda0d-b1af-42fc-b40c-0eab0835a6d8' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c85bda0d-b1af-42fc-b40c-0eab0835a6d8' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-64b5b762-f95b-466a-9c1b-af7efe222ffd' class='xr-var-data-in' type='checkbox'><label for='data-64b5b762-f95b-466a-9c1b-af7efe222ffd' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;, ..., &#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;], dtype=object)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-bdf355c3-a32b-4518-825d-ad5684263666' class='xr-section-summary-in' type='checkbox'  ><label for='section-bdf355c3-a32b-4518-825d-ad5684263666' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>sample_id</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-02eccd7a-da4b-4de5-9f21-5ae5013f0604' class='xr-index-data-in' type='checkbox'/><label for='index-02eccd7a-da4b-4de5-9f21-5ae5013f0604' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;sample_0&#x27;], dtype=&#x27;object&#x27;, name=&#x27;sample_id&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-226b58a8-bd79-4b6f-bd7a-aad9589946d3' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-226b58a8-bd79-4b6f-bd7a-aad9589946d3' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.Dataset>\n",
       "Dimensions:                   (ranges: 180974, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) <U2 '1' '1' '1' '1' ... 'Y' 'Y' 'Y' 'Y'\n",
       "    start                     (ranges) int64 4808454 4828583 ... 89743993\n",
       "    end                       (ranges) int64 4808485 4828648 ... 89744124\n",
       "    range_id                  (ranges) int64 5 9 13 14 ... 408327 408328 408329\n",
       "    num_contexts              (ranges) int64 1 0 0 0 0 1 1 0 ... 0 0 0 0 0 0 0 0\n",
       "    range_length              (ranges) int64 31 65 65 47 47 ... 34 66 34 66 131\n",
       "  * sample_id                 (sample_id) <U8 'sample_0'\n",
       "    strand                    (ranges) <U1 '+' '+' '+' '+' ... '-' '-' '-' '-'\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 126.0 0.0 ... 0.0 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 126.0 nan ... nan nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    Source                    (ranges) object 'HAVANA' 'HAVANA' ... 'HAVANA'\n",
       "    Type                      (ranges) object 'exon' 'exon' ... 'exon' 'exon'\n",
       "    Score                     (ranges) object '.' '.' '.' '.' ... '.' '.' '.'\n",
       "    Phase                     (ranges) object '.' '.' '.' '.' ... '.' '.' '.'\n",
       "    Id                        (ranges) object 'exon:ENSMUST00000027036.10:2' ...\n",
       "    Parent                    (ranges) object 'ENSMUST00000027036.10' ... 'EN...\n",
       "    Gene_id                   (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Transcript_id             (ranges) object 'ENSMUST00000027036.10' ... 'EN...\n",
       "    Gene_type                 (ranges) object 'protein_coding' ... 'protein_c...\n",
       "    Gene_name                 (ranges) object 'Lypla1' 'Lypla1' ... 'Gm21996'\n",
       "    Transcript_type           (ranges) object 'protein_coding' ... 'protein_c...\n",
       "    Transcript_name           (ranges) object 'Lypla1-201' ... 'Gm21996-201'\n",
       "    Exon_number               (ranges) object '2' '3' '2' '4' ... '4' '3' '2'\n",
       "    Exon_id                   (ranges) object 'ENSMUSE00001255376.1' ... 'ENS...\n",
       "    Level                     (ranges) object '2' '2' '2' '2' ... '2' '2' '2'\n",
       "    Transcript_support_level  (ranges) object '1' '1' '5' '1' ... '1' '1' '1'\n",
       "    Mgi_id                    (ranges) object 'MGI:1344588' ... 'MGI:5440224'\n",
       "    Tag                       (ranges) object 'basic,appris_principal_1,CCDS'...\n",
       "    Havana_gene               (ranges) object 'OTTMUSG00000021562.4' ... 'OTT...\n",
       "    Havana_transcript         (ranges) object 'OTTMUST00000051162.1' ... 'OTT...\n",
       "    Protein_id                (ranges) object 'ENSMUSP00000027036.4' ... 'ENS...\n",
       "    Ccdsid                    (ranges) object 'CCDS14806.1' 'CCDS14806.1' ... ''\n",
       "    Ont                       (ranges) object '' '' '' '' '' ... '' '' '' '' ''"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "antisense_methylation['exons']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;\n",
       "Dimensions:                   (ranges: 180974, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) &lt;U2 &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;\n",
       "    start                     (ranges) int64 4808454 4828583 ... 89743993\n",
       "    end                       (ranges) int64 4808485 4828648 ... 89744124\n",
       "    range_id                  (ranges) int64 5 9 13 14 ... 408327 408328 408329\n",
       "    num_contexts              (ranges) int64 1 0 0 0 0 2 2 0 ... 0 0 0 0 1 0 0 0\n",
       "    range_length              (ranges) int64 31 65 65 47 47 ... 34 66 34 66 131\n",
       "  * sample_id                 (sample_id) &lt;U8 &#x27;sample_0&#x27;\n",
       "    strand                    (ranges) &lt;U1 &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 113.0 0.0 ... 0.0 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 113.0 nan ... nan nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    Source                    (ranges) object &#x27;HAVANA&#x27; &#x27;HAVANA&#x27; ... &#x27;HAVANA&#x27;\n",
       "    Type                      (ranges) object &#x27;exon&#x27; &#x27;exon&#x27; ... &#x27;exon&#x27; &#x27;exon&#x27;\n",
       "    Score                     (ranges) object &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;\n",
       "    Phase                     (ranges) object &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;\n",
       "    Id                        (ranges) object &#x27;exon:ENSMUST00000027036.10:2&#x27; ...\n",
       "    Parent                    (ranges) object &#x27;ENSMUST00000027036.10&#x27; ... &#x27;EN...\n",
       "    Gene_id                   (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Transcript_id             (ranges) object &#x27;ENSMUST00000027036.10&#x27; ... &#x27;EN...\n",
       "    Gene_type                 (ranges) object &#x27;protein_coding&#x27; ... &#x27;protein_c...\n",
       "    Gene_name                 (ranges) object &#x27;Lypla1&#x27; &#x27;Lypla1&#x27; ... &#x27;Gm21996&#x27;\n",
       "    Transcript_type           (ranges) object &#x27;protein_coding&#x27; ... &#x27;protein_c...\n",
       "    Transcript_name           (ranges) object &#x27;Lypla1-201&#x27; ... &#x27;Gm21996-201&#x27;\n",
       "    Exon_number               (ranges) object &#x27;2&#x27; &#x27;3&#x27; &#x27;2&#x27; &#x27;4&#x27; ... &#x27;4&#x27; &#x27;3&#x27; &#x27;2&#x27;\n",
       "    Exon_id                   (ranges) object &#x27;ENSMUSE00001255376.1&#x27; ... &#x27;ENS...\n",
       "    Level                     (ranges) object &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;\n",
       "    Transcript_support_level  (ranges) object &#x27;1&#x27; &#x27;1&#x27; &#x27;5&#x27; &#x27;1&#x27; ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;\n",
       "    Mgi_id                    (ranges) object &#x27;MGI:1344588&#x27; ... &#x27;MGI:5440224&#x27;\n",
       "    Tag                       (ranges) object &#x27;basic,appris_principal_1,CCDS&#x27;...\n",
       "    Havana_gene               (ranges) object &#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTT...\n",
       "    Havana_transcript         (ranges) object &#x27;OTTMUST00000051162.1&#x27; ... &#x27;OTT...\n",
       "    Protein_id                (ranges) object &#x27;ENSMUSP00000027036.4&#x27; ... &#x27;ENS...\n",
       "    Ccdsid                    (ranges) object &#x27;CCDS14806.1&#x27; &#x27;CCDS14806.1&#x27; ... &#x27;&#x27;\n",
       "    Ont                       (ranges) object &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; ... &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27;</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-ff0ecc66-9b6e-4f6b-bfa7-0b3b86aecf44' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-ff0ecc66-9b6e-4f6b-bfa7-0b3b86aecf44' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>ranges</span>: 180974</li><li><span class='xr-has-index'>sample_id</span>: 1</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-711241ae-70ce-43ec-bb49-bd60dd731d71' class='xr-section-summary-in' type='checkbox'  checked><label for='section-711241ae-70ce-43ec-bb49-bd60dd731d71' class='xr-section-summary' >Coordinates: <span>(8)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>contig</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U2</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;</div><input id='attrs-29e0540b-0da3-4d4b-b443-fde3a4742c7b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-29e0540b-0da3-4d4b-b443-fde3a4742c7b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-66ea0484-7e72-472b-83df-981fcf4257cf' class='xr-var-data-in' type='checkbox'><label for='data-66ea0484-7e72-472b-83df-981fcf4257cf' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;, ..., &#x27;Y&#x27;, &#x27;Y&#x27;, &#x27;Y&#x27;], dtype=&#x27;&lt;U2&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>start</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4808454 4828583 ... 89743993</div><input id='attrs-ef619465-290c-4ace-a98a-4938488633f2' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ef619465-290c-4ace-a98a-4938488633f2' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b8a05710-a0b4-448f-b02b-16afc8d0c6a0' class='xr-var-data-in' type='checkbox'><label for='data-b8a05710-a0b4-448f-b02b-16afc8d0c6a0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4808454,  4828583,  4828583, ..., 89731936, 89732605, 89743993])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>end</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4808485 4828648 ... 89744124</div><input id='attrs-76c6f7cf-2dd4-44e3-9c8f-1e9e38a0b94a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-76c6f7cf-2dd4-44e3-9c8f-1e9e38a0b94a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1c485d32-17a6-4830-b738-db26464366df' class='xr-var-data-in' type='checkbox'><label for='data-1c485d32-17a6-4830-b738-db26464366df' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4808485,  4828648,  4828648, ..., 89731970, 89732671, 89744124])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>5 9 13 14 ... 408327 408328 408329</div><input id='attrs-b1ad56c0-3a3d-4427-abec-8273212f1255' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b1ad56c0-3a3d-4427-abec-8273212f1255' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-6b8696fb-87b0-455b-9826-d24183f9b2da' class='xr-var-data-in' type='checkbox'><label for='data-6b8696fb-87b0-455b-9826-d24183f9b2da' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([     5,      9,     13, ..., 408327, 408328, 408329])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_contexts</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>1 0 0 0 0 2 2 0 ... 0 0 0 0 1 0 0 0</div><input id='attrs-e7ca2fc9-810f-4571-a7ec-e0fa652d5dc0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e7ca2fc9-810f-4571-a7ec-e0fa652d5dc0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-298ca3d1-d5be-4b47-815d-c63bd38fd234' class='xr-var-data-in' type='checkbox'><label for='data-298ca3d1-d5be-4b47-815d-c63bd38fd234' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([1, 0, 0, ..., 0, 0, 0])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_length</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>31 65 65 47 47 ... 34 66 34 66 131</div><input id='attrs-42399218-ffd8-4ac0-8527-d5fbe9d22cd1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-42399218-ffd8-4ac0-8527-d5fbe9d22cd1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8aaaddae-555a-4267-bffe-62d30cdef440' class='xr-var-data-in' type='checkbox'><label for='data-8aaaddae-555a-4267-bffe-62d30cdef440' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 31,  65,  65, ...,  34,  66, 131])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample_id</span></div><div class='xr-var-dims'>(sample_id)</div><div class='xr-var-dtype'>&lt;U8</div><div class='xr-var-preview xr-preview'>&#x27;sample_0&#x27;</div><input id='attrs-00d4b461-6918-4406-b1d9-eebfdb75a300' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-00d4b461-6918-4406-b1d9-eebfdb75a300' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4a0b5032-6d34-4b24-b324-1e699b565746' class='xr-var-data-in' type='checkbox'><label for='data-4a0b5032-6d34-4b24-b324-1e699b565746' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;sample_0&#x27;], dtype=&#x27;&lt;U8&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U1</div><div class='xr-var-preview xr-preview'>&#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;</div><input id='attrs-b62fe158-dc67-464e-9c1b-c4a2dce5e749' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b62fe158-dc67-464e-9c1b-c4a2dce5e749' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d78c529e-e547-4b39-b5cf-65d96d375d9b' class='xr-var-data-in' type='checkbox'><label for='data-d78c529e-e547-4b39-b5cf-65d96d375d9b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;+&#x27;, &#x27;+&#x27;, &#x27;+&#x27;, ..., &#x27;-&#x27;, &#x27;-&#x27;, &#x27;-&#x27;], dtype=&#x27;&lt;U1&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-b92cc8ec-d654-472d-84c9-1c4b4056e671' class='xr-section-summary-in' type='checkbox'  ><label for='section-b92cc8ec-d654-472d-84c9-1c4b4056e671' class='xr-section-summary' >Data variables: <span>(35)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0</div><input id='attrs-8c65985e-97ae-4b30-af93-b46664ee0446' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8c65985e-97ae-4b30-af93-b46664ee0446' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-cf5cb7ec-793d-41f6-8b80-0c0470fc1b02' class='xr-var-data-in' type='checkbox'><label for='data-cf5cb7ec-793d-41f6-8b80-0c0470fc1b02' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 nan nan nan ... nan nan nan</div><input id='attrs-babf7712-6031-4fe9-afc8-582273b3f2a3' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-babf7712-6031-4fe9-afc8-582273b3f2a3' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e0cf8956-db1a-48e2-867b-f18eb948cdf7' class='xr-var-data-in' type='checkbox'><label for='data-e0cf8956-db1a-48e2-867b-f18eb948cdf7' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [nan],\n",
       "       [nan],\n",
       "       ...,\n",
       "       [nan],\n",
       "       [nan],\n",
       "       [nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 1.0 0.0 0.0 0.0</div><input id='attrs-31de348d-e94e-4530-8b6d-23d6f9fe3e4f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-31de348d-e94e-4530-8b6d-23d6f9fe3e4f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9cbac08c-c9dd-4a96-90f3-a7ca65a17fed' class='xr-var-data-in' type='checkbox'><label for='data-9cbac08c-c9dd-4a96-90f3-a7ca65a17fed' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0</div><input id='attrs-4229eea5-55a6-4bc7-9c11-fb95ae1281b0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4229eea5-55a6-4bc7-9c11-fb95ae1281b0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0e1a5813-10f0-4977-98fc-8df477405e0c' class='xr-var-data-in' type='checkbox'><label for='data-0e1a5813-10f0-4977-98fc-8df477405e0c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 nan nan nan ... 0.0 nan nan nan</div><input id='attrs-f045a8ab-7666-495d-8988-542d8cae94e1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f045a8ab-7666-495d-8988-542d8cae94e1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-98413805-bea8-4986-9d76-f2064e47a861' class='xr-var-data-in' type='checkbox'><label for='data-98413805-bea8-4986-9d76-f2064e47a861' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [nan],\n",
       "       [nan],\n",
       "       ...,\n",
       "       [nan],\n",
       "       [nan],\n",
       "       [nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 1.0 0.0 0.0 0.0</div><input id='attrs-8524e164-b769-4c7b-9dc2-6725faf1b0df' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8524e164-b769-4c7b-9dc2-6725faf1b0df' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-09749a79-28c5-42c2-936a-f0d98090271d' class='xr-var-data-in' type='checkbox'><label for='data-09749a79-28c5-42c2-936a-f0d98090271d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0</div><input id='attrs-b000501f-33b5-4ba7-936d-806660a58883' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b000501f-33b5-4ba7-936d-806660a58883' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-09758059-dd13-4863-9394-f7ca1f214bb4' class='xr-var-data-in' type='checkbox'><label for='data-09758059-dd13-4863-9394-f7ca1f214bb4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 nan nan nan ... nan nan nan</div><input id='attrs-c9d1aefa-4541-4a50-9339-d84242b567cb' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c9d1aefa-4541-4a50-9339-d84242b567cb' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-6c2d5b7c-c06e-4a4d-ae8e-22242b653988' class='xr-var-data-in' type='checkbox'><label for='data-6c2d5b7c-c06e-4a4d-ae8e-22242b653988' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [nan],\n",
       "       [nan],\n",
       "       ...,\n",
       "       [nan],\n",
       "       [nan],\n",
       "       [nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 1.0 0.0 0.0 0.0</div><input id='attrs-653c8a4d-ce98-4489-8649-de75959ab1c0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-653c8a4d-ce98-4489-8649-de75959ab1c0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d12e483d-e566-4845-83bd-c0cb18f6b42d' class='xr-var-data-in' type='checkbox'><label for='data-d12e483d-e566-4845-83bd-c0cb18f6b42d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>113.0 0.0 0.0 0.0 ... 0.0 0.0 0.0</div><input id='attrs-1a8c044b-fc8a-47a3-926b-745e6595273c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1a8c044b-fc8a-47a3-926b-745e6595273c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5667b687-2009-40e0-adb4-a570e71f214e' class='xr-var-data-in' type='checkbox'><label for='data-5667b687-2009-40e0-adb4-a570e71f214e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[113.],\n",
       "       [  0.],\n",
       "       [  0.],\n",
       "       ...,\n",
       "       [  0.],\n",
       "       [  0.],\n",
       "       [  0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>113.0 nan nan nan ... nan nan nan</div><input id='attrs-8ac3b239-35bb-44c6-a61c-aee4d17de266' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8ac3b239-35bb-44c6-a61c-aee4d17de266' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-eef3770b-dce9-40ee-b7e9-9ea862681c08' class='xr-var-data-in' type='checkbox'><label for='data-eef3770b-dce9-40ee-b7e9-9ea862681c08' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[113.],\n",
       "       [ nan],\n",
       "       [ nan],\n",
       "       ...,\n",
       "       [ nan],\n",
       "       [ nan],\n",
       "       [ nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 1.0 0.0 0.0 0.0</div><input id='attrs-b9975d6f-7dfa-471f-9d0a-0c82fc792d86' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b9975d6f-7dfa-471f-9d0a-0c82fc792d86' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-242c3af2-7c15-4410-b0fa-6f26bb28a6c5' class='xr-var-data-in' type='checkbox'><label for='data-242c3af2-7c15-4410-b0fa-6f26bb28a6c5' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Source</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;HAVANA&#x27; &#x27;HAVANA&#x27; ... &#x27;HAVANA&#x27;</div><input id='attrs-0baa9aab-df06-428a-ae63-8efefe2e39b6' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0baa9aab-df06-428a-ae63-8efefe2e39b6' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-83519934-faf4-4b21-9919-84186aab9c3a' class='xr-var-data-in' type='checkbox'><label for='data-83519934-faf4-4b21-9919-84186aab9c3a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;, ..., &#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;exon&#x27; &#x27;exon&#x27; ... &#x27;exon&#x27; &#x27;exon&#x27;</div><input id='attrs-3740831e-98ef-4885-80da-184bd62c277a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3740831e-98ef-4885-80da-184bd62c277a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3619729e-eead-4030-a79d-2380342223e0' class='xr-var-data-in' type='checkbox'><label for='data-3619729e-eead-4030-a79d-2380342223e0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;exon&#x27;, &#x27;exon&#x27;, &#x27;exon&#x27;, ..., &#x27;exon&#x27;, &#x27;exon&#x27;, &#x27;exon&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Score</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;</div><input id='attrs-b3191eb2-b295-4c54-9bd5-9f2d5932dac8' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b3191eb2-b295-4c54-9bd5-9f2d5932dac8' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1e7cae54-5d0e-4ada-8295-34d125cd920e' class='xr-var-data-in' type='checkbox'><label for='data-1e7cae54-5d0e-4ada-8295-34d125cd920e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;, ..., &#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Phase</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;</div><input id='attrs-c45edfe1-4a57-4e89-8226-a6fe7ce8266d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c45edfe1-4a57-4e89-8226-a6fe7ce8266d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ccdf3346-5a95-4b2e-83d6-fc253bc15cb3' class='xr-var-data-in' type='checkbox'><label for='data-ccdf3346-5a95-4b2e-83d6-fc253bc15cb3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;, ..., &#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;exon:ENSMUST00000027036.10:2&#x27; ....</div><input id='attrs-211320ee-b57a-412a-b7eb-ce0c08fd7830' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-211320ee-b57a-412a-b7eb-ce0c08fd7830' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ddd5a57a-3ca3-4d3d-9cb4-2a09318970c5' class='xr-var-data-in' type='checkbox'><label for='data-ddd5a57a-3ca3-4d3d-9cb4-2a09318970c5' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;exon:ENSMUST00000027036.10:2&#x27;, &#x27;exon:ENSMUST00000027036.10:3&#x27;,\n",
       "       &#x27;exon:ENSMUST00000155020.1:2&#x27;, ..., &#x27;exon:ENSMUST00000188269.1:4&#x27;,\n",
       "       &#x27;exon:ENSMUST00000188269.1:3&#x27;, &#x27;exon:ENSMUST00000188269.1:2&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Parent</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUST00000027036.10&#x27; ... &#x27;ENS...</div><input id='attrs-2956497d-5ed9-4fd3-b7ef-05dc02c5186a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2956497d-5ed9-4fd3-b7ef-05dc02c5186a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-99544fba-f0c5-4645-bbd9-f6def65c2a8d' class='xr-var-data-in' type='checkbox'><label for='data-99544fba-f0c5-4645-bbd9-f6def65c2a8d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUST00000027036.10&#x27;, &#x27;ENSMUST00000027036.10&#x27;,\n",
       "       &#x27;ENSMUST00000155020.1&#x27;, ..., &#x27;ENSMUST00000188269.1&#x27;,\n",
       "       &#x27;ENSMUST00000188269.1&#x27;, &#x27;ENSMUST00000188269.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-30dfc0ee-ae4e-4e18-8173-bb10d46e8ae6' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-30dfc0ee-ae4e-4e18-8173-bb10d46e8ae6' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a3fec75b-5271-4505-b06a-0929f986afa8' class='xr-var-data-in' type='checkbox'><label for='data-a3fec75b-5271-4505-b06a-0929f986afa8' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000025903.14&#x27;,\n",
       "       &#x27;ENSMUSG00000104217.1&#x27;, ..., &#x27;ENSMUSG00000100608.1&#x27;,\n",
       "       &#x27;ENSMUSG00000100608.1&#x27;, &#x27;ENSMUSG00000100608.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUST00000027036.10&#x27; ... &#x27;ENS...</div><input id='attrs-772563d2-22cf-4484-a86d-475260ca6e05' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-772563d2-22cf-4484-a86d-475260ca6e05' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-69a78739-317a-4e36-beac-ae023ea128e4' class='xr-var-data-in' type='checkbox'><label for='data-69a78739-317a-4e36-beac-ae023ea128e4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUST00000027036.10&#x27;, &#x27;ENSMUST00000027036.10&#x27;,\n",
       "       &#x27;ENSMUST00000155020.1&#x27;, ..., &#x27;ENSMUST00000188269.1&#x27;,\n",
       "       &#x27;ENSMUST00000188269.1&#x27;, &#x27;ENSMUST00000188269.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;protein_coding&#x27; ... &#x27;protein_co...</div><input id='attrs-83b66ea0-4d6b-4d73-90d4-cbe27d41231a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-83b66ea0-4d6b-4d73-90d4-cbe27d41231a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-708fb7a3-85c4-4c3e-bbbb-eca0da740e24' class='xr-var-data-in' type='checkbox'><label for='data-708fb7a3-85c4-4c3e-bbbb-eca0da740e24' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, ...,\n",
       "       &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Lypla1&#x27; &#x27;Lypla1&#x27; ... &#x27;Gm21996&#x27;</div><input id='attrs-d985a492-bba7-400f-8dd1-2604be82f450' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d985a492-bba7-400f-8dd1-2604be82f450' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8d34707d-a791-4dc6-bd14-135d057005ec' class='xr-var-data-in' type='checkbox'><label for='data-8d34707d-a791-4dc6-bd14-135d057005ec' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Lypla1&#x27;, &#x27;Lypla1&#x27;, &#x27;Gm37988&#x27;, ..., &#x27;Gm21996&#x27;, &#x27;Gm21996&#x27;,\n",
       "       &#x27;Gm21996&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;protein_coding&#x27; ... &#x27;protein_co...</div><input id='attrs-5f2a73b5-a667-4fd3-9db4-e132db3c0d16' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5f2a73b5-a667-4fd3-9db4-e132db3c0d16' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9c4d3638-3277-4474-ab63-c187568cda03' class='xr-var-data-in' type='checkbox'><label for='data-9c4d3638-3277-4474-ab63-c187568cda03' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;nonsense_mediated_decay&#x27;, ...,\n",
       "       &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Lypla1-201&#x27; ... &#x27;Gm21996-201&#x27;</div><input id='attrs-90a9e6b1-ea74-490a-aaed-84fc149380b8' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-90a9e6b1-ea74-490a-aaed-84fc149380b8' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7d0756bc-36e7-4107-9a49-658f0f94c43b' class='xr-var-data-in' type='checkbox'><label for='data-7d0756bc-36e7-4107-9a49-658f0f94c43b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Lypla1-201&#x27;, &#x27;Lypla1-201&#x27;, &#x27;Gm37988-201&#x27;, ..., &#x27;Gm21996-201&#x27;,\n",
       "       &#x27;Gm21996-201&#x27;, &#x27;Gm21996-201&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;2&#x27; &#x27;3&#x27; &#x27;2&#x27; &#x27;4&#x27; ... &#x27;5&#x27; &#x27;4&#x27; &#x27;3&#x27; &#x27;2&#x27;</div><input id='attrs-2fc82528-7eb9-441c-bbc8-eeb0159fec7f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2fc82528-7eb9-441c-bbc8-eeb0159fec7f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ad182daf-98e9-4504-b8b3-0b718b14eac0' class='xr-var-data-in' type='checkbox'><label for='data-ad182daf-98e9-4504-b8b3-0b718b14eac0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;2&#x27;, &#x27;3&#x27;, &#x27;2&#x27;, ..., &#x27;4&#x27;, &#x27;3&#x27;, &#x27;2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSE00001255376.1&#x27; ... &#x27;ENSM...</div><input id='attrs-19043b29-c2b8-44f9-a044-a9400adf8989' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-19043b29-c2b8-44f9-a044-a9400adf8989' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-86906730-829d-41c1-8ca1-93b66f6bd8d3' class='xr-var-data-in' type='checkbox'><label for='data-86906730-829d-41c1-8ca1-93b66f6bd8d3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSE00001255376.1&#x27;, &#x27;ENSMUSE00001222304.1&#x27;,\n",
       "       &#x27;ENSMUSE00001241075.1&#x27;, ..., &#x27;ENSMUSE00001333847.1&#x27;,\n",
       "       &#x27;ENSMUSE00001331494.1&#x27;, &#x27;ENSMUSE00001332880.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;</div><input id='attrs-6060e53a-f384-4d30-9966-e5439fc0115f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6060e53a-f384-4d30-9966-e5439fc0115f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c65db9ee-ebf5-4b5c-bdbb-38c7192142b2' class='xr-var-data-in' type='checkbox'><label for='data-c65db9ee-ebf5-4b5c-bdbb-38c7192142b2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;, ..., &#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_support_level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;5&#x27; &#x27;1&#x27; ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;</div><input id='attrs-afbf5053-71ca-4f96-9ca2-885c1ee9dcac' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-afbf5053-71ca-4f96-9ca2-885c1ee9dcac' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-eb4b1cf2-68a4-4dbe-a2c1-f87d7e0f77a1' class='xr-var-data-in' type='checkbox'><label for='data-eb4b1cf2-68a4-4dbe-a2c1-f87d7e0f77a1' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;5&#x27;, ..., &#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Mgi_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;MGI:1344588&#x27; ... &#x27;MGI:5440224&#x27;</div><input id='attrs-32038a94-cac1-4b8a-9674-63840b4924f9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-32038a94-cac1-4b8a-9674-63840b4924f9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-dcd90594-c253-4e67-9433-864731787434' class='xr-var-data-in' type='checkbox'><label for='data-dcd90594-c253-4e67-9433-864731787434' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;MGI:1344588&#x27;, &#x27;MGI:1344588&#x27;, &#x27;MGI:5611216&#x27;, ..., &#x27;MGI:5440224&#x27;,\n",
       "       &#x27;MGI:5440224&#x27;, &#x27;MGI:5440224&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Tag</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;basic,appris_principal_1,CCDS&#x27; ...</div><input id='attrs-cb528e44-8237-4a6c-8aea-282dcd8a47c0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-cb528e44-8237-4a6c-8aea-282dcd8a47c0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-de20ffd4-d9a5-418a-b680-c0e9219d699c' class='xr-var-data-in' type='checkbox'><label for='data-de20ffd4-d9a5-418a-b680-c0e9219d699c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;basic,appris_principal_1,CCDS&#x27;, &#x27;basic,appris_principal_1,CCDS&#x27;,\n",
       "       &#x27;basic,appris_principal_1,readthrough_transcript&#x27;, ...,\n",
       "       &#x27;basic,appris_principal_1&#x27;, &#x27;basic,appris_principal_1&#x27;,\n",
       "       &#x27;basic,appris_principal_1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_gene</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTTM...</div><input id='attrs-381cbbaa-54d6-43e0-abbc-b18c8bc62843' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-381cbbaa-54d6-43e0-abbc-b18c8bc62843' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-2f773747-e93e-452e-b6cc-51f3adc5ea4c' class='xr-var-data-in' type='checkbox'><label for='data-2f773747-e93e-452e-b6cc-51f3adc5ea4c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;OTTMUSG00000021562.4&#x27;, &#x27;OTTMUSG00000021562.4&#x27;,\n",
       "       &#x27;OTTMUSG00000050100.1&#x27;, ..., &#x27;OTTMUSG00000047316.1&#x27;,\n",
       "       &#x27;OTTMUSG00000047316.1&#x27;, &#x27;OTTMUSG00000047316.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_transcript</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;OTTMUST00000051162.1&#x27; ... &#x27;OTTM...</div><input id='attrs-aa9e0ec6-f00e-440e-b813-e493549c9884' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-aa9e0ec6-f00e-440e-b813-e493549c9884' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3eeb5b80-90a9-4a8b-9ee4-be4a70b2adb7' class='xr-var-data-in' type='checkbox'><label for='data-3eeb5b80-90a9-4a8b-9ee4-be4a70b2adb7' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;OTTMUST00000051162.1&#x27;, &#x27;OTTMUST00000051162.1&#x27;,\n",
       "       &#x27;OTTMUST00000127419.1&#x27;, ..., &#x27;OTTMUST00000122318.1&#x27;,\n",
       "       &#x27;OTTMUST00000122318.1&#x27;, &#x27;OTTMUST00000122318.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Protein_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSP00000027036.4&#x27; ... &#x27;ENSM...</div><input id='attrs-0ce296fe-0a88-422e-9273-4a6ab375db21' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0ce296fe-0a88-422e-9273-4a6ab375db21' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-186f0ffa-2ea4-4b94-88e2-9c3eb7553809' class='xr-var-data-in' type='checkbox'><label for='data-186f0ffa-2ea4-4b94-88e2-9c3eb7553809' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSP00000027036.4&#x27;, &#x27;ENSMUSP00000027036.4&#x27;,\n",
       "       &#x27;ENSMUSP00000136108.1&#x27;, ..., &#x27;ENSMUSP00000141173.1&#x27;,\n",
       "       &#x27;ENSMUSP00000141173.1&#x27;, &#x27;ENSMUSP00000141173.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ccdsid</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;CCDS14806.1&#x27; &#x27;CCDS14806.1&#x27; ... &#x27;&#x27;</div><input id='attrs-8660a87e-3354-429e-bdf1-193ae4228bdd' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8660a87e-3354-429e-bdf1-193ae4228bdd' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-552ba290-5160-40b6-8e90-69efc709d35f' class='xr-var-data-in' type='checkbox'><label for='data-552ba290-5160-40b6-8e90-69efc709d35f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;CCDS14806.1&#x27;, &#x27;CCDS14806.1&#x27;, &#x27;&#x27;, ..., &#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ont</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; ... &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27;</div><input id='attrs-5151416d-b225-4e97-9256-88be1c44e0d9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5151416d-b225-4e97-9256-88be1c44e0d9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7f123c78-344f-489e-88f1-ac417967a6de' class='xr-var-data-in' type='checkbox'><label for='data-7f123c78-344f-489e-88f1-ac417967a6de' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;, ..., &#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;], dtype=object)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-038e55c1-f9d4-45c7-ab01-e9b35333e571' class='xr-section-summary-in' type='checkbox'  ><label for='section-038e55c1-f9d4-45c7-ab01-e9b35333e571' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>sample_id</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-db1a3935-482b-46ca-b5d5-69ba7c683506' class='xr-index-data-in' type='checkbox'/><label for='index-db1a3935-482b-46ca-b5d5-69ba7c683506' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;sample_0&#x27;], dtype=&#x27;object&#x27;, name=&#x27;sample_id&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-ca91a99d-7ee4-4ace-8f9d-fdb7c6c584dc' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-ca91a99d-7ee4-4ace-8f9d-fdb7c6c584dc' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.Dataset>\n",
       "Dimensions:                   (ranges: 180974, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) <U2 '1' '1' '1' '1' ... 'Y' 'Y' 'Y' 'Y'\n",
       "    start                     (ranges) int64 4808454 4828583 ... 89743993\n",
       "    end                       (ranges) int64 4808485 4828648 ... 89744124\n",
       "    range_id                  (ranges) int64 5 9 13 14 ... 408327 408328 408329\n",
       "    num_contexts              (ranges) int64 1 0 0 0 0 2 2 0 ... 0 0 0 0 1 0 0 0\n",
       "    range_length              (ranges) int64 31 65 65 47 47 ... 34 66 34 66 131\n",
       "  * sample_id                 (sample_id) <U8 'sample_0'\n",
       "    strand                    (ranges) <U1 '+' '+' '+' '+' ... '-' '-' '-' '-'\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 113.0 0.0 ... 0.0 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 113.0 nan ... nan nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    Source                    (ranges) object 'HAVANA' 'HAVANA' ... 'HAVANA'\n",
       "    Type                      (ranges) object 'exon' 'exon' ... 'exon' 'exon'\n",
       "    Score                     (ranges) object '.' '.' '.' '.' ... '.' '.' '.'\n",
       "    Phase                     (ranges) object '.' '.' '.' '.' ... '.' '.' '.'\n",
       "    Id                        (ranges) object 'exon:ENSMUST00000027036.10:2' ...\n",
       "    Parent                    (ranges) object 'ENSMUST00000027036.10' ... 'EN...\n",
       "    Gene_id                   (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Transcript_id             (ranges) object 'ENSMUST00000027036.10' ... 'EN...\n",
       "    Gene_type                 (ranges) object 'protein_coding' ... 'protein_c...\n",
       "    Gene_name                 (ranges) object 'Lypla1' 'Lypla1' ... 'Gm21996'\n",
       "    Transcript_type           (ranges) object 'protein_coding' ... 'protein_c...\n",
       "    Transcript_name           (ranges) object 'Lypla1-201' ... 'Gm21996-201'\n",
       "    Exon_number               (ranges) object '2' '3' '2' '4' ... '4' '3' '2'\n",
       "    Exon_id                   (ranges) object 'ENSMUSE00001255376.1' ... 'ENS...\n",
       "    Level                     (ranges) object '2' '2' '2' '2' ... '2' '2' '2'\n",
       "    Transcript_support_level  (ranges) object '1' '1' '5' '1' ... '1' '1' '1'\n",
       "    Mgi_id                    (ranges) object 'MGI:1344588' ... 'MGI:5440224'\n",
       "    Tag                       (ranges) object 'basic,appris_principal_1,CCDS'...\n",
       "    Havana_gene               (ranges) object 'OTTMUSG00000021562.4' ... 'OTT...\n",
       "    Havana_transcript         (ranges) object 'OTTMUST00000051162.1' ... 'OTT...\n",
       "    Protein_id                (ranges) object 'ENSMUSP00000027036.4' ... 'ENS...\n",
       "    Ccdsid                    (ranges) object 'CCDS14806.1' 'CCDS14806.1' ... ''\n",
       "    Ont                       (ranges) object '' '' '' '' '' ... '' '' '' '' ''"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sense_methylation['exons']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "calculate methlyation fractions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>total_c</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.807980</td>\n",
       "      <td>0.025436</td>\n",
       "      <td>0.847215</td>\n",
       "      <td>51.0</td>\n",
       "      <td>6015.0</td>\n",
       "      <td>2995</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003.15</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.014888</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>7.0</td>\n",
       "      <td>403.0</td>\n",
       "      <td>681</td>\n",
       "      <td>Pbsn</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000003.13</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028.15</td>\n",
       "      <td>0.581295</td>\n",
       "      <td>0.031589</td>\n",
       "      <td>0.619697</td>\n",
       "      <td>60.0</td>\n",
       "      <td>6458.0</td>\n",
       "      <td>1955</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000028.13</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000037.17</td>\n",
       "      <td>0.758328</td>\n",
       "      <td>0.017636</td>\n",
       "      <td>0.786414</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1531.0</td>\n",
       "      <td>2777</td>\n",
       "      <td>Scml2</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000101113.8</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000049.11</td>\n",
       "      <td>0.783618</td>\n",
       "      <td>0.036519</td>\n",
       "      <td>0.835495</td>\n",
       "      <td>28.0</td>\n",
       "      <td>2930.0</td>\n",
       "      <td>1068</td>\n",
       "      <td>Apoh</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000049.5</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  total_c  \\\n",
       "0   ENSMUSG00000000001.4  0.807980  0.025436   0.847215       51.0   6015.0   \n",
       "1  ENSMUSG00000000003.15  0.774194  0.014888   0.806452        7.0    403.0   \n",
       "2  ENSMUSG00000000028.15  0.581295  0.031589   0.619697       60.0   6458.0   \n",
       "3  ENSMUSG00000000037.17  0.758328  0.017636   0.786414       30.0   1531.0   \n",
       "4  ENSMUSG00000000049.11  0.783618  0.036519   0.835495       28.0   2930.0   \n",
       "\n",
       "   range_length Gene_name contig strand          Transcript_id Region  \n",
       "0          2995     Gnai3      3      -   ENSMUST00000000001.4  exons  \n",
       "1           681      Pbsn      X      -  ENSMUST00000000003.13  exons  \n",
       "2          1955     Cdc45     16      -  ENSMUST00000000028.13  exons  \n",
       "3          2777     Scml2      X      +   ENSMUST00000101113.8  exons  \n",
       "4          1068      Apoh     11      +   ENSMUST00000000049.5  exons  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict_df_sense = {}\n",
    "dict_df_antisense = {}\n",
    "\n",
    "for region in dict_regions_to_patch_plus.keys():\n",
    "    dict_df_sense[region] = summarise_across_region(sense_methylation[region], region)\n",
    "    dict_df_antisense[region] = summarise_across_region(antisense_methylation[region], region)\n",
    "\n",
    "stitched_sense_regions = pd.concat(dict_df_sense.values())\n",
    "stitched_antisense_regions = pd.concat(dict_df_antisense.values())\n",
    "\n",
    "\n",
    "stitched_sense_regions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>total_c</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.807980</td>\n",
       "      <td>0.025436</td>\n",
       "      <td>0.847215</td>\n",
       "      <td>51.0</td>\n",
       "      <td>6015.0</td>\n",
       "      <td>2995</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.780352</td>\n",
       "      <td>0.038773</td>\n",
       "      <td>0.832180</td>\n",
       "      <td>144.0</td>\n",
       "      <td>15320.0</td>\n",
       "      <td>13562</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>introns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.000784</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000784</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1275.0</td>\n",
       "      <td>140</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.813870</td>\n",
       "      <td>0.023717</td>\n",
       "      <td>0.849446</td>\n",
       "      <td>32.0</td>\n",
       "      <td>3879.0</td>\n",
       "      <td>2054</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  total_c  \\\n",
       "0  ENSMUSG00000000001.4  0.807980  0.025436   0.847215       51.0   6015.0   \n",
       "0  ENSMUSG00000000001.4  0.780352  0.038773   0.832180      144.0  15320.0   \n",
       "0  ENSMUSG00000000001.4  0.000784  0.000000   0.000784       15.0   1275.0   \n",
       "0  ENSMUSG00000000001.4  0.813870  0.023717   0.849446       32.0   3879.0   \n",
       "\n",
       "   range_length Gene_name contig strand         Transcript_id  \\\n",
       "0          2995     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0         13562     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0           140     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0          2054     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "\n",
       "             Region  \n",
       "0             exons  \n",
       "0           introns  \n",
       "0   five_prime_utrs  \n",
       "0  three_prime_utrs  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stitched_sense_regions[stitched_sense_regions['Transcript_id']==\"ENSMUST00000000001.4\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>total_c</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.810206</td>\n",
       "      <td>0.026502</td>\n",
       "      <td>0.848889</td>\n",
       "      <td>51.0</td>\n",
       "      <td>6075.0</td>\n",
       "      <td>2995</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.772727</td>\n",
       "      <td>0.045917</td>\n",
       "      <td>0.834170</td>\n",
       "      <td>148.0</td>\n",
       "      <td>15136.0</td>\n",
       "      <td>13562</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>introns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1205.0</td>\n",
       "      <td>140</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.816244</td>\n",
       "      <td>0.028680</td>\n",
       "      <td>0.856091</td>\n",
       "      <td>32.0</td>\n",
       "      <td>3940.0</td>\n",
       "      <td>2054</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  total_c  \\\n",
       "0  ENSMUSG00000000001.4  0.810206  0.026502   0.848889       51.0   6075.0   \n",
       "0  ENSMUSG00000000001.4  0.772727  0.045917   0.834170      148.0  15136.0   \n",
       "0  ENSMUSG00000000001.4  0.000000  0.000000   0.000000       15.0   1205.0   \n",
       "0  ENSMUSG00000000001.4  0.816244  0.028680   0.856091       32.0   3940.0   \n",
       "\n",
       "   range_length Gene_name contig strand         Transcript_id  \\\n",
       "0          2995     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0         13562     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0           140     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0          2054     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "\n",
       "             Region  \n",
       "0             exons  \n",
       "0           introns  \n",
       "0   five_prime_utrs  \n",
       "0  three_prime_utrs  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stitched_antisense_regions[stitched_antisense_regions['Transcript_id']==\"ENSMUST00000000001.4\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(69769, 12)"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stitched_sense_regions.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "dropping regions with no coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "summarised regions on sense strand : (69769, 12)\n",
      "of which has no coverage: 4587\n",
      "proportion no coverage: 0.06574553168312575\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>total_c</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSMUSG00000000325.16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4406</td>\n",
       "      <td>Arvcf</td>\n",
       "      <td>16</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000115614.8</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>ENSMUSG00000000326.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1365</td>\n",
       "      <td>Comt</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000115609.9</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>ENSMUSG00000000791.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2688</td>\n",
       "      <td>Il12rb1</td>\n",
       "      <td>8</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000808.7</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>ENSMUSG00000002104.11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1251</td>\n",
       "      <td>Rapsn</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000050323.5</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>ENSMUSG00000002910.11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1755</td>\n",
       "      <td>Arrdc2</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000002989.10</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15443</th>\n",
       "      <td>ENSMUSG00000115219.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>333</td>\n",
       "      <td>Eef1akmt4</td>\n",
       "      <td>16</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000115522.9</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15453</th>\n",
       "      <td>ENSMUSG00000116780.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>Gm3417</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000097399.5</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15454</th>\n",
       "      <td>ENSMUSG00000116953.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100</td>\n",
       "      <td>9030025P20Rik</td>\n",
       "      <td>17</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000061688.10</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15455</th>\n",
       "      <td>ENSMUSG00000117081.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>62</td>\n",
       "      <td>Gm49948</td>\n",
       "      <td>16</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000113795.7</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15463</th>\n",
       "      <td>ENSMUSG00000118346.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>120</td>\n",
       "      <td>Tmem179b</td>\n",
       "      <td>19</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000010249.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4587 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_id  mean_mc  mean_hmc  mean_modc  cpg_count  \\\n",
       "59     ENSMUSG00000000325.16      NaN       NaN        NaN        0.0   \n",
       "60     ENSMUSG00000000326.13      NaN       NaN        NaN        0.0   \n",
       "134     ENSMUSG00000000791.9      NaN       NaN        NaN        0.0   \n",
       "376    ENSMUSG00000002104.11      NaN       NaN        NaN        0.0   \n",
       "497    ENSMUSG00000002910.11      NaN       NaN        NaN        0.0   \n",
       "...                      ...      ...       ...        ...        ...   \n",
       "15443   ENSMUSG00000115219.2      NaN       NaN        NaN        0.0   \n",
       "15453   ENSMUSG00000116780.1      NaN       NaN        NaN        0.0   \n",
       "15454   ENSMUSG00000116953.1      NaN       NaN        NaN        0.0   \n",
       "15455   ENSMUSG00000117081.1      NaN       NaN        NaN        0.0   \n",
       "15463   ENSMUSG00000118346.1      NaN       NaN        NaN        0.0   \n",
       "\n",
       "       total_c  range_length      Gene_name contig strand  \\\n",
       "59         0.0          4406          Arvcf     16      +   \n",
       "60         0.0          1365           Comt     16      -   \n",
       "134        0.0          2688        Il12rb1      8      +   \n",
       "376        0.0          1251          Rapsn      2      +   \n",
       "497        0.0          1755         Arrdc2      8      -   \n",
       "...        ...           ...            ...    ...    ...   \n",
       "15443      0.0           333      Eef1akmt4     16      +   \n",
       "15453      0.0             2         Gm3417     17      -   \n",
       "15454      0.0           100  9030025P20Rik     17      +   \n",
       "15455      0.0            62        Gm49948     16      +   \n",
       "15463      0.0           120       Tmem179b     19      -   \n",
       "\n",
       "               Transcript_id            Region  \n",
       "59      ENSMUST00000115614.8             exons  \n",
       "60      ENSMUST00000115609.9             exons  \n",
       "134     ENSMUST00000000808.7             exons  \n",
       "376     ENSMUST00000050323.5             exons  \n",
       "497    ENSMUST00000002989.10             exons  \n",
       "...                      ...               ...  \n",
       "15443   ENSMUST00000115522.9  three_prime_utrs  \n",
       "15453   ENSMUST00000097399.5  three_prime_utrs  \n",
       "15454  ENSMUST00000061688.10  three_prime_utrs  \n",
       "15455   ENSMUST00000113795.7  three_prime_utrs  \n",
       "15463   ENSMUST00000010249.6  three_prime_utrs  \n",
       "\n",
       "[4587 rows x 12 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"summarised regions on sense strand : {stitched_sense_regions.shape}\")\n",
    "\n",
    "all_count = stitched_sense_regions.shape[0]\n",
    "nocov_count = stitched_sense_regions[stitched_sense_regions['total_c']<=0].shape[0]\n",
    "print(f\"of which has no coverage: {nocov_count}\")\n",
    "print(f\"proportion no coverage: {nocov_count/all_count}\")\n",
    "\n",
    "stitched_sense_regions[stitched_sense_regions['total_c']<=0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop regions with no coverage:\n",
    "stitched_sense_regions = stitched_sense_regions[stitched_sense_regions['total_c']>0]\n",
    "stitched_antisense_regions = stitched_antisense_regions[stitched_antisense_regions['total_c']>0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>total_c</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.807980</td>\n",
       "      <td>0.025436</td>\n",
       "      <td>0.847215</td>\n",
       "      <td>51.0</td>\n",
       "      <td>6015.0</td>\n",
       "      <td>2995</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.780352</td>\n",
       "      <td>0.038773</td>\n",
       "      <td>0.832180</td>\n",
       "      <td>144.0</td>\n",
       "      <td>15320.0</td>\n",
       "      <td>13562</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>introns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.000784</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000784</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1275.0</td>\n",
       "      <td>140</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.813870</td>\n",
       "      <td>0.023717</td>\n",
       "      <td>0.849446</td>\n",
       "      <td>32.0</td>\n",
       "      <td>3879.0</td>\n",
       "      <td>2054</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  total_c  \\\n",
       "0  ENSMUSG00000000001.4  0.807980  0.025436   0.847215       51.0   6015.0   \n",
       "0  ENSMUSG00000000001.4  0.780352  0.038773   0.832180      144.0  15320.0   \n",
       "0  ENSMUSG00000000001.4  0.000784  0.000000   0.000784       15.0   1275.0   \n",
       "0  ENSMUSG00000000001.4  0.813870  0.023717   0.849446       32.0   3879.0   \n",
       "\n",
       "   range_length Gene_name contig strand         Transcript_id  \\\n",
       "0          2995     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0         13562     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0           140     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0          2054     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "\n",
       "             Region  \n",
       "0             exons  \n",
       "0           introns  \n",
       "0   five_prime_utrs  \n",
       "0  three_prime_utrs  "
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stitched_sense_regions[stitched_sense_regions['Gene_id']==\"ENSMUSG00000000001.4\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>total_c</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.810206</td>\n",
       "      <td>0.026502</td>\n",
       "      <td>0.848889</td>\n",
       "      <td>51.0</td>\n",
       "      <td>6075.0</td>\n",
       "      <td>2995</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.772727</td>\n",
       "      <td>0.045917</td>\n",
       "      <td>0.834170</td>\n",
       "      <td>148.0</td>\n",
       "      <td>15136.0</td>\n",
       "      <td>13562</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>introns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1205.0</td>\n",
       "      <td>140</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.816244</td>\n",
       "      <td>0.028680</td>\n",
       "      <td>0.856091</td>\n",
       "      <td>32.0</td>\n",
       "      <td>3940.0</td>\n",
       "      <td>2054</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  total_c  \\\n",
       "0  ENSMUSG00000000001.4  0.810206  0.026502   0.848889       51.0   6075.0   \n",
       "0  ENSMUSG00000000001.4  0.772727  0.045917   0.834170      148.0  15136.0   \n",
       "0  ENSMUSG00000000001.4  0.000000  0.000000   0.000000       15.0   1205.0   \n",
       "0  ENSMUSG00000000001.4  0.816244  0.028680   0.856091       32.0   3940.0   \n",
       "\n",
       "   range_length Gene_name contig strand         Transcript_id  \\\n",
       "0          2995     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0         13562     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0           140     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0          2054     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "\n",
       "             Region  \n",
       "0             exons  \n",
       "0           introns  \n",
       "0   five_prime_utrs  \n",
       "0  three_prime_utrs  "
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stitched_antisense_regions[stitched_antisense_regions['Gene_id']==\"ENSMUSG00000000001.4\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>id</th>\n",
       "      <th>parent</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>mgi_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>757650</th>\n",
       "      <td>3</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>108107279</td>\n",
       "      <td>108146145</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>...</td>\n",
       "      <td>Gnai3-201</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:95773</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>OTTMUSG00000007197.1</td>\n",
       "      <td>OTTMUST00000016610.1</td>\n",
       "      <td>ENSMUSP00000000001.4</td>\n",
       "      <td>CCDS17751.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seq_id  source        type      start        end score strand phase  \\\n",
       "757650      3  HAVANA  transcript  108107279  108146145     .      -     .   \n",
       "\n",
       "                          id                parent  ... transcript_name level  \\\n",
       "757650  ENSMUST00000000001.4  ENSMUSG00000000001.4  ...       Gnai3-201     2   \n",
       "\n",
       "       transcript_support_level     mgi_id                            tag  \\\n",
       "757650                        1  MGI:95773  basic,appris_principal_1,CCDS   \n",
       "\n",
       "                 havana_gene     havana_transcript            protein_id  \\\n",
       "757650  OTTMUSG00000007197.1  OTTMUST00000016610.1  ENSMUSP00000000001.4   \n",
       "\n",
       "             ccdsid ont  \n",
       "757650  CCDS17751.1      \n",
       "\n",
       "[1 rows x 25 columns]"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts[transcripts[\"id\"]==\"ENSMUST00000000001.4\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### no stitching"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### TSS, TES, first exon, first intron, genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plus strand total regions: 53146\n",
      "minus strand total regions: 53416\n"
     ]
    }
   ],
   "source": [
    "print(f\"plus strand total regions: {len(before_tss_plus)+ len(after_tes_plus) + len(genes_plus) + len(first_exons_plus) + len(first_introns_plus)}\")\n",
    "print(f\"minus strand total regions: {len(before_tss_minus)+ len(after_tes_minus) + len(genes_minus) + len(first_exons_minus) + len(first_introns_minus)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop columns\n",
    "to_drop = [\"Score\", \"Source\", \"Phase\", \"Type\"]\n",
    "\n",
    "regions_dict_plus ={\n",
    "        \"before_tss\": before_tss_plus,\n",
    "        \"after_tes\": after_tes_plus,\n",
    "        \"genes\": genes_plus,\n",
    "        \"first_exons\": first_exons_plus,\n",
    "        \"first_introns\": first_introns_plus,\n",
    "    }\n",
    "\n",
    "for region in regions_dict_plus:\n",
    "    regions_dict_plus[region].Region = region\n",
    "    try:\n",
    "        regions_dict_plus[region] = regions_dict_plus[region].drop(to_drop)\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "regions_dict_minus ={\n",
    "        \"before_tss\": before_tss_minus,\n",
    "        \"after_tes\": after_tes_minus,\n",
    "        \"genes\": genes_minus,\n",
    "        \"first_exons\": first_exons_minus,\n",
    "        \"first_introns\": first_introns_minus,\n",
    "    }\n",
    "\n",
    "for region in regions_dict_minus:\n",
    "    regions_dict_minus[region].Region = region\n",
    "    try:\n",
    "        regions_dict_minus[region] = regions_dict_minus[region].drop(to_drop)\n",
    "    except:\n",
    "        pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarise_methylation_across_other_regions(plus_strand_cpgs, minus_strand_cpgs, regions_dict, region_strand, sense_methylation=None, antisense_methylation=None):\n",
    "    \"\"\"\n",
    "    Summarise the methylation data across multiple regions.\n",
    "\n",
    "    Parameters\n",
    "        rdr: pyranges object\n",
    "        regions_dict: dict\n",
    "\n",
    "    Returns\n",
    "        grouped_df: pd.DataFrame\n",
    "    \"\"\"\n",
    "    print(f\"summarising over {regions_dict.keys()}\")\n",
    "\n",
    "    rds_plus = plus_strand_cpgs.reduce_byranges(\n",
    "                ranges=list(regions_dict.values()), \n",
    "                var=[\"num_mc\", \"num_hmc\", \"num_modc\", \"num_total_c\"]\n",
    "            )\n",
    "    rds_minus = minus_strand_cpgs.reduce_byranges(\n",
    "                ranges=list(regions_dict.values()),      \n",
    "                var=[\"num_mc\", \"num_hmc\", \"num_modc\", \"num_total_c\"]\n",
    "            )\n",
    "   \n",
    "    if region_strand == \"+\":\n",
    "        # sense to the + region\n",
    "        sense_plus_ds = rds_plus\n",
    "        plus_sense = ['+'] * sense_plus_ds.dims['ranges']\n",
    "        sense_plus_ds = sense_plus_ds.assign_coords(strand=('ranges', plus_sense))\n",
    "\n",
    "        # antisense to the + region\n",
    "        antisense_plus_ds = rds_minus\n",
    "        plus_antisense = ['+'] * antisense_plus_ds.dims['ranges']\n",
    "        antisense_plus_ds = antisense_plus_ds.assign_coords(strand=('ranges', plus_antisense))\n",
    "        \n",
    "        # Add the modified dataset to the output xarray\n",
    "        if sense_methylation is None:\n",
    "            sense_methylation= sense_plus_ds\n",
    "        else:\n",
    "            sense_methylation= xr.concat([sense_methylation, sense_plus_ds], dim=\"ranges\")\n",
    "            \n",
    "        if antisense_methylation is None:\n",
    "            antisense_methylation = antisense_plus_ds\n",
    "        else:\n",
    "            antisense_methylation= xr.concat([antisense_methylation, antisense_plus_ds], dim=\"ranges\")\n",
    "    else:\n",
    "        # sense to the - region\n",
    "        sense_minus_ds = rds_minus\n",
    "        minus_sense = ['-'] * sense_minus_ds.dims['ranges']\n",
    "        sense_minus_ds = sense_minus_ds.assign_coords(strand=('ranges', minus_sense))\n",
    "        \n",
    "        # antisense to the - region\n",
    "        antisense_minus_ds = rds_plus\n",
    "        minus_antisense = ['-'] * antisense_minus_ds.dims['ranges']\n",
    "        antisense_minus_ds = antisense_minus_ds.assign_coords(strand=('ranges', minus_antisense))\n",
    "    \n",
    "\n",
    "        # Add the modified dataset to the output xarray\n",
    "        if sense_methylation is None:\n",
    "            sense_methylation = sense_minus_ds\n",
    "        else:\n",
    "            sense_methylation = xr.concat([sense_methylation, sense_minus_ds], dim=\"ranges\")\n",
    "\n",
    "        if antisense_methylation is None:\n",
    "            antisense_methylation= antisense_minus_ds\n",
    "        else:\n",
    "            antisense_methylation = xr.concat([antisense_methylation, antisense_minus_ds], dim=\"ranges\")\n",
    "\n",
    "    return sense_methylation, antisense_methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "summarising over dict_keys(['before_tss', 'after_tes', 'genes', 'first_exons', 'first_introns'])\n"
     ]
    }
   ],
   "source": [
    "# sense_ds, antisense_ds = summarise_methylation_across_other_regions(plus_strand_cov, minus_strand_cov, regions_dict_plus, \"+\")\n",
    "sense_ds, antisense_ds = summarise_methylation_across_other_regions(plus_strand_cov, minus_strand_cov, regions_dict_minus, \"-\", sense_ds, antisense_ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;\n",
       "Dimensions:                   (ranges: 106562, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) &lt;U2 &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;\n",
       "    start                     (ranges) int64 4807587 4807691 ... 22859847\n",
       "    end                       (ranges) int64 4807787 4807891 ... 22861212\n",
       "    range_id                  (ranges) int64 0 1 2 3 ... 53412 53413 53414 53415\n",
       "    num_contexts              (ranges) int64 27 35 21 0 7 2 11 ... 0 1 0 1 1 0 0\n",
       "    range_length              (ranges) int64 200 200 200 200 ... 1361 1365 1365\n",
       "  * sample_id                 (sample_id) &lt;U8 &#x27;sample_0&#x27;\n",
       "    strand                    (ranges) &lt;U1 &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 2.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 0.05714 ... nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 1.0 4.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.03704 0.1143 ... nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 1.0 6.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.03704 0.1714 ... nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 2.122e+03 ... 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 78.59 82.97 ... nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    Id                        (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Gene_id                   (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Gene_type                 (ranges) object &#x27;protein_coding&#x27; ... &#x27;protein_c...\n",
       "    Gene_name                 (ranges) object &#x27;Lypla1&#x27; &#x27;Gm37988&#x27; ... &#x27;Gm21366&#x27;\n",
       "    Level                     (ranges) object &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;\n",
       "    Mgi_id                    (ranges) object &#x27;MGI:1344588&#x27; ... &#x27;MGI:5434721&#x27;\n",
       "    Havana_gene               (ranges) object &#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTT...\n",
       "    Tag                       (ranges) object &#x27;overlapping_locus&#x27; ... &#x27;&#x27;\n",
       "    Transcript_id             (ranges) object &#x27;ENSMUST00000027036.10&#x27; ... &#x27;EN...\n",
       "    Region                    (ranges) object &#x27;before_tss&#x27; ... &#x27;first_introns&#x27;\n",
       "    Strand                    (ranges) object &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;\n",
       "    Parent                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_type           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_name           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_number               (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_id                   (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_support_level  (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Havana_transcript         (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Protein_id                (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ccdsid                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ont                       (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Level_1                   (ranges) float64 nan nan nan nan ... 8.0 8.0 8.0\n",
       "    Ranges_id                 (ranges) float64 nan nan ... 2.164e+04 2.164e+04\n",
       "    Intron_number             (ranges) object nan nan nan nan ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;\n",
       "    mean_mc                   (ranges, sample_id) float64 0.0 0.0006887 ... nan\n",
       "    mean_hmc                  (ranges, sample_id) float64 0.0004713 ... nan\n",
       "    mean_modc                 (ranges, sample_id) float64 0.0004713 ... nan</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-31147f6d-a8fa-4dfe-b354-cdbf7e1bee16' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-31147f6d-a8fa-4dfe-b354-cdbf7e1bee16' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>ranges</span>: 106562</li><li><span class='xr-has-index'>sample_id</span>: 1</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-1067057e-8211-4301-b1c1-e316f5fcba77' class='xr-section-summary-in' type='checkbox'  checked><label for='section-1067057e-8211-4301-b1c1-e316f5fcba77' class='xr-section-summary' >Coordinates: <span>(8)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>contig</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U2</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;</div><input id='attrs-2e3878b5-e7aa-4ec5-afc9-e3b448d7459f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2e3878b5-e7aa-4ec5-afc9-e3b448d7459f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e674a1f1-ea26-4d30-b588-2b9bc7b6d025' class='xr-var-data-in' type='checkbox'><label for='data-e674a1f1-ea26-4d30-b588-2b9bc7b6d025' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;, ..., &#x27;Y&#x27;, &#x27;Y&#x27;, &#x27;Y&#x27;], dtype=&#x27;&lt;U2&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>start</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4807587 4807691 ... 22859847</div><input id='attrs-724a1a40-afbe-4935-b421-17af4b80f89d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-724a1a40-afbe-4935-b421-17af4b80f89d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-872dc82d-2eff-4d21-a8db-a7424f0bee45' class='xr-var-data-in' type='checkbox'><label for='data-872dc82d-2eff-4d21-a8db-a7424f0bee45' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4807587,  4807691,  4857613, ..., 20992252, 22049944, 22859847])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>end</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4807787 4807891 ... 22861212</div><input id='attrs-a18cae6c-ca3b-4661-8ff5-e3f3c5d9b210' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-a18cae6c-ca3b-4661-8ff5-e3f3c5d9b210' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3e38e2b7-d458-43da-bdda-bd9e3a9f1919' class='xr-var-data-in' type='checkbox'><label for='data-3e38e2b7-d458-43da-bdda-bd9e3a9f1919' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4807787,  4807891,  4857813, ..., 20993613, 22051309, 22861212])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 ... 53412 53413 53414 53415</div><input id='attrs-b48001c8-3d60-4f26-a5f3-adba8b8e1c41' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b48001c8-3d60-4f26-a5f3-adba8b8e1c41' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9ee9e55d-bcfb-440a-9336-7878f0a4553c' class='xr-var-data-in' type='checkbox'><label for='data-9ee9e55d-bcfb-440a-9336-7878f0a4553c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([    0,     1,     2, ..., 53413, 53414, 53415])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_contexts</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>27 35 21 0 7 2 11 ... 0 1 0 1 1 0 0</div><input id='attrs-88647910-a1e9-4bce-a3d7-446739b32568' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-88647910-a1e9-4bce-a3d7-446739b32568' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-71931ba1-667d-4552-a3c4-cd9dda6d0370' class='xr-var-data-in' type='checkbox'><label for='data-71931ba1-667d-4552-a3c4-cd9dda6d0370' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([27, 35, 21, ...,  1,  0,  0])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_length</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>200 200 200 200 ... 1361 1365 1365</div><input id='attrs-b453aed7-1161-4c9f-b0e7-d4e28a73029c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b453aed7-1161-4c9f-b0e7-d4e28a73029c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-00829e8c-6a5f-4497-85c7-d20d1161dc71' class='xr-var-data-in' type='checkbox'><label for='data-00829e8c-6a5f-4497-85c7-d20d1161dc71' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 200,  200,  200, ..., 1361, 1365, 1365])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample_id</span></div><div class='xr-var-dims'>(sample_id)</div><div class='xr-var-dtype'>&lt;U8</div><div class='xr-var-preview xr-preview'>&#x27;sample_0&#x27;</div><input id='attrs-2d96aa2f-c27b-49bb-85de-fda0fc43b769' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2d96aa2f-c27b-49bb-85de-fda0fc43b769' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d6f86216-c1a7-48d4-8ec4-91135a6f6713' class='xr-var-data-in' type='checkbox'><label for='data-d6f86216-c1a7-48d4-8ec4-91135a6f6713' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;sample_0&#x27;], dtype=&#x27;&lt;U8&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U1</div><div class='xr-var-preview xr-preview'>&#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;</div><input id='attrs-1ba79bac-2c88-4c59-8e3f-4bb2d7373bb9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1ba79bac-2c88-4c59-8e3f-4bb2d7373bb9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ab15e215-f00f-4a97-9be2-13d6690a83d2' class='xr-var-data-in' type='checkbox'><label for='data-ab15e215-f00f-4a97-9be2-13d6690a83d2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;+&#x27;, &#x27;+&#x27;, &#x27;+&#x27;, ..., &#x27;-&#x27;, &#x27;-&#x27;, &#x27;-&#x27;], dtype=&#x27;&lt;U1&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-6c023c27-674b-4a50-a0fb-6053b41bcaf9' class='xr-section-summary-in' type='checkbox'  ><label for='section-6c023c27-674b-4a50-a0fb-6053b41bcaf9' class='xr-section-summary' >Data variables: <span>(39)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 2.0 4.0 0.0 ... 19.0 0.0 0.0</div><input id='attrs-53ba451f-c0de-4a35-ac20-77aaf8d8d061' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-53ba451f-c0de-4a35-ac20-77aaf8d8d061' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-2811e44c-1683-466f-8b37-b57e0b58ca52' class='xr-var-data-in' type='checkbox'><label for='data-2811e44c-1683-466f-8b37-b57e0b58ca52' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [ 2.],\n",
       "       [ 4.],\n",
       "       ...,\n",
       "       [19.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.05714 0.1905 ... 19.0 nan nan</div><input id='attrs-964fdbb2-839a-4442-a0fb-8ac39e87d4e3' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-964fdbb2-839a-4442-a0fb-8ac39e87d4e3' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-54f13990-4902-40ff-8441-f5957b542218' class='xr-var-data-in' type='checkbox'><label for='data-54f13990-4902-40ff-8441-f5957b542218' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.        ],\n",
       "       [ 0.05714286],\n",
       "       [ 0.19047619],\n",
       "       ...,\n",
       "       [19.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-4bf57e53-f9ce-4ee8-b3d0-87c8e9790935' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4bf57e53-f9ce-4ee8-b3d0-87c8e9790935' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5ba6f0f4-7751-4e2f-8760-4b0d238d4b56' class='xr-var-data-in' type='checkbox'><label for='data-5ba6f0f4-7751-4e2f-8760-4b0d238d4b56' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 4.0 2.0 0.0 ... 0.0 1.0 0.0 0.0</div><input id='attrs-df103408-4fd0-43e9-8827-7b6144a2fd0c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-df103408-4fd0-43e9-8827-7b6144a2fd0c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9c518803-36c5-4cf9-a5a9-2d9e9afa06f4' class='xr-var-data-in' type='checkbox'><label for='data-9c518803-36c5-4cf9-a5a9-2d9e9afa06f4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [4.],\n",
       "       [2.],\n",
       "       ...,\n",
       "       [1.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.03704 0.1143 0.09524 ... nan nan</div><input id='attrs-c3e3f546-9469-4cd1-a933-f11e70d1d923' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c3e3f546-9469-4cd1-a933-f11e70d1d923' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-71f7c168-dfcb-4625-ac18-00ba3101dc0e' class='xr-var-data-in' type='checkbox'><label for='data-71f7c168-dfcb-4625-ac18-00ba3101dc0e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.03703704],\n",
       "       [0.11428571],\n",
       "       [0.0952381 ],\n",
       "       ...,\n",
       "       [1.        ],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-ddf435e7-931c-47f1-a1b5-b927dcd4c1c7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ddf435e7-931c-47f1-a1b5-b927dcd4c1c7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c3635cbe-090e-4497-ae67-548291c8ff08' class='xr-var-data-in' type='checkbox'><label for='data-c3635cbe-090e-4497-ae67-548291c8ff08' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 6.0 6.0 0.0 ... 20.0 0.0 0.0</div><input id='attrs-7d8e6c44-a29d-48a9-bd92-943d450f9e66' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7d8e6c44-a29d-48a9-bd92-943d450f9e66' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5c5d8d47-35b5-49da-a625-f24597aeb222' class='xr-var-data-in' type='checkbox'><label for='data-5c5d8d47-35b5-49da-a625-f24597aeb222' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 1.],\n",
       "       [ 6.],\n",
       "       [ 6.],\n",
       "       ...,\n",
       "       [20.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.03704 0.1714 0.2857 ... nan nan</div><input id='attrs-53486d7c-9ed9-46ac-9b15-88147e2f82f7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-53486d7c-9ed9-46ac-9b15-88147e2f82f7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-12040866-0394-4958-99dc-331deca6886a' class='xr-var-data-in' type='checkbox'><label for='data-12040866-0394-4958-99dc-331deca6886a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.03703704],\n",
       "       [ 0.17142857],\n",
       "       [ 0.28571429],\n",
       "       ...,\n",
       "       [20.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-1a1ebd79-060c-4a0d-8e93-71e7f1680f34' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1a1ebd79-060c-4a0d-8e93-71e7f1680f34' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b7ee9d09-0373-426c-a24a-267f7906ad25' class='xr-var-data-in' type='checkbox'><label for='data-b7ee9d09-0373-426c-a24a-267f7906ad25' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>2.122e+03 2.904e+03 ... 0.0 0.0</div><input id='attrs-365a3d36-a023-44fb-90fd-4f9c22a831b0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-365a3d36-a023-44fb-90fd-4f9c22a831b0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-055eef67-1df9-44a9-ac31-995b7a8f836e' class='xr-var-data-in' type='checkbox'><label for='data-055eef67-1df9-44a9-ac31-995b7a8f836e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[2122.],\n",
       "       [2904.],\n",
       "       [2032.],\n",
       "       ...,\n",
       "       [  28.],\n",
       "       [   0.],\n",
       "       [   0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>78.59 82.97 96.76 ... 28.0 nan nan</div><input id='attrs-a8706458-e858-4c6c-a8fa-9a241460105e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-a8706458-e858-4c6c-a8fa-9a241460105e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3703923a-4329-45b7-a344-91ab9a848750' class='xr-var-data-in' type='checkbox'><label for='data-3703923a-4329-45b7-a344-91ab9a848750' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[78.59259259],\n",
       "       [82.97142857],\n",
       "       [96.76190476],\n",
       "       ...,\n",
       "       [28.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-e5a015c8-5aad-49be-a9d9-41914d1a55ef' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e5a015c8-5aad-49be-a9d9-41914d1a55ef' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c50e4385-481a-4024-906d-24fd28da069f' class='xr-var-data-in' type='checkbox'><label for='data-c50e4385-481a-4024-906d-24fd28da069f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-fc0c10b6-f993-4e46-83ea-858b146cc777' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-fc0c10b6-f993-4e46-83ea-858b146cc777' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5e1a7a1d-fa75-470a-af5e-e455a5b2b68a' class='xr-var-data-in' type='checkbox'><label for='data-5e1a7a1d-fa75-470a-af5e-e455a5b2b68a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000104217.1&#x27;,\n",
       "       &#x27;ENSMUSG00000033813.15&#x27;, ..., &#x27;ENSMUSG00000102668.1&#x27;,\n",
       "       &#x27;ENSMUSG00000103528.1&#x27;, &#x27;ENSMUSG00000103919.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-4cdb60fc-0abf-41fc-b3ff-6111a39f2b1e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4cdb60fc-0abf-41fc-b3ff-6111a39f2b1e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b38507f9-5576-4bd6-8b3d-b17eb08b37b2' class='xr-var-data-in' type='checkbox'><label for='data-b38507f9-5576-4bd6-8b3d-b17eb08b37b2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000104217.1&#x27;,\n",
       "       &#x27;ENSMUSG00000033813.15&#x27;, ..., &#x27;ENSMUSG00000102668.1&#x27;,\n",
       "       &#x27;ENSMUSG00000103528.1&#x27;, &#x27;ENSMUSG00000103919.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;protein_coding&#x27; ... &#x27;protein_co...</div><input id='attrs-68d8db8f-be87-4226-9c56-99645fc95153' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-68d8db8f-be87-4226-9c56-99645fc95153' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e89e968a-3d93-48ff-9c49-b8db3d291ab8' class='xr-var-data-in' type='checkbox'><label for='data-e89e968a-3d93-48ff-9c49-b8db3d291ab8' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, ...,\n",
       "       &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Lypla1&#x27; &#x27;Gm37988&#x27; ... &#x27;Gm21366&#x27;</div><input id='attrs-e9fb67ce-f3c1-401e-b72f-9e7a6e25b251' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e9fb67ce-f3c1-401e-b72f-9e7a6e25b251' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-cf408d4f-e143-4384-bf74-68caede174eb' class='xr-var-data-in' type='checkbox'><label for='data-cf408d4f-e143-4384-bf74-68caede174eb' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Lypla1&#x27;, &#x27;Gm37988&#x27;, &#x27;Tcea1&#x27;, ..., &#x27;Gm29866&#x27;, &#x27;Gm31571&#x27;, &#x27;Gm21366&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;</div><input id='attrs-883181c9-46fb-42e9-91ee-520928e907a5' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-883181c9-46fb-42e9-91ee-520928e907a5' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d83617cf-ff10-452e-87d8-403e125e7bce' class='xr-var-data-in' type='checkbox'><label for='data-d83617cf-ff10-452e-87d8-403e125e7bce' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;, ..., &#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Mgi_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;MGI:1344588&#x27; ... &#x27;MGI:5434721&#x27;</div><input id='attrs-ca6739b7-4818-4cd7-ab2c-5767fe903cdc' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ca6739b7-4818-4cd7-ab2c-5767fe903cdc' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-da1b6ad6-4b07-4894-85d8-314a0b6dec83' class='xr-var-data-in' type='checkbox'><label for='data-da1b6ad6-4b07-4894-85d8-314a0b6dec83' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;MGI:1344588&#x27;, &#x27;MGI:5611216&#x27;, &#x27;MGI:1196624&#x27;, ..., &#x27;MGI:5589025&#x27;,\n",
       "       &#x27;MGI:5590730&#x27;, &#x27;MGI:5434721&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_gene</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTTM...</div><input id='attrs-c0bc561b-e7e9-416c-a75c-f0499b0c5ad1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c0bc561b-e7e9-416c-a75c-f0499b0c5ad1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-cd7ad736-fa47-48de-883f-9940d4e734ac' class='xr-var-data-in' type='checkbox'><label for='data-cd7ad736-fa47-48de-883f-9940d4e734ac' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;OTTMUSG00000021562.4&#x27;, &#x27;OTTMUSG00000050100.1&#x27;,\n",
       "       &#x27;OTTMUSG00000042348.1&#x27;, ..., &#x27;OTTMUSG00000050180.1&#x27;,\n",
       "       &#x27;OTTMUSG00000050134.1&#x27;, &#x27;OTTMUSG00000050112.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Tag</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;overlapping_locus&#x27; ... &#x27;&#x27;</div><input id='attrs-4f9b5529-973c-4725-9ce6-07c7a2510580' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4f9b5529-973c-4725-9ce6-07c7a2510580' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-55f42200-7322-4ec0-8ae6-3ac9be7a99ec' class='xr-var-data-in' type='checkbox'><label for='data-55f42200-7322-4ec0-8ae6-3ac9be7a99ec' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;overlapping_locus&#x27;, &#x27;overlapping_locus&#x27;, &#x27;overlapping_locus&#x27;, ...,\n",
       "       &#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUST00000027036.10&#x27; ... &#x27;ENS...</div><input id='attrs-20147b88-ea26-4b0f-a9b7-8fbe810ecaf3' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-20147b88-ea26-4b0f-a9b7-8fbe810ecaf3' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e1101dce-000c-4d11-9469-1aff67792bac' class='xr-var-data-in' type='checkbox'><label for='data-e1101dce-000c-4d11-9469-1aff67792bac' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUST00000027036.10&#x27;, &#x27;ENSMUST00000155020.1&#x27;,\n",
       "       &#x27;ENSMUST00000081551.13&#x27;, ..., &#x27;ENSMUST00000191675.1&#x27;,\n",
       "       &#x27;ENSMUST00000192521.1&#x27;, &#x27;ENSMUST00000192892.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Region</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;before_tss&#x27; ... &#x27;first_introns&#x27;</div><input id='attrs-b682ae1f-43f5-4935-a831-44e8ce4474dc' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b682ae1f-43f5-4935-a831-44e8ce4474dc' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3ff4110d-010d-4f0f-9cf8-2b9fc0bc8a01' class='xr-var-data-in' type='checkbox'><label for='data-3ff4110d-010d-4f0f-9cf8-2b9fc0bc8a01' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;before_tss&#x27;, &#x27;before_tss&#x27;, &#x27;before_tss&#x27;, ..., &#x27;first_introns&#x27;,\n",
       "       &#x27;first_introns&#x27;, &#x27;first_introns&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;</div><input id='attrs-38a2b719-e66d-4d16-a03a-b1c3e3a88926' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-38a2b719-e66d-4d16-a03a-b1c3e3a88926' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e818933e-28d3-4c25-8935-7f646b429c74' class='xr-var-data-in' type='checkbox'><label for='data-e818933e-28d3-4c25-8935-7f646b429c74' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;, ..., &#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Parent</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-4630c39e-60d0-4303-983f-cc58f4e1afd0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4630c39e-60d0-4303-983f-cc58f4e1afd0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ee4c3ab6-3ac8-4d17-b610-8f7708cc54da' class='xr-var-data-in' type='checkbox'><label for='data-ee4c3ab6-3ac8-4d17-b610-8f7708cc54da' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-9bde9990-5c6d-43f5-84cd-8ad953f5d9b5' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-9bde9990-5c6d-43f5-84cd-8ad953f5d9b5' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-31dd80be-92d4-4f59-9835-00e16b12daf9' class='xr-var-data-in' type='checkbox'><label for='data-31dd80be-92d4-4f59-9835-00e16b12daf9' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-b0620d9d-ea9d-4f79-ac47-79401fd1a27f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b0620d9d-ea9d-4f79-ac47-79401fd1a27f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-426a0d21-0b35-4a48-b548-df9e1937b930' class='xr-var-data-in' type='checkbox'><label for='data-426a0d21-0b35-4a48-b548-df9e1937b930' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-ef252f06-4bdf-4a84-9029-bbb2064d26cc' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ef252f06-4bdf-4a84-9029-bbb2064d26cc' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7e264f74-b0dc-42d8-9fe3-b04eb9e5c253' class='xr-var-data-in' type='checkbox'><label for='data-7e264f74-b0dc-42d8-9fe3-b04eb9e5c253' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-adab8c6f-2c19-4b88-8278-97f17668da3a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-adab8c6f-2c19-4b88-8278-97f17668da3a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4e16c0f3-1042-4fc2-a576-c43ea2300666' class='xr-var-data-in' type='checkbox'><label for='data-4e16c0f3-1042-4fc2-a576-c43ea2300666' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_support_level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-c994920a-20d1-4086-926c-d6247e1019aa' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c994920a-20d1-4086-926c-d6247e1019aa' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-19b732d9-be55-4491-b413-7df5c6c77653' class='xr-var-data-in' type='checkbox'><label for='data-19b732d9-be55-4491-b413-7df5c6c77653' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_transcript</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-deaaaff6-ce67-4d00-bbc5-09cdfce0c438' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-deaaaff6-ce67-4d00-bbc5-09cdfce0c438' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8a1034fb-6ca8-4efc-b4cb-b6de77953f18' class='xr-var-data-in' type='checkbox'><label for='data-8a1034fb-6ca8-4efc-b4cb-b6de77953f18' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Protein_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-0fc3f0d7-f9db-444e-a96c-6ce35d901c1f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0fc3f0d7-f9db-444e-a96c-6ce35d901c1f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7ca5f40c-f7f5-4606-9a8d-80368cfeba3e' class='xr-var-data-in' type='checkbox'><label for='data-7ca5f40c-f7f5-4606-9a8d-80368cfeba3e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ccdsid</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-99b4debf-5b4f-4166-af53-edcd1b4dc78e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-99b4debf-5b4f-4166-af53-edcd1b4dc78e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-05317126-f9b4-4f1d-8a38-04a166006390' class='xr-var-data-in' type='checkbox'><label for='data-05317126-f9b4-4f1d-8a38-04a166006390' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ont</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-6d03d8de-fa61-4a12-9e11-af9857f190dd' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6d03d8de-fa61-4a12-9e11-af9857f190dd' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-15269cf6-9e17-4e45-bc55-90ffa53c77af' class='xr-var-data-in' type='checkbox'><label for='data-15269cf6-9e17-4e45-bc55-90ffa53c77af' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level_1</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... 8.0 8.0 8.0 8.0</div><input id='attrs-99eb0fa3-0cfc-4321-b6cd-f4cf2890a19a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-99eb0fa3-0cfc-4321-b6cd-f4cf2890a19a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-31ac69f2-19ec-4e14-b35b-ff6976efdc72' class='xr-var-data-in' type='checkbox'><label for='data-31ac69f2-19ec-4e14-b35b-ff6976efdc72' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ...,  8.,  8.,  8.])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ranges_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>nan nan nan ... 2.164e+04 2.164e+04</div><input id='attrs-d57bbe97-fdc1-404c-a1a9-41d87ae2043d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d57bbe97-fdc1-404c-a1a9-41d87ae2043d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-45f6fa60-3194-4c0a-b009-bbee41a7585e' class='xr-var-data-in' type='checkbox'><label for='data-45f6fa60-3194-4c0a-b009-bbee41a7585e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([   nan,    nan,    nan, ..., 21635., 21636., 21637.])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Intron_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;</div><input id='attrs-f5397743-5c26-49d1-a77d-baf334346e7b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f5397743-5c26-49d1-a77d-baf334346e7b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0c3269c0-57a4-4cdc-967e-92cd5b2f3091' class='xr-var-data-in' type='checkbox'><label for='data-0c3269c0-57a4-4cdc-967e-92cd5b2f3091' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., &#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_mc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0006887 0.001969 ... nan nan</div><input id='attrs-eb3876c4-d68a-4df7-bbc0-e40090ce7ff9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-eb3876c4-d68a-4df7-bbc0-e40090ce7ff9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-889d091d-1a0a-44f6-9a84-d51e4cb3346a' class='xr-var-data-in' type='checkbox'><label for='data-889d091d-1a0a-44f6-9a84-d51e4cb3346a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.        ],\n",
       "       [0.00068871],\n",
       "       [0.0019685 ],\n",
       "       ...,\n",
       "       [0.67857143],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_hmc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0004713 0.001377 ... nan nan</div><input id='attrs-fb0456bb-0586-4e83-9382-653259074a38' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-fb0456bb-0586-4e83-9382-653259074a38' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d96a5299-3383-4016-a840-28c82dc5c187' class='xr-var-data-in' type='checkbox'><label for='data-d96a5299-3383-4016-a840-28c82dc5c187' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.00047125],\n",
       "       [0.00137741],\n",
       "       [0.00098425],\n",
       "       ...,\n",
       "       [0.03571429],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_modc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0004713 0.002066 ... nan nan</div><input id='attrs-2dde4b9a-acab-4d19-961a-2feca0c87390' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2dde4b9a-acab-4d19-961a-2feca0c87390' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-37106b97-3c96-4ea0-9a1f-52815017c9e0' class='xr-var-data-in' type='checkbox'><label for='data-37106b97-3c96-4ea0-9a1f-52815017c9e0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[4.71253534e-04],\n",
       "       [2.06611570e-03],\n",
       "       [2.95275591e-03],\n",
       "       ...,\n",
       "       [7.14285714e-01],\n",
       "       [           nan],\n",
       "       [           nan]])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-7efc9c75-dd1e-48c3-b1d1-6ee399e99cfe' class='xr-section-summary-in' type='checkbox'  ><label for='section-7efc9c75-dd1e-48c3-b1d1-6ee399e99cfe' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>sample_id</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-79e7cb09-25ff-4745-be4b-71f661903cc7' class='xr-index-data-in' type='checkbox'/><label for='index-79e7cb09-25ff-4745-be4b-71f661903cc7' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;sample_0&#x27;], dtype=&#x27;object&#x27;, name=&#x27;sample_id&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-c30fbf29-a261-439c-99fc-756c57b99815' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-c30fbf29-a261-439c-99fc-756c57b99815' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.Dataset>\n",
       "Dimensions:                   (ranges: 106562, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) <U2 '1' '1' '1' '1' ... 'Y' 'Y' 'Y' 'Y'\n",
       "    start                     (ranges) int64 4807587 4807691 ... 22859847\n",
       "    end                       (ranges) int64 4807787 4807891 ... 22861212\n",
       "    range_id                  (ranges) int64 0 1 2 3 ... 53412 53413 53414 53415\n",
       "    num_contexts              (ranges) int64 27 35 21 0 7 2 11 ... 0 1 0 1 1 0 0\n",
       "    range_length              (ranges) int64 200 200 200 200 ... 1361 1365 1365\n",
       "  * sample_id                 (sample_id) <U8 'sample_0'\n",
       "    strand                    (ranges) <U1 '+' '+' '+' '+' ... '-' '-' '-' '-'\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 2.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 0.05714 ... nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 1.0 4.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.03704 0.1143 ... nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 1.0 6.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.03704 0.1714 ... nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 2.122e+03 ... 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 78.59 82.97 ... nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    Id                        (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Gene_id                   (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Gene_type                 (ranges) object 'protein_coding' ... 'protein_c...\n",
       "    Gene_name                 (ranges) object 'Lypla1' 'Gm37988' ... 'Gm21366'\n",
       "    Level                     (ranges) object '2' '2' '2' '2' ... '2' '2' '2'\n",
       "    Mgi_id                    (ranges) object 'MGI:1344588' ... 'MGI:5434721'\n",
       "    Havana_gene               (ranges) object 'OTTMUSG00000021562.4' ... 'OTT...\n",
       "    Tag                       (ranges) object 'overlapping_locus' ... ''\n",
       "    Transcript_id             (ranges) object 'ENSMUST00000027036.10' ... 'EN...\n",
       "    Region                    (ranges) object 'before_tss' ... 'first_introns'\n",
       "    Strand                    (ranges) object '.' '.' '.' '.' ... '.' '.' '.'\n",
       "    Parent                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_type           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_name           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_number               (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_id                   (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_support_level  (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Havana_transcript         (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Protein_id                (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ccdsid                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ont                       (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Level_1                   (ranges) float64 nan nan nan nan ... 8.0 8.0 8.0\n",
       "    Ranges_id                 (ranges) float64 nan nan ... 2.164e+04 2.164e+04\n",
       "    Intron_number             (ranges) object nan nan nan nan ... '1' '1' '1'\n",
       "    mean_mc                   (ranges, sample_id) float64 0.0 0.0006887 ... nan\n",
       "    mean_hmc                  (ranges, sample_id) float64 0.0004713 ... nan\n",
       "    mean_modc                 (ranges, sample_id) float64 0.0004713 ... nan"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compute mean methylation levels as mc / total_c\n",
    "sense_ds = sense_ds.assign(\n",
    "    mean_mc = sense_ds[\"num_mc_sum\"] / sense_ds[\"num_total_c_sum\"],\n",
    "    mean_hmc = sense_ds[\"num_hmc_sum\"] / sense_ds[\"num_total_c_sum\"],\n",
    "    mean_modc = sense_ds[\"num_modc_sum\"] / sense_ds[\"num_total_c_sum\"],\n",
    ")\n",
    "sense_ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;\n",
       "Dimensions:                   (ranges: 106562, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) &lt;U2 &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;\n",
       "    start                     (ranges) int64 4807587 4807691 ... 22859847\n",
       "    end                       (ranges) int64 4807787 4807891 ... 22861212\n",
       "    range_id                  (ranges) int64 0 1 2 3 ... 53412 53413 53414 53415\n",
       "    num_contexts              (ranges) int64 27 35 22 0 7 2 11 ... 0 3 0 1 1 0 0\n",
       "    range_length              (ranges) int64 200 200 200 200 ... 1361 1365 1365\n",
       "  * sample_id                 (sample_id) &lt;U8 &#x27;sample_0&#x27;\n",
       "    strand                    (ranges) &lt;U1 &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 1.0 4.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.03704 0.1143 ... nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 3.0 4.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.1111 0.1143 ... nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 4.0 8.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.1481 0.2286 ... nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 2.111e+03 ... 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 78.19 74.09 ... nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    Id                        (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Gene_id                   (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Gene_type                 (ranges) object &#x27;protein_coding&#x27; ... &#x27;protein_c...\n",
       "    Gene_name                 (ranges) object &#x27;Lypla1&#x27; &#x27;Gm37988&#x27; ... &#x27;Gm21366&#x27;\n",
       "    Level                     (ranges) object &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;\n",
       "    Mgi_id                    (ranges) object &#x27;MGI:1344588&#x27; ... &#x27;MGI:5434721&#x27;\n",
       "    Havana_gene               (ranges) object &#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTT...\n",
       "    Tag                       (ranges) object &#x27;overlapping_locus&#x27; ... &#x27;&#x27;\n",
       "    Transcript_id             (ranges) object &#x27;ENSMUST00000027036.10&#x27; ... &#x27;EN...\n",
       "    Region                    (ranges) object &#x27;before_tss&#x27; ... &#x27;first_introns&#x27;\n",
       "    Strand                    (ranges) object &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;\n",
       "    Parent                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_type           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_name           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_number               (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_id                   (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_support_level  (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Havana_transcript         (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Protein_id                (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ccdsid                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ont                       (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Level_1                   (ranges) float64 nan nan nan nan ... 8.0 8.0 8.0\n",
       "    Ranges_id                 (ranges) float64 nan nan ... 2.164e+04 2.164e+04\n",
       "    Intron_number             (ranges) object nan nan nan nan ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;\n",
       "    mean_mc                   (ranges, sample_id) float64 0.0004737 ... nan\n",
       "    mean_hmc                  (ranges, sample_id) float64 0.001421 ... nan\n",
       "    mean_modc                 (ranges, sample_id) float64 0.001895 ... nan</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-3b64a4a4-8d52-469f-bba5-a7683faf4e7e' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-3b64a4a4-8d52-469f-bba5-a7683faf4e7e' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>ranges</span>: 106562</li><li><span class='xr-has-index'>sample_id</span>: 1</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-abae1e23-c783-4a82-8770-f60e517f6a77' class='xr-section-summary-in' type='checkbox'  checked><label for='section-abae1e23-c783-4a82-8770-f60e517f6a77' class='xr-section-summary' >Coordinates: <span>(8)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>contig</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U2</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;</div><input id='attrs-f675b9a9-2afe-4057-b31e-cebf5a1d52e6' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f675b9a9-2afe-4057-b31e-cebf5a1d52e6' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3bf78118-6eef-4f22-b646-7d6cfbb68f81' class='xr-var-data-in' type='checkbox'><label for='data-3bf78118-6eef-4f22-b646-7d6cfbb68f81' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;, ..., &#x27;Y&#x27;, &#x27;Y&#x27;, &#x27;Y&#x27;], dtype=&#x27;&lt;U2&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>start</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4807587 4807691 ... 22859847</div><input id='attrs-af164497-ebd8-4240-948c-fa24f4a6a49e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-af164497-ebd8-4240-948c-fa24f4a6a49e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-6cb06a7b-0a61-4d6d-bf5a-b776bc8d5796' class='xr-var-data-in' type='checkbox'><label for='data-6cb06a7b-0a61-4d6d-bf5a-b776bc8d5796' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4807587,  4807691,  4857613, ..., 20992252, 22049944, 22859847])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>end</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4807787 4807891 ... 22861212</div><input id='attrs-6870d5a8-0cab-4549-a062-3c1c37f42123' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6870d5a8-0cab-4549-a062-3c1c37f42123' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9b3c3ee1-504a-4b71-85fb-8ac281f8f5c6' class='xr-var-data-in' type='checkbox'><label for='data-9b3c3ee1-504a-4b71-85fb-8ac281f8f5c6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4807787,  4807891,  4857813, ..., 20993613, 22051309, 22861212])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 ... 53412 53413 53414 53415</div><input id='attrs-30dd120c-e23c-4fdf-896e-96a181c04286' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-30dd120c-e23c-4fdf-896e-96a181c04286' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7b16d27d-68fb-4d76-af1c-fba3681d0d5d' class='xr-var-data-in' type='checkbox'><label for='data-7b16d27d-68fb-4d76-af1c-fba3681d0d5d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([    0,     1,     2, ..., 53413, 53414, 53415])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_contexts</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>27 35 22 0 7 2 11 ... 0 3 0 1 1 0 0</div><input id='attrs-e3a54bc6-c5e8-4020-8fa2-5e187e08e675' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e3a54bc6-c5e8-4020-8fa2-5e187e08e675' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-dd9e81f3-76cc-4607-9bcb-b009484c253e' class='xr-var-data-in' type='checkbox'><label for='data-dd9e81f3-76cc-4607-9bcb-b009484c253e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([27, 35, 22, ...,  1,  0,  0])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_length</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>200 200 200 200 ... 1361 1365 1365</div><input id='attrs-a78122aa-6915-4a14-9fe3-2a4194a1648a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-a78122aa-6915-4a14-9fe3-2a4194a1648a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-303b077b-b787-4f74-8ef7-7a17288828ce' class='xr-var-data-in' type='checkbox'><label for='data-303b077b-b787-4f74-8ef7-7a17288828ce' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 200,  200,  200, ..., 1361, 1365, 1365])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample_id</span></div><div class='xr-var-dims'>(sample_id)</div><div class='xr-var-dtype'>&lt;U8</div><div class='xr-var-preview xr-preview'>&#x27;sample_0&#x27;</div><input id='attrs-022d8a6c-2648-4514-a9d6-c02a0a65e4ba' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-022d8a6c-2648-4514-a9d6-c02a0a65e4ba' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-cc8bd16f-adb6-40f2-9367-a3b5c48131e6' class='xr-var-data-in' type='checkbox'><label for='data-cc8bd16f-adb6-40f2-9367-a3b5c48131e6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;sample_0&#x27;], dtype=&#x27;&lt;U8&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U1</div><div class='xr-var-preview xr-preview'>&#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;</div><input id='attrs-149eb246-9394-40ff-9f3d-f1376e844572' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-149eb246-9394-40ff-9f3d-f1376e844572' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d77d93c2-cd7e-4979-bfd8-b0e9f0ec556c' class='xr-var-data-in' type='checkbox'><label for='data-d77d93c2-cd7e-4979-bfd8-b0e9f0ec556c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;+&#x27;, &#x27;+&#x27;, &#x27;+&#x27;, ..., &#x27;-&#x27;, &#x27;-&#x27;, &#x27;-&#x27;], dtype=&#x27;&lt;U1&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-aec70633-88da-4f01-a805-fa29a4c6214f' class='xr-section-summary-in' type='checkbox'  ><label for='section-aec70633-88da-4f01-a805-fa29a4c6214f' class='xr-section-summary' >Data variables: <span>(39)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 4.0 3.0 0.0 ... 19.0 0.0 0.0</div><input id='attrs-b75c49ca-06fb-49a3-8572-592e2705e69d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b75c49ca-06fb-49a3-8572-592e2705e69d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ed77a065-d8ce-4b92-8e0c-0ec877cf0f9e' class='xr-var-data-in' type='checkbox'><label for='data-ed77a065-d8ce-4b92-8e0c-0ec877cf0f9e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 1.],\n",
       "       [ 4.],\n",
       "       [ 3.],\n",
       "       ...,\n",
       "       [19.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.03704 0.1143 0.1364 ... nan nan</div><input id='attrs-b18d2886-bf51-4897-bab4-f3e8ad86397e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b18d2886-bf51-4897-bab4-f3e8ad86397e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-eda27542-ccd6-4df9-9017-8ebe9c2e7861' class='xr-var-data-in' type='checkbox'><label for='data-eda27542-ccd6-4df9-9017-8ebe9c2e7861' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.03703704],\n",
       "       [ 0.11428571],\n",
       "       [ 0.13636364],\n",
       "       ...,\n",
       "       [19.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 22.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-46a4852e-a07a-4342-b036-e5b7775c5e2b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-46a4852e-a07a-4342-b036-e5b7775c5e2b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0effa668-6985-4b3d-8f04-181ffa33d833' class='xr-var-data-in' type='checkbox'><label for='data-0effa668-6985-4b3d-8f04-181ffa33d833' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [22.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>3.0 4.0 1.0 0.0 ... 0.0 1.0 0.0 0.0</div><input id='attrs-a161f7eb-ed6d-4685-a3db-1148d153220b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-a161f7eb-ed6d-4685-a3db-1148d153220b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-bb870405-e1de-40ca-9e47-0f73b06fb7dc' class='xr-var-data-in' type='checkbox'><label for='data-bb870405-e1de-40ca-9e47-0f73b06fb7dc' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[3.],\n",
       "       [4.],\n",
       "       [1.],\n",
       "       ...,\n",
       "       [1.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.1111 0.1143 0.04545 ... nan nan</div><input id='attrs-4d77e6b3-5d1f-425d-88a4-10d4abf3cf84' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4d77e6b3-5d1f-425d-88a4-10d4abf3cf84' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c11d0a35-822b-43d9-991e-e91496fead05' class='xr-var-data-in' type='checkbox'><label for='data-c11d0a35-822b-43d9-991e-e91496fead05' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.11111111],\n",
       "       [0.11428571],\n",
       "       [0.04545455],\n",
       "       ...,\n",
       "       [1.        ],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 22.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-1a8a4cf9-4e9d-4338-89c8-f65cb1129222' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1a8a4cf9-4e9d-4338-89c8-f65cb1129222' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-16f1bbe1-7447-4afa-99e2-617cedfc9bc1' class='xr-var-data-in' type='checkbox'><label for='data-16f1bbe1-7447-4afa-99e2-617cedfc9bc1' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [22.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>4.0 8.0 4.0 0.0 ... 21.0 0.0 0.0</div><input id='attrs-88c76f11-fa1f-42de-adbf-12769dc9a557' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-88c76f11-fa1f-42de-adbf-12769dc9a557' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-09c62026-058a-4a81-9444-13408a702d4a' class='xr-var-data-in' type='checkbox'><label for='data-09c62026-058a-4a81-9444-13408a702d4a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 4.],\n",
       "       [ 8.],\n",
       "       [ 4.],\n",
       "       ...,\n",
       "       [21.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.1481 0.2286 0.1818 ... nan nan</div><input id='attrs-ee4f76af-0218-4230-9b05-92119aefa30c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ee4f76af-0218-4230-9b05-92119aefa30c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-de3264b9-0b7f-4af8-9845-c4d49cc456d5' class='xr-var-data-in' type='checkbox'><label for='data-de3264b9-0b7f-4af8-9845-c4d49cc456d5' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.14814815],\n",
       "       [ 0.22857143],\n",
       "       [ 0.18181818],\n",
       "       ...,\n",
       "       [21.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 22.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-b39e9ffa-d1de-44c0-bf93-aa6a6ec8e4a0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b39e9ffa-d1de-44c0-bf93-aa6a6ec8e4a0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-34f8f6b2-4431-4af0-9198-29d7a15f01fc' class='xr-var-data-in' type='checkbox'><label for='data-34f8f6b2-4431-4af0-9198-29d7a15f01fc' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [22.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>2.111e+03 2.593e+03 ... 0.0 0.0</div><input id='attrs-a666bd7f-33be-4dd8-8540-7e66211d796b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-a666bd7f-33be-4dd8-8540-7e66211d796b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-446f29a3-283d-4a21-a0bc-301e6f9c8d8f' class='xr-var-data-in' type='checkbox'><label for='data-446f29a3-283d-4a21-a0bc-301e6f9c8d8f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[2111.],\n",
       "       [2593.],\n",
       "       [1982.],\n",
       "       ...,\n",
       "       [  29.],\n",
       "       [   0.],\n",
       "       [   0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>78.19 74.09 90.09 ... 29.0 nan nan</div><input id='attrs-8a313dcc-9e6d-4aea-973c-722ea99d56f8' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8a313dcc-9e6d-4aea-973c-722ea99d56f8' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b86481a7-9ba3-41fd-ae6c-63672911ed55' class='xr-var-data-in' type='checkbox'><label for='data-b86481a7-9ba3-41fd-ae6c-63672911ed55' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[78.18518519],\n",
       "       [74.08571429],\n",
       "       [90.09090909],\n",
       "       ...,\n",
       "       [29.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 22.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-2423ee02-0282-4270-8794-ecf7d46cdcb1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2423ee02-0282-4270-8794-ecf7d46cdcb1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-148d8163-4fc5-43b7-b3cd-d45a3319acbd' class='xr-var-data-in' type='checkbox'><label for='data-148d8163-4fc5-43b7-b3cd-d45a3319acbd' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [22.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-96db8f82-3cf1-42ea-8219-2d5b85c3e582' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-96db8f82-3cf1-42ea-8219-2d5b85c3e582' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a5322f0a-1386-4cbf-948d-2260cc45b0b4' class='xr-var-data-in' type='checkbox'><label for='data-a5322f0a-1386-4cbf-948d-2260cc45b0b4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000104217.1&#x27;,\n",
       "       &#x27;ENSMUSG00000033813.15&#x27;, ..., &#x27;ENSMUSG00000102668.1&#x27;,\n",
       "       &#x27;ENSMUSG00000103528.1&#x27;, &#x27;ENSMUSG00000103919.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-04193d01-9c53-4c5c-9950-a933be919f99' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-04193d01-9c53-4c5c-9950-a933be919f99' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-2aa6b50e-5794-488a-b880-c32cafdae779' class='xr-var-data-in' type='checkbox'><label for='data-2aa6b50e-5794-488a-b880-c32cafdae779' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000104217.1&#x27;,\n",
       "       &#x27;ENSMUSG00000033813.15&#x27;, ..., &#x27;ENSMUSG00000102668.1&#x27;,\n",
       "       &#x27;ENSMUSG00000103528.1&#x27;, &#x27;ENSMUSG00000103919.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;protein_coding&#x27; ... &#x27;protein_co...</div><input id='attrs-35500636-7816-4635-bf7e-fefc67a55f81' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-35500636-7816-4635-bf7e-fefc67a55f81' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-af740fdb-39cd-4eb3-bf3a-e174d3501b10' class='xr-var-data-in' type='checkbox'><label for='data-af740fdb-39cd-4eb3-bf3a-e174d3501b10' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, ...,\n",
       "       &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Lypla1&#x27; &#x27;Gm37988&#x27; ... &#x27;Gm21366&#x27;</div><input id='attrs-271bb459-2713-440e-9b14-d671fb627576' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-271bb459-2713-440e-9b14-d671fb627576' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d80b4cc4-6e9c-461a-a880-d6b75bebb38c' class='xr-var-data-in' type='checkbox'><label for='data-d80b4cc4-6e9c-461a-a880-d6b75bebb38c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Lypla1&#x27;, &#x27;Gm37988&#x27;, &#x27;Tcea1&#x27;, ..., &#x27;Gm29866&#x27;, &#x27;Gm31571&#x27;, &#x27;Gm21366&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;</div><input id='attrs-61bdea9a-be90-4445-a208-9b23829d9af7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-61bdea9a-be90-4445-a208-9b23829d9af7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-89c5bb5a-57e7-4e0c-bd00-e648c3860429' class='xr-var-data-in' type='checkbox'><label for='data-89c5bb5a-57e7-4e0c-bd00-e648c3860429' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;, ..., &#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Mgi_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;MGI:1344588&#x27; ... &#x27;MGI:5434721&#x27;</div><input id='attrs-55d0f94c-ce52-4d8e-bad4-d652a9999c7c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-55d0f94c-ce52-4d8e-bad4-d652a9999c7c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-caaabd7b-f668-4ec5-a472-f0b4e1b5d875' class='xr-var-data-in' type='checkbox'><label for='data-caaabd7b-f668-4ec5-a472-f0b4e1b5d875' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;MGI:1344588&#x27;, &#x27;MGI:5611216&#x27;, &#x27;MGI:1196624&#x27;, ..., &#x27;MGI:5589025&#x27;,\n",
       "       &#x27;MGI:5590730&#x27;, &#x27;MGI:5434721&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_gene</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTTM...</div><input id='attrs-41d4c1ca-e9e6-49eb-8f5e-ae11c790ab16' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-41d4c1ca-e9e6-49eb-8f5e-ae11c790ab16' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c9b02ee2-3c5f-4bcd-852b-80097a83ff30' class='xr-var-data-in' type='checkbox'><label for='data-c9b02ee2-3c5f-4bcd-852b-80097a83ff30' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;OTTMUSG00000021562.4&#x27;, &#x27;OTTMUSG00000050100.1&#x27;,\n",
       "       &#x27;OTTMUSG00000042348.1&#x27;, ..., &#x27;OTTMUSG00000050180.1&#x27;,\n",
       "       &#x27;OTTMUSG00000050134.1&#x27;, &#x27;OTTMUSG00000050112.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Tag</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;overlapping_locus&#x27; ... &#x27;&#x27;</div><input id='attrs-3650da46-76a2-4f0c-bf69-d169b043ec7a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3650da46-76a2-4f0c-bf69-d169b043ec7a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-01c728c9-e1e1-4080-8a1d-ed79f8550e4f' class='xr-var-data-in' type='checkbox'><label for='data-01c728c9-e1e1-4080-8a1d-ed79f8550e4f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;overlapping_locus&#x27;, &#x27;overlapping_locus&#x27;, &#x27;overlapping_locus&#x27;, ...,\n",
       "       &#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUST00000027036.10&#x27; ... &#x27;ENS...</div><input id='attrs-a0dae531-f4bd-43f5-a759-b2f22158533a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-a0dae531-f4bd-43f5-a759-b2f22158533a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-cce40f20-711a-45ed-82f3-4220f132754c' class='xr-var-data-in' type='checkbox'><label for='data-cce40f20-711a-45ed-82f3-4220f132754c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUST00000027036.10&#x27;, &#x27;ENSMUST00000155020.1&#x27;,\n",
       "       &#x27;ENSMUST00000081551.13&#x27;, ..., &#x27;ENSMUST00000191675.1&#x27;,\n",
       "       &#x27;ENSMUST00000192521.1&#x27;, &#x27;ENSMUST00000192892.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Region</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;before_tss&#x27; ... &#x27;first_introns&#x27;</div><input id='attrs-310cffe7-0c36-4487-b6c2-371880d33165' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-310cffe7-0c36-4487-b6c2-371880d33165' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7cfd756a-aa6f-427c-8737-62f7239dc926' class='xr-var-data-in' type='checkbox'><label for='data-7cfd756a-aa6f-427c-8737-62f7239dc926' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;before_tss&#x27;, &#x27;before_tss&#x27;, &#x27;before_tss&#x27;, ..., &#x27;first_introns&#x27;,\n",
       "       &#x27;first_introns&#x27;, &#x27;first_introns&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;</div><input id='attrs-3300606c-42ea-4226-a536-1174e5ebdac3' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3300606c-42ea-4226-a536-1174e5ebdac3' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-51bda875-cf63-4951-8fa9-1b0125b24ca9' class='xr-var-data-in' type='checkbox'><label for='data-51bda875-cf63-4951-8fa9-1b0125b24ca9' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;, ..., &#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Parent</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-4f831866-b6c2-43c2-9b0e-5cd35812cbbd' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4f831866-b6c2-43c2-9b0e-5cd35812cbbd' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-f1889465-e506-45a1-8a95-ea7d9b6709ee' class='xr-var-data-in' type='checkbox'><label for='data-f1889465-e506-45a1-8a95-ea7d9b6709ee' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-f4f3178e-194e-46ac-8a88-20e5170cbabc' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f4f3178e-194e-46ac-8a88-20e5170cbabc' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-27e4de55-c3f1-4520-9165-690651789668' class='xr-var-data-in' type='checkbox'><label for='data-27e4de55-c3f1-4520-9165-690651789668' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-14d7c352-56ec-4b3b-9c92-c50815ebe8d5' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-14d7c352-56ec-4b3b-9c92-c50815ebe8d5' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-cea98e6e-1372-4aed-8377-c5a4a7eb71d6' class='xr-var-data-in' type='checkbox'><label for='data-cea98e6e-1372-4aed-8377-c5a4a7eb71d6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-452ae13c-60e7-4aa0-bef1-7b585e339955' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-452ae13c-60e7-4aa0-bef1-7b585e339955' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3f3fb4cf-b8b2-4ce5-8b6b-f2df12b502de' class='xr-var-data-in' type='checkbox'><label for='data-3f3fb4cf-b8b2-4ce5-8b6b-f2df12b502de' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-7276a40d-1b50-4259-9e5c-bf9dc5aeb184' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7276a40d-1b50-4259-9e5c-bf9dc5aeb184' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ffa9535d-55e6-4b5d-9d64-93490f79a9f4' class='xr-var-data-in' type='checkbox'><label for='data-ffa9535d-55e6-4b5d-9d64-93490f79a9f4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_support_level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-87be6e3a-ee97-417c-b25c-d337e605dec0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-87be6e3a-ee97-417c-b25c-d337e605dec0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-181a6c6a-6cd6-4ef8-8404-daafbb0c8cc4' class='xr-var-data-in' type='checkbox'><label for='data-181a6c6a-6cd6-4ef8-8404-daafbb0c8cc4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_transcript</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-322a3ddf-c46c-4911-b1c5-6bbb637f81cb' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-322a3ddf-c46c-4911-b1c5-6bbb637f81cb' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-dc556180-ff09-4fe8-95a3-b739c35736da' class='xr-var-data-in' type='checkbox'><label for='data-dc556180-ff09-4fe8-95a3-b739c35736da' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Protein_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-2669ae34-cde7-4e81-87c3-cf8411bc7878' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2669ae34-cde7-4e81-87c3-cf8411bc7878' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-407b090f-c3f1-4075-9e14-0ea9b0a873b4' class='xr-var-data-in' type='checkbox'><label for='data-407b090f-c3f1-4075-9e14-0ea9b0a873b4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ccdsid</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-33eea375-9a70-49cf-b7ea-82259bcfb5dd' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-33eea375-9a70-49cf-b7ea-82259bcfb5dd' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c938eec4-9033-49a3-b6a4-395b0d69018b' class='xr-var-data-in' type='checkbox'><label for='data-c938eec4-9033-49a3-b6a4-395b0d69018b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ont</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-6dad3d43-57bd-4555-a221-347d4d64b563' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6dad3d43-57bd-4555-a221-347d4d64b563' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0ee43b83-bc9f-4af7-9fb8-29351a55034f' class='xr-var-data-in' type='checkbox'><label for='data-0ee43b83-bc9f-4af7-9fb8-29351a55034f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level_1</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... 8.0 8.0 8.0 8.0</div><input id='attrs-e9d0f872-0660-4efe-b478-4dbe7217ee16' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e9d0f872-0660-4efe-b478-4dbe7217ee16' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-31ce054a-299e-4e4b-ad69-0aff070ac158' class='xr-var-data-in' type='checkbox'><label for='data-31ce054a-299e-4e4b-ad69-0aff070ac158' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ...,  8.,  8.,  8.])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ranges_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>nan nan nan ... 2.164e+04 2.164e+04</div><input id='attrs-428efec1-3521-437b-9b75-a1f890ab8e87' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-428efec1-3521-437b-9b75-a1f890ab8e87' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3b2d79e6-617f-484d-937f-ac67ee31a1e6' class='xr-var-data-in' type='checkbox'><label for='data-3b2d79e6-617f-484d-937f-ac67ee31a1e6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([   nan,    nan,    nan, ..., 21635., 21636., 21637.])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Intron_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;</div><input id='attrs-ba4d0072-dd81-4042-8dd2-876f389a62cb' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ba4d0072-dd81-4042-8dd2-876f389a62cb' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5b89a6d4-da98-4da7-ae74-d86d2f2ee6ef' class='xr-var-data-in' type='checkbox'><label for='data-5b89a6d4-da98-4da7-ae74-d86d2f2ee6ef' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., &#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_mc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0004737 0.001543 ... nan nan</div><input id='attrs-e6283395-0c3d-4e1c-904b-07b6bd74afc9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e6283395-0c3d-4e1c-904b-07b6bd74afc9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0db6a075-5975-4170-8691-74e2a32fb666' class='xr-var-data-in' type='checkbox'><label for='data-0db6a075-5975-4170-8691-74e2a32fb666' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[4.73709143e-04],\n",
       "       [1.54261473e-03],\n",
       "       [1.51362260e-03],\n",
       "       ...,\n",
       "       [6.55172414e-01],\n",
       "       [           nan],\n",
       "       [           nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_hmc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.001421 0.001543 ... nan nan</div><input id='attrs-51e0987a-a9b6-430b-8899-4f6c8fa18234' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-51e0987a-a9b6-430b-8899-4f6c8fa18234' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-df6e7b6f-9122-4217-8b5d-51e5d660d28b' class='xr-var-data-in' type='checkbox'><label for='data-df6e7b6f-9122-4217-8b5d-51e5d660d28b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.00142113],\n",
       "       [0.00154261],\n",
       "       [0.00050454],\n",
       "       ...,\n",
       "       [0.03448276],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_modc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.001895 0.003085 ... nan nan</div><input id='attrs-85b9ac3a-2c50-4598-a028-7fea33165936' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-85b9ac3a-2c50-4598-a028-7fea33165936' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-07f6e951-caac-4f3f-bf0e-573ed4ecf46b' class='xr-var-data-in' type='checkbox'><label for='data-07f6e951-caac-4f3f-bf0e-573ed4ecf46b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.00189484],\n",
       "       [0.00308523],\n",
       "       [0.00201816],\n",
       "       ...,\n",
       "       [0.72413793],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-3c7f72d7-dec0-4982-947a-6637b81618ba' class='xr-section-summary-in' type='checkbox'  ><label for='section-3c7f72d7-dec0-4982-947a-6637b81618ba' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>sample_id</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-b617fb45-ff0a-4dec-8aef-5f6afed8734f' class='xr-index-data-in' type='checkbox'/><label for='index-b617fb45-ff0a-4dec-8aef-5f6afed8734f' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;sample_0&#x27;], dtype=&#x27;object&#x27;, name=&#x27;sample_id&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-3b5e0251-d13a-4332-a2cd-e4d3f7c2cf70' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-3b5e0251-d13a-4332-a2cd-e4d3f7c2cf70' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.Dataset>\n",
       "Dimensions:                   (ranges: 106562, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) <U2 '1' '1' '1' '1' ... 'Y' 'Y' 'Y' 'Y'\n",
       "    start                     (ranges) int64 4807587 4807691 ... 22859847\n",
       "    end                       (ranges) int64 4807787 4807891 ... 22861212\n",
       "    range_id                  (ranges) int64 0 1 2 3 ... 53412 53413 53414 53415\n",
       "    num_contexts              (ranges) int64 27 35 22 0 7 2 11 ... 0 3 0 1 1 0 0\n",
       "    range_length              (ranges) int64 200 200 200 200 ... 1361 1365 1365\n",
       "  * sample_id                 (sample_id) <U8 'sample_0'\n",
       "    strand                    (ranges) <U1 '+' '+' '+' '+' ... '-' '-' '-' '-'\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 1.0 4.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.03704 0.1143 ... nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 3.0 4.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.1111 0.1143 ... nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 4.0 8.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.1481 0.2286 ... nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 2.111e+03 ... 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 78.19 74.09 ... nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    Id                        (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Gene_id                   (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Gene_type                 (ranges) object 'protein_coding' ... 'protein_c...\n",
       "    Gene_name                 (ranges) object 'Lypla1' 'Gm37988' ... 'Gm21366'\n",
       "    Level                     (ranges) object '2' '2' '2' '2' ... '2' '2' '2'\n",
       "    Mgi_id                    (ranges) object 'MGI:1344588' ... 'MGI:5434721'\n",
       "    Havana_gene               (ranges) object 'OTTMUSG00000021562.4' ... 'OTT...\n",
       "    Tag                       (ranges) object 'overlapping_locus' ... ''\n",
       "    Transcript_id             (ranges) object 'ENSMUST00000027036.10' ... 'EN...\n",
       "    Region                    (ranges) object 'before_tss' ... 'first_introns'\n",
       "    Strand                    (ranges) object '.' '.' '.' '.' ... '.' '.' '.'\n",
       "    Parent                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_type           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_name           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_number               (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_id                   (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_support_level  (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Havana_transcript         (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Protein_id                (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ccdsid                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ont                       (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Level_1                   (ranges) float64 nan nan nan nan ... 8.0 8.0 8.0\n",
       "    Ranges_id                 (ranges) float64 nan nan ... 2.164e+04 2.164e+04\n",
       "    Intron_number             (ranges) object nan nan nan nan ... '1' '1' '1'\n",
       "    mean_mc                   (ranges, sample_id) float64 0.0004737 ... nan\n",
       "    mean_hmc                  (ranges, sample_id) float64 0.001421 ... nan\n",
       "    mean_modc                 (ranges, sample_id) float64 0.001895 ... nan"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "antisense_ds = antisense_ds.assign(\n",
    "    mean_mc = antisense_ds[\"num_mc_sum\"] / antisense_ds[\"num_total_c_sum\"],\n",
    "    mean_hmc = antisense_ds[\"num_hmc_sum\"] / antisense_ds[\"num_total_c_sum\"],\n",
    "    mean_modc = antisense_ds[\"num_modc_sum\"] / antisense_ds[\"num_total_c_sum\"],\n",
    ")\n",
    "antisense_ds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "drop ranges where there are not C coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Count of instances where num_total_c_sum == 0: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(8952)\n",
      "Total number of instances: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(106562)\n",
      "Proportion of instances where num_total_c_sum (sense_ds) == 0: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(0.08400743)\n",
      "Count of instances where num_total_c_sum == 0 after dropping: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(0)\n"
     ]
    }
   ],
   "source": [
    "# Count instances where num_total_c_sum equals 0\n",
    "count_zeros = (sense_ds['num_total_c_sum'] == 0).sum().compute()\n",
    "print(f\"Count of instances where num_total_c_sum == 0: {count_zeros}\")\n",
    "\n",
    "# Total number of instances\n",
    "total_count = sense_ds['num_total_c_sum'].count().compute()\n",
    "print(f\"Total number of instances: {total_count}\")\n",
    "\n",
    "# Calculate the proportion of instances where num_total_c_sum equals 0\n",
    "proportion_zeros = count_zeros / total_count\n",
    "print(f\"Proportion of instances where num_total_c_sum (sense_ds) == 0: {proportion_zeros}\")\n",
    "\n",
    "# Create a mask where num_total_c_sum is greater than 0\n",
    "mask = sense_ds['num_total_c_sum'] > 0\n",
    "\n",
    "# Apply the mask to filter out rows where num_total_c_sum is 0 and drop them\n",
    "filtered_sense_ds = sense_ds.where(mask, drop=True)\n",
    "\n",
    "count_zeros = (filtered_sense_ds['num_total_c_sum'] == 0).sum().compute()\n",
    "print(f\"Count of instances where num_total_c_sum == 0 after dropping: {count_zeros}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Count of instances where num_total_c_sum == 0: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(8978)\n",
      "Total number of instances: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(106562)\n",
      "Proportion of instances where num_total_c_sum (antisense_ds) == 0: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(0.08425142)\n",
      "Count of instances where num_total_c_sum == 0 after dropping: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(0)\n"
     ]
    }
   ],
   "source": [
    "# Count instances where num_total_c_sum equals 0\n",
    "count_zeros = (antisense_ds['num_total_c_sum'] == 0).sum().compute()\n",
    "print(f\"Count of instances where num_total_c_sum == 0: {count_zeros}\")\n",
    "\n",
    "# Total number of instances\n",
    "total_count = antisense_ds['num_total_c_sum'].count().compute()\n",
    "print(f\"Total number of instances: {total_count}\")\n",
    "\n",
    "# Calculate the proportion of instances where num_total_c_sum equals 0\n",
    "proportion_zeros = count_zeros / total_count\n",
    "print(f\"Proportion of instances where num_total_c_sum (antisense_ds) == 0: {proportion_zeros}\")\n",
    "\n",
    "# Create a mask where num_total_c_sum is greater than 0\n",
    "mask = antisense_ds['num_total_c_sum'] > 0\n",
    "\n",
    "# Apply the mask to filter out rows where num_total_c_sum is 0 and drop them\n",
    "filtered_antisense_ds = antisense_ds.where(mask, drop=True)\n",
    "\n",
    "count_zeros = (filtered_antisense_ds['num_total_c_sum'] == 0).sum().compute()\n",
    "print(f\"Count of instances where num_total_c_sum == 0 after dropping: {count_zeros}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "no missing transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_transcript_genes = filtered_sense_ds[\"Transcript_id\"].where(filtered_sense_ds[\"Transcript_id\"].isnull(), drop=True)\n",
    "if len(missing_transcript_genes) > 0:\n",
    "    print(\"SENSE Gene IDs with missing Transcript IDs:\")\n",
    "    print(missing_transcript_genes)\n",
    "\n",
    "missing_transcript_genes = filtered_antisense_ds[\"Transcript_id\"].where(filtered_antisense_ds[\"Transcript_id\"].isnull(), drop=True)\n",
    "if len(missing_transcript_genes) > 0:\n",
    "    print(\"ANTISENSE Gene IDs with missing Transcript IDs:\")\n",
    "    print(missing_transcript_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "exon introns 3\\`UTRs 5\\`UTRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "stitched_sense_regions['Sense'] = \"sense\"\n",
    "stitched_antisense_regions['Sense'] = \"antisense\"\n",
    "stitched_antisense_regions = stitched_antisense_regions.drop([\"range_length\", \"cpg_count\"], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>total_c</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "      <th>Sense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.810206</td>\n",
       "      <td>0.026502</td>\n",
       "      <td>0.848889</td>\n",
       "      <td>6075.0</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003.15</td>\n",
       "      <td>0.747642</td>\n",
       "      <td>0.023585</td>\n",
       "      <td>0.780660</td>\n",
       "      <td>424.0</td>\n",
       "      <td>Pbsn</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000003.13</td>\n",
       "      <td>exons</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028.15</td>\n",
       "      <td>0.557015</td>\n",
       "      <td>0.031306</td>\n",
       "      <td>0.596918</td>\n",
       "      <td>6165.0</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000028.13</td>\n",
       "      <td>exons</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000037.17</td>\n",
       "      <td>0.767530</td>\n",
       "      <td>0.013898</td>\n",
       "      <td>0.791535</td>\n",
       "      <td>1583.0</td>\n",
       "      <td>Scml2</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000101113.8</td>\n",
       "      <td>exons</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000049.11</td>\n",
       "      <td>0.793698</td>\n",
       "      <td>0.026632</td>\n",
       "      <td>0.833458</td>\n",
       "      <td>2666.0</td>\n",
       "      <td>Apoh</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000049.5</td>\n",
       "      <td>exons</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15460</th>\n",
       "      <td>ENSMUSG00000117975.1</td>\n",
       "      <td>0.778394</td>\n",
       "      <td>0.042044</td>\n",
       "      <td>0.833577</td>\n",
       "      <td>3425.0</td>\n",
       "      <td>Itprip</td>\n",
       "      <td>19</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000095998.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15461</th>\n",
       "      <td>ENSMUSG00000118219.1</td>\n",
       "      <td>0.424810</td>\n",
       "      <td>0.083937</td>\n",
       "      <td>0.515992</td>\n",
       "      <td>5659.0</td>\n",
       "      <td>Gm29695</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000188175.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15462</th>\n",
       "      <td>ENSMUSG00000118332.1</td>\n",
       "      <td>0.813403</td>\n",
       "      <td>0.045638</td>\n",
       "      <td>0.870884</td>\n",
       "      <td>3462.0</td>\n",
       "      <td>Fam220a</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000119488.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15464</th>\n",
       "      <td>ENSMUSG00000118537.1</td>\n",
       "      <td>0.766355</td>\n",
       "      <td>0.009346</td>\n",
       "      <td>0.813084</td>\n",
       "      <td>107.0</td>\n",
       "      <td>Shld3</td>\n",
       "      <td>13</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000133280.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15465</th>\n",
       "      <td>ENSMUSG00000118560.1</td>\n",
       "      <td>0.603896</td>\n",
       "      <td>0.077922</td>\n",
       "      <td>0.694805</td>\n",
       "      <td>154.0</td>\n",
       "      <td>Lim2</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000004732.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65190 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_id   mean_mc  mean_hmc  mean_modc  total_c  \\\n",
       "0       ENSMUSG00000000001.4  0.810206  0.026502   0.848889   6075.0   \n",
       "1      ENSMUSG00000000003.15  0.747642  0.023585   0.780660    424.0   \n",
       "2      ENSMUSG00000000028.15  0.557015  0.031306   0.596918   6165.0   \n",
       "3      ENSMUSG00000000037.17  0.767530  0.013898   0.791535   1583.0   \n",
       "4      ENSMUSG00000000049.11  0.793698  0.026632   0.833458   2666.0   \n",
       "...                      ...       ...       ...        ...      ...   \n",
       "15460   ENSMUSG00000117975.1  0.778394  0.042044   0.833577   3425.0   \n",
       "15461   ENSMUSG00000118219.1  0.424810  0.083937   0.515992   5659.0   \n",
       "15462   ENSMUSG00000118332.1  0.813403  0.045638   0.870884   3462.0   \n",
       "15464   ENSMUSG00000118537.1  0.766355  0.009346   0.813084    107.0   \n",
       "15465   ENSMUSG00000118560.1  0.603896  0.077922   0.694805    154.0   \n",
       "\n",
       "      Gene_name contig strand          Transcript_id            Region  \\\n",
       "0         Gnai3      3      -   ENSMUST00000000001.4             exons   \n",
       "1          Pbsn      X      -  ENSMUST00000000003.13             exons   \n",
       "2         Cdc45     16      -  ENSMUST00000000028.13             exons   \n",
       "3         Scml2      X      +   ENSMUST00000101113.8             exons   \n",
       "4          Apoh     11      +   ENSMUST00000000049.5             exons   \n",
       "...         ...    ...    ...                    ...               ...   \n",
       "15460    Itprip     19      -   ENSMUST00000095998.6  three_prime_utrs   \n",
       "15461   Gm29695      1      +   ENSMUST00000188175.1  three_prime_utrs   \n",
       "15462   Fam220a      5      +   ENSMUST00000119488.1  three_prime_utrs   \n",
       "15464     Shld3     13      -   ENSMUST00000133280.1  three_prime_utrs   \n",
       "15465      Lim2      7      +   ENSMUST00000004732.6  three_prime_utrs   \n",
       "\n",
       "           Sense  \n",
       "0      antisense  \n",
       "1      antisense  \n",
       "2      antisense  \n",
       "3      antisense  \n",
       "4      antisense  \n",
       "...          ...  \n",
       "15460  antisense  \n",
       "15461  antisense  \n",
       "15462  antisense  \n",
       "15464  antisense  \n",
       "15465  antisense  \n",
       "\n",
       "[65190 rows x 11 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stitched_antisense_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>total_c</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "      <th>Sense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.807980</td>\n",
       "      <td>0.025436</td>\n",
       "      <td>0.847215</td>\n",
       "      <td>51.0</td>\n",
       "      <td>6015.0</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003.15</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.014888</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>7.0</td>\n",
       "      <td>403.0</td>\n",
       "      <td>681.0</td>\n",
       "      <td>Pbsn</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000003.13</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028.15</td>\n",
       "      <td>0.581295</td>\n",
       "      <td>0.031589</td>\n",
       "      <td>0.619697</td>\n",
       "      <td>60.0</td>\n",
       "      <td>6458.0</td>\n",
       "      <td>1955.0</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000028.13</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000037.17</td>\n",
       "      <td>0.758328</td>\n",
       "      <td>0.017636</td>\n",
       "      <td>0.786414</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1531.0</td>\n",
       "      <td>2777.0</td>\n",
       "      <td>Scml2</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000101113.8</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000049.11</td>\n",
       "      <td>0.783618</td>\n",
       "      <td>0.036519</td>\n",
       "      <td>0.835495</td>\n",
       "      <td>28.0</td>\n",
       "      <td>2930.0</td>\n",
       "      <td>1068.0</td>\n",
       "      <td>Apoh</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000049.5</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15460</th>\n",
       "      <td>ENSMUSG00000117975.1</td>\n",
       "      <td>0.778394</td>\n",
       "      <td>0.042044</td>\n",
       "      <td>0.833577</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3425.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Itprip</td>\n",
       "      <td>19</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000095998.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15461</th>\n",
       "      <td>ENSMUSG00000118219.1</td>\n",
       "      <td>0.424810</td>\n",
       "      <td>0.083937</td>\n",
       "      <td>0.515992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5659.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gm29695</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000188175.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15462</th>\n",
       "      <td>ENSMUSG00000118332.1</td>\n",
       "      <td>0.813403</td>\n",
       "      <td>0.045638</td>\n",
       "      <td>0.870884</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3462.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fam220a</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000119488.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15464</th>\n",
       "      <td>ENSMUSG00000118537.1</td>\n",
       "      <td>0.766355</td>\n",
       "      <td>0.009346</td>\n",
       "      <td>0.813084</td>\n",
       "      <td>NaN</td>\n",
       "      <td>107.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shld3</td>\n",
       "      <td>13</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000133280.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15465</th>\n",
       "      <td>ENSMUSG00000118560.1</td>\n",
       "      <td>0.603896</td>\n",
       "      <td>0.077922</td>\n",
       "      <td>0.694805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>154.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lim2</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000004732.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>130372 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  \\\n",
       "0       ENSMUSG00000000001.4  0.807980  0.025436   0.847215       51.0   \n",
       "1      ENSMUSG00000000003.15  0.774194  0.014888   0.806452        7.0   \n",
       "2      ENSMUSG00000000028.15  0.581295  0.031589   0.619697       60.0   \n",
       "3      ENSMUSG00000000037.17  0.758328  0.017636   0.786414       30.0   \n",
       "4      ENSMUSG00000000049.11  0.783618  0.036519   0.835495       28.0   \n",
       "...                      ...       ...       ...        ...        ...   \n",
       "15460   ENSMUSG00000117975.1  0.778394  0.042044   0.833577        NaN   \n",
       "15461   ENSMUSG00000118219.1  0.424810  0.083937   0.515992        NaN   \n",
       "15462   ENSMUSG00000118332.1  0.813403  0.045638   0.870884        NaN   \n",
       "15464   ENSMUSG00000118537.1  0.766355  0.009346   0.813084        NaN   \n",
       "15465   ENSMUSG00000118560.1  0.603896  0.077922   0.694805        NaN   \n",
       "\n",
       "       total_c  range_length Gene_name contig strand          Transcript_id  \\\n",
       "0       6015.0        2995.0     Gnai3      3      -   ENSMUST00000000001.4   \n",
       "1        403.0         681.0      Pbsn      X      -  ENSMUST00000000003.13   \n",
       "2       6458.0        1955.0     Cdc45     16      -  ENSMUST00000000028.13   \n",
       "3       1531.0        2777.0     Scml2      X      +   ENSMUST00000101113.8   \n",
       "4       2930.0        1068.0      Apoh     11      +   ENSMUST00000000049.5   \n",
       "...        ...           ...       ...    ...    ...                    ...   \n",
       "15460   3425.0           NaN    Itprip     19      -   ENSMUST00000095998.6   \n",
       "15461   5659.0           NaN   Gm29695      1      +   ENSMUST00000188175.1   \n",
       "15462   3462.0           NaN   Fam220a      5      +   ENSMUST00000119488.1   \n",
       "15464    107.0           NaN     Shld3     13      -   ENSMUST00000133280.1   \n",
       "15465    154.0           NaN      Lim2      7      +   ENSMUST00000004732.6   \n",
       "\n",
       "                 Region      Sense  \n",
       "0                 exons      sense  \n",
       "1                 exons      sense  \n",
       "2                 exons      sense  \n",
       "3                 exons      sense  \n",
       "4                 exons      sense  \n",
       "...                 ...        ...  \n",
       "15460  three_prime_utrs  antisense  \n",
       "15461  three_prime_utrs  antisense  \n",
       "15462  three_prime_utrs  antisense  \n",
       "15464  three_prime_utrs  antisense  \n",
       "15465  three_prime_utrs  antisense  \n",
       "\n",
       "[130372 rows x 13 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stitched_regions = pd.concat([stitched_sense_regions, stitched_antisense_regions])\n",
    "stitched_regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "combining TSS, TES, first exon, first intron, genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sense = filtered_sense_ds.to_dataframe().reset_index(level=\"sample_id\", drop=True)\n",
    "df_antisense = filtered_antisense_ds.to_dataframe().reset_index(level=\"sample_id\", drop=True)\n",
    "\n",
    "df_sense[\"Sense\"] = \"sense\"\n",
    "df_antisense[\"Sense\"] = \"antisense\"\n",
    "df_antisense = df_antisense.drop(['range_length','num_total_c_cpg_count'],axis=1)\n",
    "df = pd.concat([df_sense, df_antisense])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_mc_sum</th>\n",
       "      <th>num_mc_mean</th>\n",
       "      <th>num_mc_cpg_count</th>\n",
       "      <th>num_hmc_sum</th>\n",
       "      <th>num_hmc_mean</th>\n",
       "      <th>num_hmc_cpg_count</th>\n",
       "      <th>num_modc_sum</th>\n",
       "      <th>num_modc_mean</th>\n",
       "      <th>num_modc_cpg_count</th>\n",
       "      <th>num_total_c_sum</th>\n",
       "      <th>...</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>range_id</th>\n",
       "      <th>num_contexts</th>\n",
       "      <th>range_length</th>\n",
       "      <th>strand</th>\n",
       "      <th>Sense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ranges</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.037037</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.037037</td>\n",
       "      <td>27.0</td>\n",
       "      <td>2122.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000471</td>\n",
       "      <td>0.000471</td>\n",
       "      <td>1</td>\n",
       "      <td>4807587</td>\n",
       "      <td>4807787</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>200.0</td>\n",
       "      <td>+</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>35.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.114286</td>\n",
       "      <td>35.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.171429</td>\n",
       "      <td>35.0</td>\n",
       "      <td>2904.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001377</td>\n",
       "      <td>0.002066</td>\n",
       "      <td>1</td>\n",
       "      <td>4807691</td>\n",
       "      <td>4807891</td>\n",
       "      <td>1</td>\n",
       "      <td>35</td>\n",
       "      <td>200.0</td>\n",
       "      <td>+</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4.0</td>\n",
       "      <td>0.190476</td>\n",
       "      <td>21.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>21.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>21.0</td>\n",
       "      <td>2032.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000984</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>1</td>\n",
       "      <td>4857613</td>\n",
       "      <td>4857813</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>200.0</td>\n",
       "      <td>+</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>136.0</td>\n",
       "      <td>19.428571</td>\n",
       "      <td>7.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>3.285714</td>\n",
       "      <td>7.0</td>\n",
       "      <td>159.0</td>\n",
       "      <td>22.714286</td>\n",
       "      <td>7.0</td>\n",
       "      <td>501.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.045908</td>\n",
       "      <td>0.317365</td>\n",
       "      <td>1</td>\n",
       "      <td>5588265</td>\n",
       "      <td>5588465</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>200.0</td>\n",
       "      <td>+</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>145.0</td>\n",
       "      <td>72.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>153.0</td>\n",
       "      <td>76.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>202.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034653</td>\n",
       "      <td>0.757426</td>\n",
       "      <td>1</td>\n",
       "      <td>6205996</td>\n",
       "      <td>6206196</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>200.0</td>\n",
       "      <td>+</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97579</th>\n",
       "      <td>43.0</td>\n",
       "      <td>21.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>23.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>0.734375</td>\n",
       "      <td>Y</td>\n",
       "      <td>6917397</td>\n",
       "      <td>6918160</td>\n",
       "      <td>53402</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97580</th>\n",
       "      <td>63.0</td>\n",
       "      <td>10.500000</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>11.166667</td>\n",
       "      <td>6.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.837500</td>\n",
       "      <td>Y</td>\n",
       "      <td>42580848</td>\n",
       "      <td>42582206</td>\n",
       "      <td>53403</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97581</th>\n",
       "      <td>21.0</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.666667</td>\n",
       "      <td>3.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>0.675000</td>\n",
       "      <td>Y</td>\n",
       "      <td>68831319</td>\n",
       "      <td>68832684</td>\n",
       "      <td>53410</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97582</th>\n",
       "      <td>31.0</td>\n",
       "      <td>31.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.939394</td>\n",
       "      <td>Y</td>\n",
       "      <td>40315473</td>\n",
       "      <td>40316859</td>\n",
       "      <td>53412</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97583</th>\n",
       "      <td>19.0</td>\n",
       "      <td>19.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>21.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.724138</td>\n",
       "      <td>Y</td>\n",
       "      <td>20992252</td>\n",
       "      <td>20993613</td>\n",
       "      <td>53413</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>195194 rows × 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        num_mc_sum  num_mc_mean  num_mc_cpg_count  num_hmc_sum  num_hmc_mean  \\\n",
       "ranges                                                                         \n",
       "0              0.0     0.000000              27.0          1.0      0.037037   \n",
       "1              2.0     0.057143              35.0          4.0      0.114286   \n",
       "2              4.0     0.190476              21.0          2.0      0.095238   \n",
       "3            136.0    19.428571               7.0         23.0      3.285714   \n",
       "4            145.0    72.500000               2.0          7.0      3.500000   \n",
       "...            ...          ...               ...          ...           ...   \n",
       "97579         43.0    21.500000               2.0          3.0      1.500000   \n",
       "97580         63.0    10.500000               6.0          0.0      0.000000   \n",
       "97581         21.0     7.000000               3.0          5.0      1.666667   \n",
       "97582         31.0    31.000000               1.0          0.0      0.000000   \n",
       "97583         19.0    19.000000               1.0          1.0      1.000000   \n",
       "\n",
       "        num_hmc_cpg_count  num_modc_sum  num_modc_mean  num_modc_cpg_count  \\\n",
       "ranges                                                                       \n",
       "0                    27.0           1.0       0.037037                27.0   \n",
       "1                    35.0           6.0       0.171429                35.0   \n",
       "2                    21.0           6.0       0.285714                21.0   \n",
       "3                     7.0         159.0      22.714286                 7.0   \n",
       "4                     2.0         153.0      76.500000                 2.0   \n",
       "...                   ...           ...            ...                 ...   \n",
       "97579                 2.0          47.0      23.500000                 2.0   \n",
       "97580                 6.0          67.0      11.166667                 6.0   \n",
       "97581                 3.0          27.0       9.000000                 3.0   \n",
       "97582                 1.0          31.0      31.000000                 1.0   \n",
       "97583                 1.0          21.0      21.000000                 1.0   \n",
       "\n",
       "        num_total_c_sum  ...  mean_hmc  mean_modc contig     start       end  \\\n",
       "ranges                   ...                                                   \n",
       "0                2122.0  ...  0.000471   0.000471      1   4807587   4807787   \n",
       "1                2904.0  ...  0.001377   0.002066      1   4807691   4807891   \n",
       "2                2032.0  ...  0.000984   0.002953      1   4857613   4857813   \n",
       "3                 501.0  ...  0.045908   0.317365      1   5588265   5588465   \n",
       "4                 202.0  ...  0.034653   0.757426      1   6205996   6206196   \n",
       "...                 ...  ...       ...        ...    ...       ...       ...   \n",
       "97579              64.0  ...  0.046875   0.734375      Y   6917397   6918160   \n",
       "97580              80.0  ...  0.000000   0.837500      Y  42580848  42582206   \n",
       "97581              40.0  ...  0.125000   0.675000      Y  68831319  68832684   \n",
       "97582              33.0  ...  0.000000   0.939394      Y  40315473  40316859   \n",
       "97583              29.0  ...  0.034483   0.724138      Y  20992252  20993613   \n",
       "\n",
       "       range_id num_contexts range_length strand      Sense  \n",
       "ranges                                                       \n",
       "0             0           27        200.0      +      sense  \n",
       "1             1           35        200.0      +      sense  \n",
       "2             2           21        200.0      +      sense  \n",
       "3             4            7        200.0      +      sense  \n",
       "4             5            2        200.0      +      sense  \n",
       "...         ...          ...          ...    ...        ...  \n",
       "97579     53402            2          NaN      -  antisense  \n",
       "97580     53403            6          NaN      -  antisense  \n",
       "97581     53410            3          NaN      -  antisense  \n",
       "97582     53412            1          NaN      -  antisense  \n",
       "97583     53413            1          NaN      -  antisense  \n",
       "\n",
       "[195194 rows x 47 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "      <th>Sense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000471</td>\n",
       "      <td>0.000471</td>\n",
       "      <td>27.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>0.000689</td>\n",
       "      <td>0.001377</td>\n",
       "      <td>0.002066</td>\n",
       "      <td>35.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000155020.1</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.000984</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>21.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.271457</td>\n",
       "      <td>0.045908</td>\n",
       "      <td>0.317365</td>\n",
       "      <td>7.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000025907.14</td>\n",
       "      <td>0.717822</td>\n",
       "      <td>0.034653</td>\n",
       "      <td>0.757426</td>\n",
       "      <td>2.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Rb1cc1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000027040.12</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325561</th>\n",
       "      <td>ENSMUSG00000117975.1</td>\n",
       "      <td>0.778394</td>\n",
       "      <td>0.042044</td>\n",
       "      <td>0.833577</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Itprip</td>\n",
       "      <td>19</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000095998.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325562</th>\n",
       "      <td>ENSMUSG00000118219.1</td>\n",
       "      <td>0.424810</td>\n",
       "      <td>0.083937</td>\n",
       "      <td>0.515992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gm29695</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000188175.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325563</th>\n",
       "      <td>ENSMUSG00000118332.1</td>\n",
       "      <td>0.813403</td>\n",
       "      <td>0.045638</td>\n",
       "      <td>0.870884</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fam220a</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000119488.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325564</th>\n",
       "      <td>ENSMUSG00000118537.1</td>\n",
       "      <td>0.766355</td>\n",
       "      <td>0.009346</td>\n",
       "      <td>0.813084</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shld3</td>\n",
       "      <td>13</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000133280.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325565</th>\n",
       "      <td>ENSMUSG00000118560.1</td>\n",
       "      <td>0.603896</td>\n",
       "      <td>0.077922</td>\n",
       "      <td>0.694805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lim2</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000004732.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>325566 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  \\\n",
       "0       ENSMUSG00000025903.14  0.000000  0.000471   0.000471       27.0   \n",
       "1        ENSMUSG00000104217.1  0.000689  0.001377   0.002066       35.0   \n",
       "2       ENSMUSG00000033813.15  0.001969  0.000984   0.002953       21.0   \n",
       "3       ENSMUSG00000025905.14  0.271457  0.045908   0.317365        7.0   \n",
       "4       ENSMUSG00000025907.14  0.717822  0.034653   0.757426        2.0   \n",
       "...                       ...       ...       ...        ...        ...   \n",
       "325561   ENSMUSG00000117975.1  0.778394  0.042044   0.833577        NaN   \n",
       "325562   ENSMUSG00000118219.1  0.424810  0.083937   0.515992        NaN   \n",
       "325563   ENSMUSG00000118332.1  0.813403  0.045638   0.870884        NaN   \n",
       "325564   ENSMUSG00000118537.1  0.766355  0.009346   0.813084        NaN   \n",
       "325565   ENSMUSG00000118560.1  0.603896  0.077922   0.694805        NaN   \n",
       "\n",
       "        range_length Gene_name contig strand          Transcript_id  \\\n",
       "0              200.0    Lypla1      1      +  ENSMUST00000027036.10   \n",
       "1              200.0   Gm37988      1      +   ENSMUST00000155020.1   \n",
       "2              200.0     Tcea1      1      +  ENSMUST00000081551.13   \n",
       "3              200.0     Oprk1      1      +   ENSMUST00000160777.7   \n",
       "4              200.0    Rb1cc1      1      +  ENSMUST00000027040.12   \n",
       "...              ...       ...    ...    ...                    ...   \n",
       "325561           NaN    Itprip     19      -   ENSMUST00000095998.6   \n",
       "325562           NaN   Gm29695      1      +   ENSMUST00000188175.1   \n",
       "325563           NaN   Fam220a      5      +   ENSMUST00000119488.1   \n",
       "325564           NaN     Shld3     13      -   ENSMUST00000133280.1   \n",
       "325565           NaN      Lim2      7      +   ENSMUST00000004732.6   \n",
       "\n",
       "                  Region      Sense  \n",
       "0             before_tss      sense  \n",
       "1             before_tss      sense  \n",
       "2             before_tss      sense  \n",
       "3             before_tss      sense  \n",
       "4             before_tss      sense  \n",
       "...                  ...        ...  \n",
       "325561  three_prime_utrs  antisense  \n",
       "325562  three_prime_utrs  antisense  \n",
       "325563  three_prime_utrs  antisense  \n",
       "325564  three_prime_utrs  antisense  \n",
       "325565  three_prime_utrs  antisense  \n",
       "\n",
       "[325566 rows x 12 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"cpg_count\"] = df[\"num_total_c_cpg_count\"]\n",
    "use_cols = [c for c in stitched_regions.columns if c != \"total_c\"]\n",
    "df=df[use_cols].reset_index(drop=True)\n",
    "df = pd.concat([df, stitched_regions]).reset_index(drop=True)\n",
    "df.drop('total_c', axis=1, inplace=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "old"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "      <th>Sense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000471</td>\n",
       "      <td>0.000471</td>\n",
       "      <td>27.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>0.000689</td>\n",
       "      <td>0.001377</td>\n",
       "      <td>0.002066</td>\n",
       "      <td>35.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000155020.1</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.000984</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>21.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.271457</td>\n",
       "      <td>0.045908</td>\n",
       "      <td>0.317365</td>\n",
       "      <td>7.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000025907.14</td>\n",
       "      <td>0.717822</td>\n",
       "      <td>0.034653</td>\n",
       "      <td>0.757426</td>\n",
       "      <td>2.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Rb1cc1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000027040.12</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367485</th>\n",
       "      <td>ENSMUSG00000117975.1</td>\n",
       "      <td>0.778394</td>\n",
       "      <td>0.042044</td>\n",
       "      <td>0.833577</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Itprip</td>\n",
       "      <td>19</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000095998.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367486</th>\n",
       "      <td>ENSMUSG00000118219.1</td>\n",
       "      <td>0.424810</td>\n",
       "      <td>0.083937</td>\n",
       "      <td>0.515992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gm29695</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000188175.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367487</th>\n",
       "      <td>ENSMUSG00000118332.1</td>\n",
       "      <td>0.813403</td>\n",
       "      <td>0.045638</td>\n",
       "      <td>0.870884</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fam220a</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000119488.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367488</th>\n",
       "      <td>ENSMUSG00000118537.1</td>\n",
       "      <td>0.766355</td>\n",
       "      <td>0.009346</td>\n",
       "      <td>0.813084</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shld3</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000133280.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367489</th>\n",
       "      <td>ENSMUSG00000118560.1</td>\n",
       "      <td>0.603896</td>\n",
       "      <td>0.077922</td>\n",
       "      <td>0.694805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lim2</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000004732.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>367490 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  \\\n",
       "0       ENSMUSG00000025903.14  0.000000  0.000471   0.000471       27.0   \n",
       "1        ENSMUSG00000104217.1  0.000689  0.001377   0.002066       35.0   \n",
       "2       ENSMUSG00000033813.15  0.001969  0.000984   0.002953       21.0   \n",
       "3       ENSMUSG00000025905.14  0.271457  0.045908   0.317365        7.0   \n",
       "4       ENSMUSG00000025907.14  0.717822  0.034653   0.757426        2.0   \n",
       "...                       ...       ...       ...        ...        ...   \n",
       "367485   ENSMUSG00000117975.1  0.778394  0.042044   0.833577        NaN   \n",
       "367486   ENSMUSG00000118219.1  0.424810  0.083937   0.515992        NaN   \n",
       "367487   ENSMUSG00000118332.1  0.813403  0.045638   0.870884        NaN   \n",
       "367488   ENSMUSG00000118537.1  0.766355  0.009346   0.813084        NaN   \n",
       "367489   ENSMUSG00000118560.1  0.603896  0.077922   0.694805        NaN   \n",
       "\n",
       "        range_length Gene_name contig strand          Transcript_id  \\\n",
       "0              200.0    Lypla1      1      +  ENSMUST00000027036.10   \n",
       "1              200.0   Gm37988      1      +   ENSMUST00000155020.1   \n",
       "2              200.0     Tcea1      1      +  ENSMUST00000081551.13   \n",
       "3              200.0     Oprk1      1      +   ENSMUST00000160777.7   \n",
       "4              200.0    Rb1cc1      1      +  ENSMUST00000027040.12   \n",
       "...              ...       ...    ...    ...                    ...   \n",
       "367485           NaN    Itprip     19      +   ENSMUST00000095998.6   \n",
       "367486           NaN   Gm29695      1      -   ENSMUST00000188175.1   \n",
       "367487           NaN   Fam220a      5      -   ENSMUST00000119488.1   \n",
       "367488           NaN     Shld3     13      +   ENSMUST00000133280.1   \n",
       "367489           NaN      Lim2      7      -   ENSMUST00000004732.6   \n",
       "\n",
       "                  Region      Sense  \n",
       "0             before_tss      sense  \n",
       "1             before_tss      sense  \n",
       "2             before_tss      sense  \n",
       "3             before_tss      sense  \n",
       "4             before_tss      sense  \n",
       "...                  ...        ...  \n",
       "367485  three_prime_utrs  antisense  \n",
       "367486  three_prime_utrs  antisense  \n",
       "367487  three_prime_utrs  antisense  \n",
       "367488  three_prime_utrs  antisense  \n",
       "367489  three_prime_utrs  antisense  \n",
       "\n",
       "[367490 rows x 12 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df[\"cpg_count\"] = df[\"num_total_c_cpg_count\"]\n",
    "# use_cols = [c for c in stitched_regions.columns if c != \"total_c\"]\n",
    "# df=df[use_cols].reset_index(drop=True)\n",
    "# df = pd.concat([df, stitched_regions]).reset_index(drop=True)\n",
    "# df.drop('total_c', axis=1, inplace=True)\n",
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "      <th>Sense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.271457</td>\n",
       "      <td>0.045908</td>\n",
       "      <td>0.317365</td>\n",
       "      <td>7.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.759259</td>\n",
       "      <td>0.018519</td>\n",
       "      <td>0.791667</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>after_tes</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.680438</td>\n",
       "      <td>0.043483</td>\n",
       "      <td>0.736818</td>\n",
       "      <td>98.0</td>\n",
       "      <td>17665.0</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>genes</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1765</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.176190</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.261905</td>\n",
       "      <td>5.0</td>\n",
       "      <td>170.0</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>first_exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2749</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.164557</td>\n",
       "      <td>0.215190</td>\n",
       "      <td>0.379747</td>\n",
       "      <td>1.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>first_introns</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97613</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.320334</td>\n",
       "      <td>0.100279</td>\n",
       "      <td>0.428969</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98171</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.786378</td>\n",
       "      <td>0.024768</td>\n",
       "      <td>0.828173</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>after_tes</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98768</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.686781</td>\n",
       "      <td>0.033950</td>\n",
       "      <td>0.732589</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>genes</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99372</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.184615</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>0.241758</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>first_exons</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100357</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>0.275000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>first_introns</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200280</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.711306</td>\n",
       "      <td>0.051391</td>\n",
       "      <td>0.773881</td>\n",
       "      <td>35.0</td>\n",
       "      <td>4501.0</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219739</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.710797</td>\n",
       "      <td>0.032089</td>\n",
       "      <td>0.758022</td>\n",
       "      <td>57.0</td>\n",
       "      <td>12967.0</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>introns</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237048</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.238000</td>\n",
       "      <td>0.086000</td>\n",
       "      <td>0.330000</td>\n",
       "      <td>6.0</td>\n",
       "      <td>183.0</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250738</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.706485</td>\n",
       "      <td>0.029863</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>12.0</td>\n",
       "      <td>3346.0</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265462</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.728581</td>\n",
       "      <td>0.039003</td>\n",
       "      <td>0.780691</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>exons</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284919</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.715674</td>\n",
       "      <td>0.026731</td>\n",
       "      <td>0.754759</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>introns</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302241</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.242308</td>\n",
       "      <td>0.053846</td>\n",
       "      <td>0.296154</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315929</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.734868</td>\n",
       "      <td>0.027280</td>\n",
       "      <td>0.780051</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  \\\n",
       "3       ENSMUSG00000025905.14  0.271457  0.045908   0.317365        7.0   \n",
       "564     ENSMUSG00000025905.14  0.759259  0.018519   0.791667        6.0   \n",
       "1161    ENSMUSG00000025905.14  0.680438  0.043483   0.736818       98.0   \n",
       "1765    ENSMUSG00000025905.14  0.176190  0.083333   0.261905        5.0   \n",
       "2749    ENSMUSG00000025905.14  0.164557  0.215190   0.379747        1.0   \n",
       "97613   ENSMUSG00000025905.14  0.320334  0.100279   0.428969        NaN   \n",
       "98171   ENSMUSG00000025905.14  0.786378  0.024768   0.828173        NaN   \n",
       "98768   ENSMUSG00000025905.14  0.686781  0.033950   0.732589        NaN   \n",
       "99372   ENSMUSG00000025905.14  0.184615  0.057143   0.241758        NaN   \n",
       "100357  ENSMUSG00000025905.14  0.125000  0.150000   0.275000        NaN   \n",
       "200280  ENSMUSG00000025905.14  0.711306  0.051391   0.773881       35.0   \n",
       "219739  ENSMUSG00000025905.14  0.710797  0.032089   0.758022       57.0   \n",
       "237048  ENSMUSG00000025905.14  0.238000  0.086000   0.330000        6.0   \n",
       "250738  ENSMUSG00000025905.14  0.706485  0.029863   0.750000       12.0   \n",
       "265462  ENSMUSG00000025905.14  0.728581  0.039003   0.780691        NaN   \n",
       "284919  ENSMUSG00000025905.14  0.715674  0.026731   0.754759        NaN   \n",
       "302241  ENSMUSG00000025905.14  0.242308  0.053846   0.296154        NaN   \n",
       "315929  ENSMUSG00000025905.14  0.734868  0.027280   0.780051        NaN   \n",
       "\n",
       "        range_length Gene_name contig strand         Transcript_id  \\\n",
       "3              200.0     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "564           1000.0     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "1161         17665.0     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "1765           170.0     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "2749            27.0     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "97613            NaN     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "98171            NaN     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "98768            NaN     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "99372            NaN     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "100357           NaN     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "200280        4501.0     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "219739       12967.0     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "237048         183.0     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "250738        3346.0     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "265462           NaN     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "284919           NaN     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "302241           NaN     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "315929           NaN     Oprk1      1      +  ENSMUST00000160777.7   \n",
       "\n",
       "                  Region      Sense  \n",
       "3             before_tss      sense  \n",
       "564            after_tes      sense  \n",
       "1161               genes      sense  \n",
       "1765         first_exons      sense  \n",
       "2749       first_introns      sense  \n",
       "97613         before_tss  antisense  \n",
       "98171          after_tes  antisense  \n",
       "98768              genes  antisense  \n",
       "99372        first_exons  antisense  \n",
       "100357     first_introns  antisense  \n",
       "200280             exons      sense  \n",
       "219739           introns      sense  \n",
       "237048   five_prime_utrs      sense  \n",
       "250738  three_prime_utrs      sense  \n",
       "265462             exons  antisense  \n",
       "284919           introns  antisense  \n",
       "302241   five_prime_utrs  antisense  \n",
       "315929  three_prime_utrs  antisense  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[\"Gene_id\"]==\"ENSMUSG00000025905.14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>Region</th>\n",
       "      <th>Sense</th>\n",
       "      <th>range_length</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_modc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>311844</th>\n",
       "      <td>-</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.816244</td>\n",
       "      <td>0.856091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246653</th>\n",
       "      <td>-</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>sense</td>\n",
       "      <td>2054.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>0.813870</td>\n",
       "      <td>0.849446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214745</th>\n",
       "      <td>-</td>\n",
       "      <td>introns</td>\n",
       "      <td>sense</td>\n",
       "      <td>13562.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>0.780352</td>\n",
       "      <td>0.832180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279924</th>\n",
       "      <td>-</td>\n",
       "      <td>introns</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.772727</td>\n",
       "      <td>0.834170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56762</th>\n",
       "      <td>-</td>\n",
       "      <td>genes</td>\n",
       "      <td>sense</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>430.0</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>0.719680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154357</th>\n",
       "      <td>-</td>\n",
       "      <td>genes</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>0.725905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298101</th>\n",
       "      <td>-</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232914</th>\n",
       "      <td>-</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "      <td>sense</td>\n",
       "      <td>140.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.000784</td>\n",
       "      <td>0.000784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154908</th>\n",
       "      <td>-</td>\n",
       "      <td>first_introns</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.625551</td>\n",
       "      <td>0.682169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57314</th>\n",
       "      <td>-</td>\n",
       "      <td>first_introns</td>\n",
       "      <td>sense</td>\n",
       "      <td>22051.0</td>\n",
       "      <td>208.0</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>0.678294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154793</th>\n",
       "      <td>-</td>\n",
       "      <td>first_exons</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57199</th>\n",
       "      <td>-</td>\n",
       "      <td>first_exons</td>\n",
       "      <td>sense</td>\n",
       "      <td>258.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.000455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195194</th>\n",
       "      <td>-</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.807980</td>\n",
       "      <td>0.847215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260376</th>\n",
       "      <td>-</td>\n",
       "      <td>exons</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.810206</td>\n",
       "      <td>0.848889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153396</th>\n",
       "      <td>-</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.001415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55802</th>\n",
       "      <td>-</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "      <td>200.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56276</th>\n",
       "      <td>-</td>\n",
       "      <td>after_tes</td>\n",
       "      <td>sense</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.744475</td>\n",
       "      <td>0.794199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153871</th>\n",
       "      <td>-</td>\n",
       "      <td>after_tes</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.751773</td>\n",
       "      <td>0.804965</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       strand            Region      Sense  range_length  cpg_count   mean_mc  \\\n",
       "311844      -  three_prime_utrs  antisense           NaN        NaN  0.816244   \n",
       "246653      -  three_prime_utrs      sense        2054.0       32.0  0.813870   \n",
       "214745      -           introns      sense       13562.0      144.0  0.780352   \n",
       "279924      -           introns  antisense           NaN        NaN  0.772727   \n",
       "56762       -             genes      sense       38866.0      430.0  0.670919   \n",
       "154357      -             genes  antisense           NaN        NaN  0.672555   \n",
       "298101      -   five_prime_utrs  antisense           NaN        NaN  0.000000   \n",
       "232914      -   five_prime_utrs      sense         140.0       15.0  0.000784   \n",
       "154908      -     first_introns  antisense           NaN        NaN  0.625551   \n",
       "57314       -     first_introns      sense       22051.0      208.0  0.624199   \n",
       "154793      -       first_exons  antisense           NaN        NaN  0.000000   \n",
       "57199       -       first_exons      sense         258.0       27.0  0.000455   \n",
       "195194      -             exons      sense        2995.0       51.0  0.807980   \n",
       "260376      -             exons  antisense           NaN        NaN  0.810206   \n",
       "153396      -        before_tss  antisense           NaN        NaN  0.001415   \n",
       "55802       -        before_tss      sense         200.0       15.0  0.000000   \n",
       "56276       -         after_tes      sense        1000.0       13.0  0.744475   \n",
       "153871      -         after_tes  antisense           NaN        NaN  0.751773   \n",
       "\n",
       "        mean_modc  \n",
       "311844   0.856091  \n",
       "246653   0.849446  \n",
       "214745   0.832180  \n",
       "279924   0.834170  \n",
       "56762    0.719680  \n",
       "154357   0.725905  \n",
       "298101   0.000000  \n",
       "232914   0.000784  \n",
       "154908   0.682169  \n",
       "57314    0.678294  \n",
       "154793   0.000506  \n",
       "57199    0.000455  \n",
       "195194   0.847215  \n",
       "260376   0.848889  \n",
       "153396   0.001415  \n",
       "55802    0.000687  \n",
       "56276    0.794199  \n",
       "153871   0.804965  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[\"Gene_id\"]==\"ENSMUSG00000000001.4\"][[\"strand\", \"Region\", \"Sense\", \"range_length\", \"cpg_count\", \"mean_mc\", \"mean_modc\"]].sort_values(by=\"Region\", ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "pivot table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pivot = df.pivot(\n",
    "    index=[\"Gene_id\", \"Gene_name\", \"contig\",\"strand\"],\n",
    "    # index = [\"Transcript_id\", \"contig\", \"strand\"],\n",
    "    columns=[\"Region\", \"Sense\"],\n",
    "    values=[\"mean_mc\", \"mean_hmc\", \"mean_modc\", \"cpg_count\",  \"range_length\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th colspan=\"10\" halign=\"left\">mean_mc</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">range_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>before_tss</th>\n",
       "      <th>after_tes</th>\n",
       "      <th>genes</th>\n",
       "      <th>first_exons</th>\n",
       "      <th>first_introns</th>\n",
       "      <th>before_tss</th>\n",
       "      <th>after_tes</th>\n",
       "      <th>genes</th>\n",
       "      <th>first_exons</th>\n",
       "      <th>first_introns</th>\n",
       "      <th>...</th>\n",
       "      <th>first_exons</th>\n",
       "      <th>first_introns</th>\n",
       "      <th>exons</th>\n",
       "      <th>introns</th>\n",
       "      <th>five_prime_utrs</th>\n",
       "      <th>three_prime_utrs</th>\n",
       "      <th>exons</th>\n",
       "      <th>introns</th>\n",
       "      <th>five_prime_utrs</th>\n",
       "      <th>three_prime_utrs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001.4</th>\n",
       "      <th>Gnai3</th>\n",
       "      <th>3</th>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.744475</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.751773</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.625551</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>13562.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>2054.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003.15</th>\n",
       "      <th>Pbsn</th>\n",
       "      <th>X</th>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.675325</td>\n",
       "      <td>0.749183</td>\n",
       "      <td>0.719298</td>\n",
       "      <td>0.751803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.718121</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>0.734848</td>\n",
       "      <td>0.767077</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>681.0</td>\n",
       "      <td>9532.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>235.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.15</th>\n",
       "      <th>Cdc45</th>\n",
       "      <th>16</th>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.786600</td>\n",
       "      <td>0.658988</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.775061</td>\n",
       "      <td>0.651237</td>\n",
       "      <td>0.002212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1955.0</td>\n",
       "      <td>29402.0</td>\n",
       "      <td>311.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037.17</th>\n",
       "      <th>Scml2</th>\n",
       "      <th>X</th>\n",
       "      <th>+</th>\n",
       "      <td>0.653846</td>\n",
       "      <td>0.697581</td>\n",
       "      <td>0.730139</td>\n",
       "      <td>0.761134</td>\n",
       "      <td>0.796358</td>\n",
       "      <td>0.616327</td>\n",
       "      <td>0.768559</td>\n",
       "      <td>0.730809</td>\n",
       "      <td>0.710692</td>\n",
       "      <td>0.787937</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2777.0</td>\n",
       "      <td>138112.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000049.11</th>\n",
       "      <th>Apoh</th>\n",
       "      <th>11</th>\n",
       "      <th>+</th>\n",
       "      <td>0.007270</td>\n",
       "      <td>0.813542</td>\n",
       "      <td>0.635118</td>\n",
       "      <td>0.819620</td>\n",
       "      <td>0.597780</td>\n",
       "      <td>0.003900</td>\n",
       "      <td>0.761813</td>\n",
       "      <td>0.651213</td>\n",
       "      <td>0.833846</td>\n",
       "      <td>0.621140</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1068.0</td>\n",
       "      <td>17921.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118623.1</th>\n",
       "      <th>AL935121.1</th>\n",
       "      <th>2</th>\n",
       "      <th>+</th>\n",
       "      <td>0.512397</td>\n",
       "      <td>0.585841</td>\n",
       "      <td>0.619217</td>\n",
       "      <td>0.764444</td>\n",
       "      <td>0.608056</td>\n",
       "      <td>0.261682</td>\n",
       "      <td>0.603261</td>\n",
       "      <td>0.603646</td>\n",
       "      <td>0.766055</td>\n",
       "      <td>0.663423</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2782.0</td>\n",
       "      <td>44783.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118638.1</th>\n",
       "      <th>AL805980.1</th>\n",
       "      <th>X</th>\n",
       "      <th>-</th>\n",
       "      <td>0.870370</td>\n",
       "      <td>0.853982</td>\n",
       "      <td>0.712744</td>\n",
       "      <td>0.793194</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.814815</td>\n",
       "      <td>0.791667</td>\n",
       "      <td>0.685039</td>\n",
       "      <td>0.744770</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>580.0</td>\n",
       "      <td>1162.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118640.1</th>\n",
       "      <th>AC167036.2</th>\n",
       "      <th>1</th>\n",
       "      <th>-</th>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.705426</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.863874</td>\n",
       "      <td>0.766129</td>\n",
       "      <td>0.809602</td>\n",
       "      <td>0.809602</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118646.1</th>\n",
       "      <th>AC160405.1</th>\n",
       "      <th>10</th>\n",
       "      <th>-</th>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.790244</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.756352</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>0.810799</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>0.829988</td>\n",
       "      <td>0.758004</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>39.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118653.1</th>\n",
       "      <th>AC159819.1</th>\n",
       "      <th>9</th>\n",
       "      <th>+</th>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.667582</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.720052</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21035 rows × 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  mean_mc                      \\\n",
       "Region                                         before_tss after_tes     genes   \n",
       "Sense                                               sense     sense     sense   \n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      -        0.000000  0.744475  0.670919   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -             NaN  0.675325  0.749183   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -        0.000000  0.786600  0.658988   \n",
       "ENSMUSG00000000037.17 Scml2      X      +        0.653846  0.697581  0.730139   \n",
       "ENSMUSG00000000049.11 Apoh       11     +        0.007270  0.813542  0.635118   \n",
       "...                                                   ...       ...       ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +        0.512397  0.585841  0.619217   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -        0.870370  0.853982  0.712744   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -        0.853868  0.705426  0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -        0.833333  0.790244  0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +        0.276243  0.667582  0.284238   \n",
       "\n",
       "                                                                          \\\n",
       "Region                                         first_exons first_introns   \n",
       "Sense                                                sense         sense   \n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      -         0.000455      0.624199   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -         0.719298      0.751803   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -         0.000000           NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +         0.761134      0.796358   \n",
       "ENSMUSG00000000049.11 Apoh       11     +         0.819620      0.597780   \n",
       "...                                                    ...           ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +         0.764444      0.608056   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -         0.793194           NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -         0.779018           NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -         0.777778      0.756352   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +         0.284238           NaN   \n",
       "\n",
       "                                                                               \\\n",
       "Region                                         before_tss after_tes     genes   \n",
       "Sense                                           antisense antisense antisense   \n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      -        0.001415  0.751773  0.672555   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -             NaN  0.718121  0.750896   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -        0.000000  0.775061  0.651237   \n",
       "ENSMUSG00000000037.17 Scml2      X      +        0.616327  0.768559  0.730809   \n",
       "ENSMUSG00000000049.11 Apoh       11     +        0.003900  0.761813  0.651213   \n",
       "...                                                   ...       ...       ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +        0.261682  0.603261  0.603646   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -        0.814815  0.791667  0.685039   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -        0.863874  0.766129  0.809602   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -        0.860000  0.810799  0.762000   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +        0.315789  0.720052  0.300631   \n",
       "\n",
       "                                                                          ...  \\\n",
       "Region                                         first_exons first_introns  ...   \n",
       "Sense                                            antisense     antisense  ...   \n",
       "Gene_id               Gene_name  contig strand                            ...   \n",
       "ENSMUSG00000000001.4  Gnai3      3      -         0.000000      0.625551  ...   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -         0.734848      0.767077  ...   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -         0.002212           NaN  ...   \n",
       "ENSMUSG00000000037.17 Scml2      X      +         0.710692      0.787937  ...   \n",
       "ENSMUSG00000000049.11 Apoh       11     +         0.833846      0.621140  ...   \n",
       "...                                                    ...           ...  ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +         0.766055      0.663423  ...   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -         0.744770           NaN  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -         0.809602           NaN  ...   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -         0.829988      0.758004  ...   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +         0.300631           NaN  ...   \n",
       "\n",
       "                                               range_length                \\\n",
       "Region                                          first_exons first_introns   \n",
       "Sense                                             antisense     antisense   \n",
       "Gene_id               Gene_name  contig strand                              \n",
       "ENSMUSG00000000001.4  Gnai3      3      -               NaN           NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -               NaN           NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -               NaN           NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +               NaN           NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +               NaN           NaN   \n",
       "...                                                     ...           ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +               NaN           NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -               NaN           NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -               NaN           NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -               NaN           NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +               NaN           NaN   \n",
       "\n",
       "                                                                  \\\n",
       "Region                                           exons   introns   \n",
       "Sense                                            sense     sense   \n",
       "Gene_id               Gene_name  contig strand                     \n",
       "ENSMUSG00000000001.4  Gnai3      3      -       2995.0   13562.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -        681.0    9532.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -       1955.0   29402.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +       2777.0  138112.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +       1068.0   17921.0   \n",
       "...                                                ...       ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +       2782.0   44783.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -        580.0    1162.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -          NaN       NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -         39.0       NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +          NaN       NaN   \n",
       "\n",
       "                                                                \\\n",
       "Region                                         five_prime_utrs   \n",
       "Sense                                                    sense   \n",
       "Gene_id               Gene_name  contig strand                   \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                140.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                139.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                311.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                  NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                 50.0   \n",
       "...                                                        ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                  NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                  NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                  NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                  NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                  NaN   \n",
       "\n",
       "                                                                           \\\n",
       "Region                                         three_prime_utrs     exons   \n",
       "Sense                                                     sense antisense   \n",
       "Gene_id               Gene_name  contig strand                              \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                2054.0       NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                 235.0       NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                   NaN       NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                   NaN       NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                 100.0       NaN   \n",
       "...                                                         ...       ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                   NaN       NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                   NaN       NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                   NaN       NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                   NaN       NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                   NaN       NaN   \n",
       "\n",
       "                                                                          \\\n",
       "Region                                           introns five_prime_utrs   \n",
       "Sense                                          antisense       antisense   \n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      -            NaN             NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -            NaN             NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -            NaN             NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +            NaN             NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +            NaN             NaN   \n",
       "...                                                  ...             ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +            NaN             NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -            NaN             NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -            NaN             NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -            NaN             NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +            NaN             NaN   \n",
       "\n",
       "                                                                 \n",
       "Region                                         three_prime_utrs  \n",
       "Sense                                                 antisense  \n",
       "Gene_id               Gene_name  contig strand                   \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                   NaN  \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                   NaN  \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                   NaN  \n",
       "ENSMUSG00000000037.17 Scml2      X      +                   NaN  \n",
       "ENSMUSG00000000049.11 Apoh       11     +                   NaN  \n",
       "...                                                         ...  \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                   NaN  \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                   NaN  \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                   NaN  \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                   NaN  \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                   NaN  \n",
       "\n",
       "[21035 rows x 90 columns]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pivot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>mean_mc_first_exons_sense</th>\n",
       "      <th>mean_mc_first_introns_sense</th>\n",
       "      <th>mean_mc_before_tss_antisense</th>\n",
       "      <th>mean_mc_after_tes_antisense</th>\n",
       "      <th>mean_mc_genes_antisense</th>\n",
       "      <th>mean_mc_first_exons_antisense</th>\n",
       "      <th>mean_mc_first_introns_antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>range_length_first_exons_antisense</th>\n",
       "      <th>range_length_first_introns_antisense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "      <th>range_length_exons_antisense</th>\n",
       "      <th>range_length_introns_antisense</th>\n",
       "      <th>range_length_five_prime_utrs_antisense</th>\n",
       "      <th>range_length_three_prime_utrs_antisense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001.4</th>\n",
       "      <th>Gnai3</th>\n",
       "      <th>3</th>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.744475</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.751773</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.625551</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>13562.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>2054.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003.15</th>\n",
       "      <th>Pbsn</th>\n",
       "      <th>X</th>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.675325</td>\n",
       "      <td>0.749183</td>\n",
       "      <td>0.719298</td>\n",
       "      <td>0.751803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.718121</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>0.734848</td>\n",
       "      <td>0.767077</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>681.0</td>\n",
       "      <td>9532.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>235.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.15</th>\n",
       "      <th>Cdc45</th>\n",
       "      <th>16</th>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.786600</td>\n",
       "      <td>0.658988</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.775061</td>\n",
       "      <td>0.651237</td>\n",
       "      <td>0.002212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1955.0</td>\n",
       "      <td>29402.0</td>\n",
       "      <td>311.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037.17</th>\n",
       "      <th>Scml2</th>\n",
       "      <th>X</th>\n",
       "      <th>+</th>\n",
       "      <td>0.653846</td>\n",
       "      <td>0.697581</td>\n",
       "      <td>0.730139</td>\n",
       "      <td>0.761134</td>\n",
       "      <td>0.796358</td>\n",
       "      <td>0.616327</td>\n",
       "      <td>0.768559</td>\n",
       "      <td>0.730809</td>\n",
       "      <td>0.710692</td>\n",
       "      <td>0.787937</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2777.0</td>\n",
       "      <td>138112.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000049.11</th>\n",
       "      <th>Apoh</th>\n",
       "      <th>11</th>\n",
       "      <th>+</th>\n",
       "      <td>0.007270</td>\n",
       "      <td>0.813542</td>\n",
       "      <td>0.635118</td>\n",
       "      <td>0.819620</td>\n",
       "      <td>0.597780</td>\n",
       "      <td>0.003900</td>\n",
       "      <td>0.761813</td>\n",
       "      <td>0.651213</td>\n",
       "      <td>0.833846</td>\n",
       "      <td>0.621140</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1068.0</td>\n",
       "      <td>17921.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118623.1</th>\n",
       "      <th>AL935121.1</th>\n",
       "      <th>2</th>\n",
       "      <th>+</th>\n",
       "      <td>0.512397</td>\n",
       "      <td>0.585841</td>\n",
       "      <td>0.619217</td>\n",
       "      <td>0.764444</td>\n",
       "      <td>0.608056</td>\n",
       "      <td>0.261682</td>\n",
       "      <td>0.603261</td>\n",
       "      <td>0.603646</td>\n",
       "      <td>0.766055</td>\n",
       "      <td>0.663423</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2782.0</td>\n",
       "      <td>44783.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118638.1</th>\n",
       "      <th>AL805980.1</th>\n",
       "      <th>X</th>\n",
       "      <th>-</th>\n",
       "      <td>0.870370</td>\n",
       "      <td>0.853982</td>\n",
       "      <td>0.712744</td>\n",
       "      <td>0.793194</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.814815</td>\n",
       "      <td>0.791667</td>\n",
       "      <td>0.685039</td>\n",
       "      <td>0.744770</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>580.0</td>\n",
       "      <td>1162.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118640.1</th>\n",
       "      <th>AC167036.2</th>\n",
       "      <th>1</th>\n",
       "      <th>-</th>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.705426</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.863874</td>\n",
       "      <td>0.766129</td>\n",
       "      <td>0.809602</td>\n",
       "      <td>0.809602</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118646.1</th>\n",
       "      <th>AC160405.1</th>\n",
       "      <th>10</th>\n",
       "      <th>-</th>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.790244</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.756352</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>0.810799</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>0.829988</td>\n",
       "      <td>0.758004</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>39.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118653.1</th>\n",
       "      <th>AC159819.1</th>\n",
       "      <th>9</th>\n",
       "      <th>+</th>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.667582</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.720052</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21035 rows × 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                mean_mc_before_tss_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                       0.000000   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                            NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                       0.000000   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                       0.653846   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                       0.007270   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                       0.512397   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                       0.870370   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                       0.853868   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                       0.833333   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                       0.276243   \n",
       "\n",
       "                                                mean_mc_after_tes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                      0.744475   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                      0.675325   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                      0.786600   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                      0.697581   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                      0.813542   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                      0.585841   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                      0.853982   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.705426   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                      0.790244   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.667582   \n",
       "\n",
       "                                                mean_mc_genes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                        \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                  0.670919   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                  0.749183   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                  0.658988   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                  0.730139   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                  0.635118   \n",
       "...                                                             ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                  0.619217   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                  0.712744   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                  0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                  0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                  0.284238   \n",
       "\n",
       "                                                mean_mc_first_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                              \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                        0.000455   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                        0.719298   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                        0.000000   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                        0.761134   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                        0.819620   \n",
       "...                                                                   ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                        0.764444   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                        0.793194   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                        0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                        0.777778   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                        0.284238   \n",
       "\n",
       "                                                mean_mc_first_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                          0.624199   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                          0.751803   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                               NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                          0.796358   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                          0.597780   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                          0.608056   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                               NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                          0.756352   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               NaN   \n",
       "\n",
       "                                                mean_mc_before_tss_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                           0.001415   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                           0.000000   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                           0.616327   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                           0.003900   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                           0.261682   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                           0.814815   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                           0.863874   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                           0.860000   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                           0.315789   \n",
       "\n",
       "                                                mean_mc_after_tes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                          0.751773   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                          0.718121   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                          0.775061   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                          0.768559   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                          0.761813   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                          0.603261   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                          0.791667   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                          0.766129   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                          0.810799   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                          0.720052   \n",
       "\n",
       "                                                mean_mc_genes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                      0.672555   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                      0.750896   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                      0.651237   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                      0.730809   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                      0.651213   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                      0.603646   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                      0.685039   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.809602   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                      0.762000   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.300631   \n",
       "\n",
       "                                                mean_mc_first_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                            0.000000   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                            0.734848   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                            0.002212   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                            0.710692   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                            0.833846   \n",
       "...                                                                       ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                            0.766055   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                            0.744770   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                            0.809602   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                            0.829988   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                            0.300631   \n",
       "\n",
       "                                                mean_mc_first_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                    \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                              0.625551   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                              0.767077   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                   NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                              0.787937   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                              0.621140   \n",
       "...                                                                         ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                              0.663423   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                   NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                   NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                              0.758004   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                   NaN   \n",
       "\n",
       "                                                ...  \\\n",
       "Gene_id               Gene_name  contig strand  ...   \n",
       "ENSMUSG00000000001.4  Gnai3      3      -       ...   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -       ...   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -       ...   \n",
       "ENSMUSG00000000037.17 Scml2      X      +       ...   \n",
       "ENSMUSG00000000049.11 Apoh       11     +       ...   \n",
       "...                                             ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +       ...   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -       ...   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -       ...   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +       ...   \n",
       "\n",
       "                                                range_length_first_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                      NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                      NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                      NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                      NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                      NaN   \n",
       "...                                                                            ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                      NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                      NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                      NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                      NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                      NaN   \n",
       "\n",
       "                                                range_length_first_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                         \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                        NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                        NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                        NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                        NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                        NaN   \n",
       "...                                                                              ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                        NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                        NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                        NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                        NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                        NaN   \n",
       "\n",
       "                                                range_length_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                         2995.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                          681.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                         1955.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                         2777.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                         1068.0   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                         2782.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                          580.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                            NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                           39.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                            NaN   \n",
       "\n",
       "                                                range_length_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                               \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                          13562.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                           9532.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                          29402.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                         138112.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                          17921.0   \n",
       "...                                                                    ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                          44783.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                           1162.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                              NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                              NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                              NaN   \n",
       "\n",
       "                                                range_length_five_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                    140.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                    139.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                    311.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                      NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                     50.0   \n",
       "...                                                                            ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                      NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                      NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                      NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                      NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                      NaN   \n",
       "\n",
       "                                                range_length_three_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                        \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                    2054.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                     235.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                       NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                       NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                     100.0   \n",
       "...                                                                             ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                       NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                       NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                       NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                       NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                       NaN   \n",
       "\n",
       "                                                range_length_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                NaN   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                NaN   \n",
       "\n",
       "                                                range_length_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                   \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                  NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                  NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                  NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                  NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                  NaN   \n",
       "...                                                                        ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                  NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                  NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                  NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                  NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                  NaN   \n",
       "\n",
       "                                                range_length_five_prime_utrs_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                           \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                          NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                          NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                          NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                          NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                          NaN   \n",
       "...                                                                                ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                          NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                          NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                          NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                          NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                          NaN   \n",
       "\n",
       "                                                range_length_three_prime_utrs_antisense  \n",
       "Gene_id               Gene_name  contig strand                                           \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                           NaN  \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                           NaN  \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                           NaN  \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                           NaN  \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                           NaN  \n",
       "...                                                                                 ...  \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                           NaN  \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                           NaN  \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                           NaN  \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                           NaN  \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                           NaN  \n",
       "\n",
       "[21035 rows x 90 columns]"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compost feature names\n",
    "df_features = df_pivot.copy()\n",
    "df_features.columns = [\" \".join(col).strip() for col in df_features.columns.values]\n",
    "features = df_features.columns\n",
    "features = [f.replace(\" \", \"_\") for f in features]\n",
    "df_features.columns = features\n",
    "df_features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "old"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>mean_mc_first_exons_sense</th>\n",
       "      <th>mean_mc_first_introns_sense</th>\n",
       "      <th>mean_mc_before_tss_antisense</th>\n",
       "      <th>mean_mc_after_tes_antisense</th>\n",
       "      <th>mean_mc_genes_antisense</th>\n",
       "      <th>mean_mc_first_exons_antisense</th>\n",
       "      <th>mean_mc_first_introns_antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>range_length_first_exons_antisense</th>\n",
       "      <th>range_length_first_introns_antisense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "      <th>range_length_exons_antisense</th>\n",
       "      <th>range_length_introns_antisense</th>\n",
       "      <th>range_length_five_prime_utrs_antisense</th>\n",
       "      <th>range_length_three_prime_utrs_antisense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000000001.4</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Gnai3</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">3</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.751773</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.625551</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.744475</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>13562.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>2054.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000000003.15</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Pbsn</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">X</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.718121</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>0.734848</td>\n",
       "      <td>0.767077</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.675325</td>\n",
       "      <td>0.749183</td>\n",
       "      <td>0.719298</td>\n",
       "      <td>0.751803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.749183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>681.0</td>\n",
       "      <td>9532.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>235.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.15</th>\n",
       "      <th>Cdc45</th>\n",
       "      <th>16</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.651237</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.775061</td>\n",
       "      <td>0.651237</td>\n",
       "      <td>0.002212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118640.1</th>\n",
       "      <th>AC167036.2</th>\n",
       "      <th>1</th>\n",
       "      <th>-</th>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.705426</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000118646.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">AC160405.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">10</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>0.810799</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>0.829988</td>\n",
       "      <td>0.758004</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.790244</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.756352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>39.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000118653.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">AC159819.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">9</th>\n",
       "      <th>+</th>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.667582</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.720052</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42046 rows × 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                mean_mc_before_tss_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            NaN   \n",
       "                                        -                       0.000000   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            NaN   \n",
       "                                        -                            NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            NaN   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                       0.853868   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            NaN   \n",
       "                                        -                       0.833333   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                       0.276243   \n",
       "                                        -                            NaN   \n",
       "\n",
       "                                                mean_mc_after_tes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                           NaN   \n",
       "                                        -                      0.744475   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                           NaN   \n",
       "                                        -                      0.675325   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                           NaN   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.705426   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                           NaN   \n",
       "                                        -                      0.790244   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.667582   \n",
       "                                        -                           NaN   \n",
       "\n",
       "                                                mean_mc_genes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                        \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                  0.672555   \n",
       "                                        -                  0.670919   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                  0.750896   \n",
       "                                        -                  0.749183   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                  0.651237   \n",
       "...                                                             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                  0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                  0.762000   \n",
       "                                        -                  0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                  0.284238   \n",
       "                                        -                  0.300631   \n",
       "\n",
       "                                                mean_mc_first_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                              \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                             NaN   \n",
       "                                        -                        0.000455   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                             NaN   \n",
       "                                        -                        0.719298   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                             NaN   \n",
       "...                                                                   ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                        0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                             NaN   \n",
       "                                        -                        0.777778   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                        0.284238   \n",
       "                                        -                             NaN   \n",
       "\n",
       "                                                mean_mc_first_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                               NaN   \n",
       "                                        -                          0.624199   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                               NaN   \n",
       "                                        -                          0.751803   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                               NaN   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                               NaN   \n",
       "                                        -                          0.756352   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               NaN   \n",
       "                                        -                               NaN   \n",
       "\n",
       "                                                mean_mc_before_tss_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                           0.001415   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                NaN   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                           0.000000   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                           0.860000   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                NaN   \n",
       "                                        -                           0.315789   \n",
       "\n",
       "                                                mean_mc_after_tes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                          0.751773   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                          0.718121   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                          0.775061   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                          0.810799   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               NaN   \n",
       "                                        -                          0.720052   \n",
       "\n",
       "                                                mean_mc_genes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                      0.672555   \n",
       "                                        -                      0.670919   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                      0.750896   \n",
       "                                        -                      0.749183   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                      0.651237   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                      0.762000   \n",
       "                                        -                      0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.284238   \n",
       "                                        -                      0.300631   \n",
       "\n",
       "                                                mean_mc_first_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            0.000000   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            0.734848   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            0.002212   \n",
       "...                                                                       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                 NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            0.829988   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                 NaN   \n",
       "                                        -                            0.300631   \n",
       "\n",
       "                                                mean_mc_first_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                    \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                              0.625551   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                              0.767077   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                   NaN   \n",
       "...                                                                         ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                   NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                              0.758004   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                   NaN   \n",
       "                                        -                                   NaN   \n",
       "\n",
       "                                                ...  \\\n",
       "Gene_id               Gene_name  contig strand  ...   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +       ...   \n",
       "...                                             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -       ...   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +       ...   \n",
       "                                        -       ...   \n",
       "\n",
       "                                                range_length_first_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                      NaN   \n",
       "...                                                                            ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                      NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "\n",
       "                                                range_length_first_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                         \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                        NaN   \n",
       "                                        -                                        NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                        NaN   \n",
       "                                        -                                        NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                        NaN   \n",
       "...                                                                              ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                        NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                        NaN   \n",
       "                                        -                                        NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                        NaN   \n",
       "                                        -                                        NaN   \n",
       "\n",
       "                                                range_length_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            NaN   \n",
       "                                        -                         2995.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            NaN   \n",
       "                                        -                          681.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            NaN   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                            NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            NaN   \n",
       "                                        -                           39.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                            NaN   \n",
       "                                        -                            NaN   \n",
       "\n",
       "                                                range_length_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                               \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                              NaN   \n",
       "                                        -                          13562.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                              NaN   \n",
       "                                        -                           9532.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                              NaN   \n",
       "...                                                                    ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                              NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                              NaN   \n",
       "                                        -                              NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                              NaN   \n",
       "                                        -                              NaN   \n",
       "\n",
       "                                                range_length_five_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                      NaN   \n",
       "                                        -                                    140.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                      NaN   \n",
       "                                        -                                    139.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                      NaN   \n",
       "...                                                                            ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                      NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "\n",
       "                                                range_length_three_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                        \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                       NaN   \n",
       "                                        -                                    2054.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                       NaN   \n",
       "                                        -                                     235.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                       NaN   \n",
       "...                                                                             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                       NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                       NaN   \n",
       "                                        -                                       NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                       NaN   \n",
       "                                        -                                       NaN   \n",
       "\n",
       "                                                range_length_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                NaN   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                NaN   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                NaN   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                NaN   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                NaN   \n",
       "                                        -                                NaN   \n",
       "\n",
       "                                                range_length_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                  NaN   \n",
       "                                        -                                  NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                  NaN   \n",
       "                                        -                                  NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                  NaN   \n",
       "...                                                                        ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                  NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                  NaN   \n",
       "                                        -                                  NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                  NaN   \n",
       "                                        -                                  NaN   \n",
       "\n",
       "                                                range_length_five_prime_utrs_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                           \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                          NaN   \n",
       "                                        -                                          NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                          NaN   \n",
       "                                        -                                          NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                          NaN   \n",
       "...                                                                                ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                          NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                          NaN   \n",
       "                                        -                                          NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                          NaN   \n",
       "                                        -                                          NaN   \n",
       "\n",
       "                                                range_length_three_prime_utrs_antisense  \n",
       "Gene_id               Gene_name  contig strand                                           \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                           NaN  \n",
       "                                        -                                           NaN  \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                           NaN  \n",
       "                                        -                                           NaN  \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                           NaN  \n",
       "...                                                                                 ...  \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                           NaN  \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                           NaN  \n",
       "                                        -                                           NaN  \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                           NaN  \n",
       "                                        -                                           NaN  \n",
       "\n",
       "[42046 rows x 90 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # compost feature names\n",
    "# df_features = df_pivot.copy()\n",
    "# df_features.columns = [\" \".join(col).strip() for col in df_features.columns.values]\n",
    "# features = df_features.columns\n",
    "# features = [f.replace(\" \", \"_\") for f in features]\n",
    "# df_features.columns = features\n",
    "# df_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cpg_count_before_tss_antisense',\n",
       " 'cpg_count_after_tes_antisense',\n",
       " 'cpg_count_genes_antisense',\n",
       " 'cpg_count_first_exons_antisense',\n",
       " 'cpg_count_first_introns_antisense',\n",
       " 'cpg_count_exons_antisense',\n",
       " 'cpg_count_introns_antisense',\n",
       " 'cpg_count_five_prime_utrs_antisense',\n",
       " 'cpg_count_three_prime_utrs_antisense',\n",
       " 'range_length_before_tss_antisense',\n",
       " 'range_length_after_tes_antisense',\n",
       " 'range_length_genes_antisense',\n",
       " 'range_length_first_exons_antisense',\n",
       " 'range_length_first_introns_antisense',\n",
       " 'range_length_exons_antisense',\n",
       " 'range_length_introns_antisense',\n",
       " 'range_length_five_prime_utrs_antisense',\n",
       " 'range_length_three_prime_utrs_antisense']"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[ c for c in df_features.columns.to_list()  if c  not in df_features.dropna(axis=1, how='all').columns.to_list()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>mean_mc_first_exons_sense</th>\n",
       "      <th>mean_mc_first_introns_sense</th>\n",
       "      <th>mean_mc_before_tss_antisense</th>\n",
       "      <th>mean_mc_after_tes_antisense</th>\n",
       "      <th>mean_mc_genes_antisense</th>\n",
       "      <th>mean_mc_first_exons_antisense</th>\n",
       "      <th>mean_mc_first_introns_antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>cpg_count_three_prime_utrs_sense</th>\n",
       "      <th>range_length_before_tss_sense</th>\n",
       "      <th>range_length_after_tes_sense</th>\n",
       "      <th>range_length_genes_sense</th>\n",
       "      <th>range_length_first_exons_sense</th>\n",
       "      <th>range_length_first_introns_sense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001.4</th>\n",
       "      <th>Gnai3</th>\n",
       "      <th>3</th>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.744475</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.751773</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.625551</td>\n",
       "      <td>...</td>\n",
       "      <td>32.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>22051.0</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>13562.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>2054.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003.15</th>\n",
       "      <th>Pbsn</th>\n",
       "      <th>X</th>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.675325</td>\n",
       "      <td>0.749183</td>\n",
       "      <td>0.719298</td>\n",
       "      <td>0.751803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.718121</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>0.734848</td>\n",
       "      <td>0.767077</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>15722.0</td>\n",
       "      <td>214.0</td>\n",
       "      <td>5295.0</td>\n",
       "      <td>681.0</td>\n",
       "      <td>9532.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>235.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.15</th>\n",
       "      <th>Cdc45</th>\n",
       "      <th>16</th>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.786600</td>\n",
       "      <td>0.658988</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.775061</td>\n",
       "      <td>0.651237</td>\n",
       "      <td>0.002212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>31540.0</td>\n",
       "      <td>168.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1955.0</td>\n",
       "      <td>29402.0</td>\n",
       "      <td>311.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037.17</th>\n",
       "      <th>Scml2</th>\n",
       "      <th>X</th>\n",
       "      <th>+</th>\n",
       "      <td>0.653846</td>\n",
       "      <td>0.697581</td>\n",
       "      <td>0.730139</td>\n",
       "      <td>0.761134</td>\n",
       "      <td>0.796358</td>\n",
       "      <td>0.616327</td>\n",
       "      <td>0.768559</td>\n",
       "      <td>0.730809</td>\n",
       "      <td>0.710692</td>\n",
       "      <td>0.787937</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>175688.0</td>\n",
       "      <td>131.0</td>\n",
       "      <td>34668.0</td>\n",
       "      <td>2777.0</td>\n",
       "      <td>138112.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000049.11</th>\n",
       "      <th>Apoh</th>\n",
       "      <th>11</th>\n",
       "      <th>+</th>\n",
       "      <td>0.007270</td>\n",
       "      <td>0.813542</td>\n",
       "      <td>0.635118</td>\n",
       "      <td>0.819620</td>\n",
       "      <td>0.597780</td>\n",
       "      <td>0.003900</td>\n",
       "      <td>0.761813</td>\n",
       "      <td>0.651213</td>\n",
       "      <td>0.833846</td>\n",
       "      <td>0.621140</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>71042.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>51939.0</td>\n",
       "      <td>1068.0</td>\n",
       "      <td>17921.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118623.1</th>\n",
       "      <th>AL935121.1</th>\n",
       "      <th>2</th>\n",
       "      <th>+</th>\n",
       "      <td>0.512397</td>\n",
       "      <td>0.585841</td>\n",
       "      <td>0.619217</td>\n",
       "      <td>0.764444</td>\n",
       "      <td>0.608056</td>\n",
       "      <td>0.261682</td>\n",
       "      <td>0.603261</td>\n",
       "      <td>0.603646</td>\n",
       "      <td>0.766055</td>\n",
       "      <td>0.663423</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>48754.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>1102.0</td>\n",
       "      <td>2782.0</td>\n",
       "      <td>44783.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118638.1</th>\n",
       "      <th>AL805980.1</th>\n",
       "      <th>X</th>\n",
       "      <th>-</th>\n",
       "      <td>0.870370</td>\n",
       "      <td>0.853982</td>\n",
       "      <td>0.712744</td>\n",
       "      <td>0.793194</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.814815</td>\n",
       "      <td>0.791667</td>\n",
       "      <td>0.685039</td>\n",
       "      <td>0.744770</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>3793.0</td>\n",
       "      <td>1432.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>580.0</td>\n",
       "      <td>1162.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118640.1</th>\n",
       "      <th>AC167036.2</th>\n",
       "      <th>1</th>\n",
       "      <th>-</th>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.705426</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.863874</td>\n",
       "      <td>0.766129</td>\n",
       "      <td>0.809602</td>\n",
       "      <td>0.809602</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>890.0</td>\n",
       "      <td>890.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118646.1</th>\n",
       "      <th>AC160405.1</th>\n",
       "      <th>10</th>\n",
       "      <th>-</th>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.790244</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.756352</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>0.810799</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>0.829988</td>\n",
       "      <td>0.758004</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>19585.0</td>\n",
       "      <td>342.0</td>\n",
       "      <td>19204.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118653.1</th>\n",
       "      <th>AC159819.1</th>\n",
       "      <th>9</th>\n",
       "      <th>+</th>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.667582</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.720052</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>987.0</td>\n",
       "      <td>987.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21035 rows × 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                mean_mc_before_tss_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                       0.000000   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                            NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                       0.000000   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                       0.653846   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                       0.007270   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                       0.512397   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                       0.870370   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                       0.853868   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                       0.833333   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                       0.276243   \n",
       "\n",
       "                                                mean_mc_after_tes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                      0.744475   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                      0.675325   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                      0.786600   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                      0.697581   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                      0.813542   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                      0.585841   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                      0.853982   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.705426   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                      0.790244   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.667582   \n",
       "\n",
       "                                                mean_mc_genes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                        \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                  0.670919   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                  0.749183   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                  0.658988   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                  0.730139   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                  0.635118   \n",
       "...                                                             ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                  0.619217   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                  0.712744   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                  0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                  0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                  0.284238   \n",
       "\n",
       "                                                mean_mc_first_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                              \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                        0.000455   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                        0.719298   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                        0.000000   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                        0.761134   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                        0.819620   \n",
       "...                                                                   ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                        0.764444   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                        0.793194   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                        0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                        0.777778   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                        0.284238   \n",
       "\n",
       "                                                mean_mc_first_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                          0.624199   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                          0.751803   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                               NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                          0.796358   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                          0.597780   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                          0.608056   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                               NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                          0.756352   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               NaN   \n",
       "\n",
       "                                                mean_mc_before_tss_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                           0.001415   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                           0.000000   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                           0.616327   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                           0.003900   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                           0.261682   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                           0.814815   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                           0.863874   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                           0.860000   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                           0.315789   \n",
       "\n",
       "                                                mean_mc_after_tes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                          0.751773   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                          0.718121   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                          0.775061   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                          0.768559   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                          0.761813   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                          0.603261   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                          0.791667   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                          0.766129   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                          0.810799   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                          0.720052   \n",
       "\n",
       "                                                mean_mc_genes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                      0.672555   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                      0.750896   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                      0.651237   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                      0.730809   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                      0.651213   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                      0.603646   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                      0.685039   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.809602   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                      0.762000   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.300631   \n",
       "\n",
       "                                                mean_mc_first_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                            0.000000   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                            0.734848   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                            0.002212   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                            0.710692   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                            0.833846   \n",
       "...                                                                       ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                            0.766055   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                            0.744770   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                            0.809602   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                            0.829988   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                            0.300631   \n",
       "\n",
       "                                                mean_mc_first_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                    \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                              0.625551   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                              0.767077   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                   NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                              0.787937   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                              0.621140   \n",
       "...                                                                         ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                              0.663423   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                   NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                   NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                              0.758004   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                   NaN   \n",
       "\n",
       "                                                ...  \\\n",
       "Gene_id               Gene_name  contig strand  ...   \n",
       "ENSMUSG00000000001.4  Gnai3      3      -       ...   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -       ...   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -       ...   \n",
       "ENSMUSG00000000037.17 Scml2      X      +       ...   \n",
       "ENSMUSG00000000049.11 Apoh       11     +       ...   \n",
       "...                                             ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +       ...   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -       ...   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -       ...   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +       ...   \n",
       "\n",
       "                                                cpg_count_three_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                     \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                   32.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                    1.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                    NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                    NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                    2.0   \n",
       "...                                                                          ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                    NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                    NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                    NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                    NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                    NaN   \n",
       "\n",
       "                                                range_length_before_tss_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                               200.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                 NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                               200.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                               200.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                               200.0   \n",
       "...                                                                       ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                               200.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                               200.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               200.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                               200.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               200.0   \n",
       "\n",
       "                                                range_length_after_tes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                             1000.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                             1000.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                             1000.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                             1000.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                             1000.0   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                             1000.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                             1000.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                             1000.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                             1000.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                             1000.0   \n",
       "\n",
       "                                                range_length_genes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                        38866.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                        15722.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                        31540.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                       175688.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                        71042.0   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                        48754.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                         3793.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                          890.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                        19585.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                          987.0   \n",
       "\n",
       "                                                range_length_first_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                   \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                258.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                214.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                168.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                131.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                114.0   \n",
       "...                                                                        ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                 87.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                               1432.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                890.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                342.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                987.0   \n",
       "\n",
       "                                                range_length_first_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                     \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                22051.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                 5295.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                    NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                34668.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                51939.0   \n",
       "...                                                                          ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                 1102.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                    NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                    NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                19204.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                    NaN   \n",
       "\n",
       "                                                range_length_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                         2995.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                          681.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                         1955.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                         2777.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                         1068.0   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                         2782.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                          580.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                            NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                           39.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                            NaN   \n",
       "\n",
       "                                                range_length_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                               \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                          13562.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                           9532.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                          29402.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                         138112.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                          17921.0   \n",
       "...                                                                    ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                          44783.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                           1162.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                              NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                              NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                              NaN   \n",
       "\n",
       "                                                range_length_five_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                    140.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                    139.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                    311.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                      NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                     50.0   \n",
       "...                                                                            ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                      NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                      NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                      NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                      NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                      NaN   \n",
       "\n",
       "                                                range_length_three_prime_utrs_sense  \n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                    2054.0  \n",
       "ENSMUSG00000000003.15 Pbsn       X      -                                     235.0  \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                       NaN  \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                       NaN  \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                     100.0  \n",
       "...                                                                             ...  \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                       NaN  \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                       NaN  \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                       NaN  \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                       NaN  \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                       NaN  \n",
       "\n",
       "[21035 rows x 72 columns]"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features_cleaned = df_features.dropna(axis=1, how='all')\n",
    "\n",
    "# Display the cleaned DataFrame without columns that have all NaN values\n",
    "df_features_cleaned\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>mean_mc_first_exons_sense</th>\n",
       "      <th>mean_mc_first_introns_sense</th>\n",
       "      <th>mean_mc_before_tss_antisense</th>\n",
       "      <th>mean_mc_after_tes_antisense</th>\n",
       "      <th>mean_mc_genes_antisense</th>\n",
       "      <th>mean_mc_first_exons_antisense</th>\n",
       "      <th>mean_mc_first_introns_antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>cpg_count_three_prime_utrs_sense</th>\n",
       "      <th>range_length_before_tss_sense</th>\n",
       "      <th>range_length_after_tes_sense</th>\n",
       "      <th>range_length_genes_sense</th>\n",
       "      <th>range_length_first_exons_sense</th>\n",
       "      <th>range_length_first_introns_sense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gnai3</th>\n",
       "      <th>3</th>\n",
       "      <th>-</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.744475</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.751773</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.625551</td>\n",
       "      <td>...</td>\n",
       "      <td>32.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>22051.0</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>13562.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>2054.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         mean_mc_before_tss_sense  mean_mc_after_tes_sense  \\\n",
       "Gene_name contig strand                                                      \n",
       "Gnai3     3      -                            0.0                 0.744475   \n",
       "\n",
       "                         mean_mc_genes_sense  mean_mc_first_exons_sense  \\\n",
       "Gene_name contig strand                                                   \n",
       "Gnai3     3      -                  0.670919                   0.000455   \n",
       "\n",
       "                         mean_mc_first_introns_sense  \\\n",
       "Gene_name contig strand                                \n",
       "Gnai3     3      -                          0.624199   \n",
       "\n",
       "                         mean_mc_before_tss_antisense  \\\n",
       "Gene_name contig strand                                 \n",
       "Gnai3     3      -                           0.001415   \n",
       "\n",
       "                         mean_mc_after_tes_antisense  mean_mc_genes_antisense  \\\n",
       "Gene_name contig strand                                                         \n",
       "Gnai3     3      -                          0.751773                 0.672555   \n",
       "\n",
       "                         mean_mc_first_exons_antisense  \\\n",
       "Gene_name contig strand                                  \n",
       "Gnai3     3      -                                 0.0   \n",
       "\n",
       "                         mean_mc_first_introns_antisense  ...  \\\n",
       "Gene_name contig strand                                   ...   \n",
       "Gnai3     3      -                              0.625551  ...   \n",
       "\n",
       "                         cpg_count_three_prime_utrs_sense  \\\n",
       "Gene_name contig strand                                     \n",
       "Gnai3     3      -                                   32.0   \n",
       "\n",
       "                         range_length_before_tss_sense  \\\n",
       "Gene_name contig strand                                  \n",
       "Gnai3     3      -                               200.0   \n",
       "\n",
       "                         range_length_after_tes_sense  \\\n",
       "Gene_name contig strand                                 \n",
       "Gnai3     3      -                             1000.0   \n",
       "\n",
       "                         range_length_genes_sense  \\\n",
       "Gene_name contig strand                             \n",
       "Gnai3     3      -                        38866.0   \n",
       "\n",
       "                         range_length_first_exons_sense  \\\n",
       "Gene_name contig strand                                   \n",
       "Gnai3     3      -                                258.0   \n",
       "\n",
       "                         range_length_first_introns_sense  \\\n",
       "Gene_name contig strand                                     \n",
       "Gnai3     3      -                                22051.0   \n",
       "\n",
       "                         range_length_exons_sense  range_length_introns_sense  \\\n",
       "Gene_name contig strand                                                         \n",
       "Gnai3     3      -                         2995.0                     13562.0   \n",
       "\n",
       "                         range_length_five_prime_utrs_sense  \\\n",
       "Gene_name contig strand                                       \n",
       "Gnai3     3      -                                    140.0   \n",
       "\n",
       "                         range_length_three_prime_utrs_sense  \n",
       "Gene_name contig strand                                       \n",
       "Gnai3     3      -                                    2054.0  \n",
       "\n",
       "[1 rows x 72 columns]"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features_cleaned.loc[\"ENSMUSG00000000001.4\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### write to file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we write this feature file as a pickle file, which preserves the variable type, and is therefore prefered over a more basic text file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Gene_id', 'Gene_name', 'contig', 'strand', 'mean_mc_before_tss_sense',\n",
       "       'mean_mc_after_tes_sense', 'mean_mc_genes_sense',\n",
       "       'mean_mc_first_exons_sense', 'mean_mc_first_introns_sense',\n",
       "       'mean_mc_before_tss_antisense', 'mean_mc_after_tes_antisense',\n",
       "       'mean_mc_genes_antisense', 'mean_mc_first_exons_antisense',\n",
       "       'mean_mc_first_introns_antisense', 'mean_mc_exons_sense',\n",
       "       'mean_mc_introns_sense', 'mean_mc_five_prime_utrs_sense',\n",
       "       'mean_mc_three_prime_utrs_sense', 'mean_mc_exons_antisense',\n",
       "       'mean_mc_introns_antisense', 'mean_mc_five_prime_utrs_antisense',\n",
       "       'mean_mc_three_prime_utrs_antisense', 'mean_hmc_before_tss_sense',\n",
       "       'mean_hmc_after_tes_sense', 'mean_hmc_genes_sense',\n",
       "       'mean_hmc_first_exons_sense', 'mean_hmc_first_introns_sense',\n",
       "       'mean_hmc_before_tss_antisense', 'mean_hmc_after_tes_antisense',\n",
       "       'mean_hmc_genes_antisense', 'mean_hmc_first_exons_antisense',\n",
       "       'mean_hmc_first_introns_antisense', 'mean_hmc_exons_sense',\n",
       "       'mean_hmc_introns_sense', 'mean_hmc_five_prime_utrs_sense',\n",
       "       'mean_hmc_three_prime_utrs_sense', 'mean_hmc_exons_antisense',\n",
       "       'mean_hmc_introns_antisense', 'mean_hmc_five_prime_utrs_antisense',\n",
       "       'mean_hmc_three_prime_utrs_antisense', 'mean_modc_before_tss_sense',\n",
       "       'mean_modc_after_tes_sense', 'mean_modc_genes_sense',\n",
       "       'mean_modc_first_exons_sense', 'mean_modc_first_introns_sense',\n",
       "       'mean_modc_before_tss_antisense', 'mean_modc_after_tes_antisense',\n",
       "       'mean_modc_genes_antisense', 'mean_modc_first_exons_antisense',\n",
       "       'mean_modc_first_introns_antisense', 'mean_modc_exons_sense',\n",
       "       'mean_modc_introns_sense', 'mean_modc_five_prime_utrs_sense',\n",
       "       'mean_modc_three_prime_utrs_sense', 'mean_modc_exons_antisense',\n",
       "       'mean_modc_introns_antisense', 'mean_modc_five_prime_utrs_antisense',\n",
       "       'mean_modc_three_prime_utrs_antisense', 'cpg_count_before_tss_sense',\n",
       "       'cpg_count_after_tes_sense', 'cpg_count_genes_sense',\n",
       "       'cpg_count_first_exons_sense', 'cpg_count_first_introns_sense',\n",
       "       'cpg_count_exons_sense', 'cpg_count_introns_sense',\n",
       "       'cpg_count_five_prime_utrs_sense', 'cpg_count_three_prime_utrs_sense',\n",
       "       'range_length_before_tss_sense', 'range_length_after_tes_sense',\n",
       "       'range_length_genes_sense', 'range_length_first_exons_sense',\n",
       "       'range_length_first_introns_sense', 'range_length_exons_sense',\n",
       "       'range_length_introns_sense', 'range_length_five_prime_utrs_sense',\n",
       "       'range_length_three_prime_utrs_sense', 'selected_transcript'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features_cleaned = df_features_cleaned.reset_index()\n",
    "# add a column with a boolean depending on whether the transcript is is selected_transcripts or not\n",
    "df_features_cleaned[\"selected_transcript\"] = df_features_cleaned.Gene_id.isin(selected_transcripts.gene_id)\n",
    "df_features_cleaned.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Source', 'Type', 'Start', 'End', 'Score', 'Strand',\n",
       "       'Phase', 'Id', 'Gene_id', 'Gene_type', 'Gene_name', 'Level', 'Mgi_id',\n",
       "       'Havana_gene', 'Tag', 'Ranges_ID'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_features_cleaned = df_features_cleaned[~df_features_cleaned['contig'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>mean_mc_first_exons_sense</th>\n",
       "      <th>mean_mc_first_introns_sense</th>\n",
       "      <th>mean_mc_before_tss_antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>range_length_before_tss_sense</th>\n",
       "      <th>range_length_after_tes_sense</th>\n",
       "      <th>range_length_genes_sense</th>\n",
       "      <th>range_length_first_exons_sense</th>\n",
       "      <th>range_length_first_introns_sense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "      <th>selected_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.744475</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>22051.0</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>13562.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>2054.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003.15</td>\n",
       "      <td>Pbsn</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.675325</td>\n",
       "      <td>0.749183</td>\n",
       "      <td>0.719298</td>\n",
       "      <td>0.751803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>15722.0</td>\n",
       "      <td>214.0</td>\n",
       "      <td>5295.0</td>\n",
       "      <td>681.0</td>\n",
       "      <td>9532.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>235.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028.15</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.786600</td>\n",
       "      <td>0.658988</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>31540.0</td>\n",
       "      <td>168.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1955.0</td>\n",
       "      <td>29402.0</td>\n",
       "      <td>311.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000037.17</td>\n",
       "      <td>Scml2</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>0.653846</td>\n",
       "      <td>0.697581</td>\n",
       "      <td>0.730139</td>\n",
       "      <td>0.761134</td>\n",
       "      <td>0.796358</td>\n",
       "      <td>0.616327</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>175688.0</td>\n",
       "      <td>131.0</td>\n",
       "      <td>34668.0</td>\n",
       "      <td>2777.0</td>\n",
       "      <td>138112.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000049.11</td>\n",
       "      <td>Apoh</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "      <td>0.007270</td>\n",
       "      <td>0.813542</td>\n",
       "      <td>0.635118</td>\n",
       "      <td>0.819620</td>\n",
       "      <td>0.597780</td>\n",
       "      <td>0.003900</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>71042.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>51939.0</td>\n",
       "      <td>1068.0</td>\n",
       "      <td>17921.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21030</th>\n",
       "      <td>ENSMUSG00000118623.1</td>\n",
       "      <td>AL935121.1</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>0.512397</td>\n",
       "      <td>0.585841</td>\n",
       "      <td>0.619217</td>\n",
       "      <td>0.764444</td>\n",
       "      <td>0.608056</td>\n",
       "      <td>0.261682</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>48754.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>1102.0</td>\n",
       "      <td>2782.0</td>\n",
       "      <td>44783.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21031</th>\n",
       "      <td>ENSMUSG00000118638.1</td>\n",
       "      <td>AL805980.1</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>0.870370</td>\n",
       "      <td>0.853982</td>\n",
       "      <td>0.712744</td>\n",
       "      <td>0.793194</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.814815</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>3793.0</td>\n",
       "      <td>1432.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>580.0</td>\n",
       "      <td>1162.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21032</th>\n",
       "      <td>ENSMUSG00000118640.1</td>\n",
       "      <td>AC167036.2</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.705426</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.863874</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>890.0</td>\n",
       "      <td>890.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21033</th>\n",
       "      <td>ENSMUSG00000118646.1</td>\n",
       "      <td>AC160405.1</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.790244</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.756352</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>19585.0</td>\n",
       "      <td>342.0</td>\n",
       "      <td>19204.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21034</th>\n",
       "      <td>ENSMUSG00000118653.1</td>\n",
       "      <td>AC159819.1</td>\n",
       "      <td>9</td>\n",
       "      <td>+</td>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.667582</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>987.0</td>\n",
       "      <td>987.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21035 rows × 77 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_id   Gene_name contig strand  \\\n",
       "0       ENSMUSG00000000001.4       Gnai3      3      -   \n",
       "1      ENSMUSG00000000003.15        Pbsn      X      -   \n",
       "2      ENSMUSG00000000028.15       Cdc45     16      -   \n",
       "3      ENSMUSG00000000037.17       Scml2      X      +   \n",
       "4      ENSMUSG00000000049.11        Apoh     11      +   \n",
       "...                      ...         ...    ...    ...   \n",
       "21030   ENSMUSG00000118623.1  AL935121.1      2      +   \n",
       "21031   ENSMUSG00000118638.1  AL805980.1      X      -   \n",
       "21032   ENSMUSG00000118640.1  AC167036.2      1      -   \n",
       "21033   ENSMUSG00000118646.1  AC160405.1     10      -   \n",
       "21034   ENSMUSG00000118653.1  AC159819.1      9      +   \n",
       "\n",
       "       mean_mc_before_tss_sense  mean_mc_after_tes_sense  mean_mc_genes_sense  \\\n",
       "0                      0.000000                 0.744475             0.670919   \n",
       "1                           NaN                 0.675325             0.749183   \n",
       "2                      0.000000                 0.786600             0.658988   \n",
       "3                      0.653846                 0.697581             0.730139   \n",
       "4                      0.007270                 0.813542             0.635118   \n",
       "...                         ...                      ...                  ...   \n",
       "21030                  0.512397                 0.585841             0.619217   \n",
       "21031                  0.870370                 0.853982             0.712744   \n",
       "21032                  0.853868                 0.705426             0.779018   \n",
       "21033                  0.833333                 0.790244             0.757576   \n",
       "21034                  0.276243                 0.667582             0.284238   \n",
       "\n",
       "       mean_mc_first_exons_sense  mean_mc_first_introns_sense  \\\n",
       "0                       0.000455                     0.624199   \n",
       "1                       0.719298                     0.751803   \n",
       "2                       0.000000                          NaN   \n",
       "3                       0.761134                     0.796358   \n",
       "4                       0.819620                     0.597780   \n",
       "...                          ...                          ...   \n",
       "21030                   0.764444                     0.608056   \n",
       "21031                   0.793194                          NaN   \n",
       "21032                   0.779018                          NaN   \n",
       "21033                   0.777778                     0.756352   \n",
       "21034                   0.284238                          NaN   \n",
       "\n",
       "       mean_mc_before_tss_antisense  ...  range_length_before_tss_sense  \\\n",
       "0                          0.001415  ...                          200.0   \n",
       "1                               NaN  ...                            NaN   \n",
       "2                          0.000000  ...                          200.0   \n",
       "3                          0.616327  ...                          200.0   \n",
       "4                          0.003900  ...                          200.0   \n",
       "...                             ...  ...                            ...   \n",
       "21030                      0.261682  ...                          200.0   \n",
       "21031                      0.814815  ...                          200.0   \n",
       "21032                      0.863874  ...                          200.0   \n",
       "21033                      0.860000  ...                          200.0   \n",
       "21034                      0.315789  ...                          200.0   \n",
       "\n",
       "       range_length_after_tes_sense  range_length_genes_sense  \\\n",
       "0                            1000.0                   38866.0   \n",
       "1                            1000.0                   15722.0   \n",
       "2                            1000.0                   31540.0   \n",
       "3                            1000.0                  175688.0   \n",
       "4                            1000.0                   71042.0   \n",
       "...                             ...                       ...   \n",
       "21030                        1000.0                   48754.0   \n",
       "21031                        1000.0                    3793.0   \n",
       "21032                        1000.0                     890.0   \n",
       "21033                        1000.0                   19585.0   \n",
       "21034                        1000.0                     987.0   \n",
       "\n",
       "       range_length_first_exons_sense  range_length_first_introns_sense  \\\n",
       "0                               258.0                           22051.0   \n",
       "1                               214.0                            5295.0   \n",
       "2                               168.0                               NaN   \n",
       "3                               131.0                           34668.0   \n",
       "4                               114.0                           51939.0   \n",
       "...                               ...                               ...   \n",
       "21030                            87.0                            1102.0   \n",
       "21031                          1432.0                               NaN   \n",
       "21032                           890.0                               NaN   \n",
       "21033                           342.0                           19204.0   \n",
       "21034                           987.0                               NaN   \n",
       "\n",
       "       range_length_exons_sense  range_length_introns_sense  \\\n",
       "0                        2995.0                     13562.0   \n",
       "1                         681.0                      9532.0   \n",
       "2                        1955.0                     29402.0   \n",
       "3                        2777.0                    138112.0   \n",
       "4                        1068.0                     17921.0   \n",
       "...                         ...                         ...   \n",
       "21030                    2782.0                     44783.0   \n",
       "21031                     580.0                      1162.0   \n",
       "21032                       NaN                         NaN   \n",
       "21033                      39.0                         NaN   \n",
       "21034                       NaN                         NaN   \n",
       "\n",
       "       range_length_five_prime_utrs_sense  \\\n",
       "0                                   140.0   \n",
       "1                                   139.0   \n",
       "2                                   311.0   \n",
       "3                                     NaN   \n",
       "4                                    50.0   \n",
       "...                                   ...   \n",
       "21030                                 NaN   \n",
       "21031                                 NaN   \n",
       "21032                                 NaN   \n",
       "21033                                 NaN   \n",
       "21034                                 NaN   \n",
       "\n",
       "       range_length_three_prime_utrs_sense  selected_transcript  \n",
       "0                                   2054.0                 True  \n",
       "1                                    235.0                 True  \n",
       "2                                      NaN                 True  \n",
       "3                                      NaN                 True  \n",
       "4                                    100.0                 True  \n",
       "...                                    ...                  ...  \n",
       "21030                                  NaN                 True  \n",
       "21031                                  NaN                 True  \n",
       "21032                                  NaN                 True  \n",
       "21033                                  NaN                 True  \n",
       "21034                                  NaN                 True  \n",
       "\n",
       "[21035 rows x 77 columns]"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features_cleaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>...</th>\n",
       "      <th>range_length_before_tss_sense</th>\n",
       "      <th>range_length_after_tes_sense</th>\n",
       "      <th>range_length_genes_sense</th>\n",
       "      <th>range_length_first_exons_sense</th>\n",
       "      <th>range_length_first_introns_sense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "      <th>selected_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>1</td>\n",
       "      <td>4807787</td>\n",
       "      <td>4848409</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>0.529368</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>40622.0</td>\n",
       "      <td>159.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>2339.0</td>\n",
       "      <td>38089.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>1722.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4807891</td>\n",
       "      <td>4886769</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0.000689</td>\n",
       "      <td>0.686047</td>\n",
       "      <td>0.472292</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>78878.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>20602.0</td>\n",
       "      <td>725.0</td>\n",
       "      <td>57461.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>1</td>\n",
       "      <td>4857813</td>\n",
       "      <td>4897908</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.800370</td>\n",
       "      <td>0.514710</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>40095.0</td>\n",
       "      <td>162.0</td>\n",
       "      <td>9494.0</td>\n",
       "      <td>2375.0</td>\n",
       "      <td>28064.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>1540.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>1</td>\n",
       "      <td>5070017</td>\n",
       "      <td>5162528</td>\n",
       "      <td>Atp6v1h</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.847909</td>\n",
       "      <td>0.648667</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>92511.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>14273.0</td>\n",
       "      <td>1523.0</td>\n",
       "      <td>76589.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>1</td>\n",
       "      <td>5588465</td>\n",
       "      <td>5606130</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0.271457</td>\n",
       "      <td>0.759259</td>\n",
       "      <td>0.680438</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>17665.0</td>\n",
       "      <td>170.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>4501.0</td>\n",
       "      <td>12967.0</td>\n",
       "      <td>183.0</td>\n",
       "      <td>3346.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21030</th>\n",
       "      <td>ENSMUSG00000096902.7</td>\n",
       "      <td>Y</td>\n",
       "      <td>72220862</td>\n",
       "      <td>72247117</td>\n",
       "      <td>Gm20843</td>\n",
       "      <td>Y</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.727273</td>\n",
       "      <td>0.629630</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>26255.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>872.0</td>\n",
       "      <td>23977.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>197.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21031</th>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>Y</td>\n",
       "      <td>79148788</td>\n",
       "      <td>79151121</td>\n",
       "      <td>Gm20917</td>\n",
       "      <td>Y</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.782609</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2333.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1111.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21032</th>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>Y</td>\n",
       "      <td>84562571</td>\n",
       "      <td>84564906</td>\n",
       "      <td>Gm21394</td>\n",
       "      <td>Y</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.679487</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2335.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1114.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21033</th>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>Y</td>\n",
       "      <td>85528516</td>\n",
       "      <td>85530907</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>Y</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.781250</td>\n",
       "      <td>0.686275</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>2391.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1231.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21034</th>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>89713423</td>\n",
       "      <td>89745531</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>Y</td>\n",
       "      <td>-</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.673913</td>\n",
       "      <td>0.765957</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>32108.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1364.0</td>\n",
       "      <td>878.0</td>\n",
       "      <td>29823.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>203.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21035 rows × 80 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_id Chromosome     Start       End Gene_name contig  \\\n",
       "0      ENSMUSG00000025903.14          1   4807787   4848409    Lypla1      1   \n",
       "1       ENSMUSG00000104217.1          1   4807891   4886769   Gm37988      1   \n",
       "2      ENSMUSG00000033813.15          1   4857813   4897908     Tcea1      1   \n",
       "3      ENSMUSG00000033793.12          1   5070017   5162528   Atp6v1h      1   \n",
       "4      ENSMUSG00000025905.14          1   5588465   5606130     Oprk1      1   \n",
       "...                      ...        ...       ...       ...       ...    ...   \n",
       "21030   ENSMUSG00000096902.7          Y  72220862  72247117   Gm20843      Y   \n",
       "21031   ENSMUSG00000095867.2          Y  79148788  79151121   Gm20917      Y   \n",
       "21032   ENSMUSG00000094660.2          Y  84562571  84564906   Gm21394      Y   \n",
       "21033   ENSMUSG00000095650.2          Y  85528516  85530907   Gm20854      Y   \n",
       "21034   ENSMUSG00000100608.1          Y  89713423  89745531   Gm21996      Y   \n",
       "\n",
       "      strand  mean_mc_before_tss_sense  mean_mc_after_tes_sense  \\\n",
       "0          +                  0.000000                 0.731707   \n",
       "1          +                  0.000689                 0.686047   \n",
       "2          +                  0.001969                 0.800370   \n",
       "3          +                       NaN                 0.847909   \n",
       "4          +                  0.271457                 0.759259   \n",
       "...      ...                       ...                      ...   \n",
       "21030      -                       NaN                 0.727273   \n",
       "21031      -                       NaN                      NaN   \n",
       "21032      -                       NaN                      NaN   \n",
       "21033      -                       NaN                 0.781250   \n",
       "21034      -                  0.833333                 0.673913   \n",
       "\n",
       "       mean_mc_genes_sense  ...  range_length_before_tss_sense  \\\n",
       "0                 0.529368  ...                          200.0   \n",
       "1                 0.472292  ...                          200.0   \n",
       "2                 0.514710  ...                          200.0   \n",
       "3                 0.648667  ...                            NaN   \n",
       "4                 0.680438  ...                          200.0   \n",
       "...                    ...  ...                            ...   \n",
       "21030             0.629630  ...                            NaN   \n",
       "21031             0.782609  ...                            NaN   \n",
       "21032             0.679487  ...                            NaN   \n",
       "21033             0.686275  ...                            NaN   \n",
       "21034             0.765957  ...                          200.0   \n",
       "\n",
       "       range_length_after_tes_sense  range_length_genes_sense  \\\n",
       "0                            1000.0                   40622.0   \n",
       "1                            1000.0                   78878.0   \n",
       "2                            1000.0                   40095.0   \n",
       "3                            1000.0                   92511.0   \n",
       "4                            1000.0                   17665.0   \n",
       "...                             ...                       ...   \n",
       "21030                        1000.0                   26255.0   \n",
       "21031                           NaN                    2333.0   \n",
       "21032                           NaN                    2335.0   \n",
       "21033                        1000.0                    2391.0   \n",
       "21034                        1000.0                   32108.0   \n",
       "\n",
       "       range_length_first_exons_sense  range_length_first_introns_sense  \\\n",
       "0                               159.0                              35.0   \n",
       "1                                90.0                           20602.0   \n",
       "2                               162.0                            9494.0   \n",
       "3                               126.0                           14273.0   \n",
       "4                               170.0                              27.0   \n",
       "...                               ...                               ...   \n",
       "21030                             NaN                               NaN   \n",
       "21031                             NaN                            1111.0   \n",
       "21032                             NaN                            1114.0   \n",
       "21033                             NaN                            1231.0   \n",
       "21034                             NaN                            1364.0   \n",
       "\n",
       "       range_length_exons_sense  range_length_introns_sense  \\\n",
       "0                        2339.0                     38089.0   \n",
       "1                         725.0                     57461.0   \n",
       "2                        2375.0                     28064.0   \n",
       "3                        1523.0                     76589.0   \n",
       "4                        4501.0                     12967.0   \n",
       "...                         ...                         ...   \n",
       "21030                     872.0                     23977.0   \n",
       "21031                       NaN                         NaN   \n",
       "21032                       NaN                         NaN   \n",
       "21033                       NaN                         NaN   \n",
       "21034                     878.0                     29823.0   \n",
       "\n",
       "       range_length_five_prime_utrs_sense  \\\n",
       "0                                    90.0   \n",
       "1                                     NaN   \n",
       "2                                    99.0   \n",
       "3                                     NaN   \n",
       "4                                   183.0   \n",
       "...                                   ...   \n",
       "21030                                 NaN   \n",
       "21031                                 NaN   \n",
       "21032                                 NaN   \n",
       "21033                                 NaN   \n",
       "21034                                 NaN   \n",
       "\n",
       "       range_length_three_prime_utrs_sense  selected_transcript  \n",
       "0                                   1722.0                 True  \n",
       "1                                      NaN                 True  \n",
       "2                                   1540.0                 True  \n",
       "3                                      NaN                 True  \n",
       "4                                   3346.0                 True  \n",
       "...                                    ...                  ...  \n",
       "21030                                197.0                 True  \n",
       "21031                                  NaN                 True  \n",
       "21032                                  NaN                 True  \n",
       "21033                                  NaN                 True  \n",
       "21034                                203.0                 True  \n",
       "\n",
       "[21035 rows x 80 columns]"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features_cleaned = genes.df[[\"Gene_id\", \"Chromosome\", \"Start\", \"End\"]].merge(\n",
    "    df_features_cleaned, on=\"Gene_id\", how=\"inner\",\n",
    "    )\n",
    "df_features_cleaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['range_length_before_tss_sense',\n",
       " 'range_length_after_tes_sense',\n",
       " 'range_length_genes_sense',\n",
       " 'range_length_first_exons_sense',\n",
       " 'range_length_first_introns_sense',\n",
       " 'range_length_exons_sense',\n",
       " 'range_length_introns_sense',\n",
       " 'range_length_five_prime_utrs_sense',\n",
       " 'range_length_three_prime_utrs_sense']"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "length_col = [x for x in df_features_cleaned.columns if x.startswith(\"range_length\")]\n",
    "length_col\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"strand\" in df_features_cleaned.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_features_cleaned.to_pickle(\"../data/CEGXRun1485_CG_fixed-strand.pickle\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "modality",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
