{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mRunning cells with 'modality (Python 3.11.3)' requires the ipykernel package.\n",
      "\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n",
      "\u001b[1;31mCommand: 'conda install -n modality ipykernel --update-deps --force-reinstall'"
     ]
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "import seaborn as sns\n",
    "import xarray as xr\n",
    "\n",
    "# Seaborn customisation\n",
    "sns.set_theme()\n",
    "sns.set_style(\"whitegrid\")\n",
    "biomodal_palette = [\"#003B49\", \"#9CDBD9\", \"#F87C56\", \"#C0DF16\", \"#05868E\"]\n",
    "sns.set_palette(biomodal_palette)\n",
    "\n",
    "from modality.datasets import load_biomodal_dataset\n",
    "from modality.annotation import (\n",
    "    get_genes,\n",
    "    get_transcription_end_region,\n",
    "    get_tss_region,\n",
    "    get_exons,\n",
    "    get_introns,\n",
    "    get_five_prime_utrs,\n",
    "    get_three_prime_utrs,\n",
    "    get_transcripts,\n",
    "    get_cpg_islands,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load ES-E14 evoC dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from modality.contig_dataset import ContigDataset\n",
    "def unify_chr_names(ds: ContigDataset) -> ContigDataset:\n",
    "    # Extract the contig column\n",
    "    selected_column = ds[\"contig\"].values\n",
    "    \n",
    "    # Check if the contigs start with \"chr\" and remove it if they do\n",
    "    unified_contigs = np.array([contig[3:] if contig.startswith(\"chr\") else contig \n",
    "                                for contig in selected_column],dtype='U20')\n",
    "    \n",
    "    # Assign the unified contigs back to the dataset\n",
    "    ds = ds.assign_coords(contig=(\"pos\", unified_contigs))\n",
    "    \n",
    "    # Print unique values of the contig column\n",
    "    print(np.unique(ds[\"contig\"].values))\n",
    "    return ds\n",
    "\n",
    "\n",
    "def unify_slice_names(ds):\n",
    "    # Iterate through the dataset attributes and update slice names\n",
    "    for key in list(ds.attrs.keys()):\n",
    "        if key.startswith('slice_chr'):\n",
    "            new_key = key.replace('slice_chr', 'slice_')\n",
    "            ds.attrs[new_key] = ds.attrs.pop(key)\n",
    "    \n",
    "    return ds\n",
    "\n",
    "def load_data(dataset=\"\"):\n",
    "    #downloads modality.contig_dataset.ContigDataset object\n",
    "    if dataset != \"\":  \n",
    "        ds = ContigDataset.from_zarrz(dataset)\n",
    "        ds = unify_chr_names(ds)\n",
    "        ds = unify_slice_names(ds)\n",
    "    else:\n",
    "        ds = load_biomodal_dataset()\n",
    "        # ds = ContigDataset.from_zarrz(\"../ES-E14.zarrz\")\n",
    "        ds = ds.drop_vars([\"Input DNA Quantity (ng/sample)\", \"tech_replicate_number\"])  \n",
    "    ds = ds.sum(dim=\"sample_id\", keep_attrs=True)\n",
    "    ds = ds.expand_dims(dim=\"sample_id\", axis=1)\n",
    "    ds = ds.assign_coords(sample_id=[\"sample_0\"])\n",
    "    ds.assign_fractions(\n",
    "        numerators=[\"num_modc\", \"num_mc\", \"num_hmc\"],\n",
    "        denominator=\"num_total_c\",\n",
    "        min_coverage=10,\n",
    "        inplace=True,\n",
    "    )\n",
    "    return ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ds = load_data(\"../data/CEGX_Run1485_CG.zarrz\")\n",
    "# ds.to_zarrz(\"../data/CEGX_Run1485_CG_renamed.zarrz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds = ContigDataset.from_zarrz(\"../data/CEGX_Run1485_CG_renamed.zarrz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### subset `ContigDataset` by strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from modality.contig_dataset import set_contig_slices, cast_result\n",
    "\n",
    "def refresh_slices(subset_data):\n",
    "    if subset_data.pos.size == 0:\n",
    "            raise ValueError(\n",
    "                \"Subset is empty. Please check the coverage values and method used.\"\n",
    "            )\n",
    "    else:\n",
    "        slices = [x for x in subset_data.attrs if x.startswith(\"slice\")]\n",
    "        for sl in slices:\n",
    "            subset_data.attrs.pop(sl)\n",
    "        return set_contig_slices(subset_data)\n",
    "\n",
    "def split_strand(data, strand):\n",
    "    \"\"\"\n",
    "    split xarray dataset into positive and negative strands\n",
    "    Parameters\n",
    "        data : xarray.Dataset\n",
    "        strand : str ('+', '-')\n",
    "    Returns\n",
    "        strand_data :xarray.Dataset\n",
    "    \"\"\"\n",
    "    strand_mask = (data['strand'].data == strand).compute()\n",
    "    masked_da = xr.DataArray(strand_mask, dims=data['strand'].dims, coords=data['strand'].coords)\n",
    "    strand_data = data.where(masked_da, drop=True)\n",
    "\n",
    "    # return strand_data\n",
    "    return cast_result(refresh_slices(strand_data.ds))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plus_strand = split_strand(ds, \"+\")\n",
    "# minus_strand = split_strand(ds, \"-\")\n",
    "\n",
    "# plus_strand.to_zarrz(\"../data/plus_strand.zarrz\")\n",
    "# minus_strand.to_zarrz(\"../data/minus_strand.zarrz\")\n",
    "\n",
    "# get original dataset for comparison\n",
    "# ds_orig = load_data()\n",
    "# plus_strand_orig = split_strand(ds_orig, \"+\")\n",
    "# minus_strand_orig = split_strand(ds_orig, \"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "plus_strand = ContigDataset.from_zarrz(\"../data/plus_strand.zarrz\")\n",
    "minus_strand= ContigDataset.from_zarrz(\"../data/minus_strand.zarrz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "slice_1\n",
      "(1477561,)\n",
      "slice_2\n",
      "(1521437,)\n",
      "slice_3\n",
      "(1191216,)\n",
      "slice_4\n",
      "(1332974,)\n",
      "slice_5\n",
      "(1352625,)\n",
      "slice_6\n",
      "(1165998,)\n",
      "slice_7\n",
      "(1224939,)\n",
      "slice_8\n",
      "(1133971,)\n",
      "slice_9\n",
      "(1070707,)\n",
      "slice_10\n",
      "(1106705,)\n",
      "slice_11\n",
      "(1163543,)\n",
      "slice_12\n",
      "(956515,)\n",
      "slice_13\n",
      "(983784,)\n",
      "slice_14\n",
      "(941369,)\n",
      "slice_15\n",
      "(870934,)\n",
      "slice_16\n",
      "(755291,)\n",
      "slice_17\n",
      "(854457,)\n",
      "slice_18\n",
      "(721541,)\n",
      "slice_19\n",
      "(558056,)\n",
      "slice_X\n",
      "(969469,)\n",
      "slice_Y\n",
      "(514458,)\n",
      "slice_M\n",
      "(287,)\n"
     ]
    }
   ],
   "source": [
    "for contig in [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,'X','Y','M']:\n",
    "    slice1 = plus_strand.slices[f\"slice_{contig}\"]\n",
    "    print(f\"slice_{contig}\")\n",
    "    print(plus_strand.ref_position[slice1].data.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Loading mm10 (GRCm38) reference genome from [GENCODE](https://www.gencodegenes.org/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-13 10:15:35 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------+------------+------------+-----------+-----------+-------+\n",
      "|   Chromosome | Source     | Type       |     Start |       End | +12   |\n",
      "|   (category) | (object)   | (object)   |   (int64) |   (int64) | ...   |\n",
      "|--------------+------------+------------+-----------+-----------+-------|\n",
      "|            1 | HAVANA     | gene       |   4807787 |   4848409 | ...   |\n",
      "|            1 | HAVANA     | gene       |   4807891 |   4886769 | ...   |\n",
      "|            1 | HAVANA     | gene       |   4857813 |   4897908 | ...   |\n",
      "|            1 | HAVANA     | gene       |   5070017 |   5162528 | ...   |\n",
      "|            1 | HAVANA     | gene       |   5588465 |   5606130 | ...   |\n",
      "+--------------+------------+------------+-----------+-----------+-------+\n",
      "Stranded PyRanges object has 5 rows and 17 columns from 1 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
      "12 hidden columns: Score, Strand, Phase, Id, Gene_id, Gene_type, Gene_name, Level, ... (+ 4 more.)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-13 10:15:48 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------------------------------------------------\n",
      " all genes: 21673\n",
      " unique gene id: 21673\n"
     ]
    }
   ],
   "source": [
    "# Get the gene annotations\n",
    "gene_filter = {\n",
    "    \"gene_type\": \"protein_coding\",\n",
    "    \"source\": \"HAVANA\",\n",
    "}\n",
    "\n",
    "genes = get_genes(\n",
    "    reference=\"mm10\",\n",
    "    as_pyranges=True,\n",
    "    filterby=gene_filter,\n",
    ")\n",
    "print(genes.head(5))\n",
    "\n",
    "# Get transcripts for mm10\n",
    "transcripts = get_transcripts(\n",
    "        reference=\"mm10\",\n",
    "        contig=None,\n",
    "        start=None,\n",
    "        end=None,\n",
    "        as_pyranges=False,\n",
    "    )\n",
    "\n",
    "print('-'*100)\n",
    "\n",
    "print(f\" all genes: {len(genes)}\")\n",
    "print(f\" unique gene id: {len(genes.df['Id'])}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Select primiary transcript for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003.15</td>\n",
       "      <td>ENSMUST00000000003.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028.15</td>\n",
       "      <td>ENSMUST00000000028.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000037.17</td>\n",
       "      <td>ENSMUST00000101113.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000049.11</td>\n",
       "      <td>ENSMUST00000000049.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene_id          transcript_id\n",
       "0   ENSMUSG00000000001.4   ENSMUST00000000001.4\n",
       "1  ENSMUSG00000000003.15  ENSMUST00000000003.13\n",
       "2  ENSMUSG00000000028.15  ENSMUST00000000028.13\n",
       "3  ENSMUSG00000000037.17   ENSMUST00000101113.8\n",
       "4  ENSMUSG00000000049.11   ENSMUST00000000049.5"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def select_transcript_based_on_tag(df):\n",
    "    # for each transcript in df, select the one with the highest priority tag\n",
    "    # priorities are:\n",
    "        # 1. 'basic,appris_principal_1,CCDS'\n",
    "        # 2. 'basic,appris_principal_1'\n",
    "        # 3. 'basic,CCDS'\n",
    "        # 4. 'basic'\n",
    "    # but with 'exp_conf' (experimentally confirmed) tag, the priority is higher.\n",
    "    \n",
    "    priorties = {\n",
    "        'basic,appris_principal_1,exp_conf,CCDS': 1,\n",
    "        'basic,appris_principal_1,CCDS': 1,\n",
    "        'basic,appris_principal_1,exp_conf': 3,\n",
    "        'basic,appris_principal_1': 4,\n",
    "        'basic,exp_conf,CCDS': 5,\n",
    "        'basic,CCDS': 6,\n",
    "        'basic,exp_conf': 7,\n",
    "        'basic': 8\n",
    "    }\n",
    "\n",
    "    # sort the dataframe by the priority of the tags\n",
    "    df['tag_priority'] = df.tag.map(priorties)\n",
    "\n",
    "    df = df.sort_values(by='tag_priority')\n",
    "\n",
    "    # drop duplicates, keeping the first one\n",
    "    df = df.drop_duplicates(subset='gene_id', keep='first')\n",
    "\n",
    "    return df[[\"gene_id\", \"transcript_id\"]]\n",
    "\n",
    "selected_transcripts = transcripts.groupby('gene_id').apply(\n",
    "    select_transcript_based_on_tag\n",
    "    ).reset_index(drop=True)\n",
    "selected_transcripts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 21541 genes with unique transcript ids\n"
     ]
    }
   ],
   "source": [
    "# Create dictionary for gene_id to transcript_id mapping\n",
    "gene_to_transcription = selected_transcripts.set_index('gene_id')['transcript_id'].to_dict()\n",
    "print(f\"There are {len(gene_to_transcription)} genes with unique transcript ids\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "review / drop genes with no transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all genes: 21673\n",
      "all genes with annotated transcripts: 21541\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807787</td>\n",
       "      <td>4848409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807891</td>\n",
       "      <td>4886769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5611216</td>\n",
       "      <td>OTTMUSG00000050100.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUST00000155020.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4857813</td>\n",
       "      <td>4897908</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5070017</td>\n",
       "      <td>5162528</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Atp6v1h</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1914864</td>\n",
       "      <td>OTTMUSG00000050145.9</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUST00000192847.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5588465</td>\n",
       "      <td>5606130</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21536</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>78835720</td>\n",
       "      <td>78838055</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20806</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434162</td>\n",
       "      <td>OTTMUSG00000046577.2</td>\n",
       "      <td></td>\n",
       "      <td>21668</td>\n",
       "      <td>ENSMUST00000190349.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21537</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>79148788</td>\n",
       "      <td>79151121</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20917</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434273</td>\n",
       "      <td>OTTMUSG00000046619.2</td>\n",
       "      <td></td>\n",
       "      <td>21669</td>\n",
       "      <td>ENSMUST00000188706.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21538</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>84562571</td>\n",
       "      <td>84564906</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21394</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434749</td>\n",
       "      <td>OTTMUSG00000045415.1</td>\n",
       "      <td></td>\n",
       "      <td>21670</td>\n",
       "      <td>ENSMUST00000189463.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21539</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>85528516</td>\n",
       "      <td>85530907</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434210</td>\n",
       "      <td>OTTMUSG00000042966.1</td>\n",
       "      <td></td>\n",
       "      <td>21671</td>\n",
       "      <td>ENSMUST00000181549.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21540</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89713423</td>\n",
       "      <td>89745531</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5440224</td>\n",
       "      <td>OTTMUSG00000047316.1</td>\n",
       "      <td></td>\n",
       "      <td>21672</td>\n",
       "      <td>ENSMUST00000188269.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21541 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +13   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 4807787   | 4848409   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4807891   | 4886769   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4857813   | 4897908   | ...   |\n",
       "| 1            | HAVANA     | gene       | 5070017   | 5162528   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 79148788  | 79151121  | ...   |\n",
       "| Y            | HAVANA     | gene       | 84562571  | 84564906  | ...   |\n",
       "| Y            | HAVANA     | gene       | 85528516  | 85530907  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89713423  | 89745531  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 21,541 rows and 18 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "13 hidden columns: Score, Strand, Phase, Id, Gene_id, Gene_type, Gene_name, Level, ... (+ 5 more.)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Function to map transcription_id\n",
    "def map_transcription_id(gene_id, gene_to_transcription):\n",
    "    return gene_to_transcription.get(gene_id, None)\n",
    "\n",
    "# Map transcription_id \n",
    "genes_df = genes.df\n",
    "genes_df['Transcript_id'] = genes_df['Gene_id'].map(lambda gene_id: map_transcription_id(gene_id, gene_to_transcription))\n",
    "genes_clean = pr.PyRanges(genes_df)\n",
    "\n",
    "# Find and Handle missing gene_ids\n",
    "missing_gene_ids = genes_clean.df[genes_clean.df['Transcript_id'].isna()]['Gene_id'].tolist()\n",
    "missing_df = genes_clean.df[genes_clean.df['Gene_id'].isin(missing_gene_ids)]\n",
    "genes_clean = genes_clean[~genes_clean.df['Transcript_id'].isna()]\n",
    "print(f\"all genes: {len(genes)}\")\n",
    "print(f\"all genes with annotated transcripts: {len(genes_clean)}\")\n",
    "genes_clean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tags for 132 missing genes:\n",
      "Tag\n",
      "                                                 87\n",
      "overlapping_locus                                13\n",
      "reference_genome_error                           10\n",
      "fragmented_locus                                  9\n",
      "ncRNA_host,fragmented_locus                       2\n",
      "ncRNA_host                                        2\n",
      "overlapping_locus,reference_genome_error          2\n",
      "ncRNA_host,reference_genome_error                 2\n",
      "ncRNA_host,overlapping_locus                      2\n",
      "fragmented_locus,reference_genome_error           1\n",
      "fragmented_locus,overlapping_locus                1\n",
      "ncRNA_host,fragmented_locus,overlapping_locus     1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(f\"Tags for {len(missing_df)} missing genes:\")\n",
    "print(missing_df['Tag'].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract Regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### requires stitching"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 200bp before Transcription Start Site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default arguments for get_tss_region and get_transcription_end_region\n",
    "default_args = {\n",
    "    \"contig\": None,\n",
    "    \"start\": None,\n",
    "    \"end\": None,\n",
    "    \"reference\": \"mm10\",\n",
    "    \"as_pyranges\": True,\n",
    "    \"protein_coding\": True,\n",
    "    \"filterby\": None,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-13 10:16:40 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "before_tss regions with transcripts: 21541\n",
      "====================================================================================================\n"
     ]
    }
   ],
   "source": [
    "before_tss = get_tss_region(\n",
    "    start_offset=-200,\n",
    "    span=200,\n",
    "    **default_args,\n",
    ")\n",
    "\n",
    "# Map transcription_id \n",
    "before_tss_df = before_tss.df\n",
    "before_tss_df['Transcript_id'] = before_tss_df['Gene_id'].map(lambda gene_id: map_transcription_id(gene_id, gene_to_transcription))\n",
    "before_tss = pr.PyRanges(before_tss_df)\n",
    "before_tss = before_tss[~before_tss.df['Transcript_id'].isna()]\n",
    "\n",
    "print(f\"before_tss regions with transcripts: {len(before_tss)}\")\n",
    "print(\"=\"*100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807587</td>\n",
       "      <td>4807787</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807691</td>\n",
       "      <td>4807891</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5611216</td>\n",
       "      <td>OTTMUSG00000050100.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUST00000155020.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4857613</td>\n",
       "      <td>4857813</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5069817</td>\n",
       "      <td>5070017</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Atp6v1h</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1914864</td>\n",
       "      <td>OTTMUSG00000050145.9</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUST00000192847.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5588265</td>\n",
       "      <td>5588465</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10735</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>87116620</td>\n",
       "      <td>87116820</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21477</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434832</td>\n",
       "      <td>OTTMUSG00000047083.1</td>\n",
       "      <td></td>\n",
       "      <td>21601</td>\n",
       "      <td>ENSMUST00000189543.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10736</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>87550765</td>\n",
       "      <td>87550965</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20906</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td></td>\n",
       "      <td>21602</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10737</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>88053114</td>\n",
       "      <td>88053314</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm28102</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td></td>\n",
       "      <td>21603</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10738</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89052605</td>\n",
       "      <td>89052805</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102045.1</td>\n",
       "      <td>ENSMUSG00000102045.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21294</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434649</td>\n",
       "      <td>OTTMUSG00000047309.1</td>\n",
       "      <td></td>\n",
       "      <td>21604</td>\n",
       "      <td>ENSMUST00000186443.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10739</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90401047</td>\n",
       "      <td>90401247</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000096178.7</td>\n",
       "      <td>ENSMUSG00000096178.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20837</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434193</td>\n",
       "      <td>OTTMUSG00000047352.1</td>\n",
       "      <td></td>\n",
       "      <td>21605</td>\n",
       "      <td>ENSMUST00000190558.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10740 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +12   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 4807587   | 4807787   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4807691   | 4807891   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4857613   | 4857813   | ...   |\n",
       "| 1            | HAVANA     | gene       | 5069817   | 5070017   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 87550765  | 87550965  | ...   |\n",
       "| Y            | HAVANA     | gene       | 88053114  | 88053314  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89052605  | 89052805  | ...   |\n",
       "| Y            | HAVANA     | gene       | 90401047  | 90401247  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Unstranded PyRanges object has 10,740 rows and 17 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "12 hidden columns: Score, Phase, Id, Gene_id, Gene_type, Gene_name, Level, Mgi_id, ... (+ 4 more.)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#split strands\n",
    "before_tss_plus = pr.PyRanges(before_tss.df[before_tss.df[\"Strand\"]==\"+\"])\n",
    "before_tss_minus = pr.PyRanges(before_tss.df[before_tss.df[\"Strand\"]==\"-\"])\n",
    "\n",
    "before_tss_plus = before_tss_plus.unstrand()\n",
    "before_tss_minus = before_tss_minus.unstrand()\n",
    "before_tss_plus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3671497</td>\n",
       "      <td>3671697</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td></td>\n",
       "      <td>611</td>\n",
       "      <td>ENSMUST00000070533.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4409240</td>\n",
       "      <td>4409440</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025900.13</td>\n",
       "      <td>ENSMUSG00000025900.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>OTTMUSG00000049985.3</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>612</td>\n",
       "      <td>ENSMUST00000027032.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4497353</td>\n",
       "      <td>4497553</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025902.13</td>\n",
       "      <td>ENSMUSG00000025902.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:107543</td>\n",
       "      <td>OTTMUSG00000050014.7</td>\n",
       "      <td></td>\n",
       "      <td>613</td>\n",
       "      <td>ENSMUST00000027035.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4785738</td>\n",
       "      <td>4785938</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033845.13</td>\n",
       "      <td>ENSMUSG00000033845.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1351639</td>\n",
       "      <td>OTTMUSG00000029329.3</td>\n",
       "      <td></td>\n",
       "      <td>614</td>\n",
       "      <td>ENSMUST00000156816.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5070284</td>\n",
       "      <td>5070484</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000002459.17</td>\n",
       "      <td>ENSMUSG00000002459.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Rgs20</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1929866</td>\n",
       "      <td>OTTMUSG00000029338.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>615</td>\n",
       "      <td>ENSMUST00000002533.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10796</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>78838055</td>\n",
       "      <td>78838255</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20806</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434162</td>\n",
       "      <td>OTTMUSG00000046577.2</td>\n",
       "      <td></td>\n",
       "      <td>21668</td>\n",
       "      <td>ENSMUST00000190349.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10797</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>79151121</td>\n",
       "      <td>79151321</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20917</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434273</td>\n",
       "      <td>OTTMUSG00000046619.2</td>\n",
       "      <td></td>\n",
       "      <td>21669</td>\n",
       "      <td>ENSMUST00000188706.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10798</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>84564906</td>\n",
       "      <td>84565106</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21394</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434749</td>\n",
       "      <td>OTTMUSG00000045415.1</td>\n",
       "      <td></td>\n",
       "      <td>21670</td>\n",
       "      <td>ENSMUST00000189463.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10799</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>85530907</td>\n",
       "      <td>85531107</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434210</td>\n",
       "      <td>OTTMUSG00000042966.1</td>\n",
       "      <td></td>\n",
       "      <td>21671</td>\n",
       "      <td>ENSMUST00000181549.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10800</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89745531</td>\n",
       "      <td>89745731</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5440224</td>\n",
       "      <td>OTTMUSG00000047316.1</td>\n",
       "      <td></td>\n",
       "      <td>21672</td>\n",
       "      <td>ENSMUST00000188269.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10801 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +12   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 3671497   | 3671697   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4409240   | 4409440   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4497353   | 4497553   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4785738   | 4785938   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 79151121  | 79151321  | ...   |\n",
       "| Y            | HAVANA     | gene       | 84564906  | 84565106  | ...   |\n",
       "| Y            | HAVANA     | gene       | 85530907  | 85531107  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89745531  | 89745731  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Unstranded PyRanges object has 10,801 rows and 17 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "12 hidden columns: Score, Phase, Id, Gene_id, Gene_type, Gene_name, Level, Mgi_id, ... (+ 4 more.)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "before_tss_minus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1000bp after Transcription End Site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-13 10:16:53 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after_tes regions with transcripts: 21541\n",
      "====================================================================================================\n"
     ]
    }
   ],
   "source": [
    "after_tes = get_transcription_end_region(\n",
    "    start_offset=0,\n",
    "    span=1000,\n",
    "    **default_args,\n",
    ")\n",
    "\n",
    "# Map transcription_id \n",
    "after_tes_df = after_tes.df\n",
    "after_tes_df['Transcript_id'] = after_tes_df['Gene_id'].map(lambda gene_id: map_transcription_id(gene_id, gene_to_transcription))\n",
    "after_tes = pr.PyRanges(before_tss_df)\n",
    "after_tes = after_tes[~after_tes.df['Transcript_id'].isna()]\n",
    "\n",
    "print(f\"after_tes regions with transcripts: {len(after_tes)}\")\n",
    "print(\"=\"*100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807587</td>\n",
       "      <td>4807787</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807691</td>\n",
       "      <td>4807891</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5611216</td>\n",
       "      <td>OTTMUSG00000050100.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUST00000155020.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4857613</td>\n",
       "      <td>4857813</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5069817</td>\n",
       "      <td>5070017</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Atp6v1h</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1914864</td>\n",
       "      <td>OTTMUSG00000050145.9</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUST00000192847.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5588265</td>\n",
       "      <td>5588465</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10735</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>87116620</td>\n",
       "      <td>87116820</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21477</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434832</td>\n",
       "      <td>OTTMUSG00000047083.1</td>\n",
       "      <td></td>\n",
       "      <td>21601</td>\n",
       "      <td>ENSMUST00000189543.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10736</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>87550765</td>\n",
       "      <td>87550965</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20906</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td></td>\n",
       "      <td>21602</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10737</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>88053114</td>\n",
       "      <td>88053314</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm28102</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td></td>\n",
       "      <td>21603</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10738</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89052605</td>\n",
       "      <td>89052805</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102045.1</td>\n",
       "      <td>ENSMUSG00000102045.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21294</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434649</td>\n",
       "      <td>OTTMUSG00000047309.1</td>\n",
       "      <td></td>\n",
       "      <td>21604</td>\n",
       "      <td>ENSMUST00000186443.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10739</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90401047</td>\n",
       "      <td>90401247</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000096178.7</td>\n",
       "      <td>ENSMUSG00000096178.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20837</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434193</td>\n",
       "      <td>OTTMUSG00000047352.1</td>\n",
       "      <td></td>\n",
       "      <td>21605</td>\n",
       "      <td>ENSMUST00000190558.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10740 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +12   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 4807587   | 4807787   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4807691   | 4807891   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4857613   | 4857813   | ...   |\n",
       "| 1            | HAVANA     | gene       | 5069817   | 5070017   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 87550765  | 87550965  | ...   |\n",
       "| Y            | HAVANA     | gene       | 88053114  | 88053314  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89052605  | 89052805  | ...   |\n",
       "| Y            | HAVANA     | gene       | 90401047  | 90401247  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Unstranded PyRanges object has 10,740 rows and 17 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "12 hidden columns: Score, Phase, Id, Gene_id, Gene_type, Gene_name, Level, Mgi_id, ... (+ 4 more.)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#split strands\n",
    "after_tes_plus = pr.PyRanges(after_tes.df[after_tes.df[\"Strand\"]==\"+\"])\n",
    "after_tes_minus = pr.PyRanges(after_tes.df[after_tes.df[\"Strand\"]==\"-\"])\n",
    "\n",
    "after_tes_plus = after_tes_plus.unstrand()\n",
    "after_tes_minus = after_tes_minus.unstrand()\n",
    "after_tes_plus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3671497</td>\n",
       "      <td>3671697</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td></td>\n",
       "      <td>611</td>\n",
       "      <td>ENSMUST00000070533.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4409240</td>\n",
       "      <td>4409440</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025900.13</td>\n",
       "      <td>ENSMUSG00000025900.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>OTTMUSG00000049985.3</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>612</td>\n",
       "      <td>ENSMUST00000027032.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4497353</td>\n",
       "      <td>4497553</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025902.13</td>\n",
       "      <td>ENSMUSG00000025902.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:107543</td>\n",
       "      <td>OTTMUSG00000050014.7</td>\n",
       "      <td></td>\n",
       "      <td>613</td>\n",
       "      <td>ENSMUST00000027035.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4785738</td>\n",
       "      <td>4785938</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033845.13</td>\n",
       "      <td>ENSMUSG00000033845.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1351639</td>\n",
       "      <td>OTTMUSG00000029329.3</td>\n",
       "      <td></td>\n",
       "      <td>614</td>\n",
       "      <td>ENSMUST00000156816.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5070284</td>\n",
       "      <td>5070484</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000002459.17</td>\n",
       "      <td>ENSMUSG00000002459.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Rgs20</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1929866</td>\n",
       "      <td>OTTMUSG00000029338.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>615</td>\n",
       "      <td>ENSMUST00000002533.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10796</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>78838055</td>\n",
       "      <td>78838255</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20806</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434162</td>\n",
       "      <td>OTTMUSG00000046577.2</td>\n",
       "      <td></td>\n",
       "      <td>21668</td>\n",
       "      <td>ENSMUST00000190349.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10797</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>79151121</td>\n",
       "      <td>79151321</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20917</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434273</td>\n",
       "      <td>OTTMUSG00000046619.2</td>\n",
       "      <td></td>\n",
       "      <td>21669</td>\n",
       "      <td>ENSMUST00000188706.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10798</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>84564906</td>\n",
       "      <td>84565106</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21394</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434749</td>\n",
       "      <td>OTTMUSG00000045415.1</td>\n",
       "      <td></td>\n",
       "      <td>21670</td>\n",
       "      <td>ENSMUST00000189463.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10799</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>85530907</td>\n",
       "      <td>85531107</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434210</td>\n",
       "      <td>OTTMUSG00000042966.1</td>\n",
       "      <td></td>\n",
       "      <td>21671</td>\n",
       "      <td>ENSMUST00000181549.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10800</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89745531</td>\n",
       "      <td>89745731</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5440224</td>\n",
       "      <td>OTTMUSG00000047316.1</td>\n",
       "      <td></td>\n",
       "      <td>21672</td>\n",
       "      <td>ENSMUST00000188269.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10801 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +12   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 3671497   | 3671697   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4409240   | 4409440   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4497353   | 4497553   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4785738   | 4785938   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 79151121  | 79151321  | ...   |\n",
       "| Y            | HAVANA     | gene       | 84564906  | 84565106  | ...   |\n",
       "| Y            | HAVANA     | gene       | 85530907  | 85531107  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89745531  | 89745731  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Unstranded PyRanges object has 10,801 rows and 17 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "12 hidden columns: Score, Phase, Id, Gene_id, Gene_type, Gene_name, Level, Mgi_id, ... (+ 4 more.)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "after_tes_minus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract 3`UTRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-13 10:17:06 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "strand shape: + (8075, 27), - (8134, 27)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>...</th>\n",
       "      <th>Level</th>\n",
       "      <th>Protein_id</th>\n",
       "      <th>Transcript_support_level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ccdsid</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Havana_transcript</th>\n",
       "      <th>Ont</th>\n",
       "      <th>Ranges_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>4845016</td>\n",
       "      <td>4846738</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000027036.10</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000027036.4</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14806.1</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>OTTMUST00000051162.1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>4896364</td>\n",
       "      <td>4897904</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000081551.13</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000080266.7</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>non_canonical_U12,basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS35505.1</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>OTTMUST00000111602.1</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>5602784</td>\n",
       "      <td>5606130</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000125105.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14809.1</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td>OTTMUST00000088255.1</td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>6274275</td>\n",
       "      <td>6276647</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000027040.12</td>\n",
       "      <td>ENSMUST00000027040.12</td>\n",
       "      <td>ENSMUSG00000025907.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000027040.6</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1341850</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS35507.1</td>\n",
       "      <td>OTTMUSG00000033467.12</td>\n",
       "      <td>OTTMUST00000084091.5</td>\n",
       "      <td></td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>6391104</td>\n",
       "      <td>6391115</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000133144.3</td>\n",
       "      <td>ENSMUST00000133144.3</td>\n",
       "      <td>ENSMUSG00000087247.3</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000137420.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:3645495</td>\n",
       "      <td>not_organism_supported,basic,appris_principal_...</td>\n",
       "      <td>CCDS56620.1</td>\n",
       "      <td>OTTMUSG00000050239.2</td>\n",
       "      <td>OTTMUST00000127625.2</td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8070</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>84772706</td>\n",
       "      <td>84772901</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000186110.1</td>\n",
       "      <td>ENSMUST00000186110.1</td>\n",
       "      <td>ENSMUSG00000099840.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140850.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434764</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000046908.1</td>\n",
       "      <td>OTTMUST00000121611.1</td>\n",
       "      <td></td>\n",
       "      <td>27079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8071</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>86087758</td>\n",
       "      <td>86087951</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000188754.1</td>\n",
       "      <td>ENSMUST00000188754.1</td>\n",
       "      <td>ENSMUSG00000100240.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000139858.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434176</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047031.1</td>\n",
       "      <td>OTTMUST00000121805.1</td>\n",
       "      <td></td>\n",
       "      <td>27080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8072</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>87142720</td>\n",
       "      <td>87142916</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000189543.6</td>\n",
       "      <td>ENSMUST00000189543.6</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140238.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434832</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047083.1</td>\n",
       "      <td>OTTMUST00000121876.1</td>\n",
       "      <td></td>\n",
       "      <td>27081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8073</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>87575495</td>\n",
       "      <td>87575690</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140305.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td>OTTMUST00000121968.1</td>\n",
       "      <td></td>\n",
       "      <td>27082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8074</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>88079297</td>\n",
       "      <td>88079493</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140242.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td>OTTMUST00000121980.1</td>\n",
       "      <td></td>\n",
       "      <td>27083</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8075 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------------+-----------+-------+\n",
       "| Chromosome   | Source     | Type            | Start     | +23   |\n",
       "| (category)   | (object)   | (object)        | (int64)   | ...   |\n",
       "|--------------+------------+-----------------+-----------+-------|\n",
       "| 1            | HAVANA     | three_prime_UTR | 4845016   | ...   |\n",
       "| 1            | HAVANA     | three_prime_UTR | 4896364   | ...   |\n",
       "| 1            | HAVANA     | three_prime_UTR | 5602784   | ...   |\n",
       "| 1            | HAVANA     | three_prime_UTR | 6274275   | ...   |\n",
       "| ...          | ...        | ...             | ...       | ...   |\n",
       "| Y            | HAVANA     | three_prime_UTR | 86087758  | ...   |\n",
       "| Y            | HAVANA     | three_prime_UTR | 87142720  | ...   |\n",
       "| Y            | HAVANA     | three_prime_UTR | 87575495  | ...   |\n",
       "| Y            | HAVANA     | three_prime_UTR | 88079297  | ...   |\n",
       "+--------------+------------+-----------------+-----------+-------+\n",
       "Unstranded PyRanges object has 8,075 rows and 27 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "23 hidden columns: End, Score, Phase, Id, Parent, Gene_id, Transcript_id, ... (+ 16 more.)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract with modality.get_three_prime_utrs\n",
    "three_prime_utrs = get_three_prime_utrs(reference=\"mm10\")\n",
    "three_prime_utrs = three_prime_utrs[\n",
    "    three_prime_utrs.Transcript_id.isin(selected_transcripts.transcript_id)\n",
    "    ]\n",
    "\n",
    "#split strands\n",
    "three_prime_utrs_plus = pr.PyRanges(three_prime_utrs.df[three_prime_utrs.df[\"Strand\"]==\"+\"])\n",
    "three_prime_utrs_plus = three_prime_utrs_plus.unstrand()\n",
    "\n",
    "three_prime_utrs_minus = pr.PyRanges(three_prime_utrs.df[three_prime_utrs.df[\"Strand\"]==\"-\"])\n",
    "three_prime_utrs_minus = three_prime_utrs_minus.unstrand()\n",
    "\n",
    "print(f\"strand shape: + {three_prime_utrs_plus.df.shape}, - {three_prime_utrs_minus.df.shape}\")\n",
    "three_prime_utrs_plus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract 5`UTRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-13 10:17:21 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "strand shape: + (11449, 27), - (11369, 27)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>...</th>\n",
       "      <th>Level</th>\n",
       "      <th>Protein_id</th>\n",
       "      <th>Transcript_support_level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ccdsid</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Havana_transcript</th>\n",
       "      <th>Ont</th>\n",
       "      <th>Ranges_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>4807822</td>\n",
       "      <td>4807912</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000027036.10</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000027036.4</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14806.1</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>OTTMUST00000051162.1</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>4857813</td>\n",
       "      <td>4857912</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000081551.13</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000080266.7</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>non_canonical_U12,basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS35505.1</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>OTTMUST00000111602.1</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>5588492</td>\n",
       "      <td>5588662</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000125105.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14809.1</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td>OTTMUST00000088255.1</td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>5589034</td>\n",
       "      <td>5589047</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000125105.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14809.1</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td>OTTMUST00000088255.1</td>\n",
       "      <td></td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>6214644</td>\n",
       "      <td>6214956</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000027040.12</td>\n",
       "      <td>ENSMUST00000027040.12</td>\n",
       "      <td>ENSMUSG00000025907.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000027040.6</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1341850</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS35507.1</td>\n",
       "      <td>OTTMUSG00000033467.12</td>\n",
       "      <td>OTTMUST00000084091.5</td>\n",
       "      <td></td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11444</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>87118231</td>\n",
       "      <td>87118244</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000189543.6</td>\n",
       "      <td>ENSMUST00000189543.6</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140238.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434832</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047083.1</td>\n",
       "      <td>OTTMUST00000121876.1</td>\n",
       "      <td></td>\n",
       "      <td>39793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11445</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>87550965</td>\n",
       "      <td>87551008</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140305.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td>OTTMUST00000121968.1</td>\n",
       "      <td></td>\n",
       "      <td>39794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11446</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>87552378</td>\n",
       "      <td>87552391</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140305.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td>OTTMUST00000121968.1</td>\n",
       "      <td></td>\n",
       "      <td>39795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11447</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>88053314</td>\n",
       "      <td>88053357</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140242.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td>OTTMUST00000121980.1</td>\n",
       "      <td></td>\n",
       "      <td>39796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11448</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>88054723</td>\n",
       "      <td>88054736</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140242.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td>OTTMUST00000121980.1</td>\n",
       "      <td></td>\n",
       "      <td>39797</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11449 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+----------------+-----------+-------+\n",
       "| Chromosome   | Source     | Type           | Start     | +23   |\n",
       "| (category)   | (object)   | (object)       | (int64)   | ...   |\n",
       "|--------------+------------+----------------+-----------+-------|\n",
       "| 1            | HAVANA     | five_prime_UTR | 4807822   | ...   |\n",
       "| 1            | HAVANA     | five_prime_UTR | 4857813   | ...   |\n",
       "| 1            | HAVANA     | five_prime_UTR | 5588492   | ...   |\n",
       "| 1            | HAVANA     | five_prime_UTR | 5589034   | ...   |\n",
       "| ...          | ...        | ...            | ...       | ...   |\n",
       "| Y            | HAVANA     | five_prime_UTR | 87550965  | ...   |\n",
       "| Y            | HAVANA     | five_prime_UTR | 87552378  | ...   |\n",
       "| Y            | HAVANA     | five_prime_UTR | 88053314  | ...   |\n",
       "| Y            | HAVANA     | five_prime_UTR | 88054723  | ...   |\n",
       "+--------------+------------+----------------+-----------+-------+\n",
       "Unstranded PyRanges object has 11,449 rows and 27 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "23 hidden columns: End, Score, Phase, Id, Parent, Gene_id, Transcript_id, ... (+ 16 more.)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract with modality.get_five_prime_utrs\n",
    "five_prime_utrs = get_five_prime_utrs(reference=\"mm10\")\n",
    "five_prime_utrs = five_prime_utrs[\n",
    "    five_prime_utrs.Transcript_id.isin(selected_transcripts.transcript_id)\n",
    "    ]\n",
    "\n",
    "#split strands\n",
    "five_prime_utrs_plus = pr.PyRanges(five_prime_utrs.df[five_prime_utrs.df[\"Strand\"]==\"+\"])\n",
    "five_prime_utrs_plus = five_prime_utrs_plus.unstrand()\n",
    "\n",
    "five_prime_utrs_minus = pr.PyRanges(five_prime_utrs.df[five_prime_utrs.df[\"Strand\"]==\"-\"])\n",
    "five_prime_utrs_minus = five_prime_utrs_minus.unstrand()\n",
    "print(f\"strand shape: + {five_prime_utrs_plus.df.shape}, - {five_prime_utrs_minus.df.shape}\")\n",
    "five_prime_utrs_plus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract exons and introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-13 10:17:51 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    }
   ],
   "source": [
    "# Exons:\n",
    "exons = get_exons(reference=\"mm10\")\n",
    "exons = exons[exons.Transcript_id.isin(selected_transcripts.transcript_id)]\n",
    "\n",
    "# exons split strands\n",
    "exons_plus = pr.PyRanges(exons.df[exons.df[\"Strand\"]==\"+\"])\n",
    "first_exons_plus = exons_plus[exons_plus.Exon_number == \"1\"].unstrand()\n",
    "exons_plus = exons_plus[exons_plus.Exon_number.astype(\"int\") > 1].unstrand()\n",
    "\n",
    "\n",
    "exons_minus = pr.PyRanges(exons.df[exons.df[\"Strand\"]==\"-\"])\n",
    "first_exons_minus = exons_minus[exons_minus.Exon_number == \"1\"].unstrand()\n",
    "exons_minus = exons_minus[exons_minus.Exon_number.astype(\"int\") > 1].unstrand()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-13 10:18:38 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n",
      "2024-09-13 10:19:02 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n",
      "2024-09-13 10:19:36 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n",
      "2024-09-13 10:19:51 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 8 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "#Introns:\n",
    "introns = get_introns(\n",
    "    reference=\"mm10\",\n",
    "    transcripts=selected_transcripts.transcript_id.values,\n",
    "    nb_workers=8,\n",
    ")\n",
    "\n",
    "# introns split strands\n",
    "introns_plus = pr.PyRanges(introns.df[introns.df[\"Strand\"]==\"+\"])\n",
    "first_introns_plus = introns_plus[introns_plus.Intron_number == \"1\"].unstrand()\n",
    "introns_plus = introns_plus[introns_plus.Intron_number.astype(\"int\") > 1].unstrand()\n",
    "\n",
    "\n",
    "introns_minus = pr.PyRanges(introns.df[introns.df[\"Strand\"]==\"-\"])\n",
    "first_introns_minus = introns_minus[introns_minus.Intron_number == \"1\"].unstrand()\n",
    "introns_minus = introns_minus[introns_minus.Intron_number.astype(\"int\") > 1].unstrand()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exons strand shape: + (90807, 27), - (90167, 27)\n",
      "Introns strand shape: + (85982, 20), - (85480, 20)\n",
      "First Exons strand shape: + (10740, 27), - (10801, 27)\n",
      "First Introns strand shape: + (10186, 20), - (10212, 20)\n"
     ]
    }
   ],
   "source": [
    "print(f\"Exons strand shape: + {exons_plus.df.shape}, - {exons_minus.df.shape}\")\n",
    "print(f\"Introns strand shape: + {introns_plus.df.shape}, - {introns_minus.df.shape}\")\n",
    "\n",
    "print(f\"First Exons strand shape: + {first_exons_plus.df.shape}, - {first_exons_minus.df.shape}\")\n",
    "print(f\"First Introns strand shape: + {first_introns_plus.df.shape}, - {first_introns_minus.df.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Source', 'Type', 'Start', 'End', 'Score', 'Strand',\n",
      "       'Phase', 'Id', 'Parent', 'Gene_id', 'Transcript_id', 'Gene_type',\n",
      "       'Gene_name', 'Transcript_type', 'Transcript_name', 'Exon_number',\n",
      "       'Exon_id', 'Level', 'Transcript_support_level', 'Mgi_id', 'Tag',\n",
      "       'Havana_gene', 'Havana_transcript', 'Protein_id', 'Ccdsid', 'Ont',\n",
      "       'Ranges_ID'],\n",
      "      dtype='object')\n",
      "Index(['Transcript_id', 'Level_1', 'Chromosome', 'Source', 'Type', 'Start',\n",
      "       'End', 'Score', 'Strand', 'Phase', 'Id', 'Gene_id', 'Gene_type',\n",
      "       'Gene_name', 'Level', 'Mgi_id', 'Havana_gene', 'Tag', 'Ranges_id',\n",
      "       'Intron_number', 'Ranges_ID'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(exons.columns) # get_exons don't return \"Transcript_support_level\", 'Havana_transcript', 'Protein_id', 'Ccdsid', 'Ont', 'Parent', 'Exon_id', 'Exon_number'\n",
    "print(introns.columns) # get_introns specific fields return 'Level_1'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Summarise 3\\` 5\\` UTRs exons and introns regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mean_across_region(df, var):\n",
    "    sum_var = df[f\"{var}_sum\"].sum()\n",
    "    sum_total_c = df[\"num_total_c_sum\"].sum()\n",
    "    return sum_var / sum_total_c\n",
    "def summarise_across_region(rdr, region):\n",
    "    \"\"\"\n",
    "    Summarise the methylation data across a region.\n",
    "\n",
    "    Parameters\n",
    "        rdr: pyranges object\n",
    "        region: str\n",
    "\n",
    "    Returns\n",
    "        grouped_df: pd.DataFrame\n",
    "    \"\"\"\n",
    "\n",
    "    df = rdr.to_dataframe().reset_index(level=\"sample_id\", drop=True)\n",
    "    grouped = df.groupby(\"Gene_id\")\n",
    "\n",
    "    mean_mc = grouped.apply(get_mean_across_region, var=\"num_mc\")\n",
    "    mean_hmc = grouped.apply(get_mean_across_region, var=\"num_hmc\")\n",
    "    mean_modc = grouped.apply(get_mean_across_region, var=\"num_modc\")\n",
    "    cpg_count = grouped[\"num_total_c_cpg_count\"].sum()\n",
    "    range_length = grouped[\"range_length\"].sum()\n",
    "    gene_name = grouped[\"Gene_name\"].first()\n",
    "    contig = grouped[\"contig\"].first()\n",
    "    strand = grouped[\"strand\"].first()\n",
    "    \n",
    "    \n",
    "    # Prepare list for storing primary_transcripts and handling missing gene_ids\n",
    "    primary_transcripts = []\n",
    "    missing_gene_ids = []\n",
    "\n",
    "    for gene_id in grouped.groups.keys():\n",
    "        if gene_id in gene_to_transcription:\n",
    "            primary_transcripts.append(gene_to_transcription[gene_id])\n",
    "        else:\n",
    "            primary_transcripts.append(None)\n",
    "            missing_gene_ids.append(gene_id)\n",
    "\n",
    "    # Print missing gene_ids\n",
    "    for gene_id in missing_gene_ids:\n",
    "        print(f\"Gene_ID {gene_id} not found in selected_transcripts\")\n",
    "\n",
    "    \n",
    "    grouped_df = pd.DataFrame(\n",
    "        {\n",
    "            \"mean_mc\": mean_mc,\n",
    "            \"mean_hmc\": mean_hmc,\n",
    "            \"mean_modc\": mean_modc,\n",
    "            \"cpg_count\": cpg_count,\n",
    "            \"range_length\": range_length,\n",
    "            \"Gene_name\": gene_name,\n",
    "            \"contig\": contig,\n",
    "            \"strand\": strand,\n",
    "            \"Transcript_id\": primary_transcripts,\n",
    "        }\n",
    "    )\n",
    "\n",
    "    grouped_df = grouped_df.reset_index()\n",
    "    grouped_df[\"Region\"] = region\n",
    "    return grouped_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'exons': +--------------+------------+------------+-----------+-----------+-------+\n",
       " | Chromosome   | Source     | Type       | Start     | End       | +22   |\n",
       " | (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       " |--------------+------------+------------+-----------+-----------+-------|\n",
       " | 1            | HAVANA     | exon       | 4808454   | 4808485   | ...   |\n",
       " | 1            | HAVANA     | exon       | 4828583   | 4828648   | ...   |\n",
       " | 1            | HAVANA     | exon       | 4828583   | 4828648   | ...   |\n",
       " | 1            | HAVANA     | exon       | 4830267   | 4830314   | ...   |\n",
       " | ...          | ...        | ...        | ...       | ...       | ...   |\n",
       " | Y            | HAVANA     | exon       | 90429592  | 90429709  | ...   |\n",
       " | Y            | HAVANA     | exon       | 90430460  | 90430559  | ...   |\n",
       " | Y            | HAVANA     | exon       | 90432572  | 90432670  | ...   |\n",
       " | Y            | HAVANA     | exon       | 90433029  | 90433262  | ...   |\n",
       " +--------------+------------+------------+-----------+-----------+-------+\n",
       " Unstranded PyRanges object has 90,807 rows and 27 columns from 21 chromosomes.\n",
       " For printing, the PyRanges was sorted on Chromosome.\n",
       " 22 hidden columns: Score, Phase, Id, Parent, Gene_id, Transcript_id, Gene_type, ... (+ 15 more.),\n",
       " 'introns': +-----------------------+-----------+--------------+------------+-------+\n",
       " | Transcript_id         | Level_1   | Chromosome   | Source     | +16   |\n",
       " | (object)              | (int64)   | (category)   | (object)   | ...   |\n",
       " |-----------------------+-----------+--------------+------------+-------|\n",
       " | ENSMUST00000000514.10 | 1         | 1            | HAVANA     | ...   |\n",
       " | ENSMUST00000000514.10 | 2         | 1            | HAVANA     | ...   |\n",
       " | ENSMUST00000000514.10 | 3         | 1            | HAVANA     | ...   |\n",
       " | ENSMUST00000000514.10 | 4         | 1            | HAVANA     | ...   |\n",
       " | ...                   | ...       | ...          | ...        | ...   |\n",
       " | ENSMUST00000194687.1  | 1         | Y            | HAVANA     | ...   |\n",
       " | ENSMUST00000195574.1  | 1         | Y            | HAVANA     | ...   |\n",
       " | ENSMUST00000195574.1  | 2         | Y            | HAVANA     | ...   |\n",
       " | ENSMUST00000195626.1  | 1         | Y            | HAVANA     | ...   |\n",
       " +-----------------------+-----------+--------------+------------+-------+\n",
       " Unstranded PyRanges object has 85,982 rows and 20 columns from 21 chromosomes.\n",
       " For printing, the PyRanges was sorted on Chromosome.\n",
       " 16 hidden columns: Type, Start, End, Score, Phase, Id, Gene_id, Gene_type, Gene_name, ... (+ 7 more.),\n",
       " 'five_prime_utrs': +--------------+------------+----------------+-----------+-------+\n",
       " | Chromosome   | Source     | Type           | Start     | +23   |\n",
       " | (category)   | (object)   | (object)       | (int64)   | ...   |\n",
       " |--------------+------------+----------------+-----------+-------|\n",
       " | 1            | HAVANA     | five_prime_UTR | 4807822   | ...   |\n",
       " | 1            | HAVANA     | five_prime_UTR | 4857813   | ...   |\n",
       " | 1            | HAVANA     | five_prime_UTR | 5588492   | ...   |\n",
       " | 1            | HAVANA     | five_prime_UTR | 5589034   | ...   |\n",
       " | ...          | ...        | ...            | ...       | ...   |\n",
       " | Y            | HAVANA     | five_prime_UTR | 87550965  | ...   |\n",
       " | Y            | HAVANA     | five_prime_UTR | 87552378  | ...   |\n",
       " | Y            | HAVANA     | five_prime_UTR | 88053314  | ...   |\n",
       " | Y            | HAVANA     | five_prime_UTR | 88054723  | ...   |\n",
       " +--------------+------------+----------------+-----------+-------+\n",
       " Unstranded PyRanges object has 11,449 rows and 27 columns from 21 chromosomes.\n",
       " For printing, the PyRanges was sorted on Chromosome.\n",
       " 23 hidden columns: End, Score, Phase, Id, Parent, Gene_id, Transcript_id, ... (+ 16 more.),\n",
       " 'three_prime_utrs': +--------------+------------+-----------------+-----------+-------+\n",
       " | Chromosome   | Source     | Type            | Start     | +23   |\n",
       " | (category)   | (object)   | (object)        | (int64)   | ...   |\n",
       " |--------------+------------+-----------------+-----------+-------|\n",
       " | 1            | HAVANA     | three_prime_UTR | 4845016   | ...   |\n",
       " | 1            | HAVANA     | three_prime_UTR | 4896364   | ...   |\n",
       " | 1            | HAVANA     | three_prime_UTR | 5602784   | ...   |\n",
       " | 1            | HAVANA     | three_prime_UTR | 6274275   | ...   |\n",
       " | ...          | ...        | ...             | ...       | ...   |\n",
       " | Y            | HAVANA     | three_prime_UTR | 86087758  | ...   |\n",
       " | Y            | HAVANA     | three_prime_UTR | 87142720  | ...   |\n",
       " | Y            | HAVANA     | three_prime_UTR | 87575495  | ...   |\n",
       " | Y            | HAVANA     | three_prime_UTR | 88079297  | ...   |\n",
       " +--------------+------------+-----------------+-----------+-------+\n",
       " Unstranded PyRanges object has 8,075 rows and 27 columns from 21 chromosomes.\n",
       " For printing, the PyRanges was sorted on Chromosome.\n",
       " 23 hidden columns: End, Score, Phase, Id, Parent, Gene_id, Transcript_id, ... (+ 16 more.)}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict_regions_to_patch_plus = {\n",
    "    \"exons\": exons_plus,\n",
    "    \"introns\": introns_plus,\n",
    "    \"five_prime_utrs\": five_prime_utrs_plus,\n",
    "    \"three_prime_utrs\": three_prime_utrs_plus,\n",
    "}\n",
    "\n",
    "dict_regions_to_patch_minus = {\n",
    "    \"exons\": exons_minus,\n",
    "    \"introns\": introns_minus,\n",
    "    \"five_prime_utrs\": five_prime_utrs_minus,\n",
    "    \"three_prime_utrs\": three_prime_utrs_minus,\n",
    "}\n",
    "\n",
    "dict_regions_to_patch_plus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Build sense and antisense methylation summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarise_methylation_across_regions(plus_strand_cpgs, minus_strand_cpgs, regions_dict, region_strand, sense_methylation={}, antisense_methylation={}):\n",
    "    \"\"\"\n",
    "    Summarise the methylation data across multiple regions.\n",
    "\n",
    "    Parameters\n",
    "        rdr: pyranges object\n",
    "        regions_dict: dict\n",
    "\n",
    "    Returns\n",
    "        grouped_df: pd.DataFrame\n",
    "    \"\"\"\n",
    "    # For each region in + strand expressed genes, summarise the methylation data\n",
    "    for region, gr in regions_dict.items():\n",
    "        rds_plus = plus_strand.reduce_byranges(\n",
    "                    ranges=gr.unstrand(), \n",
    "                    var=[\"num_mc\", \"num_hmc\", \"num_modc\", \"num_total_c\"]\n",
    "                )\n",
    "        \n",
    "        rds_minus = minus_strand.reduce_byranges(\n",
    "                ranges=gr.unstrand(), \n",
    "                var=[\"num_mc\", \"num_hmc\", \"num_modc\", \"num_total_c\"]\n",
    "            )\n",
    "        \n",
    "        if region_strand == \"+\":\n",
    "            sense_plus_ds = rds_plus\n",
    "            plus_sense = ['+'] * sense_plus_ds.dims['ranges']\n",
    "            sense_plus_ds = sense_plus_ds.assign_coords(strand=('ranges', plus_sense))\n",
    "        \n",
    "            antisense_plus_ds = rds_minus\n",
    "            plus_antisense = ['-'] * antisense_plus_ds.dims['ranges']\n",
    "            antisense_plus_ds = antisense_plus_ds.assign_coords(strand=('ranges', plus_antisense))\n",
    "\n",
    "            # Add the modified dataset to the dictionary\n",
    "            if region not in sense_methylation:\n",
    "                sense_methylation[region] = sense_plus_ds\n",
    "                antisense_methylation[region] = antisense_plus_ds\n",
    "            else:\n",
    "                sense_methylation[region] = xr.concat([sense_methylation[region], sense_plus_ds], dim=\"ranges\")\n",
    "                antisense_methylation[region] = xr.concat([antisense_methylation[region], antisense_plus_ds], dim=\"ranges\")\n",
    "        else:\n",
    "            antisense_minus_ds = rds_plus\n",
    "            minus_antisense = ['+'] * antisense_minus_ds.dims['ranges']\n",
    "            antisense_minus_ds = antisense_minus_ds.assign_coords(strand=('ranges', minus_antisense))\n",
    "            \n",
    "            sense_minus_ds = rds_minus\n",
    "            minus_sense = ['-'] * sense_minus_ds.dims['ranges']\n",
    "            sense_minus_ds = sense_minus_ds.assign_coords(strand=('ranges', minus_sense))\n",
    "\n",
    "            # Add the modified dataset to the dictionary\n",
    "            if region not in antisense_methylation:\n",
    "                sense_methylation[region] = sense_minus_ds\n",
    "                antisense_methylation[region] = antisense_minus_ds\n",
    "            else:\n",
    "                sense_methylation[region] = xr.concat([sense_methylation[region], sense_minus_ds], dim=\"ranges\")\n",
    "                antisense_methylation[region] = xr.concat([antisense_methylation[region], antisense_minus_ds], dim=\"ranges\")\n",
    "\n",
    "\n",
    "    return sense_methylation, antisense_methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "sense_methylation, antisense_methylation = summarise_methylation_across_regions(plus_strand, minus_strand, dict_regions_to_patch_plus, \"+\")\n",
    "sense_methylation, antisense_methylation = summarise_methylation_across_regions(plus_strand, minus_strand, dict_regions_to_patch_minus, \"-\", sense_methylation, antisense_methylation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;\n",
       "Dimensions:                   (ranges: 361948, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) &lt;U2 &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;\n",
       "    start                     (ranges) int64 4808454 4828583 ... 89743993\n",
       "    end                       (ranges) int64 4808485 4828648 ... 89744124\n",
       "    range_id                  (ranges) int64 5 9 13 14 ... 408327 408328 408329\n",
       "    num_contexts              (ranges) int64 1 0 0 0 0 2 2 0 ... 1 0 3 0 1 0 1 1\n",
       "    range_length              (ranges) int64 31 65 65 47 47 ... 34 66 34 66 131\n",
       "  * sample_id                 (sample_id) &lt;U8 &#x27;sample_0&#x27;\n",
       "    strand                    (ranges) &lt;U1 &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 0.0 ... 0.0 6.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 nan ... 0.0 6.0\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 1.0 0.0 ... 1.0 1.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.0 nan ... 0.0 0.0\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 1.0 0.0 ... 1.0 1.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 0.0 0.0 ... 0.0 6.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.0 nan ... 0.0 6.0\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 1.0 0.0 ... 1.0 1.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 113.0 0.0 ... 0.0 7.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 113.0 nan ... 0.0 7.0\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 1.0 0.0 ... 1.0 1.0\n",
       "    Source                    (ranges) object &#x27;HAVANA&#x27; &#x27;HAVANA&#x27; ... &#x27;HAVANA&#x27;\n",
       "    Type                      (ranges) object &#x27;exon&#x27; &#x27;exon&#x27; ... &#x27;exon&#x27; &#x27;exon&#x27;\n",
       "    Score                     (ranges) object &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;\n",
       "    Phase                     (ranges) object &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;\n",
       "    Id                        (ranges) object &#x27;exon:ENSMUST00000027036.10:2&#x27; ...\n",
       "    Parent                    (ranges) object &#x27;ENSMUST00000027036.10&#x27; ... &#x27;EN...\n",
       "    Gene_id                   (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Transcript_id             (ranges) object &#x27;ENSMUST00000027036.10&#x27; ... &#x27;EN...\n",
       "    Gene_type                 (ranges) object &#x27;protein_coding&#x27; ... &#x27;protein_c...\n",
       "    Gene_name                 (ranges) object &#x27;Lypla1&#x27; &#x27;Lypla1&#x27; ... &#x27;Gm21996&#x27;\n",
       "    Transcript_type           (ranges) object &#x27;protein_coding&#x27; ... &#x27;protein_c...\n",
       "    Transcript_name           (ranges) object &#x27;Lypla1-201&#x27; ... &#x27;Gm21996-201&#x27;\n",
       "    Exon_number               (ranges) object &#x27;2&#x27; &#x27;3&#x27; &#x27;2&#x27; &#x27;4&#x27; ... &#x27;4&#x27; &#x27;3&#x27; &#x27;2&#x27;\n",
       "    Exon_id                   (ranges) object &#x27;ENSMUSE00001255376.1&#x27; ... &#x27;ENS...\n",
       "    Level                     (ranges) object &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;\n",
       "    Transcript_support_level  (ranges) object &#x27;1&#x27; &#x27;1&#x27; &#x27;5&#x27; &#x27;1&#x27; ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;\n",
       "    Mgi_id                    (ranges) object &#x27;MGI:1344588&#x27; ... &#x27;MGI:5440224&#x27;\n",
       "    Tag                       (ranges) object &#x27;basic,appris_principal_1,CCDS&#x27;...\n",
       "    Havana_gene               (ranges) object &#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTT...\n",
       "    Havana_transcript         (ranges) object &#x27;OTTMUST00000051162.1&#x27; ... &#x27;OTT...\n",
       "    Protein_id                (ranges) object &#x27;ENSMUSP00000027036.4&#x27; ... &#x27;ENS...\n",
       "    Ccdsid                    (ranges) object &#x27;CCDS14806.1&#x27; &#x27;CCDS14806.1&#x27; ... &#x27;&#x27;\n",
       "    Ont                       (ranges) object &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; ... &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27;</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-73c244a1-8e0b-4343-8586-73c5d8702370' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-73c244a1-8e0b-4343-8586-73c5d8702370' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>ranges</span>: 361948</li><li><span class='xr-has-index'>sample_id</span>: 1</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-ae959a0d-2628-4ca7-9d62-679303ab6398' class='xr-section-summary-in' type='checkbox'  checked><label for='section-ae959a0d-2628-4ca7-9d62-679303ab6398' class='xr-section-summary' >Coordinates: <span>(8)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>contig</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U2</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;</div><input id='attrs-0663ab32-3908-4a84-ae0d-e485a44227b2' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0663ab32-3908-4a84-ae0d-e485a44227b2' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5157d80b-7f07-4219-a83f-9d1b5fdb03b0' class='xr-var-data-in' type='checkbox'><label for='data-5157d80b-7f07-4219-a83f-9d1b5fdb03b0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;, ..., &#x27;Y&#x27;, &#x27;Y&#x27;, &#x27;Y&#x27;], dtype=&#x27;&lt;U2&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>start</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4808454 4828583 ... 89743993</div><input id='attrs-ce48c680-de52-4aa0-90e4-e1a960b68446' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ce48c680-de52-4aa0-90e4-e1a960b68446' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5b798ebf-04af-4723-9d98-43f337e13766' class='xr-var-data-in' type='checkbox'><label for='data-5b798ebf-04af-4723-9d98-43f337e13766' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4808454,  4828583,  4828583, ..., 89731936, 89732605, 89743993])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>end</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4808485 4828648 ... 89744124</div><input id='attrs-8521cf94-5a73-429c-ad24-f691b7fca474' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8521cf94-5a73-429c-ad24-f691b7fca474' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c1b0b25e-2a82-445b-86d1-6d18c7e56945' class='xr-var-data-in' type='checkbox'><label for='data-c1b0b25e-2a82-445b-86d1-6d18c7e56945' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4808485,  4828648,  4828648, ..., 89731970, 89732671, 89744124])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>5 9 13 14 ... 408327 408328 408329</div><input id='attrs-67e3cd1c-0181-44d1-bebf-79603d3cbb13' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-67e3cd1c-0181-44d1-bebf-79603d3cbb13' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-49742e9c-8aa0-4f8a-90ad-3cd73ef3316b' class='xr-var-data-in' type='checkbox'><label for='data-49742e9c-8aa0-4f8a-90ad-3cd73ef3316b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([     5,      9,     13, ..., 408327, 408328, 408329])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_contexts</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>1 0 0 0 0 2 2 0 ... 1 0 3 0 1 0 1 1</div><input id='attrs-ca1a143e-c6a4-4059-8878-4ec8f0ce4043' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ca1a143e-c6a4-4059-8878-4ec8f0ce4043' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-eef3de30-b1a7-4e6d-a78f-e5d95c8a03e0' class='xr-var-data-in' type='checkbox'><label for='data-eef3de30-b1a7-4e6d-a78f-e5d95c8a03e0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([1, 0, 0, ..., 0, 1, 1])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_length</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>31 65 65 47 47 ... 34 66 34 66 131</div><input id='attrs-39957131-a3d6-46c3-82c7-c522e828f56c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-39957131-a3d6-46c3-82c7-c522e828f56c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-14ea01df-cebf-4561-9d08-5d0e4dc9e279' class='xr-var-data-in' type='checkbox'><label for='data-14ea01df-cebf-4561-9d08-5d0e4dc9e279' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 31,  65,  65, ...,  34,  66, 131])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample_id</span></div><div class='xr-var-dims'>(sample_id)</div><div class='xr-var-dtype'>&lt;U8</div><div class='xr-var-preview xr-preview'>&#x27;sample_0&#x27;</div><input id='attrs-c2048f4f-5d6a-4254-bb51-9d77637a801a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c2048f4f-5d6a-4254-bb51-9d77637a801a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7b09e3da-2b07-42b9-b0aa-b4fd84789cf9' class='xr-var-data-in' type='checkbox'><label for='data-7b09e3da-2b07-42b9-b0aa-b4fd84789cf9' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;sample_0&#x27;], dtype=&#x27;&lt;U8&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U1</div><div class='xr-var-preview xr-preview'>&#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;</div><input id='attrs-b283ef15-98f6-4b74-a571-9c3ff78dde64' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b283ef15-98f6-4b74-a571-9c3ff78dde64' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-55ca2881-9d57-4728-ba7b-8eed73049466' class='xr-var-data-in' type='checkbox'><label for='data-55ca2881-9d57-4728-ba7b-8eed73049466' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;+&#x27;, &#x27;+&#x27;, &#x27;+&#x27;, ..., &#x27;-&#x27;, &#x27;-&#x27;, &#x27;-&#x27;], dtype=&#x27;&lt;U1&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-249ce93c-9860-490c-807b-87660992366a' class='xr-section-summary-in' type='checkbox'  ><label for='section-249ce93c-9860-490c-807b-87660992366a' class='xr-section-summary' >Data variables: <span>(35)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 0.0 ... 0.0 0.0 6.0</div><input id='attrs-f1d2142c-6ee1-4ab3-85d2-60ee67f0bb1d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f1d2142c-6ee1-4ab3-85d2-60ee67f0bb1d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-60718c5e-4764-47da-8064-eaea838e85c7' class='xr-var-data-in' type='checkbox'><label for='data-60718c5e-4764-47da-8064-eaea838e85c7' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [6.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 nan nan nan ... nan 0.0 6.0</div><input id='attrs-dd0b7f28-d144-4a9e-917f-ff93729b21db' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-dd0b7f28-d144-4a9e-917f-ff93729b21db' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c2e4d73b-4a87-4d4f-b271-fd5a8ab126c9' class='xr-var-data-in' type='checkbox'><label for='data-c2e4d73b-4a87-4d4f-b271-fd5a8ab126c9' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [nan],\n",
       "       [nan],\n",
       "       ...,\n",
       "       [nan],\n",
       "       [ 0.],\n",
       "       [ 6.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 1.0 0.0 1.0 1.0</div><input id='attrs-f0ba1786-7cec-40dc-8355-288c3fc92e5b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f0ba1786-7cec-40dc-8355-288c3fc92e5b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8f2ae085-ab91-4580-a260-cb7d143c26ea' class='xr-var-data-in' type='checkbox'><label for='data-8f2ae085-ab91-4580-a260-cb7d143c26ea' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [1.],\n",
       "       [1.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0</div><input id='attrs-808d6773-d6e2-43ca-a8cc-5119040fab3d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-808d6773-d6e2-43ca-a8cc-5119040fab3d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-09ef651d-2a12-47eb-b4d5-b7c641ed5659' class='xr-var-data-in' type='checkbox'><label for='data-09ef651d-2a12-47eb-b4d5-b7c641ed5659' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 nan nan nan ... 0.0 nan 0.0 0.0</div><input id='attrs-be87bfac-c289-4b98-9a51-7d6b36084ffc' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-be87bfac-c289-4b98-9a51-7d6b36084ffc' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d9cf5d56-468a-4911-ae58-45651b5abea3' class='xr-var-data-in' type='checkbox'><label for='data-d9cf5d56-468a-4911-ae58-45651b5abea3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [nan],\n",
       "       [nan],\n",
       "       ...,\n",
       "       [nan],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 1.0 0.0 1.0 1.0</div><input id='attrs-c50cf3c7-7850-4f76-baa2-069e6589e5df' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c50cf3c7-7850-4f76-baa2-069e6589e5df' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-773380d3-ccda-4d43-b0b0-b8d9d06f300c' class='xr-var-data-in' type='checkbox'><label for='data-773380d3-ccda-4d43-b0b0-b8d9d06f300c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [1.],\n",
       "       [1.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 0.0 ... 0.0 0.0 6.0</div><input id='attrs-95dc5212-d555-40c4-97bf-796309bae268' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-95dc5212-d555-40c4-97bf-796309bae268' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-13934136-6d1c-4635-8ae7-e306abe05c89' class='xr-var-data-in' type='checkbox'><label for='data-13934136-6d1c-4635-8ae7-e306abe05c89' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [6.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 nan nan nan ... nan 0.0 6.0</div><input id='attrs-b7b75b71-78ef-4171-995f-2695a53a2fec' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b7b75b71-78ef-4171-995f-2695a53a2fec' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5ac325a6-5c5c-4447-992a-06074440600f' class='xr-var-data-in' type='checkbox'><label for='data-5ac325a6-5c5c-4447-992a-06074440600f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [nan],\n",
       "       [nan],\n",
       "       ...,\n",
       "       [nan],\n",
       "       [ 0.],\n",
       "       [ 6.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 1.0 0.0 1.0 1.0</div><input id='attrs-ecd00b88-ba53-4479-a0be-4505a0f2d5d1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ecd00b88-ba53-4479-a0be-4505a0f2d5d1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1a52e244-596b-4379-ac85-9e48e1c916a3' class='xr-var-data-in' type='checkbox'><label for='data-1a52e244-596b-4379-ac85-9e48e1c916a3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [1.],\n",
       "       [1.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>113.0 0.0 0.0 0.0 ... 0.0 0.0 7.0</div><input id='attrs-f2572326-8e5b-4012-81f6-73aea09e3ea2' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f2572326-8e5b-4012-81f6-73aea09e3ea2' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-104a84f5-c8e4-400a-9e47-55850ebd14f7' class='xr-var-data-in' type='checkbox'><label for='data-104a84f5-c8e4-400a-9e47-55850ebd14f7' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[113.],\n",
       "       [  0.],\n",
       "       [  0.],\n",
       "       ...,\n",
       "       [  0.],\n",
       "       [  0.],\n",
       "       [  7.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>113.0 nan nan nan ... nan 0.0 7.0</div><input id='attrs-e6f8cc93-0458-4d99-9af3-b528c234d43c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e6f8cc93-0458-4d99-9af3-b528c234d43c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d800cf67-dad2-4a49-889d-492e95425228' class='xr-var-data-in' type='checkbox'><label for='data-d800cf67-dad2-4a49-889d-492e95425228' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[113.],\n",
       "       [ nan],\n",
       "       [ nan],\n",
       "       ...,\n",
       "       [ nan],\n",
       "       [  0.],\n",
       "       [  7.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 1.0 0.0 1.0 1.0</div><input id='attrs-82210f53-9f71-4253-a24d-9a751ff44d9a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-82210f53-9f71-4253-a24d-9a751ff44d9a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-91ae0465-5031-4f82-a3e2-2b370680068d' class='xr-var-data-in' type='checkbox'><label for='data-91ae0465-5031-4f82-a3e2-2b370680068d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [1.],\n",
       "       [1.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Source</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;HAVANA&#x27; &#x27;HAVANA&#x27; ... &#x27;HAVANA&#x27;</div><input id='attrs-00bbd58c-03c2-4db3-9789-5d08e7383b1a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-00bbd58c-03c2-4db3-9789-5d08e7383b1a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4bbb2b2f-696a-4688-b4d2-d6da08a8d5a8' class='xr-var-data-in' type='checkbox'><label for='data-4bbb2b2f-696a-4688-b4d2-d6da08a8d5a8' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;, ..., &#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;exon&#x27; &#x27;exon&#x27; ... &#x27;exon&#x27; &#x27;exon&#x27;</div><input id='attrs-b6b2a7bb-9d91-4697-a3c2-7d807b98597c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b6b2a7bb-9d91-4697-a3c2-7d807b98597c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-77798b60-6ed2-4329-a79b-685fd90f8ac9' class='xr-var-data-in' type='checkbox'><label for='data-77798b60-6ed2-4329-a79b-685fd90f8ac9' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;exon&#x27;, &#x27;exon&#x27;, &#x27;exon&#x27;, ..., &#x27;exon&#x27;, &#x27;exon&#x27;, &#x27;exon&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Score</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;</div><input id='attrs-59d7cb0f-6e7a-4aa7-b032-5f69f6db431c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-59d7cb0f-6e7a-4aa7-b032-5f69f6db431c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d8dc2b14-28d5-45fd-814f-6ea7b4319ff4' class='xr-var-data-in' type='checkbox'><label for='data-d8dc2b14-28d5-45fd-814f-6ea7b4319ff4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;, ..., &#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Phase</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;</div><input id='attrs-c44f440f-414a-492e-8fa1-eee6825bbf38' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c44f440f-414a-492e-8fa1-eee6825bbf38' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-20390a9b-5b0f-4f1c-88cd-9bda143b6a92' class='xr-var-data-in' type='checkbox'><label for='data-20390a9b-5b0f-4f1c-88cd-9bda143b6a92' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;, ..., &#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;exon:ENSMUST00000027036.10:2&#x27; ....</div><input id='attrs-01f896de-f5fe-4065-87ff-c26e66ef3737' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-01f896de-f5fe-4065-87ff-c26e66ef3737' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-48993ef9-6386-4605-bb5f-386dcee81521' class='xr-var-data-in' type='checkbox'><label for='data-48993ef9-6386-4605-bb5f-386dcee81521' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;exon:ENSMUST00000027036.10:2&#x27;, &#x27;exon:ENSMUST00000027036.10:3&#x27;,\n",
       "       &#x27;exon:ENSMUST00000155020.1:2&#x27;, ..., &#x27;exon:ENSMUST00000188269.1:4&#x27;,\n",
       "       &#x27;exon:ENSMUST00000188269.1:3&#x27;, &#x27;exon:ENSMUST00000188269.1:2&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Parent</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUST00000027036.10&#x27; ... &#x27;ENS...</div><input id='attrs-ecc05df8-d023-4266-8074-2c1e261ce1cd' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ecc05df8-d023-4266-8074-2c1e261ce1cd' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9145f9a9-75f7-4301-af88-ab84ce11e130' class='xr-var-data-in' type='checkbox'><label for='data-9145f9a9-75f7-4301-af88-ab84ce11e130' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUST00000027036.10&#x27;, &#x27;ENSMUST00000027036.10&#x27;,\n",
       "       &#x27;ENSMUST00000155020.1&#x27;, ..., &#x27;ENSMUST00000188269.1&#x27;,\n",
       "       &#x27;ENSMUST00000188269.1&#x27;, &#x27;ENSMUST00000188269.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-06afe442-2f54-46ea-9f6a-b4638ba4d605' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-06afe442-2f54-46ea-9f6a-b4638ba4d605' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-514dc604-761e-4d4a-870f-cf786e2ae14b' class='xr-var-data-in' type='checkbox'><label for='data-514dc604-761e-4d4a-870f-cf786e2ae14b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000025903.14&#x27;,\n",
       "       &#x27;ENSMUSG00000104217.1&#x27;, ..., &#x27;ENSMUSG00000100608.1&#x27;,\n",
       "       &#x27;ENSMUSG00000100608.1&#x27;, &#x27;ENSMUSG00000100608.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUST00000027036.10&#x27; ... &#x27;ENS...</div><input id='attrs-f774d3a3-34a6-4daf-9989-56ef4f393a16' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f774d3a3-34a6-4daf-9989-56ef4f393a16' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e973d29e-b008-4813-8166-1071632934e2' class='xr-var-data-in' type='checkbox'><label for='data-e973d29e-b008-4813-8166-1071632934e2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUST00000027036.10&#x27;, &#x27;ENSMUST00000027036.10&#x27;,\n",
       "       &#x27;ENSMUST00000155020.1&#x27;, ..., &#x27;ENSMUST00000188269.1&#x27;,\n",
       "       &#x27;ENSMUST00000188269.1&#x27;, &#x27;ENSMUST00000188269.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;protein_coding&#x27; ... &#x27;protein_co...</div><input id='attrs-2f9210fe-e40f-4f9b-9274-52b539454882' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2f9210fe-e40f-4f9b-9274-52b539454882' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-83aee5df-595a-412f-a8fe-92c29c7d2e26' class='xr-var-data-in' type='checkbox'><label for='data-83aee5df-595a-412f-a8fe-92c29c7d2e26' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, ...,\n",
       "       &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Lypla1&#x27; &#x27;Lypla1&#x27; ... &#x27;Gm21996&#x27;</div><input id='attrs-d543141b-1352-44f2-a6f9-3da997fda69d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d543141b-1352-44f2-a6f9-3da997fda69d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-f25f0b60-91a0-45cf-8081-130243196930' class='xr-var-data-in' type='checkbox'><label for='data-f25f0b60-91a0-45cf-8081-130243196930' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Lypla1&#x27;, &#x27;Lypla1&#x27;, &#x27;Gm37988&#x27;, ..., &#x27;Gm21996&#x27;, &#x27;Gm21996&#x27;,\n",
       "       &#x27;Gm21996&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;protein_coding&#x27; ... &#x27;protein_co...</div><input id='attrs-6ee6d8f2-6258-4e60-9ac9-b6d4b3a50f0f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6ee6d8f2-6258-4e60-9ac9-b6d4b3a50f0f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1be8e545-3254-4a44-b30e-ae17e8da349d' class='xr-var-data-in' type='checkbox'><label for='data-1be8e545-3254-4a44-b30e-ae17e8da349d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;nonsense_mediated_decay&#x27;, ...,\n",
       "       &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Lypla1-201&#x27; ... &#x27;Gm21996-201&#x27;</div><input id='attrs-f193a318-23f1-4cd9-922d-0da18dcaf9b9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f193a318-23f1-4cd9-922d-0da18dcaf9b9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a7a2d2ce-4765-4abf-b3fa-33df50e69579' class='xr-var-data-in' type='checkbox'><label for='data-a7a2d2ce-4765-4abf-b3fa-33df50e69579' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Lypla1-201&#x27;, &#x27;Lypla1-201&#x27;, &#x27;Gm37988-201&#x27;, ..., &#x27;Gm21996-201&#x27;,\n",
       "       &#x27;Gm21996-201&#x27;, &#x27;Gm21996-201&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;2&#x27; &#x27;3&#x27; &#x27;2&#x27; &#x27;4&#x27; ... &#x27;5&#x27; &#x27;4&#x27; &#x27;3&#x27; &#x27;2&#x27;</div><input id='attrs-f60840ed-9441-4795-a07c-5fb326b4254c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f60840ed-9441-4795-a07c-5fb326b4254c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-84095f1b-ba82-43b7-9a9e-c6171d89aafd' class='xr-var-data-in' type='checkbox'><label for='data-84095f1b-ba82-43b7-9a9e-c6171d89aafd' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;2&#x27;, &#x27;3&#x27;, &#x27;2&#x27;, ..., &#x27;4&#x27;, &#x27;3&#x27;, &#x27;2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSE00001255376.1&#x27; ... &#x27;ENSM...</div><input id='attrs-1955bc75-da07-42de-8dc9-effcfa75d63f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1955bc75-da07-42de-8dc9-effcfa75d63f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-68111546-e667-4e7a-927d-3fa07bdd9212' class='xr-var-data-in' type='checkbox'><label for='data-68111546-e667-4e7a-927d-3fa07bdd9212' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSE00001255376.1&#x27;, &#x27;ENSMUSE00001222304.1&#x27;,\n",
       "       &#x27;ENSMUSE00001241075.1&#x27;, ..., &#x27;ENSMUSE00001333847.1&#x27;,\n",
       "       &#x27;ENSMUSE00001331494.1&#x27;, &#x27;ENSMUSE00001332880.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;</div><input id='attrs-a84d7d3b-2be3-493a-a87b-db8b2a92cce6' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-a84d7d3b-2be3-493a-a87b-db8b2a92cce6' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-62bcf3e2-1e00-4aff-b161-37153b0a0e1c' class='xr-var-data-in' type='checkbox'><label for='data-62bcf3e2-1e00-4aff-b161-37153b0a0e1c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;, ..., &#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_support_level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;5&#x27; &#x27;1&#x27; ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;</div><input id='attrs-b74b54b7-f71b-4236-9812-a4a750106bf7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b74b54b7-f71b-4236-9812-a4a750106bf7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-953bab67-c2c6-497e-b1cf-2750fd62371b' class='xr-var-data-in' type='checkbox'><label for='data-953bab67-c2c6-497e-b1cf-2750fd62371b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;5&#x27;, ..., &#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Mgi_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;MGI:1344588&#x27; ... &#x27;MGI:5440224&#x27;</div><input id='attrs-cef8f63b-1357-4cfd-822b-b2a6650c18c8' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-cef8f63b-1357-4cfd-822b-b2a6650c18c8' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-f10c4a33-d6a8-470a-af18-920729c42910' class='xr-var-data-in' type='checkbox'><label for='data-f10c4a33-d6a8-470a-af18-920729c42910' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;MGI:1344588&#x27;, &#x27;MGI:1344588&#x27;, &#x27;MGI:5611216&#x27;, ..., &#x27;MGI:5440224&#x27;,\n",
       "       &#x27;MGI:5440224&#x27;, &#x27;MGI:5440224&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Tag</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;basic,appris_principal_1,CCDS&#x27; ...</div><input id='attrs-b7de0616-09f2-4a0b-ac15-f1102b07df84' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b7de0616-09f2-4a0b-ac15-f1102b07df84' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7d1e798f-be19-4786-8097-c42be235477c' class='xr-var-data-in' type='checkbox'><label for='data-7d1e798f-be19-4786-8097-c42be235477c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;basic,appris_principal_1,CCDS&#x27;, &#x27;basic,appris_principal_1,CCDS&#x27;,\n",
       "       &#x27;basic,appris_principal_1,readthrough_transcript&#x27;, ...,\n",
       "       &#x27;basic,appris_principal_1&#x27;, &#x27;basic,appris_principal_1&#x27;,\n",
       "       &#x27;basic,appris_principal_1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_gene</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTTM...</div><input id='attrs-39412533-85f2-4c39-becc-1e9aaba7994c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-39412533-85f2-4c39-becc-1e9aaba7994c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-172287c7-4d82-454e-8faa-05b1fcdcf921' class='xr-var-data-in' type='checkbox'><label for='data-172287c7-4d82-454e-8faa-05b1fcdcf921' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;OTTMUSG00000021562.4&#x27;, &#x27;OTTMUSG00000021562.4&#x27;,\n",
       "       &#x27;OTTMUSG00000050100.1&#x27;, ..., &#x27;OTTMUSG00000047316.1&#x27;,\n",
       "       &#x27;OTTMUSG00000047316.1&#x27;, &#x27;OTTMUSG00000047316.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_transcript</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;OTTMUST00000051162.1&#x27; ... &#x27;OTTM...</div><input id='attrs-ea79fbb3-2b32-4b66-8bfd-003cb73e2130' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ea79fbb3-2b32-4b66-8bfd-003cb73e2130' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-27b9fee0-ec72-43ad-98d3-166e00fe417d' class='xr-var-data-in' type='checkbox'><label for='data-27b9fee0-ec72-43ad-98d3-166e00fe417d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;OTTMUST00000051162.1&#x27;, &#x27;OTTMUST00000051162.1&#x27;,\n",
       "       &#x27;OTTMUST00000127419.1&#x27;, ..., &#x27;OTTMUST00000122318.1&#x27;,\n",
       "       &#x27;OTTMUST00000122318.1&#x27;, &#x27;OTTMUST00000122318.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Protein_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSP00000027036.4&#x27; ... &#x27;ENSM...</div><input id='attrs-95f30317-f25b-429d-8a20-795a176a4846' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-95f30317-f25b-429d-8a20-795a176a4846' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c2af547a-9578-4b2f-a3f4-bea2aea542f0' class='xr-var-data-in' type='checkbox'><label for='data-c2af547a-9578-4b2f-a3f4-bea2aea542f0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSP00000027036.4&#x27;, &#x27;ENSMUSP00000027036.4&#x27;,\n",
       "       &#x27;ENSMUSP00000136108.1&#x27;, ..., &#x27;ENSMUSP00000141173.1&#x27;,\n",
       "       &#x27;ENSMUSP00000141173.1&#x27;, &#x27;ENSMUSP00000141173.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ccdsid</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;CCDS14806.1&#x27; &#x27;CCDS14806.1&#x27; ... &#x27;&#x27;</div><input id='attrs-7950f50d-dca5-46f7-ac7d-703319a081cf' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7950f50d-dca5-46f7-ac7d-703319a081cf' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-69f4e050-4529-4406-8cf5-e73df726e263' class='xr-var-data-in' type='checkbox'><label for='data-69f4e050-4529-4406-8cf5-e73df726e263' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;CCDS14806.1&#x27;, &#x27;CCDS14806.1&#x27;, &#x27;&#x27;, ..., &#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ont</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; ... &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27;</div><input id='attrs-e7d8e7cd-08c8-4fab-a8af-752ac95ecaec' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e7d8e7cd-08c8-4fab-a8af-752ac95ecaec' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1d109504-a391-417d-80c8-29f28b5f2b3d' class='xr-var-data-in' type='checkbox'><label for='data-1d109504-a391-417d-80c8-29f28b5f2b3d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;, ..., &#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;], dtype=object)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-093a000b-cbdb-4f5b-a5a3-974f44dc05dd' class='xr-section-summary-in' type='checkbox'  ><label for='section-093a000b-cbdb-4f5b-a5a3-974f44dc05dd' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>sample_id</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-81ae2078-4b63-456b-a277-f24e734b60b9' class='xr-index-data-in' type='checkbox'/><label for='index-81ae2078-4b63-456b-a277-f24e734b60b9' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;sample_0&#x27;], dtype=&#x27;object&#x27;, name=&#x27;sample_id&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-93c8b259-eb9b-49e9-ae82-283fe9bc2789' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-93c8b259-eb9b-49e9-ae82-283fe9bc2789' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.Dataset>\n",
       "Dimensions:                   (ranges: 361948, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) <U2 '1' '1' '1' '1' ... 'Y' 'Y' 'Y' 'Y'\n",
       "    start                     (ranges) int64 4808454 4828583 ... 89743993\n",
       "    end                       (ranges) int64 4808485 4828648 ... 89744124\n",
       "    range_id                  (ranges) int64 5 9 13 14 ... 408327 408328 408329\n",
       "    num_contexts              (ranges) int64 1 0 0 0 0 2 2 0 ... 1 0 3 0 1 0 1 1\n",
       "    range_length              (ranges) int64 31 65 65 47 47 ... 34 66 34 66 131\n",
       "  * sample_id                 (sample_id) <U8 'sample_0'\n",
       "    strand                    (ranges) <U1 '+' '+' '+' '+' ... '-' '-' '-' '-'\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 0.0 ... 0.0 6.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 nan ... 0.0 6.0\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 1.0 0.0 ... 1.0 1.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.0 nan ... 0.0 0.0\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 1.0 0.0 ... 1.0 1.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 0.0 0.0 ... 0.0 6.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.0 nan ... 0.0 6.0\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 1.0 0.0 ... 1.0 1.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 113.0 0.0 ... 0.0 7.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 113.0 nan ... 0.0 7.0\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 1.0 0.0 ... 1.0 1.0\n",
       "    Source                    (ranges) object 'HAVANA' 'HAVANA' ... 'HAVANA'\n",
       "    Type                      (ranges) object 'exon' 'exon' ... 'exon' 'exon'\n",
       "    Score                     (ranges) object '.' '.' '.' '.' ... '.' '.' '.'\n",
       "    Phase                     (ranges) object '.' '.' '.' '.' ... '.' '.' '.'\n",
       "    Id                        (ranges) object 'exon:ENSMUST00000027036.10:2' ...\n",
       "    Parent                    (ranges) object 'ENSMUST00000027036.10' ... 'EN...\n",
       "    Gene_id                   (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Transcript_id             (ranges) object 'ENSMUST00000027036.10' ... 'EN...\n",
       "    Gene_type                 (ranges) object 'protein_coding' ... 'protein_c...\n",
       "    Gene_name                 (ranges) object 'Lypla1' 'Lypla1' ... 'Gm21996'\n",
       "    Transcript_type           (ranges) object 'protein_coding' ... 'protein_c...\n",
       "    Transcript_name           (ranges) object 'Lypla1-201' ... 'Gm21996-201'\n",
       "    Exon_number               (ranges) object '2' '3' '2' '4' ... '4' '3' '2'\n",
       "    Exon_id                   (ranges) object 'ENSMUSE00001255376.1' ... 'ENS...\n",
       "    Level                     (ranges) object '2' '2' '2' '2' ... '2' '2' '2'\n",
       "    Transcript_support_level  (ranges) object '1' '1' '5' '1' ... '1' '1' '1'\n",
       "    Mgi_id                    (ranges) object 'MGI:1344588' ... 'MGI:5440224'\n",
       "    Tag                       (ranges) object 'basic,appris_principal_1,CCDS'...\n",
       "    Havana_gene               (ranges) object 'OTTMUSG00000021562.4' ... 'OTT...\n",
       "    Havana_transcript         (ranges) object 'OTTMUST00000051162.1' ... 'OTT...\n",
       "    Protein_id                (ranges) object 'ENSMUSP00000027036.4' ... 'ENS...\n",
       "    Ccdsid                    (ranges) object 'CCDS14806.1' 'CCDS14806.1' ... ''\n",
       "    Ont                       (ranges) object '' '' '' '' '' ... '' '' '' '' ''"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sense_methylation['exons']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "calculate methlyation fractions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.807980</td>\n",
       "      <td>0.025436</td>\n",
       "      <td>0.847215</td>\n",
       "      <td>102.0</td>\n",
       "      <td>5990</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003.15</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.014888</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>14.0</td>\n",
       "      <td>1362</td>\n",
       "      <td>Pbsn</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000003.13</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028.15</td>\n",
       "      <td>0.581295</td>\n",
       "      <td>0.031589</td>\n",
       "      <td>0.619697</td>\n",
       "      <td>120.0</td>\n",
       "      <td>3910</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000028.13</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000037.17</td>\n",
       "      <td>0.759115</td>\n",
       "      <td>0.017578</td>\n",
       "      <td>0.787109</td>\n",
       "      <td>66.0</td>\n",
       "      <td>5554</td>\n",
       "      <td>Scml2</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000101113.8</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000049.11</td>\n",
       "      <td>0.783618</td>\n",
       "      <td>0.036519</td>\n",
       "      <td>0.835495</td>\n",
       "      <td>56.0</td>\n",
       "      <td>2136</td>\n",
       "      <td>Apoh</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000049.5</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  \\\n",
       "0   ENSMUSG00000000001.4  0.807980  0.025436   0.847215      102.0   \n",
       "1  ENSMUSG00000000003.15  0.774194  0.014888   0.806452       14.0   \n",
       "2  ENSMUSG00000000028.15  0.581295  0.031589   0.619697      120.0   \n",
       "3  ENSMUSG00000000037.17  0.759115  0.017578   0.787109       66.0   \n",
       "4  ENSMUSG00000000049.11  0.783618  0.036519   0.835495       56.0   \n",
       "\n",
       "   range_length Gene_name contig strand          Transcript_id Region  \n",
       "0          5990     Gnai3      3      -   ENSMUST00000000001.4  exons  \n",
       "1          1362      Pbsn      X      -  ENSMUST00000000003.13  exons  \n",
       "2          3910     Cdc45     16      -  ENSMUST00000000028.13  exons  \n",
       "3          5554     Scml2      X      +   ENSMUST00000101113.8  exons  \n",
       "4          2136      Apoh     11      +   ENSMUST00000000049.5  exons  "
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict_df_sense = {}\n",
    "for region, gr in dict_regions_to_patch_plus.items():\n",
    "    dict_df_sense[region] = summarise_across_region(sense_methylation[region], region)\n",
    "stitched_sense_regions = pd.concat(dict_df_sense.values())\n",
    "\n",
    "dict_df_antisense = {}\n",
    "for region, gr in dict_regions_to_patch_minus.items():\n",
    "    dict_df_antisense[region] = summarise_across_region(antisense_methylation[region], region)\n",
    "stitched_antisense_regions = pd.concat(dict_df_antisense.values())\n",
    "\n",
    "\n",
    "stitched_sense_regions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.807980</td>\n",
       "      <td>0.025436</td>\n",
       "      <td>0.847215</td>\n",
       "      <td>102.0</td>\n",
       "      <td>5990</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.780437</td>\n",
       "      <td>0.038735</td>\n",
       "      <td>0.832214</td>\n",
       "      <td>296.0</td>\n",
       "      <td>27124</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>introns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.000784</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000784</td>\n",
       "      <td>30.0</td>\n",
       "      <td>280</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.813870</td>\n",
       "      <td>0.023717</td>\n",
       "      <td>0.849446</td>\n",
       "      <td>64.0</td>\n",
       "      <td>4108</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  \\\n",
       "0  ENSMUSG00000000001.4  0.807980  0.025436   0.847215      102.0   \n",
       "0  ENSMUSG00000000001.4  0.780437  0.038735   0.832214      296.0   \n",
       "0  ENSMUSG00000000001.4  0.000784  0.000000   0.000784       30.0   \n",
       "0  ENSMUSG00000000001.4  0.813870  0.023717   0.849446       64.0   \n",
       "\n",
       "   range_length Gene_name contig strand         Transcript_id  \\\n",
       "0          5990     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0         27124     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0           280     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0          4108     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "\n",
       "             Region  \n",
       "0             exons  \n",
       "0           introns  \n",
       "0   five_prime_utrs  \n",
       "0  three_prime_utrs  "
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stitched_sense_regions[stitched_sense_regions['Gene_id']==\"ENSMUSG00000000001.4\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.810206</td>\n",
       "      <td>0.026502</td>\n",
       "      <td>0.848889</td>\n",
       "      <td>102.0</td>\n",
       "      <td>5990</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.772727</td>\n",
       "      <td>0.045917</td>\n",
       "      <td>0.834170</td>\n",
       "      <td>296.0</td>\n",
       "      <td>27124</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>introns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>30.0</td>\n",
       "      <td>280</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.816244</td>\n",
       "      <td>0.028680</td>\n",
       "      <td>0.856091</td>\n",
       "      <td>64.0</td>\n",
       "      <td>4108</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  \\\n",
       "0  ENSMUSG00000000001.4  0.810206  0.026502   0.848889      102.0   \n",
       "0  ENSMUSG00000000001.4  0.772727  0.045917   0.834170      296.0   \n",
       "0  ENSMUSG00000000001.4  0.000000  0.000000   0.000000       30.0   \n",
       "0  ENSMUSG00000000001.4  0.816244  0.028680   0.856091       64.0   \n",
       "\n",
       "   range_length Gene_name contig strand         Transcript_id  \\\n",
       "0          5990     Gnai3      3      +  ENSMUST00000000001.4   \n",
       "0         27124     Gnai3      3      +  ENSMUST00000000001.4   \n",
       "0           280     Gnai3      3      +  ENSMUST00000000001.4   \n",
       "0          4108     Gnai3      3      +  ENSMUST00000000001.4   \n",
       "\n",
       "             Region  \n",
       "0             exons  \n",
       "0           introns  \n",
       "0   five_prime_utrs  \n",
       "0  three_prime_utrs  "
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stitched_antisense_regions[stitched_antisense_regions['Gene_id']==\"ENSMUSG00000000001.4\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### no stitching"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### TSS, TES, first exon, first intron"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plus strand total regions: 63947\n",
      "minus strand total regions: 64156\n"
     ]
    }
   ],
   "source": [
    "print(f\"plus strand total regions: {len(before_tss_plus)+ len(after_tes_plus) + len(genes_clean) + len(first_exons_plus) + len(first_introns_plus)}\")\n",
    "print(f\"minus strand total regions: {len(before_tss_minus)+ len(after_tes_minus) + len(genes_clean) + len(first_exons_minus) + len(first_introns_minus)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop columns\n",
    "to_drop = [\"Score\", \"Source\", \"Phase\", \"Type\"]\n",
    "\n",
    "regions_dict_plus ={\n",
    "        \"before_tss\": before_tss_plus,\n",
    "        \"after_tes\": after_tes_plus,\n",
    "        \"genes\": genes_clean,\n",
    "        \"first_exons\": first_exons_plus,\n",
    "        \"first_introns\": first_introns_plus,\n",
    "    }\n",
    "\n",
    "for region in regions_dict_plus:\n",
    "    regions_dict_plus[region].Region = region\n",
    "    try:\n",
    "        regions_dict_plus[region] = regions_dict_plus[region].drop(to_drop)\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "regions_dict_minus ={\n",
    "        \"before_tss\": before_tss_minus,\n",
    "        \"after_tes\": after_tes_minus,\n",
    "        \"genes\": genes_clean,\n",
    "        \"first_exons\": first_exons_minus,\n",
    "        \"first_introns\": first_introns_minus,\n",
    "    }\n",
    "\n",
    "for region in regions_dict_minus:\n",
    "    regions_dict_minus[region].Region = region\n",
    "    try:\n",
    "        regions_dict_minus[region] = regions_dict_minus[region].drop(to_drop)\n",
    "    except:\n",
    "        pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarise_methylation_across_other_regions(plus_strand_cpgs, minus_strand_cpgs, regions_dict, strand, sense_methylation=None, antisense_methylation=None):\n",
    "    \"\"\"\n",
    "    Summarise the methylation data across multiple regions.\n",
    "\n",
    "    Parameters\n",
    "        rdr: pyranges object\n",
    "        regions_dict: dict\n",
    "\n",
    "    Returns\n",
    "        grouped_df: pd.DataFrame\n",
    "    \"\"\"\n",
    "\n",
    "    rds_plus = plus_strand_cpgs.reduce_byranges(\n",
    "                ranges=list(regions_dict.values()), \n",
    "                var=[\"num_mc\", \"num_hmc\", \"num_modc\", \"num_total_c\"]\n",
    "            )\n",
    "    rds_minus = minus_strand_cpgs.reduce_byranges(\n",
    "                ranges=list(regions_dict.values()),      \n",
    "                var=[\"num_mc\", \"num_hmc\", \"num_modc\", \"num_total_c\"]\n",
    "            )\n",
    "\n",
    "    if strand == \"+\":\n",
    "        sense_plus_ds = rds_plus\n",
    "        plus_sense = ['+'] * sense_plus_ds.dims['ranges']\n",
    "        sense_plus_ds = sense_plus_ds.assign_coords(strand=('ranges', plus_sense))\n",
    "    \n",
    "        antisense_plus_ds = rds_minus\n",
    "        plus_antisense = ['-'] * antisense_plus_ds.dims['ranges']\n",
    "        antisense_plus_ds = antisense_plus_ds.assign_coords(strand=('ranges', plus_antisense))\n",
    "\n",
    "        # Add the modified dataset to the dictionary\n",
    "        if sense_methylation is None:\n",
    "            sense_methylation= sense_plus_ds\n",
    "            antisense_methylation= antisense_plus_ds\n",
    "        else:\n",
    "            # print(\"hi:\" ,sense_methylation)\n",
    "            sense_methylation= xr.concat([sense_methylation, sense_plus_ds], dim=\"ranges\")\n",
    "            antisense_methylation= xr.concat([antisense_methylation, antisense_plus_ds], dim=\"ranges\")\n",
    "    else:\n",
    "        antisense_minus_ds = rds_plus\n",
    "        minus_antisense = ['+'] * antisense_minus_ds.dims['ranges']\n",
    "        antisense_minus_ds = antisense_minus_ds.assign_coords(strand=('ranges', minus_antisense))\n",
    "        \n",
    "        sense_minus_ds = rds_minus\n",
    "        minus_sense = ['-'] * sense_minus_ds.dims['ranges']\n",
    "        sense_minus_ds = sense_minus_ds.assign_coords(strand=('ranges', minus_sense))\n",
    "\n",
    "        # Add the modified dataset to the dictionary\n",
    "        if antisense_methylation is None:\n",
    "            sense_methylation = sense_minus_ds\n",
    "            antisense_methylation= antisense_minus_ds\n",
    "        else:\n",
    "            sense_methylation = xr.concat([sense_methylation, sense_minus_ds], dim=\"ranges\")\n",
    "            antisense_methylation = xr.concat([antisense_methylation, antisense_minus_ds], dim=\"ranges\")\n",
    "\n",
    "\n",
    "    return sense_methylation, antisense_methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "sense_ds, antisense_ds = summarise_methylation_across_other_regions(plus_strand, minus_strand, regions_dict_plus, \"+\")\n",
    "sense_ds, antisense_ds = summarise_methylation_across_other_regions(plus_strand, minus_strand, regions_dict_minus, \"-\", sense_ds, antisense_ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;\n",
       "Dimensions:                   (ranges: 128103, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) &lt;U2 &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;\n",
       "    start                     (ranges) int64 4807587 4807691 ... 22859847\n",
       "    end                       (ranges) int64 4807787 4807891 ... 22861212\n",
       "    range_id                  (ranges) int64 0 1 2 3 ... 64152 64153 64154 64155\n",
       "    num_contexts              (ranges) int64 27 35 21 0 7 2 ... 10 12 11 9 11 10\n",
       "    range_length              (ranges) int64 200 200 200 200 ... 1361 1365 1365\n",
       "  * sample_id                 (sample_id) &lt;U8 &#x27;sample_0&#x27;\n",
       "    strand                    (ranges) &lt;U1 &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 2.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 0.05714 ... 0.0\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 27.0 35.0 ... 10.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 1.0 4.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.03704 0.1143 ... 0.0\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 27.0 35.0 ... 10.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 1.0 6.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.03704 0.1714 ... 0.0\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 27.0 35.0 ... 10.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 2.122e+03 ... 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 78.59 82.97 ... 0.0\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 27.0 35.0 ... 10.0\n",
       "    Id                        (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Gene_id                   (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Gene_type                 (ranges) object &#x27;protein_coding&#x27; ... &#x27;protein_c...\n",
       "    Gene_name                 (ranges) object &#x27;Lypla1&#x27; &#x27;Gm37988&#x27; ... &#x27;Gm21366&#x27;\n",
       "    Level                     (ranges) object &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;\n",
       "    Mgi_id                    (ranges) object &#x27;MGI:1344588&#x27; ... &#x27;MGI:5434721&#x27;\n",
       "    Havana_gene               (ranges) object &#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTT...\n",
       "    Tag                       (ranges) object &#x27;overlapping_locus&#x27; ... &#x27;&#x27;\n",
       "    Transcript_id             (ranges) object &#x27;ENSMUST00000027036.10&#x27; ... &#x27;EN...\n",
       "    Region                    (ranges) object &#x27;before_tss&#x27; ... &#x27;first_introns&#x27;\n",
       "    Strand                    (ranges) object &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;\n",
       "    Parent                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_type           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_name           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_number               (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_id                   (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_support_level  (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Havana_transcript         (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Protein_id                (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ccdsid                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ont                       (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Level_1                   (ranges) float64 nan nan nan nan ... 8.0 8.0 8.0\n",
       "    Ranges_id                 (ranges) float64 nan nan ... 2.164e+04 2.164e+04\n",
       "    Intron_number             (ranges) object nan nan nan nan ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;\n",
       "    mean_mc                   (ranges, sample_id) float64 0.0 0.0006887 ... nan\n",
       "    mean_hmc                  (ranges, sample_id) float64 0.0004713 ... nan\n",
       "    mean_modc                 (ranges, sample_id) float64 0.0004713 ... nan</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-3afe208c-80e7-4cf8-bfce-b8ff0f90179b' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-3afe208c-80e7-4cf8-bfce-b8ff0f90179b' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>ranges</span>: 128103</li><li><span class='xr-has-index'>sample_id</span>: 1</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-1313c07e-d338-40a7-b9d5-121e0d3b5e72' class='xr-section-summary-in' type='checkbox'  checked><label for='section-1313c07e-d338-40a7-b9d5-121e0d3b5e72' class='xr-section-summary' >Coordinates: <span>(8)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>contig</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U2</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;</div><input id='attrs-5c6aa753-bebb-4f04-a4ae-77f170e3d391' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5c6aa753-bebb-4f04-a4ae-77f170e3d391' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-356914f9-f800-42b0-9264-60b53855e89a' class='xr-var-data-in' type='checkbox'><label for='data-356914f9-f800-42b0-9264-60b53855e89a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;, ..., &#x27;Y&#x27;, &#x27;Y&#x27;, &#x27;Y&#x27;], dtype=&#x27;&lt;U2&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>start</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4807587 4807691 ... 22859847</div><input id='attrs-28c1ebdc-915f-43c9-bad2-0090a95c9667' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-28c1ebdc-915f-43c9-bad2-0090a95c9667' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8eb2351a-9ffb-4e9f-b924-e550fc54cc99' class='xr-var-data-in' type='checkbox'><label for='data-8eb2351a-9ffb-4e9f-b924-e550fc54cc99' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4807587,  4807691,  4857613, ..., 20992252, 22049944, 22859847])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>end</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4807787 4807891 ... 22861212</div><input id='attrs-1a530a1d-420c-4c00-bc79-e451a075442c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1a530a1d-420c-4c00-bc79-e451a075442c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e084315f-69d3-462f-aafa-91b18349d076' class='xr-var-data-in' type='checkbox'><label for='data-e084315f-69d3-462f-aafa-91b18349d076' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4807787,  4807891,  4857813, ..., 20993613, 22051309, 22861212])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 ... 64152 64153 64154 64155</div><input id='attrs-e48dbc78-1a02-496a-bcbf-76db2c350a9d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e48dbc78-1a02-496a-bcbf-76db2c350a9d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8e535261-6085-426e-8479-996b57606e77' class='xr-var-data-in' type='checkbox'><label for='data-8e535261-6085-426e-8479-996b57606e77' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([    0,     1,     2, ..., 64153, 64154, 64155])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_contexts</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>27 35 21 0 7 2 ... 10 12 11 9 11 10</div><input id='attrs-8ed6ad48-b305-4e00-9dd8-f3c1aa5a3763' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8ed6ad48-b305-4e00-9dd8-f3c1aa5a3763' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-000a1eb1-0926-400a-a5e7-b001d75f78ca' class='xr-var-data-in' type='checkbox'><label for='data-000a1eb1-0926-400a-a5e7-b001d75f78ca' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([27, 35, 21, ...,  9, 11, 10])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_length</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>200 200 200 200 ... 1361 1365 1365</div><input id='attrs-b1cd70cc-6b78-4502-8f98-79a4c2ea056e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b1cd70cc-6b78-4502-8f98-79a4c2ea056e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-75a77ff9-03d5-4b49-ad89-46a1b3ced2dd' class='xr-var-data-in' type='checkbox'><label for='data-75a77ff9-03d5-4b49-ad89-46a1b3ced2dd' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 200,  200,  200, ..., 1361, 1365, 1365])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample_id</span></div><div class='xr-var-dims'>(sample_id)</div><div class='xr-var-dtype'>&lt;U8</div><div class='xr-var-preview xr-preview'>&#x27;sample_0&#x27;</div><input id='attrs-7ef8f47d-d8a6-4b04-8a4b-e17bac8c3e83' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7ef8f47d-d8a6-4b04-8a4b-e17bac8c3e83' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5edde3a2-0113-4ccb-8c90-c5e02d667062' class='xr-var-data-in' type='checkbox'><label for='data-5edde3a2-0113-4ccb-8c90-c5e02d667062' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;sample_0&#x27;], dtype=&#x27;&lt;U8&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U1</div><div class='xr-var-preview xr-preview'>&#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;</div><input id='attrs-02bf22a4-422e-4858-9627-d3cfe8c988d4' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-02bf22a4-422e-4858-9627-d3cfe8c988d4' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0d251cf9-f515-4f7f-beb5-c97975264bea' class='xr-var-data-in' type='checkbox'><label for='data-0d251cf9-f515-4f7f-beb5-c97975264bea' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;+&#x27;, &#x27;+&#x27;, &#x27;+&#x27;, ..., &#x27;-&#x27;, &#x27;-&#x27;, &#x27;-&#x27;], dtype=&#x27;&lt;U1&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-8930f863-05cd-44e7-bd59-96bfa9504d35' class='xr-section-summary-in' type='checkbox'  ><label for='section-8930f863-05cd-44e7-bd59-96bfa9504d35' class='xr-section-summary' >Data variables: <span>(39)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 2.0 4.0 0.0 ... 22.0 0.0 0.0</div><input id='attrs-8ec868a8-7dd0-472c-85d7-d883136f4ace' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8ec868a8-7dd0-472c-85d7-d883136f4ace' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-f03b92f1-6bbd-4b46-afc5-903f47b6d227' class='xr-var-data-in' type='checkbox'><label for='data-f03b92f1-6bbd-4b46-afc5-903f47b6d227' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [ 2.],\n",
       "       [ 4.],\n",
       "       ...,\n",
       "       [22.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.05714 0.1905 ... 0.0 0.0</div><input id='attrs-cbdd17b2-0ac9-4828-b233-9e2118bb37a7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-cbdd17b2-0ac9-4828-b233-9e2118bb37a7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a593bf08-1ae7-4ebc-bc2e-b2f453cb421b' class='xr-var-data-in' type='checkbox'><label for='data-a593bf08-1ae7-4ebc-bc2e-b2f453cb421b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.        ],\n",
       "       [0.05714286],\n",
       "       [0.19047619],\n",
       "       ...,\n",
       "       [2.44444444],\n",
       "       [0.        ],\n",
       "       [0.        ]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 ... 9.0 11.0 10.0</div><input id='attrs-7d007677-9f02-40fa-ae83-40d657d51054' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7d007677-9f02-40fa-ae83-40d657d51054' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4274f206-acae-43bd-b097-b792139cd54e' class='xr-var-data-in' type='checkbox'><label for='data-4274f206-acae-43bd-b097-b792139cd54e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 9.],\n",
       "       [11.],\n",
       "       [10.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 4.0 2.0 0.0 ... 1.0 1.0 0.0 0.0</div><input id='attrs-2596581a-7769-40ec-9aeb-8c2d0891d614' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2596581a-7769-40ec-9aeb-8c2d0891d614' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9530de42-30eb-4e51-8a59-34d61ec4417e' class='xr-var-data-in' type='checkbox'><label for='data-9530de42-30eb-4e51-8a59-34d61ec4417e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [4.],\n",
       "       [2.],\n",
       "       ...,\n",
       "       [1.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.03704 0.1143 0.09524 ... 0.0 0.0</div><input id='attrs-9cdfb2a0-eca4-4d51-bb62-339774c197d0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-9cdfb2a0-eca4-4d51-bb62-339774c197d0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d816a04a-8517-484c-a7e1-8c2784578399' class='xr-var-data-in' type='checkbox'><label for='data-d816a04a-8517-484c-a7e1-8c2784578399' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.03703704],\n",
       "       [0.11428571],\n",
       "       [0.0952381 ],\n",
       "       ...,\n",
       "       [0.11111111],\n",
       "       [0.        ],\n",
       "       [0.        ]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 ... 9.0 11.0 10.0</div><input id='attrs-ecd1951d-51f0-49ed-beb4-a43876d63adc' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ecd1951d-51f0-49ed-beb4-a43876d63adc' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-aa3cf8b5-efd8-4459-98fb-222ab1a17903' class='xr-var-data-in' type='checkbox'><label for='data-aa3cf8b5-efd8-4459-98fb-222ab1a17903' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 9.],\n",
       "       [11.],\n",
       "       [10.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 6.0 6.0 0.0 ... 23.0 0.0 0.0</div><input id='attrs-d91e99d1-d863-4a38-b71f-46c99c8322dd' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d91e99d1-d863-4a38-b71f-46c99c8322dd' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9b65730d-0699-4d5a-840e-19796b353878' class='xr-var-data-in' type='checkbox'><label for='data-9b65730d-0699-4d5a-840e-19796b353878' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 1.],\n",
       "       [ 6.],\n",
       "       [ 6.],\n",
       "       ...,\n",
       "       [23.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.03704 0.1714 0.2857 ... 0.0 0.0</div><input id='attrs-eb7023ce-9fd6-405b-99c0-48dd49c1ee13' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-eb7023ce-9fd6-405b-99c0-48dd49c1ee13' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-cc99ee73-14dd-4f3b-9b19-5048b37fe222' class='xr-var-data-in' type='checkbox'><label for='data-cc99ee73-14dd-4f3b-9b19-5048b37fe222' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.03703704],\n",
       "       [0.17142857],\n",
       "       [0.28571429],\n",
       "       ...,\n",
       "       [2.55555556],\n",
       "       [0.        ],\n",
       "       [0.        ]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 ... 9.0 11.0 10.0</div><input id='attrs-9d2a868d-96fe-4eb5-8a5f-b945efa86ac1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-9d2a868d-96fe-4eb5-8a5f-b945efa86ac1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-6a795f36-fddd-4caa-b8e7-16cd34a80782' class='xr-var-data-in' type='checkbox'><label for='data-6a795f36-fddd-4caa-b8e7-16cd34a80782' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 9.],\n",
       "       [11.],\n",
       "       [10.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>2.122e+03 2.904e+03 ... 0.0 0.0</div><input id='attrs-6970b99a-4e8b-4832-81d4-2bb8f280e29a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6970b99a-4e8b-4832-81d4-2bb8f280e29a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-2042b635-5aca-4af4-8679-db419bcf4863' class='xr-var-data-in' type='checkbox'><label for='data-2042b635-5aca-4af4-8679-db419bcf4863' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[2122.],\n",
       "       [2904.],\n",
       "       [2032.],\n",
       "       ...,\n",
       "       [  32.],\n",
       "       [   0.],\n",
       "       [   0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>78.59 82.97 96.76 ... 3.556 0.0 0.0</div><input id='attrs-6bcb104b-5e89-496a-9d38-7dda08349aad' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6bcb104b-5e89-496a-9d38-7dda08349aad' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-db5a23a6-2627-4c67-9936-bcd968da5dbe' class='xr-var-data-in' type='checkbox'><label for='data-db5a23a6-2627-4c67-9936-bcd968da5dbe' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[78.59259259],\n",
       "       [82.97142857],\n",
       "       [96.76190476],\n",
       "       ...,\n",
       "       [ 3.55555556],\n",
       "       [ 0.        ],\n",
       "       [ 0.        ]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 ... 9.0 11.0 10.0</div><input id='attrs-f7c24272-9b81-44bd-b368-ca74246341d2' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f7c24272-9b81-44bd-b368-ca74246341d2' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-43bb8b99-1813-45e0-a11c-4ccfefd5d6a3' class='xr-var-data-in' type='checkbox'><label for='data-43bb8b99-1813-45e0-a11c-4ccfefd5d6a3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 9.],\n",
       "       [11.],\n",
       "       [10.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-0d2f56e9-befa-42d5-a142-420d1dd2d035' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0d2f56e9-befa-42d5-a142-420d1dd2d035' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-536f40d8-170f-4e92-91a2-a3fde0c64d02' class='xr-var-data-in' type='checkbox'><label for='data-536f40d8-170f-4e92-91a2-a3fde0c64d02' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000104217.1&#x27;,\n",
       "       &#x27;ENSMUSG00000033813.15&#x27;, ..., &#x27;ENSMUSG00000102668.1&#x27;,\n",
       "       &#x27;ENSMUSG00000103528.1&#x27;, &#x27;ENSMUSG00000103919.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-47731439-99a1-49dc-b580-a3a090117c69' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-47731439-99a1-49dc-b580-a3a090117c69' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-fdac5594-d755-47ae-8e6e-a5e0d9f300df' class='xr-var-data-in' type='checkbox'><label for='data-fdac5594-d755-47ae-8e6e-a5e0d9f300df' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000104217.1&#x27;,\n",
       "       &#x27;ENSMUSG00000033813.15&#x27;, ..., &#x27;ENSMUSG00000102668.1&#x27;,\n",
       "       &#x27;ENSMUSG00000103528.1&#x27;, &#x27;ENSMUSG00000103919.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;protein_coding&#x27; ... &#x27;protein_co...</div><input id='attrs-1849c0a6-789b-4ecf-bdc2-44b7ad11897f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1849c0a6-789b-4ecf-bdc2-44b7ad11897f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-01db31af-3046-4e25-9ef2-22682e59046e' class='xr-var-data-in' type='checkbox'><label for='data-01db31af-3046-4e25-9ef2-22682e59046e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, ...,\n",
       "       &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Lypla1&#x27; &#x27;Gm37988&#x27; ... &#x27;Gm21366&#x27;</div><input id='attrs-65c00c54-ab11-4957-9916-310beb8e0f78' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-65c00c54-ab11-4957-9916-310beb8e0f78' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e7b0d8f9-eb47-458d-a264-4d1fe90b25b2' class='xr-var-data-in' type='checkbox'><label for='data-e7b0d8f9-eb47-458d-a264-4d1fe90b25b2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Lypla1&#x27;, &#x27;Gm37988&#x27;, &#x27;Tcea1&#x27;, ..., &#x27;Gm29866&#x27;, &#x27;Gm31571&#x27;, &#x27;Gm21366&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;</div><input id='attrs-ccf09c28-d1ac-4ef6-bbc2-ba80d477a0ce' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ccf09c28-d1ac-4ef6-bbc2-ba80d477a0ce' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-6ae6c9a5-f260-4a8f-b947-b858eb519fc4' class='xr-var-data-in' type='checkbox'><label for='data-6ae6c9a5-f260-4a8f-b947-b858eb519fc4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;, ..., &#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Mgi_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;MGI:1344588&#x27; ... &#x27;MGI:5434721&#x27;</div><input id='attrs-eb8d1cb2-5d23-4b27-a836-241babf02f79' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-eb8d1cb2-5d23-4b27-a836-241babf02f79' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b00fe517-8c2b-414f-82f2-2f43340595ce' class='xr-var-data-in' type='checkbox'><label for='data-b00fe517-8c2b-414f-82f2-2f43340595ce' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;MGI:1344588&#x27;, &#x27;MGI:5611216&#x27;, &#x27;MGI:1196624&#x27;, ..., &#x27;MGI:5589025&#x27;,\n",
       "       &#x27;MGI:5590730&#x27;, &#x27;MGI:5434721&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_gene</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTTM...</div><input id='attrs-2ad070a8-ef7c-42db-b4a2-be2044e47abc' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2ad070a8-ef7c-42db-b4a2-be2044e47abc' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-228e6483-41c8-4dfa-b1ef-54833dff8a25' class='xr-var-data-in' type='checkbox'><label for='data-228e6483-41c8-4dfa-b1ef-54833dff8a25' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;OTTMUSG00000021562.4&#x27;, &#x27;OTTMUSG00000050100.1&#x27;,\n",
       "       &#x27;OTTMUSG00000042348.1&#x27;, ..., &#x27;OTTMUSG00000050180.1&#x27;,\n",
       "       &#x27;OTTMUSG00000050134.1&#x27;, &#x27;OTTMUSG00000050112.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Tag</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;overlapping_locus&#x27; ... &#x27;&#x27;</div><input id='attrs-6eafe468-a7d0-47eb-96dd-4b0e0f389c04' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6eafe468-a7d0-47eb-96dd-4b0e0f389c04' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b5e49048-68ca-4160-a1f7-8e9813c4062e' class='xr-var-data-in' type='checkbox'><label for='data-b5e49048-68ca-4160-a1f7-8e9813c4062e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;overlapping_locus&#x27;, &#x27;overlapping_locus&#x27;, &#x27;overlapping_locus&#x27;, ...,\n",
       "       &#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUST00000027036.10&#x27; ... &#x27;ENS...</div><input id='attrs-09055ad9-a292-4834-b6ee-b8d79c2dc624' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-09055ad9-a292-4834-b6ee-b8d79c2dc624' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ab30fdfc-6e7d-4795-b6ef-8860d14f5d31' class='xr-var-data-in' type='checkbox'><label for='data-ab30fdfc-6e7d-4795-b6ef-8860d14f5d31' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUST00000027036.10&#x27;, &#x27;ENSMUST00000155020.1&#x27;,\n",
       "       &#x27;ENSMUST00000081551.13&#x27;, ..., &#x27;ENSMUST00000191675.1&#x27;,\n",
       "       &#x27;ENSMUST00000192521.1&#x27;, &#x27;ENSMUST00000192892.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Region</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;before_tss&#x27; ... &#x27;first_introns&#x27;</div><input id='attrs-90eb4a68-353d-4449-8749-e01a797e653d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-90eb4a68-353d-4449-8749-e01a797e653d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-fa6c4f1d-4587-45bf-835b-746e1dfb040d' class='xr-var-data-in' type='checkbox'><label for='data-fa6c4f1d-4587-45bf-835b-746e1dfb040d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;before_tss&#x27;, &#x27;before_tss&#x27;, &#x27;before_tss&#x27;, ..., &#x27;first_introns&#x27;,\n",
       "       &#x27;first_introns&#x27;, &#x27;first_introns&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;</div><input id='attrs-5f65ad74-8f8d-4694-9414-72363170009c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5f65ad74-8f8d-4694-9414-72363170009c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-eacafd3e-5d82-4fc0-9f25-3018babba142' class='xr-var-data-in' type='checkbox'><label for='data-eacafd3e-5d82-4fc0-9f25-3018babba142' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;, ..., &#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Parent</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-6c9bead8-6b64-485e-aaa6-d01b27f34f9c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6c9bead8-6b64-485e-aaa6-d01b27f34f9c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-f749fc82-251f-48fa-8cbb-7f11744e8a9f' class='xr-var-data-in' type='checkbox'><label for='data-f749fc82-251f-48fa-8cbb-7f11744e8a9f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-6cc8a66a-12d9-4fd5-b3a6-9ee8e0958dd1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6cc8a66a-12d9-4fd5-b3a6-9ee8e0958dd1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-26e4a710-ca7a-4d5d-9d58-ffbd0812f6b0' class='xr-var-data-in' type='checkbox'><label for='data-26e4a710-ca7a-4d5d-9d58-ffbd0812f6b0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-44a6a426-501b-48be-a8ba-4b4415eaeb26' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-44a6a426-501b-48be-a8ba-4b4415eaeb26' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b9c56926-ab29-4964-923f-492f05a61a07' class='xr-var-data-in' type='checkbox'><label for='data-b9c56926-ab29-4964-923f-492f05a61a07' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-ee4c465f-f3ef-4e44-85f6-041a16b0fd99' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ee4c465f-f3ef-4e44-85f6-041a16b0fd99' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-312db73e-9a73-48b0-aa28-8e3631f5d633' class='xr-var-data-in' type='checkbox'><label for='data-312db73e-9a73-48b0-aa28-8e3631f5d633' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-91034c4a-00f5-4645-be59-b61603eef30e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-91034c4a-00f5-4645-be59-b61603eef30e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-95dfa036-58fd-465e-98e1-e3fc18810e4e' class='xr-var-data-in' type='checkbox'><label for='data-95dfa036-58fd-465e-98e1-e3fc18810e4e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_support_level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-146d08ee-48ce-4acc-9f76-c8bb21a05b28' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-146d08ee-48ce-4acc-9f76-c8bb21a05b28' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c611ea40-54d5-423f-a8cc-51b9f001ae14' class='xr-var-data-in' type='checkbox'><label for='data-c611ea40-54d5-423f-a8cc-51b9f001ae14' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_transcript</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-117e921d-2882-4f4c-9f9e-14ed6597cef9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-117e921d-2882-4f4c-9f9e-14ed6597cef9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-df753ea3-7e74-4db0-aa2b-95bcd491e1d3' class='xr-var-data-in' type='checkbox'><label for='data-df753ea3-7e74-4db0-aa2b-95bcd491e1d3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Protein_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-99c02a90-2138-42bc-8af3-d743cf83a97c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-99c02a90-2138-42bc-8af3-d743cf83a97c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-46c485f8-f645-4e3e-a043-2f27605d1ce7' class='xr-var-data-in' type='checkbox'><label for='data-46c485f8-f645-4e3e-a043-2f27605d1ce7' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ccdsid</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-3154b4e3-d847-49cc-9b49-8fc1359b2dcb' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3154b4e3-d847-49cc-9b49-8fc1359b2dcb' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e26ff0ab-4e60-473d-ae77-b6c963f99f35' class='xr-var-data-in' type='checkbox'><label for='data-e26ff0ab-4e60-473d-ae77-b6c963f99f35' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ont</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-6205a972-986c-4065-ac8e-414274da33f5' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6205a972-986c-4065-ac8e-414274da33f5' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-de4ebe6c-7e3f-492b-993c-6b4a35273866' class='xr-var-data-in' type='checkbox'><label for='data-de4ebe6c-7e3f-492b-993c-6b4a35273866' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level_1</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... 8.0 8.0 8.0 8.0</div><input id='attrs-045ee3c1-9793-4b3d-b680-7993f7fdf24d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-045ee3c1-9793-4b3d-b680-7993f7fdf24d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-61b63f54-5e32-4dc2-84b5-4786f98fc3b9' class='xr-var-data-in' type='checkbox'><label for='data-61b63f54-5e32-4dc2-84b5-4786f98fc3b9' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ...,  8.,  8.,  8.])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ranges_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>nan nan nan ... 2.164e+04 2.164e+04</div><input id='attrs-6c9cad1d-c86d-4375-874f-ccff9ce25b55' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6c9cad1d-c86d-4375-874f-ccff9ce25b55' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3bd7ba24-1489-4ce0-a81f-1af087a9624c' class='xr-var-data-in' type='checkbox'><label for='data-3bd7ba24-1489-4ce0-a81f-1af087a9624c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([   nan,    nan,    nan, ..., 21635., 21636., 21637.])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Intron_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;</div><input id='attrs-05522ef3-aedd-4530-a42c-a409df2e51e3' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-05522ef3-aedd-4530-a42c-a409df2e51e3' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-37b64801-6418-4e45-a19a-8873d7f36503' class='xr-var-data-in' type='checkbox'><label for='data-37b64801-6418-4e45-a19a-8873d7f36503' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., &#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_mc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0006887 0.001969 ... nan nan</div><input id='attrs-441a12ef-4328-4cf9-9d3d-8810829ab38e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-441a12ef-4328-4cf9-9d3d-8810829ab38e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d0022f66-f8e5-40e1-9a25-c49807c087d2' class='xr-var-data-in' type='checkbox'><label for='data-d0022f66-f8e5-40e1-9a25-c49807c087d2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.        ],\n",
       "       [0.00068871],\n",
       "       [0.0019685 ],\n",
       "       ...,\n",
       "       [0.6875    ],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_hmc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0004713 0.001377 ... nan nan</div><input id='attrs-e8cafed8-f8bf-4105-b49f-56ba44ed2b41' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e8cafed8-f8bf-4105-b49f-56ba44ed2b41' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-148e27ac-5ca4-47c0-857e-d83809aea86c' class='xr-var-data-in' type='checkbox'><label for='data-148e27ac-5ca4-47c0-857e-d83809aea86c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.00047125],\n",
       "       [0.00137741],\n",
       "       [0.00098425],\n",
       "       ...,\n",
       "       [0.03125   ],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_modc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0004713 0.002066 ... nan nan</div><input id='attrs-440acdf6-3733-476a-aaca-52dac35449c4' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-440acdf6-3733-476a-aaca-52dac35449c4' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-40bd548e-7b22-44a5-a745-e760fed318bf' class='xr-var-data-in' type='checkbox'><label for='data-40bd548e-7b22-44a5-a745-e760fed318bf' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[4.71253534e-04],\n",
       "       [2.06611570e-03],\n",
       "       [2.95275591e-03],\n",
       "       ...,\n",
       "       [7.18750000e-01],\n",
       "       [           nan],\n",
       "       [           nan]])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-52be7f59-69e3-495f-b0be-2b77a523c37f' class='xr-section-summary-in' type='checkbox'  ><label for='section-52be7f59-69e3-495f-b0be-2b77a523c37f' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>sample_id</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-3c6e413d-1b92-494a-8638-b1a1c8f5d30f' class='xr-index-data-in' type='checkbox'/><label for='index-3c6e413d-1b92-494a-8638-b1a1c8f5d30f' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;sample_0&#x27;], dtype=&#x27;object&#x27;, name=&#x27;sample_id&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-4fcf8867-f48c-4a1d-bd17-2ccd9a21cf95' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-4fcf8867-f48c-4a1d-bd17-2ccd9a21cf95' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.Dataset>\n",
       "Dimensions:                   (ranges: 128103, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) <U2 '1' '1' '1' '1' ... 'Y' 'Y' 'Y' 'Y'\n",
       "    start                     (ranges) int64 4807587 4807691 ... 22859847\n",
       "    end                       (ranges) int64 4807787 4807891 ... 22861212\n",
       "    range_id                  (ranges) int64 0 1 2 3 ... 64152 64153 64154 64155\n",
       "    num_contexts              (ranges) int64 27 35 21 0 7 2 ... 10 12 11 9 11 10\n",
       "    range_length              (ranges) int64 200 200 200 200 ... 1361 1365 1365\n",
       "  * sample_id                 (sample_id) <U8 'sample_0'\n",
       "    strand                    (ranges) <U1 '+' '+' '+' '+' ... '-' '-' '-' '-'\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 2.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 0.05714 ... 0.0\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 27.0 35.0 ... 10.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 1.0 4.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.03704 0.1143 ... 0.0\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 27.0 35.0 ... 10.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 1.0 6.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.03704 0.1714 ... 0.0\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 27.0 35.0 ... 10.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 2.122e+03 ... 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 78.59 82.97 ... 0.0\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 27.0 35.0 ... 10.0\n",
       "    Id                        (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Gene_id                   (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Gene_type                 (ranges) object 'protein_coding' ... 'protein_c...\n",
       "    Gene_name                 (ranges) object 'Lypla1' 'Gm37988' ... 'Gm21366'\n",
       "    Level                     (ranges) object '2' '2' '2' '2' ... '2' '2' '2'\n",
       "    Mgi_id                    (ranges) object 'MGI:1344588' ... 'MGI:5434721'\n",
       "    Havana_gene               (ranges) object 'OTTMUSG00000021562.4' ... 'OTT...\n",
       "    Tag                       (ranges) object 'overlapping_locus' ... ''\n",
       "    Transcript_id             (ranges) object 'ENSMUST00000027036.10' ... 'EN...\n",
       "    Region                    (ranges) object 'before_tss' ... 'first_introns'\n",
       "    Strand                    (ranges) object '.' '.' '.' '.' ... '.' '.' '.'\n",
       "    Parent                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_type           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_name           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_number               (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_id                   (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_support_level  (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Havana_transcript         (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Protein_id                (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ccdsid                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ont                       (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Level_1                   (ranges) float64 nan nan nan nan ... 8.0 8.0 8.0\n",
       "    Ranges_id                 (ranges) float64 nan nan ... 2.164e+04 2.164e+04\n",
       "    Intron_number             (ranges) object nan nan nan nan ... '1' '1' '1'\n",
       "    mean_mc                   (ranges, sample_id) float64 0.0 0.0006887 ... nan\n",
       "    mean_hmc                  (ranges, sample_id) float64 0.0004713 ... nan\n",
       "    mean_modc                 (ranges, sample_id) float64 0.0004713 ... nan"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compute mean methylation levels as mc / total_c\n",
    "sense_ds = sense_ds.assign(\n",
    "    mean_mc = sense_ds[\"num_mc_sum\"] / sense_ds[\"num_total_c_sum\"],\n",
    "    mean_hmc = sense_ds[\"num_hmc_sum\"] / sense_ds[\"num_total_c_sum\"],\n",
    "    mean_modc = sense_ds[\"num_modc_sum\"] / sense_ds[\"num_total_c_sum\"],\n",
    ")\n",
    "sense_ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [],
   "source": [
    "antisense_ds = antisense_ds.assign(\n",
    "    mean_mc = antisense_ds[\"num_mc_sum\"] / antisense_ds[\"num_total_c_sum\"],\n",
    "    mean_hmc = antisense_ds[\"num_hmc_sum\"] / antisense_ds[\"num_total_c_sum\"],\n",
    "    mean_modc = antisense_ds[\"num_modc_sum\"] / antisense_ds[\"num_total_c_sum\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_transcript_genes = sense_ds[\"Transcript_id\"].where(sense_ds[\"Transcript_id\"].isnull(), drop=True)\n",
    "if len(missing_transcript_genes) > 0:\n",
    "    print(\"SENSE Gene IDs with missing Transcript IDs:\")\n",
    "    print(missing_transcript_genes)\n",
    "\n",
    "missing_transcript_genes = antisense_ds[\"Transcript_id\"].where(antisense_ds[\"Transcript_id\"].isnull(), drop=True)\n",
    "if len(missing_transcript_genes) > 0:\n",
    "    print(\"ANTISENSE Gene IDs with missing Transcript IDs:\")\n",
    "    print(missing_transcript_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "exon introns 3\\`UTRs 5\\`UTRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "stitched_sense_regions['Sense'] = \"sense\"\n",
    "stitched_antisense_regions['Sense'] = \"antisense\"\n",
    "stitched_antisense_regions = stitched_antisense_regions.drop([\"range_length\", \"cpg_count\"], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "      <th>Sense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.807980</td>\n",
       "      <td>0.025436</td>\n",
       "      <td>0.847215</td>\n",
       "      <td>102.0</td>\n",
       "      <td>5990.0</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003.15</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.014888</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>14.0</td>\n",
       "      <td>1362.0</td>\n",
       "      <td>Pbsn</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000003.13</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028.15</td>\n",
       "      <td>0.581295</td>\n",
       "      <td>0.031589</td>\n",
       "      <td>0.619697</td>\n",
       "      <td>120.0</td>\n",
       "      <td>3910.0</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000028.13</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000037.17</td>\n",
       "      <td>0.759115</td>\n",
       "      <td>0.017578</td>\n",
       "      <td>0.787109</td>\n",
       "      <td>66.0</td>\n",
       "      <td>5554.0</td>\n",
       "      <td>Scml2</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000101113.8</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000049.11</td>\n",
       "      <td>0.783618</td>\n",
       "      <td>0.036519</td>\n",
       "      <td>0.835495</td>\n",
       "      <td>56.0</td>\n",
       "      <td>2136.0</td>\n",
       "      <td>Apoh</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000049.5</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15461</th>\n",
       "      <td>ENSMUSG00000118219.1</td>\n",
       "      <td>0.424810</td>\n",
       "      <td>0.083937</td>\n",
       "      <td>0.515992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gm29695</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000188175.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15462</th>\n",
       "      <td>ENSMUSG00000118332.1</td>\n",
       "      <td>0.813403</td>\n",
       "      <td>0.045638</td>\n",
       "      <td>0.870884</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fam220a</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000119488.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15463</th>\n",
       "      <td>ENSMUSG00000118346.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tmem179b</td>\n",
       "      <td>19</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000010249.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15464</th>\n",
       "      <td>ENSMUSG00000118537.1</td>\n",
       "      <td>0.766355</td>\n",
       "      <td>0.009346</td>\n",
       "      <td>0.813084</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shld3</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000133280.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15465</th>\n",
       "      <td>ENSMUSG00000118560.1</td>\n",
       "      <td>0.603896</td>\n",
       "      <td>0.077922</td>\n",
       "      <td>0.694805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lim2</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000004732.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>139538 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  \\\n",
       "0       ENSMUSG00000000001.4  0.807980  0.025436   0.847215      102.0   \n",
       "1      ENSMUSG00000000003.15  0.774194  0.014888   0.806452       14.0   \n",
       "2      ENSMUSG00000000028.15  0.581295  0.031589   0.619697      120.0   \n",
       "3      ENSMUSG00000000037.17  0.759115  0.017578   0.787109       66.0   \n",
       "4      ENSMUSG00000000049.11  0.783618  0.036519   0.835495       56.0   \n",
       "...                      ...       ...       ...        ...        ...   \n",
       "15461   ENSMUSG00000118219.1  0.424810  0.083937   0.515992        NaN   \n",
       "15462   ENSMUSG00000118332.1  0.813403  0.045638   0.870884        NaN   \n",
       "15463   ENSMUSG00000118346.1       NaN       NaN        NaN        NaN   \n",
       "15464   ENSMUSG00000118537.1  0.766355  0.009346   0.813084        NaN   \n",
       "15465   ENSMUSG00000118560.1  0.603896  0.077922   0.694805        NaN   \n",
       "\n",
       "       range_length Gene_name contig strand          Transcript_id  \\\n",
       "0            5990.0     Gnai3      3      -   ENSMUST00000000001.4   \n",
       "1            1362.0      Pbsn      X      -  ENSMUST00000000003.13   \n",
       "2            3910.0     Cdc45     16      -  ENSMUST00000000028.13   \n",
       "3            5554.0     Scml2      X      +   ENSMUST00000101113.8   \n",
       "4            2136.0      Apoh     11      +   ENSMUST00000000049.5   \n",
       "...             ...       ...    ...    ...                    ...   \n",
       "15461           NaN   Gm29695      1      -   ENSMUST00000188175.1   \n",
       "15462           NaN   Fam220a      5      -   ENSMUST00000119488.1   \n",
       "15463           NaN  Tmem179b     19      +   ENSMUST00000010249.6   \n",
       "15464           NaN     Shld3     13      +   ENSMUST00000133280.1   \n",
       "15465           NaN      Lim2      7      -   ENSMUST00000004732.6   \n",
       "\n",
       "                 Region      Sense  \n",
       "0                 exons      sense  \n",
       "1                 exons      sense  \n",
       "2                 exons      sense  \n",
       "3                 exons      sense  \n",
       "4                 exons      sense  \n",
       "...                 ...        ...  \n",
       "15461  three_prime_utrs  antisense  \n",
       "15462  three_prime_utrs  antisense  \n",
       "15463  three_prime_utrs  antisense  \n",
       "15464  three_prime_utrs  antisense  \n",
       "15465  three_prime_utrs  antisense  \n",
       "\n",
       "[139538 rows x 12 columns]"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stitched_regions = pd.concat([stitched_sense_regions, stitched_antisense_regions])\n",
    "stitched_regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TSS TES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sense = sense_ds.to_dataframe().reset_index(level=\"sample_id\", drop=True)\n",
    "df_antisense = antisense_ds.to_dataframe().reset_index(level=\"sample_id\", drop=True)\n",
    "\n",
    "df_sense[\"Sense\"] = \"sense\"\n",
    "df_antisense[\"Sense\"] = \"antisense\"\n",
    "df_antisense = df_antisense.drop(['range_length','num_total_c_cpg_count'],axis=1)\n",
    "df = pd.concat([df_sense, df_antisense])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_mc_sum</th>\n",
       "      <th>num_mc_mean</th>\n",
       "      <th>num_mc_cpg_count</th>\n",
       "      <th>num_hmc_sum</th>\n",
       "      <th>num_hmc_mean</th>\n",
       "      <th>num_hmc_cpg_count</th>\n",
       "      <th>num_modc_sum</th>\n",
       "      <th>num_modc_mean</th>\n",
       "      <th>num_modc_cpg_count</th>\n",
       "      <th>num_total_c_sum</th>\n",
       "      <th>...</th>\n",
       "      <th>Ccdsid</th>\n",
       "      <th>Ont</th>\n",
       "      <th>Level_1</th>\n",
       "      <th>Ranges_id</th>\n",
       "      <th>Intron_number</th>\n",
       "      <th>strand</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>Sense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ranges</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.037037</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.037037</td>\n",
       "      <td>27.0</td>\n",
       "      <td>2122.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000471</td>\n",
       "      <td>0.000471</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>35.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.114286</td>\n",
       "      <td>35.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.171429</td>\n",
       "      <td>35.0</td>\n",
       "      <td>2904.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0.000689</td>\n",
       "      <td>0.001377</td>\n",
       "      <td>0.002066</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4.0</td>\n",
       "      <td>0.190476</td>\n",
       "      <td>21.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>21.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>21.0</td>\n",
       "      <td>2032.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.000984</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>136.0</td>\n",
       "      <td>19.428571</td>\n",
       "      <td>7.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>3.285714</td>\n",
       "      <td>7.0</td>\n",
       "      <td>159.0</td>\n",
       "      <td>22.714286</td>\n",
       "      <td>7.0</td>\n",
       "      <td>501.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0.271457</td>\n",
       "      <td>0.045908</td>\n",
       "      <td>0.317365</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128098</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>21666.0</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128099</th>\n",
       "      <td>39.0</td>\n",
       "      <td>3.545455</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>3.545455</td>\n",
       "      <td>11.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>21650.0</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0.951220</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.951220</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128100</th>\n",
       "      <td>22.0</td>\n",
       "      <td>2.444444</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>9.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>2.666667</td>\n",
       "      <td>9.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>21635.0</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0.687500</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128101</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>21636.0</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128102</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>21637.0</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>256206 rows × 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        num_mc_sum  num_mc_mean  num_mc_cpg_count  num_hmc_sum  num_hmc_mean  \\\n",
       "ranges                                                                         \n",
       "0              0.0     0.000000              27.0          1.0      0.037037   \n",
       "1              2.0     0.057143              35.0          4.0      0.114286   \n",
       "2              4.0     0.190476              21.0          2.0      0.095238   \n",
       "3              0.0          NaN               0.0          0.0           NaN   \n",
       "4            136.0    19.428571               7.0         23.0      3.285714   \n",
       "...            ...          ...               ...          ...           ...   \n",
       "128098         0.0     0.000000              12.0          0.0      0.000000   \n",
       "128099        39.0     3.545455              11.0          0.0      0.000000   \n",
       "128100        22.0     2.444444               9.0          1.0      0.111111   \n",
       "128101         0.0     0.000000              11.0          0.0      0.000000   \n",
       "128102         0.0     0.000000              10.0          0.0      0.000000   \n",
       "\n",
       "        num_hmc_cpg_count  num_modc_sum  num_modc_mean  num_modc_cpg_count  \\\n",
       "ranges                                                                       \n",
       "0                    27.0           1.0       0.037037                27.0   \n",
       "1                    35.0           6.0       0.171429                35.0   \n",
       "2                    21.0           6.0       0.285714                21.0   \n",
       "3                     0.0           0.0            NaN                 0.0   \n",
       "4                     7.0         159.0      22.714286                 7.0   \n",
       "...                   ...           ...            ...                 ...   \n",
       "128098               12.0           0.0       0.000000                12.0   \n",
       "128099               11.0          39.0       3.545455                11.0   \n",
       "128100                9.0          24.0       2.666667                 9.0   \n",
       "128101               11.0           0.0       0.000000                11.0   \n",
       "128102               10.0           0.0       0.000000                10.0   \n",
       "\n",
       "        num_total_c_sum  ...  Ccdsid  Ont Level_1  Ranges_id  Intron_number  \\\n",
       "ranges                   ...                                                  \n",
       "0                2122.0  ...     NaN  NaN     NaN        NaN            NaN   \n",
       "1                2904.0  ...     NaN  NaN     NaN        NaN            NaN   \n",
       "2                2032.0  ...     NaN  NaN     NaN        NaN            NaN   \n",
       "3                   0.0  ...     NaN  NaN     NaN        NaN            NaN   \n",
       "4                 501.0  ...     NaN  NaN     NaN        NaN            NaN   \n",
       "...                 ...  ...     ...  ...     ...        ...            ...   \n",
       "128098              0.0  ...     NaN  NaN     8.0    21666.0              1   \n",
       "128099             41.0  ...     NaN  NaN     8.0    21650.0              1   \n",
       "128100             32.0  ...     NaN  NaN     8.0    21635.0              1   \n",
       "128101              0.0  ...     NaN  NaN     8.0    21636.0              1   \n",
       "128102              0.0  ...     NaN  NaN     8.0    21637.0              1   \n",
       "\n",
       "        strand   mean_mc  mean_hmc mean_modc      Sense  \n",
       "ranges                                                   \n",
       "0            +  0.000000  0.000471  0.000471      sense  \n",
       "1            +  0.000689  0.001377  0.002066      sense  \n",
       "2            +  0.001969  0.000984  0.002953      sense  \n",
       "3            +       NaN       NaN       NaN      sense  \n",
       "4            +  0.271457  0.045908  0.317365      sense  \n",
       "...        ...       ...       ...       ...        ...  \n",
       "128098       +       NaN       NaN       NaN  antisense  \n",
       "128099       +  0.951220  0.000000  0.951220  antisense  \n",
       "128100       +  0.687500  0.031250  0.750000  antisense  \n",
       "128101       +       NaN       NaN       NaN  antisense  \n",
       "128102       +       NaN       NaN       NaN  antisense  \n",
       "\n",
       "[256206 rows x 47 columns]"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "      <th>Sense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000471</td>\n",
       "      <td>0.000471</td>\n",
       "      <td>27.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>0.000689</td>\n",
       "      <td>0.001377</td>\n",
       "      <td>0.002066</td>\n",
       "      <td>35.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000155020.1</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.000984</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>21.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Atp6v1h</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000192847.5</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.271457</td>\n",
       "      <td>0.045908</td>\n",
       "      <td>0.317365</td>\n",
       "      <td>7.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395739</th>\n",
       "      <td>ENSMUSG00000118219.1</td>\n",
       "      <td>0.424810</td>\n",
       "      <td>0.083937</td>\n",
       "      <td>0.515992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gm29695</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000188175.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395740</th>\n",
       "      <td>ENSMUSG00000118332.1</td>\n",
       "      <td>0.813403</td>\n",
       "      <td>0.045638</td>\n",
       "      <td>0.870884</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fam220a</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000119488.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395741</th>\n",
       "      <td>ENSMUSG00000118346.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tmem179b</td>\n",
       "      <td>19</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000010249.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395742</th>\n",
       "      <td>ENSMUSG00000118537.1</td>\n",
       "      <td>0.766355</td>\n",
       "      <td>0.009346</td>\n",
       "      <td>0.813084</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shld3</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000133280.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395743</th>\n",
       "      <td>ENSMUSG00000118560.1</td>\n",
       "      <td>0.603896</td>\n",
       "      <td>0.077922</td>\n",
       "      <td>0.694805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lim2</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000004732.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>395744 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  \\\n",
       "0       ENSMUSG00000025903.14  0.000000  0.000471   0.000471       27.0   \n",
       "1        ENSMUSG00000104217.1  0.000689  0.001377   0.002066       35.0   \n",
       "2       ENSMUSG00000033813.15  0.001969  0.000984   0.002953       21.0   \n",
       "3       ENSMUSG00000033793.12       NaN       NaN        NaN        0.0   \n",
       "4       ENSMUSG00000025905.14  0.271457  0.045908   0.317365        7.0   \n",
       "...                       ...       ...       ...        ...        ...   \n",
       "395739   ENSMUSG00000118219.1  0.424810  0.083937   0.515992        NaN   \n",
       "395740   ENSMUSG00000118332.1  0.813403  0.045638   0.870884        NaN   \n",
       "395741   ENSMUSG00000118346.1       NaN       NaN        NaN        NaN   \n",
       "395742   ENSMUSG00000118537.1  0.766355  0.009346   0.813084        NaN   \n",
       "395743   ENSMUSG00000118560.1  0.603896  0.077922   0.694805        NaN   \n",
       "\n",
       "        range_length Gene_name contig strand          Transcript_id  \\\n",
       "0              200.0    Lypla1      1      +  ENSMUST00000027036.10   \n",
       "1              200.0   Gm37988      1      +   ENSMUST00000155020.1   \n",
       "2              200.0     Tcea1      1      +  ENSMUST00000081551.13   \n",
       "3              200.0   Atp6v1h      1      +   ENSMUST00000192847.5   \n",
       "4              200.0     Oprk1      1      +   ENSMUST00000160777.7   \n",
       "...              ...       ...    ...    ...                    ...   \n",
       "395739           NaN   Gm29695      1      -   ENSMUST00000188175.1   \n",
       "395740           NaN   Fam220a      5      -   ENSMUST00000119488.1   \n",
       "395741           NaN  Tmem179b     19      +   ENSMUST00000010249.6   \n",
       "395742           NaN     Shld3     13      +   ENSMUST00000133280.1   \n",
       "395743           NaN      Lim2      7      -   ENSMUST00000004732.6   \n",
       "\n",
       "                  Region      Sense  \n",
       "0             before_tss      sense  \n",
       "1             before_tss      sense  \n",
       "2             before_tss      sense  \n",
       "3             before_tss      sense  \n",
       "4             before_tss      sense  \n",
       "...                  ...        ...  \n",
       "395739  three_prime_utrs  antisense  \n",
       "395740  three_prime_utrs  antisense  \n",
       "395741  three_prime_utrs  antisense  \n",
       "395742  three_prime_utrs  antisense  \n",
       "395743  three_prime_utrs  antisense  \n",
       "\n",
       "[395744 rows x 12 columns]"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"cpg_count\"] = df[\"num_total_c_cpg_count\"]\n",
    "df=df[stitched_regions.columns].reset_index(drop=True)\n",
    "df = pd.concat([df, stitched_regions]).reset_index(drop=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>Region</th>\n",
       "      <th>Sense</th>\n",
       "      <th>range_length</th>\n",
       "      <th>cpg_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>380278</th>\n",
       "      <td>+</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310509</th>\n",
       "      <td>-</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>sense</td>\n",
       "      <td>4108.0</td>\n",
       "      <td>64.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346157</th>\n",
       "      <td>+</td>\n",
       "      <td>introns</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276388</th>\n",
       "      <td>-</td>\n",
       "      <td>introns</td>\n",
       "      <td>sense</td>\n",
       "      <td>27124.0</td>\n",
       "      <td>296.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138946</th>\n",
       "      <td>-</td>\n",
       "      <td>genes</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202835</th>\n",
       "      <td>+</td>\n",
       "      <td>genes</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10843</th>\n",
       "      <td>+</td>\n",
       "      <td>genes</td>\n",
       "      <td>sense</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>434.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74732</th>\n",
       "      <td>-</td>\n",
       "      <td>genes</td>\n",
       "      <td>sense</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>434.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295103</th>\n",
       "      <td>-</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "      <td>sense</td>\n",
       "      <td>280.0</td>\n",
       "      <td>30.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364872</th>\n",
       "      <td>+</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75353</th>\n",
       "      <td>-</td>\n",
       "      <td>first_introns</td>\n",
       "      <td>sense</td>\n",
       "      <td>22051.0</td>\n",
       "      <td>208.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203456</th>\n",
       "      <td>+</td>\n",
       "      <td>first_introns</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75228</th>\n",
       "      <td>-</td>\n",
       "      <td>first_exons</td>\n",
       "      <td>sense</td>\n",
       "      <td>258.0</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203331</th>\n",
       "      <td>+</td>\n",
       "      <td>first_exons</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256206</th>\n",
       "      <td>-</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "      <td>5990.0</td>\n",
       "      <td>102.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325975</th>\n",
       "      <td>+</td>\n",
       "      <td>exons</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201349</th>\n",
       "      <td>+</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73246</th>\n",
       "      <td>-</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "      <td>200.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201845</th>\n",
       "      <td>+</td>\n",
       "      <td>after_tes</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73742</th>\n",
       "      <td>-</td>\n",
       "      <td>after_tes</td>\n",
       "      <td>sense</td>\n",
       "      <td>200.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       strand            Region      Sense  range_length  cpg_count\n",
       "380278      +  three_prime_utrs  antisense           NaN        NaN\n",
       "310509      -  three_prime_utrs      sense        4108.0       64.0\n",
       "346157      +           introns  antisense           NaN        NaN\n",
       "276388      -           introns      sense       27124.0      296.0\n",
       "138946      -             genes  antisense           NaN        NaN\n",
       "202835      +             genes  antisense           NaN        NaN\n",
       "10843       +             genes      sense       38866.0      434.0\n",
       "74732       -             genes      sense       38866.0      434.0\n",
       "295103      -   five_prime_utrs      sense         280.0       30.0\n",
       "364872      +   five_prime_utrs  antisense           NaN        NaN\n",
       "75353       -     first_introns      sense       22051.0      208.0\n",
       "203456      +     first_introns  antisense           NaN        NaN\n",
       "75228       -       first_exons      sense         258.0       27.0\n",
       "203331      +       first_exons  antisense           NaN        NaN\n",
       "256206      -             exons      sense        5990.0      102.0\n",
       "325975      +             exons  antisense           NaN        NaN\n",
       "201349      +        before_tss  antisense           NaN        NaN\n",
       "73246       -        before_tss      sense         200.0       15.0\n",
       "201845      +         after_tes  antisense           NaN        NaN\n",
       "73742       -         after_tes      sense         200.0       15.0"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[\"Gene_id\"]==\"ENSMUSG00000000001.4\"][[\"strand\", \"Region\", \"Sense\", \"range_length\", \"cpg_count\"]].sort_values(by=\"Region\", ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "pivot table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pivot = df.pivot(\n",
    "    index=[\"Gene_id\", \"Gene_name\", \"contig\", \"strand\"],\n",
    "    # index = [\"Transcript_id\", \"contig\", \"strand\"],\n",
    "    columns=[\"Region\", \"Sense\"],\n",
    "    values=[\"mean_mc\", \"mean_hmc\", \"mean_modc\", \"cpg_count\",  \"range_length\"], #\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th colspan=\"10\" halign=\"left\">mean_mc</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">range_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>before_tss</th>\n",
       "      <th>after_tes</th>\n",
       "      <th>genes</th>\n",
       "      <th>first_exons</th>\n",
       "      <th>first_introns</th>\n",
       "      <th>before_tss</th>\n",
       "      <th>after_tes</th>\n",
       "      <th>genes</th>\n",
       "      <th>first_exons</th>\n",
       "      <th>first_introns</th>\n",
       "      <th>...</th>\n",
       "      <th>first_exons</th>\n",
       "      <th>first_introns</th>\n",
       "      <th>exons</th>\n",
       "      <th>introns</th>\n",
       "      <th>five_prime_utrs</th>\n",
       "      <th>three_prime_utrs</th>\n",
       "      <th>exons</th>\n",
       "      <th>introns</th>\n",
       "      <th>five_prime_utrs</th>\n",
       "      <th>three_prime_utrs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000000001.4</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Gnai3</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">3</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.625551</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.670983</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.670983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5990.0</td>\n",
       "      <td>27124.0</td>\n",
       "      <td>280.0</td>\n",
       "      <td>4108.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000000003.15</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Pbsn</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">X</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>0.734848</td>\n",
       "      <td>0.767077</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.748778</td>\n",
       "      <td>0.719298</td>\n",
       "      <td>0.750745</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.748778</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1362.0</td>\n",
       "      <td>19064.0</td>\n",
       "      <td>278.0</td>\n",
       "      <td>470.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.15</th>\n",
       "      <th>Cdc45</th>\n",
       "      <th>16</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.651294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.651294</td>\n",
       "      <td>0.002212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118640.1</th>\n",
       "      <th>AC167036.2</th>\n",
       "      <th>1</th>\n",
       "      <th>-</th>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000118646.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">AC160405.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">10</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>0.829988</td>\n",
       "      <td>0.758004</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.756352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>78.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000118653.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">AC159819.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">9</th>\n",
       "      <th>+</th>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.281619</td>\n",
       "      <td>0.281619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.281619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43082 rows × 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  mean_mc                      \\\n",
       "Region                                         before_tss after_tes     genes   \n",
       "Sense                                               sense     sense     sense   \n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      +             NaN       NaN  0.672555   \n",
       "                                        -        0.000000  0.000000  0.670983   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +             NaN       NaN  0.750896   \n",
       "                                        -             NaN       NaN  0.748778   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +             NaN       NaN  0.651294   \n",
       "...                                                   ...       ...       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -        0.853868  0.853868  0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +             NaN       NaN  0.762000   \n",
       "                                        -        0.833333  0.833333  0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +        0.276243  0.276243  0.281619   \n",
       "                                        -             NaN       NaN  0.300631   \n",
       "\n",
       "                                                                          \\\n",
       "Region                                         first_exons first_introns   \n",
       "Sense                                                sense         sense   \n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +              NaN           NaN   \n",
       "                                        -         0.000455      0.624199   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +              NaN           NaN   \n",
       "                                        -         0.719298      0.750745   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +              NaN           NaN   \n",
       "...                                                    ...           ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -         0.779018           NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +              NaN           NaN   \n",
       "                                        -         0.777778      0.756352   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +         0.281619           NaN   \n",
       "                                        -              NaN           NaN   \n",
       "\n",
       "                                                                               \\\n",
       "Region                                         before_tss after_tes     genes   \n",
       "Sense                                           antisense antisense antisense   \n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      +        0.001415  0.001415  0.672555   \n",
       "                                        -             NaN       NaN  0.670983   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +             NaN       NaN  0.750896   \n",
       "                                        -             NaN       NaN  0.748778   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +        0.000000  0.000000  0.651294   \n",
       "...                                                   ...       ...       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -             NaN       NaN  0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +        0.860000  0.860000  0.762000   \n",
       "                                        -             NaN       NaN  0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +             NaN       NaN  0.281619   \n",
       "                                        -        0.315789  0.315789  0.300631   \n",
       "\n",
       "                                                                          ...  \\\n",
       "Region                                         first_exons first_introns  ...   \n",
       "Sense                                            antisense     antisense  ...   \n",
       "Gene_id               Gene_name  contig strand                            ...   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +         0.000000      0.625551  ...   \n",
       "                                        -              NaN           NaN  ...   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +         0.734848      0.767077  ...   \n",
       "                                        -              NaN           NaN  ...   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +         0.002212           NaN  ...   \n",
       "...                                                    ...           ...  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -              NaN           NaN  ...   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +         0.829988      0.758004  ...   \n",
       "                                        -              NaN           NaN  ...   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +              NaN           NaN  ...   \n",
       "                                        -         0.300631           NaN  ...   \n",
       "\n",
       "                                               range_length                \\\n",
       "Region                                          first_exons first_introns   \n",
       "Sense                                             antisense     antisense   \n",
       "Gene_id               Gene_name  contig strand                              \n",
       "ENSMUSG00000000001.4  Gnai3      3      +               NaN           NaN   \n",
       "                                        -               NaN           NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +               NaN           NaN   \n",
       "                                        -               NaN           NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +               NaN           NaN   \n",
       "...                                                     ...           ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -               NaN           NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +               NaN           NaN   \n",
       "                                        -               NaN           NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +               NaN           NaN   \n",
       "                                        -               NaN           NaN   \n",
       "\n",
       "                                                                 \\\n",
       "Region                                           exons  introns   \n",
       "Sense                                            sense    sense   \n",
       "Gene_id               Gene_name  contig strand                    \n",
       "ENSMUSG00000000001.4  Gnai3      3      +          NaN      NaN   \n",
       "                                        -       5990.0  27124.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +          NaN      NaN   \n",
       "                                        -       1362.0  19064.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +          NaN      NaN   \n",
       "...                                                ...      ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -          NaN      NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +          NaN      NaN   \n",
       "                                        -         78.0      NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +          NaN      NaN   \n",
       "                                        -          NaN      NaN   \n",
       "\n",
       "                                                                \\\n",
       "Region                                         five_prime_utrs   \n",
       "Sense                                                    sense   \n",
       "Gene_id               Gene_name  contig strand                   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                  NaN   \n",
       "                                        -                280.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                  NaN   \n",
       "                                        -                278.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                  NaN   \n",
       "...                                                        ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                  NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                  NaN   \n",
       "                                        -                  NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                  NaN   \n",
       "                                        -                  NaN   \n",
       "\n",
       "                                                                           \\\n",
       "Region                                         three_prime_utrs     exons   \n",
       "Sense                                                     sense antisense   \n",
       "Gene_id               Gene_name  contig strand                              \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                   NaN       NaN   \n",
       "                                        -                4108.0       NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                   NaN       NaN   \n",
       "                                        -                 470.0       NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                   NaN       NaN   \n",
       "...                                                         ...       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                   NaN       NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                   NaN       NaN   \n",
       "                                        -                   NaN       NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                   NaN       NaN   \n",
       "                                        -                   NaN       NaN   \n",
       "\n",
       "                                                                          \\\n",
       "Region                                           introns five_prime_utrs   \n",
       "Sense                                          antisense       antisense   \n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +            NaN             NaN   \n",
       "                                        -            NaN             NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +            NaN             NaN   \n",
       "                                        -            NaN             NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +            NaN             NaN   \n",
       "...                                                  ...             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -            NaN             NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +            NaN             NaN   \n",
       "                                        -            NaN             NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +            NaN             NaN   \n",
       "                                        -            NaN             NaN   \n",
       "\n",
       "                                                                 \n",
       "Region                                         three_prime_utrs  \n",
       "Sense                                                 antisense  \n",
       "Gene_id               Gene_name  contig strand                   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                   NaN  \n",
       "                                        -                   NaN  \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                   NaN  \n",
       "                                        -                   NaN  \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                   NaN  \n",
       "...                                                         ...  \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                   NaN  \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                   NaN  \n",
       "                                        -                   NaN  \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                   NaN  \n",
       "                                        -                   NaN  \n",
       "\n",
       "[43082 rows x 90 columns]"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pivot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>mean_mc_first_exons_sense</th>\n",
       "      <th>mean_mc_first_introns_sense</th>\n",
       "      <th>mean_mc_before_tss_antisense</th>\n",
       "      <th>mean_mc_after_tes_antisense</th>\n",
       "      <th>mean_mc_genes_antisense</th>\n",
       "      <th>mean_mc_first_exons_antisense</th>\n",
       "      <th>mean_mc_first_introns_antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>range_length_first_exons_antisense</th>\n",
       "      <th>range_length_first_introns_antisense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "      <th>range_length_exons_antisense</th>\n",
       "      <th>range_length_introns_antisense</th>\n",
       "      <th>range_length_five_prime_utrs_antisense</th>\n",
       "      <th>range_length_three_prime_utrs_antisense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000000001.4</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Gnai3</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">3</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.625551</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.670983</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.670983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5990.0</td>\n",
       "      <td>27124.0</td>\n",
       "      <td>280.0</td>\n",
       "      <td>4108.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000000003.15</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Pbsn</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">X</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>0.734848</td>\n",
       "      <td>0.767077</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.748778</td>\n",
       "      <td>0.719298</td>\n",
       "      <td>0.750745</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.748778</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1362.0</td>\n",
       "      <td>19064.0</td>\n",
       "      <td>278.0</td>\n",
       "      <td>470.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.15</th>\n",
       "      <th>Cdc45</th>\n",
       "      <th>16</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.651294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.651294</td>\n",
       "      <td>0.002212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118640.1</th>\n",
       "      <th>AC167036.2</th>\n",
       "      <th>1</th>\n",
       "      <th>-</th>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000118646.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">AC160405.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">10</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>0.829988</td>\n",
       "      <td>0.758004</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.756352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>78.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000118653.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">AC159819.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">9</th>\n",
       "      <th>+</th>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.281619</td>\n",
       "      <td>0.281619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.281619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43082 rows × 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                mean_mc_before_tss_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            NaN   \n",
       "                                        -                       0.000000   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            NaN   \n",
       "                                        -                            NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            NaN   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                       0.853868   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            NaN   \n",
       "                                        -                       0.833333   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                       0.276243   \n",
       "                                        -                            NaN   \n",
       "\n",
       "                                                mean_mc_after_tes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                           NaN   \n",
       "                                        -                      0.000000   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                           NaN   \n",
       "                                        -                           NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                           NaN   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.853868   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                           NaN   \n",
       "                                        -                      0.833333   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.276243   \n",
       "                                        -                           NaN   \n",
       "\n",
       "                                                mean_mc_genes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                        \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                  0.672555   \n",
       "                                        -                  0.670983   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                  0.750896   \n",
       "                                        -                  0.748778   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                  0.651294   \n",
       "...                                                             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                  0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                  0.762000   \n",
       "                                        -                  0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                  0.281619   \n",
       "                                        -                  0.300631   \n",
       "\n",
       "                                                mean_mc_first_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                              \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                             NaN   \n",
       "                                        -                        0.000455   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                             NaN   \n",
       "                                        -                        0.719298   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                             NaN   \n",
       "...                                                                   ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                        0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                             NaN   \n",
       "                                        -                        0.777778   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                        0.281619   \n",
       "                                        -                             NaN   \n",
       "\n",
       "                                                mean_mc_first_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                               NaN   \n",
       "                                        -                          0.624199   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                               NaN   \n",
       "                                        -                          0.750745   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                               NaN   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                               NaN   \n",
       "                                        -                          0.756352   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               NaN   \n",
       "                                        -                               NaN   \n",
       "\n",
       "                                                mean_mc_before_tss_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                           0.001415   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                NaN   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                           0.000000   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                           0.860000   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                NaN   \n",
       "                                        -                           0.315789   \n",
       "\n",
       "                                                mean_mc_after_tes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                          0.001415   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                               NaN   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                          0.000000   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                          0.860000   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               NaN   \n",
       "                                        -                          0.315789   \n",
       "\n",
       "                                                mean_mc_genes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                      0.672555   \n",
       "                                        -                      0.670983   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                      0.750896   \n",
       "                                        -                      0.748778   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                      0.651294   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                      0.762000   \n",
       "                                        -                      0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.281619   \n",
       "                                        -                      0.300631   \n",
       "\n",
       "                                                mean_mc_first_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            0.000000   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            0.734848   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            0.002212   \n",
       "...                                                                       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                 NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            0.829988   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                 NaN   \n",
       "                                        -                            0.300631   \n",
       "\n",
       "                                                mean_mc_first_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                    \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                              0.625551   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                              0.767077   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                   NaN   \n",
       "...                                                                         ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                   NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                              0.758004   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                   NaN   \n",
       "                                        -                                   NaN   \n",
       "\n",
       "                                                ...  \\\n",
       "Gene_id               Gene_name  contig strand  ...   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +       ...   \n",
       "...                                             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -       ...   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +       ...   \n",
       "                                        -       ...   \n",
       "\n",
       "                                                range_length_first_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                      NaN   \n",
       "...                                                                            ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                      NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "\n",
       "                                                range_length_first_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                         \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                        NaN   \n",
       "                                        -                                        NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                        NaN   \n",
       "                                        -                                        NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                        NaN   \n",
       "...                                                                              ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                        NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                        NaN   \n",
       "                                        -                                        NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                        NaN   \n",
       "                                        -                                        NaN   \n",
       "\n",
       "                                                range_length_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            NaN   \n",
       "                                        -                         5990.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            NaN   \n",
       "                                        -                         1362.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            NaN   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                            NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            NaN   \n",
       "                                        -                           78.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                            NaN   \n",
       "                                        -                            NaN   \n",
       "\n",
       "                                                range_length_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                               \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                              NaN   \n",
       "                                        -                          27124.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                              NaN   \n",
       "                                        -                          19064.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                              NaN   \n",
       "...                                                                    ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                              NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                              NaN   \n",
       "                                        -                              NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                              NaN   \n",
       "                                        -                              NaN   \n",
       "\n",
       "                                                range_length_five_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                      NaN   \n",
       "                                        -                                    280.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                      NaN   \n",
       "                                        -                                    278.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                      NaN   \n",
       "...                                                                            ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                      NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "\n",
       "                                                range_length_three_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                        \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                       NaN   \n",
       "                                        -                                    4108.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                       NaN   \n",
       "                                        -                                     470.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                       NaN   \n",
       "...                                                                             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                       NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                       NaN   \n",
       "                                        -                                       NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                       NaN   \n",
       "                                        -                                       NaN   \n",
       "\n",
       "                                                range_length_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                NaN   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                NaN   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                NaN   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                NaN   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                NaN   \n",
       "                                        -                                NaN   \n",
       "\n",
       "                                                range_length_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                  NaN   \n",
       "                                        -                                  NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                  NaN   \n",
       "                                        -                                  NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                  NaN   \n",
       "...                                                                        ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                  NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                  NaN   \n",
       "                                        -                                  NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                  NaN   \n",
       "                                        -                                  NaN   \n",
       "\n",
       "                                                range_length_five_prime_utrs_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                           \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                          NaN   \n",
       "                                        -                                          NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                          NaN   \n",
       "                                        -                                          NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                          NaN   \n",
       "...                                                                                ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                          NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                          NaN   \n",
       "                                        -                                          NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                          NaN   \n",
       "                                        -                                          NaN   \n",
       "\n",
       "                                                range_length_three_prime_utrs_antisense  \n",
       "Gene_id               Gene_name  contig strand                                           \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                           NaN  \n",
       "                                        -                                           NaN  \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                           NaN  \n",
       "                                        -                                           NaN  \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                           NaN  \n",
       "...                                                                                 ...  \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                           NaN  \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                           NaN  \n",
       "                                        -                                           NaN  \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                           NaN  \n",
       "                                        -                                           NaN  \n",
       "\n",
       "[43082 rows x 90 columns]"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compost feature names\n",
    "df_features = df_pivot.copy()\n",
    "df_features.columns = [\" \".join(col).strip() for col in df_features.columns.values]\n",
    "features = df_features.columns\n",
    "features = [f.replace(\" \", \"_\") for f in features]\n",
    "df_features.columns = features\n",
    "df_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>mean_mc_first_exons_sense</th>\n",
       "      <th>mean_mc_first_introns_sense</th>\n",
       "      <th>mean_mc_before_tss_antisense</th>\n",
       "      <th>mean_mc_after_tes_antisense</th>\n",
       "      <th>mean_mc_genes_antisense</th>\n",
       "      <th>mean_mc_first_exons_antisense</th>\n",
       "      <th>mean_mc_first_introns_antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>cpg_count_three_prime_utrs_sense</th>\n",
       "      <th>range_length_before_tss_sense</th>\n",
       "      <th>range_length_after_tes_sense</th>\n",
       "      <th>range_length_genes_sense</th>\n",
       "      <th>range_length_first_exons_sense</th>\n",
       "      <th>range_length_first_introns_sense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000000001.4</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Gnai3</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">3</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.625551</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.670983</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.670983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>64.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>22051.0</td>\n",
       "      <td>5990.0</td>\n",
       "      <td>27124.0</td>\n",
       "      <td>280.0</td>\n",
       "      <td>4108.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000000003.15</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Pbsn</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">X</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>0.734848</td>\n",
       "      <td>0.767077</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15722.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.748778</td>\n",
       "      <td>0.719298</td>\n",
       "      <td>0.750745</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.748778</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>15722.0</td>\n",
       "      <td>214.0</td>\n",
       "      <td>5295.0</td>\n",
       "      <td>1362.0</td>\n",
       "      <td>19064.0</td>\n",
       "      <td>278.0</td>\n",
       "      <td>470.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.15</th>\n",
       "      <th>Cdc45</th>\n",
       "      <th>16</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.651294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.651294</td>\n",
       "      <td>0.002212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31540.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118640.1</th>\n",
       "      <th>AC167036.2</th>\n",
       "      <th>1</th>\n",
       "      <th>-</th>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>890.0</td>\n",
       "      <td>890.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000118646.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">AC160405.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">10</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>0.829988</td>\n",
       "      <td>0.758004</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19585.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.756352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>19585.0</td>\n",
       "      <td>342.0</td>\n",
       "      <td>19204.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000118653.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">AC159819.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">9</th>\n",
       "      <th>+</th>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.281619</td>\n",
       "      <td>0.281619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.281619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>987.0</td>\n",
       "      <td>987.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>987.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43082 rows × 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                mean_mc_before_tss_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            NaN   \n",
       "                                        -                       0.000000   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            NaN   \n",
       "                                        -                            NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            NaN   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                       0.853868   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            NaN   \n",
       "                                        -                       0.833333   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                       0.276243   \n",
       "                                        -                            NaN   \n",
       "\n",
       "                                                mean_mc_after_tes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                           NaN   \n",
       "                                        -                      0.000000   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                           NaN   \n",
       "                                        -                           NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                           NaN   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.853868   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                           NaN   \n",
       "                                        -                      0.833333   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.276243   \n",
       "                                        -                           NaN   \n",
       "\n",
       "                                                mean_mc_genes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                        \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                  0.672555   \n",
       "                                        -                  0.670983   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                  0.750896   \n",
       "                                        -                  0.748778   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                  0.651294   \n",
       "...                                                             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                  0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                  0.762000   \n",
       "                                        -                  0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                  0.281619   \n",
       "                                        -                  0.300631   \n",
       "\n",
       "                                                mean_mc_first_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                              \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                             NaN   \n",
       "                                        -                        0.000455   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                             NaN   \n",
       "                                        -                        0.719298   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                             NaN   \n",
       "...                                                                   ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                        0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                             NaN   \n",
       "                                        -                        0.777778   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                        0.281619   \n",
       "                                        -                             NaN   \n",
       "\n",
       "                                                mean_mc_first_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                               NaN   \n",
       "                                        -                          0.624199   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                               NaN   \n",
       "                                        -                          0.750745   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                               NaN   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                               NaN   \n",
       "                                        -                          0.756352   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               NaN   \n",
       "                                        -                               NaN   \n",
       "\n",
       "                                                mean_mc_before_tss_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                           0.001415   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                NaN   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                           0.000000   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                           0.860000   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                NaN   \n",
       "                                        -                           0.315789   \n",
       "\n",
       "                                                mean_mc_after_tes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                          0.001415   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                               NaN   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                          0.000000   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                          0.860000   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               NaN   \n",
       "                                        -                          0.315789   \n",
       "\n",
       "                                                mean_mc_genes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                      0.672555   \n",
       "                                        -                      0.670983   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                      0.750896   \n",
       "                                        -                      0.748778   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                      0.651294   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                      0.762000   \n",
       "                                        -                      0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.281619   \n",
       "                                        -                      0.300631   \n",
       "\n",
       "                                                mean_mc_first_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            0.000000   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            0.734848   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            0.002212   \n",
       "...                                                                       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                 NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            0.829988   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                 NaN   \n",
       "                                        -                            0.300631   \n",
       "\n",
       "                                                mean_mc_first_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                    \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                              0.625551   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                              0.767077   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                   NaN   \n",
       "...                                                                         ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                   NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                              0.758004   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                   NaN   \n",
       "                                        -                                   NaN   \n",
       "\n",
       "                                                ...  \\\n",
       "Gene_id               Gene_name  contig strand  ...   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +       ...   \n",
       "...                                             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -       ...   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +       ...   \n",
       "                                        -       ...   \n",
       "\n",
       "                                                cpg_count_three_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                     \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                    NaN   \n",
       "                                        -                                   64.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                    NaN   \n",
       "                                        -                                    2.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                    NaN   \n",
       "...                                                                          ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                    NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                    NaN   \n",
       "                                        -                                    NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                    NaN   \n",
       "                                        -                                    NaN   \n",
       "\n",
       "                                                range_length_before_tss_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                 NaN   \n",
       "                                        -                               200.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                 NaN   \n",
       "                                        -                               200.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                 NaN   \n",
       "...                                                                       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               200.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                 NaN   \n",
       "                                        -                               200.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               200.0   \n",
       "                                        -                                 NaN   \n",
       "\n",
       "                                                range_length_after_tes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                NaN   \n",
       "                                        -                              200.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                NaN   \n",
       "                                        -                              200.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                NaN   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                              200.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                NaN   \n",
       "                                        -                              200.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                              200.0   \n",
       "                                        -                                NaN   \n",
       "\n",
       "                                                range_length_genes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                        38866.0   \n",
       "                                        -                        38866.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                        15722.0   \n",
       "                                        -                        15722.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                        31540.0   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                          890.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                        19585.0   \n",
       "                                        -                        19585.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                          987.0   \n",
       "                                        -                          987.0   \n",
       "\n",
       "                                                range_length_first_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                  NaN   \n",
       "                                        -                                258.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                  NaN   \n",
       "                                        -                                214.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                  NaN   \n",
       "...                                                                        ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                890.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                  NaN   \n",
       "                                        -                                342.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                987.0   \n",
       "                                        -                                  NaN   \n",
       "\n",
       "                                                range_length_first_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                     \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                    NaN   \n",
       "                                        -                                22051.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                    NaN   \n",
       "                                        -                                 5295.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                    NaN   \n",
       "...                                                                          ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                    NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                    NaN   \n",
       "                                        -                                19204.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                    NaN   \n",
       "                                        -                                    NaN   \n",
       "\n",
       "                                                range_length_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            NaN   \n",
       "                                        -                         5990.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            NaN   \n",
       "                                        -                         1362.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            NaN   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                            NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            NaN   \n",
       "                                        -                           78.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                            NaN   \n",
       "                                        -                            NaN   \n",
       "\n",
       "                                                range_length_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                               \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                              NaN   \n",
       "                                        -                          27124.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                              NaN   \n",
       "                                        -                          19064.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                              NaN   \n",
       "...                                                                    ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                              NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                              NaN   \n",
       "                                        -                              NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                              NaN   \n",
       "                                        -                              NaN   \n",
       "\n",
       "                                                range_length_five_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                      NaN   \n",
       "                                        -                                    280.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                      NaN   \n",
       "                                        -                                    278.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                      NaN   \n",
       "...                                                                            ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                      NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "\n",
       "                                                range_length_three_prime_utrs_sense  \n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                       NaN  \n",
       "                                        -                                    4108.0  \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                       NaN  \n",
       "                                        -                                     470.0  \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                       NaN  \n",
       "...                                                                             ...  \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                       NaN  \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                       NaN  \n",
       "                                        -                                       NaN  \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                       NaN  \n",
       "                                        -                                       NaN  \n",
       "\n",
       "[43082 rows x 72 columns]"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features_cleaned = df_features.dropna(axis=1, how='all')\n",
    "\n",
    "# Display the cleaned DataFrame without columns that have all NaN values\n",
    "df_features_cleaned\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>mean_mc_first_exons_sense</th>\n",
       "      <th>mean_mc_first_introns_sense</th>\n",
       "      <th>mean_mc_before_tss_antisense</th>\n",
       "      <th>mean_mc_after_tes_antisense</th>\n",
       "      <th>mean_mc_genes_antisense</th>\n",
       "      <th>mean_mc_first_exons_antisense</th>\n",
       "      <th>mean_mc_first_introns_antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>range_length_first_exons_antisense</th>\n",
       "      <th>range_length_first_introns_antisense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "      <th>range_length_exons_antisense</th>\n",
       "      <th>range_length_introns_antisense</th>\n",
       "      <th>range_length_five_prime_utrs_antisense</th>\n",
       "      <th>range_length_three_prime_utrs_antisense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Gnai3</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">3</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.625551</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.670983</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.670983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5990.0</td>\n",
       "      <td>27124.0</td>\n",
       "      <td>280.0</td>\n",
       "      <td>4108.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         mean_mc_before_tss_sense  mean_mc_after_tes_sense  \\\n",
       "Gene_name contig strand                                                      \n",
       "Gnai3     3      +                            NaN                      NaN   \n",
       "                 -                            0.0                      0.0   \n",
       "\n",
       "                         mean_mc_genes_sense  mean_mc_first_exons_sense  \\\n",
       "Gene_name contig strand                                                   \n",
       "Gnai3     3      +                  0.672555                        NaN   \n",
       "                 -                  0.670983                   0.000455   \n",
       "\n",
       "                         mean_mc_first_introns_sense  \\\n",
       "Gene_name contig strand                                \n",
       "Gnai3     3      +                               NaN   \n",
       "                 -                          0.624199   \n",
       "\n",
       "                         mean_mc_before_tss_antisense  \\\n",
       "Gene_name contig strand                                 \n",
       "Gnai3     3      +                           0.001415   \n",
       "                 -                                NaN   \n",
       "\n",
       "                         mean_mc_after_tes_antisense  mean_mc_genes_antisense  \\\n",
       "Gene_name contig strand                                                         \n",
       "Gnai3     3      +                          0.001415                 0.672555   \n",
       "                 -                               NaN                 0.670983   \n",
       "\n",
       "                         mean_mc_first_exons_antisense  \\\n",
       "Gene_name contig strand                                  \n",
       "Gnai3     3      +                                 0.0   \n",
       "                 -                                 NaN   \n",
       "\n",
       "                         mean_mc_first_introns_antisense  ...  \\\n",
       "Gene_name contig strand                                   ...   \n",
       "Gnai3     3      +                              0.625551  ...   \n",
       "                 -                                   NaN  ...   \n",
       "\n",
       "                         range_length_first_exons_antisense  \\\n",
       "Gene_name contig strand                                       \n",
       "Gnai3     3      +                                      NaN   \n",
       "                 -                                      NaN   \n",
       "\n",
       "                         range_length_first_introns_antisense  \\\n",
       "Gene_name contig strand                                         \n",
       "Gnai3     3      +                                        NaN   \n",
       "                 -                                        NaN   \n",
       "\n",
       "                         range_length_exons_sense  range_length_introns_sense  \\\n",
       "Gene_name contig strand                                                         \n",
       "Gnai3     3      +                            NaN                         NaN   \n",
       "                 -                         5990.0                     27124.0   \n",
       "\n",
       "                         range_length_five_prime_utrs_sense  \\\n",
       "Gene_name contig strand                                       \n",
       "Gnai3     3      +                                      NaN   \n",
       "                 -                                    280.0   \n",
       "\n",
       "                         range_length_three_prime_utrs_sense  \\\n",
       "Gene_name contig strand                                        \n",
       "Gnai3     3      +                                       NaN   \n",
       "                 -                                    4108.0   \n",
       "\n",
       "                         range_length_exons_antisense  \\\n",
       "Gene_name contig strand                                 \n",
       "Gnai3     3      +                                NaN   \n",
       "                 -                                NaN   \n",
       "\n",
       "                         range_length_introns_antisense  \\\n",
       "Gene_name contig strand                                   \n",
       "Gnai3     3      +                                  NaN   \n",
       "                 -                                  NaN   \n",
       "\n",
       "                         range_length_five_prime_utrs_antisense  \\\n",
       "Gene_name contig strand                                           \n",
       "Gnai3     3      +                                          NaN   \n",
       "                 -                                          NaN   \n",
       "\n",
       "                         range_length_three_prime_utrs_antisense  \n",
       "Gene_name contig strand                                           \n",
       "Gnai3     3      +                                           NaN  \n",
       "                 -                                           NaN  \n",
       "\n",
       "[2 rows x 90 columns]"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features_cleaned.loc[\"ENSMUSG00000000001.4\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### write to file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we write this feature file as a pickle file, which preserves the variable type, and is therefore prefered over a more basic text file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Gene_id', 'Gene_name', 'contig', 'strand', 'mean_mc_before_tss_sense',\n",
       "       'mean_mc_after_tes_sense', 'mean_mc_genes_sense',\n",
       "       'mean_mc_first_exons_sense', 'mean_mc_first_introns_sense',\n",
       "       'mean_mc_before_tss_antisense', 'mean_mc_after_tes_antisense',\n",
       "       'mean_mc_genes_antisense', 'mean_mc_first_exons_antisense',\n",
       "       'mean_mc_first_introns_antisense', 'mean_mc_exons_sense',\n",
       "       'mean_mc_introns_sense', 'mean_mc_five_prime_utrs_sense',\n",
       "       'mean_mc_three_prime_utrs_sense', 'mean_mc_exons_antisense',\n",
       "       'mean_mc_introns_antisense', 'mean_mc_five_prime_utrs_antisense',\n",
       "       'mean_mc_three_prime_utrs_antisense', 'mean_hmc_before_tss_sense',\n",
       "       'mean_hmc_after_tes_sense', 'mean_hmc_genes_sense',\n",
       "       'mean_hmc_first_exons_sense', 'mean_hmc_first_introns_sense',\n",
       "       'mean_hmc_before_tss_antisense', 'mean_hmc_after_tes_antisense',\n",
       "       'mean_hmc_genes_antisense', 'mean_hmc_first_exons_antisense',\n",
       "       'mean_hmc_first_introns_antisense', 'mean_hmc_exons_sense',\n",
       "       'mean_hmc_introns_sense', 'mean_hmc_five_prime_utrs_sense',\n",
       "       'mean_hmc_three_prime_utrs_sense', 'mean_hmc_exons_antisense',\n",
       "       'mean_hmc_introns_antisense', 'mean_hmc_five_prime_utrs_antisense',\n",
       "       'mean_hmc_three_prime_utrs_antisense', 'mean_modc_before_tss_sense',\n",
       "       'mean_modc_after_tes_sense', 'mean_modc_genes_sense',\n",
       "       'mean_modc_first_exons_sense', 'mean_modc_first_introns_sense',\n",
       "       'mean_modc_before_tss_antisense', 'mean_modc_after_tes_antisense',\n",
       "       'mean_modc_genes_antisense', 'mean_modc_first_exons_antisense',\n",
       "       'mean_modc_first_introns_antisense', 'mean_modc_exons_sense',\n",
       "       'mean_modc_introns_sense', 'mean_modc_five_prime_utrs_sense',\n",
       "       'mean_modc_three_prime_utrs_sense', 'mean_modc_exons_antisense',\n",
       "       'mean_modc_introns_antisense', 'mean_modc_five_prime_utrs_antisense',\n",
       "       'mean_modc_three_prime_utrs_antisense', 'cpg_count_before_tss_sense',\n",
       "       'cpg_count_after_tes_sense', 'cpg_count_genes_sense',\n",
       "       'cpg_count_first_exons_sense', 'cpg_count_first_introns_sense',\n",
       "       'cpg_count_exons_sense', 'cpg_count_introns_sense',\n",
       "       'cpg_count_five_prime_utrs_sense', 'cpg_count_three_prime_utrs_sense',\n",
       "       'range_length_before_tss_sense', 'range_length_after_tes_sense',\n",
       "       'range_length_genes_sense', 'range_length_first_exons_sense',\n",
       "       'range_length_first_introns_sense', 'range_length_exons_sense',\n",
       "       'range_length_introns_sense', 'range_length_five_prime_utrs_sense',\n",
       "       'range_length_three_prime_utrs_sense', 'selected_transcript'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features_cleaned = df_features_cleaned.reset_index()\n",
    "# add a column with a boolean depending on whether the transcript is is selected_transcripts or not\n",
    "df_features_cleaned[\"selected_transcript\"] = df_features_cleaned.Gene_id.isin(selected_transcripts.gene_id)\n",
    "df_features_cleaned.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Source', 'Type', 'Start', 'End', 'Score', 'Strand',\n",
       "       'Phase', 'Id', 'Gene_id', 'Gene_type', 'Gene_name', 'Level', 'Mgi_id',\n",
       "       'Havana_gene', 'Tag', 'Ranges_ID'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>...</th>\n",
       "      <th>range_length_before_tss_sense</th>\n",
       "      <th>range_length_after_tes_sense</th>\n",
       "      <th>range_length_genes_sense</th>\n",
       "      <th>range_length_first_exons_sense</th>\n",
       "      <th>range_length_first_introns_sense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "      <th>selected_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>1</td>\n",
       "      <td>4807787</td>\n",
       "      <td>4848409</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.529368</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>40622.0</td>\n",
       "      <td>159.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>4678.0</td>\n",
       "      <td>76178.0</td>\n",
       "      <td>180.0</td>\n",
       "      <td>3444.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>1</td>\n",
       "      <td>4807787</td>\n",
       "      <td>4848409</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.538405</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40622.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4807891</td>\n",
       "      <td>4886769</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0.000689</td>\n",
       "      <td>0.000689</td>\n",
       "      <td>0.472305</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>78878.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>20602.0</td>\n",
       "      <td>1450.0</td>\n",
       "      <td>114922.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4807891</td>\n",
       "      <td>4886769</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.483284</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>78878.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>1</td>\n",
       "      <td>4857813</td>\n",
       "      <td>4897908</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.514725</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>40095.0</td>\n",
       "      <td>162.0</td>\n",
       "      <td>9494.0</td>\n",
       "      <td>4750.0</td>\n",
       "      <td>56128.0</td>\n",
       "      <td>198.0</td>\n",
       "      <td>3080.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43209</th>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>Y</td>\n",
       "      <td>84562571</td>\n",
       "      <td>84564906</td>\n",
       "      <td>Gm21394</td>\n",
       "      <td>Y</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.683544</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>2335.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>1114.0</td>\n",
       "      <td>2206.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>796.0</td>\n",
       "      <td>276.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43210</th>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>Y</td>\n",
       "      <td>85528516</td>\n",
       "      <td>85530907</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>Y</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.606061</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2391.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43211</th>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>Y</td>\n",
       "      <td>85528516</td>\n",
       "      <td>85530907</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>Y</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.642857</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>2391.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>1231.0</td>\n",
       "      <td>2202.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43212</th>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>89713423</td>\n",
       "      <td>89745531</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>Y</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.734607</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32108.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43213</th>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>89713423</td>\n",
       "      <td>89745531</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>Y</td>\n",
       "      <td>-</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.773050</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>32108.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>1364.0</td>\n",
       "      <td>1756.0</td>\n",
       "      <td>59646.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>406.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43214 rows × 80 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_id Chromosome     Start       End Gene_name contig  \\\n",
       "0      ENSMUSG00000025903.14          1   4807787   4848409    Lypla1      1   \n",
       "1      ENSMUSG00000025903.14          1   4807787   4848409    Lypla1      1   \n",
       "2       ENSMUSG00000104217.1          1   4807891   4886769   Gm37988      1   \n",
       "3       ENSMUSG00000104217.1          1   4807891   4886769   Gm37988      1   \n",
       "4      ENSMUSG00000033813.15          1   4857813   4897908     Tcea1      1   \n",
       "...                      ...        ...       ...       ...       ...    ...   \n",
       "43209   ENSMUSG00000094660.2          Y  84562571  84564906   Gm21394      Y   \n",
       "43210   ENSMUSG00000095650.2          Y  85528516  85530907   Gm20854      Y   \n",
       "43211   ENSMUSG00000095650.2          Y  85528516  85530907   Gm20854      Y   \n",
       "43212   ENSMUSG00000100608.1          Y  89713423  89745531   Gm21996      Y   \n",
       "43213   ENSMUSG00000100608.1          Y  89713423  89745531   Gm21996      Y   \n",
       "\n",
       "      strand  mean_mc_before_tss_sense  mean_mc_after_tes_sense  \\\n",
       "0          +                  0.000000                 0.000000   \n",
       "1          -                       NaN                      NaN   \n",
       "2          +                  0.000689                 0.000689   \n",
       "3          -                       NaN                      NaN   \n",
       "4          +                  0.001969                 0.001969   \n",
       "...      ...                       ...                      ...   \n",
       "43209      -                       NaN                      NaN   \n",
       "43210      +                       NaN                      NaN   \n",
       "43211      -                       NaN                      NaN   \n",
       "43212      +                       NaN                      NaN   \n",
       "43213      -                  0.800000                 0.800000   \n",
       "\n",
       "       mean_mc_genes_sense  ...  range_length_before_tss_sense  \\\n",
       "0                 0.529368  ...                          200.0   \n",
       "1                 0.538405  ...                            NaN   \n",
       "2                 0.472305  ...                          200.0   \n",
       "3                 0.483284  ...                            NaN   \n",
       "4                 0.514725  ...                          200.0   \n",
       "...                    ...  ...                            ...   \n",
       "43209             0.683544  ...                          200.0   \n",
       "43210             0.606061  ...                            NaN   \n",
       "43211             0.642857  ...                          200.0   \n",
       "43212             0.734607  ...                            NaN   \n",
       "43213             0.773050  ...                          200.0   \n",
       "\n",
       "       range_length_after_tes_sense  range_length_genes_sense  \\\n",
       "0                             200.0                   40622.0   \n",
       "1                               NaN                   40622.0   \n",
       "2                             200.0                   78878.0   \n",
       "3                               NaN                   78878.0   \n",
       "4                             200.0                   40095.0   \n",
       "...                             ...                       ...   \n",
       "43209                         200.0                    2335.0   \n",
       "43210                           NaN                    2391.0   \n",
       "43211                         200.0                    2391.0   \n",
       "43212                           NaN                   32108.0   \n",
       "43213                         200.0                   32108.0   \n",
       "\n",
       "       range_length_first_exons_sense  range_length_first_introns_sense  \\\n",
       "0                               159.0                              35.0   \n",
       "1                                 NaN                               NaN   \n",
       "2                                90.0                           20602.0   \n",
       "3                                 NaN                               NaN   \n",
       "4                               162.0                            9494.0   \n",
       "...                               ...                               ...   \n",
       "43209                           118.0                            1114.0   \n",
       "43210                             NaN                               NaN   \n",
       "43211                            53.0                            1231.0   \n",
       "43212                             NaN                               NaN   \n",
       "43213                            43.0                            1364.0   \n",
       "\n",
       "       range_length_exons_sense  range_length_introns_sense  \\\n",
       "0                        4678.0                     76178.0   \n",
       "1                           NaN                         NaN   \n",
       "2                        1450.0                    114922.0   \n",
       "3                           NaN                         NaN   \n",
       "4                        4750.0                     56128.0   \n",
       "...                         ...                         ...   \n",
       "43209                    2206.0                         NaN   \n",
       "43210                       NaN                         NaN   \n",
       "43211                    2202.0                        12.0   \n",
       "43212                       NaN                         NaN   \n",
       "43213                    1756.0                     59646.0   \n",
       "\n",
       "       range_length_five_prime_utrs_sense  \\\n",
       "0                                   180.0   \n",
       "1                                     NaN   \n",
       "2                                     NaN   \n",
       "3                                     NaN   \n",
       "4                                   198.0   \n",
       "...                                   ...   \n",
       "43209                               796.0   \n",
       "43210                                 NaN   \n",
       "43211                                 NaN   \n",
       "43212                                 NaN   \n",
       "43213                               112.0   \n",
       "\n",
       "       range_length_three_prime_utrs_sense  selected_transcript  \n",
       "0                                   3444.0                 True  \n",
       "1                                      NaN                 True  \n",
       "2                                      NaN                 True  \n",
       "3                                      NaN                 True  \n",
       "4                                   3080.0                 True  \n",
       "...                                    ...                  ...  \n",
       "43209                                276.0                 True  \n",
       "43210                                  NaN                 True  \n",
       "43211                                  NaN                 True  \n",
       "43212                                  NaN                 True  \n",
       "43213                                406.0                 True  \n",
       "\n",
       "[43214 rows x 80 columns]"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features_cleaned = genes.df[[\"Gene_id\", \"Chromosome\", \"Start\", \"End\"]].merge(\n",
    "    df_features_cleaned, on=\"Gene_id\", how=\"left\",\n",
    "    )\n",
    "df_features_cleaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['range_length_before_tss_sense',\n",
       " 'range_length_after_tes_sense',\n",
       " 'range_length_genes_sense',\n",
       " 'range_length_first_exons_sense',\n",
       " 'range_length_first_introns_sense',\n",
       " 'range_length_before_tss_antisense',\n",
       " 'range_length_after_tes_antisense',\n",
       " 'range_length_genes_antisense',\n",
       " 'range_length_first_exons_antisense',\n",
       " 'range_length_first_introns_antisense',\n",
       " 'range_length_exons_sense',\n",
       " 'range_length_introns_sense',\n",
       " 'range_length_five_prime_utrs_sense',\n",
       " 'range_length_three_prime_utrs_sense',\n",
       " 'range_length_exons_antisense',\n",
       " 'range_length_introns_antisense',\n",
       " 'range_length_five_prime_utrs_antisense',\n",
       " 'range_length_three_prime_utrs_antisense']"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "length_col = [x for x in df_features.columns if x.startswith(\"range_length\")]\n",
    "length_col\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_features_cleaned.to_pickle(\"rerun_CEGXRun1485.pickle\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "modality",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
