{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "import seaborn as sns\n",
    "import xarray as xr\n",
    "\n",
    "# Seaborn customisation\n",
    "sns.set_theme()\n",
    "sns.set_style(\"whitegrid\")\n",
    "biomodal_palette = [\"#003B49\", \"#9CDBD9\", \"#F87C56\", \"#C0DF16\", \"#05868E\"]\n",
    "sns.set_palette(biomodal_palette)\n",
    "\n",
    "from modality.datasets import load_biomodal_dataset\n",
    "from modality.annotation import (\n",
    "    get_genes,\n",
    "    get_transcription_end_region,\n",
    "    get_tss_region,\n",
    "    get_exons,\n",
    "    get_introns,\n",
    "    get_five_prime_utrs,\n",
    "    get_three_prime_utrs,\n",
    "    get_transcripts,\n",
    "    get_cpg_islands,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load ES-E14 evoC dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from modality.contig_dataset import ContigDataset\n",
    "def unify_chr_names(ds: ContigDataset) -> ContigDataset:\n",
    "    # Extract the contig column\n",
    "    selected_column = ds[\"contig\"].values\n",
    "    \n",
    "    # Check if the contigs start with \"chr\" and remove it if they do\n",
    "    unified_contigs = np.array([contig[3:] if contig.startswith(\"chr\") else contig \n",
    "                                for contig in selected_column],dtype='U20')\n",
    "    \n",
    "    # Assign the unified contigs back to the dataset\n",
    "    ds = ds.assign_coords(contig=(\"pos\", unified_contigs))\n",
    "    \n",
    "    # Print unique values of the contig column\n",
    "    print(np.unique(ds[\"contig\"].values))\n",
    "    return ds\n",
    "\n",
    "\n",
    "def unify_slice_names(ds):\n",
    "    # Iterate through the dataset attributes and update slice names\n",
    "    for key in list(ds.attrs.keys()):\n",
    "        if key.startswith('slice_chr'):\n",
    "            new_key = key.replace('slice_chr', 'slice_')\n",
    "            ds.attrs[new_key] = ds.attrs.pop(key)\n",
    "    \n",
    "    return ds\n",
    "\n",
    "def load_data(dataset=\"\"):\n",
    "    #downloads modality.contig_dataset.ContigDataset object\n",
    "    if dataset != \"\":  \n",
    "        ds = ContigDataset.from_zarrz(dataset)\n",
    "        ds = unify_chr_names(ds)\n",
    "        ds = unify_slice_names(ds)\n",
    "    else:\n",
    "        ds = load_biomodal_dataset()\n",
    "        # ds = ContigDataset.from_zarrz(\"../ES-E14.zarrz\")\n",
    "        ds = ds.drop_vars([\"Input DNA Quantity (ng/sample)\", \"tech_replicate_number\"])  \n",
    "    ds = ds.sum(dim=\"sample_id\", keep_attrs=True)\n",
    "    ds = ds.expand_dims(dim=\"sample_id\", axis=1)\n",
    "    ds = ds.assign_coords(sample_id=[\"sample_0\"])\n",
    "    ds.assign_fractions(\n",
    "        numerators=[\"num_modc\", \"num_mc\", \"num_hmc\"],\n",
    "        denominator=\"num_total_c\",\n",
    "        min_coverage=10,\n",
    "        inplace=True,\n",
    "    )\n",
    "    return ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ds = load_data(\"../data/CEGX_Run1485_CG.zarrz\")\n",
    "# ds.to_zarrz(\"../data/CEGX_Run1485_CG_renamed.zarrz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds = ContigDataset.from_zarrz(\"../data/CEGX_Run1485_CG_renamed.zarrz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### subset `ContigDataset` by strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from modality.contig_dataset import set_contig_slices, cast_result\n",
    "\n",
    "def refresh_slices(subset_data):\n",
    "    if subset_data.pos.size == 0:\n",
    "            raise ValueError(\n",
    "                \"Subset is empty. Please check the coverage values and method used.\"\n",
    "            )\n",
    "    else:\n",
    "        slices = [x for x in subset_data.attrs if x.startswith(\"slice\")]\n",
    "        for sl in slices:\n",
    "            subset_data.attrs.pop(sl)\n",
    "        return set_contig_slices(subset_data)\n",
    "\n",
    "def split_strand(data, strand):\n",
    "    \"\"\"\n",
    "    split xarray dataset into positive and negative strands\n",
    "    Parameters\n",
    "        data : xarray.Dataset\n",
    "        strand : str ('+', '-')\n",
    "    Returns\n",
    "        strand_data :xarray.Dataset\n",
    "    \"\"\"\n",
    "    strand_mask = (data['strand'].data == strand).compute()\n",
    "    masked_da = xr.DataArray(strand_mask, dims=data['strand'].dims, coords=data['strand'].coords)\n",
    "    strand_data = data.where(masked_da, drop=True)\n",
    "\n",
    "    # return strand_data\n",
    "    return cast_result(refresh_slices(strand_data.ds))\n",
    "\n",
    "def remove_low_coverage(data: xr, threshold: int) -> xr.Dataset:\n",
    "    \"\"\"\n",
    "    Removes data points with num_total_c less than the given threshold.\n",
    "\n",
    "    Parameters:\n",
    "    - data (xr.Dataset): The input dataset containing the variable `num_total_c`.\n",
    "    - threshold (int): The threshold value. Rows with `num_total_c` less than this will be removed.\n",
    "\n",
    "    Returns:\n",
    "    - xr.Dataset: The dataset with rows filtered by the threshold.\n",
    "    \"\"\"\n",
    "    # Ensure that num_total_c exists and has valid dimensions\n",
    "    if 'num_total_c' not in data.data_vars:\n",
    "        raise ValueError(\"The dataset does not contain the 'num_total_c' variable.\")\n",
    "    \n",
    "    coverage_mask = (data['num_total_c'].data >= threshold).compute()\n",
    "    masked_da = xr.DataArray(coverage_mask, dims=data['num_total_c'].dims, coords=data['num_total_c'].coords)\n",
    "    # Apply the mask to the dataset and drop positions that don't meet the threshold\n",
    "    filtered_data = data.where(masked_da, drop=True)\n",
    "    \n",
    "    return cast_result(refresh_slices(filtered_data.ds))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plus_strand = split_strand(ds, \"+\")\n",
    "# minus_strand = split_strand(ds, \"-\")\n",
    "\n",
    "# plus_strand.to_zarrz(\"../data/plus_strand.zarrz\")\n",
    "# minus_strand.to_zarrz(\"../data/minus_strand.zarrz\")\n",
    "\n",
    "# get original dataset for comparison\n",
    "# ds_orig = load_data()\n",
    "# plus_strand_orig = split_strand(ds_orig, \"+\")\n",
    "# minus_strand_orig = split_strand(ds_orig, \"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "plus_strand = ContigDataset.from_zarrz(\"../data/plus_strand.zarrz\")\n",
    "minus_strand= ContigDataset.from_zarrz(\"../data/minus_strand.zarrz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### removing low coverage positions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Compute the dask array into a numpy array\n",
    "# num_total_c_computed = ds['num_total_c'].compute()\n",
    "\n",
    "# # Plot the distribution using seaborn or matplotlib\n",
    "# sns.histplot(num_total_c_computed, bins=50, kde=True)\n",
    "# plt.title(\"Distribution of num_total_c, full dataset\")\n",
    "# plt.xlabel(\"num_total_c values\")\n",
    "# plt.ylabel(\"Frequency\")\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "working with strand aware subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-21 18:13:02 | INFO | [modality/contig_slices.py:25] Computing available chromosomes.\n",
      "2024-09-21 18:13:05 | INFO | [modality/contig_slices.py:44] Found available contigs: 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,X,Y,M.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proportion of instances where num_total_c == 0 in plus_strand: 0.11373456826114078\n"
     ]
    }
   ],
   "source": [
    "plus_all = plus_strand[\"num_total_c\"].size\n",
    "plus_strand_cov = remove_low_coverage(plus_strand, 10)\n",
    "plus_filtered = plus_strand_cov[\"num_total_c\"].size\n",
    "\n",
    "print(f\"Proportion of instances where num_total_c == 0 in plus_strand: {(plus_all - plus_filtered)/plus_all}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-21 18:17:14 | INFO | [modality/contig_slices.py:25] Computing available chromosomes.\n",
      "2024-09-21 18:17:18 | INFO | [modality/contig_slices.py:44] Found available contigs: 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,X,Y,M.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proportion of instances where num_total_c == 0 in minus_strand: 0.11376635009671968\n"
     ]
    }
   ],
   "source": [
    "minus_all = minus_strand[\"num_total_c\"].size\n",
    "minus_strand_cov = remove_low_coverage(minus_strand, 10)\n",
    "minus_filtered = minus_strand_cov[\"num_total_c\"].size\n",
    "\n",
    "print(f\"Proportion of instances where num_total_c == 0 in minus_strand: {(minus_all - minus_filtered)/minus_all}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Compute the dask array into a numpy array\n",
    "# num_total_c_computed = plus_strand_cov['num_total_c'].compute()\n",
    "\n",
    "# # Plot the distribution using seaborn or matplotlib\n",
    "# sns.histplot(num_total_c_computed, bins=50, kde=True)\n",
    "# plt.title(\"Distribution of num_total_c, cleaned Plus\")\n",
    "# plt.xlabel(\"num_total_c values\")\n",
    "# plt.ylabel(\"Frequency\")\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "double-checking the counts of remaining positions per contig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "slice_1\n",
      "(1341826,)\n",
      "slice_2\n",
      "(1414344,)\n",
      "slice_3\n",
      "(1086495,)\n",
      "slice_4\n",
      "(1174582,)\n",
      "slice_5\n",
      "(1238054,)\n",
      "slice_6\n",
      "(1069125,)\n",
      "slice_7\n",
      "(1055303,)\n",
      "slice_8\n",
      "(1029525,)\n",
      "slice_9\n",
      "(999321,)\n",
      "slice_10\n",
      "(1029648,)\n",
      "slice_11\n",
      "(1098901,)\n",
      "slice_12\n",
      "(850345,)\n",
      "slice_13\n",
      "(898712,)\n",
      "slice_14\n",
      "(812156,)\n",
      "slice_15\n",
      "(791927,)\n",
      "slice_16\n",
      "(694729,)\n",
      "slice_17\n",
      "(768263,)\n",
      "slice_18\n",
      "(658766,)\n",
      "slice_19\n",
      "(514211,)\n",
      "slice_X\n",
      "(784165,)\n",
      "slice_Y\n",
      "(70093,)\n",
      "slice_M\n",
      "(217,)\n"
     ]
    }
   ],
   "source": [
    "for contig in [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,'X','Y','M']:\n",
    "    slice1 = plus_strand_cov.slices[f\"slice_{contig}\"]\n",
    "    print(f\"slice_{contig}\")\n",
    "    print(plus_strand_cov.ref_position[slice1].data.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Loading mm10 (GRCm38) reference genome from [GENCODE](https://www.gencodegenes.org/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-21 18:17:29 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------+------------+------------+-----------+-----------+-------+\n",
      "|   Chromosome | Source     | Type       |     Start |       End | +12   |\n",
      "|   (category) | (object)   | (object)   |   (int64) |   (int64) | ...   |\n",
      "|--------------+------------+------------+-----------+-----------+-------|\n",
      "|            1 | HAVANA     | gene       |   4807787 |   4848409 | ...   |\n",
      "|            1 | HAVANA     | gene       |   4807891 |   4886769 | ...   |\n",
      "|            1 | HAVANA     | gene       |   4857813 |   4897908 | ...   |\n",
      "|            1 | HAVANA     | gene       |   5070017 |   5162528 | ...   |\n",
      "|            1 | HAVANA     | gene       |   5588465 |   5606130 | ...   |\n",
      "+--------------+------------+------------+-----------+-----------+-------+\n",
      "Stranded PyRanges object has 5 rows and 17 columns from 1 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
      "12 hidden columns: Score, Strand, Phase, Id, Gene_id, Gene_type, Gene_name, Level, ... (+ 4 more.)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-21 18:17:42 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------------------------------------------------\n",
      " all genes: 21673\n",
      " unique gene id: 21673\n"
     ]
    }
   ],
   "source": [
    "# Get the gene annotations\n",
    "gene_filter = {\n",
    "    \"gene_type\": \"protein_coding\",\n",
    "    \"source\": \"HAVANA\",\n",
    "}\n",
    "\n",
    "genes = get_genes(\n",
    "    reference=\"mm10\",\n",
    "    as_pyranges=True,\n",
    "    filterby=gene_filter,\n",
    ")\n",
    "print(genes.head(5))\n",
    "\n",
    "# Get transcripts for mm10\n",
    "transcripts = get_transcripts(\n",
    "        reference=\"mm10\",\n",
    "        contig=None,\n",
    "        start=None,\n",
    "        end=None,\n",
    "        as_pyranges=False,\n",
    "    )\n",
    "\n",
    "print('-'*100)\n",
    "\n",
    "print(f\" all genes: {len(genes)}\")\n",
    "print(f\" unique gene id: {len(genes.df['Id'])}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Select primiary transcript for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003.15</td>\n",
       "      <td>ENSMUST00000000003.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028.15</td>\n",
       "      <td>ENSMUST00000000028.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000037.17</td>\n",
       "      <td>ENSMUST00000101113.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000049.11</td>\n",
       "      <td>ENSMUST00000000049.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene_id          transcript_id\n",
       "0   ENSMUSG00000000001.4   ENSMUST00000000001.4\n",
       "1  ENSMUSG00000000003.15  ENSMUST00000000003.13\n",
       "2  ENSMUSG00000000028.15  ENSMUST00000000028.13\n",
       "3  ENSMUSG00000000037.17   ENSMUST00000101113.8\n",
       "4  ENSMUSG00000000049.11   ENSMUST00000000049.5"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def select_transcript_based_on_tag(df):\n",
    "    # for each transcript in df, select the one with the highest priority tag\n",
    "    # priorities are:\n",
    "        # 1. 'basic,appris_principal_1,CCDS'\n",
    "        # 2. 'basic,appris_principal_1'\n",
    "        # 3. 'basic,CCDS'\n",
    "        # 4. 'basic'\n",
    "    # but with 'exp_conf' (experimentally confirmed) tag, the priority is higher.\n",
    "    \n",
    "    priorties = {\n",
    "        'basic,appris_principal_1,exp_conf,CCDS': 1,\n",
    "        'basic,appris_principal_1,CCDS': 1,\n",
    "        'basic,appris_principal_1,exp_conf': 3,\n",
    "        'basic,appris_principal_1': 4,\n",
    "        'basic,exp_conf,CCDS': 5,\n",
    "        'basic,CCDS': 6,\n",
    "        'basic,exp_conf': 7,\n",
    "        'basic': 8\n",
    "    }\n",
    "\n",
    "    # sort the dataframe by the priority of the tags\n",
    "    df['tag_priority'] = df.tag.map(priorties)\n",
    "\n",
    "    df = df.sort_values(by='tag_priority')\n",
    "\n",
    "    # drop duplicates, keeping the first one\n",
    "    df = df.drop_duplicates(subset='gene_id', keep='first')\n",
    "\n",
    "    return df[[\"gene_id\", \"transcript_id\"]]\n",
    "\n",
    "selected_transcripts = transcripts.groupby('gene_id').apply(\n",
    "    select_transcript_based_on_tag\n",
    "    ).reset_index(drop=True)\n",
    "selected_transcripts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 21541 genes with unique transcript ids\n"
     ]
    }
   ],
   "source": [
    "# Create dictionary for gene_id to transcript_id mapping\n",
    "gene_to_transcription = selected_transcripts.set_index('gene_id')['transcript_id'].to_dict()\n",
    "print(f\"There are {len(gene_to_transcription)} genes with unique transcript ids\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "review / drop genes with no transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all genes: 21673\n",
      "all genes with annotated transcripts: 21541\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807787</td>\n",
       "      <td>4848409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807891</td>\n",
       "      <td>4886769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5611216</td>\n",
       "      <td>OTTMUSG00000050100.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUST00000155020.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4857813</td>\n",
       "      <td>4897908</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5070017</td>\n",
       "      <td>5162528</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Atp6v1h</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1914864</td>\n",
       "      <td>OTTMUSG00000050145.9</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUST00000192847.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5588465</td>\n",
       "      <td>5606130</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21536</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>78835720</td>\n",
       "      <td>78838055</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20806</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434162</td>\n",
       "      <td>OTTMUSG00000046577.2</td>\n",
       "      <td></td>\n",
       "      <td>21668</td>\n",
       "      <td>ENSMUST00000190349.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21537</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>79148788</td>\n",
       "      <td>79151121</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20917</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434273</td>\n",
       "      <td>OTTMUSG00000046619.2</td>\n",
       "      <td></td>\n",
       "      <td>21669</td>\n",
       "      <td>ENSMUST00000188706.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21538</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>84562571</td>\n",
       "      <td>84564906</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21394</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434749</td>\n",
       "      <td>OTTMUSG00000045415.1</td>\n",
       "      <td></td>\n",
       "      <td>21670</td>\n",
       "      <td>ENSMUST00000189463.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21539</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>85528516</td>\n",
       "      <td>85530907</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434210</td>\n",
       "      <td>OTTMUSG00000042966.1</td>\n",
       "      <td></td>\n",
       "      <td>21671</td>\n",
       "      <td>ENSMUST00000181549.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21540</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89713423</td>\n",
       "      <td>89745531</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5440224</td>\n",
       "      <td>OTTMUSG00000047316.1</td>\n",
       "      <td></td>\n",
       "      <td>21672</td>\n",
       "      <td>ENSMUST00000188269.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21541 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +13   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 4807787   | 4848409   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4807891   | 4886769   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4857813   | 4897908   | ...   |\n",
       "| 1            | HAVANA     | gene       | 5070017   | 5162528   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 79148788  | 79151121  | ...   |\n",
       "| Y            | HAVANA     | gene       | 84562571  | 84564906  | ...   |\n",
       "| Y            | HAVANA     | gene       | 85528516  | 85530907  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89713423  | 89745531  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 21,541 rows and 18 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "13 hidden columns: Score, Strand, Phase, Id, Gene_id, Gene_type, Gene_name, Level, ... (+ 5 more.)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Function to map transcription_id\n",
    "def map_transcription_id(gene_id, gene_to_transcription):\n",
    "    return gene_to_transcription.get(gene_id, None)\n",
    "\n",
    "# Map transcription_id \n",
    "genes_df = genes.df\n",
    "genes_df['Transcript_id'] = genes_df['Gene_id'].map(lambda gene_id: map_transcription_id(gene_id, gene_to_transcription))\n",
    "genes_clean = pr.PyRanges(genes_df)\n",
    "\n",
    "# Find and Handle missing gene_ids\n",
    "missing_gene_ids = genes_clean.df[genes_clean.df['Transcript_id'].isna()]['Gene_id'].tolist()\n",
    "missing_df = genes_clean.df[genes_clean.df['Gene_id'].isin(missing_gene_ids)]\n",
    "genes_clean = genes_clean[~genes_clean.df['Transcript_id'].isna()]\n",
    "print(f\"all genes: {len(genes)}\")\n",
    "print(f\"all genes with annotated transcripts: {len(genes_clean)}\")\n",
    "genes_clean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tags for 132 missing genes:\n",
      "Tag\n",
      "                                                 87\n",
      "overlapping_locus                                13\n",
      "reference_genome_error                           10\n",
      "fragmented_locus                                  9\n",
      "ncRNA_host,fragmented_locus                       2\n",
      "ncRNA_host                                        2\n",
      "overlapping_locus,reference_genome_error          2\n",
      "ncRNA_host,reference_genome_error                 2\n",
      "ncRNA_host,overlapping_locus                      2\n",
      "fragmented_locus,reference_genome_error           1\n",
      "fragmented_locus,overlapping_locus                1\n",
      "ncRNA_host,fragmented_locus,overlapping_locus     1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(f\"Tags for {len(missing_df)} missing genes:\")\n",
    "print(missing_df['Tag'].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract Regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### requires stitching"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 200bp before Transcription Start Site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default arguments for get_tss_region and get_transcription_end_region\n",
    "default_args = {\n",
    "    \"contig\": None,\n",
    "    \"start\": None,\n",
    "    \"end\": None,\n",
    "    \"reference\": \"mm10\",\n",
    "    \"as_pyranges\": True,\n",
    "    \"protein_coding\": True,\n",
    "    \"filterby\": None,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-21 18:18:35 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "before_tss regions with transcripts: 21541\n",
      "====================================================================================================\n"
     ]
    }
   ],
   "source": [
    "before_tss = get_tss_region(\n",
    "    start_offset=-200,\n",
    "    span=200,\n",
    "    **default_args,\n",
    ")\n",
    "\n",
    "# Map transcription_id \n",
    "before_tss_df = before_tss.df\n",
    "before_tss_df['Transcript_id'] = before_tss_df['Gene_id'].map(lambda gene_id: map_transcription_id(gene_id, gene_to_transcription))\n",
    "before_tss = pr.PyRanges(before_tss_df)\n",
    "before_tss = before_tss[~before_tss.df['Transcript_id'].isna()]\n",
    "\n",
    "print(f\"before_tss regions with transcripts: {len(before_tss)}\")\n",
    "print(\"=\"*100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807587</td>\n",
       "      <td>4807787</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807691</td>\n",
       "      <td>4807891</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5611216</td>\n",
       "      <td>OTTMUSG00000050100.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUST00000155020.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4857613</td>\n",
       "      <td>4857813</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5069817</td>\n",
       "      <td>5070017</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Atp6v1h</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1914864</td>\n",
       "      <td>OTTMUSG00000050145.9</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUST00000192847.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5588265</td>\n",
       "      <td>5588465</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10735</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>87116620</td>\n",
       "      <td>87116820</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21477</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434832</td>\n",
       "      <td>OTTMUSG00000047083.1</td>\n",
       "      <td></td>\n",
       "      <td>21601</td>\n",
       "      <td>ENSMUST00000189543.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10736</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>87550765</td>\n",
       "      <td>87550965</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20906</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td></td>\n",
       "      <td>21602</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10737</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>88053114</td>\n",
       "      <td>88053314</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm28102</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td></td>\n",
       "      <td>21603</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10738</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89052605</td>\n",
       "      <td>89052805</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102045.1</td>\n",
       "      <td>ENSMUSG00000102045.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21294</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434649</td>\n",
       "      <td>OTTMUSG00000047309.1</td>\n",
       "      <td></td>\n",
       "      <td>21604</td>\n",
       "      <td>ENSMUST00000186443.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10739</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90401047</td>\n",
       "      <td>90401247</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000096178.7</td>\n",
       "      <td>ENSMUSG00000096178.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20837</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434193</td>\n",
       "      <td>OTTMUSG00000047352.1</td>\n",
       "      <td></td>\n",
       "      <td>21605</td>\n",
       "      <td>ENSMUST00000190558.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10740 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +12   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 4807587   | 4807787   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4807691   | 4807891   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4857613   | 4857813   | ...   |\n",
       "| 1            | HAVANA     | gene       | 5069817   | 5070017   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 87550765  | 87550965  | ...   |\n",
       "| Y            | HAVANA     | gene       | 88053114  | 88053314  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89052605  | 89052805  | ...   |\n",
       "| Y            | HAVANA     | gene       | 90401047  | 90401247  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Unstranded PyRanges object has 10,740 rows and 17 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "12 hidden columns: Score, Phase, Id, Gene_id, Gene_type, Gene_name, Level, Mgi_id, ... (+ 4 more.)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#split strands\n",
    "before_tss_plus = pr.PyRanges(before_tss.df[before_tss.df[\"Strand\"]==\"+\"])\n",
    "before_tss_minus = pr.PyRanges(before_tss.df[before_tss.df[\"Strand\"]==\"-\"])\n",
    "\n",
    "before_tss_plus = before_tss_plus.unstrand()\n",
    "before_tss_minus = before_tss_minus.unstrand()\n",
    "before_tss_plus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3671497</td>\n",
       "      <td>3671697</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td></td>\n",
       "      <td>611</td>\n",
       "      <td>ENSMUST00000070533.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4409240</td>\n",
       "      <td>4409440</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025900.13</td>\n",
       "      <td>ENSMUSG00000025900.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>OTTMUSG00000049985.3</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>612</td>\n",
       "      <td>ENSMUST00000027032.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4497353</td>\n",
       "      <td>4497553</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025902.13</td>\n",
       "      <td>ENSMUSG00000025902.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:107543</td>\n",
       "      <td>OTTMUSG00000050014.7</td>\n",
       "      <td></td>\n",
       "      <td>613</td>\n",
       "      <td>ENSMUST00000027035.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4785738</td>\n",
       "      <td>4785938</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033845.13</td>\n",
       "      <td>ENSMUSG00000033845.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1351639</td>\n",
       "      <td>OTTMUSG00000029329.3</td>\n",
       "      <td></td>\n",
       "      <td>614</td>\n",
       "      <td>ENSMUST00000156816.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5070284</td>\n",
       "      <td>5070484</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000002459.17</td>\n",
       "      <td>ENSMUSG00000002459.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Rgs20</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1929866</td>\n",
       "      <td>OTTMUSG00000029338.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>615</td>\n",
       "      <td>ENSMUST00000002533.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10796</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>78838055</td>\n",
       "      <td>78838255</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20806</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434162</td>\n",
       "      <td>OTTMUSG00000046577.2</td>\n",
       "      <td></td>\n",
       "      <td>21668</td>\n",
       "      <td>ENSMUST00000190349.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10797</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>79151121</td>\n",
       "      <td>79151321</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20917</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434273</td>\n",
       "      <td>OTTMUSG00000046619.2</td>\n",
       "      <td></td>\n",
       "      <td>21669</td>\n",
       "      <td>ENSMUST00000188706.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10798</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>84564906</td>\n",
       "      <td>84565106</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21394</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434749</td>\n",
       "      <td>OTTMUSG00000045415.1</td>\n",
       "      <td></td>\n",
       "      <td>21670</td>\n",
       "      <td>ENSMUST00000189463.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10799</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>85530907</td>\n",
       "      <td>85531107</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434210</td>\n",
       "      <td>OTTMUSG00000042966.1</td>\n",
       "      <td></td>\n",
       "      <td>21671</td>\n",
       "      <td>ENSMUST00000181549.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10800</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89745531</td>\n",
       "      <td>89745731</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5440224</td>\n",
       "      <td>OTTMUSG00000047316.1</td>\n",
       "      <td></td>\n",
       "      <td>21672</td>\n",
       "      <td>ENSMUST00000188269.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10801 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +12   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 3671497   | 3671697   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4409240   | 4409440   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4497353   | 4497553   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4785738   | 4785938   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 79151121  | 79151321  | ...   |\n",
       "| Y            | HAVANA     | gene       | 84564906  | 84565106  | ...   |\n",
       "| Y            | HAVANA     | gene       | 85530907  | 85531107  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89745531  | 89745731  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Unstranded PyRanges object has 10,801 rows and 17 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "12 hidden columns: Score, Phase, Id, Gene_id, Gene_type, Gene_name, Level, Mgi_id, ... (+ 4 more.)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "before_tss_minus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1000bp after Transcription End Site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-21 18:18:48 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after_tes regions with transcripts: 21541\n",
      "====================================================================================================\n"
     ]
    }
   ],
   "source": [
    "after_tes = get_transcription_end_region(\n",
    "    start_offset=0,\n",
    "    span=1000,\n",
    "    **default_args,\n",
    ")\n",
    "\n",
    "# Map transcription_id \n",
    "after_tes_df = after_tes.df\n",
    "after_tes_df['Transcript_id'] = after_tes_df['Gene_id'].map(lambda gene_id: map_transcription_id(gene_id, gene_to_transcription))\n",
    "after_tes = pr.PyRanges(after_tes_df)\n",
    "after_tes = after_tes[~after_tes.df['Transcript_id'].isna()]\n",
    "\n",
    "print(f\"after_tes regions with transcripts: {len(after_tes)}\")\n",
    "print(\"=\"*100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4848409</td>\n",
       "      <td>4849409</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4886769</td>\n",
       "      <td>4887769</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5611216</td>\n",
       "      <td>OTTMUSG00000050100.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUST00000155020.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4897908</td>\n",
       "      <td>4898908</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5162528</td>\n",
       "      <td>5163528</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Atp6v1h</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1914864</td>\n",
       "      <td>OTTMUSG00000050145.9</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUST00000192847.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5606130</td>\n",
       "      <td>5607130</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10735</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>87142916</td>\n",
       "      <td>87143916</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21477</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434832</td>\n",
       "      <td>OTTMUSG00000047083.1</td>\n",
       "      <td></td>\n",
       "      <td>21601</td>\n",
       "      <td>ENSMUST00000189543.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10736</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>87575690</td>\n",
       "      <td>87576690</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20906</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td></td>\n",
       "      <td>21602</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10737</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>88079493</td>\n",
       "      <td>88080493</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm28102</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td></td>\n",
       "      <td>21603</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10738</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89079014</td>\n",
       "      <td>89080014</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102045.1</td>\n",
       "      <td>ENSMUSG00000102045.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21294</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434649</td>\n",
       "      <td>OTTMUSG00000047309.1</td>\n",
       "      <td></td>\n",
       "      <td>21604</td>\n",
       "      <td>ENSMUST00000186443.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10739</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90433262</td>\n",
       "      <td>90434262</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000096178.7</td>\n",
       "      <td>ENSMUSG00000096178.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20837</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434193</td>\n",
       "      <td>OTTMUSG00000047352.1</td>\n",
       "      <td></td>\n",
       "      <td>21605</td>\n",
       "      <td>ENSMUST00000190558.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10740 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +12   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 4848409   | 4849409   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4886769   | 4887769   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4897908   | 4898908   | ...   |\n",
       "| 1            | HAVANA     | gene       | 5162528   | 5163528   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 87575690  | 87576690  | ...   |\n",
       "| Y            | HAVANA     | gene       | 88079493  | 88080493  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89079014  | 89080014  | ...   |\n",
       "| Y            | HAVANA     | gene       | 90433262  | 90434262  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Unstranded PyRanges object has 10,740 rows and 17 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "12 hidden columns: Score, Phase, Id, Gene_id, Gene_type, Gene_name, Level, Mgi_id, ... (+ 4 more.)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#split strands\n",
    "after_tes_plus = pr.PyRanges(after_tes.df[after_tes.df[\"Strand\"]==\"+\"])\n",
    "after_tes_minus = pr.PyRanges(after_tes.df[after_tes.df[\"Strand\"]==\"-\"])\n",
    "\n",
    "after_tes_plus = after_tes_plus.unstrand()\n",
    "after_tes_minus = after_tes_minus.unstrand()\n",
    "after_tes_plus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ranges_ID</th>\n",
       "      <th>Transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3204900</td>\n",
       "      <td>3205900</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td></td>\n",
       "      <td>611</td>\n",
       "      <td>ENSMUST00000070533.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3998556</td>\n",
       "      <td>3999556</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025900.13</td>\n",
       "      <td>ENSMUSG00000025900.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>OTTMUSG00000049985.3</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>612</td>\n",
       "      <td>ENSMUST00000027032.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4489930</td>\n",
       "      <td>4490930</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025902.13</td>\n",
       "      <td>ENSMUSG00000025902.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:107543</td>\n",
       "      <td>OTTMUSG00000050014.7</td>\n",
       "      <td></td>\n",
       "      <td>613</td>\n",
       "      <td>ENSMUST00000027035.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4772205</td>\n",
       "      <td>4773205</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033845.13</td>\n",
       "      <td>ENSMUSG00000033845.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1351639</td>\n",
       "      <td>OTTMUSG00000029329.3</td>\n",
       "      <td></td>\n",
       "      <td>614</td>\n",
       "      <td>ENSMUST00000156816.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4908575</td>\n",
       "      <td>4909575</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000002459.17</td>\n",
       "      <td>ENSMUSG00000002459.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Rgs20</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1929866</td>\n",
       "      <td>OTTMUSG00000029338.4</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>615</td>\n",
       "      <td>ENSMUST00000002533.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10796</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>78834720</td>\n",
       "      <td>78835720</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>ENSMUSG00000094739.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20806</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434162</td>\n",
       "      <td>OTTMUSG00000046577.2</td>\n",
       "      <td></td>\n",
       "      <td>21668</td>\n",
       "      <td>ENSMUST00000190349.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10797</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>79147788</td>\n",
       "      <td>79148788</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20917</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434273</td>\n",
       "      <td>OTTMUSG00000046619.2</td>\n",
       "      <td></td>\n",
       "      <td>21669</td>\n",
       "      <td>ENSMUST00000188706.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10798</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>84561571</td>\n",
       "      <td>84562571</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21394</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434749</td>\n",
       "      <td>OTTMUSG00000045415.1</td>\n",
       "      <td></td>\n",
       "      <td>21670</td>\n",
       "      <td>ENSMUST00000189463.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10799</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>85527516</td>\n",
       "      <td>85528516</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5434210</td>\n",
       "      <td>OTTMUSG00000042966.1</td>\n",
       "      <td></td>\n",
       "      <td>21671</td>\n",
       "      <td>ENSMUST00000181549.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10800</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89712423</td>\n",
       "      <td>89713423</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:5440224</td>\n",
       "      <td>OTTMUSG00000047316.1</td>\n",
       "      <td></td>\n",
       "      <td>21672</td>\n",
       "      <td>ENSMUST00000188269.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10801 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Type       | Start     | End       | +12   |\n",
       "| (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+-------|\n",
       "| 1            | HAVANA     | gene       | 3204900   | 3205900   | ...   |\n",
       "| 1            | HAVANA     | gene       | 3998556   | 3999556   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4489930   | 4490930   | ...   |\n",
       "| 1            | HAVANA     | gene       | 4772205   | 4773205   | ...   |\n",
       "| ...          | ...        | ...        | ...       | ...       | ...   |\n",
       "| Y            | HAVANA     | gene       | 79147788  | 79148788  | ...   |\n",
       "| Y            | HAVANA     | gene       | 84561571  | 84562571  | ...   |\n",
       "| Y            | HAVANA     | gene       | 85527516  | 85528516  | ...   |\n",
       "| Y            | HAVANA     | gene       | 89712423  | 89713423  | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+-------+\n",
       "Unstranded PyRanges object has 10,801 rows and 17 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "12 hidden columns: Score, Phase, Id, Gene_id, Gene_type, Gene_name, Level, Mgi_id, ... (+ 4 more.)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "after_tes_minus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract 3`UTRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-21 18:19:01 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "strand shape: + (8075, 27), - (8134, 27)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>...</th>\n",
       "      <th>Level</th>\n",
       "      <th>Protein_id</th>\n",
       "      <th>Transcript_support_level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ccdsid</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Havana_transcript</th>\n",
       "      <th>Ont</th>\n",
       "      <th>Ranges_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>4845016</td>\n",
       "      <td>4846738</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000027036.10</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000027036.4</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14806.1</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>OTTMUST00000051162.1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>4896364</td>\n",
       "      <td>4897904</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000081551.13</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000080266.7</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>non_canonical_U12,basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS35505.1</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>OTTMUST00000111602.1</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>5602784</td>\n",
       "      <td>5606130</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000125105.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14809.1</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td>OTTMUST00000088255.1</td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>6274275</td>\n",
       "      <td>6276647</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000027040.12</td>\n",
       "      <td>ENSMUST00000027040.12</td>\n",
       "      <td>ENSMUSG00000025907.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000027040.6</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1341850</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS35507.1</td>\n",
       "      <td>OTTMUSG00000033467.12</td>\n",
       "      <td>OTTMUST00000084091.5</td>\n",
       "      <td></td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>6391104</td>\n",
       "      <td>6391115</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000133144.3</td>\n",
       "      <td>ENSMUST00000133144.3</td>\n",
       "      <td>ENSMUSG00000087247.3</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000137420.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:3645495</td>\n",
       "      <td>not_organism_supported,basic,appris_principal_...</td>\n",
       "      <td>CCDS56620.1</td>\n",
       "      <td>OTTMUSG00000050239.2</td>\n",
       "      <td>OTTMUST00000127625.2</td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8070</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>84772706</td>\n",
       "      <td>84772901</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000186110.1</td>\n",
       "      <td>ENSMUST00000186110.1</td>\n",
       "      <td>ENSMUSG00000099840.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140850.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434764</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000046908.1</td>\n",
       "      <td>OTTMUST00000121611.1</td>\n",
       "      <td></td>\n",
       "      <td>27079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8071</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>86087758</td>\n",
       "      <td>86087951</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000188754.1</td>\n",
       "      <td>ENSMUST00000188754.1</td>\n",
       "      <td>ENSMUSG00000100240.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000139858.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434176</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047031.1</td>\n",
       "      <td>OTTMUST00000121805.1</td>\n",
       "      <td></td>\n",
       "      <td>27080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8072</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>87142720</td>\n",
       "      <td>87142916</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000189543.6</td>\n",
       "      <td>ENSMUST00000189543.6</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140238.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434832</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047083.1</td>\n",
       "      <td>OTTMUST00000121876.1</td>\n",
       "      <td></td>\n",
       "      <td>27081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8073</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>87575495</td>\n",
       "      <td>87575690</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140305.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td>OTTMUST00000121968.1</td>\n",
       "      <td></td>\n",
       "      <td>27082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8074</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>88079297</td>\n",
       "      <td>88079493</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR3:ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140242.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td>OTTMUST00000121980.1</td>\n",
       "      <td></td>\n",
       "      <td>27083</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8075 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-----------------+-----------+-------+\n",
       "| Chromosome   | Source     | Type            | Start     | +23   |\n",
       "| (category)   | (object)   | (object)        | (int64)   | ...   |\n",
       "|--------------+------------+-----------------+-----------+-------|\n",
       "| 1            | HAVANA     | three_prime_UTR | 4845016   | ...   |\n",
       "| 1            | HAVANA     | three_prime_UTR | 4896364   | ...   |\n",
       "| 1            | HAVANA     | three_prime_UTR | 5602784   | ...   |\n",
       "| 1            | HAVANA     | three_prime_UTR | 6274275   | ...   |\n",
       "| ...          | ...        | ...             | ...       | ...   |\n",
       "| Y            | HAVANA     | three_prime_UTR | 86087758  | ...   |\n",
       "| Y            | HAVANA     | three_prime_UTR | 87142720  | ...   |\n",
       "| Y            | HAVANA     | three_prime_UTR | 87575495  | ...   |\n",
       "| Y            | HAVANA     | three_prime_UTR | 88079297  | ...   |\n",
       "+--------------+------------+-----------------+-----------+-------+\n",
       "Unstranded PyRanges object has 8,075 rows and 27 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "23 hidden columns: End, Score, Phase, Id, Parent, Gene_id, Transcript_id, ... (+ 16 more.)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract with modality.get_three_prime_utrs\n",
    "three_prime_utrs = get_three_prime_utrs(reference=\"mm10\")\n",
    "three_prime_utrs = three_prime_utrs[\n",
    "    three_prime_utrs.Transcript_id.isin(selected_transcripts.transcript_id)\n",
    "    ]\n",
    "\n",
    "#split strands\n",
    "three_prime_utrs_plus = pr.PyRanges(three_prime_utrs.df[three_prime_utrs.df[\"Strand\"]==\"+\"])\n",
    "three_prime_utrs_plus = three_prime_utrs_plus.unstrand()\n",
    "\n",
    "three_prime_utrs_minus = pr.PyRanges(three_prime_utrs.df[three_prime_utrs.df[\"Strand\"]==\"-\"])\n",
    "three_prime_utrs_minus = three_prime_utrs_minus.unstrand()\n",
    "\n",
    "print(f\"strand shape: + {three_prime_utrs_plus.df.shape}, - {three_prime_utrs_minus.df.shape}\")\n",
    "three_prime_utrs_plus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract 5`UTRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-21 18:19:16 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "strand shape: + (11449, 27), - (11369, 27)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Id</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>...</th>\n",
       "      <th>Level</th>\n",
       "      <th>Protein_id</th>\n",
       "      <th>Transcript_support_level</th>\n",
       "      <th>Mgi_id</th>\n",
       "      <th>Tag</th>\n",
       "      <th>Ccdsid</th>\n",
       "      <th>Havana_gene</th>\n",
       "      <th>Havana_transcript</th>\n",
       "      <th>Ont</th>\n",
       "      <th>Ranges_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>4807822</td>\n",
       "      <td>4807912</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000027036.10</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000027036.4</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14806.1</td>\n",
       "      <td>OTTMUSG00000021562.4</td>\n",
       "      <td>OTTMUST00000051162.1</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>4857813</td>\n",
       "      <td>4857912</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000081551.13</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000080266.7</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1196624</td>\n",
       "      <td>non_canonical_U12,basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS35505.1</td>\n",
       "      <td>OTTMUSG00000042348.1</td>\n",
       "      <td>OTTMUST00000111602.1</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>5588492</td>\n",
       "      <td>5588662</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000125105.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14809.1</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td>OTTMUST00000088255.1</td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>5589034</td>\n",
       "      <td>5589047</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000125105.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:97439</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS14809.1</td>\n",
       "      <td>OTTMUSG00000034734.3</td>\n",
       "      <td>OTTMUST00000088255.1</td>\n",
       "      <td></td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>6214644</td>\n",
       "      <td>6214956</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000027040.12</td>\n",
       "      <td>ENSMUST00000027040.12</td>\n",
       "      <td>ENSMUSG00000025907.14</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000027040.6</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:1341850</td>\n",
       "      <td>basic,appris_principal_1,CCDS</td>\n",
       "      <td>CCDS35507.1</td>\n",
       "      <td>OTTMUSG00000033467.12</td>\n",
       "      <td>OTTMUST00000084091.5</td>\n",
       "      <td></td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11444</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>87118231</td>\n",
       "      <td>87118244</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000189543.6</td>\n",
       "      <td>ENSMUST00000189543.6</td>\n",
       "      <td>ENSMUSG00000094399.7</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140238.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434832</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047083.1</td>\n",
       "      <td>OTTMUST00000121876.1</td>\n",
       "      <td></td>\n",
       "      <td>39793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11445</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>87550965</td>\n",
       "      <td>87551008</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140305.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td>OTTMUST00000121968.1</td>\n",
       "      <td></td>\n",
       "      <td>39794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11446</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>87552378</td>\n",
       "      <td>87552391</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUST00000186493.1</td>\n",
       "      <td>ENSMUSG00000099856.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140305.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5434262</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047138.1</td>\n",
       "      <td>OTTMUST00000121968.1</td>\n",
       "      <td></td>\n",
       "      <td>39795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11447</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>88053314</td>\n",
       "      <td>88053357</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140242.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td>OTTMUST00000121980.1</td>\n",
       "      <td></td>\n",
       "      <td>39796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11448</th>\n",
       "      <td>Y</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>88054723</td>\n",
       "      <td>88054736</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>UTR5:ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUST00000187146.1</td>\n",
       "      <td>ENSMUSG00000101915.1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSP00000140242.1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5578808</td>\n",
       "      <td>not_best_in_genome_evidence,basic,appris_princ...</td>\n",
       "      <td></td>\n",
       "      <td>OTTMUSG00000047149.1</td>\n",
       "      <td>OTTMUST00000121980.1</td>\n",
       "      <td></td>\n",
       "      <td>39797</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11449 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+----------------+-----------+-------+\n",
       "| Chromosome   | Source     | Type           | Start     | +23   |\n",
       "| (category)   | (object)   | (object)       | (int64)   | ...   |\n",
       "|--------------+------------+----------------+-----------+-------|\n",
       "| 1            | HAVANA     | five_prime_UTR | 4807822   | ...   |\n",
       "| 1            | HAVANA     | five_prime_UTR | 4857813   | ...   |\n",
       "| 1            | HAVANA     | five_prime_UTR | 5588492   | ...   |\n",
       "| 1            | HAVANA     | five_prime_UTR | 5589034   | ...   |\n",
       "| ...          | ...        | ...            | ...       | ...   |\n",
       "| Y            | HAVANA     | five_prime_UTR | 87550965  | ...   |\n",
       "| Y            | HAVANA     | five_prime_UTR | 87552378  | ...   |\n",
       "| Y            | HAVANA     | five_prime_UTR | 88053314  | ...   |\n",
       "| Y            | HAVANA     | five_prime_UTR | 88054723  | ...   |\n",
       "+--------------+------------+----------------+-----------+-------+\n",
       "Unstranded PyRanges object has 11,449 rows and 27 columns from 21 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "23 hidden columns: End, Score, Phase, Id, Parent, Gene_id, Transcript_id, ... (+ 16 more.)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract with modality.get_five_prime_utrs\n",
    "five_prime_utrs = get_five_prime_utrs(reference=\"mm10\")\n",
    "five_prime_utrs = five_prime_utrs[\n",
    "    five_prime_utrs.Transcript_id.isin(selected_transcripts.transcript_id)\n",
    "    ]\n",
    "\n",
    "#split strands\n",
    "five_prime_utrs_plus = pr.PyRanges(five_prime_utrs.df[five_prime_utrs.df[\"Strand\"]==\"+\"])\n",
    "five_prime_utrs_plus = five_prime_utrs_plus.unstrand()\n",
    "\n",
    "five_prime_utrs_minus = pr.PyRanges(five_prime_utrs.df[five_prime_utrs.df[\"Strand\"]==\"-\"])\n",
    "five_prime_utrs_minus = five_prime_utrs_minus.unstrand()\n",
    "print(f\"strand shape: + {five_prime_utrs_plus.df.shape}, - {five_prime_utrs_minus.df.shape}\")\n",
    "five_prime_utrs_plus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract exons and introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-21 18:19:42 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    }
   ],
   "source": [
    "# Exons:\n",
    "exons = get_exons(reference=\"mm10\")\n",
    "exons = exons[exons.Transcript_id.isin(selected_transcripts.transcript_id)]\n",
    "\n",
    "# exons split strands\n",
    "exons_plus = pr.PyRanges(exons.df[exons.df[\"Strand\"]==\"+\"])\n",
    "first_exons_plus = exons_plus[exons_plus.Exon_number == \"1\"].unstrand()\n",
    "exons_plus = exons_plus[exons_plus.Exon_number.astype(\"int\") > 1].unstrand()\n",
    "\n",
    "\n",
    "exons_minus = pr.PyRanges(exons.df[exons.df[\"Strand\"]==\"-\"])\n",
    "first_exons_minus = exons_minus[exons_minus.Exon_number == \"1\"].unstrand()\n",
    "exons_minus = exons_minus[exons_minus.Exon_number.astype(\"int\") > 1].unstrand()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-21 18:20:27 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n",
      "2024-09-21 18:20:51 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n",
      "2024-09-21 18:21:26 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n",
      "2024-09-21 18:21:41 | INFO | [modality/annotation.py:418] Removing readthrough_gene transcripts for gff (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.basic.annotation.gff3.gz)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 8 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "#Introns:\n",
    "introns = get_introns(\n",
    "    reference=\"mm10\",\n",
    "    transcripts=selected_transcripts.transcript_id.values,\n",
    "    nb_workers=8,\n",
    ")\n",
    "\n",
    "# introns split strands\n",
    "introns_plus = pr.PyRanges(introns.df[introns.df[\"Strand\"]==\"+\"])\n",
    "first_introns_plus = introns_plus[introns_plus.Intron_number == \"1\"].unstrand()\n",
    "introns_plus = introns_plus[introns_plus.Intron_number.astype(\"int\") > 1].unstrand()\n",
    "\n",
    "\n",
    "introns_minus = pr.PyRanges(introns.df[introns.df[\"Strand\"]==\"-\"])\n",
    "first_introns_minus = introns_minus[introns_minus.Intron_number == \"1\"].unstrand()\n",
    "introns_minus = introns_minus[introns_minus.Intron_number.astype(\"int\") > 1].unstrand()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exons strand shape: + (90807, 27), - (90167, 27)\n",
      "Introns strand shape: + (85982, 20), - (85480, 20)\n",
      "First Exons strand shape: + (10740, 27), - (10801, 27)\n",
      "First Introns strand shape: + (10186, 20), - (10212, 20)\n"
     ]
    }
   ],
   "source": [
    "print(f\"Exons strand shape: + {exons_plus.df.shape}, - {exons_minus.df.shape}\")\n",
    "print(f\"Introns strand shape: + {introns_plus.df.shape}, - {introns_minus.df.shape}\")\n",
    "\n",
    "print(f\"First Exons strand shape: + {first_exons_plus.df.shape}, - {first_exons_minus.df.shape}\")\n",
    "print(f\"First Introns strand shape: + {first_introns_plus.df.shape}, - {first_introns_minus.df.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Source', 'Type', 'Start', 'End', 'Score', 'Strand',\n",
      "       'Phase', 'Id', 'Parent', 'Gene_id', 'Transcript_id', 'Gene_type',\n",
      "       'Gene_name', 'Transcript_type', 'Transcript_name', 'Exon_number',\n",
      "       'Exon_id', 'Level', 'Transcript_support_level', 'Mgi_id', 'Tag',\n",
      "       'Havana_gene', 'Havana_transcript', 'Protein_id', 'Ccdsid', 'Ont',\n",
      "       'Ranges_ID'],\n",
      "      dtype='object')\n",
      "Index(['Transcript_id', 'Level_1', 'Chromosome', 'Source', 'Type', 'Start',\n",
      "       'End', 'Score', 'Strand', 'Phase', 'Id', 'Gene_id', 'Gene_type',\n",
      "       'Gene_name', 'Level', 'Mgi_id', 'Havana_gene', 'Tag', 'Ranges_id',\n",
      "       'Intron_number', 'Ranges_ID'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(exons.columns) # get_exons don't return \"Transcript_support_level\", 'Havana_transcript', 'Protein_id', 'Ccdsid', 'Ont', 'Parent', 'Exon_id', 'Exon_number'\n",
    "print(introns.columns) # get_introns specific fields return 'Level_1'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Summarise 3\\` 5\\` UTRs exons and introns regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mean_across_region(df, var):\n",
    "    sum_var = df[f\"{var}_sum\"].sum()\n",
    "    sum_total_c = df[\"num_total_c_sum\"].sum()\n",
    "    return sum_var / sum_total_c\n",
    "def summarise_across_region(rdr, region):\n",
    "    \"\"\"\n",
    "    Summarise the methylation data across a region.\n",
    "\n",
    "    Parameters\n",
    "        rdr: pyranges object\n",
    "        region: str\n",
    "\n",
    "    Returns\n",
    "        grouped_df: pd.DataFrame\n",
    "    \"\"\"\n",
    "\n",
    "    df = rdr.to_dataframe().reset_index(level=\"sample_id\", drop=True)\n",
    "    grouped = df.groupby(\"Gene_id\")\n",
    "\n",
    "    mean_mc = grouped.apply(get_mean_across_region, var=\"num_mc\")\n",
    "    mean_hmc = grouped.apply(get_mean_across_region, var=\"num_hmc\")\n",
    "    mean_modc = grouped.apply(get_mean_across_region, var=\"num_modc\")\n",
    "    cpg_count = grouped[\"num_total_c_cpg_count\"].sum()\n",
    "    sum_c = grouped[\"num_total_c_sum\"].sum()\n",
    "    range_length = grouped[\"range_length\"].sum()\n",
    "    gene_name = grouped[\"Gene_name\"].first()\n",
    "    contig = grouped[\"contig\"].first()\n",
    "    strand = grouped[\"strand\"].first()\n",
    "    \n",
    "    \n",
    "    # Prepare list for storing primary_transcripts and handling missing gene_ids\n",
    "    primary_transcripts = []\n",
    "    missing_gene_ids = []\n",
    "\n",
    "    for gene_id in grouped.groups.keys():\n",
    "        if gene_id in gene_to_transcription:\n",
    "            primary_transcripts.append(gene_to_transcription[gene_id])\n",
    "        else:\n",
    "            primary_transcripts.append(None)\n",
    "            missing_gene_ids.append(gene_id)\n",
    "\n",
    "    # Print missing gene_ids\n",
    "    for gene_id in missing_gene_ids:\n",
    "        print(f\"Gene_ID {gene_id} not found in selected_transcripts\")\n",
    "\n",
    "    \n",
    "    grouped_df = pd.DataFrame(\n",
    "        {\n",
    "            \"mean_mc\": mean_mc,\n",
    "            \"mean_hmc\": mean_hmc,\n",
    "            \"mean_modc\": mean_modc,\n",
    "            \"cpg_count\": cpg_count,\n",
    "            \"total_c\" : sum_c,\n",
    "            \"range_length\": range_length,\n",
    "            \"Gene_name\": gene_name,\n",
    "            \"contig\": contig,\n",
    "            \"strand\": strand,\n",
    "            \"Transcript_id\": primary_transcripts,\n",
    "        }\n",
    "    )\n",
    "\n",
    "    grouped_df = grouped_df.reset_index()\n",
    "    grouped_df[\"Region\"] = region\n",
    "    return grouped_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "regions for plus strand:\n",
      "196313\n"
     ]
    }
   ],
   "source": [
    "five_prime_utrs_count = five_prime_utrs_plus.df.shape[0]\n",
    "exon_count = exons_plus.df.shape[0]\n",
    "intron_count = introns_plus.df.shape[0]\n",
    "three_prime_utrs_count = three_prime_utrs_plus.df.shape[0]\n",
    "\n",
    "print(\"regions for plus strand:\")\n",
    "print(five_prime_utrs_count+exon_count+intron_count+three_prime_utrs_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "regions for minus strand:\n",
      "195150\n"
     ]
    }
   ],
   "source": [
    "five_prime_utrs_count = five_prime_utrs_minus.df.shape[0]\n",
    "exon_count = exons_minus.df.shape[0]\n",
    "intron_count = introns_minus.df.shape[0]\n",
    "three_prime_utrs_count = three_prime_utrs_minus.df.shape[0]\n",
    "\n",
    "print(\"regions for minus strand:\")\n",
    "print(five_prime_utrs_count+exon_count+intron_count+three_prime_utrs_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'exons': +--------------+------------+------------+-----------+-----------+-------+\n",
       " | Chromosome   | Source     | Type       | Start     | End       | +22   |\n",
       " | (category)   | (object)   | (object)   | (int64)   | (int64)   | ...   |\n",
       " |--------------+------------+------------+-----------+-----------+-------|\n",
       " | 1            | HAVANA     | exon       | 4808454   | 4808485   | ...   |\n",
       " | 1            | HAVANA     | exon       | 4828583   | 4828648   | ...   |\n",
       " | 1            | HAVANA     | exon       | 4828583   | 4828648   | ...   |\n",
       " | 1            | HAVANA     | exon       | 4830267   | 4830314   | ...   |\n",
       " | ...          | ...        | ...        | ...       | ...       | ...   |\n",
       " | Y            | HAVANA     | exon       | 90429592  | 90429709  | ...   |\n",
       " | Y            | HAVANA     | exon       | 90430460  | 90430559  | ...   |\n",
       " | Y            | HAVANA     | exon       | 90432572  | 90432670  | ...   |\n",
       " | Y            | HAVANA     | exon       | 90433029  | 90433262  | ...   |\n",
       " +--------------+------------+------------+-----------+-----------+-------+\n",
       " Unstranded PyRanges object has 90,807 rows and 27 columns from 21 chromosomes.\n",
       " For printing, the PyRanges was sorted on Chromosome.\n",
       " 22 hidden columns: Score, Phase, Id, Parent, Gene_id, Transcript_id, Gene_type, ... (+ 15 more.),\n",
       " 'introns': +-----------------------+-----------+--------------+------------+-------+\n",
       " | Transcript_id         | Level_1   | Chromosome   | Source     | +16   |\n",
       " | (object)              | (int64)   | (category)   | (object)   | ...   |\n",
       " |-----------------------+-----------+--------------+------------+-------|\n",
       " | ENSMUST00000000514.10 | 1         | 1            | HAVANA     | ...   |\n",
       " | ENSMUST00000000514.10 | 2         | 1            | HAVANA     | ...   |\n",
       " | ENSMUST00000000514.10 | 3         | 1            | HAVANA     | ...   |\n",
       " | ENSMUST00000000514.10 | 4         | 1            | HAVANA     | ...   |\n",
       " | ...                   | ...       | ...          | ...        | ...   |\n",
       " | ENSMUST00000194687.1  | 1         | Y            | HAVANA     | ...   |\n",
       " | ENSMUST00000195574.1  | 1         | Y            | HAVANA     | ...   |\n",
       " | ENSMUST00000195574.1  | 2         | Y            | HAVANA     | ...   |\n",
       " | ENSMUST00000195626.1  | 1         | Y            | HAVANA     | ...   |\n",
       " +-----------------------+-----------+--------------+------------+-------+\n",
       " Unstranded PyRanges object has 85,982 rows and 20 columns from 21 chromosomes.\n",
       " For printing, the PyRanges was sorted on Chromosome.\n",
       " 16 hidden columns: Type, Start, End, Score, Phase, Id, Gene_id, Gene_type, Gene_name, ... (+ 7 more.),\n",
       " 'five_prime_utrs': +--------------+------------+----------------+-----------+-------+\n",
       " | Chromosome   | Source     | Type           | Start     | +23   |\n",
       " | (category)   | (object)   | (object)       | (int64)   | ...   |\n",
       " |--------------+------------+----------------+-----------+-------|\n",
       " | 1            | HAVANA     | five_prime_UTR | 4807822   | ...   |\n",
       " | 1            | HAVANA     | five_prime_UTR | 4857813   | ...   |\n",
       " | 1            | HAVANA     | five_prime_UTR | 5588492   | ...   |\n",
       " | 1            | HAVANA     | five_prime_UTR | 5589034   | ...   |\n",
       " | ...          | ...        | ...            | ...       | ...   |\n",
       " | Y            | HAVANA     | five_prime_UTR | 87550965  | ...   |\n",
       " | Y            | HAVANA     | five_prime_UTR | 87552378  | ...   |\n",
       " | Y            | HAVANA     | five_prime_UTR | 88053314  | ...   |\n",
       " | Y            | HAVANA     | five_prime_UTR | 88054723  | ...   |\n",
       " +--------------+------------+----------------+-----------+-------+\n",
       " Unstranded PyRanges object has 11,449 rows and 27 columns from 21 chromosomes.\n",
       " For printing, the PyRanges was sorted on Chromosome.\n",
       " 23 hidden columns: End, Score, Phase, Id, Parent, Gene_id, Transcript_id, ... (+ 16 more.),\n",
       " 'three_prime_utrs': +--------------+------------+-----------------+-----------+-------+\n",
       " | Chromosome   | Source     | Type            | Start     | +23   |\n",
       " | (category)   | (object)   | (object)        | (int64)   | ...   |\n",
       " |--------------+------------+-----------------+-----------+-------|\n",
       " | 1            | HAVANA     | three_prime_UTR | 4845016   | ...   |\n",
       " | 1            | HAVANA     | three_prime_UTR | 4896364   | ...   |\n",
       " | 1            | HAVANA     | three_prime_UTR | 5602784   | ...   |\n",
       " | 1            | HAVANA     | three_prime_UTR | 6274275   | ...   |\n",
       " | ...          | ...        | ...             | ...       | ...   |\n",
       " | Y            | HAVANA     | three_prime_UTR | 86087758  | ...   |\n",
       " | Y            | HAVANA     | three_prime_UTR | 87142720  | ...   |\n",
       " | Y            | HAVANA     | three_prime_UTR | 87575495  | ...   |\n",
       " | Y            | HAVANA     | three_prime_UTR | 88079297  | ...   |\n",
       " +--------------+------------+-----------------+-----------+-------+\n",
       " Unstranded PyRanges object has 8,075 rows and 27 columns from 21 chromosomes.\n",
       " For printing, the PyRanges was sorted on Chromosome.\n",
       " 23 hidden columns: End, Score, Phase, Id, Parent, Gene_id, Transcript_id, ... (+ 16 more.)}"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict_regions_to_patch_plus = {\n",
    "    \"exons\": exons_plus,\n",
    "    \"introns\": introns_plus,\n",
    "    \"five_prime_utrs\": five_prime_utrs_plus,\n",
    "    \"three_prime_utrs\": three_prime_utrs_plus,\n",
    "}\n",
    "\n",
    "dict_regions_to_patch_minus = {\n",
    "    \"exons\": exons_minus,\n",
    "    \"introns\": introns_minus,\n",
    "    \"five_prime_utrs\": five_prime_utrs_minus,\n",
    "    \"three_prime_utrs\": three_prime_utrs_minus,\n",
    "}\n",
    "\n",
    "dict_regions_to_patch_plus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Build sense and antisense methylation summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarise_methylation_across_regions(plus_strand_cpgs, minus_strand_cpgs, regions_dict, region_strand, sense_methylation={}, antisense_methylation={}):\n",
    "    \"\"\"\n",
    "    Summarise the methylation data across multiple regions.\n",
    "\n",
    "    Parameters\n",
    "        rdr: pyranges object\n",
    "        regions_dict: dict\n",
    "\n",
    "    Returns\n",
    "        grouped_df: pd.DataFrame\n",
    "    \"\"\"\n",
    "    # For each region in + strand expressed genes, summarise the methylation data\n",
    "    for region, gr in regions_dict.items():\n",
    "        rds_plus = plus_strand_cpgs.reduce_byranges(\n",
    "                    ranges=gr.unstrand(), \n",
    "                    var=[\"num_mc\", \"num_hmc\", \"num_modc\", \"num_total_c\"]\n",
    "                )\n",
    "        \n",
    "        rds_minus = minus_strand_cpgs.reduce_byranges(\n",
    "                ranges=gr.unstrand(), \n",
    "                var=[\"num_mc\", \"num_hmc\", \"num_modc\", \"num_total_c\"]\n",
    "            )\n",
    "        \n",
    "        if region_strand == \"+\":\n",
    "            # sense to the + region\n",
    "            sense_plus_ds = rds_plus\n",
    "            plus_sense = ['+'] * sense_plus_ds.dims['ranges']\n",
    "            sense_plus_ds = sense_plus_ds.assign_coords(strand=('ranges', plus_sense))\n",
    "\n",
    "            # antisense to the + region\n",
    "            antisense_plus_ds = rds_minus\n",
    "            plus_antisense = ['-'] * antisense_plus_ds.dims['ranges']\n",
    "            antisense_plus_ds = antisense_plus_ds.assign_coords(strand=('ranges', plus_antisense))\n",
    "            \n",
    "            # Add the modified dataset to the dictionary\n",
    "            if region not in sense_methylation:\n",
    "                sense_methylation[region] = sense_plus_ds\n",
    "                antisense_methylation[region] = antisense_plus_ds\n",
    "            else:\n",
    "                sense_methylation[region] = xr.concat([sense_methylation[region], sense_plus_ds], dim=\"ranges\")\n",
    "                antisense_methylation[region] = xr.concat([antisense_methylation[region], antisense_plus_ds], dim=\"ranges\")\n",
    "        else:\n",
    "            # sense to the - region\n",
    "            sense_minus_ds = rds_minus\n",
    "            minus_sense = ['-'] * sense_minus_ds.dims['ranges']\n",
    "            sense_minus_ds = sense_minus_ds.assign_coords(strand=('ranges', minus_sense))\n",
    "            \n",
    "            # antisense to the - region\n",
    "            antisense_minus_ds = rds_plus\n",
    "            minus_antisense = ['+'] * antisense_minus_ds.dims['ranges']\n",
    "            antisense_minus_ds = antisense_minus_ds.assign_coords(strand=('ranges', minus_antisense))\n",
    "            \n",
    "\n",
    "            # Add the modified dataset to the dictionary\n",
    "            if region not in antisense_methylation:\n",
    "                sense_methylation[region] = sense_minus_ds\n",
    "                antisense_methylation[region] = antisense_minus_ds\n",
    "            else:\n",
    "                sense_methylation[region] = xr.concat([sense_methylation[region], sense_minus_ds], dim=\"ranges\")\n",
    "                antisense_methylation[region] = xr.concat([antisense_methylation[region], antisense_minus_ds], dim=\"ranges\")\n",
    "\n",
    "\n",
    "    return sense_methylation, antisense_methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "sense_methylation, antisense_methylation = summarise_methylation_across_regions(plus_strand_cov, minus_strand_cov, dict_regions_to_patch_plus, \"+\")\n",
    "sense_methylation, antisense_methylation = summarise_methylation_across_regions(plus_strand_cov, minus_strand_cov, dict_regions_to_patch_minus, \"-\", sense_methylation, antisense_methylation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;\n",
       "Dimensions:                   (ranges: 180974, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) &lt;U2 &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;\n",
       "    start                     (ranges) int64 4808454 4828583 ... 89743993\n",
       "    end                       (ranges) int64 4808485 4828648 ... 89744124\n",
       "    range_id                  (ranges) int64 5 9 13 14 ... 408327 408328 408329\n",
       "    num_contexts              (ranges) int64 1 0 0 0 0 2 2 0 ... 0 0 0 0 1 0 0 0\n",
       "    range_length              (ranges) int64 31 65 65 47 47 ... 34 66 34 66 131\n",
       "  * sample_id                 (sample_id) &lt;U8 &#x27;sample_0&#x27;\n",
       "    strand                    (ranges) &lt;U1 &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 113.0 0.0 ... 0.0 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 113.0 nan ... nan nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    Source                    (ranges) object &#x27;HAVANA&#x27; &#x27;HAVANA&#x27; ... &#x27;HAVANA&#x27;\n",
       "    Type                      (ranges) object &#x27;exon&#x27; &#x27;exon&#x27; ... &#x27;exon&#x27; &#x27;exon&#x27;\n",
       "    Score                     (ranges) object &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;\n",
       "    Phase                     (ranges) object &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;\n",
       "    Id                        (ranges) object &#x27;exon:ENSMUST00000027036.10:2&#x27; ...\n",
       "    Parent                    (ranges) object &#x27;ENSMUST00000027036.10&#x27; ... &#x27;EN...\n",
       "    Gene_id                   (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Transcript_id             (ranges) object &#x27;ENSMUST00000027036.10&#x27; ... &#x27;EN...\n",
       "    Gene_type                 (ranges) object &#x27;protein_coding&#x27; ... &#x27;protein_c...\n",
       "    Gene_name                 (ranges) object &#x27;Lypla1&#x27; &#x27;Lypla1&#x27; ... &#x27;Gm21996&#x27;\n",
       "    Transcript_type           (ranges) object &#x27;protein_coding&#x27; ... &#x27;protein_c...\n",
       "    Transcript_name           (ranges) object &#x27;Lypla1-201&#x27; ... &#x27;Gm21996-201&#x27;\n",
       "    Exon_number               (ranges) object &#x27;2&#x27; &#x27;3&#x27; &#x27;2&#x27; &#x27;4&#x27; ... &#x27;4&#x27; &#x27;3&#x27; &#x27;2&#x27;\n",
       "    Exon_id                   (ranges) object &#x27;ENSMUSE00001255376.1&#x27; ... &#x27;ENS...\n",
       "    Level                     (ranges) object &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;\n",
       "    Transcript_support_level  (ranges) object &#x27;1&#x27; &#x27;1&#x27; &#x27;5&#x27; &#x27;1&#x27; ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;\n",
       "    Mgi_id                    (ranges) object &#x27;MGI:1344588&#x27; ... &#x27;MGI:5440224&#x27;\n",
       "    Tag                       (ranges) object &#x27;basic,appris_principal_1,CCDS&#x27;...\n",
       "    Havana_gene               (ranges) object &#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTT...\n",
       "    Havana_transcript         (ranges) object &#x27;OTTMUST00000051162.1&#x27; ... &#x27;OTT...\n",
       "    Protein_id                (ranges) object &#x27;ENSMUSP00000027036.4&#x27; ... &#x27;ENS...\n",
       "    Ccdsid                    (ranges) object &#x27;CCDS14806.1&#x27; &#x27;CCDS14806.1&#x27; ... &#x27;&#x27;\n",
       "    Ont                       (ranges) object &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; ... &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27;</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-ff0ecc66-9b6e-4f6b-bfa7-0b3b86aecf44' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-ff0ecc66-9b6e-4f6b-bfa7-0b3b86aecf44' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>ranges</span>: 180974</li><li><span class='xr-has-index'>sample_id</span>: 1</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-711241ae-70ce-43ec-bb49-bd60dd731d71' class='xr-section-summary-in' type='checkbox'  checked><label for='section-711241ae-70ce-43ec-bb49-bd60dd731d71' class='xr-section-summary' >Coordinates: <span>(8)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>contig</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U2</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;</div><input id='attrs-29e0540b-0da3-4d4b-b443-fde3a4742c7b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-29e0540b-0da3-4d4b-b443-fde3a4742c7b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-66ea0484-7e72-472b-83df-981fcf4257cf' class='xr-var-data-in' type='checkbox'><label for='data-66ea0484-7e72-472b-83df-981fcf4257cf' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;, ..., &#x27;Y&#x27;, &#x27;Y&#x27;, &#x27;Y&#x27;], dtype=&#x27;&lt;U2&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>start</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4808454 4828583 ... 89743993</div><input id='attrs-ef619465-290c-4ace-a98a-4938488633f2' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ef619465-290c-4ace-a98a-4938488633f2' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b8a05710-a0b4-448f-b02b-16afc8d0c6a0' class='xr-var-data-in' type='checkbox'><label for='data-b8a05710-a0b4-448f-b02b-16afc8d0c6a0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4808454,  4828583,  4828583, ..., 89731936, 89732605, 89743993])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>end</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4808485 4828648 ... 89744124</div><input id='attrs-76c6f7cf-2dd4-44e3-9c8f-1e9e38a0b94a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-76c6f7cf-2dd4-44e3-9c8f-1e9e38a0b94a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1c485d32-17a6-4830-b738-db26464366df' class='xr-var-data-in' type='checkbox'><label for='data-1c485d32-17a6-4830-b738-db26464366df' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4808485,  4828648,  4828648, ..., 89731970, 89732671, 89744124])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>5 9 13 14 ... 408327 408328 408329</div><input id='attrs-b1ad56c0-3a3d-4427-abec-8273212f1255' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b1ad56c0-3a3d-4427-abec-8273212f1255' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-6b8696fb-87b0-455b-9826-d24183f9b2da' class='xr-var-data-in' type='checkbox'><label for='data-6b8696fb-87b0-455b-9826-d24183f9b2da' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([     5,      9,     13, ..., 408327, 408328, 408329])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_contexts</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>1 0 0 0 0 2 2 0 ... 0 0 0 0 1 0 0 0</div><input id='attrs-e7ca2fc9-810f-4571-a7ec-e0fa652d5dc0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e7ca2fc9-810f-4571-a7ec-e0fa652d5dc0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-298ca3d1-d5be-4b47-815d-c63bd38fd234' class='xr-var-data-in' type='checkbox'><label for='data-298ca3d1-d5be-4b47-815d-c63bd38fd234' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([1, 0, 0, ..., 0, 0, 0])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_length</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>31 65 65 47 47 ... 34 66 34 66 131</div><input id='attrs-42399218-ffd8-4ac0-8527-d5fbe9d22cd1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-42399218-ffd8-4ac0-8527-d5fbe9d22cd1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8aaaddae-555a-4267-bffe-62d30cdef440' class='xr-var-data-in' type='checkbox'><label for='data-8aaaddae-555a-4267-bffe-62d30cdef440' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 31,  65,  65, ...,  34,  66, 131])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample_id</span></div><div class='xr-var-dims'>(sample_id)</div><div class='xr-var-dtype'>&lt;U8</div><div class='xr-var-preview xr-preview'>&#x27;sample_0&#x27;</div><input id='attrs-00d4b461-6918-4406-b1d9-eebfdb75a300' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-00d4b461-6918-4406-b1d9-eebfdb75a300' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4a0b5032-6d34-4b24-b324-1e699b565746' class='xr-var-data-in' type='checkbox'><label for='data-4a0b5032-6d34-4b24-b324-1e699b565746' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;sample_0&#x27;], dtype=&#x27;&lt;U8&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U1</div><div class='xr-var-preview xr-preview'>&#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;</div><input id='attrs-b62fe158-dc67-464e-9c1b-c4a2dce5e749' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b62fe158-dc67-464e-9c1b-c4a2dce5e749' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d78c529e-e547-4b39-b5cf-65d96d375d9b' class='xr-var-data-in' type='checkbox'><label for='data-d78c529e-e547-4b39-b5cf-65d96d375d9b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;+&#x27;, &#x27;+&#x27;, &#x27;+&#x27;, ..., &#x27;-&#x27;, &#x27;-&#x27;, &#x27;-&#x27;], dtype=&#x27;&lt;U1&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-b92cc8ec-d654-472d-84c9-1c4b4056e671' class='xr-section-summary-in' type='checkbox'  ><label for='section-b92cc8ec-d654-472d-84c9-1c4b4056e671' class='xr-section-summary' >Data variables: <span>(35)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0</div><input id='attrs-8c65985e-97ae-4b30-af93-b46664ee0446' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8c65985e-97ae-4b30-af93-b46664ee0446' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-cf5cb7ec-793d-41f6-8b80-0c0470fc1b02' class='xr-var-data-in' type='checkbox'><label for='data-cf5cb7ec-793d-41f6-8b80-0c0470fc1b02' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 nan nan nan ... nan nan nan</div><input id='attrs-babf7712-6031-4fe9-afc8-582273b3f2a3' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-babf7712-6031-4fe9-afc8-582273b3f2a3' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e0cf8956-db1a-48e2-867b-f18eb948cdf7' class='xr-var-data-in' type='checkbox'><label for='data-e0cf8956-db1a-48e2-867b-f18eb948cdf7' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [nan],\n",
       "       [nan],\n",
       "       ...,\n",
       "       [nan],\n",
       "       [nan],\n",
       "       [nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 1.0 0.0 0.0 0.0</div><input id='attrs-31de348d-e94e-4530-8b6d-23d6f9fe3e4f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-31de348d-e94e-4530-8b6d-23d6f9fe3e4f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9cbac08c-c9dd-4a96-90f3-a7ca65a17fed' class='xr-var-data-in' type='checkbox'><label for='data-9cbac08c-c9dd-4a96-90f3-a7ca65a17fed' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0</div><input id='attrs-4229eea5-55a6-4bc7-9c11-fb95ae1281b0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4229eea5-55a6-4bc7-9c11-fb95ae1281b0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0e1a5813-10f0-4977-98fc-8df477405e0c' class='xr-var-data-in' type='checkbox'><label for='data-0e1a5813-10f0-4977-98fc-8df477405e0c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 nan nan nan ... 0.0 nan nan nan</div><input id='attrs-f045a8ab-7666-495d-8988-542d8cae94e1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f045a8ab-7666-495d-8988-542d8cae94e1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-98413805-bea8-4986-9d76-f2064e47a861' class='xr-var-data-in' type='checkbox'><label for='data-98413805-bea8-4986-9d76-f2064e47a861' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [nan],\n",
       "       [nan],\n",
       "       ...,\n",
       "       [nan],\n",
       "       [nan],\n",
       "       [nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 1.0 0.0 0.0 0.0</div><input id='attrs-8524e164-b769-4c7b-9dc2-6725faf1b0df' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8524e164-b769-4c7b-9dc2-6725faf1b0df' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-09749a79-28c5-42c2-936a-f0d98090271d' class='xr-var-data-in' type='checkbox'><label for='data-09749a79-28c5-42c2-936a-f0d98090271d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0</div><input id='attrs-b000501f-33b5-4ba7-936d-806660a58883' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b000501f-33b5-4ba7-936d-806660a58883' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-09758059-dd13-4863-9394-f7ca1f214bb4' class='xr-var-data-in' type='checkbox'><label for='data-09758059-dd13-4863-9394-f7ca1f214bb4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 nan nan nan ... nan nan nan</div><input id='attrs-c9d1aefa-4541-4a50-9339-d84242b567cb' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c9d1aefa-4541-4a50-9339-d84242b567cb' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-6c2d5b7c-c06e-4a4d-ae8e-22242b653988' class='xr-var-data-in' type='checkbox'><label for='data-6c2d5b7c-c06e-4a4d-ae8e-22242b653988' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [nan],\n",
       "       [nan],\n",
       "       ...,\n",
       "       [nan],\n",
       "       [nan],\n",
       "       [nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 1.0 0.0 0.0 0.0</div><input id='attrs-653c8a4d-ce98-4489-8649-de75959ab1c0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-653c8a4d-ce98-4489-8649-de75959ab1c0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d12e483d-e566-4845-83bd-c0cb18f6b42d' class='xr-var-data-in' type='checkbox'><label for='data-d12e483d-e566-4845-83bd-c0cb18f6b42d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>113.0 0.0 0.0 0.0 ... 0.0 0.0 0.0</div><input id='attrs-1a8c044b-fc8a-47a3-926b-745e6595273c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1a8c044b-fc8a-47a3-926b-745e6595273c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5667b687-2009-40e0-adb4-a570e71f214e' class='xr-var-data-in' type='checkbox'><label for='data-5667b687-2009-40e0-adb4-a570e71f214e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[113.],\n",
       "       [  0.],\n",
       "       [  0.],\n",
       "       ...,\n",
       "       [  0.],\n",
       "       [  0.],\n",
       "       [  0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>113.0 nan nan nan ... nan nan nan</div><input id='attrs-8ac3b239-35bb-44c6-a61c-aee4d17de266' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8ac3b239-35bb-44c6-a61c-aee4d17de266' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-eef3770b-dce9-40ee-b7e9-9ea862681c08' class='xr-var-data-in' type='checkbox'><label for='data-eef3770b-dce9-40ee-b7e9-9ea862681c08' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[113.],\n",
       "       [ nan],\n",
       "       [ nan],\n",
       "       ...,\n",
       "       [ nan],\n",
       "       [ nan],\n",
       "       [ nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 0.0 0.0 0.0 ... 1.0 0.0 0.0 0.0</div><input id='attrs-b9975d6f-7dfa-471f-9d0a-0c82fc792d86' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b9975d6f-7dfa-471f-9d0a-0c82fc792d86' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-242c3af2-7c15-4410-b0fa-6f26bb28a6c5' class='xr-var-data-in' type='checkbox'><label for='data-242c3af2-7c15-4410-b0fa-6f26bb28a6c5' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [0.],\n",
       "       [0.],\n",
       "       ...,\n",
       "       [0.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Source</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;HAVANA&#x27; &#x27;HAVANA&#x27; ... &#x27;HAVANA&#x27;</div><input id='attrs-0baa9aab-df06-428a-ae63-8efefe2e39b6' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0baa9aab-df06-428a-ae63-8efefe2e39b6' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-83519934-faf4-4b21-9919-84186aab9c3a' class='xr-var-data-in' type='checkbox'><label for='data-83519934-faf4-4b21-9919-84186aab9c3a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;, ..., &#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;, &#x27;HAVANA&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;exon&#x27; &#x27;exon&#x27; ... &#x27;exon&#x27; &#x27;exon&#x27;</div><input id='attrs-3740831e-98ef-4885-80da-184bd62c277a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3740831e-98ef-4885-80da-184bd62c277a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3619729e-eead-4030-a79d-2380342223e0' class='xr-var-data-in' type='checkbox'><label for='data-3619729e-eead-4030-a79d-2380342223e0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;exon&#x27;, &#x27;exon&#x27;, &#x27;exon&#x27;, ..., &#x27;exon&#x27;, &#x27;exon&#x27;, &#x27;exon&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Score</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;</div><input id='attrs-b3191eb2-b295-4c54-9bd5-9f2d5932dac8' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b3191eb2-b295-4c54-9bd5-9f2d5932dac8' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1e7cae54-5d0e-4ada-8295-34d125cd920e' class='xr-var-data-in' type='checkbox'><label for='data-1e7cae54-5d0e-4ada-8295-34d125cd920e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;, ..., &#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Phase</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;</div><input id='attrs-c45edfe1-4a57-4e89-8226-a6fe7ce8266d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c45edfe1-4a57-4e89-8226-a6fe7ce8266d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ccdf3346-5a95-4b2e-83d6-fc253bc15cb3' class='xr-var-data-in' type='checkbox'><label for='data-ccdf3346-5a95-4b2e-83d6-fc253bc15cb3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;, ..., &#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;exon:ENSMUST00000027036.10:2&#x27; ....</div><input id='attrs-211320ee-b57a-412a-b7eb-ce0c08fd7830' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-211320ee-b57a-412a-b7eb-ce0c08fd7830' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ddd5a57a-3ca3-4d3d-9cb4-2a09318970c5' class='xr-var-data-in' type='checkbox'><label for='data-ddd5a57a-3ca3-4d3d-9cb4-2a09318970c5' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;exon:ENSMUST00000027036.10:2&#x27;, &#x27;exon:ENSMUST00000027036.10:3&#x27;,\n",
       "       &#x27;exon:ENSMUST00000155020.1:2&#x27;, ..., &#x27;exon:ENSMUST00000188269.1:4&#x27;,\n",
       "       &#x27;exon:ENSMUST00000188269.1:3&#x27;, &#x27;exon:ENSMUST00000188269.1:2&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Parent</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUST00000027036.10&#x27; ... &#x27;ENS...</div><input id='attrs-2956497d-5ed9-4fd3-b7ef-05dc02c5186a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2956497d-5ed9-4fd3-b7ef-05dc02c5186a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-99544fba-f0c5-4645-bbd9-f6def65c2a8d' class='xr-var-data-in' type='checkbox'><label for='data-99544fba-f0c5-4645-bbd9-f6def65c2a8d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUST00000027036.10&#x27;, &#x27;ENSMUST00000027036.10&#x27;,\n",
       "       &#x27;ENSMUST00000155020.1&#x27;, ..., &#x27;ENSMUST00000188269.1&#x27;,\n",
       "       &#x27;ENSMUST00000188269.1&#x27;, &#x27;ENSMUST00000188269.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-30dfc0ee-ae4e-4e18-8173-bb10d46e8ae6' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-30dfc0ee-ae4e-4e18-8173-bb10d46e8ae6' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a3fec75b-5271-4505-b06a-0929f986afa8' class='xr-var-data-in' type='checkbox'><label for='data-a3fec75b-5271-4505-b06a-0929f986afa8' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000025903.14&#x27;,\n",
       "       &#x27;ENSMUSG00000104217.1&#x27;, ..., &#x27;ENSMUSG00000100608.1&#x27;,\n",
       "       &#x27;ENSMUSG00000100608.1&#x27;, &#x27;ENSMUSG00000100608.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUST00000027036.10&#x27; ... &#x27;ENS...</div><input id='attrs-772563d2-22cf-4484-a86d-475260ca6e05' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-772563d2-22cf-4484-a86d-475260ca6e05' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-69a78739-317a-4e36-beac-ae023ea128e4' class='xr-var-data-in' type='checkbox'><label for='data-69a78739-317a-4e36-beac-ae023ea128e4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUST00000027036.10&#x27;, &#x27;ENSMUST00000027036.10&#x27;,\n",
       "       &#x27;ENSMUST00000155020.1&#x27;, ..., &#x27;ENSMUST00000188269.1&#x27;,\n",
       "       &#x27;ENSMUST00000188269.1&#x27;, &#x27;ENSMUST00000188269.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;protein_coding&#x27; ... &#x27;protein_co...</div><input id='attrs-83b66ea0-4d6b-4d73-90d4-cbe27d41231a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-83b66ea0-4d6b-4d73-90d4-cbe27d41231a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-708fb7a3-85c4-4c3e-bbbb-eca0da740e24' class='xr-var-data-in' type='checkbox'><label for='data-708fb7a3-85c4-4c3e-bbbb-eca0da740e24' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, ...,\n",
       "       &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Lypla1&#x27; &#x27;Lypla1&#x27; ... &#x27;Gm21996&#x27;</div><input id='attrs-d985a492-bba7-400f-8dd1-2604be82f450' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d985a492-bba7-400f-8dd1-2604be82f450' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8d34707d-a791-4dc6-bd14-135d057005ec' class='xr-var-data-in' type='checkbox'><label for='data-8d34707d-a791-4dc6-bd14-135d057005ec' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Lypla1&#x27;, &#x27;Lypla1&#x27;, &#x27;Gm37988&#x27;, ..., &#x27;Gm21996&#x27;, &#x27;Gm21996&#x27;,\n",
       "       &#x27;Gm21996&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;protein_coding&#x27; ... &#x27;protein_co...</div><input id='attrs-5f2a73b5-a667-4fd3-9db4-e132db3c0d16' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5f2a73b5-a667-4fd3-9db4-e132db3c0d16' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9c4d3638-3277-4474-ab63-c187568cda03' class='xr-var-data-in' type='checkbox'><label for='data-9c4d3638-3277-4474-ab63-c187568cda03' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;nonsense_mediated_decay&#x27;, ...,\n",
       "       &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Lypla1-201&#x27; ... &#x27;Gm21996-201&#x27;</div><input id='attrs-90a9e6b1-ea74-490a-aaed-84fc149380b8' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-90a9e6b1-ea74-490a-aaed-84fc149380b8' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7d0756bc-36e7-4107-9a49-658f0f94c43b' class='xr-var-data-in' type='checkbox'><label for='data-7d0756bc-36e7-4107-9a49-658f0f94c43b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Lypla1-201&#x27;, &#x27;Lypla1-201&#x27;, &#x27;Gm37988-201&#x27;, ..., &#x27;Gm21996-201&#x27;,\n",
       "       &#x27;Gm21996-201&#x27;, &#x27;Gm21996-201&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;2&#x27; &#x27;3&#x27; &#x27;2&#x27; &#x27;4&#x27; ... &#x27;5&#x27; &#x27;4&#x27; &#x27;3&#x27; &#x27;2&#x27;</div><input id='attrs-2fc82528-7eb9-441c-bbc8-eeb0159fec7f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2fc82528-7eb9-441c-bbc8-eeb0159fec7f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ad182daf-98e9-4504-b8b3-0b718b14eac0' class='xr-var-data-in' type='checkbox'><label for='data-ad182daf-98e9-4504-b8b3-0b718b14eac0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;2&#x27;, &#x27;3&#x27;, &#x27;2&#x27;, ..., &#x27;4&#x27;, &#x27;3&#x27;, &#x27;2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSE00001255376.1&#x27; ... &#x27;ENSM...</div><input id='attrs-19043b29-c2b8-44f9-a044-a9400adf8989' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-19043b29-c2b8-44f9-a044-a9400adf8989' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-86906730-829d-41c1-8ca1-93b66f6bd8d3' class='xr-var-data-in' type='checkbox'><label for='data-86906730-829d-41c1-8ca1-93b66f6bd8d3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSE00001255376.1&#x27;, &#x27;ENSMUSE00001222304.1&#x27;,\n",
       "       &#x27;ENSMUSE00001241075.1&#x27;, ..., &#x27;ENSMUSE00001333847.1&#x27;,\n",
       "       &#x27;ENSMUSE00001331494.1&#x27;, &#x27;ENSMUSE00001332880.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;</div><input id='attrs-6060e53a-f384-4d30-9966-e5439fc0115f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6060e53a-f384-4d30-9966-e5439fc0115f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c65db9ee-ebf5-4b5c-bdbb-38c7192142b2' class='xr-var-data-in' type='checkbox'><label for='data-c65db9ee-ebf5-4b5c-bdbb-38c7192142b2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;, ..., &#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_support_level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;5&#x27; &#x27;1&#x27; ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;</div><input id='attrs-afbf5053-71ca-4f96-9ca2-885c1ee9dcac' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-afbf5053-71ca-4f96-9ca2-885c1ee9dcac' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-eb4b1cf2-68a4-4dbe-a2c1-f87d7e0f77a1' class='xr-var-data-in' type='checkbox'><label for='data-eb4b1cf2-68a4-4dbe-a2c1-f87d7e0f77a1' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;5&#x27;, ..., &#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Mgi_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;MGI:1344588&#x27; ... &#x27;MGI:5440224&#x27;</div><input id='attrs-32038a94-cac1-4b8a-9674-63840b4924f9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-32038a94-cac1-4b8a-9674-63840b4924f9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-dcd90594-c253-4e67-9433-864731787434' class='xr-var-data-in' type='checkbox'><label for='data-dcd90594-c253-4e67-9433-864731787434' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;MGI:1344588&#x27;, &#x27;MGI:1344588&#x27;, &#x27;MGI:5611216&#x27;, ..., &#x27;MGI:5440224&#x27;,\n",
       "       &#x27;MGI:5440224&#x27;, &#x27;MGI:5440224&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Tag</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;basic,appris_principal_1,CCDS&#x27; ...</div><input id='attrs-cb528e44-8237-4a6c-8aea-282dcd8a47c0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-cb528e44-8237-4a6c-8aea-282dcd8a47c0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-de20ffd4-d9a5-418a-b680-c0e9219d699c' class='xr-var-data-in' type='checkbox'><label for='data-de20ffd4-d9a5-418a-b680-c0e9219d699c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;basic,appris_principal_1,CCDS&#x27;, &#x27;basic,appris_principal_1,CCDS&#x27;,\n",
       "       &#x27;basic,appris_principal_1,readthrough_transcript&#x27;, ...,\n",
       "       &#x27;basic,appris_principal_1&#x27;, &#x27;basic,appris_principal_1&#x27;,\n",
       "       &#x27;basic,appris_principal_1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_gene</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTTM...</div><input id='attrs-381cbbaa-54d6-43e0-abbc-b18c8bc62843' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-381cbbaa-54d6-43e0-abbc-b18c8bc62843' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-2f773747-e93e-452e-b6cc-51f3adc5ea4c' class='xr-var-data-in' type='checkbox'><label for='data-2f773747-e93e-452e-b6cc-51f3adc5ea4c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;OTTMUSG00000021562.4&#x27;, &#x27;OTTMUSG00000021562.4&#x27;,\n",
       "       &#x27;OTTMUSG00000050100.1&#x27;, ..., &#x27;OTTMUSG00000047316.1&#x27;,\n",
       "       &#x27;OTTMUSG00000047316.1&#x27;, &#x27;OTTMUSG00000047316.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_transcript</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;OTTMUST00000051162.1&#x27; ... &#x27;OTTM...</div><input id='attrs-aa9e0ec6-f00e-440e-b813-e493549c9884' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-aa9e0ec6-f00e-440e-b813-e493549c9884' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3eeb5b80-90a9-4a8b-9ee4-be4a70b2adb7' class='xr-var-data-in' type='checkbox'><label for='data-3eeb5b80-90a9-4a8b-9ee4-be4a70b2adb7' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;OTTMUST00000051162.1&#x27;, &#x27;OTTMUST00000051162.1&#x27;,\n",
       "       &#x27;OTTMUST00000127419.1&#x27;, ..., &#x27;OTTMUST00000122318.1&#x27;,\n",
       "       &#x27;OTTMUST00000122318.1&#x27;, &#x27;OTTMUST00000122318.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Protein_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSP00000027036.4&#x27; ... &#x27;ENSM...</div><input id='attrs-0ce296fe-0a88-422e-9273-4a6ab375db21' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0ce296fe-0a88-422e-9273-4a6ab375db21' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-186f0ffa-2ea4-4b94-88e2-9c3eb7553809' class='xr-var-data-in' type='checkbox'><label for='data-186f0ffa-2ea4-4b94-88e2-9c3eb7553809' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSP00000027036.4&#x27;, &#x27;ENSMUSP00000027036.4&#x27;,\n",
       "       &#x27;ENSMUSP00000136108.1&#x27;, ..., &#x27;ENSMUSP00000141173.1&#x27;,\n",
       "       &#x27;ENSMUSP00000141173.1&#x27;, &#x27;ENSMUSP00000141173.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ccdsid</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;CCDS14806.1&#x27; &#x27;CCDS14806.1&#x27; ... &#x27;&#x27;</div><input id='attrs-8660a87e-3354-429e-bdf1-193ae4228bdd' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8660a87e-3354-429e-bdf1-193ae4228bdd' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-552ba290-5160-40b6-8e90-69efc709d35f' class='xr-var-data-in' type='checkbox'><label for='data-552ba290-5160-40b6-8e90-69efc709d35f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;CCDS14806.1&#x27;, &#x27;CCDS14806.1&#x27;, &#x27;&#x27;, ..., &#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ont</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; ... &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27; &#x27;&#x27;</div><input id='attrs-5151416d-b225-4e97-9256-88be1c44e0d9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5151416d-b225-4e97-9256-88be1c44e0d9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7f123c78-344f-489e-88f1-ac417967a6de' class='xr-var-data-in' type='checkbox'><label for='data-7f123c78-344f-489e-88f1-ac417967a6de' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;, ..., &#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;], dtype=object)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-038e55c1-f9d4-45c7-ab01-e9b35333e571' class='xr-section-summary-in' type='checkbox'  ><label for='section-038e55c1-f9d4-45c7-ab01-e9b35333e571' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>sample_id</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-db1a3935-482b-46ca-b5d5-69ba7c683506' class='xr-index-data-in' type='checkbox'/><label for='index-db1a3935-482b-46ca-b5d5-69ba7c683506' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;sample_0&#x27;], dtype=&#x27;object&#x27;, name=&#x27;sample_id&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-ca91a99d-7ee4-4ace-8f9d-fdb7c6c584dc' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-ca91a99d-7ee4-4ace-8f9d-fdb7c6c584dc' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.Dataset>\n",
       "Dimensions:                   (ranges: 180974, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) <U2 '1' '1' '1' '1' ... 'Y' 'Y' 'Y' 'Y'\n",
       "    start                     (ranges) int64 4808454 4828583 ... 89743993\n",
       "    end                       (ranges) int64 4808485 4828648 ... 89744124\n",
       "    range_id                  (ranges) int64 5 9 13 14 ... 408327 408328 408329\n",
       "    num_contexts              (ranges) int64 1 0 0 0 0 2 2 0 ... 0 0 0 0 1 0 0 0\n",
       "    range_length              (ranges) int64 31 65 65 47 47 ... 34 66 34 66 131\n",
       "  * sample_id                 (sample_id) <U8 'sample_0'\n",
       "    strand                    (ranges) <U1 '+' '+' '+' '+' ... '-' '-' '-' '-'\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 0.0 0.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.0 nan ... nan nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 113.0 0.0 ... 0.0 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 113.0 nan ... nan nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 1.0 0.0 ... 0.0 0.0\n",
       "    Source                    (ranges) object 'HAVANA' 'HAVANA' ... 'HAVANA'\n",
       "    Type                      (ranges) object 'exon' 'exon' ... 'exon' 'exon'\n",
       "    Score                     (ranges) object '.' '.' '.' '.' ... '.' '.' '.'\n",
       "    Phase                     (ranges) object '.' '.' '.' '.' ... '.' '.' '.'\n",
       "    Id                        (ranges) object 'exon:ENSMUST00000027036.10:2' ...\n",
       "    Parent                    (ranges) object 'ENSMUST00000027036.10' ... 'EN...\n",
       "    Gene_id                   (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Transcript_id             (ranges) object 'ENSMUST00000027036.10' ... 'EN...\n",
       "    Gene_type                 (ranges) object 'protein_coding' ... 'protein_c...\n",
       "    Gene_name                 (ranges) object 'Lypla1' 'Lypla1' ... 'Gm21996'\n",
       "    Transcript_type           (ranges) object 'protein_coding' ... 'protein_c...\n",
       "    Transcript_name           (ranges) object 'Lypla1-201' ... 'Gm21996-201'\n",
       "    Exon_number               (ranges) object '2' '3' '2' '4' ... '4' '3' '2'\n",
       "    Exon_id                   (ranges) object 'ENSMUSE00001255376.1' ... 'ENS...\n",
       "    Level                     (ranges) object '2' '2' '2' '2' ... '2' '2' '2'\n",
       "    Transcript_support_level  (ranges) object '1' '1' '5' '1' ... '1' '1' '1'\n",
       "    Mgi_id                    (ranges) object 'MGI:1344588' ... 'MGI:5440224'\n",
       "    Tag                       (ranges) object 'basic,appris_principal_1,CCDS'...\n",
       "    Havana_gene               (ranges) object 'OTTMUSG00000021562.4' ... 'OTT...\n",
       "    Havana_transcript         (ranges) object 'OTTMUST00000051162.1' ... 'OTT...\n",
       "    Protein_id                (ranges) object 'ENSMUSP00000027036.4' ... 'ENS...\n",
       "    Ccdsid                    (ranges) object 'CCDS14806.1' 'CCDS14806.1' ... ''\n",
       "    Ont                       (ranges) object '' '' '' '' '' ... '' '' '' '' ''"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sense_methylation['exons']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "calculate methlyation fractions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>total_c</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.807980</td>\n",
       "      <td>0.025436</td>\n",
       "      <td>0.847215</td>\n",
       "      <td>51.0</td>\n",
       "      <td>6015.0</td>\n",
       "      <td>2995</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003.15</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.014888</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>7.0</td>\n",
       "      <td>403.0</td>\n",
       "      <td>681</td>\n",
       "      <td>Pbsn</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000003.13</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028.15</td>\n",
       "      <td>0.581295</td>\n",
       "      <td>0.031589</td>\n",
       "      <td>0.619697</td>\n",
       "      <td>60.0</td>\n",
       "      <td>6458.0</td>\n",
       "      <td>1955</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000028.13</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000037.17</td>\n",
       "      <td>0.758328</td>\n",
       "      <td>0.017636</td>\n",
       "      <td>0.786414</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1531.0</td>\n",
       "      <td>2777</td>\n",
       "      <td>Scml2</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000101113.8</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000049.11</td>\n",
       "      <td>0.783618</td>\n",
       "      <td>0.036519</td>\n",
       "      <td>0.835495</td>\n",
       "      <td>28.0</td>\n",
       "      <td>2930.0</td>\n",
       "      <td>1068</td>\n",
       "      <td>Apoh</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000049.5</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  total_c  \\\n",
       "0   ENSMUSG00000000001.4  0.807980  0.025436   0.847215       51.0   6015.0   \n",
       "1  ENSMUSG00000000003.15  0.774194  0.014888   0.806452        7.0    403.0   \n",
       "2  ENSMUSG00000000028.15  0.581295  0.031589   0.619697       60.0   6458.0   \n",
       "3  ENSMUSG00000000037.17  0.758328  0.017636   0.786414       30.0   1531.0   \n",
       "4  ENSMUSG00000000049.11  0.783618  0.036519   0.835495       28.0   2930.0   \n",
       "\n",
       "   range_length Gene_name contig strand          Transcript_id Region  \n",
       "0          2995     Gnai3      3      -   ENSMUST00000000001.4  exons  \n",
       "1           681      Pbsn      X      -  ENSMUST00000000003.13  exons  \n",
       "2          1955     Cdc45     16      -  ENSMUST00000000028.13  exons  \n",
       "3          2777     Scml2      X      +   ENSMUST00000101113.8  exons  \n",
       "4          1068      Apoh     11      +   ENSMUST00000000049.5  exons  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict_df_sense = {}\n",
    "dict_df_antisense = {}\n",
    "\n",
    "for region in dict_regions_to_patch_plus.keys():\n",
    "    dict_df_sense[region] = summarise_across_region(sense_methylation[region], region)\n",
    "    dict_df_antisense[region] = summarise_across_region(antisense_methylation[region], region)\n",
    "\n",
    "stitched_sense_regions = pd.concat(dict_df_sense.values())\n",
    "stitched_antisense_regions = pd.concat(dict_df_antisense.values())\n",
    "\n",
    "\n",
    "stitched_sense_regions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(69769, 12)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stitched_sense_regions.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "dropping regions with no coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "summarised regions on sense strand : (69769, 12)\n",
      "of which has no coverage: 4587\n",
      "proportion no coverage: 0.06574553168312575\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>total_c</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSMUSG00000000325.16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4406</td>\n",
       "      <td>Arvcf</td>\n",
       "      <td>16</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000115614.8</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>ENSMUSG00000000326.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1365</td>\n",
       "      <td>Comt</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000115609.9</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>ENSMUSG00000000791.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2688</td>\n",
       "      <td>Il12rb1</td>\n",
       "      <td>8</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000808.7</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>ENSMUSG00000002104.11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1251</td>\n",
       "      <td>Rapsn</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000050323.5</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>ENSMUSG00000002910.11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1755</td>\n",
       "      <td>Arrdc2</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000002989.10</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15443</th>\n",
       "      <td>ENSMUSG00000115219.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>333</td>\n",
       "      <td>Eef1akmt4</td>\n",
       "      <td>16</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000115522.9</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15453</th>\n",
       "      <td>ENSMUSG00000116780.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>Gm3417</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000097399.5</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15454</th>\n",
       "      <td>ENSMUSG00000116953.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100</td>\n",
       "      <td>9030025P20Rik</td>\n",
       "      <td>17</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000061688.10</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15455</th>\n",
       "      <td>ENSMUSG00000117081.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>62</td>\n",
       "      <td>Gm49948</td>\n",
       "      <td>16</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000113795.7</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15463</th>\n",
       "      <td>ENSMUSG00000118346.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>120</td>\n",
       "      <td>Tmem179b</td>\n",
       "      <td>19</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000010249.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4587 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_id  mean_mc  mean_hmc  mean_modc  cpg_count  \\\n",
       "59     ENSMUSG00000000325.16      NaN       NaN        NaN        0.0   \n",
       "60     ENSMUSG00000000326.13      NaN       NaN        NaN        0.0   \n",
       "134     ENSMUSG00000000791.9      NaN       NaN        NaN        0.0   \n",
       "376    ENSMUSG00000002104.11      NaN       NaN        NaN        0.0   \n",
       "497    ENSMUSG00000002910.11      NaN       NaN        NaN        0.0   \n",
       "...                      ...      ...       ...        ...        ...   \n",
       "15443   ENSMUSG00000115219.2      NaN       NaN        NaN        0.0   \n",
       "15453   ENSMUSG00000116780.1      NaN       NaN        NaN        0.0   \n",
       "15454   ENSMUSG00000116953.1      NaN       NaN        NaN        0.0   \n",
       "15455   ENSMUSG00000117081.1      NaN       NaN        NaN        0.0   \n",
       "15463   ENSMUSG00000118346.1      NaN       NaN        NaN        0.0   \n",
       "\n",
       "       total_c  range_length      Gene_name contig strand  \\\n",
       "59         0.0          4406          Arvcf     16      +   \n",
       "60         0.0          1365           Comt     16      -   \n",
       "134        0.0          2688        Il12rb1      8      +   \n",
       "376        0.0          1251          Rapsn      2      +   \n",
       "497        0.0          1755         Arrdc2      8      -   \n",
       "...        ...           ...            ...    ...    ...   \n",
       "15443      0.0           333      Eef1akmt4     16      +   \n",
       "15453      0.0             2         Gm3417     17      -   \n",
       "15454      0.0           100  9030025P20Rik     17      +   \n",
       "15455      0.0            62        Gm49948     16      +   \n",
       "15463      0.0           120       Tmem179b     19      -   \n",
       "\n",
       "               Transcript_id            Region  \n",
       "59      ENSMUST00000115614.8             exons  \n",
       "60      ENSMUST00000115609.9             exons  \n",
       "134     ENSMUST00000000808.7             exons  \n",
       "376     ENSMUST00000050323.5             exons  \n",
       "497    ENSMUST00000002989.10             exons  \n",
       "...                      ...               ...  \n",
       "15443   ENSMUST00000115522.9  three_prime_utrs  \n",
       "15453   ENSMUST00000097399.5  three_prime_utrs  \n",
       "15454  ENSMUST00000061688.10  three_prime_utrs  \n",
       "15455   ENSMUST00000113795.7  three_prime_utrs  \n",
       "15463   ENSMUST00000010249.6  three_prime_utrs  \n",
       "\n",
       "[4587 rows x 12 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"summarised regions on sense strand : {stitched_sense_regions.shape}\")\n",
    "\n",
    "all_count = stitched_sense_regions.shape[0]\n",
    "nocov_count = stitched_sense_regions[stitched_sense_regions['total_c']<=0].shape[0]\n",
    "print(f\"of which has no coverage: {nocov_count}\")\n",
    "print(f\"proportion no coverage: {nocov_count/all_count}\")\n",
    "\n",
    "stitched_sense_regions[stitched_sense_regions['total_c']<=0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "summarised regions on antisense strand : (69769, 12)\n",
      "of which has no coverage: 4579\n",
      "proportion no coverage: 0.06563086757729078\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>total_c</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSMUSG00000000325.16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4406</td>\n",
       "      <td>Arvcf</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000115614.8</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>ENSMUSG00000000326.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1365</td>\n",
       "      <td>Comt</td>\n",
       "      <td>16</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000115609.9</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>ENSMUSG00000000791.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2688</td>\n",
       "      <td>Il12rb1</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000808.7</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>ENSMUSG00000002104.11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1251</td>\n",
       "      <td>Rapsn</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000050323.5</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>ENSMUSG00000002910.11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1755</td>\n",
       "      <td>Arrdc2</td>\n",
       "      <td>8</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000002989.10</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15443</th>\n",
       "      <td>ENSMUSG00000115219.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>333</td>\n",
       "      <td>Eef1akmt4</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000115522.9</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15453</th>\n",
       "      <td>ENSMUSG00000116780.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>Gm3417</td>\n",
       "      <td>17</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000097399.5</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15454</th>\n",
       "      <td>ENSMUSG00000116953.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100</td>\n",
       "      <td>9030025P20Rik</td>\n",
       "      <td>17</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000061688.10</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15455</th>\n",
       "      <td>ENSMUSG00000117081.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>62</td>\n",
       "      <td>Gm49948</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000113795.7</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15463</th>\n",
       "      <td>ENSMUSG00000118346.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>120</td>\n",
       "      <td>Tmem179b</td>\n",
       "      <td>19</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000010249.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4579 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_id  mean_mc  mean_hmc  mean_modc  cpg_count  \\\n",
       "59     ENSMUSG00000000325.16      NaN       NaN        NaN        0.0   \n",
       "60     ENSMUSG00000000326.13      NaN       NaN        NaN        0.0   \n",
       "134     ENSMUSG00000000791.9      NaN       NaN        NaN        0.0   \n",
       "376    ENSMUSG00000002104.11      NaN       NaN        NaN        0.0   \n",
       "497    ENSMUSG00000002910.11      NaN       NaN        NaN        0.0   \n",
       "...                      ...      ...       ...        ...        ...   \n",
       "15443   ENSMUSG00000115219.2      NaN       NaN        NaN        0.0   \n",
       "15453   ENSMUSG00000116780.1      NaN       NaN        NaN        0.0   \n",
       "15454   ENSMUSG00000116953.1      NaN       NaN        NaN        0.0   \n",
       "15455   ENSMUSG00000117081.1      NaN       NaN        NaN        0.0   \n",
       "15463   ENSMUSG00000118346.1      NaN       NaN        NaN        0.0   \n",
       "\n",
       "       total_c  range_length      Gene_name contig strand  \\\n",
       "59         0.0          4406          Arvcf     16      -   \n",
       "60         0.0          1365           Comt     16      +   \n",
       "134        0.0          2688        Il12rb1      8      -   \n",
       "376        0.0          1251          Rapsn      2      -   \n",
       "497        0.0          1755         Arrdc2      8      +   \n",
       "...        ...           ...            ...    ...    ...   \n",
       "15443      0.0           333      Eef1akmt4     16      -   \n",
       "15453      0.0             2         Gm3417     17      +   \n",
       "15454      0.0           100  9030025P20Rik     17      -   \n",
       "15455      0.0            62        Gm49948     16      -   \n",
       "15463      0.0           120       Tmem179b     19      +   \n",
       "\n",
       "               Transcript_id            Region  \n",
       "59      ENSMUST00000115614.8             exons  \n",
       "60      ENSMUST00000115609.9             exons  \n",
       "134     ENSMUST00000000808.7             exons  \n",
       "376     ENSMUST00000050323.5             exons  \n",
       "497    ENSMUST00000002989.10             exons  \n",
       "...                      ...               ...  \n",
       "15443   ENSMUST00000115522.9  three_prime_utrs  \n",
       "15453   ENSMUST00000097399.5  three_prime_utrs  \n",
       "15454  ENSMUST00000061688.10  three_prime_utrs  \n",
       "15455   ENSMUST00000113795.7  three_prime_utrs  \n",
       "15463   ENSMUST00000010249.6  three_prime_utrs  \n",
       "\n",
       "[4579 rows x 12 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"summarised regions on antisense strand : {stitched_antisense_regions.shape}\")\n",
    "\n",
    "all_count = stitched_antisense_regions.shape[0]\n",
    "nocov_count = stitched_antisense_regions[stitched_antisense_regions['total_c']<=0].shape[0]\n",
    "print(f\"of which has no coverage: {nocov_count}\")\n",
    "print(f\"proportion no coverage: {nocov_count/all_count}\")\n",
    "\n",
    "stitched_antisense_regions[stitched_antisense_regions['total_c']<=0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop regions with no coverage:\n",
    "stitched_sense_regions = stitched_sense_regions[stitched_sense_regions['total_c']>0]\n",
    "stitched_antisense_regions = stitched_antisense_regions[stitched_antisense_regions['total_c']>0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>total_c</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.807980</td>\n",
       "      <td>0.025436</td>\n",
       "      <td>0.847215</td>\n",
       "      <td>51.0</td>\n",
       "      <td>6015.0</td>\n",
       "      <td>2995</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.780352</td>\n",
       "      <td>0.038773</td>\n",
       "      <td>0.832180</td>\n",
       "      <td>144.0</td>\n",
       "      <td>15320.0</td>\n",
       "      <td>13562</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>introns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.000784</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000784</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1275.0</td>\n",
       "      <td>140</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.813870</td>\n",
       "      <td>0.023717</td>\n",
       "      <td>0.849446</td>\n",
       "      <td>32.0</td>\n",
       "      <td>3879.0</td>\n",
       "      <td>2054</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  total_c  \\\n",
       "0  ENSMUSG00000000001.4  0.807980  0.025436   0.847215       51.0   6015.0   \n",
       "0  ENSMUSG00000000001.4  0.780352  0.038773   0.832180      144.0  15320.0   \n",
       "0  ENSMUSG00000000001.4  0.000784  0.000000   0.000784       15.0   1275.0   \n",
       "0  ENSMUSG00000000001.4  0.813870  0.023717   0.849446       32.0   3879.0   \n",
       "\n",
       "   range_length Gene_name contig strand         Transcript_id  \\\n",
       "0          2995     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0         13562     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0           140     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "0          2054     Gnai3      3      -  ENSMUST00000000001.4   \n",
       "\n",
       "             Region  \n",
       "0             exons  \n",
       "0           introns  \n",
       "0   five_prime_utrs  \n",
       "0  three_prime_utrs  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stitched_sense_regions[stitched_sense_regions['Gene_id']==\"ENSMUSG00000000001.4\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>total_c</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.810206</td>\n",
       "      <td>0.026502</td>\n",
       "      <td>0.848889</td>\n",
       "      <td>51.0</td>\n",
       "      <td>6075.0</td>\n",
       "      <td>2995</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.772727</td>\n",
       "      <td>0.045917</td>\n",
       "      <td>0.834170</td>\n",
       "      <td>148.0</td>\n",
       "      <td>15136.0</td>\n",
       "      <td>13562</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>introns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1205.0</td>\n",
       "      <td>140</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.816244</td>\n",
       "      <td>0.028680</td>\n",
       "      <td>0.856091</td>\n",
       "      <td>32.0</td>\n",
       "      <td>3940.0</td>\n",
       "      <td>2054</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  total_c  \\\n",
       "0  ENSMUSG00000000001.4  0.810206  0.026502   0.848889       51.0   6075.0   \n",
       "0  ENSMUSG00000000001.4  0.772727  0.045917   0.834170      148.0  15136.0   \n",
       "0  ENSMUSG00000000001.4  0.000000  0.000000   0.000000       15.0   1205.0   \n",
       "0  ENSMUSG00000000001.4  0.816244  0.028680   0.856091       32.0   3940.0   \n",
       "\n",
       "   range_length Gene_name contig strand         Transcript_id  \\\n",
       "0          2995     Gnai3      3      +  ENSMUST00000000001.4   \n",
       "0         13562     Gnai3      3      +  ENSMUST00000000001.4   \n",
       "0           140     Gnai3      3      +  ENSMUST00000000001.4   \n",
       "0          2054     Gnai3      3      +  ENSMUST00000000001.4   \n",
       "\n",
       "             Region  \n",
       "0             exons  \n",
       "0           introns  \n",
       "0   five_prime_utrs  \n",
       "0  three_prime_utrs  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stitched_antisense_regions[stitched_antisense_regions['Gene_id']==\"ENSMUSG00000000001.4\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### no stitching"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### TSS, TES, first exon, first intron"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plus strand total regions: 63947\n",
      "minus strand total regions: 64156\n"
     ]
    }
   ],
   "source": [
    "print(f\"plus strand total regions: {len(before_tss_plus)+ len(after_tes_plus) + len(genes_clean) + len(first_exons_plus) + len(first_introns_plus)}\")\n",
    "print(f\"minus strand total regions: {len(before_tss_minus)+ len(after_tes_minus) + len(genes_clean) + len(first_exons_minus) + len(first_introns_minus)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop columns\n",
    "to_drop = [\"Score\", \"Source\", \"Phase\", \"Type\"]\n",
    "\n",
    "regions_dict_plus ={\n",
    "        \"before_tss\": before_tss_plus,\n",
    "        \"after_tes\": after_tes_plus,\n",
    "        \"genes\": genes_clean,\n",
    "        \"first_exons\": first_exons_plus,\n",
    "        \"first_introns\": first_introns_plus,\n",
    "    }\n",
    "\n",
    "for region in regions_dict_plus:\n",
    "    regions_dict_plus[region].Region = region\n",
    "    try:\n",
    "        regions_dict_plus[region] = regions_dict_plus[region].drop(to_drop)\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "regions_dict_minus ={\n",
    "        \"before_tss\": before_tss_minus,\n",
    "        \"after_tes\": after_tes_minus,\n",
    "        \"genes\": genes_clean,\n",
    "        \"first_exons\": first_exons_minus,\n",
    "        \"first_introns\": first_introns_minus,\n",
    "    }\n",
    "\n",
    "for region in regions_dict_minus:\n",
    "    regions_dict_minus[region].Region = region\n",
    "    try:\n",
    "        regions_dict_minus[region] = regions_dict_minus[region].drop(to_drop)\n",
    "    except:\n",
    "        pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarise_methylation_across_other_regions(plus_strand_cpgs, minus_strand_cpgs, regions_dict, region_strand, sense_methylation=None, antisense_methylation=None):\n",
    "    \"\"\"\n",
    "    Summarise the methylation data across multiple regions.\n",
    "\n",
    "    Parameters\n",
    "        rdr: pyranges object\n",
    "        regions_dict: dict\n",
    "\n",
    "    Returns\n",
    "        grouped_df: pd.DataFrame\n",
    "    \"\"\"\n",
    "\n",
    "    rds_plus = plus_strand_cpgs.reduce_byranges(\n",
    "                ranges=list(regions_dict.values()), \n",
    "                var=[\"num_mc\", \"num_hmc\", \"num_modc\", \"num_total_c\"]\n",
    "            )\n",
    "    rds_minus = minus_strand_cpgs.reduce_byranges(\n",
    "                ranges=list(regions_dict.values()),      \n",
    "                var=[\"num_mc\", \"num_hmc\", \"num_modc\", \"num_total_c\"]\n",
    "            )\n",
    "   \n",
    "    if region_strand == \"+\":\n",
    "        # sense to the + region\n",
    "        sense_plus_ds = rds_plus\n",
    "        plus_sense = ['+'] * sense_plus_ds.dims['ranges']\n",
    "        sense_plus_ds = sense_plus_ds.assign_coords(strand=('ranges', plus_sense))\n",
    "\n",
    "        # antisense to the + region\n",
    "        antisense_plus_ds = rds_minus\n",
    "        plus_antisense = ['-'] * antisense_plus_ds.dims['ranges']\n",
    "        antisense_plus_ds = antisense_plus_ds.assign_coords(strand=('ranges', plus_antisense))\n",
    "        \n",
    "        # Add the modified dataset to the output xarray\n",
    "        if sense_methylation is None:\n",
    "            sense_methylation= sense_plus_ds\n",
    "        else:\n",
    "            sense_methylation= xr.concat([sense_methylation, sense_plus_ds], dim=\"ranges\")\n",
    "            \n",
    "        if antisense_methylation is None:\n",
    "            antisense_methylation = antisense_plus_ds\n",
    "        else:\n",
    "            antisense_methylation= xr.concat([antisense_methylation, antisense_plus_ds], dim=\"ranges\")\n",
    "    else:\n",
    "        # sense to the - region\n",
    "        sense_minus_ds = rds_minus\n",
    "        minus_sense = ['-'] * sense_minus_ds.dims['ranges']\n",
    "        sense_minus_ds = sense_minus_ds.assign_coords(strand=('ranges', minus_sense))\n",
    "        \n",
    "        # antisense to the - region\n",
    "        antisense_minus_ds = rds_plus\n",
    "        minus_antisense = ['+'] * antisense_minus_ds.dims['ranges']\n",
    "        antisense_minus_ds = antisense_minus_ds.assign_coords(strand=('ranges', minus_antisense))\n",
    "    \n",
    "\n",
    "        # Add the modified dataset to the output xarray\n",
    "        if sense_methylation is None:\n",
    "            sense_methylation = sense_minus_ds\n",
    "        else:\n",
    "            sense_methylation = xr.concat([sense_methylation, sense_minus_ds], dim=\"ranges\")\n",
    "\n",
    "        if antisense_methylation is None:\n",
    "            antisense_methylation= antisense_minus_ds\n",
    "        else:\n",
    "            antisense_methylation = xr.concat([antisense_methylation, antisense_minus_ds], dim=\"ranges\")\n",
    "\n",
    "    return sense_methylation, antisense_methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "sense_ds, antisense_ds = summarise_methylation_across_other_regions(plus_strand_cov, minus_strand_cov, regions_dict_plus, \"+\")\n",
    "sense_ds, antisense_ds = summarise_methylation_across_other_regions(plus_strand_cov, minus_strand_cov, regions_dict_minus, \"-\", sense_ds, antisense_ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;\n",
       "Dimensions:                   (ranges: 128103, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) &lt;U2 &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;\n",
       "    start                     (ranges) int64 4807587 4807691 ... 22859847\n",
       "    end                       (ranges) int64 4807787 4807891 ... 22861212\n",
       "    range_id                  (ranges) int64 0 1 2 3 ... 64152 64153 64154 64155\n",
       "    num_contexts              (ranges) int64 27 35 21 0 7 2 11 ... 0 1 0 1 1 0 0\n",
       "    range_length              (ranges) int64 200 200 200 200 ... 1361 1365 1365\n",
       "  * sample_id                 (sample_id) &lt;U8 &#x27;sample_0&#x27;\n",
       "    strand                    (ranges) &lt;U1 &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 2.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 0.05714 ... nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 1.0 4.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.03704 0.1143 ... nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 1.0 6.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.03704 0.1714 ... nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 2.122e+03 ... 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 78.59 82.97 ... nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    Id                        (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Gene_id                   (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Gene_type                 (ranges) object &#x27;protein_coding&#x27; ... &#x27;protein_c...\n",
       "    Gene_name                 (ranges) object &#x27;Lypla1&#x27; &#x27;Gm37988&#x27; ... &#x27;Gm21366&#x27;\n",
       "    Level                     (ranges) object &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;\n",
       "    Mgi_id                    (ranges) object &#x27;MGI:1344588&#x27; ... &#x27;MGI:5434721&#x27;\n",
       "    Havana_gene               (ranges) object &#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTT...\n",
       "    Tag                       (ranges) object &#x27;overlapping_locus&#x27; ... &#x27;&#x27;\n",
       "    Transcript_id             (ranges) object &#x27;ENSMUST00000027036.10&#x27; ... &#x27;EN...\n",
       "    Region                    (ranges) object &#x27;before_tss&#x27; ... &#x27;first_introns&#x27;\n",
       "    Strand                    (ranges) object &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;\n",
       "    Parent                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_type           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_name           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_number               (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_id                   (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_support_level  (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Havana_transcript         (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Protein_id                (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ccdsid                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ont                       (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Level_1                   (ranges) float64 nan nan nan nan ... 8.0 8.0 8.0\n",
       "    Ranges_id                 (ranges) float64 nan nan ... 2.164e+04 2.164e+04\n",
       "    Intron_number             (ranges) object nan nan nan nan ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-3e30b4b2-f821-4e02-a593-0eef6e1af1e3' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-3e30b4b2-f821-4e02-a593-0eef6e1af1e3' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>ranges</span>: 128103</li><li><span class='xr-has-index'>sample_id</span>: 1</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-14a89ead-849f-4389-add4-9dc745a99d48' class='xr-section-summary-in' type='checkbox'  checked><label for='section-14a89ead-849f-4389-add4-9dc745a99d48' class='xr-section-summary' >Coordinates: <span>(8)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>contig</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U2</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;</div><input id='attrs-02edfb5d-cce9-4233-884a-ea298c250d41' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-02edfb5d-cce9-4233-884a-ea298c250d41' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c060416b-67de-425f-a04d-1171b1f5f67b' class='xr-var-data-in' type='checkbox'><label for='data-c060416b-67de-425f-a04d-1171b1f5f67b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;, ..., &#x27;Y&#x27;, &#x27;Y&#x27;, &#x27;Y&#x27;], dtype=&#x27;&lt;U2&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>start</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4807587 4807691 ... 22859847</div><input id='attrs-560af856-d9b5-43ac-9888-d061900b071f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-560af856-d9b5-43ac-9888-d061900b071f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1080f998-0b56-4bf7-83b3-ee8a81f12274' class='xr-var-data-in' type='checkbox'><label for='data-1080f998-0b56-4bf7-83b3-ee8a81f12274' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4807587,  4807691,  4857613, ..., 20992252, 22049944, 22859847])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>end</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4807787 4807891 ... 22861212</div><input id='attrs-ca46bf3f-98ad-483a-9ec8-a161b697c0a2' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ca46bf3f-98ad-483a-9ec8-a161b697c0a2' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1e8aa27c-a033-4a53-a768-4574dae73390' class='xr-var-data-in' type='checkbox'><label for='data-1e8aa27c-a033-4a53-a768-4574dae73390' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4807787,  4807891,  4857813, ..., 20993613, 22051309, 22861212])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 ... 64152 64153 64154 64155</div><input id='attrs-2dab5d0e-e8bb-4151-99f5-ee3d2a29c85c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2dab5d0e-e8bb-4151-99f5-ee3d2a29c85c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-bdb118f0-6828-45f6-a577-877306280a52' class='xr-var-data-in' type='checkbox'><label for='data-bdb118f0-6828-45f6-a577-877306280a52' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([    0,     1,     2, ..., 64153, 64154, 64155])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_contexts</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>27 35 21 0 7 2 11 ... 0 1 0 1 1 0 0</div><input id='attrs-0784e29e-a39a-4146-9879-41e0bd798f44' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0784e29e-a39a-4146-9879-41e0bd798f44' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ffa0e3ae-3efe-425b-8797-983a211d6542' class='xr-var-data-in' type='checkbox'><label for='data-ffa0e3ae-3efe-425b-8797-983a211d6542' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([27, 35, 21, ...,  1,  0,  0])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_length</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>200 200 200 200 ... 1361 1365 1365</div><input id='attrs-a634ff6d-2619-4561-aa61-efeb1ac69ea7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-a634ff6d-2619-4561-aa61-efeb1ac69ea7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-efc14c13-cd85-4b7c-8c7b-f5a28867b1b4' class='xr-var-data-in' type='checkbox'><label for='data-efc14c13-cd85-4b7c-8c7b-f5a28867b1b4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 200,  200,  200, ..., 1361, 1365, 1365])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample_id</span></div><div class='xr-var-dims'>(sample_id)</div><div class='xr-var-dtype'>&lt;U8</div><div class='xr-var-preview xr-preview'>&#x27;sample_0&#x27;</div><input id='attrs-31c385c4-0c35-4931-b2d2-a8394f1ef439' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-31c385c4-0c35-4931-b2d2-a8394f1ef439' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4d4e12a3-c902-4f0e-b962-ec0e71918de5' class='xr-var-data-in' type='checkbox'><label for='data-4d4e12a3-c902-4f0e-b962-ec0e71918de5' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;sample_0&#x27;], dtype=&#x27;&lt;U8&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U1</div><div class='xr-var-preview xr-preview'>&#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;</div><input id='attrs-142b42fb-71a7-4073-9e7c-d56ec93d6375' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-142b42fb-71a7-4073-9e7c-d56ec93d6375' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-792cf106-5ce2-49ef-bcda-9fce59980341' class='xr-var-data-in' type='checkbox'><label for='data-792cf106-5ce2-49ef-bcda-9fce59980341' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;+&#x27;, &#x27;+&#x27;, &#x27;+&#x27;, ..., &#x27;-&#x27;, &#x27;-&#x27;, &#x27;-&#x27;], dtype=&#x27;&lt;U1&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-6ba6ac00-2688-40c2-a490-e8cbdb889d0b' class='xr-section-summary-in' type='checkbox'  ><label for='section-6ba6ac00-2688-40c2-a490-e8cbdb889d0b' class='xr-section-summary' >Data variables: <span>(36)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 2.0 4.0 0.0 ... 19.0 0.0 0.0</div><input id='attrs-5453d388-dd31-4ca0-b415-34c52647db08' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5453d388-dd31-4ca0-b415-34c52647db08' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a10a9ccc-cf87-42e1-bbae-6af544f0d8ec' class='xr-var-data-in' type='checkbox'><label for='data-a10a9ccc-cf87-42e1-bbae-6af544f0d8ec' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [ 2.],\n",
       "       [ 4.],\n",
       "       ...,\n",
       "       [19.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.05714 0.1905 ... 19.0 nan nan</div><input id='attrs-47a21e71-c7e3-4835-858b-eba90b5341ef' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-47a21e71-c7e3-4835-858b-eba90b5341ef' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d63e8fa2-825b-4a68-87a2-38f76fdd0bf3' class='xr-var-data-in' type='checkbox'><label for='data-d63e8fa2-825b-4a68-87a2-38f76fdd0bf3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.        ],\n",
       "       [ 0.05714286],\n",
       "       [ 0.19047619],\n",
       "       ...,\n",
       "       [19.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-3b9c848b-7af0-436f-b86d-58599f0b30da' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3b9c848b-7af0-436f-b86d-58599f0b30da' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0dfd9a95-fa8a-42d7-943f-0ba350d2cc31' class='xr-var-data-in' type='checkbox'><label for='data-0dfd9a95-fa8a-42d7-943f-0ba350d2cc31' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 4.0 2.0 0.0 ... 0.0 1.0 0.0 0.0</div><input id='attrs-d911f04a-0dba-47e9-9aa2-3a02f0a6ce23' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d911f04a-0dba-47e9-9aa2-3a02f0a6ce23' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5e825ed4-c97b-48c8-8f4f-f9ff521f5d24' class='xr-var-data-in' type='checkbox'><label for='data-5e825ed4-c97b-48c8-8f4f-f9ff521f5d24' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [4.],\n",
       "       [2.],\n",
       "       ...,\n",
       "       [1.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.03704 0.1143 0.09524 ... nan nan</div><input id='attrs-092426d2-c3bb-45c2-b10d-21d95ae89c3f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-092426d2-c3bb-45c2-b10d-21d95ae89c3f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8682a68d-6aed-4725-8f77-f72bc22a4add' class='xr-var-data-in' type='checkbox'><label for='data-8682a68d-6aed-4725-8f77-f72bc22a4add' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.03703704],\n",
       "       [0.11428571],\n",
       "       [0.0952381 ],\n",
       "       ...,\n",
       "       [1.        ],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-4862be21-c149-4996-82cb-44b720e20ea7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4862be21-c149-4996-82cb-44b720e20ea7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e4ba55da-fd01-4871-bd34-3c9fa9bcd029' class='xr-var-data-in' type='checkbox'><label for='data-e4ba55da-fd01-4871-bd34-3c9fa9bcd029' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 6.0 6.0 0.0 ... 20.0 0.0 0.0</div><input id='attrs-7f8e4ec9-ad88-44ea-ae8b-469052b50959' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7f8e4ec9-ad88-44ea-ae8b-469052b50959' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8c8278b2-5965-4b86-9c4e-6765e35e405a' class='xr-var-data-in' type='checkbox'><label for='data-8c8278b2-5965-4b86-9c4e-6765e35e405a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 1.],\n",
       "       [ 6.],\n",
       "       [ 6.],\n",
       "       ...,\n",
       "       [20.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.03704 0.1714 0.2857 ... nan nan</div><input id='attrs-4ab448b6-ef4a-402f-a347-222796603291' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4ab448b6-ef4a-402f-a347-222796603291' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0f25e844-a2a4-495a-9779-c87f0b5b8ced' class='xr-var-data-in' type='checkbox'><label for='data-0f25e844-a2a4-495a-9779-c87f0b5b8ced' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.03703704],\n",
       "       [ 0.17142857],\n",
       "       [ 0.28571429],\n",
       "       ...,\n",
       "       [20.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-069017e9-4734-4df4-8536-4e3933974e76' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-069017e9-4734-4df4-8536-4e3933974e76' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-f5407116-22ca-41ce-a120-4c950e20e417' class='xr-var-data-in' type='checkbox'><label for='data-f5407116-22ca-41ce-a120-4c950e20e417' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>2.122e+03 2.904e+03 ... 0.0 0.0</div><input id='attrs-ae954335-1fd2-4ed3-b2f0-2d8886ce2a00' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ae954335-1fd2-4ed3-b2f0-2d8886ce2a00' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-87577f43-cbc2-493d-aa5c-f3dbd7fda9c6' class='xr-var-data-in' type='checkbox'><label for='data-87577f43-cbc2-493d-aa5c-f3dbd7fda9c6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[2122.],\n",
       "       [2904.],\n",
       "       [2032.],\n",
       "       ...,\n",
       "       [  28.],\n",
       "       [   0.],\n",
       "       [   0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>78.59 82.97 96.76 ... 28.0 nan nan</div><input id='attrs-6ebb22c8-3918-4833-866e-afca6f991fda' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6ebb22c8-3918-4833-866e-afca6f991fda' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5c4590a0-b3ec-42eb-b66c-254f0fb16ed2' class='xr-var-data-in' type='checkbox'><label for='data-5c4590a0-b3ec-42eb-b66c-254f0fb16ed2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[78.59259259],\n",
       "       [82.97142857],\n",
       "       [96.76190476],\n",
       "       ...,\n",
       "       [28.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-7b1bf048-0173-49e9-a7f6-25caad60599a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7b1bf048-0173-49e9-a7f6-25caad60599a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-bb02cae1-bfd0-4667-bc21-da3db8af064a' class='xr-var-data-in' type='checkbox'><label for='data-bb02cae1-bfd0-4667-bc21-da3db8af064a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-f174e0e4-7346-44aa-810b-95e2563e4c35' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f174e0e4-7346-44aa-810b-95e2563e4c35' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-32d31be1-031b-4e5c-81bc-1f325f5d81f9' class='xr-var-data-in' type='checkbox'><label for='data-32d31be1-031b-4e5c-81bc-1f325f5d81f9' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000104217.1&#x27;,\n",
       "       &#x27;ENSMUSG00000033813.15&#x27;, ..., &#x27;ENSMUSG00000102668.1&#x27;,\n",
       "       &#x27;ENSMUSG00000103528.1&#x27;, &#x27;ENSMUSG00000103919.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-f3fdc79f-75aa-4284-b4af-3bb04b7656c0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f3fdc79f-75aa-4284-b4af-3bb04b7656c0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-79ef02e2-fe83-4717-b056-b66d51b827de' class='xr-var-data-in' type='checkbox'><label for='data-79ef02e2-fe83-4717-b056-b66d51b827de' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000104217.1&#x27;,\n",
       "       &#x27;ENSMUSG00000033813.15&#x27;, ..., &#x27;ENSMUSG00000102668.1&#x27;,\n",
       "       &#x27;ENSMUSG00000103528.1&#x27;, &#x27;ENSMUSG00000103919.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;protein_coding&#x27; ... &#x27;protein_co...</div><input id='attrs-c084b879-1917-42fd-9a8f-64a13d066831' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c084b879-1917-42fd-9a8f-64a13d066831' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-2a66e17d-4fdd-4b30-8e5c-15e8b50dd2b7' class='xr-var-data-in' type='checkbox'><label for='data-2a66e17d-4fdd-4b30-8e5c-15e8b50dd2b7' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, ...,\n",
       "       &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Lypla1&#x27; &#x27;Gm37988&#x27; ... &#x27;Gm21366&#x27;</div><input id='attrs-70f8eebb-f069-4c0f-b700-e31002859346' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-70f8eebb-f069-4c0f-b700-e31002859346' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-36c57ec0-7444-4cfa-862b-2a2d8ad5e607' class='xr-var-data-in' type='checkbox'><label for='data-36c57ec0-7444-4cfa-862b-2a2d8ad5e607' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Lypla1&#x27;, &#x27;Gm37988&#x27;, &#x27;Tcea1&#x27;, ..., &#x27;Gm29866&#x27;, &#x27;Gm31571&#x27;, &#x27;Gm21366&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;</div><input id='attrs-ca784dc2-17c0-4d33-be48-f4d4c90df6a7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ca784dc2-17c0-4d33-be48-f4d4c90df6a7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4662f90c-4e78-4f5b-8a69-dd361a5d1df1' class='xr-var-data-in' type='checkbox'><label for='data-4662f90c-4e78-4f5b-8a69-dd361a5d1df1' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;, ..., &#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Mgi_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;MGI:1344588&#x27; ... &#x27;MGI:5434721&#x27;</div><input id='attrs-5dc9bfa7-9bdd-4e25-a48d-1de292e168c1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5dc9bfa7-9bdd-4e25-a48d-1de292e168c1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3bcc49dc-3baa-47a1-bcae-6d42f49bdd87' class='xr-var-data-in' type='checkbox'><label for='data-3bcc49dc-3baa-47a1-bcae-6d42f49bdd87' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;MGI:1344588&#x27;, &#x27;MGI:5611216&#x27;, &#x27;MGI:1196624&#x27;, ..., &#x27;MGI:5589025&#x27;,\n",
       "       &#x27;MGI:5590730&#x27;, &#x27;MGI:5434721&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_gene</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTTM...</div><input id='attrs-7d7877f2-72da-4040-823f-29f792908411' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7d7877f2-72da-4040-823f-29f792908411' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7e6ac68c-1c25-4a12-9979-785b8c800015' class='xr-var-data-in' type='checkbox'><label for='data-7e6ac68c-1c25-4a12-9979-785b8c800015' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;OTTMUSG00000021562.4&#x27;, &#x27;OTTMUSG00000050100.1&#x27;,\n",
       "       &#x27;OTTMUSG00000042348.1&#x27;, ..., &#x27;OTTMUSG00000050180.1&#x27;,\n",
       "       &#x27;OTTMUSG00000050134.1&#x27;, &#x27;OTTMUSG00000050112.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Tag</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;overlapping_locus&#x27; ... &#x27;&#x27;</div><input id='attrs-24b3fec4-6512-476d-b684-77fbf497bfc9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-24b3fec4-6512-476d-b684-77fbf497bfc9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-71f69359-c396-4992-8636-6f3454224621' class='xr-var-data-in' type='checkbox'><label for='data-71f69359-c396-4992-8636-6f3454224621' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;overlapping_locus&#x27;, &#x27;overlapping_locus&#x27;, &#x27;overlapping_locus&#x27;, ...,\n",
       "       &#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUST00000027036.10&#x27; ... &#x27;ENS...</div><input id='attrs-b7f6d7e3-7445-43bd-a8f6-efa45b3aab94' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b7f6d7e3-7445-43bd-a8f6-efa45b3aab94' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a7c8f4a9-6ffa-4618-ae04-d4884bd42dc6' class='xr-var-data-in' type='checkbox'><label for='data-a7c8f4a9-6ffa-4618-ae04-d4884bd42dc6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUST00000027036.10&#x27;, &#x27;ENSMUST00000155020.1&#x27;,\n",
       "       &#x27;ENSMUST00000081551.13&#x27;, ..., &#x27;ENSMUST00000191675.1&#x27;,\n",
       "       &#x27;ENSMUST00000192521.1&#x27;, &#x27;ENSMUST00000192892.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Region</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;before_tss&#x27; ... &#x27;first_introns&#x27;</div><input id='attrs-05e94f6c-a90c-433f-b971-046fe9d50366' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-05e94f6c-a90c-433f-b971-046fe9d50366' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4f5f4f60-4bc0-4cca-a758-4411750e475f' class='xr-var-data-in' type='checkbox'><label for='data-4f5f4f60-4bc0-4cca-a758-4411750e475f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;before_tss&#x27;, &#x27;before_tss&#x27;, &#x27;before_tss&#x27;, ..., &#x27;first_introns&#x27;,\n",
       "       &#x27;first_introns&#x27;, &#x27;first_introns&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;</div><input id='attrs-f8b07fe0-5cbb-4a4f-868d-3031a2fa9c03' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f8b07fe0-5cbb-4a4f-868d-3031a2fa9c03' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-37d5086f-643d-484a-9567-6cb3e2bece91' class='xr-var-data-in' type='checkbox'><label for='data-37d5086f-643d-484a-9567-6cb3e2bece91' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;, ..., &#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Parent</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-e81e4b84-a936-4f1d-9f6a-982a9d733f7a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e81e4b84-a936-4f1d-9f6a-982a9d733f7a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d6305db3-fa11-4320-90b3-e47afbcc597f' class='xr-var-data-in' type='checkbox'><label for='data-d6305db3-fa11-4320-90b3-e47afbcc597f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-d3c85412-307a-41ec-bef9-70b859acd8f0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d3c85412-307a-41ec-bef9-70b859acd8f0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1d0757e7-2d0b-4387-bf89-1bfcf1daac0a' class='xr-var-data-in' type='checkbox'><label for='data-1d0757e7-2d0b-4387-bf89-1bfcf1daac0a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-df66d2f6-e1fc-48f1-be67-e70b66e30e14' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-df66d2f6-e1fc-48f1-be67-e70b66e30e14' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5e633eb5-446c-4ab5-85dc-a11d38d65208' class='xr-var-data-in' type='checkbox'><label for='data-5e633eb5-446c-4ab5-85dc-a11d38d65208' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-d32faf4f-5e49-4130-a1e2-40ee33a87f63' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d32faf4f-5e49-4130-a1e2-40ee33a87f63' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c38752cf-7566-4106-97ea-78e3e5c4258b' class='xr-var-data-in' type='checkbox'><label for='data-c38752cf-7566-4106-97ea-78e3e5c4258b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-417ffed8-ccce-46c9-8916-51c34612a6e0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-417ffed8-ccce-46c9-8916-51c34612a6e0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-abbf6b67-bd19-4a6f-853f-cd563d547c30' class='xr-var-data-in' type='checkbox'><label for='data-abbf6b67-bd19-4a6f-853f-cd563d547c30' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_support_level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-88a062d0-e26d-4599-9f8a-87a8c9cee27d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-88a062d0-e26d-4599-9f8a-87a8c9cee27d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-084b5982-3e60-4886-a0c8-fb7351b13d37' class='xr-var-data-in' type='checkbox'><label for='data-084b5982-3e60-4886-a0c8-fb7351b13d37' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_transcript</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-dfc0d380-efc7-4482-954e-ba7e85f5ef75' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-dfc0d380-efc7-4482-954e-ba7e85f5ef75' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4923f10c-1418-48db-a784-39f90a0a1786' class='xr-var-data-in' type='checkbox'><label for='data-4923f10c-1418-48db-a784-39f90a0a1786' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Protein_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-fbcf1db4-5f3c-44ee-8248-d1308a831ace' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-fbcf1db4-5f3c-44ee-8248-d1308a831ace' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-579d8b19-40c4-49ca-8464-1892d7da6a12' class='xr-var-data-in' type='checkbox'><label for='data-579d8b19-40c4-49ca-8464-1892d7da6a12' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ccdsid</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-337c60b0-8dcb-4177-8615-94a4ca44891a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-337c60b0-8dcb-4177-8615-94a4ca44891a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-f63edcd7-0616-4265-9f20-c53b8e7e3078' class='xr-var-data-in' type='checkbox'><label for='data-f63edcd7-0616-4265-9f20-c53b8e7e3078' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ont</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-f265f8f3-ee38-4237-95d5-dbb1a5d04830' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f265f8f3-ee38-4237-95d5-dbb1a5d04830' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9301fdc0-0a38-443d-80d5-28dbe92ace6e' class='xr-var-data-in' type='checkbox'><label for='data-9301fdc0-0a38-443d-80d5-28dbe92ace6e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level_1</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... 8.0 8.0 8.0 8.0</div><input id='attrs-ef7c76f5-1ce4-49a4-bf22-052a8183390d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ef7c76f5-1ce4-49a4-bf22-052a8183390d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e38b95d1-5391-4252-b2b2-41d18adb9623' class='xr-var-data-in' type='checkbox'><label for='data-e38b95d1-5391-4252-b2b2-41d18adb9623' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ...,  8.,  8.,  8.])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ranges_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>nan nan nan ... 2.164e+04 2.164e+04</div><input id='attrs-dce4e228-4c95-4b57-82ab-18c5313be6e5' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-dce4e228-4c95-4b57-82ab-18c5313be6e5' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b5dec235-a5c6-4da6-8ce0-bd57f4ceead7' class='xr-var-data-in' type='checkbox'><label for='data-b5dec235-a5c6-4da6-8ce0-bd57f4ceead7' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([   nan,    nan,    nan, ..., 21635., 21636., 21637.])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Intron_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;</div><input id='attrs-d96ea4df-82bf-49ec-9673-efecf58064eb' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d96ea4df-82bf-49ec-9673-efecf58064eb' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-6d4b3b70-9103-4cab-a7eb-7e33f92da1e8' class='xr-var-data-in' type='checkbox'><label for='data-6d4b3b70-9103-4cab-a7eb-7e33f92da1e8' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., &#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;], dtype=object)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-b182b091-f62d-4585-943f-a56b4d3424ad' class='xr-section-summary-in' type='checkbox'  ><label for='section-b182b091-f62d-4585-943f-a56b4d3424ad' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>sample_id</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-dcf421a1-dd1f-46ae-8372-9bb19dea01d0' class='xr-index-data-in' type='checkbox'/><label for='index-dcf421a1-dd1f-46ae-8372-9bb19dea01d0' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;sample_0&#x27;], dtype=&#x27;object&#x27;, name=&#x27;sample_id&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-a32ca93b-da4b-403f-9a79-8f92db84a29a' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-a32ca93b-da4b-403f-9a79-8f92db84a29a' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.Dataset>\n",
       "Dimensions:                   (ranges: 128103, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) <U2 '1' '1' '1' '1' ... 'Y' 'Y' 'Y' 'Y'\n",
       "    start                     (ranges) int64 4807587 4807691 ... 22859847\n",
       "    end                       (ranges) int64 4807787 4807891 ... 22861212\n",
       "    range_id                  (ranges) int64 0 1 2 3 ... 64152 64153 64154 64155\n",
       "    num_contexts              (ranges) int64 27 35 21 0 7 2 11 ... 0 1 0 1 1 0 0\n",
       "    range_length              (ranges) int64 200 200 200 200 ... 1361 1365 1365\n",
       "  * sample_id                 (sample_id) <U8 'sample_0'\n",
       "    strand                    (ranges) <U1 '+' '+' '+' '+' ... '-' '-' '-' '-'\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 2.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 0.05714 ... nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 1.0 4.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.03704 0.1143 ... nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 1.0 6.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.03704 0.1714 ... nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 2.122e+03 ... 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 78.59 82.97 ... nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    Id                        (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Gene_id                   (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Gene_type                 (ranges) object 'protein_coding' ... 'protein_c...\n",
       "    Gene_name                 (ranges) object 'Lypla1' 'Gm37988' ... 'Gm21366'\n",
       "    Level                     (ranges) object '2' '2' '2' '2' ... '2' '2' '2'\n",
       "    Mgi_id                    (ranges) object 'MGI:1344588' ... 'MGI:5434721'\n",
       "    Havana_gene               (ranges) object 'OTTMUSG00000021562.4' ... 'OTT...\n",
       "    Tag                       (ranges) object 'overlapping_locus' ... ''\n",
       "    Transcript_id             (ranges) object 'ENSMUST00000027036.10' ... 'EN...\n",
       "    Region                    (ranges) object 'before_tss' ... 'first_introns'\n",
       "    Strand                    (ranges) object '.' '.' '.' '.' ... '.' '.' '.'\n",
       "    Parent                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_type           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_name           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_number               (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_id                   (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_support_level  (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Havana_transcript         (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Protein_id                (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ccdsid                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ont                       (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Level_1                   (ranges) float64 nan nan nan nan ... 8.0 8.0 8.0\n",
       "    Ranges_id                 (ranges) float64 nan nan ... 2.164e+04 2.164e+04\n",
       "    Intron_number             (ranges) object nan nan nan nan ... '1' '1' '1'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sense_ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;\n",
       "Dimensions:                   (ranges: 128103, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) &lt;U2 &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;\n",
       "    start                     (ranges) int64 4807587 4807691 ... 22859847\n",
       "    end                       (ranges) int64 4807787 4807891 ... 22861212\n",
       "    range_id                  (ranges) int64 0 1 2 3 ... 64152 64153 64154 64155\n",
       "    num_contexts              (ranges) int64 27 35 21 0 7 2 11 ... 0 1 0 1 1 0 0\n",
       "    range_length              (ranges) int64 200 200 200 200 ... 1361 1365 1365\n",
       "  * sample_id                 (sample_id) &lt;U8 &#x27;sample_0&#x27;\n",
       "    strand                    (ranges) &lt;U1 &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 2.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 0.05714 ... nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 1.0 4.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.03704 0.1143 ... nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 1.0 6.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.03704 0.1714 ... nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 2.122e+03 ... 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 78.59 82.97 ... nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    Id                        (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Gene_id                   (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Gene_type                 (ranges) object &#x27;protein_coding&#x27; ... &#x27;protein_c...\n",
       "    Gene_name                 (ranges) object &#x27;Lypla1&#x27; &#x27;Gm37988&#x27; ... &#x27;Gm21366&#x27;\n",
       "    Level                     (ranges) object &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;\n",
       "    Mgi_id                    (ranges) object &#x27;MGI:1344588&#x27; ... &#x27;MGI:5434721&#x27;\n",
       "    Havana_gene               (ranges) object &#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTT...\n",
       "    Tag                       (ranges) object &#x27;overlapping_locus&#x27; ... &#x27;&#x27;\n",
       "    Transcript_id             (ranges) object &#x27;ENSMUST00000027036.10&#x27; ... &#x27;EN...\n",
       "    Region                    (ranges) object &#x27;before_tss&#x27; ... &#x27;first_introns&#x27;\n",
       "    Strand                    (ranges) object &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;\n",
       "    Parent                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_type           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_name           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_number               (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_id                   (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_support_level  (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Havana_transcript         (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Protein_id                (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ccdsid                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ont                       (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Level_1                   (ranges) float64 nan nan nan nan ... 8.0 8.0 8.0\n",
       "    Ranges_id                 (ranges) float64 nan nan ... 2.164e+04 2.164e+04\n",
       "    Intron_number             (ranges) object nan nan nan nan ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;\n",
       "    mean_mc                   (ranges, sample_id) float64 0.0 0.0006887 ... nan\n",
       "    mean_hmc                  (ranges, sample_id) float64 0.0004713 ... nan\n",
       "    mean_modc                 (ranges, sample_id) float64 0.0004713 ... nan</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-897c62c8-bf2f-420a-95e8-2b32874ee290' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-897c62c8-bf2f-420a-95e8-2b32874ee290' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>ranges</span>: 128103</li><li><span class='xr-has-index'>sample_id</span>: 1</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-14f3ac91-ba42-4ac7-b609-9b1be5f9d777' class='xr-section-summary-in' type='checkbox'  checked><label for='section-14f3ac91-ba42-4ac7-b609-9b1be5f9d777' class='xr-section-summary' >Coordinates: <span>(8)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>contig</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U2</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;</div><input id='attrs-6db4826b-683e-43cc-9ac3-76405ea3d457' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6db4826b-683e-43cc-9ac3-76405ea3d457' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c5835a9a-e9a2-4e32-bebd-ea3a7e0585ad' class='xr-var-data-in' type='checkbox'><label for='data-c5835a9a-e9a2-4e32-bebd-ea3a7e0585ad' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;, ..., &#x27;Y&#x27;, &#x27;Y&#x27;, &#x27;Y&#x27;], dtype=&#x27;&lt;U2&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>start</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4807587 4807691 ... 22859847</div><input id='attrs-17031c27-1ad9-4f99-b715-b852c964daa4' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-17031c27-1ad9-4f99-b715-b852c964daa4' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1eb63e8a-d4a6-4e42-8a96-524e036083d9' class='xr-var-data-in' type='checkbox'><label for='data-1eb63e8a-d4a6-4e42-8a96-524e036083d9' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4807587,  4807691,  4857613, ..., 20992252, 22049944, 22859847])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>end</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4807787 4807891 ... 22861212</div><input id='attrs-200f7ba2-a588-4aec-a996-09f01b4af704' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-200f7ba2-a588-4aec-a996-09f01b4af704' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-406de91d-4140-49fd-8689-64804635620d' class='xr-var-data-in' type='checkbox'><label for='data-406de91d-4140-49fd-8689-64804635620d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4807787,  4807891,  4857813, ..., 20993613, 22051309, 22861212])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 ... 64152 64153 64154 64155</div><input id='attrs-30352033-fb29-4f8a-9463-4a3b03555190' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-30352033-fb29-4f8a-9463-4a3b03555190' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7a1c4a8e-167a-4b7d-8021-c2f194acf78b' class='xr-var-data-in' type='checkbox'><label for='data-7a1c4a8e-167a-4b7d-8021-c2f194acf78b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([    0,     1,     2, ..., 64153, 64154, 64155])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_contexts</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>27 35 21 0 7 2 11 ... 0 1 0 1 1 0 0</div><input id='attrs-f40dd68f-8ada-4358-bc71-488458c08250' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f40dd68f-8ada-4358-bc71-488458c08250' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-29321089-be1d-481b-bab2-971cf2b5519b' class='xr-var-data-in' type='checkbox'><label for='data-29321089-be1d-481b-bab2-971cf2b5519b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([27, 35, 21, ...,  1,  0,  0])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_length</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>200 200 200 200 ... 1361 1365 1365</div><input id='attrs-495be536-e271-4816-9c50-61ba77e14af1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-495be536-e271-4816-9c50-61ba77e14af1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-413fb43e-54f8-45ef-acd2-098ba8c13e57' class='xr-var-data-in' type='checkbox'><label for='data-413fb43e-54f8-45ef-acd2-098ba8c13e57' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 200,  200,  200, ..., 1361, 1365, 1365])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample_id</span></div><div class='xr-var-dims'>(sample_id)</div><div class='xr-var-dtype'>&lt;U8</div><div class='xr-var-preview xr-preview'>&#x27;sample_0&#x27;</div><input id='attrs-3e2dd279-8168-45a4-8c21-3dfb8f318359' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3e2dd279-8168-45a4-8c21-3dfb8f318359' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a3b5a199-bfdf-413e-b5e5-385fd5fb47f1' class='xr-var-data-in' type='checkbox'><label for='data-a3b5a199-bfdf-413e-b5e5-385fd5fb47f1' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;sample_0&#x27;], dtype=&#x27;&lt;U8&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U1</div><div class='xr-var-preview xr-preview'>&#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; ... &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27;</div><input id='attrs-ea817c78-5f87-449f-8903-b48aab9944cc' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ea817c78-5f87-449f-8903-b48aab9944cc' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-657e1728-abcd-4e99-ae0b-eca255f87a1e' class='xr-var-data-in' type='checkbox'><label for='data-657e1728-abcd-4e99-ae0b-eca255f87a1e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;+&#x27;, &#x27;+&#x27;, &#x27;+&#x27;, ..., &#x27;-&#x27;, &#x27;-&#x27;, &#x27;-&#x27;], dtype=&#x27;&lt;U1&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-359f2ab5-ec54-4ab4-ba37-4eaeb2f8c259' class='xr-section-summary-in' type='checkbox'  ><label for='section-359f2ab5-ec54-4ab4-ba37-4eaeb2f8c259' class='xr-section-summary' >Data variables: <span>(39)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 2.0 4.0 0.0 ... 19.0 0.0 0.0</div><input id='attrs-051a8b61-884b-4ab8-9cdd-466014d32d96' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-051a8b61-884b-4ab8-9cdd-466014d32d96' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-eba7d400-665b-4ebd-93cb-a6b4181fd3d0' class='xr-var-data-in' type='checkbox'><label for='data-eba7d400-665b-4ebd-93cb-a6b4181fd3d0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.],\n",
       "       [ 2.],\n",
       "       [ 4.],\n",
       "       ...,\n",
       "       [19.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.05714 0.1905 ... 19.0 nan nan</div><input id='attrs-c22e2829-7e67-4919-bae1-106c0e797464' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c22e2829-7e67-4919-bae1-106c0e797464' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-07b4f65c-0fe3-4992-aa14-241020b87483' class='xr-var-data-in' type='checkbox'><label for='data-07b4f65c-0fe3-4992-aa14-241020b87483' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.        ],\n",
       "       [ 0.05714286],\n",
       "       [ 0.19047619],\n",
       "       ...,\n",
       "       [19.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-40db01da-14d9-4bab-ad94-f6dc6304d432' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-40db01da-14d9-4bab-ad94-f6dc6304d432' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d25dc807-4674-45a0-9688-d56a11eef430' class='xr-var-data-in' type='checkbox'><label for='data-d25dc807-4674-45a0-9688-d56a11eef430' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 4.0 2.0 0.0 ... 0.0 1.0 0.0 0.0</div><input id='attrs-5db8df04-32af-4ec4-b05f-360608e3548a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5db8df04-32af-4ec4-b05f-360608e3548a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-966e48c7-ed6d-4ba4-b325-5b9961a04265' class='xr-var-data-in' type='checkbox'><label for='data-966e48c7-ed6d-4ba4-b325-5b9961a04265' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.],\n",
       "       [4.],\n",
       "       [2.],\n",
       "       ...,\n",
       "       [1.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.03704 0.1143 0.09524 ... nan nan</div><input id='attrs-77fcb7be-0be9-44b2-80a9-cbeca3a7d479' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-77fcb7be-0be9-44b2-80a9-cbeca3a7d479' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5ae789c8-2e95-44c5-aa4a-3350578a2e08' class='xr-var-data-in' type='checkbox'><label for='data-5ae789c8-2e95-44c5-aa4a-3350578a2e08' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.03703704],\n",
       "       [0.11428571],\n",
       "       [0.0952381 ],\n",
       "       ...,\n",
       "       [1.        ],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-e3697219-e1b7-4b06-92a0-fdcd403a6d9e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e3697219-e1b7-4b06-92a0-fdcd403a6d9e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9c6728e3-18e0-4ff9-a182-b38d9a24e365' class='xr-var-data-in' type='checkbox'><label for='data-9c6728e3-18e0-4ff9-a182-b38d9a24e365' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 6.0 6.0 0.0 ... 20.0 0.0 0.0</div><input id='attrs-050766f6-c2d6-4ef7-acf4-a8a8d813d39f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-050766f6-c2d6-4ef7-acf4-a8a8d813d39f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ee3e6748-f7cc-435f-8b63-a828681a2ad6' class='xr-var-data-in' type='checkbox'><label for='data-ee3e6748-f7cc-435f-8b63-a828681a2ad6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 1.],\n",
       "       [ 6.],\n",
       "       [ 6.],\n",
       "       ...,\n",
       "       [20.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.03704 0.1714 0.2857 ... nan nan</div><input id='attrs-0cf04929-da25-4a98-ac4f-526395c91e05' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0cf04929-da25-4a98-ac4f-526395c91e05' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-15856ad0-67fb-42f1-86df-d15314db0efe' class='xr-var-data-in' type='checkbox'><label for='data-15856ad0-67fb-42f1-86df-d15314db0efe' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.03703704],\n",
       "       [ 0.17142857],\n",
       "       [ 0.28571429],\n",
       "       ...,\n",
       "       [20.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-c7705fff-74bf-4933-862b-d54792a64c3a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c7705fff-74bf-4933-862b-d54792a64c3a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0d874c9e-10bc-4e11-8f27-02b8c340264e' class='xr-var-data-in' type='checkbox'><label for='data-0d874c9e-10bc-4e11-8f27-02b8c340264e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>2.122e+03 2.904e+03 ... 0.0 0.0</div><input id='attrs-d6c5e608-bedc-438f-85ee-358c09bb7c3d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d6c5e608-bedc-438f-85ee-358c09bb7c3d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5fab5461-58e1-446f-bb07-de1570b0f641' class='xr-var-data-in' type='checkbox'><label for='data-5fab5461-58e1-446f-bb07-de1570b0f641' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[2122.],\n",
       "       [2904.],\n",
       "       [2032.],\n",
       "       ...,\n",
       "       [  28.],\n",
       "       [   0.],\n",
       "       [   0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>78.59 82.97 96.76 ... 28.0 nan nan</div><input id='attrs-3d8981d7-f45f-4e4e-ab57-b95cdd354ce9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3d8981d7-f45f-4e4e-ab57-b95cdd354ce9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d0acffd1-9ab6-4dc2-a214-a6b367f1e62a' class='xr-var-data-in' type='checkbox'><label for='data-d0acffd1-9ab6-4dc2-a214-a6b367f1e62a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[78.59259259],\n",
       "       [82.97142857],\n",
       "       [96.76190476],\n",
       "       ...,\n",
       "       [28.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 21.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-741c907e-2bd8-4326-9dc3-bc815247aef6' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-741c907e-2bd8-4326-9dc3-bc815247aef6' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1a4150bd-c5c9-4c88-a69d-9618f97954fd' class='xr-var-data-in' type='checkbox'><label for='data-1a4150bd-c5c9-4c88-a69d-9618f97954fd' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [21.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-c9a62335-e1bc-4609-8c87-c46243491637' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c9a62335-e1bc-4609-8c87-c46243491637' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-2ea0e436-f122-426a-8844-7cb007e1bf9d' class='xr-var-data-in' type='checkbox'><label for='data-2ea0e436-f122-426a-8844-7cb007e1bf9d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000104217.1&#x27;,\n",
       "       &#x27;ENSMUSG00000033813.15&#x27;, ..., &#x27;ENSMUSG00000102668.1&#x27;,\n",
       "       &#x27;ENSMUSG00000103528.1&#x27;, &#x27;ENSMUSG00000103919.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-1f4cc55d-7193-4e38-a829-10a87675dcaa' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1f4cc55d-7193-4e38-a829-10a87675dcaa' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0afb96c4-a3bd-4d2a-9531-1a32981ba25c' class='xr-var-data-in' type='checkbox'><label for='data-0afb96c4-a3bd-4d2a-9531-1a32981ba25c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000104217.1&#x27;,\n",
       "       &#x27;ENSMUSG00000033813.15&#x27;, ..., &#x27;ENSMUSG00000102668.1&#x27;,\n",
       "       &#x27;ENSMUSG00000103528.1&#x27;, &#x27;ENSMUSG00000103919.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;protein_coding&#x27; ... &#x27;protein_co...</div><input id='attrs-6ce035eb-9e9e-468c-acb2-baf94237eb04' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6ce035eb-9e9e-468c-acb2-baf94237eb04' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-04553a9d-ab4f-444c-ad42-015c1d6ac05e' class='xr-var-data-in' type='checkbox'><label for='data-04553a9d-ab4f-444c-ad42-015c1d6ac05e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, ...,\n",
       "       &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Lypla1&#x27; &#x27;Gm37988&#x27; ... &#x27;Gm21366&#x27;</div><input id='attrs-c14510ed-8ab0-47cb-990d-d51e246d9721' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c14510ed-8ab0-47cb-990d-d51e246d9721' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9f98539f-6615-4cb6-b726-6d031031f3f8' class='xr-var-data-in' type='checkbox'><label for='data-9f98539f-6615-4cb6-b726-6d031031f3f8' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Lypla1&#x27;, &#x27;Gm37988&#x27;, &#x27;Tcea1&#x27;, ..., &#x27;Gm29866&#x27;, &#x27;Gm31571&#x27;, &#x27;Gm21366&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;</div><input id='attrs-40265744-c76f-4b25-927d-6ee92f03d50b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-40265744-c76f-4b25-927d-6ee92f03d50b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0439016b-d76e-4849-abc1-35dd041df1d7' class='xr-var-data-in' type='checkbox'><label for='data-0439016b-d76e-4849-abc1-35dd041df1d7' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;, ..., &#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Mgi_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;MGI:1344588&#x27; ... &#x27;MGI:5434721&#x27;</div><input id='attrs-5a7072b0-60d5-417d-88a9-2f332917c353' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5a7072b0-60d5-417d-88a9-2f332917c353' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-55e43362-2377-4799-888a-36a0ed0fd3df' class='xr-var-data-in' type='checkbox'><label for='data-55e43362-2377-4799-888a-36a0ed0fd3df' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;MGI:1344588&#x27;, &#x27;MGI:5611216&#x27;, &#x27;MGI:1196624&#x27;, ..., &#x27;MGI:5589025&#x27;,\n",
       "       &#x27;MGI:5590730&#x27;, &#x27;MGI:5434721&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_gene</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTTM...</div><input id='attrs-5a7a84e5-c44e-420b-8ac7-cefd897c03c8' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5a7a84e5-c44e-420b-8ac7-cefd897c03c8' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-65e82ea6-bed1-40c2-bd83-a9e1e72977c3' class='xr-var-data-in' type='checkbox'><label for='data-65e82ea6-bed1-40c2-bd83-a9e1e72977c3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;OTTMUSG00000021562.4&#x27;, &#x27;OTTMUSG00000050100.1&#x27;,\n",
       "       &#x27;OTTMUSG00000042348.1&#x27;, ..., &#x27;OTTMUSG00000050180.1&#x27;,\n",
       "       &#x27;OTTMUSG00000050134.1&#x27;, &#x27;OTTMUSG00000050112.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Tag</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;overlapping_locus&#x27; ... &#x27;&#x27;</div><input id='attrs-c30a3034-4d47-4d9f-9302-331d0a58da8a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c30a3034-4d47-4d9f-9302-331d0a58da8a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b64afb8d-03c6-4900-81e1-fa85cf7cbbee' class='xr-var-data-in' type='checkbox'><label for='data-b64afb8d-03c6-4900-81e1-fa85cf7cbbee' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;overlapping_locus&#x27;, &#x27;overlapping_locus&#x27;, &#x27;overlapping_locus&#x27;, ...,\n",
       "       &#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUST00000027036.10&#x27; ... &#x27;ENS...</div><input id='attrs-03bacd51-c6b9-4118-846a-83ea75e7e386' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-03bacd51-c6b9-4118-846a-83ea75e7e386' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ecf37459-29b8-4908-bb6d-b1e20d08bb9e' class='xr-var-data-in' type='checkbox'><label for='data-ecf37459-29b8-4908-bb6d-b1e20d08bb9e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUST00000027036.10&#x27;, &#x27;ENSMUST00000155020.1&#x27;,\n",
       "       &#x27;ENSMUST00000081551.13&#x27;, ..., &#x27;ENSMUST00000191675.1&#x27;,\n",
       "       &#x27;ENSMUST00000192521.1&#x27;, &#x27;ENSMUST00000192892.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Region</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;before_tss&#x27; ... &#x27;first_introns&#x27;</div><input id='attrs-155e663c-4de6-483a-91f7-3b19c495ca23' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-155e663c-4de6-483a-91f7-3b19c495ca23' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c834d886-6100-4dd4-ad90-f8bfb827bcb2' class='xr-var-data-in' type='checkbox'><label for='data-c834d886-6100-4dd4-ad90-f8bfb827bcb2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;before_tss&#x27;, &#x27;before_tss&#x27;, &#x27;before_tss&#x27;, ..., &#x27;first_introns&#x27;,\n",
       "       &#x27;first_introns&#x27;, &#x27;first_introns&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;</div><input id='attrs-119a0e34-6f21-4ce0-b642-5f05fb04729a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-119a0e34-6f21-4ce0-b642-5f05fb04729a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-10aeb70d-e2ff-4eca-9534-f40f10d52c5f' class='xr-var-data-in' type='checkbox'><label for='data-10aeb70d-e2ff-4eca-9534-f40f10d52c5f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;, ..., &#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Parent</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-8e1fa191-2296-47a3-90ea-16f3705332d5' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8e1fa191-2296-47a3-90ea-16f3705332d5' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b0063d1d-e75d-443d-877f-b6a59116db24' class='xr-var-data-in' type='checkbox'><label for='data-b0063d1d-e75d-443d-877f-b6a59116db24' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-32bdd257-94c8-42a3-8035-4c3707c18435' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-32bdd257-94c8-42a3-8035-4c3707c18435' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-898bd15e-58e1-42f5-bbc4-216ec019af8c' class='xr-var-data-in' type='checkbox'><label for='data-898bd15e-58e1-42f5-bbc4-216ec019af8c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-894e18ba-ebc7-4526-b897-ad7c65d0bb3c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-894e18ba-ebc7-4526-b897-ad7c65d0bb3c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-68e169bb-4b7f-4d25-bbaf-eead6011c23e' class='xr-var-data-in' type='checkbox'><label for='data-68e169bb-4b7f-4d25-bbaf-eead6011c23e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-0f73b6a5-b315-4c18-9d40-87ba5527b3b1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0f73b6a5-b315-4c18-9d40-87ba5527b3b1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-43e997ff-0814-4edd-8e8d-acc0a67134bd' class='xr-var-data-in' type='checkbox'><label for='data-43e997ff-0814-4edd-8e8d-acc0a67134bd' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-8975ed1d-93bf-4773-a753-d3144eae2abc' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8975ed1d-93bf-4773-a753-d3144eae2abc' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-95939efc-5570-4c96-82ab-08a0228fca68' class='xr-var-data-in' type='checkbox'><label for='data-95939efc-5570-4c96-82ab-08a0228fca68' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_support_level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-bfc0de28-bb77-4f8b-8950-4e40da36e16f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-bfc0de28-bb77-4f8b-8950-4e40da36e16f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0feade1e-7838-44e3-99b3-5e15262c5c32' class='xr-var-data-in' type='checkbox'><label for='data-0feade1e-7838-44e3-99b3-5e15262c5c32' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_transcript</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-4f3a3f5f-80c6-4352-9b39-1fdb89e1044b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4f3a3f5f-80c6-4352-9b39-1fdb89e1044b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8d722f2b-f719-4c85-bd53-a2548fab6303' class='xr-var-data-in' type='checkbox'><label for='data-8d722f2b-f719-4c85-bd53-a2548fab6303' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Protein_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-b8bf1376-cd00-47c1-9a6f-06894f80c6c0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b8bf1376-cd00-47c1-9a6f-06894f80c6c0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e549f342-ef69-4b6a-9326-2320e261a135' class='xr-var-data-in' type='checkbox'><label for='data-e549f342-ef69-4b6a-9326-2320e261a135' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ccdsid</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-7f2477fe-9c8c-4aeb-a171-190cb029c2cc' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7f2477fe-9c8c-4aeb-a171-190cb029c2cc' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-564e112a-f7bf-44f1-8c92-1c6d40f85dbb' class='xr-var-data-in' type='checkbox'><label for='data-564e112a-f7bf-44f1-8c92-1c6d40f85dbb' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ont</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-880230ad-af13-4105-8288-991e0244517b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-880230ad-af13-4105-8288-991e0244517b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-6a926108-765c-46c6-ae58-5caf9a8cbd30' class='xr-var-data-in' type='checkbox'><label for='data-6a926108-765c-46c6-ae58-5caf9a8cbd30' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level_1</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... 8.0 8.0 8.0 8.0</div><input id='attrs-097e563a-0ad9-4b67-9e68-f500e4129900' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-097e563a-0ad9-4b67-9e68-f500e4129900' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-27ba0f6d-33d2-4f22-b97c-629a8c555011' class='xr-var-data-in' type='checkbox'><label for='data-27ba0f6d-33d2-4f22-b97c-629a8c555011' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ...,  8.,  8.,  8.])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ranges_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>nan nan nan ... 2.164e+04 2.164e+04</div><input id='attrs-8a64c9c1-6906-4a16-a3d9-0aecb6e7bb3c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8a64c9c1-6906-4a16-a3d9-0aecb6e7bb3c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4eefe1b4-2b18-4989-b22c-549a53b18c68' class='xr-var-data-in' type='checkbox'><label for='data-4eefe1b4-2b18-4989-b22c-549a53b18c68' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([   nan,    nan,    nan, ..., 21635., 21636., 21637.])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Intron_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;</div><input id='attrs-1f31e57f-90ff-4fff-82c0-0359f22ea6da' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1f31e57f-90ff-4fff-82c0-0359f22ea6da' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-313c4b48-76a6-49d9-bd19-c08c8731628a' class='xr-var-data-in' type='checkbox'><label for='data-313c4b48-76a6-49d9-bd19-c08c8731628a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., &#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_mc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0006887 0.001969 ... nan nan</div><input id='attrs-353be7da-3583-4da9-8b90-3c9e458497bd' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-353be7da-3583-4da9-8b90-3c9e458497bd' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-02987f4c-42be-48be-9ed8-0cb7acbd0218' class='xr-var-data-in' type='checkbox'><label for='data-02987f4c-42be-48be-9ed8-0cb7acbd0218' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.        ],\n",
       "       [0.00068871],\n",
       "       [0.0019685 ],\n",
       "       ...,\n",
       "       [0.67857143],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_hmc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0004713 0.001377 ... nan nan</div><input id='attrs-6546d1ab-6648-4273-a4ee-8606d9ca0780' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6546d1ab-6648-4273-a4ee-8606d9ca0780' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d8268951-4e0a-452d-9e95-a41758a5718e' class='xr-var-data-in' type='checkbox'><label for='data-d8268951-4e0a-452d-9e95-a41758a5718e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.00047125],\n",
       "       [0.00137741],\n",
       "       [0.00098425],\n",
       "       ...,\n",
       "       [0.03571429],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_modc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0004713 0.002066 ... nan nan</div><input id='attrs-0d14ddfa-3b40-4690-9eb4-a01ff1b01189' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0d14ddfa-3b40-4690-9eb4-a01ff1b01189' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-dcbbddb7-0617-4006-823c-2f3179cc0a0e' class='xr-var-data-in' type='checkbox'><label for='data-dcbbddb7-0617-4006-823c-2f3179cc0a0e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[4.71253534e-04],\n",
       "       [2.06611570e-03],\n",
       "       [2.95275591e-03],\n",
       "       ...,\n",
       "       [7.14285714e-01],\n",
       "       [           nan],\n",
       "       [           nan]])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-a904cae4-6fc3-4688-8973-f0c865061607' class='xr-section-summary-in' type='checkbox'  ><label for='section-a904cae4-6fc3-4688-8973-f0c865061607' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>sample_id</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-3196f721-6d35-4456-a9c4-2326658b4dff' class='xr-index-data-in' type='checkbox'/><label for='index-3196f721-6d35-4456-a9c4-2326658b4dff' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;sample_0&#x27;], dtype=&#x27;object&#x27;, name=&#x27;sample_id&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-950c0236-870d-4bec-88f5-1e0773a04064' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-950c0236-870d-4bec-88f5-1e0773a04064' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.Dataset>\n",
       "Dimensions:                   (ranges: 128103, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) <U2 '1' '1' '1' '1' ... 'Y' 'Y' 'Y' 'Y'\n",
       "    start                     (ranges) int64 4807587 4807691 ... 22859847\n",
       "    end                       (ranges) int64 4807787 4807891 ... 22861212\n",
       "    range_id                  (ranges) int64 0 1 2 3 ... 64152 64153 64154 64155\n",
       "    num_contexts              (ranges) int64 27 35 21 0 7 2 11 ... 0 1 0 1 1 0 0\n",
       "    range_length              (ranges) int64 200 200 200 200 ... 1361 1365 1365\n",
       "  * sample_id                 (sample_id) <U8 'sample_0'\n",
       "    strand                    (ranges) <U1 '+' '+' '+' '+' ... '-' '-' '-' '-'\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 0.0 2.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.0 0.05714 ... nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 1.0 4.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.03704 0.1143 ... nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 1.0 6.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.03704 0.1714 ... nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 2.122e+03 ... 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 78.59 82.97 ... nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    Id                        (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Gene_id                   (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Gene_type                 (ranges) object 'protein_coding' ... 'protein_c...\n",
       "    Gene_name                 (ranges) object 'Lypla1' 'Gm37988' ... 'Gm21366'\n",
       "    Level                     (ranges) object '2' '2' '2' '2' ... '2' '2' '2'\n",
       "    Mgi_id                    (ranges) object 'MGI:1344588' ... 'MGI:5434721'\n",
       "    Havana_gene               (ranges) object 'OTTMUSG00000021562.4' ... 'OTT...\n",
       "    Tag                       (ranges) object 'overlapping_locus' ... ''\n",
       "    Transcript_id             (ranges) object 'ENSMUST00000027036.10' ... 'EN...\n",
       "    Region                    (ranges) object 'before_tss' ... 'first_introns'\n",
       "    Strand                    (ranges) object '.' '.' '.' '.' ... '.' '.' '.'\n",
       "    Parent                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_type           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_name           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_number               (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_id                   (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_support_level  (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Havana_transcript         (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Protein_id                (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ccdsid                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ont                       (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Level_1                   (ranges) float64 nan nan nan nan ... 8.0 8.0 8.0\n",
       "    Ranges_id                 (ranges) float64 nan nan ... 2.164e+04 2.164e+04\n",
       "    Intron_number             (ranges) object nan nan nan nan ... '1' '1' '1'\n",
       "    mean_mc                   (ranges, sample_id) float64 0.0 0.0006887 ... nan\n",
       "    mean_hmc                  (ranges, sample_id) float64 0.0004713 ... nan\n",
       "    mean_modc                 (ranges, sample_id) float64 0.0004713 ... nan"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compute mean methylation levels as mc / total_c\n",
    "sense_ds = sense_ds.assign(\n",
    "    mean_mc = sense_ds[\"num_mc_sum\"] / sense_ds[\"num_total_c_sum\"],\n",
    "    mean_hmc = sense_ds[\"num_hmc_sum\"] / sense_ds[\"num_total_c_sum\"],\n",
    "    mean_modc = sense_ds[\"num_modc_sum\"] / sense_ds[\"num_total_c_sum\"],\n",
    ")\n",
    "sense_ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;\n",
       "Dimensions:                   (ranges: 128103, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) &lt;U2 &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;\n",
       "    start                     (ranges) int64 4807587 4807691 ... 22859847\n",
       "    end                       (ranges) int64 4807787 4807891 ... 22861212\n",
       "    range_id                  (ranges) int64 0 1 2 3 ... 64152 64153 64154 64155\n",
       "    num_contexts              (ranges) int64 27 35 22 0 7 2 11 ... 0 3 0 1 1 0 0\n",
       "    range_length              (ranges) int64 200 200 200 200 ... 1361 1365 1365\n",
       "  * sample_id                 (sample_id) &lt;U8 &#x27;sample_0&#x27;\n",
       "    strand                    (ranges) &lt;U1 &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; ... &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27;\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 1.0 4.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.03704 0.1143 ... nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 3.0 4.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.1111 0.1143 ... nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 4.0 8.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.1481 0.2286 ... nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 2.111e+03 ... 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 78.19 74.09 ... nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    Id                        (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Gene_id                   (ranges) object &#x27;ENSMUSG00000025903.14&#x27; ... &#x27;EN...\n",
       "    Gene_type                 (ranges) object &#x27;protein_coding&#x27; ... &#x27;protein_c...\n",
       "    Gene_name                 (ranges) object &#x27;Lypla1&#x27; &#x27;Gm37988&#x27; ... &#x27;Gm21366&#x27;\n",
       "    Level                     (ranges) object &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;\n",
       "    Mgi_id                    (ranges) object &#x27;MGI:1344588&#x27; ... &#x27;MGI:5434721&#x27;\n",
       "    Havana_gene               (ranges) object &#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTT...\n",
       "    Tag                       (ranges) object &#x27;overlapping_locus&#x27; ... &#x27;&#x27;\n",
       "    Transcript_id             (ranges) object &#x27;ENSMUST00000027036.10&#x27; ... &#x27;EN...\n",
       "    Region                    (ranges) object &#x27;before_tss&#x27; ... &#x27;first_introns&#x27;\n",
       "    Strand                    (ranges) object &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;\n",
       "    Parent                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_type           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_name           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_number               (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_id                   (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_support_level  (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Havana_transcript         (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Protein_id                (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ccdsid                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ont                       (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Level_1                   (ranges) float64 nan nan nan nan ... 8.0 8.0 8.0\n",
       "    Ranges_id                 (ranges) float64 nan nan ... 2.164e+04 2.164e+04\n",
       "    Intron_number             (ranges) object nan nan nan nan ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;\n",
       "    mean_mc                   (ranges, sample_id) float64 0.0004737 ... nan\n",
       "    mean_hmc                  (ranges, sample_id) float64 0.001421 ... nan\n",
       "    mean_modc                 (ranges, sample_id) float64 0.001895 ... nan</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-a101b390-1d3f-4442-b679-290f29b021ae' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-a101b390-1d3f-4442-b679-290f29b021ae' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>ranges</span>: 128103</li><li><span class='xr-has-index'>sample_id</span>: 1</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-aa3d0a88-a36d-4a75-b816-e2a784c42c2d' class='xr-section-summary-in' type='checkbox'  checked><label for='section-aa3d0a88-a36d-4a75-b816-e2a784c42c2d' class='xr-section-summary' >Coordinates: <span>(8)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>contig</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U2</div><div class='xr-var-preview xr-preview'>&#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; ... &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27; &#x27;Y&#x27;</div><input id='attrs-d2f34694-684d-4113-a065-4f91fcea0b4c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d2f34694-684d-4113-a065-4f91fcea0b4c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b5b4be73-2396-4a4d-a921-be054a1465c4' class='xr-var-data-in' type='checkbox'><label for='data-b5b4be73-2396-4a4d-a921-be054a1465c4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;, ..., &#x27;Y&#x27;, &#x27;Y&#x27;, &#x27;Y&#x27;], dtype=&#x27;&lt;U2&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>start</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4807587 4807691 ... 22859847</div><input id='attrs-14729cfb-58bb-40f7-92f9-7f9e60a2f638' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-14729cfb-58bb-40f7-92f9-7f9e60a2f638' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b0ef268e-99da-424e-8618-71899c13365d' class='xr-var-data-in' type='checkbox'><label for='data-b0ef268e-99da-424e-8618-71899c13365d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4807587,  4807691,  4857613, ..., 20992252, 22049944, 22859847])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>end</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>4807787 4807891 ... 22861212</div><input id='attrs-222b661c-b15a-40a5-9604-b4b7de7fc010' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-222b661c-b15a-40a5-9604-b4b7de7fc010' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-eb4f1b48-2f7a-4776-a3db-0ff95b414cc9' class='xr-var-data-in' type='checkbox'><label for='data-eb4f1b48-2f7a-4776-a3db-0ff95b414cc9' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 4807787,  4807891,  4857813, ..., 20993613, 22051309, 22861212])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 ... 64152 64153 64154 64155</div><input id='attrs-e00e06fd-3226-455f-b3ea-af0377079102' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e00e06fd-3226-455f-b3ea-af0377079102' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-75f30a92-886a-4fb9-8fed-58bdb3579c7b' class='xr-var-data-in' type='checkbox'><label for='data-75f30a92-886a-4fb9-8fed-58bdb3579c7b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([    0,     1,     2, ..., 64153, 64154, 64155])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_contexts</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>27 35 22 0 7 2 11 ... 0 3 0 1 1 0 0</div><input id='attrs-db0164d9-fd89-4382-bf5f-51bdc83c9eec' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-db0164d9-fd89-4382-bf5f-51bdc83c9eec' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8a5e72fb-7654-4449-aec1-59e09e2ebef2' class='xr-var-data-in' type='checkbox'><label for='data-8a5e72fb-7654-4449-aec1-59e09e2ebef2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([27, 35, 22, ...,  1,  0,  0])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>range_length</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>200 200 200 200 ... 1361 1365 1365</div><input id='attrs-9df72995-8615-4041-a279-82e9e204d65a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-9df72995-8615-4041-a279-82e9e204d65a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-325f7783-8339-4229-a71e-6eb37da27190' class='xr-var-data-in' type='checkbox'><label for='data-325f7783-8339-4229-a71e-6eb37da27190' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 200,  200,  200, ..., 1361, 1365, 1365])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample_id</span></div><div class='xr-var-dims'>(sample_id)</div><div class='xr-var-dtype'>&lt;U8</div><div class='xr-var-preview xr-preview'>&#x27;sample_0&#x27;</div><input id='attrs-7838cc83-110a-430b-b8a5-6a7dad724830' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7838cc83-110a-430b-b8a5-6a7dad724830' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-68d8e885-f8b1-4a55-997a-6e3f0525ccb3' class='xr-var-data-in' type='checkbox'><label for='data-68d8e885-f8b1-4a55-997a-6e3f0525ccb3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;sample_0&#x27;], dtype=&#x27;&lt;U8&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>&lt;U1</div><div class='xr-var-preview xr-preview'>&#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; &#x27;-&#x27; ... &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27; &#x27;+&#x27;</div><input id='attrs-2d0b38ee-a832-4b68-bbb7-5ad50b7c76e6' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2d0b38ee-a832-4b68-bbb7-5ad50b7c76e6' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-06f798d8-dcb5-4fae-92be-0eece16c1b8b' class='xr-var-data-in' type='checkbox'><label for='data-06f798d8-dcb5-4fae-92be-0eece16c1b8b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;-&#x27;, &#x27;-&#x27;, &#x27;-&#x27;, ..., &#x27;+&#x27;, &#x27;+&#x27;, &#x27;+&#x27;], dtype=&#x27;&lt;U1&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-2b67d171-3871-4e8f-9c26-79a590c31b65' class='xr-section-summary-in' type='checkbox'  ><label for='section-2b67d171-3871-4e8f-9c26-79a590c31b65' class='xr-section-summary' >Data variables: <span>(39)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 4.0 3.0 0.0 ... 19.0 0.0 0.0</div><input id='attrs-9a39628a-f554-4636-a806-bc6ca745304d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-9a39628a-f554-4636-a806-bc6ca745304d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-31ab53f3-3d61-4fac-9e7b-645d47dca01f' class='xr-var-data-in' type='checkbox'><label for='data-31ab53f3-3d61-4fac-9e7b-645d47dca01f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 1.],\n",
       "       [ 4.],\n",
       "       [ 3.],\n",
       "       ...,\n",
       "       [19.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.03704 0.1143 0.1364 ... nan nan</div><input id='attrs-2d4f60f7-bd45-4d74-bcb4-ec8602c5dbf7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2d4f60f7-bd45-4d74-bcb4-ec8602c5dbf7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ab4fc88a-b7ca-4b3a-a96d-440532ff3f8f' class='xr-var-data-in' type='checkbox'><label for='data-ab4fc88a-b7ca-4b3a-a96d-440532ff3f8f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.03703704],\n",
       "       [ 0.11428571],\n",
       "       [ 0.13636364],\n",
       "       ...,\n",
       "       [19.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_mc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 22.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-39e39d8d-1c01-4853-a729-dfe7d528f93b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-39e39d8d-1c01-4853-a729-dfe7d528f93b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-db6747c4-9973-4bbb-aadd-7caf3f74f2a5' class='xr-var-data-in' type='checkbox'><label for='data-db6747c4-9973-4bbb-aadd-7caf3f74f2a5' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [22.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>3.0 4.0 1.0 0.0 ... 0.0 1.0 0.0 0.0</div><input id='attrs-d6e9de94-0824-498d-a5d9-9d16dd21bd19' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d6e9de94-0824-498d-a5d9-9d16dd21bd19' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1eeb96be-cb89-42b9-8fa2-a91900f0e288' class='xr-var-data-in' type='checkbox'><label for='data-1eeb96be-cb89-42b9-8fa2-a91900f0e288' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[3.],\n",
       "       [4.],\n",
       "       [1.],\n",
       "       ...,\n",
       "       [1.],\n",
       "       [0.],\n",
       "       [0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.1111 0.1143 0.04545 ... nan nan</div><input id='attrs-596fa1f4-6f12-433b-92d1-e9bfa5dfb697' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-596fa1f4-6f12-433b-92d1-e9bfa5dfb697' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7a79daff-c79d-4661-9353-b4769666655e' class='xr-var-data-in' type='checkbox'><label for='data-7a79daff-c79d-4661-9353-b4769666655e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.11111111],\n",
       "       [0.11428571],\n",
       "       [0.04545455],\n",
       "       ...,\n",
       "       [1.        ],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_hmc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 22.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-7671e4a9-f07f-4264-bdaa-790e3d98ed4b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7671e4a9-f07f-4264-bdaa-790e3d98ed4b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-33649ae3-c8bf-4421-9c89-11f4c4726083' class='xr-var-data-in' type='checkbox'><label for='data-33649ae3-c8bf-4421-9c89-11f4c4726083' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [22.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>4.0 8.0 4.0 0.0 ... 21.0 0.0 0.0</div><input id='attrs-b443a446-f493-4d33-8a16-669f25e4b360' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b443a446-f493-4d33-8a16-669f25e4b360' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1260d3cd-3784-4d0f-bb61-0864a83dbd07' class='xr-var-data-in' type='checkbox'><label for='data-1260d3cd-3784-4d0f-bb61-0864a83dbd07' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 4.],\n",
       "       [ 8.],\n",
       "       [ 4.],\n",
       "       ...,\n",
       "       [21.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.1481 0.2286 0.1818 ... nan nan</div><input id='attrs-c6dcf90b-72a5-436f-a90a-75eed3aac028' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c6dcf90b-72a5-436f-a90a-75eed3aac028' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3d5cab12-db6f-49ee-923e-125842da8ee9' class='xr-var-data-in' type='checkbox'><label for='data-3d5cab12-db6f-49ee-923e-125842da8ee9' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 0.14814815],\n",
       "       [ 0.22857143],\n",
       "       [ 0.18181818],\n",
       "       ...,\n",
       "       [21.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_modc_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 22.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-b93e6c04-1142-4d04-b070-37d707a62e89' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b93e6c04-1142-4d04-b070-37d707a62e89' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5750065d-80f1-4a76-a833-5f086c51abe0' class='xr-var-data-in' type='checkbox'><label for='data-5750065d-80f1-4a76-a833-5f086c51abe0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [22.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_sum</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>2.111e+03 2.593e+03 ... 0.0 0.0</div><input id='attrs-c044447d-575f-4804-a512-1ffb817d202b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c044447d-575f-4804-a512-1ffb817d202b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3ec8fdb3-37b9-4898-bb56-c657cd782b0c' class='xr-var-data-in' type='checkbox'><label for='data-3ec8fdb3-37b9-4898-bb56-c657cd782b0c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[2111.],\n",
       "       [2593.],\n",
       "       [1982.],\n",
       "       ...,\n",
       "       [  29.],\n",
       "       [   0.],\n",
       "       [   0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_mean</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>78.19 74.09 90.09 ... 29.0 nan nan</div><input id='attrs-b33e410e-9e19-4273-858a-d0d61d944d8f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b33e410e-9e19-4273-858a-d0d61d944d8f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9ad231c4-d9f9-4e89-b6d5-270b130b41c3' class='xr-var-data-in' type='checkbox'><label for='data-9ad231c4-d9f9-4e89-b6d5-270b130b41c3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[78.18518519],\n",
       "       [74.08571429],\n",
       "       [90.09090909],\n",
       "       ...,\n",
       "       [29.        ],\n",
       "       [        nan],\n",
       "       [        nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>num_total_c_cpg_count</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>27.0 35.0 22.0 0.0 ... 1.0 0.0 0.0</div><input id='attrs-f0e209da-8f90-40e1-bad2-cd846458bef7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f0e209da-8f90-40e1-bad2-cd846458bef7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5e255b1b-c96a-4a52-a8fa-729942dd843e' class='xr-var-data-in' type='checkbox'><label for='data-5e255b1b-c96a-4a52-a8fa-729942dd843e' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[27.],\n",
       "       [35.],\n",
       "       [22.],\n",
       "       ...,\n",
       "       [ 1.],\n",
       "       [ 0.],\n",
       "       [ 0.]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-b8800199-c929-47a3-8c07-a51725c46f02' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b8800199-c929-47a3-8c07-a51725c46f02' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-40ee0641-afba-4553-8a4e-cd1ad33a13a9' class='xr-var-data-in' type='checkbox'><label for='data-40ee0641-afba-4553-8a4e-cd1ad33a13a9' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000104217.1&#x27;,\n",
       "       &#x27;ENSMUSG00000033813.15&#x27;, ..., &#x27;ENSMUSG00000102668.1&#x27;,\n",
       "       &#x27;ENSMUSG00000103528.1&#x27;, &#x27;ENSMUSG00000103919.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUSG00000025903.14&#x27; ... &#x27;ENS...</div><input id='attrs-793a7e38-73a8-4838-9195-9a1533f70c71' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-793a7e38-73a8-4838-9195-9a1533f70c71' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ec92aec3-fa3a-4853-a861-b26d5c48ee29' class='xr-var-data-in' type='checkbox'><label for='data-ec92aec3-fa3a-4853-a861-b26d5c48ee29' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUSG00000025903.14&#x27;, &#x27;ENSMUSG00000104217.1&#x27;,\n",
       "       &#x27;ENSMUSG00000033813.15&#x27;, ..., &#x27;ENSMUSG00000102668.1&#x27;,\n",
       "       &#x27;ENSMUSG00000103528.1&#x27;, &#x27;ENSMUSG00000103919.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;protein_coding&#x27; ... &#x27;protein_co...</div><input id='attrs-e73e9f0c-7c18-4f06-ab49-1f8641978203' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e73e9f0c-7c18-4f06-ab49-1f8641978203' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-f813e085-be0e-4a10-aed4-5a9e47689972' class='xr-var-data-in' type='checkbox'><label for='data-f813e085-be0e-4a10-aed4-5a9e47689972' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, ...,\n",
       "       &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;, &#x27;protein_coding&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Gene_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Lypla1&#x27; &#x27;Gm37988&#x27; ... &#x27;Gm21366&#x27;</div><input id='attrs-09d907f4-ad8b-45c7-a800-d4718b01fac6' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-09d907f4-ad8b-45c7-a800-d4718b01fac6' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-f33f4a14-a5a2-4dc2-ba67-b49ede76e907' class='xr-var-data-in' type='checkbox'><label for='data-f33f4a14-a5a2-4dc2-ba67-b49ede76e907' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Lypla1&#x27;, &#x27;Gm37988&#x27;, &#x27;Tcea1&#x27;, ..., &#x27;Gm29866&#x27;, &#x27;Gm31571&#x27;, &#x27;Gm21366&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; ... &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27; &#x27;2&#x27;</div><input id='attrs-679222c6-3f1c-412b-95ad-adf8ed87fcbc' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-679222c6-3f1c-412b-95ad-adf8ed87fcbc' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-252b5863-6d33-45cd-bae4-63f1f0602d0b' class='xr-var-data-in' type='checkbox'><label for='data-252b5863-6d33-45cd-bae4-63f1f0602d0b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;, ..., &#x27;2&#x27;, &#x27;2&#x27;, &#x27;2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Mgi_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;MGI:1344588&#x27; ... &#x27;MGI:5434721&#x27;</div><input id='attrs-c1420ebf-5f6f-4a4e-925f-0a084089f3fd' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c1420ebf-5f6f-4a4e-925f-0a084089f3fd' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-854f7e98-9249-46d3-b675-2e0d53949486' class='xr-var-data-in' type='checkbox'><label for='data-854f7e98-9249-46d3-b675-2e0d53949486' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;MGI:1344588&#x27;, &#x27;MGI:5611216&#x27;, &#x27;MGI:1196624&#x27;, ..., &#x27;MGI:5589025&#x27;,\n",
       "       &#x27;MGI:5590730&#x27;, &#x27;MGI:5434721&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_gene</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;OTTMUSG00000021562.4&#x27; ... &#x27;OTTM...</div><input id='attrs-69e8a9cf-264d-4bc9-9b2f-d9f97267b2ef' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-69e8a9cf-264d-4bc9-9b2f-d9f97267b2ef' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-054ad3d6-acd2-4ab1-ad85-dcc9c5572d51' class='xr-var-data-in' type='checkbox'><label for='data-054ad3d6-acd2-4ab1-ad85-dcc9c5572d51' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;OTTMUSG00000021562.4&#x27;, &#x27;OTTMUSG00000050100.1&#x27;,\n",
       "       &#x27;OTTMUSG00000042348.1&#x27;, ..., &#x27;OTTMUSG00000050180.1&#x27;,\n",
       "       &#x27;OTTMUSG00000050134.1&#x27;, &#x27;OTTMUSG00000050112.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Tag</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;overlapping_locus&#x27; ... &#x27;&#x27;</div><input id='attrs-2f13bfcd-22cf-45c2-ab02-076470486de2' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2f13bfcd-22cf-45c2-ab02-076470486de2' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-534255b2-d466-4e3b-aec7-ec8e81afd44d' class='xr-var-data-in' type='checkbox'><label for='data-534255b2-d466-4e3b-aec7-ec8e81afd44d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;overlapping_locus&#x27;, &#x27;overlapping_locus&#x27;, &#x27;overlapping_locus&#x27;, ...,\n",
       "       &#x27;&#x27;, &#x27;&#x27;, &#x27;&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;ENSMUST00000027036.10&#x27; ... &#x27;ENS...</div><input id='attrs-b542663a-afaa-42c0-b128-17967e1110c9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b542663a-afaa-42c0-b128-17967e1110c9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c10a4251-7d20-4872-98fc-fbe7b0b12280' class='xr-var-data-in' type='checkbox'><label for='data-c10a4251-7d20-4872-98fc-fbe7b0b12280' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;ENSMUST00000027036.10&#x27;, &#x27;ENSMUST00000155020.1&#x27;,\n",
       "       &#x27;ENSMUST00000081551.13&#x27;, ..., &#x27;ENSMUST00000191675.1&#x27;,\n",
       "       &#x27;ENSMUST00000192521.1&#x27;, &#x27;ENSMUST00000192892.1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Region</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;before_tss&#x27; ... &#x27;first_introns&#x27;</div><input id='attrs-48b2276e-a0b4-4eb0-82a0-c86b90dab004' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-48b2276e-a0b4-4eb0-82a0-c86b90dab004' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5fd7b181-923b-4d22-969e-aff237c1a145' class='xr-var-data-in' type='checkbox'><label for='data-5fd7b181-923b-4d22-969e-aff237c1a145' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;before_tss&#x27;, &#x27;before_tss&#x27;, &#x27;before_tss&#x27;, ..., &#x27;first_introns&#x27;,\n",
       "       &#x27;first_introns&#x27;, &#x27;first_introns&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Strand</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; ... &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27; &#x27;.&#x27;</div><input id='attrs-1c01dbbf-2117-49dc-8210-0436e372af27' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1c01dbbf-2117-49dc-8210-0436e372af27' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-01448e9c-5f07-4966-8341-a02c96c6af5b' class='xr-var-data-in' type='checkbox'><label for='data-01448e9c-5f07-4966-8341-a02c96c6af5b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;, ..., &#x27;.&#x27;, &#x27;.&#x27;, &#x27;.&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Parent</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-23e9f340-07c7-44c7-abff-85909d98a287' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-23e9f340-07c7-44c7-abff-85909d98a287' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-cbb53d67-5a4a-462f-8b68-df962c25a8bd' class='xr-var-data-in' type='checkbox'><label for='data-cbb53d67-5a4a-462f-8b68-df962c25a8bd' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_type</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-c78a3213-4c71-4249-8126-2816a10adb7f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c78a3213-4c71-4249-8126-2816a10adb7f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-abe8bac5-4f15-4c23-88fb-1662b44a18ba' class='xr-var-data-in' type='checkbox'><label for='data-abe8bac5-4f15-4c23-88fb-1662b44a18ba' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_name</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-4eb1f81e-2544-404d-844a-50087dfe1e79' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4eb1f81e-2544-404d-844a-50087dfe1e79' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8a545e45-19b0-4ff8-8545-3aa545b21e9f' class='xr-var-data-in' type='checkbox'><label for='data-8a545e45-19b0-4ff8-8545-3aa545b21e9f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-08f83562-6ede-462b-be75-a4e0fb8fc38c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-08f83562-6ede-462b-be75-a4e0fb8fc38c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-2c3bf54b-8701-466f-9d45-c080c6c638b4' class='xr-var-data-in' type='checkbox'><label for='data-2c3bf54b-8701-466f-9d45-c080c6c638b4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Exon_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-e3af1ca8-3912-437a-b3e5-25e8b9e62a12' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e3af1ca8-3912-437a-b3e5-25e8b9e62a12' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-76312126-615f-4f9f-9a29-238797ee8f12' class='xr-var-data-in' type='checkbox'><label for='data-76312126-615f-4f9f-9a29-238797ee8f12' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Transcript_support_level</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-82553b71-513c-403c-baf9-86d1cd628b9b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-82553b71-513c-403c-baf9-86d1cd628b9b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ffff9c83-412c-4c0c-9c6e-a6766ab0939c' class='xr-var-data-in' type='checkbox'><label for='data-ffff9c83-412c-4c0c-9c6e-a6766ab0939c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Havana_transcript</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-f069df81-da3f-4015-9c7c-8e987faa44fd' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f069df81-da3f-4015-9c7c-8e987faa44fd' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-35470bec-2945-40c6-ae22-defd23623a93' class='xr-var-data-in' type='checkbox'><label for='data-35470bec-2945-40c6-ae22-defd23623a93' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Protein_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-880bc4fa-f8f1-466f-8030-66a14e50818f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-880bc4fa-f8f1-466f-8030-66a14e50818f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0ee5c956-7e22-47ce-b052-29c4172702db' class='xr-var-data-in' type='checkbox'><label for='data-0ee5c956-7e22-47ce-b052-29c4172702db' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ccdsid</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-86e1d1ee-3cfc-4e0d-aa74-24c4f9293291' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-86e1d1ee-3cfc-4e0d-aa74-24c4f9293291' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b9c28ca6-61e4-4ad7-bb55-e456fcf83f19' class='xr-var-data-in' type='checkbox'><label for='data-b9c28ca6-61e4-4ad7-bb55-e456fcf83f19' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ont</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... nan nan nan nan</div><input id='attrs-5470bc4b-af87-430e-87e9-14d176af0eeb' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5470bc4b-af87-430e-87e9-14d176af0eeb' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d7836edd-75bc-49d8-8f38-918b775a0fdf' class='xr-var-data-in' type='checkbox'><label for='data-d7836edd-75bc-49d8-8f38-918b775a0fdf' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., nan, nan, nan], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Level_1</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... 8.0 8.0 8.0 8.0</div><input id='attrs-b3653b8a-b444-42ed-a64a-eee212fd8278' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b3653b8a-b444-42ed-a64a-eee212fd8278' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-060d01fb-7916-4b9d-b606-71b1648d658c' class='xr-var-data-in' type='checkbox'><label for='data-060d01fb-7916-4b9d-b606-71b1648d658c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ...,  8.,  8.,  8.])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Ranges_id</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>nan nan nan ... 2.164e+04 2.164e+04</div><input id='attrs-56fe091c-3cb2-4720-947d-c8bafc582431' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-56fe091c-3cb2-4720-947d-c8bafc582431' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-27e52b99-fb10-4850-82b9-3a97cda79837' class='xr-var-data-in' type='checkbox'><label for='data-27e52b99-fb10-4850-82b9-3a97cda79837' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([   nan,    nan,    nan, ..., 21635., 21636., 21637.])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>Intron_number</span></div><div class='xr-var-dims'>(ranges)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan nan ... &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27; &#x27;1&#x27;</div><input id='attrs-2543e430-a869-4bd0-b72e-3de675c6bef1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2543e430-a869-4bd0-b72e-3de675c6bef1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d233cb04-2674-45be-a90e-a9957babca52' class='xr-var-data-in' type='checkbox'><label for='data-d233cb04-2674-45be-a90e-a9957babca52' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([nan, nan, nan, ..., &#x27;1&#x27;, &#x27;1&#x27;, &#x27;1&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_mc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0004737 0.001543 ... nan nan</div><input id='attrs-919ae1ba-46f6-4560-95c0-50e0990a2df4' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-919ae1ba-46f6-4560-95c0-50e0990a2df4' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e063538a-8fe8-46f1-b3b4-d771b1bbcc32' class='xr-var-data-in' type='checkbox'><label for='data-e063538a-8fe8-46f1-b3b4-d771b1bbcc32' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[4.73709143e-04],\n",
       "       [1.54261473e-03],\n",
       "       [1.51362260e-03],\n",
       "       ...,\n",
       "       [6.55172414e-01],\n",
       "       [           nan],\n",
       "       [           nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_hmc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.001421 0.001543 ... nan nan</div><input id='attrs-9aed9166-ede0-485a-abb5-fafdd107ffc4' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-9aed9166-ede0-485a-abb5-fafdd107ffc4' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-f1226938-0ac5-4ed8-9404-7eacbd111ab2' class='xr-var-data-in' type='checkbox'><label for='data-f1226938-0ac5-4ed8-9404-7eacbd111ab2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.00142113],\n",
       "       [0.00154261],\n",
       "       [0.00050454],\n",
       "       ...,\n",
       "       [0.03448276],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mean_modc</span></div><div class='xr-var-dims'>(ranges, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.001895 0.003085 ... nan nan</div><input id='attrs-4f9247d6-5ad1-4aad-ad03-81623abdfc33' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4f9247d6-5ad1-4aad-ad03-81623abdfc33' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9c3b6c5f-71b7-404f-ab63-58df78b88087' class='xr-var-data-in' type='checkbox'><label for='data-9c3b6c5f-71b7-404f-ab63-58df78b88087' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.00189484],\n",
       "       [0.00308523],\n",
       "       [0.00201816],\n",
       "       ...,\n",
       "       [0.72413793],\n",
       "       [       nan],\n",
       "       [       nan]])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-074ec1b4-6727-4dab-b5ca-723d3aa7b66c' class='xr-section-summary-in' type='checkbox'  ><label for='section-074ec1b4-6727-4dab-b5ca-723d3aa7b66c' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>sample_id</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-e80f6b91-b2c6-49c8-8cce-2694fc961117' class='xr-index-data-in' type='checkbox'/><label for='index-e80f6b91-b2c6-49c8-8cce-2694fc961117' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;sample_0&#x27;], dtype=&#x27;object&#x27;, name=&#x27;sample_id&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-00feccec-b07b-4cfe-92a4-d5ff8230d9d6' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-00feccec-b07b-4cfe-92a4-d5ff8230d9d6' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.Dataset>\n",
       "Dimensions:                   (ranges: 128103, sample_id: 1)\n",
       "Coordinates:\n",
       "    contig                    (ranges) <U2 '1' '1' '1' '1' ... 'Y' 'Y' 'Y' 'Y'\n",
       "    start                     (ranges) int64 4807587 4807691 ... 22859847\n",
       "    end                       (ranges) int64 4807787 4807891 ... 22861212\n",
       "    range_id                  (ranges) int64 0 1 2 3 ... 64152 64153 64154 64155\n",
       "    num_contexts              (ranges) int64 27 35 22 0 7 2 11 ... 0 3 0 1 1 0 0\n",
       "    range_length              (ranges) int64 200 200 200 200 ... 1361 1365 1365\n",
       "  * sample_id                 (sample_id) <U8 'sample_0'\n",
       "    strand                    (ranges) <U1 '-' '-' '-' '-' ... '+' '+' '+' '+'\n",
       "Dimensions without coordinates: ranges\n",
       "Data variables:\n",
       "    num_mc_sum                (ranges, sample_id) float64 1.0 4.0 ... 0.0 0.0\n",
       "    num_mc_mean               (ranges, sample_id) float64 0.03704 0.1143 ... nan\n",
       "    num_mc_cpg_count          (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_hmc_sum               (ranges, sample_id) float64 3.0 4.0 ... 0.0 0.0\n",
       "    num_hmc_mean              (ranges, sample_id) float64 0.1111 0.1143 ... nan\n",
       "    num_hmc_cpg_count         (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_modc_sum              (ranges, sample_id) float64 4.0 8.0 ... 0.0 0.0\n",
       "    num_modc_mean             (ranges, sample_id) float64 0.1481 0.2286 ... nan\n",
       "    num_modc_cpg_count        (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    num_total_c_sum           (ranges, sample_id) float64 2.111e+03 ... 0.0\n",
       "    num_total_c_mean          (ranges, sample_id) float64 78.19 74.09 ... nan\n",
       "    num_total_c_cpg_count     (ranges, sample_id) float64 27.0 35.0 ... 0.0 0.0\n",
       "    Id                        (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Gene_id                   (ranges) object 'ENSMUSG00000025903.14' ... 'EN...\n",
       "    Gene_type                 (ranges) object 'protein_coding' ... 'protein_c...\n",
       "    Gene_name                 (ranges) object 'Lypla1' 'Gm37988' ... 'Gm21366'\n",
       "    Level                     (ranges) object '2' '2' '2' '2' ... '2' '2' '2'\n",
       "    Mgi_id                    (ranges) object 'MGI:1344588' ... 'MGI:5434721'\n",
       "    Havana_gene               (ranges) object 'OTTMUSG00000021562.4' ... 'OTT...\n",
       "    Tag                       (ranges) object 'overlapping_locus' ... ''\n",
       "    Transcript_id             (ranges) object 'ENSMUST00000027036.10' ... 'EN...\n",
       "    Region                    (ranges) object 'before_tss' ... 'first_introns'\n",
       "    Strand                    (ranges) object '.' '.' '.' '.' ... '.' '.' '.'\n",
       "    Parent                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_type           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_name           (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_number               (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Exon_id                   (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Transcript_support_level  (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Havana_transcript         (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Protein_id                (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ccdsid                    (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Ont                       (ranges) object nan nan nan nan ... nan nan nan\n",
       "    Level_1                   (ranges) float64 nan nan nan nan ... 8.0 8.0 8.0\n",
       "    Ranges_id                 (ranges) float64 nan nan ... 2.164e+04 2.164e+04\n",
       "    Intron_number             (ranges) object nan nan nan nan ... '1' '1' '1'\n",
       "    mean_mc                   (ranges, sample_id) float64 0.0004737 ... nan\n",
       "    mean_hmc                  (ranges, sample_id) float64 0.001421 ... nan\n",
       "    mean_modc                 (ranges, sample_id) float64 0.001895 ... nan"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "antisense_ds = antisense_ds.assign(\n",
    "    mean_mc = antisense_ds[\"num_mc_sum\"] / antisense_ds[\"num_total_c_sum\"],\n",
    "    mean_hmc = antisense_ds[\"num_hmc_sum\"] / antisense_ds[\"num_total_c_sum\"],\n",
    "    mean_modc = antisense_ds[\"num_modc_sum\"] / antisense_ds[\"num_total_c_sum\"],\n",
    ")\n",
    "antisense_ds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "drop ranges where there are not C coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Count of instances where num_total_c_sum == 0: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(9529)\n",
      "Total number of instances: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(128103)\n",
      "Proportion of instances where num_total_c_sum (sense_ds) == 0: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(0.07438546)\n",
      "Count of instances where num_total_c_sum == 0 after dropping: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(0)\n"
     ]
    }
   ],
   "source": [
    "# Count instances where num_total_c_sum equals 0\n",
    "count_zeros = (sense_ds['num_total_c_sum'] == 0).sum().compute()\n",
    "print(f\"Count of instances where num_total_c_sum == 0: {count_zeros}\")\n",
    "\n",
    "# Total number of instances\n",
    "total_count = sense_ds['num_total_c_sum'].count().compute()\n",
    "print(f\"Total number of instances: {total_count}\")\n",
    "\n",
    "# Calculate the proportion of instances where num_total_c_sum equals 0\n",
    "proportion_zeros = count_zeros / total_count\n",
    "print(f\"Proportion of instances where num_total_c_sum (sense_ds) == 0: {proportion_zeros}\")\n",
    "\n",
    "# Create a mask where num_total_c_sum is greater than 0\n",
    "mask = sense_ds['num_total_c_sum'] > 0\n",
    "\n",
    "# Apply the mask to filter out rows where num_total_c_sum is 0 and drop them\n",
    "filtered_sense_ds = sense_ds.where(mask, drop=True)\n",
    "\n",
    "count_zeros = (filtered_sense_ds['num_total_c_sum'] == 0).sum().compute()\n",
    "print(f\"Count of instances where num_total_c_sum == 0 after dropping: {count_zeros}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Count of instances where num_total_c_sum == 0: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(9559)\n",
      "Total number of instances: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(128103)\n",
      "Proportion of instances where num_total_c_sum (antisense_ds) == 0: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(0.07461964)\n",
      "Count of instances where num_total_c_sum == 0 after dropping: <xarray.DataArray 'num_total_c_sum' ()>\n",
      "array(0)\n"
     ]
    }
   ],
   "source": [
    "# Count instances where num_total_c_sum equals 0\n",
    "count_zeros = (antisense_ds['num_total_c_sum'] == 0).sum().compute()\n",
    "print(f\"Count of instances where num_total_c_sum == 0: {count_zeros}\")\n",
    "\n",
    "# Total number of instances\n",
    "total_count = antisense_ds['num_total_c_sum'].count().compute()\n",
    "print(f\"Total number of instances: {total_count}\")\n",
    "\n",
    "# Calculate the proportion of instances where num_total_c_sum equals 0\n",
    "proportion_zeros = count_zeros / total_count\n",
    "print(f\"Proportion of instances where num_total_c_sum (antisense_ds) == 0: {proportion_zeros}\")\n",
    "\n",
    "# Create a mask where num_total_c_sum is greater than 0\n",
    "mask = antisense_ds['num_total_c_sum'] > 0\n",
    "\n",
    "# Apply the mask to filter out rows where num_total_c_sum is 0 and drop them\n",
    "filtered_antisense_ds = antisense_ds.where(mask, drop=True)\n",
    "\n",
    "count_zeros = (filtered_antisense_ds['num_total_c_sum'] == 0).sum().compute()\n",
    "print(f\"Count of instances where num_total_c_sum == 0 after dropping: {count_zeros}\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "no missing transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_transcript_genes = filtered_sense_ds[\"Transcript_id\"].where(filtered_sense_ds[\"Transcript_id\"].isnull(), drop=True)\n",
    "if len(missing_transcript_genes) > 0:\n",
    "    print(\"SENSE Gene IDs with missing Transcript IDs:\")\n",
    "    print(missing_transcript_genes)\n",
    "\n",
    "missing_transcript_genes = filtered_antisense_ds[\"Transcript_id\"].where(filtered_antisense_ds[\"Transcript_id\"].isnull(), drop=True)\n",
    "if len(missing_transcript_genes) > 0:\n",
    "    print(\"ANTISENSE Gene IDs with missing Transcript IDs:\")\n",
    "    print(missing_transcript_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "exon introns 3\\`UTRs 5\\`UTRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "stitched_sense_regions['Sense'] = \"sense\"\n",
    "stitched_antisense_regions['Sense'] = \"antisense\"\n",
    "stitched_antisense_regions = stitched_antisense_regions.drop([\"range_length\", \"cpg_count\"], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>total_c</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "      <th>Sense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.4</td>\n",
       "      <td>0.807980</td>\n",
       "      <td>0.025436</td>\n",
       "      <td>0.847215</td>\n",
       "      <td>51.0</td>\n",
       "      <td>6015.0</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000001.4</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003.15</td>\n",
       "      <td>0.774194</td>\n",
       "      <td>0.014888</td>\n",
       "      <td>0.806452</td>\n",
       "      <td>7.0</td>\n",
       "      <td>403.0</td>\n",
       "      <td>681.0</td>\n",
       "      <td>Pbsn</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000003.13</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028.15</td>\n",
       "      <td>0.581295</td>\n",
       "      <td>0.031589</td>\n",
       "      <td>0.619697</td>\n",
       "      <td>60.0</td>\n",
       "      <td>6458.0</td>\n",
       "      <td>1955.0</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000028.13</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000037.17</td>\n",
       "      <td>0.758328</td>\n",
       "      <td>0.017636</td>\n",
       "      <td>0.786414</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1531.0</td>\n",
       "      <td>2777.0</td>\n",
       "      <td>Scml2</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000101113.8</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000049.11</td>\n",
       "      <td>0.783618</td>\n",
       "      <td>0.036519</td>\n",
       "      <td>0.835495</td>\n",
       "      <td>28.0</td>\n",
       "      <td>2930.0</td>\n",
       "      <td>1068.0</td>\n",
       "      <td>Apoh</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000000049.5</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15460</th>\n",
       "      <td>ENSMUSG00000117975.1</td>\n",
       "      <td>0.778394</td>\n",
       "      <td>0.042044</td>\n",
       "      <td>0.833577</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3425.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Itprip</td>\n",
       "      <td>19</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000095998.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15461</th>\n",
       "      <td>ENSMUSG00000118219.1</td>\n",
       "      <td>0.424810</td>\n",
       "      <td>0.083937</td>\n",
       "      <td>0.515992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5659.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gm29695</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000188175.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15462</th>\n",
       "      <td>ENSMUSG00000118332.1</td>\n",
       "      <td>0.813403</td>\n",
       "      <td>0.045638</td>\n",
       "      <td>0.870884</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3462.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fam220a</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000119488.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15464</th>\n",
       "      <td>ENSMUSG00000118537.1</td>\n",
       "      <td>0.766355</td>\n",
       "      <td>0.009346</td>\n",
       "      <td>0.813084</td>\n",
       "      <td>NaN</td>\n",
       "      <td>107.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shld3</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000133280.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15465</th>\n",
       "      <td>ENSMUSG00000118560.1</td>\n",
       "      <td>0.603896</td>\n",
       "      <td>0.077922</td>\n",
       "      <td>0.694805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>154.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lim2</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000004732.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>130372 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  \\\n",
       "0       ENSMUSG00000000001.4  0.807980  0.025436   0.847215       51.0   \n",
       "1      ENSMUSG00000000003.15  0.774194  0.014888   0.806452        7.0   \n",
       "2      ENSMUSG00000000028.15  0.581295  0.031589   0.619697       60.0   \n",
       "3      ENSMUSG00000000037.17  0.758328  0.017636   0.786414       30.0   \n",
       "4      ENSMUSG00000000049.11  0.783618  0.036519   0.835495       28.0   \n",
       "...                      ...       ...       ...        ...        ...   \n",
       "15460   ENSMUSG00000117975.1  0.778394  0.042044   0.833577        NaN   \n",
       "15461   ENSMUSG00000118219.1  0.424810  0.083937   0.515992        NaN   \n",
       "15462   ENSMUSG00000118332.1  0.813403  0.045638   0.870884        NaN   \n",
       "15464   ENSMUSG00000118537.1  0.766355  0.009346   0.813084        NaN   \n",
       "15465   ENSMUSG00000118560.1  0.603896  0.077922   0.694805        NaN   \n",
       "\n",
       "       total_c  range_length Gene_name contig strand          Transcript_id  \\\n",
       "0       6015.0        2995.0     Gnai3      3      -   ENSMUST00000000001.4   \n",
       "1        403.0         681.0      Pbsn      X      -  ENSMUST00000000003.13   \n",
       "2       6458.0        1955.0     Cdc45     16      -  ENSMUST00000000028.13   \n",
       "3       1531.0        2777.0     Scml2      X      +   ENSMUST00000101113.8   \n",
       "4       2930.0        1068.0      Apoh     11      +   ENSMUST00000000049.5   \n",
       "...        ...           ...       ...    ...    ...                    ...   \n",
       "15460   3425.0           NaN    Itprip     19      +   ENSMUST00000095998.6   \n",
       "15461   5659.0           NaN   Gm29695      1      -   ENSMUST00000188175.1   \n",
       "15462   3462.0           NaN   Fam220a      5      -   ENSMUST00000119488.1   \n",
       "15464    107.0           NaN     Shld3     13      +   ENSMUST00000133280.1   \n",
       "15465    154.0           NaN      Lim2      7      -   ENSMUST00000004732.6   \n",
       "\n",
       "                 Region      Sense  \n",
       "0                 exons      sense  \n",
       "1                 exons      sense  \n",
       "2                 exons      sense  \n",
       "3                 exons      sense  \n",
       "4                 exons      sense  \n",
       "...                 ...        ...  \n",
       "15460  three_prime_utrs  antisense  \n",
       "15461  three_prime_utrs  antisense  \n",
       "15462  three_prime_utrs  antisense  \n",
       "15464  three_prime_utrs  antisense  \n",
       "15465  three_prime_utrs  antisense  \n",
       "\n",
       "[130372 rows x 13 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stitched_regions = pd.concat([stitched_sense_regions, stitched_antisense_regions])\n",
    "stitched_regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TSS TES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sense = filtered_sense_ds.to_dataframe().reset_index(level=\"sample_id\", drop=True)\n",
    "df_antisense = filtered_antisense_ds.to_dataframe().reset_index(level=\"sample_id\", drop=True)\n",
    "\n",
    "df_sense[\"Sense\"] = \"sense\"\n",
    "df_antisense[\"Sense\"] = \"antisense\"\n",
    "df_antisense = df_antisense.drop(['range_length','num_total_c_cpg_count'],axis=1)\n",
    "df = pd.concat([df_sense, df_antisense])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_mc_sum</th>\n",
       "      <th>num_mc_mean</th>\n",
       "      <th>num_mc_cpg_count</th>\n",
       "      <th>num_hmc_sum</th>\n",
       "      <th>num_hmc_mean</th>\n",
       "      <th>num_hmc_cpg_count</th>\n",
       "      <th>num_modc_sum</th>\n",
       "      <th>num_modc_mean</th>\n",
       "      <th>num_modc_cpg_count</th>\n",
       "      <th>num_total_c_sum</th>\n",
       "      <th>...</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>range_id</th>\n",
       "      <th>num_contexts</th>\n",
       "      <th>range_length</th>\n",
       "      <th>strand</th>\n",
       "      <th>Sense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ranges</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.037037</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.037037</td>\n",
       "      <td>27.0</td>\n",
       "      <td>2122.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000471</td>\n",
       "      <td>0.000471</td>\n",
       "      <td>1</td>\n",
       "      <td>4807587</td>\n",
       "      <td>4807787</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>200.0</td>\n",
       "      <td>+</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>35.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.114286</td>\n",
       "      <td>35.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.171429</td>\n",
       "      <td>35.0</td>\n",
       "      <td>2904.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001377</td>\n",
       "      <td>0.002066</td>\n",
       "      <td>1</td>\n",
       "      <td>4807691</td>\n",
       "      <td>4807891</td>\n",
       "      <td>1</td>\n",
       "      <td>35</td>\n",
       "      <td>200.0</td>\n",
       "      <td>+</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4.0</td>\n",
       "      <td>0.190476</td>\n",
       "      <td>21.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>21.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>21.0</td>\n",
       "      <td>2032.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000984</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>1</td>\n",
       "      <td>4857613</td>\n",
       "      <td>4857813</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>200.0</td>\n",
       "      <td>+</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>136.0</td>\n",
       "      <td>19.428571</td>\n",
       "      <td>7.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>3.285714</td>\n",
       "      <td>7.0</td>\n",
       "      <td>159.0</td>\n",
       "      <td>22.714286</td>\n",
       "      <td>7.0</td>\n",
       "      <td>501.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.045908</td>\n",
       "      <td>0.317365</td>\n",
       "      <td>1</td>\n",
       "      <td>5588265</td>\n",
       "      <td>5588465</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>200.0</td>\n",
       "      <td>+</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>145.0</td>\n",
       "      <td>72.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>153.0</td>\n",
       "      <td>76.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>202.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034653</td>\n",
       "      <td>0.757426</td>\n",
       "      <td>1</td>\n",
       "      <td>6205996</td>\n",
       "      <td>6206196</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>200.0</td>\n",
       "      <td>+</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118539</th>\n",
       "      <td>43.0</td>\n",
       "      <td>21.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>23.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>0.734375</td>\n",
       "      <td>Y</td>\n",
       "      <td>6917397</td>\n",
       "      <td>6918160</td>\n",
       "      <td>64142</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118540</th>\n",
       "      <td>63.0</td>\n",
       "      <td>10.500000</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>11.166667</td>\n",
       "      <td>6.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.837500</td>\n",
       "      <td>Y</td>\n",
       "      <td>42580848</td>\n",
       "      <td>42582206</td>\n",
       "      <td>64143</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118541</th>\n",
       "      <td>21.0</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.666667</td>\n",
       "      <td>3.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>0.675000</td>\n",
       "      <td>Y</td>\n",
       "      <td>68831319</td>\n",
       "      <td>68832684</td>\n",
       "      <td>64150</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118542</th>\n",
       "      <td>31.0</td>\n",
       "      <td>31.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.939394</td>\n",
       "      <td>Y</td>\n",
       "      <td>40315473</td>\n",
       "      <td>40316859</td>\n",
       "      <td>64152</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118543</th>\n",
       "      <td>19.0</td>\n",
       "      <td>19.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>21.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.724138</td>\n",
       "      <td>Y</td>\n",
       "      <td>20992252</td>\n",
       "      <td>20993613</td>\n",
       "      <td>64153</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237118 rows × 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        num_mc_sum  num_mc_mean  num_mc_cpg_count  num_hmc_sum  num_hmc_mean  \\\n",
       "ranges                                                                         \n",
       "0              0.0     0.000000              27.0          1.0      0.037037   \n",
       "1              2.0     0.057143              35.0          4.0      0.114286   \n",
       "2              4.0     0.190476              21.0          2.0      0.095238   \n",
       "3            136.0    19.428571               7.0         23.0      3.285714   \n",
       "4            145.0    72.500000               2.0          7.0      3.500000   \n",
       "...            ...          ...               ...          ...           ...   \n",
       "118539        43.0    21.500000               2.0          3.0      1.500000   \n",
       "118540        63.0    10.500000               6.0          0.0      0.000000   \n",
       "118541        21.0     7.000000               3.0          5.0      1.666667   \n",
       "118542        31.0    31.000000               1.0          0.0      0.000000   \n",
       "118543        19.0    19.000000               1.0          1.0      1.000000   \n",
       "\n",
       "        num_hmc_cpg_count  num_modc_sum  num_modc_mean  num_modc_cpg_count  \\\n",
       "ranges                                                                       \n",
       "0                    27.0           1.0       0.037037                27.0   \n",
       "1                    35.0           6.0       0.171429                35.0   \n",
       "2                    21.0           6.0       0.285714                21.0   \n",
       "3                     7.0         159.0      22.714286                 7.0   \n",
       "4                     2.0         153.0      76.500000                 2.0   \n",
       "...                   ...           ...            ...                 ...   \n",
       "118539                2.0          47.0      23.500000                 2.0   \n",
       "118540                6.0          67.0      11.166667                 6.0   \n",
       "118541                3.0          27.0       9.000000                 3.0   \n",
       "118542                1.0          31.0      31.000000                 1.0   \n",
       "118543                1.0          21.0      21.000000                 1.0   \n",
       "\n",
       "        num_total_c_sum  ...  mean_hmc  mean_modc contig     start       end  \\\n",
       "ranges                   ...                                                   \n",
       "0                2122.0  ...  0.000471   0.000471      1   4807587   4807787   \n",
       "1                2904.0  ...  0.001377   0.002066      1   4807691   4807891   \n",
       "2                2032.0  ...  0.000984   0.002953      1   4857613   4857813   \n",
       "3                 501.0  ...  0.045908   0.317365      1   5588265   5588465   \n",
       "4                 202.0  ...  0.034653   0.757426      1   6205996   6206196   \n",
       "...                 ...  ...       ...        ...    ...       ...       ...   \n",
       "118539             64.0  ...  0.046875   0.734375      Y   6917397   6918160   \n",
       "118540             80.0  ...  0.000000   0.837500      Y  42580848  42582206   \n",
       "118541             40.0  ...  0.125000   0.675000      Y  68831319  68832684   \n",
       "118542             33.0  ...  0.000000   0.939394      Y  40315473  40316859   \n",
       "118543             29.0  ...  0.034483   0.724138      Y  20992252  20993613   \n",
       "\n",
       "       range_id num_contexts range_length strand      Sense  \n",
       "ranges                                                       \n",
       "0             0           27        200.0      +      sense  \n",
       "1             1           35        200.0      +      sense  \n",
       "2             2           21        200.0      +      sense  \n",
       "3             4            7        200.0      +      sense  \n",
       "4             5            2        200.0      +      sense  \n",
       "...         ...          ...          ...    ...        ...  \n",
       "118539    64142            2          NaN      +  antisense  \n",
       "118540    64143            6          NaN      +  antisense  \n",
       "118541    64150            3          NaN      +  antisense  \n",
       "118542    64152            1          NaN      +  antisense  \n",
       "118543    64153            1          NaN      +  antisense  \n",
       "\n",
       "[237118 rows x 47 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>mean_mc</th>\n",
       "      <th>mean_hmc</th>\n",
       "      <th>mean_modc</th>\n",
       "      <th>cpg_count</th>\n",
       "      <th>range_length</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>Transcript_id</th>\n",
       "      <th>Region</th>\n",
       "      <th>Sense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000471</td>\n",
       "      <td>0.000471</td>\n",
       "      <td>27.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>0.000689</td>\n",
       "      <td>0.001377</td>\n",
       "      <td>0.002066</td>\n",
       "      <td>35.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000155020.1</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.000984</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>21.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000081551.13</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>0.271457</td>\n",
       "      <td>0.045908</td>\n",
       "      <td>0.317365</td>\n",
       "      <td>7.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Oprk1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000160777.7</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000025907.14</td>\n",
       "      <td>0.717822</td>\n",
       "      <td>0.034653</td>\n",
       "      <td>0.757426</td>\n",
       "      <td>2.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Rb1cc1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000027040.12</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367485</th>\n",
       "      <td>ENSMUSG00000117975.1</td>\n",
       "      <td>0.778394</td>\n",
       "      <td>0.042044</td>\n",
       "      <td>0.833577</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Itprip</td>\n",
       "      <td>19</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000095998.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367486</th>\n",
       "      <td>ENSMUSG00000118219.1</td>\n",
       "      <td>0.424810</td>\n",
       "      <td>0.083937</td>\n",
       "      <td>0.515992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gm29695</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000188175.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367487</th>\n",
       "      <td>ENSMUSG00000118332.1</td>\n",
       "      <td>0.813403</td>\n",
       "      <td>0.045638</td>\n",
       "      <td>0.870884</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fam220a</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000119488.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367488</th>\n",
       "      <td>ENSMUSG00000118537.1</td>\n",
       "      <td>0.766355</td>\n",
       "      <td>0.009346</td>\n",
       "      <td>0.813084</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shld3</td>\n",
       "      <td>13</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUST00000133280.1</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367489</th>\n",
       "      <td>ENSMUSG00000118560.1</td>\n",
       "      <td>0.603896</td>\n",
       "      <td>0.077922</td>\n",
       "      <td>0.694805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lim2</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000004732.6</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>367490 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene_id   mean_mc  mean_hmc  mean_modc  cpg_count  \\\n",
       "0       ENSMUSG00000025903.14  0.000000  0.000471   0.000471       27.0   \n",
       "1        ENSMUSG00000104217.1  0.000689  0.001377   0.002066       35.0   \n",
       "2       ENSMUSG00000033813.15  0.001969  0.000984   0.002953       21.0   \n",
       "3       ENSMUSG00000025905.14  0.271457  0.045908   0.317365        7.0   \n",
       "4       ENSMUSG00000025907.14  0.717822  0.034653   0.757426        2.0   \n",
       "...                       ...       ...       ...        ...        ...   \n",
       "367485   ENSMUSG00000117975.1  0.778394  0.042044   0.833577        NaN   \n",
       "367486   ENSMUSG00000118219.1  0.424810  0.083937   0.515992        NaN   \n",
       "367487   ENSMUSG00000118332.1  0.813403  0.045638   0.870884        NaN   \n",
       "367488   ENSMUSG00000118537.1  0.766355  0.009346   0.813084        NaN   \n",
       "367489   ENSMUSG00000118560.1  0.603896  0.077922   0.694805        NaN   \n",
       "\n",
       "        range_length Gene_name contig strand          Transcript_id  \\\n",
       "0              200.0    Lypla1      1      +  ENSMUST00000027036.10   \n",
       "1              200.0   Gm37988      1      +   ENSMUST00000155020.1   \n",
       "2              200.0     Tcea1      1      +  ENSMUST00000081551.13   \n",
       "3              200.0     Oprk1      1      +   ENSMUST00000160777.7   \n",
       "4              200.0    Rb1cc1      1      +  ENSMUST00000027040.12   \n",
       "...              ...       ...    ...    ...                    ...   \n",
       "367485           NaN    Itprip     19      +   ENSMUST00000095998.6   \n",
       "367486           NaN   Gm29695      1      -   ENSMUST00000188175.1   \n",
       "367487           NaN   Fam220a      5      -   ENSMUST00000119488.1   \n",
       "367488           NaN     Shld3     13      +   ENSMUST00000133280.1   \n",
       "367489           NaN      Lim2      7      -   ENSMUST00000004732.6   \n",
       "\n",
       "                  Region      Sense  \n",
       "0             before_tss      sense  \n",
       "1             before_tss      sense  \n",
       "2             before_tss      sense  \n",
       "3             before_tss      sense  \n",
       "4             before_tss      sense  \n",
       "...                  ...        ...  \n",
       "367485  three_prime_utrs  antisense  \n",
       "367486  three_prime_utrs  antisense  \n",
       "367487  three_prime_utrs  antisense  \n",
       "367488  three_prime_utrs  antisense  \n",
       "367489  three_prime_utrs  antisense  \n",
       "\n",
       "[367490 rows x 12 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"cpg_count\"] = df[\"num_total_c_cpg_count\"]\n",
    "use_cols = [c for c in stitched_regions.columns if c != \"total_c\"]\n",
    "df=df[use_cols].reset_index(drop=True)\n",
    "df = pd.concat([df, stitched_regions]).reset_index(drop=True)\n",
    "df.drop('total_c', axis=1, inplace=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>Region</th>\n",
       "      <th>Sense</th>\n",
       "      <th>range_length</th>\n",
       "      <th>cpg_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>353768</th>\n",
       "      <td>+</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288577</th>\n",
       "      <td>-</td>\n",
       "      <td>three_prime_utrs</td>\n",
       "      <td>sense</td>\n",
       "      <td>2054.0</td>\n",
       "      <td>32.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321848</th>\n",
       "      <td>+</td>\n",
       "      <td>introns</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256669</th>\n",
       "      <td>-</td>\n",
       "      <td>introns</td>\n",
       "      <td>sense</td>\n",
       "      <td>13562.0</td>\n",
       "      <td>144.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128784</th>\n",
       "      <td>-</td>\n",
       "      <td>genes</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187828</th>\n",
       "      <td>+</td>\n",
       "      <td>genes</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10215</th>\n",
       "      <td>+</td>\n",
       "      <td>genes</td>\n",
       "      <td>sense</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>434.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69274</th>\n",
       "      <td>-</td>\n",
       "      <td>genes</td>\n",
       "      <td>sense</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>430.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274838</th>\n",
       "      <td>-</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "      <td>sense</td>\n",
       "      <td>140.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340025</th>\n",
       "      <td>+</td>\n",
       "      <td>five_prime_utrs</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69826</th>\n",
       "      <td>-</td>\n",
       "      <td>first_introns</td>\n",
       "      <td>sense</td>\n",
       "      <td>22051.0</td>\n",
       "      <td>208.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188379</th>\n",
       "      <td>+</td>\n",
       "      <td>first_introns</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69711</th>\n",
       "      <td>-</td>\n",
       "      <td>first_exons</td>\n",
       "      <td>sense</td>\n",
       "      <td>258.0</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188264</th>\n",
       "      <td>+</td>\n",
       "      <td>first_exons</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237118</th>\n",
       "      <td>-</td>\n",
       "      <td>exons</td>\n",
       "      <td>sense</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>51.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302300</th>\n",
       "      <td>+</td>\n",
       "      <td>exons</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186379</th>\n",
       "      <td>+</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67826</th>\n",
       "      <td>-</td>\n",
       "      <td>before_tss</td>\n",
       "      <td>sense</td>\n",
       "      <td>200.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186854</th>\n",
       "      <td>+</td>\n",
       "      <td>after_tes</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68300</th>\n",
       "      <td>-</td>\n",
       "      <td>after_tes</td>\n",
       "      <td>sense</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       strand            Region      Sense  range_length  cpg_count\n",
       "353768      +  three_prime_utrs  antisense           NaN        NaN\n",
       "288577      -  three_prime_utrs      sense        2054.0       32.0\n",
       "321848      +           introns  antisense           NaN        NaN\n",
       "256669      -           introns      sense       13562.0      144.0\n",
       "128784      -             genes  antisense           NaN        NaN\n",
       "187828      +             genes  antisense           NaN        NaN\n",
       "10215       +             genes      sense       38866.0      434.0\n",
       "69274       -             genes      sense       38866.0      430.0\n",
       "274838      -   five_prime_utrs      sense         140.0       15.0\n",
       "340025      +   five_prime_utrs  antisense           NaN        NaN\n",
       "69826       -     first_introns      sense       22051.0      208.0\n",
       "188379      +     first_introns  antisense           NaN        NaN\n",
       "69711       -       first_exons      sense         258.0       27.0\n",
       "188264      +       first_exons  antisense           NaN        NaN\n",
       "237118      -             exons      sense        2995.0       51.0\n",
       "302300      +             exons  antisense           NaN        NaN\n",
       "186379      +        before_tss  antisense           NaN        NaN\n",
       "67826       -        before_tss      sense         200.0       15.0\n",
       "186854      +         after_tes  antisense           NaN        NaN\n",
       "68300       -         after_tes      sense        1000.0       13.0"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[\"Gene_id\"]==\"ENSMUSG00000000001.4\"][[\"strand\", \"Region\", \"Sense\", \"range_length\", \"cpg_count\"]].sort_values(by=\"Region\", ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "pivot table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pivot = df.pivot(\n",
    "    index=[\"Gene_id\", \"Gene_name\", \"contig\", \"strand\"],\n",
    "    # index = [\"Transcript_id\", \"contig\", \"strand\"],\n",
    "    columns=[\"Region\", \"Sense\"],\n",
    "    values=[\"mean_mc\", \"mean_hmc\", \"mean_modc\", \"cpg_count\",  \"range_length\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th colspan=\"10\" halign=\"left\">mean_mc</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">range_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>before_tss</th>\n",
       "      <th>after_tes</th>\n",
       "      <th>genes</th>\n",
       "      <th>first_exons</th>\n",
       "      <th>first_introns</th>\n",
       "      <th>before_tss</th>\n",
       "      <th>after_tes</th>\n",
       "      <th>genes</th>\n",
       "      <th>first_exons</th>\n",
       "      <th>first_introns</th>\n",
       "      <th>...</th>\n",
       "      <th>first_exons</th>\n",
       "      <th>first_introns</th>\n",
       "      <th>exons</th>\n",
       "      <th>introns</th>\n",
       "      <th>five_prime_utrs</th>\n",
       "      <th>three_prime_utrs</th>\n",
       "      <th>exons</th>\n",
       "      <th>introns</th>\n",
       "      <th>five_prime_utrs</th>\n",
       "      <th>three_prime_utrs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>sense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "      <th>antisense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000000001.4</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Gnai3</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">3</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.751773</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.625551</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.744475</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>13562.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>2054.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000000003.15</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Pbsn</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">X</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.718121</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>0.734848</td>\n",
       "      <td>0.767077</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.675325</td>\n",
       "      <td>0.749183</td>\n",
       "      <td>0.719298</td>\n",
       "      <td>0.751803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.749183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>681.0</td>\n",
       "      <td>9532.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>235.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.15</th>\n",
       "      <th>Cdc45</th>\n",
       "      <th>16</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.651237</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.775061</td>\n",
       "      <td>0.651237</td>\n",
       "      <td>0.002212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118640.1</th>\n",
       "      <th>AC167036.2</th>\n",
       "      <th>1</th>\n",
       "      <th>-</th>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.705426</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000118646.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">AC160405.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">10</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>0.810799</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>0.829988</td>\n",
       "      <td>0.758004</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.790244</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.756352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>39.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000118653.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">AC159819.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">9</th>\n",
       "      <th>+</th>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.667582</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.720052</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42046 rows × 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  mean_mc                      \\\n",
       "Region                                         before_tss after_tes     genes   \n",
       "Sense                                               sense     sense     sense   \n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      +             NaN       NaN  0.672555   \n",
       "                                        -        0.000000  0.744475  0.670919   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +             NaN       NaN  0.750896   \n",
       "                                        -             NaN  0.675325  0.749183   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +             NaN       NaN  0.651237   \n",
       "...                                                   ...       ...       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -        0.853868  0.705426  0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +             NaN       NaN  0.762000   \n",
       "                                        -        0.833333  0.790244  0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +        0.276243  0.667582  0.284238   \n",
       "                                        -             NaN       NaN  0.300631   \n",
       "\n",
       "                                                                          \\\n",
       "Region                                         first_exons first_introns   \n",
       "Sense                                                sense         sense   \n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +              NaN           NaN   \n",
       "                                        -         0.000455      0.624199   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +              NaN           NaN   \n",
       "                                        -         0.719298      0.751803   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +              NaN           NaN   \n",
       "...                                                    ...           ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -         0.779018           NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +              NaN           NaN   \n",
       "                                        -         0.777778      0.756352   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +         0.284238           NaN   \n",
       "                                        -              NaN           NaN   \n",
       "\n",
       "                                                                               \\\n",
       "Region                                         before_tss after_tes     genes   \n",
       "Sense                                           antisense antisense antisense   \n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      +        0.001415  0.751773  0.672555   \n",
       "                                        -             NaN       NaN  0.670919   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +             NaN  0.718121  0.750896   \n",
       "                                        -             NaN       NaN  0.749183   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +        0.000000  0.775061  0.651237   \n",
       "...                                                   ...       ...       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -             NaN       NaN  0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +        0.860000  0.810799  0.762000   \n",
       "                                        -             NaN       NaN  0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +             NaN       NaN  0.284238   \n",
       "                                        -        0.315789  0.720052  0.300631   \n",
       "\n",
       "                                                                          ...  \\\n",
       "Region                                         first_exons first_introns  ...   \n",
       "Sense                                            antisense     antisense  ...   \n",
       "Gene_id               Gene_name  contig strand                            ...   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +         0.000000      0.625551  ...   \n",
       "                                        -              NaN           NaN  ...   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +         0.734848      0.767077  ...   \n",
       "                                        -              NaN           NaN  ...   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +         0.002212           NaN  ...   \n",
       "...                                                    ...           ...  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -              NaN           NaN  ...   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +         0.829988      0.758004  ...   \n",
       "                                        -              NaN           NaN  ...   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +              NaN           NaN  ...   \n",
       "                                        -         0.300631           NaN  ...   \n",
       "\n",
       "                                               range_length                \\\n",
       "Region                                          first_exons first_introns   \n",
       "Sense                                             antisense     antisense   \n",
       "Gene_id               Gene_name  contig strand                              \n",
       "ENSMUSG00000000001.4  Gnai3      3      +               NaN           NaN   \n",
       "                                        -               NaN           NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +               NaN           NaN   \n",
       "                                        -               NaN           NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +               NaN           NaN   \n",
       "...                                                     ...           ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -               NaN           NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +               NaN           NaN   \n",
       "                                        -               NaN           NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +               NaN           NaN   \n",
       "                                        -               NaN           NaN   \n",
       "\n",
       "                                                                 \\\n",
       "Region                                           exons  introns   \n",
       "Sense                                            sense    sense   \n",
       "Gene_id               Gene_name  contig strand                    \n",
       "ENSMUSG00000000001.4  Gnai3      3      +          NaN      NaN   \n",
       "                                        -       2995.0  13562.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +          NaN      NaN   \n",
       "                                        -        681.0   9532.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +          NaN      NaN   \n",
       "...                                                ...      ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -          NaN      NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +          NaN      NaN   \n",
       "                                        -         39.0      NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +          NaN      NaN   \n",
       "                                        -          NaN      NaN   \n",
       "\n",
       "                                                                \\\n",
       "Region                                         five_prime_utrs   \n",
       "Sense                                                    sense   \n",
       "Gene_id               Gene_name  contig strand                   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                  NaN   \n",
       "                                        -                140.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                  NaN   \n",
       "                                        -                139.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                  NaN   \n",
       "...                                                        ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                  NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                  NaN   \n",
       "                                        -                  NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                  NaN   \n",
       "                                        -                  NaN   \n",
       "\n",
       "                                                                           \\\n",
       "Region                                         three_prime_utrs     exons   \n",
       "Sense                                                     sense antisense   \n",
       "Gene_id               Gene_name  contig strand                              \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                   NaN       NaN   \n",
       "                                        -                2054.0       NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                   NaN       NaN   \n",
       "                                        -                 235.0       NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                   NaN       NaN   \n",
       "...                                                         ...       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                   NaN       NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                   NaN       NaN   \n",
       "                                        -                   NaN       NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                   NaN       NaN   \n",
       "                                        -                   NaN       NaN   \n",
       "\n",
       "                                                                          \\\n",
       "Region                                           introns five_prime_utrs   \n",
       "Sense                                          antisense       antisense   \n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +            NaN             NaN   \n",
       "                                        -            NaN             NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +            NaN             NaN   \n",
       "                                        -            NaN             NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +            NaN             NaN   \n",
       "...                                                  ...             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -            NaN             NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +            NaN             NaN   \n",
       "                                        -            NaN             NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +            NaN             NaN   \n",
       "                                        -            NaN             NaN   \n",
       "\n",
       "                                                                 \n",
       "Region                                         three_prime_utrs  \n",
       "Sense                                                 antisense  \n",
       "Gene_id               Gene_name  contig strand                   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                   NaN  \n",
       "                                        -                   NaN  \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                   NaN  \n",
       "                                        -                   NaN  \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                   NaN  \n",
       "...                                                         ...  \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                   NaN  \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                   NaN  \n",
       "                                        -                   NaN  \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                   NaN  \n",
       "                                        -                   NaN  \n",
       "\n",
       "[42046 rows x 90 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pivot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>mean_mc_first_exons_sense</th>\n",
       "      <th>mean_mc_first_introns_sense</th>\n",
       "      <th>mean_mc_before_tss_antisense</th>\n",
       "      <th>mean_mc_after_tes_antisense</th>\n",
       "      <th>mean_mc_genes_antisense</th>\n",
       "      <th>mean_mc_first_exons_antisense</th>\n",
       "      <th>mean_mc_first_introns_antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>range_length_first_exons_antisense</th>\n",
       "      <th>range_length_first_introns_antisense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "      <th>range_length_exons_antisense</th>\n",
       "      <th>range_length_introns_antisense</th>\n",
       "      <th>range_length_five_prime_utrs_antisense</th>\n",
       "      <th>range_length_three_prime_utrs_antisense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000000001.4</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Gnai3</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">3</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.751773</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.625551</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.744475</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>13562.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>2054.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000000003.15</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Pbsn</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">X</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.718121</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>0.734848</td>\n",
       "      <td>0.767077</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.675325</td>\n",
       "      <td>0.749183</td>\n",
       "      <td>0.719298</td>\n",
       "      <td>0.751803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.749183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>681.0</td>\n",
       "      <td>9532.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>235.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.15</th>\n",
       "      <th>Cdc45</th>\n",
       "      <th>16</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.651237</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.775061</td>\n",
       "      <td>0.651237</td>\n",
       "      <td>0.002212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118640.1</th>\n",
       "      <th>AC167036.2</th>\n",
       "      <th>1</th>\n",
       "      <th>-</th>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.705426</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000118646.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">AC160405.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">10</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>0.810799</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>0.829988</td>\n",
       "      <td>0.758004</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.790244</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.756352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>39.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000118653.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">AC159819.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">9</th>\n",
       "      <th>+</th>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.667582</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.720052</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42046 rows × 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                mean_mc_before_tss_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            NaN   \n",
       "                                        -                       0.000000   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            NaN   \n",
       "                                        -                            NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            NaN   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                       0.853868   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            NaN   \n",
       "                                        -                       0.833333   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                       0.276243   \n",
       "                                        -                            NaN   \n",
       "\n",
       "                                                mean_mc_after_tes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                           NaN   \n",
       "                                        -                      0.744475   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                           NaN   \n",
       "                                        -                      0.675325   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                           NaN   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.705426   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                           NaN   \n",
       "                                        -                      0.790244   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.667582   \n",
       "                                        -                           NaN   \n",
       "\n",
       "                                                mean_mc_genes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                        \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                  0.672555   \n",
       "                                        -                  0.670919   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                  0.750896   \n",
       "                                        -                  0.749183   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                  0.651237   \n",
       "...                                                             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                  0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                  0.762000   \n",
       "                                        -                  0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                  0.284238   \n",
       "                                        -                  0.300631   \n",
       "\n",
       "                                                mean_mc_first_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                              \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                             NaN   \n",
       "                                        -                        0.000455   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                             NaN   \n",
       "                                        -                        0.719298   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                             NaN   \n",
       "...                                                                   ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                        0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                             NaN   \n",
       "                                        -                        0.777778   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                        0.284238   \n",
       "                                        -                             NaN   \n",
       "\n",
       "                                                mean_mc_first_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                               NaN   \n",
       "                                        -                          0.624199   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                               NaN   \n",
       "                                        -                          0.751803   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                               NaN   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                               NaN   \n",
       "                                        -                          0.756352   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               NaN   \n",
       "                                        -                               NaN   \n",
       "\n",
       "                                                mean_mc_before_tss_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                           0.001415   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                NaN   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                           0.000000   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                           0.860000   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                NaN   \n",
       "                                        -                           0.315789   \n",
       "\n",
       "                                                mean_mc_after_tes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                          0.751773   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                          0.718121   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                          0.775061   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                          0.810799   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               NaN   \n",
       "                                        -                          0.720052   \n",
       "\n",
       "                                                mean_mc_genes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                      0.672555   \n",
       "                                        -                      0.670919   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                      0.750896   \n",
       "                                        -                      0.749183   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                      0.651237   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                      0.762000   \n",
       "                                        -                      0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.284238   \n",
       "                                        -                      0.300631   \n",
       "\n",
       "                                                mean_mc_first_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            0.000000   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            0.734848   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            0.002212   \n",
       "...                                                                       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                 NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            0.829988   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                 NaN   \n",
       "                                        -                            0.300631   \n",
       "\n",
       "                                                mean_mc_first_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                    \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                              0.625551   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                              0.767077   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                   NaN   \n",
       "...                                                                         ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                   NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                              0.758004   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                   NaN   \n",
       "                                        -                                   NaN   \n",
       "\n",
       "                                                ...  \\\n",
       "Gene_id               Gene_name  contig strand  ...   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +       ...   \n",
       "...                                             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -       ...   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +       ...   \n",
       "                                        -       ...   \n",
       "\n",
       "                                                range_length_first_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                      NaN   \n",
       "...                                                                            ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                      NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "\n",
       "                                                range_length_first_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                         \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                        NaN   \n",
       "                                        -                                        NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                        NaN   \n",
       "                                        -                                        NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                        NaN   \n",
       "...                                                                              ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                        NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                        NaN   \n",
       "                                        -                                        NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                        NaN   \n",
       "                                        -                                        NaN   \n",
       "\n",
       "                                                range_length_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            NaN   \n",
       "                                        -                         2995.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            NaN   \n",
       "                                        -                          681.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            NaN   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                            NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            NaN   \n",
       "                                        -                           39.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                            NaN   \n",
       "                                        -                            NaN   \n",
       "\n",
       "                                                range_length_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                               \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                              NaN   \n",
       "                                        -                          13562.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                              NaN   \n",
       "                                        -                           9532.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                              NaN   \n",
       "...                                                                    ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                              NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                              NaN   \n",
       "                                        -                              NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                              NaN   \n",
       "                                        -                              NaN   \n",
       "\n",
       "                                                range_length_five_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                      NaN   \n",
       "                                        -                                    140.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                      NaN   \n",
       "                                        -                                    139.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                      NaN   \n",
       "...                                                                            ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                      NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "\n",
       "                                                range_length_three_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                        \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                       NaN   \n",
       "                                        -                                    2054.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                       NaN   \n",
       "                                        -                                     235.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                       NaN   \n",
       "...                                                                             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                       NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                       NaN   \n",
       "                                        -                                       NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                       NaN   \n",
       "                                        -                                       NaN   \n",
       "\n",
       "                                                range_length_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                NaN   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                NaN   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                NaN   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                NaN   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                NaN   \n",
       "                                        -                                NaN   \n",
       "\n",
       "                                                range_length_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                  NaN   \n",
       "                                        -                                  NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                  NaN   \n",
       "                                        -                                  NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                  NaN   \n",
       "...                                                                        ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                  NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                  NaN   \n",
       "                                        -                                  NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                  NaN   \n",
       "                                        -                                  NaN   \n",
       "\n",
       "                                                range_length_five_prime_utrs_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                           \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                          NaN   \n",
       "                                        -                                          NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                          NaN   \n",
       "                                        -                                          NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                          NaN   \n",
       "...                                                                                ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                          NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                          NaN   \n",
       "                                        -                                          NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                          NaN   \n",
       "                                        -                                          NaN   \n",
       "\n",
       "                                                range_length_three_prime_utrs_antisense  \n",
       "Gene_id               Gene_name  contig strand                                           \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                           NaN  \n",
       "                                        -                                           NaN  \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                           NaN  \n",
       "                                        -                                           NaN  \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                           NaN  \n",
       "...                                                                                 ...  \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                           NaN  \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                           NaN  \n",
       "                                        -                                           NaN  \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                           NaN  \n",
       "                                        -                                           NaN  \n",
       "\n",
       "[42046 rows x 90 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compost feature names\n",
    "df_features = df_pivot.copy()\n",
    "df_features.columns = [\" \".join(col).strip() for col in df_features.columns.values]\n",
    "features = df_features.columns\n",
    "features = [f.replace(\" \", \"_\") for f in features]\n",
    "df_features.columns = features\n",
    "df_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>mean_mc_first_exons_sense</th>\n",
       "      <th>mean_mc_first_introns_sense</th>\n",
       "      <th>mean_mc_before_tss_antisense</th>\n",
       "      <th>mean_mc_after_tes_antisense</th>\n",
       "      <th>mean_mc_genes_antisense</th>\n",
       "      <th>mean_mc_first_exons_antisense</th>\n",
       "      <th>mean_mc_first_introns_antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>cpg_count_three_prime_utrs_sense</th>\n",
       "      <th>range_length_before_tss_sense</th>\n",
       "      <th>range_length_after_tes_sense</th>\n",
       "      <th>range_length_genes_sense</th>\n",
       "      <th>range_length_first_exons_sense</th>\n",
       "      <th>range_length_first_introns_sense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000000001.4</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Gnai3</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">3</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.751773</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.625551</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.744475</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>32.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>22051.0</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>13562.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>2054.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000000003.15</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Pbsn</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">X</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.718121</td>\n",
       "      <td>0.750896</td>\n",
       "      <td>0.734848</td>\n",
       "      <td>0.767077</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15722.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.675325</td>\n",
       "      <td>0.749183</td>\n",
       "      <td>0.719298</td>\n",
       "      <td>0.751803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.749183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>15722.0</td>\n",
       "      <td>214.0</td>\n",
       "      <td>5295.0</td>\n",
       "      <td>681.0</td>\n",
       "      <td>9532.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>235.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.15</th>\n",
       "      <th>Cdc45</th>\n",
       "      <th>16</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.651237</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.775061</td>\n",
       "      <td>0.651237</td>\n",
       "      <td>0.002212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31540.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118640.1</th>\n",
       "      <th>AC167036.2</th>\n",
       "      <th>1</th>\n",
       "      <th>-</th>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.705426</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>890.0</td>\n",
       "      <td>890.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000118646.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">AC160405.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">10</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.860000</td>\n",
       "      <td>0.810799</td>\n",
       "      <td>0.762000</td>\n",
       "      <td>0.829988</td>\n",
       "      <td>0.758004</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19585.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.790244</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.756352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>19585.0</td>\n",
       "      <td>342.0</td>\n",
       "      <td>19204.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ENSMUSG00000118653.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">AC159819.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">9</th>\n",
       "      <th>+</th>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.667582</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>987.0</td>\n",
       "      <td>987.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.720052</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>0.300631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>987.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42046 rows × 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                mean_mc_before_tss_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            NaN   \n",
       "                                        -                       0.000000   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            NaN   \n",
       "                                        -                            NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            NaN   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                       0.853868   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            NaN   \n",
       "                                        -                       0.833333   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                       0.276243   \n",
       "                                        -                            NaN   \n",
       "\n",
       "                                                mean_mc_after_tes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                           NaN   \n",
       "                                        -                      0.744475   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                           NaN   \n",
       "                                        -                      0.675325   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                           NaN   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.705426   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                           NaN   \n",
       "                                        -                      0.790244   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.667582   \n",
       "                                        -                           NaN   \n",
       "\n",
       "                                                mean_mc_genes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                        \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                  0.672555   \n",
       "                                        -                  0.670919   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                  0.750896   \n",
       "                                        -                  0.749183   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                  0.651237   \n",
       "...                                                             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                  0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                  0.762000   \n",
       "                                        -                  0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                  0.284238   \n",
       "                                        -                  0.300631   \n",
       "\n",
       "                                                mean_mc_first_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                              \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                             NaN   \n",
       "                                        -                        0.000455   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                             NaN   \n",
       "                                        -                        0.719298   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                             NaN   \n",
       "...                                                                   ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                        0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                             NaN   \n",
       "                                        -                        0.777778   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                        0.284238   \n",
       "                                        -                             NaN   \n",
       "\n",
       "                                                mean_mc_first_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                               NaN   \n",
       "                                        -                          0.624199   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                               NaN   \n",
       "                                        -                          0.751803   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                               NaN   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                               NaN   \n",
       "                                        -                          0.756352   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               NaN   \n",
       "                                        -                               NaN   \n",
       "\n",
       "                                                mean_mc_before_tss_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                           0.001415   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                NaN   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                           0.000000   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                           0.860000   \n",
       "                                        -                                NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                NaN   \n",
       "                                        -                           0.315789   \n",
       "\n",
       "                                                mean_mc_after_tes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                          0.751773   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                          0.718121   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                          0.775061   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                          0.810799   \n",
       "                                        -                               NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               NaN   \n",
       "                                        -                          0.720052   \n",
       "\n",
       "                                                mean_mc_genes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                      0.672555   \n",
       "                                        -                      0.670919   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                      0.750896   \n",
       "                                        -                      0.749183   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                      0.651237   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                      0.762000   \n",
       "                                        -                      0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.284238   \n",
       "                                        -                      0.300631   \n",
       "\n",
       "                                                mean_mc_first_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            0.000000   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            0.734848   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            0.002212   \n",
       "...                                                                       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                 NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            0.829988   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                 NaN   \n",
       "                                        -                            0.300631   \n",
       "\n",
       "                                                mean_mc_first_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                    \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                              0.625551   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                              0.767077   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                   NaN   \n",
       "...                                                                         ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                   NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                              0.758004   \n",
       "                                        -                                   NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                   NaN   \n",
       "                                        -                                   NaN   \n",
       "\n",
       "                                                ...  \\\n",
       "Gene_id               Gene_name  contig strand  ...   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +       ...   \n",
       "...                                             ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -       ...   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +       ...   \n",
       "                                        -       ...   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +       ...   \n",
       "                                        -       ...   \n",
       "\n",
       "                                                cpg_count_three_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                     \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                    NaN   \n",
       "                                        -                                   32.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                    NaN   \n",
       "                                        -                                    1.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                    NaN   \n",
       "...                                                                          ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                    NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                    NaN   \n",
       "                                        -                                    NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                    NaN   \n",
       "                                        -                                    NaN   \n",
       "\n",
       "                                                range_length_before_tss_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                 NaN   \n",
       "                                        -                               200.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                 NaN   \n",
       "                                        -                                 NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                 NaN   \n",
       "...                                                                       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               200.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                 NaN   \n",
       "                                        -                               200.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               200.0   \n",
       "                                        -                                 NaN   \n",
       "\n",
       "                                                range_length_after_tes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                NaN   \n",
       "                                        -                             1000.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                NaN   \n",
       "                                        -                             1000.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                NaN   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                             1000.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                NaN   \n",
       "                                        -                             1000.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                             1000.0   \n",
       "                                        -                                NaN   \n",
       "\n",
       "                                                range_length_genes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                        38866.0   \n",
       "                                        -                        38866.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                        15722.0   \n",
       "                                        -                        15722.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                        31540.0   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                          890.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                        19585.0   \n",
       "                                        -                        19585.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                          987.0   \n",
       "                                        -                          987.0   \n",
       "\n",
       "                                                range_length_first_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                   \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                  NaN   \n",
       "                                        -                                258.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                  NaN   \n",
       "                                        -                                214.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                  NaN   \n",
       "...                                                                        ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                890.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                  NaN   \n",
       "                                        -                                342.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                987.0   \n",
       "                                        -                                  NaN   \n",
       "\n",
       "                                                range_length_first_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                     \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                    NaN   \n",
       "                                        -                                22051.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                    NaN   \n",
       "                                        -                                 5295.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                    NaN   \n",
       "...                                                                          ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                    NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                    NaN   \n",
       "                                        -                                19204.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                    NaN   \n",
       "                                        -                                    NaN   \n",
       "\n",
       "                                                range_length_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                            NaN   \n",
       "                                        -                         2995.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                            NaN   \n",
       "                                        -                          681.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                            NaN   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                            NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                            NaN   \n",
       "                                        -                           39.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                            NaN   \n",
       "                                        -                            NaN   \n",
       "\n",
       "                                                range_length_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                               \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                              NaN   \n",
       "                                        -                          13562.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                              NaN   \n",
       "                                        -                           9532.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                              NaN   \n",
       "...                                                                    ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                              NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                              NaN   \n",
       "                                        -                              NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                              NaN   \n",
       "                                        -                              NaN   \n",
       "\n",
       "                                                range_length_five_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                      NaN   \n",
       "                                        -                                    140.0   \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                      NaN   \n",
       "                                        -                                    139.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                      NaN   \n",
       "...                                                                            ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                      NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                      NaN   \n",
       "                                        -                                      NaN   \n",
       "\n",
       "                                                range_length_three_prime_utrs_sense  \n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      +                                       NaN  \n",
       "                                        -                                    2054.0  \n",
       "ENSMUSG00000000003.15 Pbsn       X      +                                       NaN  \n",
       "                                        -                                     235.0  \n",
       "ENSMUSG00000000028.15 Cdc45      16     +                                       NaN  \n",
       "...                                                                             ...  \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                       NaN  \n",
       "ENSMUSG00000118646.1  AC160405.1 10     +                                       NaN  \n",
       "                                        -                                       NaN  \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                       NaN  \n",
       "                                        -                                       NaN  \n",
       "\n",
       "[42046 rows x 72 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features_cleaned = df_features.dropna(axis=1, how='all')\n",
    "\n",
    "# Display the cleaned DataFrame without columns that have all NaN values\n",
    "df_features_cleaned\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>mean_mc_first_exons_sense</th>\n",
       "      <th>mean_mc_first_introns_sense</th>\n",
       "      <th>mean_mc_before_tss_antisense</th>\n",
       "      <th>mean_mc_after_tes_antisense</th>\n",
       "      <th>mean_mc_genes_antisense</th>\n",
       "      <th>mean_mc_first_exons_antisense</th>\n",
       "      <th>mean_mc_first_introns_antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>cpg_count_three_prime_utrs_sense</th>\n",
       "      <th>range_length_before_tss_sense</th>\n",
       "      <th>range_length_after_tes_sense</th>\n",
       "      <th>range_length_genes_sense</th>\n",
       "      <th>range_length_first_exons_sense</th>\n",
       "      <th>range_length_first_introns_sense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Gnai3</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">3</th>\n",
       "      <th>+</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.751773</td>\n",
       "      <td>0.672555</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.625551</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.744475</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>32.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>22051.0</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>13562.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>2054.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         mean_mc_before_tss_sense  mean_mc_after_tes_sense  \\\n",
       "Gene_name contig strand                                                      \n",
       "Gnai3     3      +                            NaN                      NaN   \n",
       "                 -                            0.0                 0.744475   \n",
       "\n",
       "                         mean_mc_genes_sense  mean_mc_first_exons_sense  \\\n",
       "Gene_name contig strand                                                   \n",
       "Gnai3     3      +                  0.672555                        NaN   \n",
       "                 -                  0.670919                   0.000455   \n",
       "\n",
       "                         mean_mc_first_introns_sense  \\\n",
       "Gene_name contig strand                                \n",
       "Gnai3     3      +                               NaN   \n",
       "                 -                          0.624199   \n",
       "\n",
       "                         mean_mc_before_tss_antisense  \\\n",
       "Gene_name contig strand                                 \n",
       "Gnai3     3      +                           0.001415   \n",
       "                 -                                NaN   \n",
       "\n",
       "                         mean_mc_after_tes_antisense  mean_mc_genes_antisense  \\\n",
       "Gene_name contig strand                                                         \n",
       "Gnai3     3      +                          0.751773                 0.672555   \n",
       "                 -                               NaN                 0.670919   \n",
       "\n",
       "                         mean_mc_first_exons_antisense  \\\n",
       "Gene_name contig strand                                  \n",
       "Gnai3     3      +                                 0.0   \n",
       "                 -                                 NaN   \n",
       "\n",
       "                         mean_mc_first_introns_antisense  ...  \\\n",
       "Gene_name contig strand                                   ...   \n",
       "Gnai3     3      +                              0.625551  ...   \n",
       "                 -                                   NaN  ...   \n",
       "\n",
       "                         cpg_count_three_prime_utrs_sense  \\\n",
       "Gene_name contig strand                                     \n",
       "Gnai3     3      +                                    NaN   \n",
       "                 -                                   32.0   \n",
       "\n",
       "                         range_length_before_tss_sense  \\\n",
       "Gene_name contig strand                                  \n",
       "Gnai3     3      +                                 NaN   \n",
       "                 -                               200.0   \n",
       "\n",
       "                         range_length_after_tes_sense  \\\n",
       "Gene_name contig strand                                 \n",
       "Gnai3     3      +                                NaN   \n",
       "                 -                             1000.0   \n",
       "\n",
       "                         range_length_genes_sense  \\\n",
       "Gene_name contig strand                             \n",
       "Gnai3     3      +                        38866.0   \n",
       "                 -                        38866.0   \n",
       "\n",
       "                         range_length_first_exons_sense  \\\n",
       "Gene_name contig strand                                   \n",
       "Gnai3     3      +                                  NaN   \n",
       "                 -                                258.0   \n",
       "\n",
       "                         range_length_first_introns_sense  \\\n",
       "Gene_name contig strand                                     \n",
       "Gnai3     3      +                                    NaN   \n",
       "                 -                                22051.0   \n",
       "\n",
       "                         range_length_exons_sense  range_length_introns_sense  \\\n",
       "Gene_name contig strand                                                         \n",
       "Gnai3     3      +                            NaN                         NaN   \n",
       "                 -                         2995.0                     13562.0   \n",
       "\n",
       "                         range_length_five_prime_utrs_sense  \\\n",
       "Gene_name contig strand                                       \n",
       "Gnai3     3      +                                      NaN   \n",
       "                 -                                    140.0   \n",
       "\n",
       "                         range_length_three_prime_utrs_sense  \n",
       "Gene_name contig strand                                       \n",
       "Gnai3     3      +                                       NaN  \n",
       "                 -                                    2054.0  \n",
       "\n",
       "[2 rows x 72 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features_cleaned.loc[\"ENSMUSG00000000001.4\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>mean_mc_first_exons_sense</th>\n",
       "      <th>mean_mc_first_introns_sense</th>\n",
       "      <th>mean_mc_before_tss_antisense</th>\n",
       "      <th>mean_mc_after_tes_antisense</th>\n",
       "      <th>mean_mc_genes_antisense</th>\n",
       "      <th>mean_mc_first_exons_antisense</th>\n",
       "      <th>mean_mc_first_introns_antisense</th>\n",
       "      <th>...</th>\n",
       "      <th>cpg_count_three_prime_utrs_sense</th>\n",
       "      <th>range_length_before_tss_sense</th>\n",
       "      <th>range_length_after_tes_sense</th>\n",
       "      <th>range_length_genes_sense</th>\n",
       "      <th>range_length_first_exons_sense</th>\n",
       "      <th>range_length_first_introns_sense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001.4</th>\n",
       "      <th>Gnai3</th>\n",
       "      <th>3</th>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.744475</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.624199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.670919</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>32.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>38866.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>22051.0</td>\n",
       "      <td>2995.0</td>\n",
       "      <td>13562.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>2054.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.15</th>\n",
       "      <th>Cdc45</th>\n",
       "      <th>16</th>\n",
       "      <th>-</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.786600</td>\n",
       "      <td>0.658988</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.658988</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>31540.0</td>\n",
       "      <td>168.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1955.0</td>\n",
       "      <td>29402.0</td>\n",
       "      <td>311.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037.17</th>\n",
       "      <th>Scml2</th>\n",
       "      <th>X</th>\n",
       "      <th>+</th>\n",
       "      <td>0.653846</td>\n",
       "      <td>0.697581</td>\n",
       "      <td>0.730139</td>\n",
       "      <td>0.761134</td>\n",
       "      <td>0.796358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.730139</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>175688.0</td>\n",
       "      <td>131.0</td>\n",
       "      <td>34668.0</td>\n",
       "      <td>2777.0</td>\n",
       "      <td>138112.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000049.11</th>\n",
       "      <th>Apoh</th>\n",
       "      <th>11</th>\n",
       "      <th>+</th>\n",
       "      <td>0.007270</td>\n",
       "      <td>0.813542</td>\n",
       "      <td>0.635118</td>\n",
       "      <td>0.819620</td>\n",
       "      <td>0.597780</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.635118</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>71042.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>51939.0</td>\n",
       "      <td>1068.0</td>\n",
       "      <td>17921.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000056.7</th>\n",
       "      <th>Narf</th>\n",
       "      <th>11</th>\n",
       "      <th>+</th>\n",
       "      <td>0.000937</td>\n",
       "      <td>0.736318</td>\n",
       "      <td>0.616608</td>\n",
       "      <td>0.001418</td>\n",
       "      <td>0.083899</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.616608</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>31.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>18603.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1068.0</td>\n",
       "      <td>4286.0</td>\n",
       "      <td>13151.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>2933.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118623.1</th>\n",
       "      <th>AL935121.1</th>\n",
       "      <th>2</th>\n",
       "      <th>+</th>\n",
       "      <td>0.512397</td>\n",
       "      <td>0.585841</td>\n",
       "      <td>0.619217</td>\n",
       "      <td>0.764444</td>\n",
       "      <td>0.608056</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.619217</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>48754.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>1102.0</td>\n",
       "      <td>2782.0</td>\n",
       "      <td>44783.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118638.1</th>\n",
       "      <th>AL805980.1</th>\n",
       "      <th>X</th>\n",
       "      <th>-</th>\n",
       "      <td>0.870370</td>\n",
       "      <td>0.853982</td>\n",
       "      <td>0.712744</td>\n",
       "      <td>0.793194</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.712744</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>3793.0</td>\n",
       "      <td>1432.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>580.0</td>\n",
       "      <td>1162.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118640.1</th>\n",
       "      <th>AC167036.2</th>\n",
       "      <th>1</th>\n",
       "      <th>-</th>\n",
       "      <td>0.853868</td>\n",
       "      <td>0.705426</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.779018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>890.0</td>\n",
       "      <td>890.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118646.1</th>\n",
       "      <th>AC160405.1</th>\n",
       "      <th>10</th>\n",
       "      <th>-</th>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.790244</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.756352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.757576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>19585.0</td>\n",
       "      <td>342.0</td>\n",
       "      <td>19204.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118653.1</th>\n",
       "      <th>AC159819.1</th>\n",
       "      <th>9</th>\n",
       "      <th>+</th>\n",
       "      <td>0.276243</td>\n",
       "      <td>0.667582</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.284238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>987.0</td>\n",
       "      <td>987.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18654 rows × 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                mean_mc_before_tss_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                       0.000000   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                       0.000000   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                       0.653846   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                       0.007270   \n",
       "ENSMUSG00000000056.7  Narf       11     +                       0.000937   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                       0.512397   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                       0.870370   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                       0.853868   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                       0.833333   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                       0.276243   \n",
       "\n",
       "                                                mean_mc_after_tes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                      0.744475   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                      0.786600   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                      0.697581   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                      0.813542   \n",
       "ENSMUSG00000000056.7  Narf       11     +                      0.736318   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                      0.585841   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                      0.853982   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.705426   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                      0.790244   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.667582   \n",
       "\n",
       "                                                mean_mc_genes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                        \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                  0.670919   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                  0.658988   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                  0.730139   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                  0.635118   \n",
       "ENSMUSG00000000056.7  Narf       11     +                  0.616608   \n",
       "...                                                             ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                  0.619217   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                  0.712744   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                  0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                  0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                  0.284238   \n",
       "\n",
       "                                                mean_mc_first_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                              \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                        0.000455   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                        0.000000   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                        0.761134   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                        0.819620   \n",
       "ENSMUSG00000000056.7  Narf       11     +                        0.001418   \n",
       "...                                                                   ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                        0.764444   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                        0.793194   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                        0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                        0.777778   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                        0.284238   \n",
       "\n",
       "                                                mean_mc_first_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                          0.624199   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                               NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                          0.796358   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                          0.597780   \n",
       "ENSMUSG00000000056.7  Narf       11     +                          0.083899   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                          0.608056   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                               NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                          0.756352   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               NaN   \n",
       "\n",
       "                                                mean_mc_before_tss_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                NaN   \n",
       "ENSMUSG00000000056.7  Narf       11     +                                NaN   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                NaN   \n",
       "\n",
       "                                                mean_mc_after_tes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                               NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                               NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                               NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                               NaN   \n",
       "ENSMUSG00000000056.7  Narf       11     +                               NaN   \n",
       "...                                                                     ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                               NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                               NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                               NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               NaN   \n",
       "\n",
       "                                                mean_mc_genes_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                            \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                      0.670919   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                      0.658988   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                      0.730139   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                      0.635118   \n",
       "ENSMUSG00000000056.7  Narf       11     +                      0.616608   \n",
       "...                                                                 ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                      0.619217   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                      0.712744   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                      0.779018   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                      0.757576   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                      0.284238   \n",
       "\n",
       "                                                mean_mc_first_exons_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                 NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                 NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                 NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                 NaN   \n",
       "ENSMUSG00000000056.7  Narf       11     +                                 NaN   \n",
       "...                                                                       ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                 NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                 NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                 NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                 NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                 NaN   \n",
       "\n",
       "                                                mean_mc_first_introns_antisense  \\\n",
       "Gene_id               Gene_name  contig strand                                    \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                   NaN   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                   NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                   NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                   NaN   \n",
       "ENSMUSG00000000056.7  Narf       11     +                                   NaN   \n",
       "...                                                                         ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                   NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                   NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                   NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                   NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                   NaN   \n",
       "\n",
       "                                                ...  \\\n",
       "Gene_id               Gene_name  contig strand  ...   \n",
       "ENSMUSG00000000001.4  Gnai3      3      -       ...   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -       ...   \n",
       "ENSMUSG00000000037.17 Scml2      X      +       ...   \n",
       "ENSMUSG00000000049.11 Apoh       11     +       ...   \n",
       "ENSMUSG00000000056.7  Narf       11     +       ...   \n",
       "...                                             ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +       ...   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -       ...   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -       ...   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -       ...   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +       ...   \n",
       "\n",
       "                                                cpg_count_three_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                     \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                   32.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                    NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                    NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                    2.0   \n",
       "ENSMUSG00000000056.7  Narf       11     +                                   31.0   \n",
       "...                                                                          ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                    NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                    NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                    NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                    NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                    NaN   \n",
       "\n",
       "                                                range_length_before_tss_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                  \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                               200.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                               200.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                               200.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                               200.0   \n",
       "ENSMUSG00000000056.7  Narf       11     +                               200.0   \n",
       "...                                                                       ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                               200.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                               200.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                               200.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                               200.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                               200.0   \n",
       "\n",
       "                                                range_length_after_tes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                 \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                             1000.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                             1000.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                             1000.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                             1000.0   \n",
       "ENSMUSG00000000056.7  Narf       11     +                             1000.0   \n",
       "...                                                                      ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                             1000.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                             1000.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                             1000.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                             1000.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                             1000.0   \n",
       "\n",
       "                                                range_length_genes_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                        38866.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                        31540.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                       175688.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                        71042.0   \n",
       "ENSMUSG00000000056.7  Narf       11     +                        18603.0   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                        48754.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                         3793.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                          890.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                        19585.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                          987.0   \n",
       "\n",
       "                                                range_length_first_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                   \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                258.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                168.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                131.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                114.0   \n",
       "ENSMUSG00000000056.7  Narf       11     +                                 98.0   \n",
       "...                                                                        ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                 87.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                               1432.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                890.0   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                342.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                987.0   \n",
       "\n",
       "                                                range_length_first_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                     \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                22051.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                    NaN   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                34668.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                51939.0   \n",
       "ENSMUSG00000000056.7  Narf       11     +                                 1068.0   \n",
       "...                                                                          ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                 1102.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                    NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                    NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                19204.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                    NaN   \n",
       "\n",
       "                                                range_length_exons_sense  \\\n",
       "Gene_id               Gene_name  contig strand                             \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                         2995.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                         1955.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                         2777.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                         1068.0   \n",
       "ENSMUSG00000000056.7  Narf       11     +                         4286.0   \n",
       "...                                                                  ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                         2782.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                          580.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                            NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                           39.0   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                            NaN   \n",
       "\n",
       "                                                range_length_introns_sense  \\\n",
       "Gene_id               Gene_name  contig strand                               \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                          13562.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                          29402.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                         138112.0   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                          17921.0   \n",
       "ENSMUSG00000000056.7  Narf       11     +                          13151.0   \n",
       "...                                                                    ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                          44783.0   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                           1162.0   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                              NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                              NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                              NaN   \n",
       "\n",
       "                                                range_length_five_prime_utrs_sense  \\\n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                    140.0   \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                    311.0   \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                      NaN   \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                     50.0   \n",
       "ENSMUSG00000000056.7  Narf       11     +                                     71.0   \n",
       "...                                                                            ...   \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                      NaN   \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                      NaN   \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                      NaN   \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                      NaN   \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                      NaN   \n",
       "\n",
       "                                                range_length_three_prime_utrs_sense  \n",
       "Gene_id               Gene_name  contig strand                                       \n",
       "ENSMUSG00000000001.4  Gnai3      3      -                                    2054.0  \n",
       "ENSMUSG00000000028.15 Cdc45      16     -                                       NaN  \n",
       "ENSMUSG00000000037.17 Scml2      X      +                                       NaN  \n",
       "ENSMUSG00000000049.11 Apoh       11     +                                     100.0  \n",
       "ENSMUSG00000000056.7  Narf       11     +                                    2933.0  \n",
       "...                                                                             ...  \n",
       "ENSMUSG00000118623.1  AL935121.1 2      +                                       NaN  \n",
       "ENSMUSG00000118638.1  AL805980.1 X      -                                       NaN  \n",
       "ENSMUSG00000118640.1  AC167036.2 1      -                                       NaN  \n",
       "ENSMUSG00000118646.1  AC160405.1 10     -                                       NaN  \n",
       "ENSMUSG00000118653.1  AC159819.1 9      +                                       NaN  \n",
       "\n",
       "[18654 rows x 72 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features_cleaned[df_features_cleaned['mean_hmc_before_tss_sense']==df_features_cleaned['mean_hmc_before_tss_sense']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### write to file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we write this feature file as a pickle file, which preserves the variable type, and is therefore prefered over a more basic text file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Gene_id', 'Gene_name', 'contig', 'strand', 'mean_mc_before_tss_sense',\n",
       "       'mean_mc_after_tes_sense', 'mean_mc_genes_sense',\n",
       "       'mean_mc_first_exons_sense', 'mean_mc_first_introns_sense',\n",
       "       'mean_mc_before_tss_antisense', 'mean_mc_after_tes_antisense',\n",
       "       'mean_mc_genes_antisense', 'mean_mc_first_exons_antisense',\n",
       "       'mean_mc_first_introns_antisense', 'mean_mc_exons_sense',\n",
       "       'mean_mc_introns_sense', 'mean_mc_five_prime_utrs_sense',\n",
       "       'mean_mc_three_prime_utrs_sense', 'mean_mc_exons_antisense',\n",
       "       'mean_mc_introns_antisense', 'mean_mc_five_prime_utrs_antisense',\n",
       "       'mean_mc_three_prime_utrs_antisense', 'mean_hmc_before_tss_sense',\n",
       "       'mean_hmc_after_tes_sense', 'mean_hmc_genes_sense',\n",
       "       'mean_hmc_first_exons_sense', 'mean_hmc_first_introns_sense',\n",
       "       'mean_hmc_before_tss_antisense', 'mean_hmc_after_tes_antisense',\n",
       "       'mean_hmc_genes_antisense', 'mean_hmc_first_exons_antisense',\n",
       "       'mean_hmc_first_introns_antisense', 'mean_hmc_exons_sense',\n",
       "       'mean_hmc_introns_sense', 'mean_hmc_five_prime_utrs_sense',\n",
       "       'mean_hmc_three_prime_utrs_sense', 'mean_hmc_exons_antisense',\n",
       "       'mean_hmc_introns_antisense', 'mean_hmc_five_prime_utrs_antisense',\n",
       "       'mean_hmc_three_prime_utrs_antisense', 'mean_modc_before_tss_sense',\n",
       "       'mean_modc_after_tes_sense', 'mean_modc_genes_sense',\n",
       "       'mean_modc_first_exons_sense', 'mean_modc_first_introns_sense',\n",
       "       'mean_modc_before_tss_antisense', 'mean_modc_after_tes_antisense',\n",
       "       'mean_modc_genes_antisense', 'mean_modc_first_exons_antisense',\n",
       "       'mean_modc_first_introns_antisense', 'mean_modc_exons_sense',\n",
       "       'mean_modc_introns_sense', 'mean_modc_five_prime_utrs_sense',\n",
       "       'mean_modc_three_prime_utrs_sense', 'mean_modc_exons_antisense',\n",
       "       'mean_modc_introns_antisense', 'mean_modc_five_prime_utrs_antisense',\n",
       "       'mean_modc_three_prime_utrs_antisense', 'cpg_count_before_tss_sense',\n",
       "       'cpg_count_after_tes_sense', 'cpg_count_genes_sense',\n",
       "       'cpg_count_first_exons_sense', 'cpg_count_first_introns_sense',\n",
       "       'cpg_count_exons_sense', 'cpg_count_introns_sense',\n",
       "       'cpg_count_five_prime_utrs_sense', 'cpg_count_three_prime_utrs_sense',\n",
       "       'range_length_before_tss_sense', 'range_length_after_tes_sense',\n",
       "       'range_length_genes_sense', 'range_length_first_exons_sense',\n",
       "       'range_length_first_introns_sense', 'range_length_exons_sense',\n",
       "       'range_length_introns_sense', 'range_length_five_prime_utrs_sense',\n",
       "       'range_length_three_prime_utrs_sense', 'selected_transcript'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features_cleaned = df_features_cleaned.reset_index()\n",
    "# add a column with a boolean depending on whether the transcript is is selected_transcripts or not\n",
    "df_features_cleaned[\"selected_transcript\"] = df_features_cleaned.Gene_id.isin(selected_transcripts.gene_id)\n",
    "df_features_cleaned.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Source', 'Type', 'Start', 'End', 'Score', 'Strand',\n",
       "       'Phase', 'Id', 'Gene_id', 'Gene_type', 'Gene_name', 'Level', 'Mgi_id',\n",
       "       'Havana_gene', 'Tag', 'Ranges_ID'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>mean_mc_before_tss_sense</th>\n",
       "      <th>mean_mc_after_tes_sense</th>\n",
       "      <th>mean_mc_genes_sense</th>\n",
       "      <th>...</th>\n",
       "      <th>range_length_before_tss_sense</th>\n",
       "      <th>range_length_after_tes_sense</th>\n",
       "      <th>range_length_genes_sense</th>\n",
       "      <th>range_length_first_exons_sense</th>\n",
       "      <th>range_length_first_introns_sense</th>\n",
       "      <th>range_length_exons_sense</th>\n",
       "      <th>range_length_introns_sense</th>\n",
       "      <th>range_length_five_prime_utrs_sense</th>\n",
       "      <th>range_length_three_prime_utrs_sense</th>\n",
       "      <th>selected_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>1</td>\n",
       "      <td>4807787</td>\n",
       "      <td>4848409</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>0.529368</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>40622.0</td>\n",
       "      <td>159.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>2339.0</td>\n",
       "      <td>38089.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>1722.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>1</td>\n",
       "      <td>4807787</td>\n",
       "      <td>4848409</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.538405</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40622.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4807891</td>\n",
       "      <td>4886769</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0.000689</td>\n",
       "      <td>0.686047</td>\n",
       "      <td>0.472292</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>78878.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>20602.0</td>\n",
       "      <td>725.0</td>\n",
       "      <td>57461.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>1</td>\n",
       "      <td>4807891</td>\n",
       "      <td>4886769</td>\n",
       "      <td>Gm37988</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.483284</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>78878.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>1</td>\n",
       "      <td>4857813</td>\n",
       "      <td>4897908</td>\n",
       "      <td>Tcea1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.800370</td>\n",
       "      <td>0.514710</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>40095.0</td>\n",
       "      <td>162.0</td>\n",
       "      <td>9494.0</td>\n",
       "      <td>2375.0</td>\n",
       "      <td>28064.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>1540.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42679</th>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>Y</td>\n",
       "      <td>84562571</td>\n",
       "      <td>84564906</td>\n",
       "      <td>Gm21394</td>\n",
       "      <td>Y</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.679487</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2335.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1114.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42680</th>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>Y</td>\n",
       "      <td>85528516</td>\n",
       "      <td>85530907</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>Y</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.550000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2391.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42681</th>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>Y</td>\n",
       "      <td>85528516</td>\n",
       "      <td>85530907</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>Y</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.781250</td>\n",
       "      <td>0.686275</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>2391.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1231.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42682</th>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>89713423</td>\n",
       "      <td>89745531</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>Y</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.721311</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32108.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42683</th>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>89713423</td>\n",
       "      <td>89745531</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>Y</td>\n",
       "      <td>-</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.673913</td>\n",
       "      <td>0.765957</td>\n",
       "      <td>...</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>32108.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1364.0</td>\n",
       "      <td>878.0</td>\n",
       "      <td>29823.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>203.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42684 rows × 80 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_id Chromosome     Start       End Gene_name contig  \\\n",
       "0      ENSMUSG00000025903.14          1   4807787   4848409    Lypla1      1   \n",
       "1      ENSMUSG00000025903.14          1   4807787   4848409    Lypla1      1   \n",
       "2       ENSMUSG00000104217.1          1   4807891   4886769   Gm37988      1   \n",
       "3       ENSMUSG00000104217.1          1   4807891   4886769   Gm37988      1   \n",
       "4      ENSMUSG00000033813.15          1   4857813   4897908     Tcea1      1   \n",
       "...                      ...        ...       ...       ...       ...    ...   \n",
       "42679   ENSMUSG00000094660.2          Y  84562571  84564906   Gm21394      Y   \n",
       "42680   ENSMUSG00000095650.2          Y  85528516  85530907   Gm20854      Y   \n",
       "42681   ENSMUSG00000095650.2          Y  85528516  85530907   Gm20854      Y   \n",
       "42682   ENSMUSG00000100608.1          Y  89713423  89745531   Gm21996      Y   \n",
       "42683   ENSMUSG00000100608.1          Y  89713423  89745531   Gm21996      Y   \n",
       "\n",
       "      strand  mean_mc_before_tss_sense  mean_mc_after_tes_sense  \\\n",
       "0          +                  0.000000                 0.731707   \n",
       "1          -                       NaN                      NaN   \n",
       "2          +                  0.000689                 0.686047   \n",
       "3          -                       NaN                      NaN   \n",
       "4          +                  0.001969                 0.800370   \n",
       "...      ...                       ...                      ...   \n",
       "42679      -                       NaN                      NaN   \n",
       "42680      +                       NaN                      NaN   \n",
       "42681      -                       NaN                 0.781250   \n",
       "42682      +                       NaN                      NaN   \n",
       "42683      -                  0.833333                 0.673913   \n",
       "\n",
       "       mean_mc_genes_sense  ...  range_length_before_tss_sense  \\\n",
       "0                 0.529368  ...                          200.0   \n",
       "1                 0.538405  ...                            NaN   \n",
       "2                 0.472292  ...                          200.0   \n",
       "3                 0.483284  ...                            NaN   \n",
       "4                 0.514710  ...                          200.0   \n",
       "...                    ...  ...                            ...   \n",
       "42679             0.679487  ...                            NaN   \n",
       "42680             0.550000  ...                            NaN   \n",
       "42681             0.686275  ...                            NaN   \n",
       "42682             0.721311  ...                            NaN   \n",
       "42683             0.765957  ...                          200.0   \n",
       "\n",
       "       range_length_after_tes_sense  range_length_genes_sense  \\\n",
       "0                            1000.0                   40622.0   \n",
       "1                               NaN                   40622.0   \n",
       "2                            1000.0                   78878.0   \n",
       "3                               NaN                   78878.0   \n",
       "4                            1000.0                   40095.0   \n",
       "...                             ...                       ...   \n",
       "42679                           NaN                    2335.0   \n",
       "42680                           NaN                    2391.0   \n",
       "42681                        1000.0                    2391.0   \n",
       "42682                           NaN                   32108.0   \n",
       "42683                        1000.0                   32108.0   \n",
       "\n",
       "       range_length_first_exons_sense  range_length_first_introns_sense  \\\n",
       "0                               159.0                              35.0   \n",
       "1                                 NaN                               NaN   \n",
       "2                                90.0                           20602.0   \n",
       "3                                 NaN                               NaN   \n",
       "4                               162.0                            9494.0   \n",
       "...                               ...                               ...   \n",
       "42679                             NaN                            1114.0   \n",
       "42680                             NaN                               NaN   \n",
       "42681                             NaN                            1231.0   \n",
       "42682                             NaN                               NaN   \n",
       "42683                             NaN                            1364.0   \n",
       "\n",
       "       range_length_exons_sense  range_length_introns_sense  \\\n",
       "0                        2339.0                     38089.0   \n",
       "1                           NaN                         NaN   \n",
       "2                         725.0                     57461.0   \n",
       "3                           NaN                         NaN   \n",
       "4                        2375.0                     28064.0   \n",
       "...                         ...                         ...   \n",
       "42679                       NaN                         NaN   \n",
       "42680                       NaN                         NaN   \n",
       "42681                       NaN                         NaN   \n",
       "42682                       NaN                         NaN   \n",
       "42683                     878.0                     29823.0   \n",
       "\n",
       "       range_length_five_prime_utrs_sense  \\\n",
       "0                                    90.0   \n",
       "1                                     NaN   \n",
       "2                                     NaN   \n",
       "3                                     NaN   \n",
       "4                                    99.0   \n",
       "...                                   ...   \n",
       "42679                                 NaN   \n",
       "42680                                 NaN   \n",
       "42681                                 NaN   \n",
       "42682                                 NaN   \n",
       "42683                                 NaN   \n",
       "\n",
       "       range_length_three_prime_utrs_sense  selected_transcript  \n",
       "0                                   1722.0                 True  \n",
       "1                                      NaN                 True  \n",
       "2                                      NaN                 True  \n",
       "3                                      NaN                 True  \n",
       "4                                   1540.0                 True  \n",
       "...                                    ...                  ...  \n",
       "42679                                  NaN                 True  \n",
       "42680                                  NaN                 True  \n",
       "42681                                  NaN                 True  \n",
       "42682                                  NaN                 True  \n",
       "42683                                203.0                 True  \n",
       "\n",
       "[42684 rows x 80 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features_cleaned = genes.df[[\"Gene_id\", \"Chromosome\", \"Start\", \"End\"]].merge(\n",
    "    df_features_cleaned, on=\"Gene_id\", how=\"left\",\n",
    "    )\n",
    "df_features_cleaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['range_length_before_tss_sense',\n",
       " 'range_length_after_tes_sense',\n",
       " 'range_length_genes_sense',\n",
       " 'range_length_first_exons_sense',\n",
       " 'range_length_first_introns_sense',\n",
       " 'range_length_before_tss_antisense',\n",
       " 'range_length_after_tes_antisense',\n",
       " 'range_length_genes_antisense',\n",
       " 'range_length_first_exons_antisense',\n",
       " 'range_length_first_introns_antisense',\n",
       " 'range_length_exons_sense',\n",
       " 'range_length_introns_sense',\n",
       " 'range_length_five_prime_utrs_sense',\n",
       " 'range_length_three_prime_utrs_sense',\n",
       " 'range_length_exons_antisense',\n",
       " 'range_length_introns_antisense',\n",
       " 'range_length_five_prime_utrs_antisense',\n",
       " 'range_length_three_prime_utrs_antisense']"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "length_col = [x for x in df_features.columns if x.startswith(\"range_length\")]\n",
    "length_col\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_features_cleaned.to_pickle(\"rerun_CEGXRun1485.pickle\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "modality",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
